

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Principal component analysis &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=c5ced968eda925caa686" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=c5ced968eda925caa686" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=c5ced968eda925caa686" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=c5ced968eda925caa686" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=c5ced968eda925caa686" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=c5ced968eda925caa686" />
  <script src="../../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=c5ced968eda925caa686"></script>

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/data_preprocessing/genotype/PCA';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Genomic relationship matrices" href="GRM.html" />
    <link rel="prev" title="Genotype PLINK file quality control" href="GWAS_QC.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../genotype_preprocessing.html">Genotype data preprocessing</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../covariate/covariate_hidden_factor.html">Hidden factor analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/qtl_association_testing.html">QTL association analysis</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">Fine-mapping with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc_dap.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/genotype/PCA.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/genotype/PCA.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/data_preprocessing/genotype/PCA.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Principal component analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#general-workflow">General workflow</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#method">Method</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-estimate-kinship-in-the-sample">Step 1: Estimate kinship in the sample</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-and-3-sample-selection-and-qc-the-genotype-data-for-pca">Step 2 and 3: Sample selection and QC the genotype data for PCA</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-pca-analysis-for-unrelated-samples">Step 4: PCA analysis for unrelated samples</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-projection-of-related-individuals">Step 5: Projection of related individuals</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-2-finalize-genotype-qc-by-pca-for-homogenous-population">Step 5-2: Finalize genotype QC by PCA for homogenous population</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-6-split-data-by-population">Step 6: Split data by population</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-7-for-each-population-do-step-2">Step 7: For each population do step 2</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-8-for-each-population-do-step-3-if-applicable">Step 8: For each population do step 3 if applicable</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-9-for-each-population-do-step-4">Step 9: For each population do step 4</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-10-for-each-population-do-step-5-if-applicable">Step 10: For each population do step 5 if applicable</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#final-step-for-each-population-do-their-own-qc-to-finalize">Final step: For each population do their own QC to finalize</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-analysis">PCA analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#project-related-individuals-back">Project related individuals back</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#plot-pca-results">Plot PCA results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#detect-outliers">Detect outliers</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#add-plot-and-outlier-detection-to-pca-steps">Add plot and outlier detection to PCA steps</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="principal-component-analysis">
<h1>Principal component analysis<a class="headerlink" href="#principal-component-analysis" title="Permalink to this heading">#</a></h1>
<p>The intention of this notebook is to perform the PCA analysis on genotype data and generate plots.</p>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">#</a></h2>
<p>Steps to generate a PCA include</p>
<ul class="simple">
<li><p>removing related individuals</p></li>
<li><p>pruning variants in linkage disequilibrium (LD)</p></li>
<li><p>perform PCA analysis on genotype of unrelated individuals</p></li>
<li><p>excluding outlier samples in the PCA space for individuals of homogeneous self-reported ancestry. These outliers may suggest poor genotyping quality or distant relatedness.</p></li>
</ul>
<p>Pitfalls</p>
<ol class="arabic simple">
<li><p>Some of the PCs may capture LD structure rather than population structure (decrease in power to detect associations in these regions of high LD)</p></li>
<li><p>When projecting a new study dataset to the PCA space computed from a reference dataset: projected PCs are shrunk toward 0 in the new dataset</p></li>
<li><p>PC scores may capture outliers that are due to family structure, population structure or other reasons; it might be beneficial to detect and remove these individuals to maximize the population structure captured by PCA (in the case of removing a few outliers) or to restrict analyses to genetically homogeneous samples</p></li>
</ol>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>Genotype in PLINK format: common variants, LD pruned.</p></li>
<li><p>Phenotype files in text format and must contain a column ‘IID’ for individual ID. Optionally it can contain a column named ‘FID’ for family ID. <strong>In this case samples will be matched by both FID and IID</strong>. If you do not have a phenotype file you can input a PLINK fam file. The phenotype file can additionally contain population information and possibly disease or other labelling information, to label the PCA plots.</p></li>
</ol>
<p>The inputs should be splitted into sets of related and unrelated individuals. Additionally you may want to prepare data per population. See “Minimal working example” section for more details.</p>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>PCA models (inside RDS file)</p></li>
<li><p>PCA scores (inside RDS file)</p></li>
<li><p>Mahalanobis distances and outliers to remove</p></li>
<li><p>Various plots</p></li>
</ol>
</section>
<section id="general-workflow">
<h2>General workflow<a class="headerlink" href="#general-workflow" title="Permalink to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>Estimate relatedness of the individuals in the sample by PLINK 2 that implements the KING algorithm</p></li>
<li><p>Select specific SNPs and samples using PLINK and remove related individuals</p></li>
<li><p>SNPs thining by doing LD-pruning</p></li>
</ol>
<p>The above steps are implemented in <code class="docutils literal notranslate"><span class="pre">GWAS_QC.ipynb</span></code> workflow.</p>
<ol class="arabic simple" start="4">
<li><p>Run PCA using only unrelated individuals for all populations, and examine the resulting plot</p></li>
<li><p>Project back related individuals, and generate a list of suggested samples to remove based on Mahalanobis distance test statistic per population. Default criteria is 0.997 percentile (two-sided) but we recommend checking the output plot before and after removal and rethink about it.</p></li>
</ol>
<p>The analysis above can be performed with reference data eg 1000 Genomes integrated, to help diagnose population substructure in data.</p>
<p>If you have subpopulations in the data, then additional steps should be applied for:</p>
<ol class="arabic simple" start="6">
<li><p>Split data into different populations, each population data should have both related vs unrelated individual data-sets</p></li>
<li><p>For each population, perform QC</p></li>
<li><p>For each population, re-calculate per population PC’s for unrelated individuals</p></li>
<li><p>For each population, project related samples back to the PC space</p></li>
<li><p>Remove outliers based on list previously generated</p></li>
</ol>
</section>
<section id="method">
<h2>Method<a class="headerlink" href="#method" title="Permalink to this heading">#</a></h2>
<p>Here is a quick recap of PCA analysis for those not immediately familiar with the method. PCA is a mathematical method to reduce dimensionality of the data while retaining most of the variation in the dataset.
This is accomplished by identifying directions or Principal Components (PC’s) that account for the maximum variation in the data.</p>
<p>One common approach to PCA is based on the singular-value decomposition of the the data matrix <span class="math notranslate nohighlight">\(X\)</span> (in our case the genotype matrix),</p>
<div class="math notranslate nohighlight">
\[X = U D V^T,\]</div>
<p>where <span class="math notranslate nohighlight">\(U\)</span> are the left eigenvectors, <span class="math notranslate nohighlight">\(D\)</span> is the diagonal matrix of singular values, and <span class="math notranslate nohighlight">\(V\)</span> are the right eigenvectors (also called loadings).</p>
<p>PCA can also be done using the eigen-decomposition of <span class="math notranslate nohighlight">\(X X^T\)</span>:</p>
<div class="math notranslate nohighlight">
\[X X^T = U S U^T,\]</div>
<p>where <span class="math notranslate nohighlight">\(S=D^2\)</span> is the diagonal matrix of eigenvalues.
<span class="math notranslate nohighlight">\(X\)</span> is usually centred (mean-subtracted) or standardised (mean subtracted, then divided by standard deviation) before PCA.</p>
<p>For PCA of SNP genotypes (at least in diploid organisms), the common standardisation is</p>
<div class="math notranslate nohighlight">
\[X_{ij}^{\prime} = \frac{X_{ij} - 2p_j}{\sqrt{2 p_j (1 - p_j)}},\]</div>
<p>where <span class="math notranslate nohighlight">\(X_{ij}\)</span> is the genotype (minor allele dosage <span class="math notranslate nohighlight">\(\{0, 1, 2\}\)</span>) for the <span class="math notranslate nohighlight">\(i\)</span>th individual and the <span class="math notranslate nohighlight">\(j\)</span>th SNP, and <span class="math notranslate nohighlight">\(p_j\)</span> is the minor allele frequency (MAF) for the <span class="math notranslate nohighlight">\(j\)</span>th SNP. In addition, the eigenvalues are scaled by the number of SNPs <span class="math notranslate nohighlight">\(m\)</span> (equivalent to performing the eigen-decomposition of <span class="math notranslate nohighlight">\(XX^T/m\)</span>).</p>
</section>
<section id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this heading">#</a></h2>
<p>The data can be found on <a class="reference external" href="https://drive.google.com/drive/u/0/folders/1_dmlPf_7-4xpBKqBmhqRhuHnOUKTAVJi">Google Drive</a>.</p>
<p>FIXME: chage this link to on synapse</p>
<p><strong>Note: parameter set for the MWE are meant to let the MWE work to show the workflow procedures. They may be unrealistic and should not be used in practice. The pipeline has reasonable default values for what we suggest to use in practice for most of the parameters.</strong></p>
<section id="step-1-estimate-kinship-in-the-sample">
<h3>Step 1: Estimate kinship in the sample<a class="headerlink" href="#step-1-estimate-kinship-in-the-sample" title="Permalink to this heading">#</a></h3>
<p>Aim: identify and output closely related individuals prior to PCA analysis.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">king</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">rename_chr22</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">kinship</span> <span class="mf">0.13</span> \
    <span class="o">--</span><span class="n">name</span> <span class="mi">20220110</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p><strong><code class="docutils literal notranslate"><span class="pre">--kinship</span> <span class="pre">0.13</span></code> is meant for use with this MWE. The default (without using <code class="docutils literal notranslate"><span class="pre">--kinship</span></code> parameter) is 0.0625 for 3rd degree of relatedness.</strong></p>
</section>
<section id="step-2-and-3-sample-selection-and-qc-the-genotype-data-for-pca">
<h3>Step 2 and 3: Sample selection and QC the genotype data for PCA<a class="headerlink" href="#step-2-and-3-sample-selection-and-qc-the-genotype-data-for-pca" title="Permalink to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>QC based on MAF, sample and variant missigness and Hardy-Weinberg Equilibrium. You can provide a list of samples to keep, or to remove. For example:</p>
<ul class="simple">
<li><p>Only extract data for one population</p></li>
<li><p>Only extract data for related individuals</p></li>
<li><p>Only extract data for unrelated individuals
In current context we would like to extract data for unrelated individuals and proceed with the rest of the QC steps.</p></li>
</ul>
</li>
<li><p>LD pruning. Prune SNPs in linkage dissequilibrium to make sure the PCA actually captures population structure and not LD structure (which could reduce the power of detecting genetic associations in these LD-regions).</p></li>
</ol>
<p><code class="docutils literal notranslate"><span class="pre">king</span></code> will split data into unrelated and related individuals. We will QC on unrelated individuals but <strong>extract the same variants from related individuals</strong>. We therefore only run <code class="docutils literal notranslate"><span class="pre">qc</span></code> with a list of variants extracted, and other filtering parameters set to 0.</p>
<p>Step 2: Variant level and sample level QC on unrelated individuals, in preparation for PCA analysis:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">maf</span><span class="o">-</span><span class="nb">filter</span> <span class="mf">0.01</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">for_pca</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>Step 3: Extract previously selected variants from related individuals in preparation for PCA, only applying missingness filter at sample level,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc_no_prune</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">variants</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">for_pca</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="ow">in</span> \
    <span class="o">--</span><span class="n">maf</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">geno</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">mind</span><span class="o">-</span><span class="nb">filter</span> <span class="mf">0.1</span> <span class="o">--</span><span class="n">hwe</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">for_pca</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
</section>
<section id="step-4-pca-analysis-for-unrelated-samples">
<h3>Step 4: PCA analysis for unrelated samples<a class="headerlink" href="#step-4-pca-analysis-for-unrelated-samples" title="Permalink to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">PCA</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">flashpca</span> \
        <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span> \
        <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">for_pca</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">bed</span> \
        <span class="o">--</span><span class="n">phenoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">txt</span> \
        <span class="o">--</span><span class="n">label</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
        <span class="o">--</span><span class="n">pop</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
        <span class="o">--</span><span class="n">maha</span><span class="o">-</span><span class="n">k</span> <span class="mi">2</span> \
        <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">flashpcaR</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>For starters we set <code class="docutils literal notranslate"><span class="pre">maha_k</span></code> abritarily to 2 for outlier calculations. You should look at the scree plot and determine how many PCs explain the most variation, then based on the scree plot generated to determine how many PCs you would like to consider for outlier checks.</p>
<p><strong>If you would like to skip outlier check and labelling, you can use <code class="docutils literal notranslate"><span class="pre">--prob</span> <span class="pre">1</span></code> to effectively not calling any samples an outlier</strong></p>
<p>If your <code class="docutils literal notranslate"><span class="pre">phenoFile</span></code> has a categorical variable with the ethnicity or population you can use <code class="docutils literal notranslate"><span class="pre">--label-col</span></code> parameter to define it, so that the plot generated has different colors depending on the population.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="o">~/</span><span class="n">tmp</span><span class="o">/</span><span class="mi">25</span><span class="o">-</span><span class="n">Jan</span><span class="o">-</span><span class="mi">2022</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">scree</span><span class="o">.</span><span class="n">png</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /home/gw/tmp/25-Jan-2022/output/pca/MWE_pheno.pca.scree.png (56.4 KiB):</div></div><img alt="../../../_images/e745848d22f44a28490ca73ce3daa8a451423d1ea5d6eed12fa2136f24bbb75a.png" src="../../../_images/e745848d22f44a28490ca73ce3daa8a451423d1ea5d6eed12fa2136f24bbb75a.png" />
</div>
</div>
</section>
<section id="step-5-projection-of-related-individuals">
<h3>Step 5: Projection of related individuals<a class="headerlink" href="#step-5-projection-of-related-individuals" title="Permalink to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">PCA</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">project_samples</span> \
  <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span> \
  <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">for_pca</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">extracted</span><span class="o">.</span><span class="n">bed</span> \
  <span class="o">--</span><span class="n">phenoFile</span> <span class="n">data</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">txt</span> \
  <span class="o">--</span><span class="n">pca</span><span class="o">-</span><span class="n">model</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">rds</span> \
  <span class="o">--</span><span class="n">label</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
  <span class="o">--</span><span class="n">pop</span><span class="o">-</span><span class="n">col</span> <span class="n">RACE</span> \
  <span class="o">--</span><span class="n">maha</span><span class="o">-</span><span class="n">k</span> <span class="mi">2</span> \
  <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">flashpcaR</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>Same as above, if you have an ethnicity variable you can specify it by using <code class="docutils literal notranslate"><span class="pre">label_col</span></code> and here you would also need to add <code class="docutils literal notranslate"><span class="pre">pop_col</span></code> that points to the same variable. This is necessary, otherwise the detect_outliers part of the workflow will error.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="o">~/</span><span class="n">tmp</span><span class="o">/</span><span class="mi">25</span><span class="o">-</span><span class="n">Jan</span><span class="o">-</span><span class="mi">2022</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">pc</span><span class="o">.</span><span class="n">png</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /home/gw/tmp/25-Jan-2022/output/pca/MWE_pheno.pca.projected.pc.png (570.8 KiB):</div></div><img alt="../../../_images/84d7dcd8b9fd53397c3d9b9e83491608cf7134c52a4df83ec67b49241283ce82.png" src="../../../_images/84d7dcd8b9fd53397c3d9b9e83491608cf7134c52a4df83ec67b49241283ce82.png" />
</div>
</div>
</section>
<section id="step-5-2-finalize-genotype-qc-by-pca-for-homogenous-population">
<h3>Step 5-2: Finalize genotype QC by PCA for homogenous population<a class="headerlink" href="#step-5-2-finalize-genotype-qc-by-pca-for-homogenous-population" title="Permalink to this heading">#</a></h3>
<p>Now pretend you analyze a homogenous population and you only did PCA to remove PC outliers. You have decided the outliers to remove and you would like to finalize the genotype QC.</p>
<p>First, remove these outliers and perform genotype data QC, similar to our initial QC but without LD pruning:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc_no_prune</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">remove</span><span class="o">-</span><span class="n">samples</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">outliers</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">no_outlier</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>Also remove outliers and keep the same variant as unrelated individuals, in related individuals:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">GWAS_QC</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">qc_no_prune</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">remove</span><span class="o">-</span><span class="n">samples</span> <span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">outliers</span> \
    <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">variants</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">no_outlier</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">bim</span> \
    <span class="o">--</span><span class="n">maf</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">geno</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> <span class="o">--</span><span class="n">mind</span><span class="o">-</span><span class="nb">filter</span> <span class="mf">0.1</span> <span class="o">--</span><span class="n">hwe</span><span class="o">-</span><span class="nb">filter</span> <span class="mi">0</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">no_outlier</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>Finally, merge back related and unrelated individuals as finalized genotype in PLINK format,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">genotype_formatting</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">merge_plink</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">unrelated</span><span class="o">.</span><span class="n">no_outlier</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">bed</span> \
               <span class="n">output</span><span class="o">/</span><span class="n">rename_chr22</span><span class="mf">.20220110</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">no_outlier</span><span class="o">.</span><span class="n">filtered</span><span class="o">.</span><span class="n">extracted</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">genotype_final</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">chr22_20220110_qced</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">container</span><span class="o">/</span><span class="n">bioinfo</span><span class="o">.</span><span class="n">sif</span>
</pre></div>
</div>
<p>You can stop here if you analyze a homogenous population.</p>
</section>
<section id="step-6-split-data-by-population">
<h3>Step 6: Split data by population<a class="headerlink" href="#step-6-split-data-by-population" title="Permalink to this heading">#</a></h3>
<p>If your data contains multiple populations, then for each population, you can still use the <code class="docutils literal notranslate"><span class="pre">GWAS_QC.ipynb</span></code> and <code class="docutils literal notranslate"><span class="pre">PCA.ipynb</span></code> workflows, except that at this step you have to split data by population using manually created files to include or to remove samples.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">pheno</span> <span class="o">=</span> <span class="n">read</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="s2">&quot;data/MWE_pheno.txt&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="n">TRUE</span><span class="p">,</span> <span class="n">stringsAsFactors</span><span class="o">=</span><span class="n">F</span><span class="p">)</span>
<span class="k">for</span> <span class="p">(</span><span class="n">i</span> <span class="ow">in</span> <span class="mi">1</span><span class="p">:</span><span class="mi">3</span><span class="p">){</span>
 <span class="n">race</span> <span class="o">=</span> <span class="n">subset</span><span class="p">(</span><span class="n">pheno</span><span class="p">,</span> <span class="n">RACE</span> <span class="o">==</span> <span class="n">i</span><span class="p">)</span>
 <span class="n">race_id</span> <span class="o">=</span> <span class="n">cbind</span><span class="p">(</span><span class="n">race</span><span class="p">[,</span><span class="mi">2</span><span class="p">],</span><span class="n">race</span><span class="p">[,</span> <span class="mi">2</span><span class="p">])</span>
 <span class="n">write</span><span class="o">.</span><span class="n">table</span><span class="p">(</span><span class="n">race_id</span><span class="p">,</span> <span class="n">paste0</span><span class="p">(</span><span class="s2">&quot;output/ID.&quot;</span><span class="p">,</span> <span class="s2">&quot;race&quot;</span><span class="p">,</span> <span class="n">i</span><span class="p">),</span> <span class="n">quote</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">col</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">,</span> <span class="n">row</span><span class="o">.</span><span class="n">names</span> <span class="o">=</span> <span class="n">FALSE</span><span class="p">)</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
<p>Below we perform analysis on each population. Here we only show it for Populations 1 and 3</p>
</section>
<section id="step-7-for-each-population-do-step-2">
<h3>Step 7: For each population do step 2<a class="headerlink" href="#step-7-for-each-population-do-step-2" title="Permalink to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>for i in race1 race3; do
    sos run GWAS_QC.ipynb qc \
        --cwd output \
        --genoFile output/rename_chr22.20220110.unrelated.bed \
        --keep-samples output/ID.$i \
        --name for_pca_$i \
        --container container/bioinfo.sif
done
</pre></div>
</div>
</section>
<section id="step-8-for-each-population-do-step-3-if-applicable">
<h3>Step 8: For each population do step 3 if applicable<a class="headerlink" href="#step-8-for-each-population-do-step-3-if-applicable" title="Permalink to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>for i in race1 race3; do
    sos run GWAS_QC.ipynb qc_no_prune \
        --cwd output \
        --genoFile output/rename_chr22.20220110.related.bed \
        --keep-variants output/rename_chr22.20220110.unrelated.for_pca_$i.filtered.prune.in \
        --keep-samples output/ID.$i \
        --maf-filter 0 --geno-filter 0 --mind-filter 0.1 --hwe-filter 0 \
        --name for_pca_$i \
        --container container/bioinfo.sif -s force
done
</pre></div>
</div>
<p>Here in fact you will run into an error message for population 1, which you can find inside <code class="docutils literal notranslate"><span class="pre">ename_chr22.20220110.related.for_pca_race1.filtered.extracted.stderr</span></code> that says: <code class="docutils literal notranslate"><span class="pre">Error:</span> <span class="pre">No</span> <span class="pre">people</span> <span class="pre">remaining</span> <span class="pre">after</span> <span class="pre">--keep.</span></code>. That means there is no samples from population 1 among the related individuals. You can safely ignore this, and only handle related individuals for population 3 in upcoming analysis.</p>
</section>
<section id="step-9-for-each-population-do-step-4">
<h3>Step 9: For each population do step 4<a class="headerlink" href="#step-9-for-each-population-do-step-4" title="Permalink to this heading">#</a></h3>
<p>Here I have to set the partial PCA to only computer 5 PCs because sample size is too small to work for the default 20 PCs.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>for i in race1 race3; do
    sos run PCA.ipynb flashpca \
        --name $i \
        --cwd output/pca \
        --genoFile output/rename_chr22.20220110.unrelated.for_pca_$i.filtered.prune.bed \
        --phenoFile data/MWE_pheno.txt \
        --label-col RACE \
        --pop-col RACE \
        --maha-k 2 \
        --k 5 \
        --container container/flashpcaR.sif
done
</pre></div>
</div>
</section>
<section id="step-10-for-each-population-do-step-5-if-applicable">
<h3>Step 10: For each population do step 5 if applicable<a class="headerlink" href="#step-10-for-each-population-do-step-5-if-applicable" title="Permalink to this heading">#</a></h3>
<p>Here we only do it for population 3 (there was no related individuals in population 1):</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>for i in race3; do
    sos run PCA.ipynb project_samples \
      --name $i \
      --cwd output/pca \
      --genoFile output/rename_chr22.20220110.related.for_pca_$i.filtered.extracted.bed \
      --phenoFile data/MWE_pheno.txt \
      --pca-model output/pca/MWE_pheno.$i.pca.rds \
      --label-col RACE \
      --pop-col RACE \
      --maha-k 2 \
      --k 5 \
      --container container/flashpcaR.sif
done
</pre></div>
</div>
<p>To visualize the results at this point,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="o">~/</span><span class="n">tmp</span><span class="o">/</span><span class="mi">19</span><span class="o">-</span><span class="n">Jan</span><span class="o">-</span><span class="mi">2022</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">pca</span><span class="o">/</span><span class="n">MWE_pheno</span><span class="o">.</span><span class="n">race3</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">pc</span><span class="o">.</span><span class="n">png</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /home/gw/tmp/19-Jan-2022/output/pca/MWE_pheno.race3.pca.projected.pc.png (124.0 KiB):</div></div><img alt="../../../_images/02871587aa834c5516433759e36ee30871615ec667f9b05186c3ca0cbcb015d5.png" src="../../../_images/02871587aa834c5516433759e36ee30871615ec667f9b05186c3ca0cbcb015d5.png" />
</div>
</div>
</section>
<section id="final-step-for-each-population-do-their-own-qc-to-finalize">
<h3>Final step: For each population do their own QC to finalize<a class="headerlink" href="#final-step-for-each-population-do-their-own-qc-to-finalize" title="Permalink to this heading">#</a></h3>
<p>Same as to Step 5-2, taking into consideration of outliers detected. Please read the <code class="docutils literal notranslate"><span class="pre">GWAS_QC.ipynb</span></code> document to see available QC options and recommendations.</p>
</section>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">PCA</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># the output directory for generated files</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># A string to identify your analysis run</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># Name of the population column in the phenoFile</span>
<span class="kn">parameter:</span> <span class="n">pop_col</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># Name of the populations (from the population column) you would like to plot and show on the PCA plot</span>
<span class="kn">parameter:</span> <span class="n">pops</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># Name of the color label column in the phenoFile; can be the same as population column. Can also be a separate column eg a &quot;super population&quot; column as a way to enable you to combine selected populations based on another column.</span>
<span class="kn">parameter:</span> <span class="n">label_col</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># Number of Principal Components to output,must be consistant between flashpca run and project samples run (flashpca partial PCA method).</span>
<span class="kn">parameter:</span> <span class="n">k</span> <span class="o">=</span> <span class="mi">20</span>
<span class="c1"># Number of Principal Components based on which outliers should be evaluated. Default is 5 but this should be based on examine the scree plot</span>
<span class="kn">parameter:</span> <span class="n">maha_k</span> <span class="o">=</span> <span class="mi">5</span>
<span class="c1"># Homogeneity of populations. Set to --homogeneous when true and --no-homogeneous when false</span>
<span class="kn">parameter:</span> <span class="n">homogeneous</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">10</span>
<span class="n">suffix</span> <span class="o">=</span> <span class="s1">&#39;_&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">pops</span><span class="p">)</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">pop_col</span><span class="p">:</span>
    <span class="n">homogeneous</span> <span class="o">=</span> <span class="kc">True</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># PCA command with PLINK, as a sanity check
[pca_plink]
# PLINK binary file
parameter: genoFile = path
input: genoFile
output: f&#39;{cwd}/{genoFile:bn}.pca.eigenvec&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: container = container, expand = &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    plink --bfile ${_input:n} --out ${_output:n} --pca ${k}
</pre></div>
</div>
</div>
</div>
</section>
<section id="pca-analysis">
<h2>PCA analysis<a class="headerlink" href="#pca-analysis" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Run PCA analysis using flashpca 
[flashpca_1]
# Plink binary file
parameter: genoFile = path
# The phenotypic file
parameter: phenoFile = path(f&#39;{genoFile}&#39;.replace(&quot;.bed&quot;,&quot;.fam&quot;))
# minimum population size to consider in the analysis
parameter: min_pop_size = 2
# How to standardize X before PCA
parameter: stand = &quot;binom2&quot;
## Input genoFile here is for unrelated samples
input: genoFile, phenoFile
output: f&#39;{cwd}/{phenoFile:bn}{(&quot;.&quot;+name) if name else &quot;&quot;}.{(suffix+&quot;.&quot;) if suffix != &quot;&quot; else &quot;&quot;}pca.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: container = container, expand = &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    # Load required libraries
    library(flashpcaR)
    library(dplyr)
    pops = c(${paths(pops):r,})
    f &lt;- flashpca(${_input[0]:nr}, ndim=${k}, stand=&quot;${stand}&quot;, do_loadings=TRUE, check_geno=TRUE)
    rownames(f$loadings) &lt;- read.table(&#39;${_input[0]:n}.bim&#39;,stringsAsFactors =F)[,2]
    # Use the projection file to generate pca plot
    pca &lt;- as.data.frame(f$projection)
    pca &lt;- tibble::rownames_to_column(pca, &quot;ID&quot;)
    colnames(pca) &lt;- c(&quot;ID&quot;,paste0(&quot;PC&quot;, 1:${k}))
  
    # Read fam file with phenotypes
    if(stringr::str_detect(${_input[1]:r},&quot;.fam$&quot;)){
        pheno &lt;- read.table(${_input[1]:r}, header=F,stringsAsFactors =F)
        colnames(pheno) = c(&quot;FID&quot;, &quot;IID&quot;, &quot;MID&quot;, &quot;PID&quot;, &quot;SEX&quot;, &quot;STATUS&quot;)
    } else {
        pheno &lt;- read.table(${_input[1]:r}, header=T,stringsAsFactors =F)
        if(&quot;IID&quot; %in% colnames(pheno) == FALSE) stop(&quot;No IID column in the phenoFile. Please rename the header of the phenoFile&quot;)
        if(&quot;FID&quot; %in% colnames(pheno) == FALSE) pheno$FID = pheno$IID
    }
    # Make the unique ID by merge FID and IID
    pheno$ID = paste(pheno$FID,pheno$IID,sep = &quot;:&quot;)
    #check duplicated ID
    if(length(unique(pheno$ID))!=length(pheno$ID)) stop(&quot;There are duplicated names in IID column of phenoFile&quot;)

    if (length(pops)&gt;0) pheno &lt;- pheno %&gt;%filter(${pop_col if pop_col else  &quot;pop&quot;} %in% pops | ${label_col if label_col else  &quot;pop&quot;} %in% pops)
    pca &lt;-merge(pheno, pca,by =&quot;ID&quot;, all=FALSE) 
    #
    if (${&quot;TRUE&quot; if pop_col else &quot;FALSE&quot;}) {
        # remove populations have less than ${min_pop_size} samples
        pop&lt;-names(table(pca$${pop_col if pop_col else &quot;pop&quot;}))
        pop_filter&lt;-pop[table(pca$pop)&lt;${min_pop_size}] # pop to be removed
        if (length(pop_filter)&gt;0) {
            warning(for (i in pop_filter){cat(i,&#39;;&#39;)},&#39;these &#39;, length(pop_filter),&quot; population will be removed due to having less than ${min_pop_size} samples in data.&quot;)
            # remove
            pca&lt;-pca%&gt;% filter(${f&#39;!{pop_col}%in%pop_filter&#39; if pop_col else pop_col})
        }
    } else {
      pca$pop &lt;- 1 
    }

    # Write the PC scores to a file
    write.table(pca,&quot;${_output:n}.txt&quot;, sep=&quot;\t&quot;, quote=FALSE, row.names=FALSE, col.names=TRUE)
    dat = list(pca_model = f, pc_scores = pca, meta = &quot;${_input[1]:bn} ${suffix}&quot;)
    # compute centroids before projecting back the samples
    # (calculate mean/median/cov per pop)
    if(${&quot;FALSE&quot; if homogeneous else &quot;TRUE&quot;}){
        pop_group &lt;- split(dat$pc_scores[ ,c(paste0(&quot;PC&quot;, 1:${maha_k}))], list(Group = dat$pc_scores$${pop_col if pop_col else &quot;pop&quot;}))
        dat$pc_cov &lt;- lapply(pop_group, function(x) cov(x))
        dat$pc_mean &lt;- lapply(pop_group, function(x) sapply(x, mean))
        dat$pc_median &lt;- lapply(pop_group, function(x) sapply(x, median))
    } else {
        dat$pc_cov &lt;- cov(f$projection[,1:${maha_k}])
        dat$pc_mean &lt;- apply(f$projection[,1:${maha_k}], 2, mean)
        dat$pc_median &lt;- apply(f$projection[,1:${maha_k}], 2, median)
    }
  
    # save results
    saveRDS(dat, ${_output:r})
  
bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container, entrypoint=entrypoint
        stdout=$[_output:n].stdout
        for i in $[_output] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;This rds file is a list containing the pca for unrelated sample&quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        done
</pre></div>
</div>
</div>
</div>
</section>
<section id="project-related-individuals-back">
<h2>Project related individuals back<a class="headerlink" href="#project-related-individuals-back" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Project back to PCA model additional samples
[project_samples_1]
# Plink binary file
parameter: genoFile = path
# The phenotypic file
parameter: phenoFile = path(f&#39;{genoFile}&#39;.replace(&quot;.bed&quot;,&quot;.fam&quot;))
parameter: pca_model = f&#39;{cwd}/{phenoFile:bn}{(&quot;.&quot;+name) if name else &quot;&quot;}.{(suffix+&quot;.&quot;) if suffix != &quot;&quot; else &quot;&quot;}pca.rds&#39;
## Input genoFile here is for related samples
input: genoFile, phenoFile, pca_model
output: f&#39;{cwd}/{phenoFile:bn}{(&quot;.&quot;+name) if name else &quot;&quot;}.pca.projected.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    # Load required libraries
    library(dplyr)
    library(flashpcaR)
    # Read the PLINK binary files
    frel &lt;- ${_input[0]:nr}
    # Read loadings, center and scale from previous PCA
    dat &lt;- readRDS(${_input[2]:r})
    f &lt;- dat$pca_model                           
    # Read the bim file to obtain reference alleles
    bim &lt;- read.table(&#39;${_input[0]:n}.bim&#39;,stringsAsFactors =F)
    ref &lt;- as.character(bim[,5])
    names(ref) &lt;- bim[,2]
    if (nrow(f$loadings) != length(ref)) stop(&quot;Input number of variants should be the same as used in the previous PCA model&quot;)
    overlapped_variants &lt;- match(names(ref), rownames(f$loadings))
    # Project related samples
    fpro &lt;- project(frel, loadings=f$loadings[overlapped_variants,], orig_mean=f$center[overlapped_variants], orig_sd=f$scale[overlapped_variants], ref_allele=ref)
    pca &lt;- fpro$projection
    k = ${k}
    pca &lt;- as.data.frame(fpro$projection)
    pca &lt;- tibble::rownames_to_column(pca, &quot;ID&quot;)
    colnames(pca) &lt;- c(&quot;ID&quot;,paste0(&quot;PC&quot;, 1:k))
    
    # Read fam file with phenotypes
    if(stringr::str_detect(${_input[1]:r},&quot;.fam$&quot;)){
        pheno &lt;- read.table(${_input[1]:r}, header=F,stringsAsFactors =F)
        colnames(pheno) = c(&quot;FID&quot;, &quot;IID&quot;, &quot;MID&quot;, &quot;PID&quot;, &quot;SEX&quot;, &quot;STATUS&quot;)
    } else {
        pheno &lt;- read.table(${_input[1]:r}, header=T,stringsAsFactors =F)
        if(&quot;IID&quot; %in% colnames(pheno) == FALSE) stop(&quot;No IID column in the phenoFile. Please rename the header of the phenoFile&quot;)
        if(&quot;FID&quot; %in% colnames(pheno) == FALSE) pheno$FID = pheno$IID
    }
    # Make the unique ID by merge FID and IID
    pheno$ID = paste(pheno$FID,pheno$IID,sep = &quot;:&quot;)
    #check duplicated ID
    if(length(unique(pheno$ID))!=length(pheno$ID)) stop(&quot;There are duplicated names in IID column of phenoFile&quot;)

    pca &lt;-merge(pheno, pca,by =&quot;ID&quot;, all=FALSE)
    dat$pc_scores = bind_rows(dat$pc_scores, pca)
    
    # Write the PC scores to a file
    write.table(dat$pc_scores,&quot;${_output:n}.txt&quot;, sep=&quot;\t&quot;, quote=FALSE, row.names=FALSE, col.names=TRUE)
    # save results
    saveRDS(dat, ${_output:r})
  
bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container, entrypoint=entrypoint
        stdout=$[_output:n].stdout
        for i in $[_output] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;This rds file is a list containing the pca for projected sample&quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        done
        for i in $[_output:n].txt ; do 
        echo &quot;output_info $i &quot; &gt;&gt; $stdout;
        echo &quot;This is the PC score&quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `cat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `cat $i | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        cat $i | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
</section>
<section id="plot-pca-results">
<h2>Plot PCA results<a class="headerlink" href="#plot-pca-results" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Plot PCA results. 
# Can be used independently as &quot;plot_pca&quot; or combined with other workflow as eg &quot;flashpca+plot_pca&quot;
[plot_pca]
parameter: outlier_file = path()
parameter: plot_data = path
parameter: min_axis = &quot;&quot;
parameter: max_axis = &quot;&quot;
input: plot_data
output: f&#39;{cwd}/{_input:bn}.pc.png&#39;,
        f&#39;{cwd}/{_input:bn}.scree.png&#39;,
        f&#39;{cwd}/{_input:bn}.scree.txt&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = 1, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: container = container, volumes = [f&quot;{outlier_file:ad}:{outlier_file:ad}&quot;], expand = &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    library(dplyr)
    library(ggplot2)
    library(gridExtra)
    library(matrixStats)
    pops = c(${paths(pops):r,})
    dat = readRDS(${_input:r})
    f = dat$pca_model
    pca_final&lt;-dat$pc_scores
    
    if (length(pops)&gt;1) pca_final &lt;- pca_final %&gt;% filter(${pop_col if pop_col else &quot;pop&quot;} %in% pops | ${label_col if label_col else  &quot;pop&quot;} %in% pops) 
    pca_final &lt;- pca_final %&gt;% mutate(${label_col if label_col else &quot;pop&quot;}=as.character(${label_col if label_col else  &quot;pop&quot;}))
    k = ${k}
    # manually set colors for PCA plotting, to avoid similar colors in one plot:
    # generated by https://mokole.com/palette.html
    set.seed(999)
    colors_40 = sample(c(&quot;#a9a9a9&quot;, &quot;#2f4f4f&quot;, &quot;#556b2f&quot;, &quot;#a0522d&quot;, &quot;#7f0000&quot;, &quot;#006400&quot;, &quot;#808000&quot;, &quot;#483d8b&quot;, &quot;#3cb371&quot;, &quot;#bdb76b&quot;, &quot;#4682b4&quot;, &quot;#9acd32&quot;, 
                          &quot;#20b2aa&quot;, &quot;#00008b&quot;, &quot;#32cd32&quot;, &quot;#daa520&quot;, &quot;#7f007f&quot;, &quot;#b03060&quot;, &quot;#ff0000&quot;, &quot;#ff8c00&quot;, &quot;#ffff00&quot;, &quot;#0000cd&quot;, &quot;#00ff00&quot;, &quot;#9400d3&quot;, 
                          &quot;#00fa9a&quot;, &quot;#00ffff&quot;, &quot;#00bfff&quot;, &quot;#f4a460&quot;, &quot;#f08080&quot;, &quot;#adff2f&quot;, &quot;#ff6347&quot;, &quot;#ff00ff&quot;, &quot;#1e90ff&quot;, &quot;#dda0dd&quot;, &quot;#7b68ee&quot;, &quot;#afeeee&quot;, 
                          &quot;#ee82ee&quot;, &quot;#ff69b4&quot;, &quot;#ffe4c4&quot;, &quot;#ffc0cb&quot;))
    colors_20 = sample(c(&quot;#2f4f4f&quot;, &quot;#2e8b57&quot;, &quot;#8b0000&quot;, &quot;#808000&quot;, &quot;#00008b&quot;, &quot;#ff0000&quot;, &quot;#ff8c00&quot;, &quot;#00ff00&quot;, &quot;#4169e1&quot;, &quot;#00ffff&quot;, &quot;#00bfff&quot;, &quot;#0000ff&quot;, 
                          &quot;#da70d6&quot;, &quot;#d8bfd8&quot;, &quot;#ff00ff&quot;, &quot;#eee8aa&quot;, &quot;#ffff54&quot;, &quot;#ff1493&quot;, &quot;#ffa07a&quot;, &quot;#98fb98&quot;))

    # assign colors to each ethnicity:
    num_col=length(unique(pca_final$${label_col if label_col else &quot;pop&quot;}))
    if (num_col &lt;= 20) {
       color_list &lt;- colors_20[1:num_col]
    } else {
       color_list &lt;- colors_40[1:num_col]
    }

    ###
    # Make the plots
    ###
    # Get the min and max values for x and y-axes
    if (${&quot;TRUE&quot; if len(min_axis) == 0 or len(max_axis) == 0 else &quot;FALSE&quot;}) {
        min_axis &lt;- round(colMins(as.matrix(f$projection[sapply(f$projection, is.numeric)])),1)
        max_axis &lt;- round(colMaxs(as.matrix(f$projection[sapply(f$projection, is.numeric)])),1)  
    } else {
        min_axis &lt;- as.double(${min_axis})
        max_axis &lt;- as.double(${max_axis})
    }
    if (${&quot;TRUE&quot; if outlier_file.is_file() else &quot;FALSE&quot;}) {
        outliers &lt;- read.table(${outlier_file:r}, col.names=c(&quot;FID&quot;, &quot;IID&quot;),stringsAsFactors =F)
        plot_pcs = function(pca_final, x, y, title=&quot;&quot;) {
            ggplot(pca_final, aes_string(x=x, y=y)) + geom_point(${f&#39;aes(color={label_col})&#39; if label_col else &quot;&quot;}) + 
              # add circles for these ouliters:
              geom_point(data=filter(pca_final, IID %in% outliers$IID, FID %in% outliers$FID), shape = 21, size=1.5, color=&#39;red&#39;, stroke = 0.9) +
              # add outliers dots:
              geom_point(data=filter(pca_final, IID %in% outliers$IID, FID %in% outliers$FID), shape = 16, size=1${f&#39;,aes(color={label_col})&#39; if label_col else &quot;&quot;} ) + 
              labs(title=title,x=x, y=y) +
              scale_y_continuous(limits=c(min_axis, max_axis)) +
              scale_x_continuous(limits=c(min_axis, max_axis)) +
              scale_color_manual(values=color_list) +
              theme_classic()
        }} else {
        plot_pcs = function(pca_final, x, y, title=&quot;&quot;) {
          ggplot(pca_final, aes_string(x=x, y=y)) + geom_point(${f&#39;aes(color={label_col})&#39; if label_col else &quot;&quot;}) + 
              labs(title=title,x=x, y=y) +
              scale_y_continuous(limits=c(min_axis, max_axis)) +
              scale_x_continuous(limits=c(min_axis, max_axis)) +
              scale_color_manual(values=color_list) +
              theme_classic()
        }}
    unit = 4
    n_col = min(4, k)
    n_row = ceiling(k / n_col)
    plots = lapply(1:(k-1), function(i) plot_pcs(pca_final, paste0(&quot;PC&quot;,i), paste0(&quot;PC&quot;,i+1), dat$meta))
    png(&#39;${_output[0]}&#39;, width = unit * n_col, height = unit * n_row, unit=&#39;in&#39;, res=300)
    do.call(gridExtra::grid.arrange, c(plots, list(ncol = n_col, nrow = n_row)))
    dev.off()
    # Create scree plot
    PVE &lt;- f$values
    PVE &lt;- round(PVE/sum(PVE), 2)
    PVEplot &lt;- qplot(c(1:length(PVE)), PVE) + geom_line() + xlab(&quot;Principal Component&quot;) + ylab(&quot;PVE&quot;) + ggtitle(&quot;Scree Plot&quot;) + ylim(0, 1) +scale_x_discrete(limits=factor(1:length(PVE)))
    PVE_cum &lt;- cumsum(PVE)/sum(PVE)
    cumPVEplot &lt;- qplot(c(1:length(PVE)), cumsum(PVE)) + geom_line() + xlab(&quot;Principal Component&quot;) + ylab(&quot;PVE&quot;) + ggtitle(&quot;Cumulative PVE Plot&quot;) + ylim(0, 1) + scale_x_discrete(limits=factor(1:length(PVE)))
    png(&#39;${_output[1]}&#39;, width = 8, height = 4, unit=&#39;in&#39;, res=300)
    grid.arrange(PVEplot, cumPVEplot, nrow = 1)
    dev.off()
    ## Textual Output
    
    PVE_output = tibble(PCs = 1:length(PVE), PVE = PVE, PVE_cum = PVE_cum )
    PVE_output%&gt;%readr::write_delim(${_output[2]:r},&quot;\t&quot;)

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39; , container = container, entrypoint=entrypoint
        stdout=$[_output[0]:n].stdout
        for i in $[_output[2]] ; do 
        echo &quot;output_info $i &quot; &gt;&gt; $stdout;
        echo &quot;This is the PC score&quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `cat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `cat $i | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        cat $i | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
</section>
<section id="detect-outliers">
<h2>Detect outliers<a class="headerlink" href="#detect-outliers" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Calculate Mahalanobis distance per population and report outliers
[detect_outliers]
# Set the probability to remove outliers eg 0.95 or 0.997
parameter: prob = 0.997
# Mahalanobis distance p-value cutoff
parameter: pval = 0.05
# Robust Mahalanobis to outliers
parameter: robust = True
parameter: pca_result = path
input: pca_result
output: distance=f&#39;{_input:n}.mahalanobis.rds&#39;,
        identified_outliers=f&#39;{_input:n}.outliers&#39;,
        analysis_summary=f&#39;{_input:n}.analysis_summary.md&#39;,
        qqplot_mahalanobis=f&#39;{_input:n}.mahalanobis_qq.png&#39;,
        hist_mahalanobis=f&#39;{_input:n}.mahalanobis_hist.png&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = 1, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: container = container, expand = &quot;${ }&quot;, entrypoint=entrypoint
    echo &#39;&#39;&#39;---
    theme: base-theme
    style: |
      img {
        height: 80%;
        display: block;
        margin-left: auto;
        margin-right: auto;
      }
    ---    
    &#39;&#39;&#39; &gt; ${_output[2]}
    
R: container = container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    # Load required libraries
    library(dplyr)
    library(ggplot2)
    library(gridExtra)
  
    # invert a known covariance matrix but allow them to be numerically singular matrix (still assuming full rank)
    robust_inv = function(s) {
        #tryCatch(solve(s), error=function(cond) solve(Matrix::nearPD(s)$mat))
        tryCatch(solve(s), error=function(cond) MASS::ginv(s))
    }

    # Calculate mahalanobis distance  
    calc_mahalanobis_dist = function(x, m, s, name = &#39;&#39;, prob=${prob}) {
        pc &lt;- x %&gt;%
          select(&quot;IID&quot;,&quot;FID&quot;, &quot;${pop_col if pop_col else &quot;pop&quot;}&quot;, starts_with(&quot;PC&quot;))
        mu_pc &lt;- pc[,4:(4 + length(m) - 1)]
        pc$mahal = mahalanobis(mu_pc, m, robust_inv(s), inverted=TRUE)
        pc$p &lt;- pchisq(pc$mahal, df=nrow(s), lower.tail=FALSE)
        manh_dis_sq_cutoff = quantile(pc$mahal, probs=prob)
        # Obtain outliers
        outliers = pc[(pc$mahal &gt; manh_dis_sq_cutoff &amp; pc$p &lt; ${pval}),]
        d_summary = paste0(capture.output(summary(pc$mahal)), collapse = &#39;\n&#39;)
        msg = paste(&#39;#&#39;, name, &quot;result summary\n## Mahalanobis distance summary:\n```\n&quot;, d_summary, &quot;\n```\n&quot;, 
            paste(&quot;The cut-off for outlier removal is set to:&quot;, manh_dis_sq_cutoff, &quot;and the number of individuals to remove is:&quot;, nrow(outliers),&quot;\n&quot;),
            paste(&quot;The new sample size after outlier removal is:&quot;,nrow(pc) - nrow(outliers),&quot;\n&quot;))
        #
        outliers &lt;- outliers %&gt;%
        select(FID,IID)
      list(pc=pc, manh_dis_sq_cutoff=manh_dis_sq_cutoff, msg=msg, outliers=outliers)
    }

    dat = readRDS(${_input:r})
    if (is.list(dat$pc_mean)) {
      pops = names(dat$pc_mean)
      pop_group = split(dat$pc_scores, f = dat$pc_scores$${pop_col if pop_col else &quot;pop&quot;})
      res = lapply(pops, function(p) calc_mahalanobis_dist(pop_group[[p]], dat$${&quot;pc_mean&quot; if not robust else &quot;pc_median&quot;}[[p]], dat$pc_cov[[p]], name = paste(dat$meta, p)))
      names(res) = pops
      res = list(
          msg = do.call(paste, c(lapply(pops, function(p) res[[p]]$msg), sep = &quot;\n&quot;)),
          manh_dis_sq_cutoff = cbind(pops, sapply(pops, function(p) res[[p]]$manh_dis_sq_cutoff)),
          outliers = do.call(rbind, c(lapply(pops, function(p) res[[p]]$outliers))),
          pc = do.call(rbind, c(lapply(pops, function(p) res[[p]]$pc)))
          )
    } else {
      res = calc_mahalanobis_dist(dat$pc_scores, dat$${&quot;pc_mean&quot; if not robust else &quot;pc_median&quot;}, dat$pc_cov, name = dat$meta)
    }
      
    write(res$msg, ${_output[2]:r})   
    # Plot mahalanobis
    k = ${k}
    png(&#39;${_output[3]}&#39;, width = 4, height = 4, unit=&#39;in&#39;, res=300)
    qqplot(qchisq(ppoints(100), df=k), res$pc$mahal, main = expression(&quot;Mahalanobis&quot; * ~D^2 * &quot; vs. quantiles of&quot; * ~ chi[k]^2), xlab = expression(chi[2]^2 * &quot;, probability points = 100&quot;), ylab = expression(D^2), pch=16)
    abline(0,1,col=&#39;red&#39;)
    dev.off() 
    png(&#39;${_output[4]}&#39;, width = 4, height = 4, unit=&#39;in&#39;, res=300)
    ggplot(res$pc, aes(x=mahal)) + geom_histogram(aes(y = ..count..), binwidth = 0.5, colour = &quot;#1F3552&quot;, fill = &quot;#4271AE&quot;) + scale_x_continuous(name = &quot;Mahalanobis distance&quot;) + theme_classic()
    dev.off()
  
    # Save results and outliers
    saveRDS(res,${_output[0]:r})
    write.table(res$outliers, ${_output[1]:r}, sep=&quot;\t&quot;, quote=FALSE, row.names=FALSE, col.names=FALSE)
  
bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint=entrypoint
        stdout=$[_output[0]:n].stdout
        for i in $[_output[0]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;This rds file detail analysis output for detect outliers&quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        done
        for i in $[_output[1]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;This text file documents the outliers samples&quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        done
        for i in $[_output[2]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;This text file documents the analysis summary&quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        done
        for i in $[_output[3]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;This plot documents the mahalanobis analysis result&quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout; 
        done
        for i in $[_output[4]] ; do 
        echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
        echo &quot;This plot documents the mahalanobis distribution result&quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        done
</pre></div>
</div>
</div>
</div>
</section>
<section id="add-plot-and-outlier-detection-to-pca-steps">
<h2>Add plot and outlier detection to PCA steps<a class="headerlink" href="#add-plot-and-outlier-detection-to-pca-steps" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">flashpca_2</span><span class="p">,</span> <span class="n">project_samples_2</span><span class="p">]</span>
<span class="c1"># Set the probability to remove outliers eg 0.95 or 0.997</span>
<span class="kn">parameter:</span> <span class="n">prob</span> <span class="o">=</span> <span class="mf">0.997</span>
<span class="c1"># Robust Mahalanobis to outliers</span>
<span class="kn">parameter:</span> <span class="n">robust</span> <span class="o">=</span> <span class="kc">True</span>
<span class="kn">output:</span> <span class="n">distance</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.mahalanobis.rds&#39;</span><span class="p">,</span>
        <span class="n">identified_outliers</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.outliers&#39;</span><span class="p">,</span>
        <span class="n">analysis_summary</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.analysis_summary.md&#39;</span><span class="p">,</span>
        <span class="n">qqplot_mahalanobis</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.mahalanobis_qq.png&#39;</span><span class="p">,</span>
        <span class="n">hist_mahalanobis</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.mahalanobis_hist.png&#39;</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s2">&quot;detect_outliers&quot;</span><span class="p">,</span> <span class="n">pca_result</span><span class="o">=</span><span class="n">_input</span><span class="p">,</span> <span class="n">prob</span><span class="o">=</span><span class="n">prob</span><span class="p">,</span> <span class="n">robust</span><span class="o">=</span><span class="n">robust</span><span class="p">)</span>

<span class="p">[</span><span class="n">flashpca_3</span><span class="p">,</span> <span class="n">project_samples_3</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="mi">1</span><span class="p">),</span> <span class="n">output_from</span><span class="p">(</span><span class="mi">2</span><span class="p">)[</span><span class="s1">&#39;identified_outliers&#39;</span><span class="p">]</span>
<span class="n">outliers</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span> <span class="k">if</span> <span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">.pc.png&quot;</span><span class="p">,</span>
        <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">.scree.png&quot;</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s2">&quot;plot_pca&quot;</span><span class="p">,</span> <span class="n">plot_data</span> <span class="o">=</span> <span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span> <span class="n">outlier_file</span> <span class="o">=</span> <span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">outliers</span><span class="p">)</span> <span class="k">else</span> <span class="n">path</span><span class="p">())</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/data_preprocessing/genotype"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="GWAS_QC.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Genotype PLINK file quality control</p>
      </div>
    </a>
    <a class="right-next"
       href="GRM.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Genomic relationship matrices</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#general-workflow">General workflow</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#method">Method</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-estimate-kinship-in-the-sample">Step 1: Estimate kinship in the sample</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-and-3-sample-selection-and-qc-the-genotype-data-for-pca">Step 2 and 3: Sample selection and QC the genotype data for PCA</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-pca-analysis-for-unrelated-samples">Step 4: PCA analysis for unrelated samples</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-projection-of-related-individuals">Step 5: Projection of related individuals</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-5-2-finalize-genotype-qc-by-pca-for-homogenous-population">Step 5-2: Finalize genotype QC by PCA for homogenous population</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-6-split-data-by-population">Step 6: Split data by population</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-7-for-each-population-do-step-2">Step 7: For each population do step 2</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-8-for-each-population-do-step-3-if-applicable">Step 8: For each population do step 3 if applicable</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-9-for-each-population-do-step-4">Step 9: For each population do step 4</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-10-for-each-population-do-step-5-if-applicable">Step 10: For each population do step 5 if applicable</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#final-step-for-each-population-do-their-own-qc-to-finalize">Final step: For each population do their own QC to finalize</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-analysis">PCA analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#project-related-individuals-back">Project related individuals back</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#plot-pca-results">Plot PCA results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#detect-outliers">Detect outliers</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#add-plot-and-outlier-detection-to-pca-steps">Add plot and outlier detection to PCA steps</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=c5ced968eda925caa686"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=c5ced968eda925caa686"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>