

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Quantifying expression from RNA-seq data &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/molecular_phenotypes/calling/RNA_calling';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Sample level RNA-seq quality control" href="../QC/bulk_expression_QC.html" />
    <link rel="prev" title="RNA-seq expression" href="../bulk_expression.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../data_preprocessing/reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../bulk_expression.html">RNA-seq expression</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mashr.html">Data overview</a></li>






</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">Fine-mapping with SuSiE model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc_dap.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/molecular_phenotypes/calling/RNA_calling.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/molecular_phenotypes/calling/RNA_calling.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/molecular_phenotypes/calling/RNA_calling.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Quantifying expression from RNA-seq data</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods-overview">Methods overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-level-quantifications">Gene-level quantifications</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exon-level-quantifications">Exon-level quantifications</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#paired-end">Paired end:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#single-end">Single end:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#strand-option"><code class="docutils literal notranslate"><span class="pre">strand</span></code> option</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#read-length-option"><code class="docutils literal notranslate"><span class="pre">read_length</span></code> option</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-and-global-parameters">Setup and global parameters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-0-qc-before-alignment">Step 0: QC before alignment</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-inputs">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-outputs">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-remove-adaptor-through-fastp">Step 1: Remove adaptor through <code class="docutils literal notranslate"><span class="pre">fastp</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Step Outputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-options">Step options</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-alternative-remove-adaptor-through-trimmomatic">Step 1 Alternative: Remove adaptor through <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-alignment-through-star">Step 2: Alignment through <code class="docutils literal notranslate"><span class="pre">STAR</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-mark-duplicates-reads-qc-through-picard">Step 3: Mark duplicates reads &amp; QC through <code class="docutils literal notranslate"><span class="pre">Picard</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id7">Step Inputs:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id8">Step Outputs:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-post-aligment-qc-through-rna-seqc">Step 4: Post aligment QC through <code class="docutils literal notranslate"><span class="pre">RNA-SeQC</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id9">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id10">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-6-quantify-expression-through-rsem">Step 6: Quantify expression through <code class="docutils literal notranslate"><span class="pre">RSEM</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-input">Step Input</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id11">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-7-summarize-with-multiqc">Step 7: summarize with MultiQC</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="quantifying-expression-from-rna-seq-data">
<h1>Quantifying expression from RNA-seq data<a class="headerlink" href="#quantifying-expression-from-rna-seq-data" title="Permalink to this heading">#</a></h1>
<p>This pipeline aims to call expression levels from RNA-seq data starting from <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> sequences. It implements the GTEx pipeline for GTEx / TOPMed project. Please refer to <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/TOPMed_RNAseq_pipeline.md">this page</a> for detail.</p>
<p><strong>Various reference data needs to be prepared before using this workflow</strong>. <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/reference_data.html">Here we provide a module</a> to download and prepare the reference data.</p>
<section id="methods-overview">
<h2>Methods overview<a class="headerlink" href="#methods-overview" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="o">%</span><span class="n">preview</span> <span class="o">../../</span><span class="n">images</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">jpeg</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> ../../images/rna_quantification.jpeg (495.9 KiB):</div></div><img alt="../../../_images/809e3fa6cf05636b9c489361bdfd6b4386ee60fda547a5977f3525466300177e.png" src="../../../_images/809e3fa6cf05636b9c489361bdfd6b4386ee60fda547a5977f3525466300177e.png" />
</div>
</div>
<p>The procedure is decribed also in <a class="reference external" href="https://gtexportal.org/home/documentationPage#staticTextAnalysisMethods">GTEx portal</a>. Here is a recap of some important details.</p>
<section id="gene-level-quantifications">
<h3>Gene-level quantifications<a class="headerlink" href="#gene-level-quantifications" title="Permalink to this heading">#</a></h3>
<p>read counts and TPM values were produced with RNA-SeQC v2.4.2 (DeLuca et al., Bioinformatics, 2012 ), using the following read-level filters:</p>
<ul class="simple">
<li><p>Reads were uniquely mapped (corresponding to a mapping quality of 255 for START BAMs).</p></li>
<li><p>Reads were aligned in proper pairs.</p></li>
<li><p>The read alignment distance was &lt;=6 (i.e., alignments must not contain more than six non-reference bases).</p></li>
<li><p>Reads were fully contained within exon boundaries. Reads overlapping introns were not counted.</p></li>
</ul>
<p>These filters were applied using the “-strictMode” flag in RNA-SeQC. The TPM values have not been normalized or corrected for any covariates.</p>
</section>
<section id="exon-level-quantifications">
<h3>Exon-level quantifications<a class="headerlink" href="#exon-level-quantifications" title="Permalink to this heading">#</a></h3>
<p>for exon-level read counts, if a read overlapped multiple exons, then a fractional value equal to the portion of the read contained within that exon was allotted. Transcript-level quantifications were calculated using RSEM v1.3.0.</p>
</section>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this heading">#</a></h2>
<section id="paired-end">
<h3>Paired end:<a class="headerlink" href="#paired-end" title="Permalink to this heading">#</a></h3>
<p>A meta-data file, tab delimited with header, containing 3 required columns: ID, fq1, fq2 and 2 optional columns: strand and read_length:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ID</span> <span class="n">fq1</span> <span class="n">fq2</span> <span class="n">strand</span> <span class="n">read_length</span>
<span class="n">sample_1</span> <span class="n">samp1_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">samp1_r2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">rf</span> <span class="mi">100</span>
<span class="n">sample_2</span> <span class="n">samp2_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">samp2_r2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">fr</span> <span class="mi">150</span>
<span class="n">sample_3</span> <span class="n">samp3_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">samp3_r2</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">strand_missing</span> <span class="mi">75</span>
</pre></div>
</div>
</section>
<section id="single-end">
<h3>Single end:<a class="headerlink" href="#single-end" title="Permalink to this heading">#</a></h3>
<p>A similar meta-data file, tab delimited with header, containing 2 required columns: ID, fq1 and 2 optional columns: strand and read_length:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ID</span> <span class="n">fq1</span> <span class="n">fq2</span> <span class="n">strand</span> <span class="n">read_length</span>
<span class="n">sample_1</span> <span class="n">samp1_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">rf</span> <span class="mi">100</span>
<span class="n">sample_2</span> <span class="n">samp2_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">fr</span> <span class="mi">150</span>
<span class="n">sample_3</span> <span class="n">samp3_r1</span><span class="o">.</span><span class="n">fq</span><span class="o">.</span><span class="n">gz</span> <span class="n">strand_missing</span> <span class="mi">75</span>
</pre></div>
</div>
<p>The interface for single end and paired end data are the same. Only difference lied in whether there is a fq2 column in the meta data.</p>
<p>All the fastq files should be available under specified folder (default assumes the same folder as where the meta-data file is).</p>
</section>
</section>
<section id="strand-option">
<h2><code class="docutils literal notranslate"><span class="pre">strand</span></code> option<a class="headerlink" href="#strand-option" title="Permalink to this heading">#</a></h2>
<p>Some steps requires you specify strand option via <code class="docutils literal notranslate"><span class="pre">--strand</span></code> and should be one of<code class="docutils literal notranslate"><span class="pre">rf</span></code>, <code class="docutils literal notranslate"><span class="pre">fr</span></code>,<code class="docutils literal notranslate"><span class="pre">unstranded</span></code> , or <code class="docutils literal notranslate"><span class="pre">strand_missing</span></code>. Please refer to <a class="reference external" href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-022-04572-7">Signal et al (2022)</a> for an explanation what these library types mean. <strong>If you are not sure about the strandedness of your data, our strand detection steps will be automatically ran to detect the strand of input fastq file levaging the <a class="reference external" href="https://physiology.med.cornell.edu/faculty/skrabanek/lab/angsd/lecture_notes/STARmanual.pdf">gene count table output from STAR</a> . The definition of strandness are following <a class="reference external" href="https://rnabio.org/module-09-appendix/0009/12/01/StrandSettings/">this page</a>. The percentage cutoff are based on <a class="reference external" href="https://github.com/signalbash/how_are_we_stranded_here">how are we stranded here</a>.</strong></p>
</section>
<section id="read-length-option">
<h2><code class="docutils literal notranslate"><span class="pre">read_length</span></code> option<a class="headerlink" href="#read-length-option" title="Permalink to this heading">#</a></h2>
<p>The the read length for each sample should be specified in the read_length column in the input meta-data file (see input section). If you don’t know the read length for a sample, enter 0 and the pipeline will use 100 as the default value. You can change this default value by using the <code class="docutils literal notranslate"><span class="pre">--sjdbOverhang</span></code> parameter. If none of the samples have read length information, please left out the read_length column and the pipeline will use the default one for each sample.</p>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading">#</a></h2>
<p>TPM count for gene level RNA expression via <code class="docutils literal notranslate"><span class="pre">rnaseqc</span></code>, and isoform level RNA expression via <code class="docutils literal notranslate"><span class="pre">RSEM</span></code>.</p>
</section>
<section id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this heading">#</a></h2>
<p>A toy <code class="docutils literal notranslate"><span class="pre">fastq</span></code> data can be found on <a class="reference external" href="https://www.synapse.org/#!Synapse:syn36417253">synapse</a>.</p>
<p>To generate <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> report,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fastqc</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_test</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">test_data</span><span class="o">/</span><span class="n">test_data</span><span class="o">.</span><span class="n">fastqlist</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">test_data</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> 
</pre></div>
</div>
</div>
</div>
<p>To cut adaptor, after cutting the adaptor, a new sample list will be generated with the trimmed fastq file. Please noted that there is no adaptors to be trimmed in the test data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fastp_trim_adaptor</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_test</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">test_data</span><span class="o">/</span><span class="n">test_data</span><span class="o">.</span><span class="n">fastqlist</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">test_data</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">STAR_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">flat</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">flat</span> <span class="o">-</span><span class="n">s</span> <span class="n">build</span> 
</pre></div>
</div>
</div>
</div>
<p>To align the reads with STAR and generate the bam_list recipe for downstream molecular phenotype count matrixes. The <code class="docutils literal notranslate"><span class="pre">-J</span> <span class="pre">4</span> <span class="pre">-c</span> <span class="pre">csg.yml</span> <span class="pre">-q</span> <span class="pre">csg</span></code> part is crucial for it ask for the required memory to conduct the STAR alignment. The STAR command can only be ran with more than 40G of memory or with the counter-part of <code class="docutils literal notranslate"><span class="pre">-J</span> <span class="pre">4</span> <span class="pre">-c</span> <span class="pre">csg.yml</span> <span class="pre">-q</span> <span class="pre">csg</span></code></p>
<p><em><strong>Warning</strong></em>: It is of crucial importance that, the STAR Alignment shall be ran with the gtf file that were generated <code class="docutils literal notranslate"><span class="pre">before</span> <span class="pre">collapsing</span> <span class="pre">to</span> <span class="pre">gene</span></code>, i.e. the one that are used to generated RSEM index. Other wise there will be an <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/issues/390">error while running RSEM later on</a></p>
<p><em><strong>Warning</strong></em>: It is also crucial that, the <code class="docutils literal notranslate"><span class="pre">STAR_align</span></code>,<code class="docutils literal notranslate"><span class="pre">picard_qc</span></code>,<code class="docutils literal notranslate"><span class="pre">strand_detect</span></code> are not to be ran saperately, but instead they need to be run together by running <code class="docutils literal notranslate"><span class="pre">STAR_output</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">STAR_output</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_test</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">test_data</span><span class="o">/</span><span class="n">test_data</span><span class="o">.</span><span class="n">fastqlist</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">test_data</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">STAR_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">flat</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">flat</span> <span class="o">-</span><span class="n">s</span> <span class="n">build</span>  <span class="o">-</span><span class="n">J</span> <span class="mi">4</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">--</span><span class="n">uncompressed</span> <span class="o">&amp;</span> 
</pre></div>
</div>
</div>
</div>
<p>To call gene-level RNA expression using <code class="docutils literal notranslate"><span class="pre">rnaseqc</span></code>.</p>
<p><em><strong>Warning</strong></em>: Unlike STAR_output and rsem, rnaseqc required the gtf <code class="docutils literal notranslate"><span class="pre">after</span> <span class="pre">collapsing</span> <span class="pre">to</span> <span class="pre">gene</span> </code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">rnaseqc_call</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_test</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">test_data</span><span class="o">/</span><span class="n">test_data</span><span class="o">.</span><span class="n">fastqlist</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">test_data</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">STAR_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">collapse_only</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy_ERCC</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">flat</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">flat</span> \
    <span class="o">--</span><span class="n">bam_list</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">finalizing</span><span class="o">/</span><span class="n">output_test</span><span class="o">/</span><span class="n">test_data_bam_list</span> 
</pre></div>
</div>
</div>
</div>
<p>To call transcript level RNA expression using <code class="docutils literal notranslate"><span class="pre">RSEM</span></code>:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">rsem_call</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">rnaseq</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">data</span><span class="o">/</span><span class="n">sample_fastq</span><span class="o">.</span><span class="n">list</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">data</span> \
    <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">STAR_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">RSEM</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">RSEM_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">flat</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">flat</span> \
    <span class="o">--</span><span class="n">bam_list</span> <span class="n">output3</span><span class="o">/</span><span class="n">sample_fastq_bam_list</span> <span class="o">-</span><span class="n">J</span> <span class="mi">3</span> <span class="o">-</span><span class="n">c</span> <span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">rsem_call</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output_test</span> \
    <span class="o">--</span><span class="n">samples</span> <span class="n">test_data</span><span class="o">/</span><span class="n">test_data</span><span class="o">.</span><span class="n">fastqlist</span> \
    <span class="o">--</span><span class="n">data</span><span class="o">-</span><span class="nb">dir</span> <span class="n">test_data</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">STAR</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">STAR_Index</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">gtf</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">collapse_only</span><span class="o">.</span><span class="n">gene</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">gtf</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">rna_quantification</span><span class="o">.</span><span class="n">sif</span> \
    <span class="o">--</span><span class="n">reference</span><span class="o">-</span><span class="n">fasta</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">GRCh38_full_analysis_set_plus_decoy_hla</span><span class="o">.</span><span class="n">noALT_noHLA_noDecoy</span><span class="o">.</span><span class="n">fasta</span> \
    <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">flat</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">Homo_sapiens</span><span class="o">.</span><span class="n">GRCh38</span><span class="mf">.103</span><span class="o">.</span><span class="n">chr</span><span class="o">.</span><span class="n">reformatted</span><span class="o">.</span><span class="n">ERCC</span><span class="o">.</span><span class="n">ref</span><span class="o">.</span><span class="n">flat</span> \
    <span class="o">--</span><span class="n">bam_list</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">finalizing</span><span class="o">/</span><span class="n">output_test</span><span class="o">/</span><span class="n">test_data_bam_list</span> \
    <span class="o">--</span><span class="n">RSEM</span><span class="o">-</span><span class="n">index</span> <span class="n">reference_data</span><span class="o">/</span><span class="n">RSEM_Index</span><span class="o">/</span> 
</pre></div>
</div>
</div>
</div>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">RNA_calling</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run RNA_calling.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  fastqc
  fastp_trim_adaptor
  trimmomatic
  STAR_align
  rnaseqc_call
  rsem_call
  picard_qc

Global Workflow Options:
  --cwd output (as path)
                        The output directory for generated files.
  --samples VAL (as path, required)
                        Sample meta data list
  --data-dir  path(f&quot;{samples:d}&quot;)

                        Raw data directory, default to the same directory as
                        sample list
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --java-mem 6G
                        Memory for Java virtual mechine (`picard`)
  --numThreads 8 (as int)
                        Number of threads
  --container &#39;&#39;
                        Software container option
  --gtf VAL (as path, required)
                        Reference gene model

Sections
  fastqc:
  fastp_trim_adaptor:
    Workflow Options:
      --window-size 4 (as int)
                        sliding window setting
      --required-quality 20 (as int)
      --leading 20 (as int)
                        the mean quality requirement option for cut_front
      --trailing 20 (as int)
                        the mean quality requirement option for cut_tail
      --min-len 15 (as int)
                        reads shorter than length_required will be discarded
  trimmomatic:
    Workflow Options:
      --trimmomatic-jar &#39;/opt/Trimmomatic-0.39/trimmomatic-0.39.jar&#39;
                        Path to the software. Default set to using our
                        rna_quantification.sif image
      --fasta-with-adapters-etc VAL (as str, required)
                        Illumina clip setting Path to the reference adaptors
      --seed-mismatches 2 (as int)
      --palindrome-clip-threshold 30 (as int)
      --simple-clip-threshold 10 (as int)
      --window-size 4 (as int)
                        sliding window setting
      --required-quality 20 (as int)
      --leading 3 (as int)
                        Other settings
      --trailing 3 (as int)
      --min-len 50 (as int)
  STAR_align:
    Workflow Options:
      --STAR-index VAL (as path, required)
                        STAR indexing file
      --outFilterMultimapNmax 20 (as int)
                        Alignment parameter
      --alignSJoverhangMin 8 (as int)
      --alignSJDBoverhangMin 1 (as int)
      --outFilterMismatchNmax 999 (as int)
      --outFilterMismatchNoverLmax 0.1 (as float)
      --alignIntronMin 20 (as int)
      --alignIntronMax 1000000 (as int)
      --alignMatesGapMax 1000000 (as int)
      --outFilterType BySJout
      --outFilterScoreMinOverLread 0.33 (as float)
      --outFilterMatchNminOverLread 0.33 (as float)
      --limitSjdbInsertNsj 1200000 (as int)
      --outSAMstrandField intronMotif
      --outFilterIntronMotifs None
      --alignSoftClipAtReferenceEnds Yes
      --quantMode TranscriptomeSAM GeneCounts (as list)
      --outSAMattrRGline ID:rg1 SM:sm1 (as list)
      --outSAMattributes NH HI AS nM NM ch (as list)
      --chimSegmentMin 15 (as int)
      --chimJunctionOverhangMin 15 (as int)
      --chimOutType Junctions WithinBAM SoftClip (as list)
      --chimMainSegmentMultNmax 1 (as int)
  rnaseqc_call_1, rsem_call_1, picard_qc:
    Workflow Options:
      --ref-flat . (as path)
                        Path to flat reference file, for computing QC metric
      --picard-jar &#39;/opt/picard-tools/picard.jar&#39;
                        Path to the software. Default set to using our
                        rna_quantification.sif image
      --reference-fasta VAL (as path, required)
                        The fasta reference file used to generate star index
      --optical-distance 100 (as int)
                        For the patterned flowcell models (HiSeq X), change to
                        2500
  rnaseqc_call_2:
    Workflow Options:
      --strand VAL (as str, required)
                        Options are fr or rf or unstranded
  rnaseqc_call_3:
  rsem_call_2:
    Workflow Options:
      --RSEM-index VAL (as path, required)
      --max-frag-len 1000 (as int)
      --strand VAL (as str, required)
                        Options are fr or rf or unstranded
  rsem_call_3:
  rsem_call_4, rnaseqc_call_4:
  rsem_call_5, rnaseqc_call_5:
    Workflow Options:
      --ref-flat . (as path)
                        Path to flat reference file, for computing QC metric
</pre></div>
</div>
</div>
</div>
</section>
<section id="setup-and-global-parameters">
<h2>Setup and global parameters<a class="headerlink" href="#setup-and-global-parameters" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># The output directory for generated files.</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># Sample meta data list</span>
<span class="kn">parameter:</span> <span class="n">samples</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Raw data directory, default to the same directory as sample list</span>
<span class="kn">parameter:</span> <span class="n">data_dir</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s2">d</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;5h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;16G&quot;</span>
<span class="c1"># Memory for Java virtual mechine (`picard`)</span>
<span class="kn">parameter:</span> <span class="n">java_mem</span> <span class="o">=</span> <span class="s2">&quot;6G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">8</span>
<span class="c1"># Software container option</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span><span class="p">{(</span><span class="s1">&#39;micromamba run -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">][:</span><span class="o">-</span><span class="mi">4</span><span class="p">])</span> <span class="k">if</span> <span class="n">container</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.sif&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="sa">f</span><span class="s1">&#39;&#39;</span><span class="p">}</span>
<span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="n">expand_size</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="c1">## Whether the fasta/fastq file is compressed or not.</span>
<span class="kn">parameter:</span> <span class="n">uncompressed</span> <span class="o">=</span> <span class="kc">False</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="c1">## FIX: The way to get sample needs to be revamped to 1. Accomodate rf/fr as column 2. accomodate single end read (Only 2 fq/samples)</span>
<span class="n">sample_inv</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="c1">## Extract strand information if user have specified the strand</span>
<span class="n">strand_inv</span> <span class="o">=</span> <span class="p">[]</span>

<span class="k">if</span> <span class="s2">&quot;strand&quot;</span> <span class="ow">in</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">strand_inv</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">strand</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">sample_inv</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span> <span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    <span class="n">stop_if</span><span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">([</span><span class="n">x</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;fr&quot;</span><span class="p">,</span> <span class="s2">&quot;rf&quot;</span><span class="p">,</span> <span class="s2">&quot;unstranded&quot;</span><span class="p">,</span><span class="s2">&quot;strand_missing&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">strand_inv</span> <span class="p">]),</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;strand columns should only include ``fr``, ``rf``, ``strand_missing`` or ``unstranded``&quot;</span><span class="p">)</span>
            
<span class="c1">## Extract read_length if user have specified read_length</span>
<span class="n">read_length</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample_inv</span><span class="o">.</span><span class="n">index</span><span class="p">)</span>
<span class="k">if</span> <span class="s2">&quot;read_length&quot;</span> <span class="ow">in</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
    <span class="n">read_length</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">read_length</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="n">sample_inv</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s2">&quot;read_length&quot;</span> <span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span>
    
    
<span class="c1">## Extract sample_id</span>
<span class="n">sample_inv_list</span> <span class="o">=</span> <span class="n">sample_inv</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">sample_id</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sample_inv_list</span><span class="p">]</span>


    
<span class="c1">## Get the file name for single/paired end data</span>
<span class="n">file_inv</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">:]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">sample_inv_list</span><span class="p">]</span>
<span class="n">file_inv</span> <span class="o">=</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">sublist</span> <span class="ow">in</span> <span class="n">file_inv</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">sublist</span><span class="p">]</span>

<span class="n">fastq</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">data_dir</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">file_inv</span><span class="p">]</span>


<span class="k">for</span> <span class="n">y</span> <span class="ow">in</span> <span class="n">fastq</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">y</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;File </span><span class="si">{</span><span class="n">y</span><span class="si">}</span><span class="s2"> does not exist&quot;</span><span class="p">)</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fastq</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">fastq</span><span class="p">)):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Duplicated files are found (but should not be allowed) in fastq file list&quot;</span><span class="p">)</span>

<span class="c1"># Is the RNA-seq data pair-end</span>
<span class="n">is_paired_end</span> <span class="o">=</span> <span class="mi">0</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">fastq</span><span class="p">)</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">sample_id</span><span class="p">)</span> <span class="k">else</span> <span class="mi">1</span> 
<span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="kp">env</span>
<span class="kp">env</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Input samples are </span><span class="si">{</span><span class="s2">&quot;paired-end&quot;</span><span class="w"> </span><span class="k">if</span><span class="w"> </span><span class="n">is_paired_end</span><span class="w"> </span><span class="k">else</span><span class="w"> </span><span class="s2">&quot;single-end&quot;</span><span class="si">}</span><span class="s1"> sequences.&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="step-0-qc-before-alignment">
<h2>Step 0: QC before alignment<a class="headerlink" href="#step-0-qc-before-alignment" title="Permalink to this heading">#</a></h2>
<p>This step utilize <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> and will generate two QC report in <code class="docutils literal notranslate"><span class="pre">html</span></code> format. It should be noted that the <a class="reference external" href="https://www.biostars.org/p/190584/">paired_end reads will also be qc separately</a></p>
<section id="step-inputs">
<h3>Step Inputs<a class="headerlink" href="#step-inputs" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fastq1</span></code> and <code class="docutils literal notranslate"><span class="pre">fastq2</span></code>: paths to original <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> file.</p></li>
</ul>
</section>
<section id="step-outputs">
<h3>Step Outputs<a class="headerlink" href="#step-outputs" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Two <code class="docutils literal notranslate"><span class="pre">html</span></code> file for QC report</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[fastqc]
input: fastq, group_by =  1
output: f&#39;{cwd}/{_input:bn}_fastqc.html&#39;,f&#39;{cwd}/{_input:bn}_fastqc.zip&#39; 
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
bash: expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container=container, entrypoint=entrypoint
    fastqc ${_input} -o ${_output[0]:d}
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-1-remove-adaptor-through-fastp">
<h2>Step 1: Remove adaptor through <code class="docutils literal notranslate"><span class="pre">fastp</span></code><a class="headerlink" href="#step-1-remove-adaptor-through-fastp" title="Permalink to this heading">#</a></h2>
<p>Documentation: <a class="reference external" href="https://github.com/OpenGene/fastp">fastp</a></p>
<p>We use <code class="docutils literal notranslate"><span class="pre">fastp</span></code> in place of the <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> for fastp’s ability to detect adapter from the reads. It was a c++ command line tool published in <a class="reference external" href="https://academic.oup.com/bioinformatics/article/34/17/i884/5093234">Sept 2018</a>. It  will use the following algorithm to detect the adaptors:</p>
<blockquote>
<div><p>The adapter-sequence detection algorithm is based on two assumptions: the first is that only one adapter exists in the data; the second is that adapter sequences exist only in the read tails. These two assumptions are valid for major next-generation sequencers like Illumina HiSeq series, NextSeq series and NovaSeq series. We compute the k-mer (k = 10) of first N reads (N = 1 M). From this k-mer, the sequences with high occurrence frequencies (&gt;0.0001) are considered as adapter seeds. Low-complexity sequences are removed because they are usually caused by sequencing artifacts. The adapter seeds are sorted by its occurrence frequencies. A tree-based algorithm is applied to extend the adapter seeds to find the real complete adapter</p>
</div></blockquote>
<p>It was demostrated that fastp can remove all the adaptor automatically and completely faster than Trimmomatic and cutadapt</p>
<section id="id1">
<h3>Step Inputs<a class="headerlink" href="#id1" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fastq</span></code>: 1 set of fq.gz file for each sample. (2 files if paired end, 1 file if single end )</p></li>
</ul>
</section>
<section id="id2">
<h3>Step Outputs<a class="headerlink" href="#id2" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>1 set of  <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> file for alignment. (2 files if paired end, 1 file if single end )</p></li>
<li><p>The unpaired reads (i.e.where a read survived, but the partner read did not.) will be discarded by default as those were not used in the following steps. This feature can be added were it was needed.</p></li>
<li><p>1 set of html documenting the quality of input reads(1 html file and 1 json file)</p></li>
</ul>
</section>
<section id="step-options">
<h3>Step options<a class="headerlink" href="#step-options" title="Permalink to this heading">#</a></h3>
<p>A few options were selected to be customizable so that fastp step can have the same level of flexiblity as the Trimmomatic step had. They are:</p>
<ul class="simple">
<li><p>min_len: length_required, reads shorter than this will be discarded, default is 15. (int [=15])</p></li>
<li><p>window_size: cut_window_size, the window size option shared by cut_front, cut_tail or cut_sliding. Range: 1~1000, default: 4 (int [=4])</p></li>
<li><p>leading/trailing : cut_front_mean_quality/cut_tail_mean_quality, the mean quality requirement option for cut_front/cut_tail, which move a sliding window from front/tail, drop the bases in the window if its mean quality &lt; threshold. <strong>Notice the choice of quality score in <code class="docutils literal notranslate"><span class="pre">fastp</span></code> (N=20) is a lot higher than that of <code class="docutils literal notranslate"><span class="pre">trimmomatic</span></code> (N=3)</strong>. Default <code class="docutils literal notranslate"><span class="pre">fastp</span></code> setting is in line with that of <a class="reference external" href="https://cutadapt.readthedocs.io/en/stable/guide.html"><code class="docutils literal notranslate"><span class="pre">cutadapt</span></code></a>.</p></li>
</ul>
<p>The default value are set to be the same as the default value of the <code class="docutils literal notranslate"><span class="pre">fastp</span></code> software.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[fastp_trim_adaptor_1]
# sliding window setting
parameter: window_size = 4
parameter: required_quality = 20
# the mean quality requirement option for cut_front
parameter: leading = 20
# the mean quality requirement option for cut_tail
parameter: trailing = 20
# reads shorter than length_required will be discarded
parameter: min_len = 15
# Path to the reference adaptors
parameter: fasta_with_adapters_etc = path(&quot;.&quot;)
warn_if(fasta_with_adapters_etc.is_file(),msg = &quot;Use input fasta and adaptor detection of paired-end read was disabled&quot; )

input: fastq, group_by = is_paired_end + 1 , group_with = &quot;sample_id&quot;
output: [f&#39;{cwd}/{path(x):bn}.trimmed.fq.gz&#39; for x in _input]
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
bash:  container=container,expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
        fastp -i ${f&#39;{_input[0]} -I {_input[1]}&#39; if is_paired_end else _input} -o ${ f&#39;{_output[0]} -O {_output[1]}&#39; if is_paired_end else _output } \
            ${f&#39;--adapter_fasta {fasta_with_adapters_etc}&#39; if fasta_with_adapters_etc.is_file() else &quot;--detect_adapter_for_pe&quot;}  -V -h ${_output[0]:n}.html -j ${_output[0]:n}.json -w ${numThreads} \
            --length_required ${min_len}  -W ${window_size} -M ${required_quality} -5 -3 --cut_front_mean_quality ${leading} --cut_tail_mean_quality ${leading}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">fastp_trim_adaptor_2</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.trimmed.txt&#39;</span>
<span class="n">sample_inv_tmp</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">samples</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">csv</span>
<span class="k">if</span> <span class="n">is_paired_end</span><span class="p">:</span>
    <span class="n">sample_inv_tmp</span><span class="o">.</span><span class="n">fq1</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">r</span><span class="si">}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">][::</span><span class="mi">2</span><span class="p">]</span>
    <span class="n">sample_inv_tmp</span><span class="o">.</span><span class="n">fq2</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">r</span><span class="si">}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">][</span><span class="mi">1</span><span class="p">::</span><span class="mi">2</span><span class="p">]</span>
<span class="kn">else:</span>
    <span class="n">sample_inv_tmp</span><span class="o">.</span><span class="n">fq1</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">r</span><span class="si">}</span><span class="s1">&#39;</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s2">&quot;&#39;&quot;</span><span class="p">,</span><span class="s2">&quot;&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">]</span>
<span class="n">sample_inv_tmp</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">_output</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-1-alternative-remove-adaptor-through-trimmomatic">
<h2>Step 1 Alternative: Remove adaptor through <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code><a class="headerlink" href="#step-1-alternative-remove-adaptor-through-trimmomatic" title="Permalink to this heading">#</a></h2>
<p>Documentation: <a class="reference external" href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a></p>
<p>We have replaced this with <code class="docutils literal notranslate"><span class="pre">fastp</span></code>, see above, which performs better than <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> in terms of removing adaptors that <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> cannot detect. <code class="docutils literal notranslate"><span class="pre">fastp</span></code> can also automatically guess the adapter sequences from data and by default no adapter sequence is required for input to <code class="docutils literal notranslate"><span class="pre">fastp</span></code>.</p>
<section id="id3">
<h3>Step Inputs<a class="headerlink" href="#id3" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">software_dir</span></code>: directory for the software</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">fasta_with_adapters_etc</span></code>: <strong>filename</strong> for the adapter reference file. According to <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> documention,</p></li>
</ul>
<blockquote>
<div><p>As a rule of thumb newer libraries will use <code class="docutils literal notranslate"><span class="pre">TruSeq3</span></code>, but this really depends on your service provider. If you use FASTQC, the “Overrepresented Sequences” report can help indicate which adapter file is best suited for your data. “Illumina Single End” or “Illumina Paired End” sequences indicate single-end or paired-end <code class="docutils literal notranslate"><span class="pre">TruSeq2</span></code> libraries, and the appropriate adapter files are <code class="docutils literal notranslate"><span class="pre">TruSeq2-SE.fa</span></code> and <code class="docutils literal notranslate"><span class="pre">TruSeq2-PE.fa</span></code> respectively. “TruSeq Universal Adapter” or “TruSeq Adapter, Index …” indicates <code class="docutils literal notranslate"><span class="pre">TruSeq-3</span></code> libraries, and the appropriate adapter files are <code class="docutils literal notranslate"><span class="pre">TruSeq3-SE.fa</span></code> or <code class="docutils literal notranslate"><span class="pre">TruSeq3-PE.fa</span></code>, for single-end and paired-end data respectively. Adapter sequences for <code class="docutils literal notranslate"><span class="pre">TruSeq2</span></code> multiplexed libraries, indicated by “Illumina Multiplexing
…”, and the various RNA library preparations are not currently included.</p>
</div></blockquote>
<p>We have <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> workflow previously defined and executed. Users should decide what fasta adapter reference to use based on <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> results (or their own knowledge).</p>
</section>
<section id="id4">
<h3>Step Outputs<a class="headerlink" href="#id4" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>Two paired <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> file for alignment</p></li>
<li><p>Two unpaired <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code></p></li>
</ul>
<blockquote>
<div><p>For single-ended data, one input and one output file are specified, plus the processing steps. For paired-end data, two input files are specified, and 4 output files, 2 for the ‘paired’ output where both reads survived the processing, and 2 for corresponding ‘unpaired’ output where a read survived, but the partner read did not.</p>
</div></blockquote>
<p><strong>You need to figure out from fastqc results what adapter reference sequence to use.</strong>, eg <code class="docutils literal notranslate"><span class="pre">--fasta_with_adapters_etc</span> <span class="pre">TruSeq3-PE.fa</span></code>. These files can be downloaded from <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> github repo.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[trimmomatic_trim_adaptor]
# Path to the software. Default set to using our rna_quantification.sif image
parameter: trimmomatic_jar = &quot;/opt/Trimmomatic-0.39/trimmomatic-0.39.jar&quot;
# Illumina clip setting
# Path to the reference adaptors
parameter: fasta_with_adapters_etc = path(&quot;.&quot;)
parameter: seed_mismatches = 2
parameter: palindrome_clip_threshold = 30
parameter: simple_clip_threshold = 10
# sliding window setting
parameter: window_size = 4
parameter: required_quality = 20
# Other settings
parameter: leading = 3
parameter: trailing = 3
parameter: min_len = 50
input: fastq, group_by = is_paired_end + 1 , group_with = &quot;sample_id&quot;
output: ([ f&#39;{cwd}/{_sample_id}_paired_{_input[0]:bn}.gz&#39;, f&#39;{cwd}/{_sample_id}_unpaired_{_input[0]:bn}.gz&#39;,  f&#39;{cwd}/{_sample_id}_paired_{_input[1]:bn}.gz&#39;,f&#39;{cwd}/{_sample_id}_unpaired_{_input[1]:bn}.gz&#39; ] if is_paired_end else f&#39;{cwd}/{_sample_id}_trimmed_{_input:bn}.gz&#39;)
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    java -jar -Xmx${java_mem} ${trimmomatic_jar} ${&quot;PE&quot; if is_paired_end else &quot;SE&quot;}  -threads ${numThreads} \
        ${_input:r}  ${_output:r} \
        ILLUMINACLIP:${fasta_with_adapters_etc}:${seed_mismatches}:${palindrome_clip_threshold}:${simple_clip_threshold} \
        LEADING:${leading} TRAILING:${trailing} SLIDINGWINDOW:${window_size}:${required_quality} MINLEN:${min_len}
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-2-alignment-through-star">
<h2>Step 2: Alignment through <code class="docutils literal notranslate"><span class="pre">STAR</span></code><a class="headerlink" href="#step-2-alignment-through-star" title="Permalink to this heading">#</a></h2>
<p>Documentation : <a class="reference external" href="https://github.com/alexdobin/STAR">STAR</a> and <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/rnaseq/src/run_STAR.py">Script in docker</a></p>
<p>This step is the main step for <code class="docutils literal notranslate"><span class="pre">STAR</span></code> alignment.</p>
<p>Originally, the STAR alignment is implemented using the run_STAR.py command from gtex. However, in order to <a class="reference external" href="https://github.com/cumc/xqtl-pipeline/issues/318">accomodate different read length among samples</a>, We reimplement what was included in the wrapper, including the re-alignment of STAR unsorted bam using samtools to avoid high mem consumption.</p>
<section id="id5">
<h3>Step Inputs<a class="headerlink" href="#id5" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>paths to trimmed <code class="docutils literal notranslate"><span class="pre">fastq.gz</span></code> file from Step 1 as documented in the new sample list.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">STAR_index</span></code>: directory for the STAR aligment index</p></li>
</ul>
</section>
<section id="id6">
<h3>Step Outputs<a class="headerlink" href="#id6" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>bam file output <code class="docutils literal notranslate"><span class="pre">${cwd}/{sample_id}.Aligned.sortedByCoord.bam</span></code>, will be used in step 3 and 4</p></li>
<li><p>bam file output <code class="docutils literal notranslate"><span class="pre">${cwd}/{sample_id}.Aligned.toTranscriptome.bam</span></code>, will be used in step 5</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[STAR_align_1]
# Reference gene model
parameter: gtf = path
# STAR indexing file
parameter: STAR_index = path
parameter: outFilterMultimapNmax = 20 
parameter: alignSJoverhangMin = 8 
parameter: alignSJDBoverhangMin = 1 
parameter: outFilterMismatchNmax = 999 
parameter: outFilterMismatchNoverLmax = 0.1
parameter: alignIntronMin = 20 
parameter: alignIntronMax = 1000000 
parameter: alignMatesGapMax = 1000000 
parameter: outFilterType =  &quot;BySJout&quot; 
parameter: outFilterScoreMinOverLread = 0.33 
parameter: outFilterMatchNminOverLread = 0.33 
parameter: limitSjdbInsertNsj = 1200000 
parameter: outSAMstrandField = &quot;intronMotif&quot; 
parameter: outFilterIntronMotifs = &quot;None&quot; 
parameter: alignSoftClipAtReferenceEnds = &quot;Yes&quot; 
parameter: quantMode = [&quot;TranscriptomeSAM&quot;, &quot;GeneCounts&quot;]
parameter: outSAMattrRGline = [&quot;ID:rg1&quot;, &quot;SM:sm1&quot;]
parameter: outSAMattributes = [&quot;NH&quot;, &quot;HI&quot;, &quot;AS&quot;, &quot;nM&quot;, &quot;NM&quot;, &quot;ch&quot;] 
parameter: chimSegmentMin = 15 
parameter: chimJunctionOverhangMin = 15 
parameter: chimOutType = [&quot;Junctions&quot;, &quot;WithinBAM&quot;, &quot;SoftClip&quot;]
parameter: chimMainSegmentMultNmax = 1 
parameter: sjdbOverhang = 100
if int(mem.replace(&quot;G&quot;,&quot;&quot;)) &lt;  40:
    print(&quot;Insufficent memory for STAR, changing to 40G&quot;)
    star_mem = &#39;40G&#39;
else:
    star_mem = mem
# This option is commented out because it will force the downstream analysis to use 40G, which significantlly slow down the process.
input: fastq,group_by = is_paired_end + 1,group_with = {&quot;sample_id&quot;,&quot;read_length&quot;}
output: cord_bam = f&#39;{cwd}/{_sample_id}.Aligned.sortedByCoord.out.bam&#39;,
        trans_bam = f&#39;{cwd}/{_sample_id}.Aligned.toTranscriptome.out.bam&#39;
if _read_length == 0:
    print(&quot;Using specified --sjdbOverhang as read length&quot;)
else:
    print(&quot;Using read length specified in the sample list&quot;)
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = star_mem, cores = numThreads
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    rm -rf ${cwd}/${_sample_id}.*.out.*.gz
    rm -rf ${cwd}/${_sample_id}._STARpass1
    set -e
    touch ${_output[0]:n}.star_start.timestamp
    STAR --runMode alignReads \
        --runThreadN ${numThreads} \
        --genomeDir  ${STAR_index} \
        --readFilesIn  ${_input:r} \
        --readFilesCommand ${&quot;cat&quot; if uncompressed else &quot;zcat&quot;} \
        --outFileNamePrefix ${_output[0]:nnnn}. \
        --outSAMstrandField ${outSAMstrandField} \
        --twopassMode Basic \
        --outFilterMultimapNmax ${outFilterMultimapNmax} \
        --alignSJoverhangMin ${alignSJoverhangMin} \
        --alignSJDBoverhangMin ${alignSJDBoverhangMin} \
        --outFilterMismatchNmax ${outFilterMismatchNmax} \
        --outFilterMismatchNoverLmax ${outFilterMismatchNoverLmax} \
        --alignIntronMin ${alignIntronMin} \
        --alignIntronMax ${alignIntronMax} \
        --alignMatesGapMax ${alignMatesGapMax} \
        --outFilterType ${outFilterType} \
        --outFilterScoreMinOverLread ${outFilterScoreMinOverLread} \
        --outFilterMatchNminOverLread ${outFilterMatchNminOverLread} \
        --limitSjdbInsertNsj ${limitSjdbInsertNsj} \
        --outFilterIntronMotifs ${outFilterIntronMotifs} \
        --alignSoftClipAtReferenceEnds ${alignSoftClipAtReferenceEnds} \
        --quantMode ${&quot; &quot;.join(quantMode)} \
        --outSAMtype BAM Unsorted \
        --outSAMunmapped Within \
        --genomeLoad NoSharedMemory \
        --chimSegmentMin ${chimSegmentMin} \
        --chimJunctionOverhangMin ${chimJunctionOverhangMin} \
        --chimOutType ${&quot; &quot;.join(chimOutType)} \
        --chimMainSegmentMultNmax ${chimMainSegmentMultNmax} \
        --chimOutJunctionFormat 0 \
        --outSAMattributes ${&quot; &quot;.join(outSAMattributes)} \
        --outSAMattrRGline ${&quot; &quot;.join(outSAMattrRGline)} \
        --sjdbOverhang ${sjdbOverhang if _read_length == 0 else _read_length  } \
        --sjdbGTFfile ${gtf} \

    rm -r ${_output[0]:nnnn}._STARgenome
    ${ f&#39;rm -rf {_output[0]:nnnn}._STARtmp&#39; if is_paired_end else &quot;&quot;}
    touch ${_output[0]:n}.sort_start.timestamp
    samtools sort --threads ${numThreads} -o ${_output[0]:nnnn}.Aligned.sortedByCoord.out.bam ${_output[0]:nnnn}.Aligned.out.bam # According to GTEX, this can help reducing the amount of memory consumption.
    rm ${_output[0]:nnnn}.Aligned.out.bam 
    samtools index ${_output[0]}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">strand_detected_1</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s2">&quot;step_strand_detected&quot;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;STAR_align_1&quot;</span><span class="p">)[</span><span class="s2">&quot;trans_bam&quot;</span><span class="p">],</span> <span class="n">group_with</span> <span class="o">=</span> <span class="s2">&quot;sample_id&quot;</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">ReadsPerGene</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">.ReadsPerGene.out.tab&#39;</span> <span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">None</span> <span class="p">)</span>
<span class="n">ReadsPerGene_list</span> <span class="o">=</span> <span class="n">ReadsPerGene</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="mi">3</span><span class="p">::,]</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">numeric_only</span> <span class="o">=</span> <span class="kc">True</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="n">strand_percentage</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span><span class="o">/</span><span class="n">ReadsPerGene_list</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">ReadsPerGene_list</span><span class="p">]</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;for sample </span><span class="si">{</span><span class="n">_sample_id</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="k">if</span> <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.9</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Counts for the 1st read strand aligned with RNA is </span><span class="si">{</span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1">, &gt; 90% of aligned count&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data is likely FR/fr-secondstrand&#39;</span><span class="p">)</span>
    <span class="n">strand_detected</span> <span class="o">=</span> <span class="s2">&quot;fr&quot;</span>
<span class="k">elif</span>  <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">&gt;</span> <span class="mf">0.9</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Counts for the 2nd read strand aligned with RNA is </span><span class="si">{</span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s1">, &gt; 90% of aligned count&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data is likely RF/fr-firststrand&#39;</span><span class="p">)</span>
    <span class="n">strand_detected</span> <span class="o">=</span> <span class="s2">&quot;rf&quot;</span>
<span class="k">elif</span> <span class="nb">max</span><span class="p">(</span> <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span>  <span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span> <span class="o">&lt;</span> <span class="mf">0.6</span><span class="p">:</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Both </span><span class="si">{</span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="si">}</span><span class="s1"> and  </span><span class="si">{</span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span><span class="si">}</span><span class="s1"> are under 60% of reads explained by one direction&#39;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Data is likely unstranded&#39;</span><span class="p">)</span>
    <span class="n">strand_detected</span> <span class="o">=</span> <span class="s2">&quot;unstranded&quot;</span>
<span class="kn">else:</span>
    <span class="n">strand_detected</span> <span class="o">=</span> <span class="s1">&#39;strand_missing&#39;</span> 
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Data does not fall into a likely stranded (max percent explained </span><span class="si">{</span><span class="nb">max</span><span class="p">(</span><span class="w"> </span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="w">  </span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="si">}</span><span class="s1"> &gt; 0.9) or unstranded layout (max percent explained </span><span class="si">{</span><span class="nb">max</span><span class="p">(</span><span class="w"> </span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="w">  </span><span class="n">strand_percentage</span><span class="p">[</span><span class="mi">2</span><span class="p">])</span><span class="si">}</span><span class="s1"> &lt; 0.6), please check your data and manually specified the ``--strand`` option as ``fr``, ``rf`` or ``unstranded``&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">strand_detected_2</span><span class="p">:</span> <span class="kp">shared</span><span class="o">=</span><span class="s2">&quot;strand&quot;</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">parameter:</span> <span class="n">strand</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="k">if</span> <span class="ow">not</span> <span class="n">strand</span><span class="p">:</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">strand_inv</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">strand</span> <span class="o">=</span> <span class="n">strand_inv</span>
        <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">strand</span><span class="p">)):</span>
            <span class="k">if</span> <span class="n">strand</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">==</span> <span class="s2">&quot;strand_missing&quot;</span><span class="p">:</span>
                <span class="n">strand</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="n">step_strand_detected</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Using strand specified in the input samples list </span><span class="si">{</span><span class="n">strand</span><span class="si">}</span><span class="s1">, replacing strand_missing with detected strand&#39;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">warn_if</span><span class="p">(</span><span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">x</span> <span class="ow">is</span> <span class="n">step_strand_detected</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">step_strand_detected</span><span class="p">),</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;strands detected are different among samples, please check your protocol, we will use the detected strand for each samples&quot;</span><span class="p">)</span>    
        <span class="n">strand</span> <span class="o">=</span> <span class="n">step_strand_detected</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Using detected strand for each samples </span><span class="si">{</span><span class="n">strand</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="n">stop_if</span><span class="p">(</span><span class="n">strand</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s2">&quot;fr&quot;</span><span class="p">,</span> <span class="s2">&quot;rf&quot;</span><span class="p">,</span> <span class="s2">&quot;unstranded&quot;</span><span class="p">],</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;``--strand`` option should be ``fr``, ``rf`` or ``unstranded``&quot;</span><span class="p">)</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;Using ``--strand`` overwrite option for all the samples </span><span class="si">{</span><span class="n">strand</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span> 
    <span class="n">strand</span> <span class="o">=</span> <span class="p">[</span><span class="n">strand</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">step_strand_detected</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-3-mark-duplicates-reads-qc-through-picard">
<h2>Step 3: Mark duplicates reads &amp; QC through <code class="docutils literal notranslate"><span class="pre">Picard</span></code><a class="headerlink" href="#step-3-mark-duplicates-reads-qc-through-picard" title="Permalink to this heading">#</a></h2>
<p>This step is the first QC step after <code class="docutils literal notranslate"><span class="pre">STAR</span></code> alignment. This step will performed QC to collect multipe metrics regarding the RNASeq using Picard. Then it will also generate a new <code class="docutils literal notranslate"><span class="pre">.bam</span></code> file with duplication marked with the hexadecimal value of <code class="docutils literal notranslate"><span class="pre">0x0400</span></code>, which corresponds to a decimal value of 1024</p>
<section id="id7">
<h3>Step Inputs:<a class="headerlink" href="#id7" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">STAR_bam</span></code>: path to the output in Step 2.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">reference_fasta</span></code>:  The fasta reference file used to generate star index, it is critical that the this fasta is <em>exactly</em> the same as those that generate the STAR Index, else this error will occurs: <a class="github reference external" href="https://github.com/cumc/xqtl-pipeline/issues/357">cumc/xqtl-pipeline#357</a></p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">RefFlat</span> <span class="pre">file</span></code></p></li>
</ul>
<p>This file is needed for picard CollectRnaSeqMetrics module, which in turn</p>
<blockquote>
<div><p>produces metrics describing the distribution of the bases within the transcripts. It calculates the total numbers and the fractions of nucleotides within specific genomic regions including untranslated regions (UTRs), introns, intergenic sequences (between discrete genes), and peptide-coding sequences (exons). This tool also determines the numbers of bases that pass quality filters that are specific to Illumina data (PF_BASES).</p>
</div></blockquote>
<p>The refFlat file can be generated by the reference_data module, RefFlat_generation step.</p>
</section>
<section id="id8">
<h3>Step Outputs:<a class="headerlink" href="#id8" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p>A collection of metrics file for each of the samples</p></li>
<li><p>A new <code class="docutils literal notranslate"><span class="pre">.bam</span></code> file with duplication marked with the hexadecimal value of <code class="docutils literal notranslate"><span class="pre">0x0400</span></code>, which corresponds to a decimal value of 1024</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[picard_qc]
# Reference gene model
parameter: gtf = path
depends: sos_variable(&#39;strand&#39;)
# Path to flat reference file, for computing QC metric
parameter: ref_flat = path
# Path to the software. Default set to using our rna_quantification.sif image
parameter: picard_jar = &quot;/opt/picard-tools/picard.jar&quot;
# The fasta reference file used to generate star index
parameter: reference_fasta = path
# For the patterned flowcell models (HiSeq X), change to 2500
parameter: optical_distance = 100
picard_strand_dict = {&quot;rf&quot;:&quot;SECOND_READ_TRANSCRIPTION_STRAND&quot;,&quot;fr&quot;: &quot;FIRST_READ_TRANSCRIPTION_STRAND&quot;,&quot;unstranded&quot;:&quot;NONE&quot; }
input: output_from(&quot;STAR_align_1&quot;),group_by = 2, group_with = {&quot;sample_id&quot;,&quot;strand&quot;}
output: picard_metrics = f&#39;{cwd}/{_sample_id}.alignment_summary_metrics&#39;,
        picard_rna_metrics = f&#39;{cwd}/{_sample_id}.rna_metrics&#39;,
        md_bam = f&#39;{cwd}/{_sample_id}.Aligned.sortedByCoord.out.md.bam&#39;,
        md_metrics = f&#39;{cwd}/{_sample_id}.Aligned.sortedByCoord.md.metrics&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
        set -e
        touch ${_output[0]:n}.CollectMultipleMetrics_start.timestamp
        java -jar -Xmx${java_mem} ${picard_jar} CollectMultipleMetrics \
            -REFERENCE_SEQUENCE ${reference_fasta} \
            -PROGRAM CollectAlignmentSummaryMetrics \
            -PROGRAM CollectInsertSizeMetrics \
            -PROGRAM QualityScoreDistribution \
            -PROGRAM MeanQualityByCycle \
            -PROGRAM CollectBaseDistributionByCycle \
            -PROGRAM CollectGcBiasMetrics \
            -VALIDATION_STRINGENCY STRICT \
            -INPUT  ${_input[&quot;cord_bam&quot;]} \
            -OUTPUT  ${_output[0]:n}

bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[-1]:n}.stderr&#39;, stdout = f&#39;{_output[-1]:n}.stdout&#39;, entrypoint=entrypoint
        set -e
        touch ${_output[2]:n}.MarkDuplicates_start.timestamp
        java -Xmx${java_mem} -jar ${picard_jar} MarkDuplicates \
            -I ${_input[&quot;cord_bam&quot;]}  \
            -O ${_output[2]} \
            -PROGRAM_RECORD_ID null \
            -M ${_output[3]} \
            -TMP_DIR ${cwd}\
            -MAX_RECORDS_IN_RAM 500000 -SORTING_COLLECTION_SIZE_RATIO 0.25 \
            -ASSUME_SORT_ORDER coordinate \
            -TAGGING_POLICY DontTag \
            -OPTICAL_DUPLICATE_PIXEL_DISTANCE ${optical_distance} \
            -CREATE_INDEX true \
            -CREATE_MD5_FILE true \
            -VALIDATION_STRINGENCY STRICT \
            -REMOVE_SEQUENCING_DUPLICATES false \
            -REMOVE_DUPLICATES false 

bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.rna_metrics.stderr&#39;, stdout = f&#39;{_output[0]:n}.rna_metrics.stderr&#39;, entrypoint=entrypoint
        set -e
        # Get only line with rRNA and transcript_id
        cat ${gtf}| grep rRNA | grep transcript_id &gt; ${gtf}.tmp.${_input[&quot;cord_bam&quot;]:bnnnn}  # To Avoid data racing problem
        samtools view -H ${_input[&quot;cord_bam&quot;]}  &gt; ${_input[&quot;cord_bam&quot;]}.RI  

python: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.rna_metrics.stderr&#39;, stdout = f&#39;{_output[0]:n}.rna_metrics.stderr&#39;, entrypoint=entrypoint
        import pandas as pd
        from collections import defaultdict
        chrom = []
        start = []
        end = []
        strand = []
        tag = []
        annotation_gtf = &quot;${gtf}.tmp.${_input[&quot;cord_bam&quot;]:bnnnn}&quot;
        with open(annotation_gtf, &#39;r&#39;) as gtf:
            for row in gtf:
                row = row.strip().split(&#39;\t&#39;)
                if row[0][0]==&#39;#&#39; or row[2]!=&quot;transcript&quot;: continue # skip header
                chrom.append(row[0])
                start.append(row[3])
                end.append(row[4])
                strand.append(row[6])
                attributes = defaultdict()
                for a in row[8].replace(&#39;&quot;&#39;, &#39;&#39;).split(&#39;;&#39;)[:-1]:
                    kv = a.strip().split(&#39; &#39;)
                    if kv[0]!=&#39;tag&#39;:
                        attributes[kv[0]] = kv[1]
                    else:
                        attributes.setdefault(&#39;tags&#39;, []).append(kv[1])
                tag.append(attributes)
        transcript_id = [x[&quot;transcript_id&quot;] for x in tag]
        RI = pd.DataFrame(data={&#39;chr&#39;:chrom, &#39;start&#39;:start, &#39;end&#39;:end, &#39;strand&#39;:strand,&#39;transcript_id&#39; : transcript_id })
        RI.to_csv(&quot;${_input[&quot;cord_bam&quot;]}.RI&quot;, index = 0, header = 0, mode = &quot;a&quot;,sep = &quot;\t&quot; )

bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.rna_metrics.stderr&#39;, stdout = f&#39;{_output[0]:n}.rna_metrics.stdout&#39;, entrypoint=entrypoint
        set -e
        touch ${_output[0]:n}.CollectRnaSeqMetrics_start.timestamp
        java -jar -Xmx${java_mem} ${picard_jar} CollectRnaSeqMetrics \
            -REF_FLAT ${ref_flat} \
            -RIBOSOMAL_INTERVALS ${_input[&quot;cord_bam&quot;]}.RI \
            -STRAND_SPECIFICITY ${picard_strand_dict[_strand]} \
            -CHART_OUTPUT ${_output[0]:n}.rna_metrics.pdf \
            -VALIDATION_STRINGENCY STRICT \
            -INPUT ${_input[&quot;cord_bam&quot;]}  \
            -OUTPUT  ${_output[0]:n}.rna_metrics
        rm ${gtf}.tmp.${_input[&quot;cord_bam&quot;]:bnnnn}

_input[&quot;cord_bam&quot;].zap()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">STAR_output</span><span class="p">]</span>
<span class="kn">input:</span> <span class="n">output_from</span><span class="p">(</span><span class="s2">&quot;picard_qc&quot;</span><span class="p">),</span>  <span class="n">group_by</span> <span class="o">=</span> <span class="s2">&quot;all&quot;</span>
<span class="kn">depends:</span> <span class="n">sos_variable</span><span class="p">(</span><span class="s2">&quot;strand&quot;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">samples</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">_bam_list&#39;</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">coord_bam_list</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">b</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">4</span><span class="p">]]</span>
<span class="n">SJ_list</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">bnnnnn</span><span class="si">}</span><span class="s1">.SJ.out.tab&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">4</span><span class="p">]]</span>
<span class="n">Trans_bam_list</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">bnnnn</span><span class="si">}</span><span class="s1">.toTranscriptome.out.bam&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">[</span><span class="mi">2</span><span class="p">::</span><span class="mi">4</span><span class="p">]]</span>
<span class="nb">print</span><span class="p">(</span><span class="n">SJ_list</span><span class="p">)</span>
<span class="n">out</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">({</span><span class="s2">&quot;sample_id&quot;</span> <span class="p">:</span> <span class="n">sample_id</span><span class="p">,</span><span class="s2">&quot;strand&quot;</span> <span class="p">:</span> <span class="n">strand</span> <span class="p">,</span> <span class="s2">&quot;coord_bam_list&quot;</span> <span class="p">:</span> <span class="n">coord_bam_list</span><span class="p">,</span> <span class="s2">&quot;SJ_list&quot;</span> <span class="p">:</span><span class="n">SJ_list</span><span class="p">,</span><span class="s2">&quot;trans_bam_list&quot;</span> <span class="p">:</span> <span class="n">Trans_bam_list</span>  <span class="p">})</span>
<span class="n">out</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">_output</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-4-post-aligment-qc-through-rna-seqc">
<h2>Step 4: Post aligment QC through <code class="docutils literal notranslate"><span class="pre">RNA-SeQC</span></code><a class="headerlink" href="#step-4-post-aligment-qc-through-rna-seqc" title="Permalink to this heading">#</a></h2>
<p>Documentation : <a class="reference external" href="https://github.com/getzlab/rnaseqc">RNA-SeQC</a> and <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/rnaseq/src/run_rnaseqc.py">Script in docker</a></p>
<p>This step is second QC step after <code class="docutils literal notranslate"><span class="pre">STAR</span></code> alignment. It will perform RNA-seq quantification as well.</p>
<section id="id9">
<h3>Step Inputs<a class="headerlink" href="#id9" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">bam_list</span></code>: path to the output of STAR_output, which outlined the strands and bam file used for each samples.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">gtf</span></code>: reference genome <code class="docutils literal notranslate"><span class="pre">.gtf</span></code> file, this gtf file need to have the same chr name format as the index used to generate the bam file and must be on collapsed gene gtf</p></li>
</ul>
</section>
<section id="id10">
<h3>Step Outputs<a class="headerlink" href="#id10" title="Permalink to this heading">#</a></h3>
<p>The following output files are generated in the output directory you provide:</p>
<ul class="simple">
<li><p>{sample}.metrics.tsv : A tab-delimited list of (Statistic, Value) pairs of all statistics and metrics recorded.</p></li>
<li><p>{sample}.exon_reads.gct : A tab-delimited GCT file with (Exon ID, Gene Name, coverage) tuples for all exons which had at least part of one read mapped.</p></li>
<li><p>{sample}.gene_reads.gct : A tab-delimited GCT file with (Gene ID, Gene Name, coverage) tuples for all genes which had at least one read map to at least one of its exons</p></li>
<li><p>{sample}.gene_tpm.gct : A tab-delimited GCT file with (Gene ID, Gene Name, TPM) tuples for all genes reported in the gene_reads.gct file. Note: this file is renamed to .gene_rpkm.gct if the –rpkm flag is present.</p></li>
<li><p>{sample}.fragmentSizes.txt : A list of fragment sizes recorded, if a BED file was provided</p></li>
<li><p>{sample}.coverage.tsv : A tab-delimited list of (Gene ID, Transcript ID, Mean Coverage, Coverage Std, Coverage CV) tuples for all transcripts encountered in the GTF.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[rnaseqc_call_1]
import os
import pandas as pd
parameter: bam_list = path
# Reference gene model
parameter: gtf = path
sample_inv = pd.read_csv(bam_list,sep = &quot;\t&quot;)
parameter: detection_threshold = 5
parameter: mapping_quality = 255
## Extract strand information if user have specified the strand
strand_list = sample_inv.strand.values.tolist()
stop_if(not all([x in [&quot;fr&quot;, &quot;rf&quot;, &quot;unstranded&quot;] for x in strand_list ]), msg = &quot;strand columns should only include ``fr``, ``rf`` or ``unstranded``, please check the bam_list&quot;)     
## Extract sample_id
sample_id = sample_inv.sample_id.values.tolist()
    
## Get the file name for cood_bam data
coord_bam_list_inv = sample_inv.coord_bam_list.values.tolist()
coord_bam_list = [f&#39;{cwd}/{x}&#39; for x in coord_bam_list_inv ]
input:  coord_bam_list, group_by = 1, group_with = {&quot;sample_id&quot;,&quot;strand_list&quot;}
output: f&#39;{cwd}/{_sample_id}.rnaseqc.gene_tpm.gct.gz&#39;,
        f&#39;{cwd}/{_sample_id}.rnaseqc.gene_reads.gct.gz&#39;,
        f&#39;{cwd}/{_sample_id}.rnaseqc.exon_reads.gct.gz&#39;,
        f&#39;{cwd}/{_sample_id}.rnaseqc.metrics.tsv&#39;

task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    cd ${cwd} &amp;&amp; \
    rnaseqc \
        ${gtf:a} \
        ${_input:a} \
        ${_output[0]:d}  \
        ${(&quot;--stranded &quot; + _strand_list) if _strand_list != &quot;unstranded&quot; else &quot;&quot;} \
        ${f&#39;--detection-threshold {detection_threshold}&#39;} ${f&#39;--mapping-quality {mapping_quality}&#39;} ${ &#39;&#39; if is_paired_end else &#39;-u&#39;}

bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    mv ${_output[0]:d}/${_input:b}.gene_tpm.gct ${_output[0]:n}
    mv ${_output[0]:d}/${_input:b}.gene_reads.gct ${_output[1]:n}
    mv ${_output[0]:d}/${_input:b}.exon_reads.gct ${_output[2]:n}
    mv ${_output[0]:d}/${_input:b}.metrics.tsv ${_output[3]}
    gzip ${_output[0]:n} ${_output[1]:n} ${_output[2]:n}
</pre></div>
</div>
</div>
</div>
<p>The RNASEQC results were merged in the following step,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[rnaseqc_call_2]
input: group_by = &quot;all&quot;
output: f&#39;{cwd}/{samples:bn}.rnaseqc.gene_tpm.gct.gz&#39;,
        f&#39;{cwd}/{samples:bn}.rnaseqc.gene_readsCount.gct.gz&#39;,
        f&#39;{cwd}/{samples:bn}.rnaseqc.exon_readsCount.gct.gz&#39;,
        f&#39;{cwd}/{samples:bn}.rnaseqc.metrics.tsv&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads
python: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    import pandas as pd
    import os
    def make_gct(gct_path):
        # sample_name
        sample_name = &quot;.&quot;.join(os.path.basename(gct_path).split(&quot;.&quot;)[:-4])
        # read_input
        pre_gct = pd.read_csv(gct_path,sep = &quot;\t&quot;,
                              skiprows= 2,index_col=&quot;Name&quot;).drop(&quot;Description&quot;,axis = 1)
        pre_gct.index.name = &quot;gene_ID&quot;
        pre_gct.columns = [sample_name]
        return(pre_gct)

    def merge_gct(gct_path_list):
        gct = pd.DataFrame()
        for gct_path in gct_path_list:
            #check duplicated indels and remove them.
            gct_col = make_gct(gct_path)
            gct = gct.merge(gct_col,right_index=True,left_index = True,how = &quot;outer&quot;)
        return gct

    input_list = [${_input:r,}]
    tpm_list = input_list[0::4]
    gc_list = input_list[1::4]
    ec_list = input_list[2::4]
    gct_path_list_list = [tpm_list,gc_list,ec_list]
    output_path = [${_output:r,}][0:3]
    for i in range(0,len(output_path)):
        output = merge_gct(gct_path_list_list[i])
        output.to_csv(output_path[i], sep = &quot;\t&quot;)
    metrics_list = input_list[3::4]
    with open(&quot;${cwd}/${samples:bn}.rnaseqc.metrics_output_list&quot;, &quot;w&quot;) as f:
        f.write(&#39;\n&#39;.join(metrics_list))

bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    aggregate_rnaseqc_metrics.py  ${_output[3]:n}_output_list ${_output[3]:nn}
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-6-quantify-expression-through-rsem">
<h2>Step 6: Quantify expression through <code class="docutils literal notranslate"><span class="pre">RSEM</span></code><a class="headerlink" href="#step-6-quantify-expression-through-rsem" title="Permalink to this heading">#</a></h2>
<p>Documentation : <a class="reference external" href="https://deweylab.github.io/RSEM/rsem-calculate-expression.html">RSEM</a> and <a class="reference external" href="https://github.com/broadinstitute/gtex-pipeline/blob/master/rnaseq/src/run_RSEM.py">Script in docker</a></p>
<p>This step generate the expression matrix from STAR output. Estimate gene and isoform expression from RNA-Seq data are generated.</p>
<section id="step-input">
<h3>Step Input<a class="headerlink" href="#step-input" title="Permalink to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">bam_list</span></code>: path to the output of STAR_output, which outlined the strands and bam file used for each samples.</p></li>
<li><p>transcript-level BAM file: path to the output of Step 3.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">RSEM_index</span></code>: path to RSEM index</p></li>
</ul>
</section>
<section id="id11">
<h3>Step Outputs<a class="headerlink" href="#id11" title="Permalink to this heading">#</a></h3>
<p>Please see the output section of <a class="reference external" href="https://deweylab.github.io/RSEM/rsem-calculate-expression.html">https://deweylab.github.io/RSEM/rsem-calculate-expression.html</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[rsem_call_1]
parameter: RSEM_index = path
parameter: max_frag_len = 1000
estimate_rspd = True

parameter: bam_list = path

sample_inv = pd.read_csv(bam_list,sep = &quot;\t&quot;)

## Extract strand information if user have specified the strand
strand_list = sample_inv.strand.values.tolist()
stop_if(not all([x in [&quot;fr&quot;, &quot;rf&quot;, &quot;unstranded&quot;] for x in strand_list ]), msg = &quot;strand columns should only include ``fr``, ``rf`` or ``unstranded``, please check the bam_list&quot;)     
## Extract sample_id
sample_id = sample_inv.sample_id.values.tolist()
    
## Get the file name for trans_bam_list data
trans_bam_list_inv = sample_inv.trans_bam_list.values.tolist()
trans_bam_list = [f&#39;{cwd}/{x}&#39; for x in trans_bam_list_inv ]
input: trans_bam_list ,  group_by = 1, group_with = {&quot;sample_id&quot;,&quot;strand_list&quot;} 
output: f&#39;{cwd}/{_sample_id}.rsem.isoforms.results&#39;, f&#39;{cwd}/{_sample_id}.rsem.genes.results&#39;,f&#39;{cwd}/{_sample_id}.rsem.stat/{_sample_id}.rsem.cnt&#39;
task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, trunk_size = job_size
bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    run_RSEM.py ${RSEM_index:a} ${_input:a} ${_sample_id} \
        -o ${_output[0]:d} \
        --max_frag_len ${max_frag_len} \
        --estimate_rspd ${&#39;true&#39; if estimate_rspd else &#39;false&#39;} \
        --paired_end ${&quot;true&quot; if is_paired_end else &quot;false&quot;} \
        --is_stranded ${&quot;true&quot; if _strand_list != &quot;unstranded&quot; else &quot;false&quot;} \
        --threads ${numThreads}
</pre></div>
</div>
</div>
</div>
<p>The RSEM results were merged in the following steps, seven files (four for each columns in the isoform output and 3 for each of the genes output) will be generated.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[rsem_call_2]
input: group_by = &quot;all&quot;
output: f&#39;{cwd}/{samples:bn}.rsem_transcripts_expected_count.txt.gz&#39;,
        f&#39;{cwd}/{samples:bn}.rsem_transcripts_tpm.txt.gz&#39;,
        f&#39;{cwd}/{samples:bn}.rsem_transcripts_fpkm.txt.gz&#39;,
        f&#39;{cwd}/{samples:bn}.rsem_transcripts_isopct.txt.gz&#39;,
        f&#39;{cwd}/{samples:bn}.rsem_genes_expected_count.txt.gz&#39;,
        f&#39;{cwd}/{samples:bn}.rsem_genes_tpm.txt.gz&#39;,
        f&#39;{cwd}/{samples:bn}.rsem_genes_fpkm.txt.gz&#39;,
        f&#39;{cwd}/{samples:bn}.rsem.aggregated_quality.metrics.tsv&#39;

task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, trunk_size = job_size
python: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint=entrypoint
    input_list = [${_input:r,}]
    with open(&#39;${cwd}/${samples:bn}.rsem.isoforms_output_list&#39;, &quot;w&quot;) as f:
        f.write(&#39;\n&#39;.join(input_list[0::3]))
    with open(&#39;${cwd}/${samples:bn}.rsem.genes_output_list&#39;, &quot;w&quot;) as f:
        f.write(&#39;\n&#39;.join(input_list[1::3]))

bash: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;
         aggregate_rsem_results.py ${cwd}/${samples:bn}.rsem.isoforms_output_list {expected_count,TPM,FPKM,IsoPct} ${_output[0]:nnn}
         aggregate_rsem_results.py ${cwd}/${samples:bn}.rsem.genes_output_list {expected_count,TPM,FPKM} ${_output[1]:nnn} 

R:  container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[-1]:n}.stderr&#39;, stdout = f&#39;{_output[-1]:n}.stdout&#39;, entrypoint=entrypoint
     readRSEM.cnt &lt;- function (source) {
            # RSEM .cnt files gives statistics about the (transcriptome) alignment passed to RSEM:
            # Row 1: N0 (# unalignable reads);
            #        N1 (# alignable reads);
            #        N2 (# filtered reads due to too many alignments);
            #        N_tot (N0+N1+N2)
            # Row 2: nUnique (# reads aligned uniquely to a gene);
            #        nMulti (# reads aligned to multiple genes);
            #        nUncertain (# reads aligned to multiple locations in the given reference sequences, which include isoform-level multi-mapping reads)
            # Row 3: nHits (# total alignments);
            #        read_type (0: single-end read, no quality; 1: single-end read, with quality score; 2: paired-end read, no quality score; 3: paired-end read, with quality score)
            # Source: https://groups.google.com/forum/#!topic/rsem-users/usmPKgsC5LU
            # Note: N1 = nUnique + nMulti

            stopifnot(file.exists(source[1]))
            isDir &lt;- file.info(source)$isdir
            if (isDir) {
                files &lt;- system(paste(&quot;find&quot;, source, &quot;-name \&quot;*.rsem.cnt\&quot;&quot;), intern=TRUE)
                stopifnot(length(files) &gt; 0)
                samples &lt;- gsub(&quot;_rsem.cnt&quot;, &quot;&quot;, basename(files), fixed=TRUE)
            } else {
                files &lt;- source
                samples &lt;- gsub(&quot;.rsem.cnt&quot;, &quot;&quot;, basename(files), fixed=TRUE)
            }
            metrics &lt;- list()
            for (i in 1:length(files)) {
                m &lt;- read.table(files[i], header=FALSE, sep=&quot; &quot;, comment.char=&quot;#&quot;, stringsAsFactors=FALSE, nrows=3, fill=TRUE)
                metrics[[i]] &lt;- data.frame(Sample=samples[i],
                File=files[i],
                TotalReads=m[1, 4],
                AlignedReads=m[1, 2],
                UniquelyAlignedReads=m[2, 1],
                stringsAsFactors=FALSE)
            }
            metrics &lt;- do.call(rbind, metrics)
            row.names(metrics) &lt;- metrics$Sample

            return(metrics)
        }
        sourceRSEM = c(${_input:r,})
        sourceRSEM = sourceRSEM[seq(3,length(sourceRSEM),3)]
        metrics.RSEM = readRSEM.cnt(sourceRSEM)
        write.table(metrics.RSEM, file=&quot;${_output[-1]}&quot;,col.names=TRUE, row.names=FALSE, quote=FALSE)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-7-summarize-with-multiqc">
<h2>Step 7: summarize with MultiQC<a class="headerlink" href="#step-7-summarize-with-multiqc" title="Permalink to this heading">#</a></h2>
<p><a class="reference external" href="https://multiqc.info/docs/#using-multiqc">MultiQC</a></p>
<blockquote>
<div><p>MultiQC is a reporting tool that parses summary statistics from results and log files generated by other bioinformatics tools. MultiQC doesn’t run other tools for you - it’s designed to be placed at the end of analysis pipelines or to be run manually when you’ve finished running your tools. When you launch MultiQC, it recursively searches through any provided file paths and finds files that it recognises. It parses relevant information from these and generates a single stand-alone HTML report file. It also saves a directory of data files with all parsed data for further downstream use.</p>
</div></blockquote>
<p>MultiQC will automatically generate QC report for anything embedded within the given directory. Therefore providing the directory containing all the output will surfice.</p>
<p>The output of MultiQC is a multi-module report each corresponding to the quality report of each step of analysis previously performed.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[rsem_call_3,rnaseqc_call_3]
output: f&#39;{cwd}/{samples:bn}.multiqc_report.html&#39;
task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, trunk_size = job_size
report: output = f&quot;{_output:n}.multiqc_config.yml&quot;
  extra_fn_clean_exts:
      - &#39;_rsem&#39;
  fn_ignore_dirs:
      - &#39;*_STARpass1&#39;
bash:  container=container,expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    multiqc ${_input:d} -v -n ${_output:b} -o ${_output:d} -c ${_output:n}.multiqc_config.yml
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[rsem_call_4, rnaseqc_call_4]
# Path to flat reference file, for computing QC metric
output: f&#39;{cwd}/{samples:b}.picard.aggregated_quality.metrics.tsv&#39;
task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, trunk_size = job_size
R: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    ## Define Function
    readPicard.alignment_summary_metrics &lt;- function (source) {
      stopifnot(length(source) == 1)
      stopifnot(file.exists(source))
      isDir &lt;- file.info(source)$isdir
      if (isDir) {
        files &lt;- system(paste(&quot;find -L&quot;, source, &quot;-name \&quot;*.alignment_summary_metrics\&quot;&quot;), intern=TRUE)
        stopifnot(length(files) &gt; 0)
        samples &lt;- gsub(&quot;.alignment_summary_metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      } else {
        files &lt;- source
        samples &lt;- gsub(&quot;.alignment_summary_metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      }
    
      metrics &lt;- list()
      for (i in 1:length(files)) {
        m &lt;- read.table(files[i], header=TRUE, sep=&quot;\t&quot;, comment.char=&quot;#&quot;, stringsAsFactors=FALSE, nrows=${is_paired_end}+1)
        metrics[[i]] &lt;- data.frame(Sample=samples[i],
                                   File=files[i],
                                   PF_READS=sum(m$PF_READS[1:2]),
                                   PF_READS_ALIGNED=sum(m$PF_READS_ALIGNED[1:2]),
                                   PCT_PF_READS_ALIGNED=sum(m$PF_READS_ALIGNED[1:2])/sum(m$PF_READS[1:2]),
                                   stringsAsFactors=FALSE)
      }
      metrics &lt;- do.call(rbind, metrics)
      row.names(metrics) &lt;- metrics$Sample
    
      return(metrics)
    }

    readPicard.rna_metrics &lt;- function(source) {

      stopifnot(length(source) == 1)
      stopifnot(file.exists(source))
      isDir &lt;- file.info(source)$isdir
      if (isDir) {
        files &lt;- system(paste(&quot;find -L&quot;, source, &quot;-name \&quot;*.rna_metrics\&quot;&quot;), intern=TRUE)
        stopifnot(length(files) &gt; 0)
        samples &lt;- gsub(&quot;.rna_metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      } else {
        files &lt;- source
        samples &lt;- gsub(&quot;.rna_metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      }

      metrics &lt;- list()
      for (i in 1:length(files)) {
        m &lt;- read.table(files[i], header=TRUE, sep=&quot;\t&quot;, comment.char=&quot;#&quot;, stringsAsFactors=FALSE, nrows=1)
        metrics[[i]] &lt;- data.frame(Sample=samples[i],
                                   File=files[i],
                                   PCT_RIBOSOMAL_BASES=m$PCT_RIBOSOMAL_BASES,
                                   PCT_CODING_BASES=m$PCT_CODING_BASES,
                                   PCT_UTR_BASES=m$PCT_UTR_BASES,
                                   PCT_INTRONIC_BASES=m$PCT_INTRONIC_BASES,
                                   PCT_INTERGENIC_BASES=m$PCT_INTERGENIC_BASES,
                                   PCT_MRNA_BASES=m$PCT_MRNA_BASES,
                                   PCT_USABLE_BASES=m$PCT_USABLE_BASES,
                                   MEDIAN_CV_COVERAGE=m$MEDIAN_CV_COVERAGE,
                                   MEDIAN_5PRIME_BIAS=m$MEDIAN_5PRIME_BIAS,
                                   MEDIAN_3PRIME_BIAS=m$MEDIAN_3PRIME_BIAS,
                                   MEDIAN_5PRIME_TO_3PRIME_BIAS=m$MEDIAN_5PRIME_TO_3PRIME_BIAS,
                                   stringsAsFactors=FALSE)
      }
      metrics &lt;- do.call(rbind, metrics)
      row.names(metrics) &lt;- metrics$Sample
    
      return(metrics)
    }


    readPicard.duplicate_metrics &lt;- function(source) {

      stopifnot(length(source) == 1)
      stopifnot(file.exists(source))
      isDir &lt;- file.info(source)$isdir
      if (isDir) {
        files &lt;- system(paste(&quot;find -L&quot;, source, &quot;-name \&quot;*.Aligned.sortedByCoord.md.metrics\&quot;&quot;), intern=TRUE)
        stopifnot(length(files) &gt; 0)
        samples &lt;- gsub(&quot;.Aligned.sortedByCoord.md.metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      } else {
        files &lt;- source
        samples &lt;- gsub(&quot;.Aligned.sortedByCoord.md.metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      }

      metrics &lt;- list()
      for (i in 1:length(files)) {
        m &lt;- read.table(files[i], header=TRUE, sep=&quot;\t&quot;, comment.char=&quot;#&quot;, stringsAsFactors=FALSE, nrows=1)
        metrics[[i]] &lt;- data.frame(Sample=samples[i],
                                   File=files[i],
                                   PERCENT_DUPLICATION=m$PERCENT_DUPLICATION,
                                   ESTIMATED_LIBRARY_SIZE=m$ESTIMATED_LIBRARY_SIZE,
                                   stringsAsFactors=FALSE)
      }
      metrics &lt;- do.call(rbind, metrics)
      row.names(metrics) &lt;- metrics$Sample
    
      return(metrics)
    }

    readPicard.wgs_metrics &lt;- function (source) {

      stopifnot(length(source) == 1)
      stopifnot(file.exists(source))
      isDir &lt;- file.info(source)$isdir
      if (isDir) {
        files &lt;- system(paste(&quot;find -L&quot;, source, &quot;-name \&quot;*.wgs_metrics\&quot;&quot;), intern=TRUE)
        stopifnot(length(files) &gt; 0)
        samples &lt;- gsub(&quot;.wgs_metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      } else {
        files &lt;- source
        samples &lt;- gsub(&quot;.wgs_metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      }

      metrics &lt;- list()
      for (i in 1:length(files)) {
        m &lt;- read.table(files[i], header=TRUE, sep=&quot;\t&quot;, comment.char=&quot;#&quot;, stringsAsFactors=FALSE, nrows=1)
        metrics[[i]] &lt;- data.frame(Sample=samples[i],
                                   File=files[i],
                                   MEDIAN_COVERAGE=m$MEDIAN_COVERAGE,
                                   MAD_COVERAGE=m$MAD_COVERAGE,
                                   PCT_EXC_DUPE=m$PCT_EXC_DUPE,
                                   PCT_EXC_TOTAL=m$PCT_EXC_TOTAL,
                                   stringsAsFactors=FALSE)
      }
      metrics &lt;- do.call(rbind, metrics)
      row.names(metrics) &lt;- metrics$Sample

      return(metrics)
    }


    readPicard.insert_size_metrics &lt;- function (source) {

      stopifnot(length(source) == 1)
      stopifnot(file.exists(source))
      isDir &lt;- file.info(source)$isdir
      if (isDir) {
        files &lt;- system(paste(&quot;find -L&quot;, source, &quot;-name \&quot;*.insert_size_metrics\&quot;&quot;), intern=TRUE)
        stopifnot(length(files) &gt; 0)
        samples &lt;- gsub(&quot;.insert_size_metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      } else {
        files &lt;- source
        samples &lt;- gsub(&quot;.insert_size_metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      }

      metrics &lt;- list()
      for (i in 1:length(files)) {
        m &lt;- read.table(files[i], header=TRUE, sep=&quot;\t&quot;, comment.char=&quot;#&quot;, stringsAsFactors=FALSE, nrows=1)
        metrics[[i]] &lt;- data.frame(Sample=samples[i],
                                   File=files[i],
                                   MEDIAN_INSERT_SIZE=m$MEDIAN_INSERT_SIZE,
                                   MODE_INSERT_SIZE=m$MODE_INSERT_SIZE,
                                   MEDIAN_ABSOLUTE_DEVIATION=m$MEDIAN_ABSOLUTE_DEVIATION,
                                   stringsAsFactors=FALSE)
      }
      metrics &lt;- do.call(rbind, metrics)
      row.names(metrics) &lt;- metrics$Sample
    
      return(metrics)
    }

    readPicard.gc_bias.summary_metrics &lt;- function (source) {

      stopifnot(length(source) == 1)
      stopifnot(file.exists(source))
      isDir &lt;- file.info(source)$isdir
      if (isDir) {
        files &lt;- system(paste(&quot;find -L&quot;, source, &quot;-name \&quot;*.gc_bias.summary_metrics\&quot;&quot;), intern=TRUE)
        stopifnot(length(files) &gt; 0)
        samples &lt;- gsub(&quot;.gc_bias.summary_metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      } else {
        files &lt;- source
        samples &lt;- gsub(&quot;.gc_bias.summary_metrics&quot;, &quot;&quot;, basename(files), fixed=TRUE)
      }

      metrics &lt;- list()
      for (i in 1:length(files)) {
        m &lt;- read.table(files[i], header=TRUE, sep=&quot;\t&quot;, comment.char=&quot;#&quot;, stringsAsFactors=FALSE, nrows=1)
        metrics[[i]] &lt;- data.frame(Sample=samples[i],
                                   File=files[i],
                                   AT_DROPOUT=m$AT_DROPOUT,
                                   GC_DROPOUT=m$GC_DROPOUT,
                                   stringsAsFactors=FALSE)
      }
      metrics &lt;- do.call(rbind, metrics)
      row.names(metrics) &lt;- metrics$Sample

      return(metrics)
    }




    readPicard &lt;- function(source) {
      metrics.aln &lt;- readPicard.alignment_summary_metrics(source)
      metrics.rna &lt;- readPicard.rna_metrics(source)
      metrics.dup &lt;- readPicard.duplicate_metrics(source)

      stopifnot(all(row.names(metrics.aln) %in% row.names(metrics.rna)) &amp;
                all(row.names(metrics.rna) %in% row.names(metrics.dup)) &amp;
                all(row.names(metrics.dup) %in% row.names(metrics.aln)))

      metrics.aln$File &lt;- NULL
      metrics.rna$File &lt;- NULL
      metrics.dup$File &lt;- NULL
      metrics.rna$Sample &lt;- NULL
      metrics.dup$Sample &lt;- NULL

      metrics &lt;- cbind(metrics.aln, metrics.rna[row.names(metrics.aln), ])
      metrics &lt;- cbind(metrics, metrics.dup[row.names(metrics.aln), ])

      return(metrics)
    }

    ## Execution  
    sourcePicard = ${_input[-1]:dr}

    Picard_qualityMetrics &lt;- readPicard(sourcePicard)
    write.table(Picard_qualityMetrics, file=&quot;${_output}&quot;,col.names=TRUE, row.names=FALSE, quote=FALSE)
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/molecular_phenotypes/calling"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../bulk_expression.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">RNA-seq expression</p>
      </div>
    </a>
    <a class="right-next"
       href="../QC/bulk_expression_QC.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Sample level RNA-seq quality control</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods-overview">Methods overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#gene-level-quantifications">Gene-level quantifications</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#exon-level-quantifications">Exon-level quantifications</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#paired-end">Paired end:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#single-end">Single end:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#strand-option"><code class="docutils literal notranslate"><span class="pre">strand</span></code> option</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#read-length-option"><code class="docutils literal notranslate"><span class="pre">read_length</span></code> option</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-and-global-parameters">Setup and global parameters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-0-qc-before-alignment">Step 0: QC before alignment</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-inputs">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-outputs">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-remove-adaptor-through-fastp">Step 1: Remove adaptor through <code class="docutils literal notranslate"><span class="pre">fastp</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id2">Step Outputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-options">Step options</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-alternative-remove-adaptor-through-trimmomatic">Step 1 Alternative: Remove adaptor through <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id3">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id4">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-alignment-through-star">Step 2: Alignment through <code class="docutils literal notranslate"><span class="pre">STAR</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id5">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id6">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-mark-duplicates-reads-qc-through-picard">Step 3: Mark duplicates reads &amp; QC through <code class="docutils literal notranslate"><span class="pre">Picard</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id7">Step Inputs:</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id8">Step Outputs:</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-post-aligment-qc-through-rna-seqc">Step 4: Post aligment QC through <code class="docutils literal notranslate"><span class="pre">RNA-SeQC</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id9">Step Inputs</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id10">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-6-quantify-expression-through-rsem">Step 6: Quantify expression through <code class="docutils literal notranslate"><span class="pre">RSEM</span></code></a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-input">Step Input</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id11">Step Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#step-7-summarize-with-multiqc">Step 7: summarize with MultiQC</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>