

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Genotype data preprocessing &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=e353d410970836974a52" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=e353d410970836974a52" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=e353d410970836974a52" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52" />

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/data_preprocessing/genotype_preprocessing';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Genotype VCF file quality control" href="genotype/VCF_QC.html" />
    <link rel="prev" title="Normalization and phenotype table generation for splicingQTL analysis" href="../molecular_phenotypes/QC/splicing_normalization.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="current reference internal" href="#">Genotype data preprocessing</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate/MASH/mashr.html">Data overview</a></li>






</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../fine_mapping/SuSiE/SuSiE.html">Fine-mapping with SuSiE model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/COLOC/fastenloc_dap.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/data_preprocessing/genotype_preprocessing.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/genotype_preprocessing.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/data_preprocessing/genotype_preprocessing.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Genotype data preprocessing</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#analysis-steps">Analysis steps</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#input-data-requirement">Input data requirement</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#qc-for-vcf-files">QC for VCF files</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-merged-output-of-vcf-qc">Optional: Merged output of VCF QC</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#get-overlap-samples-for-genotypes-and-phenotype-data">Get overlap samples for genotypes and phenotype data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-qc-for-pca">Genotype QC for PCA</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#principle-component-analysis">Principle component analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#finalized-genotype-qc">Finalized genotype QC</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#for-common-and-infrequent-variants">For common and infrequent variants</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-remove-outlier-not-part-of-the-protocol-data-analysis">Optional: Remove outlier (Not part of the protocol data analysis)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-extract-common-variants-not-part-of-the-protocol-data-analysis">Optional: Extract common variants (Not part of the protocol data analysis)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-reformatting">Genotype Reformatting</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#partition-by-chromosome">Partition by Chromosome</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-convert-to-vcf-not-part-of-the-protocol-data-analysis">Optional: convert to VCF (Not part of the protocol data analysis)</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="genotype-data-preprocessing">
<h1>Genotype data preprocessing<a class="headerlink" href="#genotype-data-preprocessing" title="Permalink to this heading">#</a></h1>
<p>This section documents output from the genotype section (colored in light yellow) of xqtl discovery pipeline MWE and explained the purpose for each of the command. The file used in this page can be found within the protocol_data folder of the synapse repo as described on the landing page.</p>
<p><strong>For mutli-theme study using same set of genotype (i.e. different tissues for the same participant) , the VCF QC and final Genotype QC steps are ran once. The related/unrelated genotype QC and PCA were ran once per tissue, as only the participant also have molecular phenotype data will be used for PCA</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%preview<span class="w"> </span>../images/eqtl_command.png
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> ../images/eqtl_command.png (254.4 KiB):</div></div><img alt="../../_images/2a1007985802c4ec80878bfc482f95d9ef647170c5e60a3f2b38b9afd8480f24.png" src="../../_images/2a1007985802c4ec80878bfc482f95d9ef647170c5e60a3f2b38b9afd8480f24.png" />
</div>
</div>
<section id="overview">
<h2>Overview<a class="headerlink" href="#overview" title="Permalink to this heading">#</a></h2>
<section id="analysis-steps">
<h3>Analysis steps<a class="headerlink" href="#analysis-steps" title="Permalink to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>Genotype data quality control (QC). See here for the <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/genotype/GWAS_QC.html">QC default settings</a>.</p></li>
<li><p>Principle component analysis (PCA) based QC, and PC computation for each sub-population available in the genotype data.</p></li>
<li><p>Genomic relationship matrix (GRM) computation (optional, when in Step 1 genotype QC detects related individuals or the input is known to have related individuals).</p></li>
</ol>
</section>
<section id="input-data-requirement">
<h3>Input data requirement<a class="headerlink" href="#input-data-requirement" title="Permalink to this heading">#</a></h3>
<ol class="arabic simple">
<li><p>Genotype data. See here for <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/genotype/genotype_formatting.html">format details</a>.</p></li>
<li><p>[Optional] a sample information file to specific population information, if external data such as HapMap or 1000 Genomes are to be integrated to the PCA analysis to visualize and assess population structure in the genotype data. See here for <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/genotype/genotype_formatting.html">format details</a>.</p></li>
</ol>
</section>
</section>
<section id="qc-for-vcf-files">
<h2>QC for VCF files<a class="headerlink" href="#qc-for-vcf-files" title="Permalink to this heading">#</a></h2>
<p>Perform QC on VCF files. The QC-ed data will also be exported to PLINK format for next steps analysis. The QC procedure includes:</p>
<ol class="arabic">
<li><p>Handling the formatting of multi-allelic sites,</p></li>
<li><p>Genotype and variant level filtering based on genotype calling qualities.</p>
<p>2.1. Genotype depth filters: For WES data, UK Biobank recommends <strong>SNPs DP&gt;10 and Indels DP&gt;10 for indels.</strong> However we think for WGS we can be less stringent, or simply rely on GQ. Users can set it to 1 eg, <code class="docutils literal notranslate"><span class="pre">--DP</span> <span class="pre">1</span> <span class="pre">--DP-indel</span> <span class="pre">1</span> </code></p>
<p>2.2. Genotype quality GQ&gt;20.</p>
<p>2.3. At least one sample per site passed the allele balance threshold &gt;= 0.15 for SNPs and &gt;=0.20 for indels (heterozygous variants). Allele balance is calculated for heterozygotes as the number of bases supporting the least-represented allele over the total number of base observations.</p>
</li>
<li><p>Known/novel variants annotation</p></li>
<li><p>Summary statistics before and after QC, in particular the ts/tv ratio, to assess the effectiveness of QC.</p></li>
</ol>
<p>3 and 4 above are for explorative analysis on the overall quality assessment of genotype data in the VCF files. We annotate known and novel variants because ts/tv are expected to be different between known and novel variants, and is important QC metric to assess the effectiveness of our QC.
After the procedure was done, the output will be converted to a plink bed/bim/fam binary format for downstream analysis.</p>
<p>It will take ~25G of memory and ~5min to complete the following step</p>
<p>Resource Usage:</p>
<ol class="arabic simple">
<li><p>Time elapsed: 780s
chr21 :  400s
chr22 :  380s</p></li>
<li><p>Max Vmem: 20G</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/VCF_QC.ipynb<span class="w"> </span>qc<span class="w"> </span><span class="se">\</span>
<span class="w">     </span>--genoFile<span class="w"> </span>input_data/Genotype/DEJ_11898_B01_GRM_WGS_2017-05-15_21.recalibrated_variants.xqtl_protocol_data.add_chr.vcf.gz<span class="w"> </span><span class="se">\</span>
<span class="w">                </span>input_data/Genotype/DEJ_11898_B01_GRM_WGS_2017-05-15_21.recalibrated_variants.xqtl_protocol_data.add_chr.vcf.gz<span class="w"> </span><span class="se">\</span>
<span class="w">     </span>--dbsnp-variants<span class="w"> </span>reference_data/00-All.add_chr.variants.gz<span class="w"> </span><span class="se">\</span>
<span class="w">     </span>--reference-genome<span class="w"> </span>reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta<span class="w"> </span><span class="se">\</span>
<span class="w">     </span>--cwd<span class="w"> </span>output/data_preprocessing/genotype<span class="w"> </span><span class="se">\</span>
<span class="w">     </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span><span class="se">\</span>
<span class="w">     </span>--walltime<span class="w"> </span><span class="s2">&quot;24h&quot;</span><span class="w"> </span><span class="se">\</span>
<span class="w">     </span>--add_chr
</pre></div>
</div>
</div>
</div>
</section>
<section id="optional-merged-output-of-vcf-qc">
<h2>Optional: Merged output of VCF QC<a class="headerlink" href="#optional-merged-output-of-vcf-qc" title="Permalink to this heading">#</a></h2>
<p>To accomodate the genotype file with extra large size,we allow our user to perform QC of vcf file on a per chromomosome basis. However, to accurately perform filtering base on minor allele count, HWE deviation, and missingness, the input to GWAS QC need to be the whole genome genotype file. Therefore we provide this utlity for the user to merge their output from vcf QC</p>
<p>Resource Usage</p>
<ol class="arabic simple">
<li><p>Time elapsed: 13s</p></li>
<li><p>Max Vmem: 0.8G</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/genotype_formatting.ipynb<span class="w"> </span>merge_plink<span class="w"> </span><span class="se">\</span>
<span class="w">           </span>--genoFile<span class="w"> </span><span class="s1">&#39;output/data_preprocessing/genotype/DEJ_11898_B01_GRM_WGS_2017-05-15_21.recalibrated_variants.xqtl_protocol_data.add_chr.add_chr.leftnorm.filtered.bed&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">                       </span><span class="s1">&#39;output/data_preprocessing/genotype/DEJ_11898_B01_GRM_WGS_2017-05-15_22.recalibrated_variants.xqtl_protocol_data.add_chr.add_chr.leftnorm.filtered.bed&#39;</span><span class="w"> </span><span class="se">\</span>
<span class="w">           </span>--cwd<span class="w"> </span>output/data_preprocessing/genotype<span class="w"> </span><span class="se">\</span>
<span class="w">           </span>--name<span class="w"> </span>xqtl_protocol_data.merged<span class="w"> </span><span class="se">\</span>
<span class="w">           </span>--container<span class="w"> </span>containers/bioinfo.sif
</pre></div>
</div>
</div>
</div>
</section>
<section id="get-overlap-samples-for-genotypes-and-phenotype-data">
<h2>Get overlap samples for genotypes and phenotype data<a class="headerlink" href="#get-overlap-samples-for-genotypes-and-phenotype-data" title="Permalink to this heading">#</a></h2>
<p>In most of the time, the number of samples in genotype and phenotype data will not be an exact match. Therefore, we provides a utility to produce a list of overlaps samples between genotypes and each phenotypes. Following command retrieve the overlapped sample between Genotypes and methylation data</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/sample_matcher.ipynb<span class="w"> </span>filtered_sample_list<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-eqtl<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--phenoFile<span class="w"> </span>output/rnaseq/xqtl_protocol_data.rnaseqc.low_expression_filtered.outlier_removed.tmm.expression.bed.gz<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--genoFile<span class="w"> </span>output/data_preprocessing/genotype/xqtl_protocol_data.merged.fam<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--sample-participant-lookup<span class="w"> </span>reference_data/sample_participant_lookup.rnaseq<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--translated_phenoFile<span class="w"> </span>





<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/sample_matcher.ipynb<span class="w"> </span>filtered_sample_list<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-sqtl<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--phenoFile<span class="w"> </span>output/leaf_cutter/xqtl_protocol_data_bam_list_intron_usage_perind.counts.gz_raw_data.qqnorm.formated.bed.gz<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--genoFile<span class="w"> </span>output/data_preprocessing/genotype/xqtl_protocol_data.merged.fam<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--sample-participant-lookup<span class="w"> </span>reference_data/sample_participant_lookup.rnaseq<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--translated_phenoFile<span class="w"> </span>




<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/sample_matcher.ipynb<span class="w"> </span>filtered_sample_list<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-mqtl<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--phenoFile<span class="w"> </span>output/methylation/xqtl_protocol_data_arrayMethylation_covariates.sesame.beta.filter_na.soft.bed.gz<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--genoFile<span class="w"> </span>output/data_preprocessing/genotype/xqtl_protocol_data.merged.fam<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--sample-participant-lookup<span class="w"> </span>reference_data/sample_participant_lookup.rnaseq<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--translated_phenoFile<span class="w"> </span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>ERROR: Failed to locate pipeline/sample_matcher.ipynb.sos
ERROR: Failed to locate pipeline/sample_matcher.ipynb.sos
ERROR: Failed to locate pipeline/sample_matcher.ipynb.sos
</pre></div>
</div>
<div class="output traceback highlight-ipythontb notranslate"><div class="highlight"><pre><span></span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="genotype-qc-for-pca">
<h2>Genotype QC for PCA<a class="headerlink" href="#genotype-qc-for-pca" title="Permalink to this heading">#</a></h2>
<p>To accuratly estimate the PCs for the genotype. We split participants based on their kinship coefficients, estimated by king. We have implemented a graph-based method to avoid removing unrelated participants that are in a family. However, this option is disable by default due to the long running time it required. User can enable this option by changing the <code class="docutils literal notranslate"><span class="pre">--no-maximize_unrelated</span></code> to <code class="docutils literal notranslate"><span class="pre">--maximize_unrelated</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>king<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-eqtl/genotype_data<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/genotype/xqtl_protocol_data.merged.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--name<span class="w"> </span>protocol-eqtl<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--keep-samples<span class="w"> </span>output/data_preprocessing/protocol-eqtl/sample_participant_lookup.filtered_geno.txt<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--walltime<span class="w"> </span>48h<span class="w">   </span>--no-maximize_unrelated




<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>king<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-sqtl/genotype_data<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/genotype/xqtl_protocol_data.merged.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--name<span class="w"> </span>protocol-sqtl<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--keep-samples<span class="w"> </span>output/data_preprocessing/protocol-sqtl/sample_participant_lookup.filtered_geno.txt<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--walltime<span class="w"> </span>48h<span class="w">   </span>--no-maximize_unrelated




<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>king<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-mqtl/genotype_data<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/genotype/xqtl_protocol_data.merged.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--name<span class="w"> </span>protocol-mqtl<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--keep-samples<span class="w"> </span>output/data_preprocessing/protocol-mqtl/sample_participant_lookup.filtered_geno.txt<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--walltime<span class="w"> </span>48h<span class="w">   </span>--no-maximize_unrelated
</pre></div>
</div>
</div>
</div>
<p>Variant level and sample level QC on unrelated individuals using missingness &gt; 10% ,MAC &gt; 5 , and LD-prunning in preparation for PCA analysis:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>qc<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-eqtl/genotype_data<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-eqtl/genotype_data/xqtl_protocol_data.merged.protocol-eqtl.unrelated.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mac-filter<span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mem<span class="w"> </span>16G<span class="w">   </span>




<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>qc<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-sqtl/genotype_data<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-sqtl/genotype_data/xqtl_protocol_data.merged.protocol-sqtl.unrelated.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mac-filter<span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mem<span class="w"> </span>16G<span class="w">   </span>




<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>qc<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-mqtl/genotype_data<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-mqtl/genotype_data/xqtl_protocol_data.merged.protocol-mqtl.unrelated.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mac-filter<span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mem<span class="w"> </span>16G<span class="w">   </span>
</pre></div>
</div>
</div>
</div>
<p>Extract previously selected variants from related individuals in preparation for PCA, only applying missingness filter at sample level,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>qc_no_prune<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-eqtl/genotype_data<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-eqtl/genotype_data/xqtl_protocol_data.merged.protocol-eqtl.related.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--maf-filter<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--geno-filter<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mind-filter<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--hwe-filter<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--keep-variants<span class="w"> </span>output/data_preprocessing/protocol-eqtl/genotype_data/xqtl_protocol_data.merged.protocol-eqtl.unrelated.filtered.prune.in<span class="w">  </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w">  </span><span class="se">\</span>
<span class="w">   </span>--mem<span class="w"> </span>16G<span class="w">    </span>




<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>qc_no_prune<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-sqtl/genotype_data<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-sqtl/genotype_data/xqtl_protocol_data.merged.protocol-sqtl.related.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--maf-filter<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--geno-filter<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mind-filter<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--hwe-filter<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--keep-variants<span class="w"> </span>output/data_preprocessing/protocol-sqtl/genotype_data/xqtl_protocol_data.merged.protocol-sqtl.unrelated.filtered.prune.in<span class="w">  </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w">  </span><span class="se">\</span>
<span class="w">   </span>--mem<span class="w"> </span>16G<span class="w">    </span>




<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>qc_no_prune<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-mqtl/genotype_data<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-mqtl/genotype_data/xqtl_protocol_data.merged.protocol-mqtl.related.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--maf-filter<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--geno-filter<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mind-filter<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--hwe-filter<span class="w"> </span><span class="m">0</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--keep-variants<span class="w"> </span>output/data_preprocessing/protocol-mqtl/genotype_data/xqtl_protocol_data.merged.protocol-mqtl.unrelated.filtered.prune.in<span class="w">  </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w">  </span><span class="se">\</span>
<span class="w">   </span>--mem<span class="w"> </span>16G<span class="w">    </span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="principle-component-analysis">
<h2>Principle component analysis<a class="headerlink" href="#principle-component-analysis" title="Permalink to this heading">#</a></h2>
<p>Analysis below uses data from previous genotype QC.</p>
<p>First, perform PCA on unrelated individuals after QC in previous step,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/PCA.ipynb<span class="w"> </span>flashpca<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-eqtl/pca<span class="w"> </span>output/data_preprocessing/protocol-eqtl/pca<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-eqtl/genotype_data/xqtl_protocol_data.merged.protocol-eqtl.unrelated.filtered.prune.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/flashpcaR.sif<span class="w">  </span>



<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/PCA.ipynb<span class="w"> </span>flashpca<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-sqtl/pca<span class="w"> </span>output/data_preprocessing/protocol-sqtl/pca<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-sqtl/genotype_data/xqtl_protocol_data.merged.protocol-sqtl.unrelated.filtered.prune.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/flashpcaR.sif<span class="w">  </span>



<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/PCA.ipynb<span class="w"> </span>flashpca<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-mqtl/pca<span class="w"> </span>output/data_preprocessing/protocol-mqtl/pca<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-mqtl/genotype_data/xqtl_protocol_data.merged.protocol-mqtl.unrelated.filtered.prune.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/flashpcaR.sif<span class="w">  </span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%preview<span class="w"> </span>/mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.pc.png
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.pc.png (611.8 KiB):</div></div><img alt="../../_images/d8c66c42080774817c94462dedb43ef8477bf9648eaebfaa7c77127624572206.png" src="../../_images/d8c66c42080774817c94462dedb43ef8477bf9648eaebfaa7c77127624572206.png" />
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%preview<span class="w"> </span>/mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.scree.png
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.scree.png (56.6 KiB):</div></div><img alt="../../_images/1445c8b7a41dd3b236e18c3dfa33f315d97a6edb2ffa6a50eeea252598bf9760.png" src="../../_images/1445c8b7a41dd3b236e18c3dfa33f315d97a6edb2ffa6a50eeea252598bf9760.png" />
</div>
</div>
<p>Then project back related individuals to the space spanned by PC from unrelated individuals, and detect PC outliers.</p>
<p><strong>Notice here that you need to decide how many PCs to use to compute outliers (the <code class="docutils literal notranslate"><span class="pre">--maha-k</span></code> parameter), based on the PC scree plot. You should include a decent number of PCs that capture enough variation in genotype data. Please manually examine the PC plot before running this command.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span><span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/PCA.ipynb<span class="w"> </span>project_samples<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-eqtl/pca<span class="w"> </span>output/data_preprocessing/protocol-eqtl/pca<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-eqtl/genotype_data/xqtl_protocol_data.merged.protocol-eqtl.related.filtered.extracted.bed<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--pca-model<span class="w">  </span>output/data_preprocessing/protocol-eqtl/pca/xqtl_protocol_data.merged.protocol-eqtl.unrelated.filtered.prune.pca.rds<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--maha-k<span class="w"> </span><span class="sb">`</span>awk<span class="w"> </span><span class="s1">&#39;$3 &lt; 0.7&#39;</span><span class="w"> </span>output/data_preprocessing/protocol-eqtl/pca/xqtl_protocol_data.merged.protocol-eqtl.unrelated.filtered.prune.pca.scree.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>tail<span class="w"> </span>-1<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w">  </span><span class="sb">`</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--container<span class="w"> </span>containers/flashpcaR.sif<span class="w">  </span>



<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/PCA.ipynb<span class="w"> </span>project_samples<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-sqtl/pca<span class="w"> </span>output/data_preprocessing/protocol-sqtl/pca<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-sqtl/genotype_data/xqtl_protocol_data.merged.protocol-sqtl.related.filtered.extracted.bed<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--pca-model<span class="w">  </span>output/data_preprocessing/protocol-sqtl/pca/xqtl_protocol_data.merged.protocol-sqtl.unrelated.filtered.prune.pca.rds<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--maha-k<span class="w"> </span><span class="sb">`</span>awk<span class="w"> </span><span class="s1">&#39;$3 &lt; 0.7&#39;</span><span class="w"> </span>output/data_preprocessing/protocol-sqtl/pca/xqtl_protocol_data.merged.protocol-sqtl.unrelated.filtered.prune.pca.scree.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>tail<span class="w"> </span>-1<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w">  </span><span class="sb">`</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--container<span class="w"> </span>containers/flashpcaR.sif<span class="w">  </span>



<span class="w"> </span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/PCA.ipynb<span class="w"> </span>project_samples<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--cwd<span class="w"> </span>output/data_preprocessing/protocol-mqtl/pca<span class="w"> </span>output/data_preprocessing/protocol-mqtl/pca<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--genoFile<span class="w"> </span>output/data_preprocessing/protocol-mqtl/genotype_data/xqtl_protocol_data.merged.protocol-mqtl.related.filtered.extracted.bed<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--pca-model<span class="w">  </span>output/data_preprocessing/protocol-mqtl/pca/xqtl_protocol_data.merged.protocol-mqtl.unrelated.filtered.prune.pca.rds<span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--maha-k<span class="w"> </span><span class="sb">`</span>awk<span class="w"> </span><span class="s1">&#39;$3 &lt; 0.7&#39;</span><span class="w"> </span>output/data_preprocessing/protocol-mqtl/pca/xqtl_protocol_data.merged.protocol-mqtl.unrelated.filtered.prune.pca.scree.txt<span class="w"> </span><span class="p">|</span><span class="w"> </span>tail<span class="w"> </span>-1<span class="w"> </span><span class="p">|</span><span class="w"> </span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w">  </span><span class="sb">`</span><span class="w"> </span><span class="se">\</span>
<span class="w">        </span>--container<span class="w"> </span>containers/flashpcaR.sif<span class="w">  </span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>%preview<span class="w"> </span>/mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.related.filtered.extracted.pca.projected.pc.png
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.related.filtered.extracted.pca.projected.pc.png (615.6 KiB):</div></div><img alt="../../_images/ef9083e21494e1015c24f5068c748f690dd16deda4b9a199d1ecb66973e594ec.png" src="../../_images/ef9083e21494e1015c24f5068c748f690dd16deda4b9a199d1ecb66973e594ec.png" />
</div>
</div>
<p>Two outliers was removed, as shown below</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>cat<span class="w"> </span>/mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/output/pca/xqtl_protocol_data.unrelated.filtered.prune.pca.outliers
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>0	SM-CJGN1
</pre></div>
</div>
</div>
</div>
</section>
<section id="finalized-genotype-qc">
<h2>Finalized genotype QC<a class="headerlink" href="#finalized-genotype-qc" title="Permalink to this heading">#</a></h2>
<section id="for-common-and-infrequent-variants">
<h3>For common and infrequent variants<a class="headerlink" href="#for-common-and-infrequent-variants" title="Permalink to this heading">#</a></h3>
<p>In our demo, we only keep variants having sample minor allele count MAC &gt;= 5, for the sake of convenience focusing our single variant analysis. It should be noted that mac and maf dont overwrite each other within our module. It is up to plink to decide how to handle the secnario of both <code class="docutils literal notranslate"><span class="pre">--maf-filter</span></code> and <code class="docutils literal notranslate"><span class="pre">--mac-filter</span></code> &gt; 0. Therefore, the user should not specify both <code class="docutils literal notranslate"><span class="pre">--maf-filter</span></code> and <code class="docutils literal notranslate"><span class="pre">--mac-filter</span></code> without good reason.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Resource<span class="w"> </span>Usage
<span class="m">1</span>.<span class="w"> </span>Time<span class="w"> </span>elapsed:<span class="w"> </span>14s
<span class="m">2</span>.<span class="w"> </span>Max<span class="w"> </span>Vmem:<span class="w"> </span><span class="m">0</span>.4G
<span class="w">    </span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>qc_no_prune<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/genotype<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/genotype/xqtl_protocol_data.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--geno-filter<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mind-filter<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--hwe-filter<span class="w"> </span>1e-06<span class="w">   </span><span class="se">\</span>
<span class="w">   </span>--mac-filter<span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w">  </span><span class="se">\</span>
<span class="w">   </span>--mem<span class="w"> </span>40G
</pre></div>
</div>
</div>
</div>
</section>
<section id="optional-remove-outlier-not-part-of-the-protocol-data-analysis">
<h3>Optional: Remove outlier (Not part of the protocol data analysis)<a class="headerlink" href="#optional-remove-outlier-not-part-of-the-protocol-data-analysis" title="Permalink to this heading">#</a></h3>
<p>PCA analysis (see covariate formatting page) can possibly provide us a list of outlier samples to be removed from the genotype data. We can optionally remove them and perform genotype data QC.</p>
<p>Removal of samples is a complicated decision that require human descretion. Therefore outlier removal is not performed automatically. The <code class="docutils literal notranslate"><span class="pre">--remove-samples</span></code> parameter take a two columns table that illustrated by  <code class="docutils literal notranslate"><span class="pre">output/pca/xqtl_protocol_data.related.filtered.extracted.pca.projected.outliers</span></code></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>qc_no_prune<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--cwd<span class="w"> </span>output/genotype<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--genoFile<span class="w"> </span>output/genotype/xqtl_protocol_data.bed<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--geno-filter<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--mind-filter<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--hwe-filter<span class="w"> </span>1e-06<span class="w">   </span><span class="se">\</span>
<span class="w">    </span>--mac-filter<span class="w"> </span><span class="m">5</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--mem<span class="w"> </span>40G
<span class="w">    </span>--remove-samples<span class="w"> </span>output/pca/xqtl_protocol_data.related.filtered.extracted.pca.projected.outliers
</pre></div>
</div>
</div>
</div>
<p>To fully illustrate how PCA and PCA based QC is done on sample with <strong>multiple populations (a more complicated but potentially very useful setting)</strong>, please try out the example documented in the <a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/genotype/PCA.html">PCA analysis module</a> with a different example data-set.</p>
</section>
<section id="optional-extract-common-variants-not-part-of-the-protocol-data-analysis">
<h3>Optional: Extract common variants (Not part of the protocol data analysis)<a class="headerlink" href="#optional-extract-common-variants-not-part-of-the-protocol-data-analysis" title="Permalink to this heading">#</a></h3>
<p>In case you would like to focus on common variants only, the genotype data can also filtered by the criteria of missingness &gt; 10% and MAF&gt;5% as shown below to retain the common variants, User can detemined the level of filtering by changing the generated command manually.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>qc_no_prune<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/genotype<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/genotype/xqtl_protocol_data.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--geno-filter<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mind-filter<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--hwe-filter<span class="w"> </span>1e-06<span class="w">   </span><span class="se">\</span>
<span class="w">   </span>--maf-filter<span class="w"> </span><span class="m">0</span>.05<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w">  </span><span class="se">\</span>
<span class="w">   </span>--mem<span class="w"> </span>40G
</pre></div>
</div>
</div>
</div>
<p>If outliers are to be removed, following command could be used:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/GWAS_QC.ipynb<span class="w"> </span>qc_no_prune<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--cwd<span class="w"> </span>output/genotype<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--genoFile<span class="w"> </span>output/genotype/xqtl_protocol_data.bed<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--geno-filter<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--mind-filter<span class="w"> </span><span class="m">0</span>.1<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--hwe-filter<span class="w"> </span>1e-06<span class="w">   </span><span class="se">\</span>
<span class="w">   </span>--maf-filter<span class="w"> </span><span class="m">0</span>.05<span class="w"> </span><span class="se">\</span>
<span class="w">   </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w">  </span><span class="se">\</span>
<span class="w">   </span>--mem<span class="w"> </span>40G<span class="w"> </span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="genotype-reformatting">
<h2>Genotype Reformatting<a class="headerlink" href="#genotype-reformatting" title="Permalink to this heading">#</a></h2>
<section id="partition-by-chromosome">
<h3>Partition by Chromosome<a class="headerlink" href="#partition-by-chromosome" title="Permalink to this heading">#</a></h3>
<p>The association analysis of TensorQTL required the chromosome be seperated by chromosome to allowed for parallel analysis.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>Resource<span class="w"> </span>Usage
<span class="m">1</span>.<span class="w"> </span>Time<span class="w"> </span>elapsed:<span class="w"> </span>24s
<span class="m">2</span>.<span class="w"> </span>Max<span class="w"> </span>Vmem:<span class="w"> </span><span class="m">0</span>.2G
<span class="w">    </span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/genotype_formatting.ipynb<span class="w"> </span>plink_by_chrom<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--genoFile<span class="w"> </span>output/genotype/xqtl_protocol_data.filtered.bed<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--cwd<span class="w"> </span>output/data_preprocessing<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--chrom<span class="w">  </span><span class="sb">`</span>cut<span class="w"> </span>-f<span class="w"> </span><span class="m">1</span><span class="w">  </span>output/genotype/xqtl_protocol_data.filtered.bim<span class="w">  </span><span class="p">|</span><span class="w"> </span>uniq<span class="w"> </span><span class="p">|</span><span class="w"> </span>sed<span class="w"> </span><span class="s2">&quot;s/chr//g&quot;</span><span class="sb">`</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="optional-convert-to-vcf-not-part-of-the-protocol-data-analysis">
<h3>Optional: convert to VCF (Not part of the protocol data analysis)<a class="headerlink" href="#optional-convert-to-vcf-not-part-of-the-protocol-data-analysis" title="Permalink to this heading">#</a></h3>
<p>If association analysis by APEX is needed, following command convert the per-chromosome plink binaray generated above into indexed vcf.gz.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>sos<span class="w"> </span>run<span class="w"> </span>pipeline/genotype_formatting.ipynb<span class="w"> </span>plink_to_vcf<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--genoFile<span class="w"> </span>plink_files_list.txt<span class="w">  </span><span class="se">\</span>
<span class="w">    </span>--cwd<span class="w"> </span>output/data_preprocessing<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--container<span class="w"> </span>containers/bioinfo.sif<span class="w"> </span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<div class="toctree-wrapper compound">
</div>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "bash"
        },
        kernelOptions: {
            name: "bash",
            path: "./code/data_preprocessing"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'bash'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  
<div class="footer-article-items footer-article__inner">
  
    <div class="footer-article-item"><!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../molecular_phenotypes/QC/splicing_normalization.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Normalization and phenotype table generation for splicingQTL analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="genotype/VCF_QC.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Genotype VCF file quality control</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div></div>
  
</div>

                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overview">Overview</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#analysis-steps">Analysis steps</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#input-data-requirement">Input data requirement</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#qc-for-vcf-files">QC for VCF files</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-merged-output-of-vcf-qc">Optional: Merged output of VCF QC</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#get-overlap-samples-for-genotypes-and-phenotype-data">Get overlap samples for genotypes and phenotype data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-qc-for-pca">Genotype QC for PCA</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#principle-component-analysis">Principle component analysis</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#finalized-genotype-qc">Finalized genotype QC</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#for-common-and-infrequent-variants">For common and infrequent variants</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-remove-outlier-not-part-of-the-protocol-data-analysis">Optional: Remove outlier (Not part of the protocol data analysis)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-extract-common-variants-not-part-of-the-protocol-data-analysis">Optional: Extract common variants (Not part of the protocol data analysis)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#genotype-reformatting">Genotype Reformatting</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#partition-by-chromosome">Partition by Chromosome</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#optional-convert-to-vcf-not-part-of-the-protocol-data-analysis">Optional: convert to VCF (Not part of the protocol data analysis)</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
       Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=e353d410970836974a52"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=e353d410970836974a52"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>