

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>RNA-seq calling and QC &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=c5ced968eda925caa686" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=c5ced968eda925caa686" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=c5ced968eda925caa686" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=c5ced968eda925caa686" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=c5ced968eda925caa686" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=c5ced968eda925caa686" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=c5ced968eda925caa686"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/commands_generator/bulk_expression_commands';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Univariate xQTL Discovery" href="eQTL_analysis_commands.html" />
    <link rel="prev" title="Illustration of xQTL protocol" href="../xqtl_protocol_demo.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active"><a class="current reference internal" href="#">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype Data Imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced cis-QTL Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../cis_analysis/cis_advanced_overview.html">Overview of advanced cis-QTL analysis</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../cis_analysis/cis_workhorse.html">Advanced cis-QTL analysis with individual level data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/SuSiE.html">SuSiE fine-mapping workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/fSuSiE.html">fSuSiE workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/mvSuSiE.html">Multivariate fine-mapping workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../cis_analysis/mrmash.html">Multivariate TWAS prediction model</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/ldsc.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL-GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../pecotmr_integration/pecotmr.html">The PECOTMR intergration framework</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">GWAS integration: enrichment and colocalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/twas_mr.html">GWAS integration: TWAS and MR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/intact.html">TWAS and COLOC integration</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">Extract genome-wide data for multivariate analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">MASH analysis pipeline with posterior computation</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Rare xQTL</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../rare_xqtl/watershed.html">watershed analysis for rare xQTL</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Post Processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../post_processing/sumstats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/rds_to_vcf.html">Summary statistics in VCF format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/association_scan_post_processing.html">Association result processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">SuSIE results post process</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/commands_generator/bulk_expression_commands.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/commands_generator/bulk_expression_commands.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/commands_generator/bulk_expression_commands.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>RNA-seq calling and QC</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#download-process-and-index-reference-data-first-time-use">Download, Process, and index reference data (first time use)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#download-reference-and-annotation-data-first-time-use">Download reference and annotation data (first time use)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reformat-reference-and-annotation-data-first-time-use">Reformat reference and annotation data (first time use)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-index-first-time-only">Generation of Index (First time only)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#perform-data-quality-summary-via-fastqc">Perform data quality summary via <code class="docutils literal notranslate"><span class="pre">fastqc</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multi-sample-rna-seq-calling-qc-and-normalizations-automatically-for-every-studies-using-the-same-set-of-ref-index-gtf">Multi-sample RNA-seq Calling, QC, and normalizations (Automatically for every studies using the same set of ref/index/gtf)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#call-gene-level-rna-expression-via-rnaseqc">Call gene-level RNA expression via <code class="docutils literal notranslate"><span class="pre">rnaseqc</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#call-transcript-level-rna-expression-via-rsem">Call transcript level RNA expression via <code class="docutils literal notranslate"><span class="pre">RSEM</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multi-sample-rna-seq-qc-for-every-set-of-samples">Multi-sample RNA-seq QC (For every set of samples)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multi-sample-read-count-normalization-for-every-set-of-samples">Multi-sample read count normalization (For every set of samples)</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="rna-seq-calling-and-qc">
<h1>RNA-seq calling and QC<a class="headerlink" href="#rna-seq-calling-and-qc" title="Permalink to this heading">#</a></h1>
<p>This document shows the use of various modules to prepare reference data, perform RNA-seq calling, expression level quantification and quality control. In particular,</p>
<ol class="arabic simple">
<li><p><a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/data_preprocessing/reference_data.html"><code class="docutils literal notranslate"><span class="pre">reference_data.ipynb</span></code></a></p></li>
<li><p><a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/molecular_phenotypes/calling/RNA_calling.html"><code class="docutils literal notranslate"><span class="pre">RNA_calling.ipynb</span></code></a></p></li>
<li><p><a class="reference external" href="https://cumc.github.io/xqtl-pipeline/code/molecular_phenotypes/QC/readCount_QC.html"><code class="docutils literal notranslate"><span class="pre">readCount_QC.ipynb</span></code></a></p></li>
</ol>
<p>A minimal working example is available on <a class="reference external" href="https://drive.google.com/drive/u/0/folders/11kQv7PXozsKkgeqADH-28bC_kZ-w_oHo">Google Drive</a>.</p>
<p>This is a master control notebook for the molecular_phenotype_calling section of our pipeline. When following example commands was excuted, it will checked if the reference data was presented and generate the reference data if not. After that, molecular phenotype calling will be done based on the recipe that were fed into it. For each rows in the recipe, It will output a bed.gz file that are ready for downstream analysis.</p>
<p>By default, the reference data is assumed to be stored in the reference_data folder under the cwd. However, user can specify whether the</p>
<ul>
<li><p>Input:</p>
<p>A recipe file, with columns: <code class="docutils literal notranslate"><span class="pre">&quot;Theme&quot;,</span> <span class="pre">&quot;Sample_fastq_list&quot;,</span> <span class="pre">&quot;Data_dir&quot;</span></code></p>
</li>
<li><p>Output:</p>
<p>A bed.gz file for each row in the Input</p>
</li>
<li><p>parameters:</p>
<ul class="simple">
<li><p>flag <code class="docutils literal notranslate"><span class="pre">--stranded</span></code> or <code class="docutils literal notranslate"><span class="pre">--no-stranded</span></code> : whether the rna seq data intended to be called is stranded or not.</p></li>
<li><p>flag <code class="docutils literal notranslate"><span class="pre">--pair-ended</span></code> or <code class="docutils literal notranslate"><span class="pre">--no-pair-ended</span></code> : whether the rna seq data intended to be called is paired-ended or not.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--exe_dir</span></code> : diretory to the workflow notebook</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--sample-participant-lookup</span></code> : a sample participant lookup table with the first column being sample ID and second column being the cooresponding sample ids in the genotype file</p></li>
</ul>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">bulk_expression_commands</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span> 
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run bulk_expression_commands.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  ref_data
  gene_annotation
  indexing
  fastqc
  rnaseqc
  RSEM
  QC
  normalization

Global Workflow Options:
  --recipe VAL (as path, required)
                        The aforementioned input recipe
  --cwd . (as path)
                        Overall wd, the file structure of analysis is
                        wd/[steps]/[sub_dir for each steps]
  --refd  path(f&#39;{cwd}/reference_data/&#39;)

                        dir to store the reference data
  --exe-dir /home/gw/GIT/xqtl-pipeline/pipeline (as path)
                        Diretory to the excutable
  --container-rnaquant &#39;/mnt/mfs/statgen/containers/rna_quantification.sif&#39;
  --yml csg.yml (as path)
  --jobs 50 (as int)
  --queue &#39;csg,q&#39;

Sections
  ref_data_1:
  ref_data_2:
  gene_annotation:
    Workflow Options:
      --[no-]stranded (required)
  indexing:
  fastqc:
  rnaseqc:
    Workflow Options:
      --fasta-with-adapters-etc &#39;TruSeq3-PE.fa&#39;
  RSEM:
    Workflow Options:
      --fasta-with-adapters-etc &#39;TruSeq3-PE.fa&#39;
      --[no-]is-paired-end (required)
  QC:
    Workflow Options:
      --low-expr-TPM 0.1 (as float)
      --low-expr-TPM-percent 0.2 (as float)
      --RLEFilterPercent 0.05 (as float)
      --DSFilterPercent 0.05 (as float)
      --topk-genes 100 (as int)
      --cluster-percent 0.6 (as float)
      --pvalues-cut 0.05 (as float)
      --cluster-level 5 (as int)
  normalization:
    Workflow Options:
      --sample-participant-lookup VAL (as path, required)
      --tpm-threshold 0.1 (as float)
      --count-threshold 6 (as int)
      --sample-frac-threshold 0.2 (as float)
      --normalization-method tmm
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1">## The aforementioned input recipe</span>
<span class="kn">parameter:</span> <span class="n">recipe</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1">## Overall wd, the file structure of analysis is wd/[steps]/[sub_dir for each steps]</span>
<span class="kn">parameter:</span>  <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1">## dir to store the reference data</span>
<span class="kn">parameter:</span> <span class="n">refd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">}</span><span class="s1">/reference_data/&#39;</span><span class="p">)</span>
<span class="c1">## Diretory to the excutable</span>
<span class="kn">parameter:</span> <span class="n">exe_dir</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;~/GIT/xqtl-pipeline/pipeline/&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">container_rnaquant</span> <span class="o">=</span> <span class="s1">&#39;/mnt/mfs/statgen/containers/rna_quantification.sif&#39;</span>
<span class="c1">## Cloud computing options:</span>
<span class="kn">parameter:</span> <span class="n">yml</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./csg.yml&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">jobs</span> <span class="o">=</span> <span class="mi">50</span> <span class="c1"># Number of jobs that are submitted to the cluster</span>
<span class="kn">parameter:</span> <span class="kp">queue</span> <span class="o">=</span> <span class="s2">&quot;csg&quot;</span> <span class="c1"># The queue that jobs are submitted to</span>
<span class="n">submission</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;-J </span><span class="si">{</span><span class="n">jobs</span><span class="si">}</span><span class="s1"> -c </span><span class="si">{</span><span class="n">yml</span><span class="si">}</span><span class="s1"> -q </span><span class="si">{</span><span class="kp">queue</span><span class="si">}</span><span class="s1">&#39;</span>
<span class="kn">parameter:</span> <span class="n">yml</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;csg.yml&quot;</span><span class="p">)</span> 
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">input_inv</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">recipe</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">to_dict</span><span class="p">(</span><span class="s2">&quot;records&quot;</span><span class="p">)</span>
<span class="kn">import</span> <span class="nn">os</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">recipe</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span> <span class="p">{</span><span class="s2">&quot;Theme&quot;</span> <span class="p">:</span> <span class="p">[</span><span class="s2">&quot;AC&quot;</span><span class="p">,</span><span class="s2">&quot;DLPFC&quot;</span><span class="p">,</span><span class="s2">&quot;PCC&quot;</span><span class="p">]</span> <span class="p">,</span>
                <span class="s2">&quot;Sample_fastq_list&quot;</span> <span class="p">:</span> <span class="p">[</span><span class="s2">&quot;AC_sample_fastq.list&quot;</span><span class="p">,</span>
                   <span class="s2">&quot;DLPFC_sample_fastq.list&quot;</span><span class="p">,</span>
                   <span class="s2">&quot;PCC_sample_fastq&quot;</span><span class="p">],</span>
                <span class="s2">&quot;Data_dir&quot;</span> <span class="p">:</span> <span class="p">[</span><span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/&quot;</span><span class="p">,</span>
                     <span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/&quot;</span><span class="p">]})</span>
<span class="n">recipe</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/mwe_recipe.txt&quot;</span><span class="p">,</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">index</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span>
<span class="n">recipe</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Theme</th>
      <th>Sample_fastq_list</th>
      <th>Data_dir</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>AC</td>
      <td>AC_sample_fastq.list</td>
      <td>/mnt/mfs/statgen/xqtl_workflow_testing/rna_qua...</td>
    </tr>
    <tr>
      <th>1</th>
      <td>DLPFC</td>
      <td>DLPFC_sample_fastq.list</td>
      <td>/mnt/mfs/statgen/xqtl_workflow_testing/rna_qua...</td>
    </tr>
    <tr>
      <th>2</th>
      <td>PCC</td>
      <td>PCC_sample_fastq</td>
      <td>/mnt/mfs/statgen/xqtl_workflow_testing/rna_qua...</td>
    </tr>
  </tbody>
</table>
</div></div></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">dryrun</span> <span class="o">./</span><span class="n">bulk_expression_master</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fastqc</span> \
    <span class="o">--</span><span class="n">recipe</span> <span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/mwe_recipe.txt&quot;</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">./</span>
    

<span class="n">sos</span> <span class="n">dryrun</span> <span class="o">./</span><span class="n">bulk_expression_master</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">normalization</span> \
    <span class="o">--</span><span class="n">recipe</span> <span class="s2">&quot;/mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/mwe_recipe.txt&quot;</span>  \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">./</span> \
    <span class="o">--</span><span class="n">sample</span><span class="o">-</span><span class="n">participant</span><span class="o">-</span><span class="n">lookup</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">data</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">sampleSheetAfterQC</span><span class="o">.</span><span class="n">txt</span>  \
    <span class="o">--</span><span class="n">fasta</span><span class="o">-</span><span class="k">with</span><span class="o">-</span><span class="n">adapters</span><span class="o">-</span><span class="n">etc</span>  <span class="s2">&quot;TruSeq3-PE.fa&quot;</span> \
    <span class="o">--</span><span class="n">stranded</span> \
    <span class="o">--</span><span class="ow">is</span><span class="o">-</span><span class="n">paired</span><span class="o">-</span><span class="n">end</span> <span class="o">&gt;</span> <span class="n">example_script</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">phenotype_master_control</span>
<span class="n">cat</span> <span class="n">example_script</span> <span class="o">|</span> <span class="n">grep</span> <span class="o">-</span><span class="n">v</span> <span class="n">HINT</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/pipeline/reference_data.ipynb download_hg_reference --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data
sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/pipeline/reference_data.ipynb download_gene_annotation --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data
sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/pipeline/reference_data.ipynb download_ercc_reference --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/reference_data.ipynb hg_reference \
    --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/GRCh38_full_analysis_set_plus_decoy_hla /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92 /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data \
    --ercc-reference /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92.fa \
    --hg-reference /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.gtf \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/reference_data.ipynb gene_annotation \
        --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data \
        --ercc-gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92.gtf \
        --hg-gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.gtf \
        --hg-reference /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/GRCh38_full_analysis_set_plus_decoy_hla /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92 /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92.noALT_noHLA_noDecoy_ERCC.fasta \
        --container /mnt/mfs/statgen/containers/rna_quantification.sif \
        --stranded  -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/reference_data.ipynb STAR_index RSEM_index \
    --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data \
    --hg-reference /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/GRCh38_full_analysis_set_plus_decoy_hla /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92 /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/ERCC92.noALT_noHLA_noDecoy_ERCC.fasta \
    --hg-gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \
    --mem 40G -J 50 -c csg.yml -q csg,q



sos run  /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  AC_sample_fastq.list \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \
    --mem 40G -J 50 -c csg.yml -q csg,q



sos run  /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  DLPFC_sample_fastq.list \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \
    --mem 40G -J 50 -c csg.yml -q csg,q



sos run  /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  PCC_sample_fastq \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --gtf /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \
    --mem 40G -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  AC_sample_fastq.list \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --RSEM-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/RSEM_Index \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \ 
    --mem 40G \
    --is_paired_end -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  DLPFC_sample_fastq.list \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --RSEM-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/RSEM_Index \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \ 
    --mem 40G \
    --is_paired_end -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/RNA_calling.ipynb rnaseqc_call \
    --cwd ./ \
    --samples  PCC_sample_fastq \
    --data-dir /mnt/mfs/statgen/xqtl_workflow_testing/rna_quant/rna_expression/ \
    --fasta_with_adapters_etc TruSeq3-PE.fa  \
    --STAR-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/STAR_Index \
    --RSEM-index /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/RSEM_Index \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif \ 
    --mem 40G \
    --is_paired_end -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_QC.ipynb qc \
    --cwd . \
    --tpm-gct AC.rnaseqc.gene_tpm.gct.gz \
    --counts-gct  AC.rnaseqc.gene_readsCount.gct.gz \
    --low_expr_TPM 0.1  \
    --low_expr_TPM_percent 0.2 \
    --RLEFilterPercent 0.05 \
    --DSFilterPercent 0.05 \
    --topk_genes 100 \
    --cluster_percent 0.6 \
    --pvalues_cut 0.05 \
    --cluster_level 5 \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_QC.ipynb qc \
    --cwd . \
    --tpm-gct DLPFC.rnaseqc.gene_tpm.gct.gz \
    --counts-gct  DLPFC.rnaseqc.gene_readsCount.gct.gz \
    --low_expr_TPM 0.1  \
    --low_expr_TPM_percent 0.2 \
    --RLEFilterPercent 0.05 \
    --DSFilterPercent 0.05 \
    --topk_genes 100 \
    --cluster_percent 0.6 \
    --pvalues_cut 0.05 \
    --cluster_level 5 \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_QC.ipynb qc \
    --cwd . \
    --tpm-gct PCC.rnaseqc.gene_tpm.gct.gz \
    --counts-gct  PCC.rnaseqc.gene_readsCount.gct.gz \
    --low_expr_TPM 0.1  \
    --low_expr_TPM_percent 0.2 \
    --RLEFilterPercent 0.05 \
    --DSFilterPercent 0.05 \
    --topk_genes 100 \
    --cluster_percent 0.6 \
    --pvalues_cut 0.05 \
    --cluster_level 5 \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_normalization.ipynb normalize \
    --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control \
    --tpm-gct AC.rnaseqc.low_expression_filtered.outlier_removed.tpm.gct.gz \
    --counts-gct AC.rnaseqc.low_expression_filtered.outlier_removed.geneCount.gct.gz \
    --annotation-gtf  /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf  \
    --sample-participant-lookup /mnt/mfs/statgen/snuc_pseudo_bulk/data/reference_data/sampleSheetAfterQC.txt \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif  \
    --tpm_threshold 0.1 \
    --count_threshold 6  \
    --sample_frac_threshold 0.2 \
    --normalization_method tmm  -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_normalization.ipynb normalize \
    --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control \
    --tpm-gct DLPFC.rnaseqc.low_expression_filtered.outlier_removed.tpm.gct.gz \
    --counts-gct DLPFC.rnaseqc.low_expression_filtered.outlier_removed.geneCount.gct.gz \
    --annotation-gtf  /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf  \
    --sample-participant-lookup /mnt/mfs/statgen/snuc_pseudo_bulk/data/reference_data/sampleSheetAfterQC.txt \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif  \
    --tpm_threshold 0.1 \
    --count_threshold 6  \
    --sample_frac_threshold 0.2 \
    --normalization_method tmm  -J 50 -c csg.yml -q csg,q



sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/bulk_expression_normalization.ipynb normalize \
    --cwd /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control \
    --tpm-gct PCC.rnaseqc.low_expression_filtered.outlier_removed.tpm.gct.gz \
    --counts-gct PCC.rnaseqc.low_expression_filtered.outlier_removed.geneCount.gct.gz \
    --annotation-gtf  /mnt/mfs/statgen/xqtl_workflow_testing/phenotype_master_control/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted..collapse_only.gene.ERCC.gtf  \
    --sample-participant-lookup /mnt/mfs/statgen/snuc_pseudo_bulk/data/reference_data/sampleSheetAfterQC.txt \
    --container /mnt/mfs/statgen/containers/rna_quantification.sif  \
    --tpm_threshold 0.1 \
    --count_threshold 6  \
    --sample_frac_threshold 0.2 \
    --normalization_method tmm  -J 50 -c csg.yml -q csg,q
</pre></div>
</div>
</div>
</div>
<section id="download-process-and-index-reference-data-first-time-use">
<h2>Download, Process, and index reference data (first time use)<a class="headerlink" href="#download-process-and-index-reference-data-first-time-use" title="Permalink to this heading">#</a></h2>
<section id="download-reference-and-annotation-data-first-time-use">
<h3>Download reference and annotation data (first time use)<a class="headerlink" href="#download-reference-and-annotation-data-first-time-use" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[ref_data_1]
output: f&#39;{refd:a}/GRCh38_full_analysis_set_plus_decoy_hla.fa&#39;,
        f&quot;{refd:a}/Homo_sapiens.GRCh38.103.chr.gtf&quot;,
        f&quot;{refd:a}/ERCC92.gtf&quot;, f&quot;{refd:a}/ERCC92.fa&quot;
report:  expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
        sos run $[exe_dir]/pipeline/reference_data.ipynb download_hg_reference --cwd $[_output[0]:d]
        sos run $[exe_dir]/pipeline/reference_data.ipynb download_gene_annotation --cwd $[_output[0]:d]
        sos run $[exe_dir]/pipeline/reference_data.ipynb download_ercc_reference --cwd $[_output[0]:d]
</pre></div>
</div>
</div>
</div>
</section>
<section id="reformat-reference-and-annotation-data-first-time-use">
<h3>Reformat reference and annotation data (first time use)<a class="headerlink" href="#reformat-reference-and-annotation-data-first-time-use" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[ref_data_2]
output:  f&#39;{_input:n}.noALT_noHLA_noDecoy_ERCC.fasta&#39;,f&#39;{_input:bn}.noALT_noHLA_noDecoy_ERCC.dict&#39;
report:  expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
        sos run $[exe_dir]/reference_data.ipynb hg_reference \
            --cwd $[_output[0]:d] \
            --ercc-reference $[_input[3]] \
            --hg-reference $[_input[1]] \
            --container $[container_rnaquant]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>**Notice: you need to know if your RNA-seq data is stranded or unstranded. If it is stranded, you should add `--stranded` to the command below.**
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[gene_annotation]
parameter: stranded = bool
input: output_from(&quot;ref_data_1&quot;),output_from(&quot;ref_data_2&quot;)[0]
output: f&#39;{_input[1]:n}.reformatted.ERCC.gtf&#39;, f&#39;{_input[1]:n}.reformatted{&quot;.collapse_only&quot; if stranded else &quot;&quot;}.gene.ERCC.gtf&#39;
report:  expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
    sos run $[exe_dir]/reference_data.ipynb gene_annotation \
            --cwd $[_output[0]:d] \
            --ercc-gtf $[_input[2]] \
            --hg-gtf $[_input[1]] \
            --hg-reference $[_input[4]] \
            --container $[container_rnaquant] \
            $[&quot;--stranded&quot; if stranded else &quot;--no-stranded&quot;]  $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
</section>
<section id="generation-of-index-first-time-only">
<h3>Generation of Index (First time only)<a class="headerlink" href="#generation-of-index-first-time-only" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[indexing]
input: output_from(&quot;gene_annotation&quot;)[1],output_from(&quot;ref_data_2&quot;)[0]
output: f&quot;{refd:a}/RSEM_Index/rsem_reference.n2g.idx.fa&quot;, 
        f&quot;{refd:a}/RSEM_Index/rsem_reference.grp&quot;, 
        f&quot;{refd:a}/RSEM_Index/rsem_reference.idx.fa&quot;,
        f&quot;{refd:a}/RSEM_Index/rsem_reference.ti&quot;, 
        f&quot;{refd:a}/RSEM_Index/rsem_reference.chrlist&quot;,
        f&quot;{refd:a}/RSEM_Index/rsem_reference.seq&quot;, 
        f&quot;{refd:a}/RSEM_Index/rsem_reference.transcripts.fa&quot;,
        f&quot;{refd:a}/STAR_Index/chrName.txt&quot;,
        f&quot;{refd:a}/STAR_Index/Log.out&quot;, 
        f&quot;{refd:a}/STAR_Index/transcriptInfo.tab&quot;,
        f&quot;{refd:a}/STAR_Index/exonGeTrInfo.tab&quot;, 
        f&quot;{refd:a}/STAR_Index/SAindex&quot;,
        f&quot;{refd:a}/STAR_Index/SA&quot;,
        f&quot;{refd:a}/STAR_Index/genomeParameters.txt&quot;, 
        f&quot;{refd:a}/STAR_Index/chrStart.txt&quot;,
        f&quot;{refd:a}/STAR_Index/sjdbList.out.tab&quot;, 
        f&quot;{refd:a}/STAR_Index/exonInfo.tab&quot;,
        f&quot;{refd:a}/STAR_Index/sjdbList.fromGTF.out.tab&quot;, 
        f&quot;{refd:a}/STAR_Index/chrLength.txt&quot;,
        f&quot;{refd:a}/STAR_Index/sjdbInfo.txt&quot;, 
        f&quot;{refd:a}/STAR_Index/Genome&quot;,
        f&quot;{refd:a}/STAR_Index/chrNameLength.txt&quot;, 
        f&quot;{refd:a}/STAR_Index/geneInfo.tab&quot;
report:  expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
        sos run $[exe_dir]/reference_data.ipynb STAR_index RSEM_index \
            --cwd $[refd:a] \
            --hg-reference $[_input[1]] \
            --hg-gtf $[_input[0]] \
            --container $[container_rnaquant] \
            --mem 40G $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="perform-data-quality-summary-via-fastqc">
<h2>Perform data quality summary via <code class="docutils literal notranslate"><span class="pre">fastqc</span></code><a class="headerlink" href="#perform-data-quality-summary-via-fastqc" title="Permalink to this heading">#</a></h2>
<p><strong>You need to figure out from fastqc results what adapter reference sequence to use for <code class="docutils literal notranslate"><span class="pre">--fasta_with_adapters_etc</span></code>. See section Step 1 below for details</strong>. We chose <code class="docutils literal notranslate"><span class="pre">--fasta_with_adapters_etc</span> <span class="pre">TruSeq3-PE.fa</span></code> for our toy example.</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">fasta_with_adapters_etc</span></code>: <strong>filename</strong> for the adapter reference file. According to <code class="docutils literal notranslate"><span class="pre">Trimmomatic</span></code> documention,</p></li>
</ul>
<blockquote>
<div><p>As a rule of thumb newer libraries will use <code class="docutils literal notranslate"><span class="pre">TruSeq3</span></code>, but this really depends on your service provider. If you use FASTQC, the Overrepresented Sequences report can help indicate which adapter file is best suited for your data. Illumina Single End or Illumina Paired End sequences indicate single-end or paired-end <code class="docutils literal notranslate"><span class="pre">TruSeq2</span></code> libraries, and the appropriate adapter files are <code class="docutils literal notranslate"><span class="pre">TruSeq2-SE.fa</span></code> and <code class="docutils literal notranslate"><span class="pre">TruSeq2-PE.fa</span></code> respectively. TruSeq Universal Adapter or TruSeq Adapter, Index  indicates <code class="docutils literal notranslate"><span class="pre">TruSeq-3</span></code> libraries, and the appropriate adapter files are <code class="docutils literal notranslate"><span class="pre">TruSeq3-SE.fa</span></code> or <code class="docutils literal notranslate"><span class="pre">TruSeq3-PE.fa</span></code>, for single-end and paired-end data respectively. Adapter sequences for <code class="docutils literal notranslate"><span class="pre">TruSeq2</span></code> multiplexed libraries, indicated by Illumina Multiplexing
, and the various RNA library preparations are not currently included.</p>
</div></blockquote>
<p>We have <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> workflow defined as followed. Users should decide what fasta adapter reference to use based on <code class="docutils literal notranslate"><span class="pre">fastqc</span></code> results (or their own knowledge).
It is assumed that all studies that were ran in one run are using the same adapter file</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[fastqc]
input: for_each = &quot;input_inv&quot;
report:  expand = &quot;$[ ]&quot;
    sos run $[exe_dir]/RNA_calling.ipynb fastqc \
        --cwd $[cwd:a]/output/rnaseq/fastqc/ \
        --samples $[_input_inv[&#39;Sample_fastq_list&#39;]] \
        --data-dir $[_input_inv[&#39;Data_dir&#39;]]   \
        --container $[container_rnaquant] $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
</section>
<section id="multi-sample-rna-seq-calling-qc-and-normalizations-automatically-for-every-studies-using-the-same-set-of-ref-index-gtf">
<h2>Multi-sample RNA-seq Calling, QC, and normalizations (Automatically for every studies using the same set of ref/index/gtf)<a class="headerlink" href="#multi-sample-rna-seq-calling-qc-and-normalizations-automatically-for-every-studies-using-the-same-set-of-ref-index-gtf" title="Permalink to this heading">#</a></h2>
<section id="call-gene-level-rna-expression-via-rnaseqc">
<h3>Call gene-level RNA expression via <code class="docutils literal notranslate"><span class="pre">rnaseqc</span></code><a class="headerlink" href="#call-gene-level-rna-expression-via-rnaseqc" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[rnaseqc]
parameter:fasta_with_adapters_etc = &quot;TruSeq3-PE.fa&quot;
input:  output_from(&quot;gene_annotation&quot;)[-1],output_from(&quot;indexing&quot;)[-1],for_each = &quot;input_inv&quot;
output: f&#39;{cwd}/{_input_inv[&quot;Theme&quot;]}.rnaseqc.gene_tpm.gct.gz&#39;,
        f&#39;{cwd}/{_input_inv[&quot;Theme&quot;]}.rnaseqc.gene_readsCount.gct.gz&#39;,
        f&#39;{cwd}/{_input_inv[&quot;Theme&quot;]}.rnaseqc.exon_readsCount.gct.gz&#39;,
        f&#39;{cwd}/{_input_inv[&quot;Theme&quot;]}.rnaseqc.metrics.tsv&#39;
report:  expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
        sos run  $[exe_dir]/RNA_calling.ipynb rnaseqc_call \
            --cwd $[_output[0]:d]/ \
            --samples  $[_input_inv[&#39;Sample_fastq_list&#39;]] \
            --data-dir $[_input_inv[&#39;Data_dir&#39;]] \
            --fasta_with_adapters_etc $[fasta_with_adapters_etc]  \
            --STAR-index $[_input[1]:d] \
            --gtf $[_input[0]] \
            --container $[container_rnaquant] \
            --mem 40G $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
</section>
<section id="call-transcript-level-rna-expression-via-rsem">
<h3>Call transcript level RNA expression via <code class="docutils literal notranslate"><span class="pre">RSEM</span></code><a class="headerlink" href="#call-transcript-level-rna-expression-via-rsem" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[RSEM]
parameter:fasta_with_adapters_etc = &quot;TruSeq3-PE.fa&quot;
parameter: paired_end = bool
input:  output_from(&quot;indexing&quot;)[0],output_from(&quot;indexing&quot;)[-1],for_each = &quot;input_inv&quot;
output: f&#39;{cwd}/{_input_inv[&quot;Theme&quot;]}.rsem_transcripts_expected_count.txt.gz&#39;,
        f&#39;{cwd}/{_input_inv[&quot;Theme&quot;]}.rsem_transcripts_tpm.txt.gz&#39;,
        f&#39;{cwd}/{_input_inv[&quot;Theme&quot;]}.rsem_transcripts_fpkm.txt.gz&#39;,
        f&#39;{cwd}/{_input_inv[&quot;Theme&quot;]}.rsem_transcripts_isopct.txt.gz&#39;,
        f&#39;{cwd}/{_input_inv[&quot;Theme&quot;]}.rsem_genes_expected_count.txt.gz&#39;,
        f&#39;{cwd}/{_input_inv[&quot;Theme&quot;]}.rsem_genes_tpm.txt.gz&#39;,
        f&#39;{cwd}/{_input_inv[&quot;Theme&quot;]}.rsem_genes_fpkm.txt.gz&#39;
report:  expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
        sos run $[exe_dir]/RNA_calling.ipynb rnaseqc_call \
            --cwd $[_output[0]:d]/ \
            --samples  $[_input_inv[&#39;Sample_fastq_list&#39;]] \
            --data-dir $[_input_inv[&#39;Data_dir&#39;]] \
            --fasta_with_adapters_etc $[fasta_with_adapters_etc]  \
            --STAR-index $[_input[1]:d] \
            --RSEM-index $[_input[0]:d] \
            --container $[container_rnaquant] \ 
            --mem 40G \
            $[&quot;--paired_end&quot; if paired_end else &quot; &quot;] $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="multi-sample-rna-seq-qc-for-every-set-of-samples">
<h2>Multi-sample RNA-seq QC (For every set of samples)<a class="headerlink" href="#multi-sample-rna-seq-qc-for-every-set-of-samples" title="Permalink to this heading">#</a></h2>
<p>We need to use a different MWE data-set that contains multiple samples  here is the <a class="reference external" href="https://drive.google.com/drive/u/0/folders/1Rv2bWHBbX_tastTh49ToYVDMV6rFP5Wk">Google Drive link</a>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[QC]
parameter: low_expr_TPM = 0.1
parameter: low_expr_TPM_percent = 0.2
parameter: RLEFilterPercent = 0.05
parameter: DSFilterPercent = 0.05
parameter: topk_genes = 100
parameter: cluster_percent = 0.6
parameter: pvalues_cut = 0.05
parameter: cluster_level = 5
depends: output_from(&quot;RSEM&quot;)
input: output_from(&quot;rnaseqc&quot;)
output: f&#39;{cwd}/{_input[0]:bnnn}.low_expression_filtered.outlier_removed.tpm.gct.gz&#39;, f&#39;{cwd}/{_input[0]:bnnn}.low_expression_filtered.outlier_removed.geneCount.gct.gz&#39;
report:  expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;
    sos run $[exe_dir]/bulk_expression_QC.ipynb qc \
        --cwd $[_output[0]:d] \
        --tpm-gct $[_input[0]] \
        --counts-gct  $[_input[1]] \
        --low_expr_TPM $[low_expr_TPM]  \
        --low_expr_TPM_percent $[low_expr_TPM_percent] \
        --RLEFilterPercent $[RLEFilterPercent] \
        --DSFilterPercent $[DSFilterPercent] \
        --topk_genes $[topk_genes] \
        --cluster_percent $[cluster_percent] \
        --pvalues_cut $[pvalues_cut] \
        --cluster_level $[cluster_level] \
        --container $[container_rnaquant] $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
</section>
<section id="multi-sample-read-count-normalization-for-every-set-of-samples">
<h2>Multi-sample read count normalization (For every set of samples)<a class="headerlink" href="#multi-sample-read-count-normalization-for-every-set-of-samples" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[normalization]
parameter: sample_participant_lookup = path
parameter: tpm_threshold = 0.1
parameter: count_threshold = 6
parameter: sample_frac_threshold = 0.2
parameter: normalization_method = &#39;tmm&#39;
input: output_from(&quot;QC&quot;),output_from(&quot;gene_annotation&quot;)
output: f&#39;{cwd:a}/{_input[0]:bnnn}.{normalization_method}.expression.bed.gz&#39;
report:  expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;
    sos run $[exe_dir]/bulk_expression_normalization.ipynb normalize \
        --cwd $[_output:d] \
        --tpm-gct $[_input[0]] \
        --counts-gct $[_input[1]] \
        --annotation-gtf  $[_input[-1]]  \
        --sample-participant-lookup $[sample_participant_lookup] \
        --container $[container_rnaquant]  \
        --tpm_threshold $[tpm_threshold] \
        --count_threshold $[count_threshold]  \
        --sample_frac_threshold $[sample_frac_threshold] \
        --normalization_method $[normalization_method]  $[submission if yml.is_file() else &quot;&quot; ]
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/commands_generator"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../xqtl_protocol_demo.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Illustration of xQTL protocol</p>
      </div>
    </a>
    <a class="right-next"
       href="eQTL_analysis_commands.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Univariate xQTL Discovery</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#download-process-and-index-reference-data-first-time-use">Download, Process, and index reference data (first time use)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#download-reference-and-annotation-data-first-time-use">Download reference and annotation data (first time use)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#reformat-reference-and-annotation-data-first-time-use">Reformat reference and annotation data (first time use)</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#generation-of-index-first-time-only">Generation of Index (First time only)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#perform-data-quality-summary-via-fastqc">Perform data quality summary via <code class="docutils literal notranslate"><span class="pre">fastqc</span></code></a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multi-sample-rna-seq-calling-qc-and-normalizations-automatically-for-every-studies-using-the-same-set-of-ref-index-gtf">Multi-sample RNA-seq Calling, QC, and normalizations (Automatically for every studies using the same set of ref/index/gtf)</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#call-gene-level-rna-expression-via-rnaseqc">Call gene-level RNA expression via <code class="docutils literal notranslate"><span class="pre">rnaseqc</span></code></a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#call-transcript-level-rna-expression-via-rsem">Call transcript level RNA expression via <code class="docutils literal notranslate"><span class="pre">RSEM</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multi-sample-rna-seq-qc-for-every-set-of-samples">Multi-sample RNA-seq QC (For every set of samples)</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multi-sample-read-count-normalization-for-every-set-of-samples">Multi-sample read count normalization (For every set of samples)</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
       Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=c5ced968eda925caa686"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=c5ced968eda925caa686"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>