
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17: http://docutils.sourceforge.net/" />

    <title>Stratified LD Score Regression &#8212; ADSP-FG Project</title>
    
  <link href="../../../_static/css/theme.css" rel="stylesheet">
  <link href="../../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Fine-mapping with PolyFun" href="../SuSiE_Ann/polyfun.html" />
    <link rel="prev" title="Enrichment analysis: TORUS and GREGOR" href="gwas_enrichment.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../README.html">
   xQTL analysis pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Command Generator
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">
   RNA-seq calling and QC
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Reference data
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../data_preprocessing/reference_data.html">
   Reference data standardization
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">
   RNA-seq expression
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">
     Quantifying expression from RNA-seq data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">
     Sample level RNA-seq quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">
     Bulk RNA-seq counts normalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/apa.html">
   Alternative polyadenylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">
     APA Calling
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../molecular_phenotypes/methylation.html">
   Methylation
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">
     Quantification of methylation data
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../molecular_phenotypes/splicing.html">
   Alternative splicing from RNA-seq data
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">
     Genotype VCF file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">
     Genotype PLINK file quality control
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">
     Gene coordinate annotation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data formatting
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">
     Covariate data formatting
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL association testing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../association_scan/APEX/APEX.html">
     APEX QTL association testing
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/>
  <label for="toctree-checkbox-10">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/>
  <label for="toctree-checkbox-11">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../COLOC/fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../COLOC/mvenloc.html">
     Multivariate SuSiE and ENLOC model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 current active has-children">
  <a class="reference internal" href="../functional_annotation.html">
   Functional annotation integration
  </a>
  <input checked="" class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/>
  <label for="toctree-checkbox-12">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul class="current">
   <li class="toctree-l2">
    <a class="reference internal" href="gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2 current active">
    <a class="current reference internal" href="#">
     Stratified LD Score Regression
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/summary_stats_standardizer.html">
   Overview of design
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../misc/rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/code/integrative_analysis/enrichment/ldsc.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/integrative_analysis/enrichment/ldsc.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/integrative_analysis/enrichment/ldsc.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#a-brief-review-on-stratified-ld-score-regression">
   A brief review on Stratified LD score regression
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#ldsc-model">
     LDSC model
    </a>
   </li>
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#stratified-ldsc">
     Stratified LDSC
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#command-interface">
   Command Interface
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#make-annotation-file">
   Make Annotation File
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#munge-summary-statistics-option-1-no-signed-summary-statistic">
   Munge Summary Statistics (Option 1: No Signed Summary Statistic)
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#munge-summary-statistics-option-2-no-signed-summary-statistic">
   Munge Summary Statistics (Option 2: No Signed Summary Statistic)
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#calculate-ld-scores">
   Calculate LD Scores
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#calculate-functional-enrichment-using-annotations">
   Calculate Functional Enrichment using Annotations
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="stratified-ld-score-regression">
<h1>Stratified LD Score Regression<a class="headerlink" href="#stratified-ld-score-regression" title="Permalink to this headline">¶</a></h1>
<p>This notebook implements the pipepline of <a class="reference external" href="https://github.com/bulik/ldsc/wiki">S-LDSC</a> for LD score and functional enrichment analysis. It is written by Anmol Singh (<a class="reference external" href="mailto:singh&#46;anmol&#37;&#52;&#48;columbia&#46;edu">singh<span>&#46;</span>anmol<span>&#64;</span>columbia<span>&#46;</span>edu</a>), with input from Dr. Gao Wang.</p>
<p><strong>FIXME: the initial draft is complete but pending Gao’s review and documentation with minimal working example</strong></p>
<p>The pipeline is developed to integrate GWAS summary statistics data, annotation data, and LD reference panel data to compute functional enrichment for each of the epigenomic annotations that the user provides using the S-LDSC model. We will first start off with an introduction, instructions to set up, and the minimal working examples. Then the workflow code that can be run using SoS on any data will be at the end.</p>
<section id="a-brief-review-on-stratified-ld-score-regression">
<h2>A brief review on Stratified LD score regression<a class="headerlink" href="#a-brief-review-on-stratified-ld-score-regression" title="Permalink to this headline">¶</a></h2>
<p>Here I briefly review LD Score Regression and what it is used for. For more in depth information on LD Score Regression please read the following three papers:</p>
<ol class="simple">
<li><p>“LD Score regression distinguishes confounding from polygenicity in genome-wide association studies” by Sullivan et al (2015)</p></li>
<li><p>“Partitioning heritability by functional annotation using genome-wide association summary statistics” by Finucane et al (2015)</p></li>
<li><p>“Linkage disequilibrium–dependent architecture of human complex traits shows action of negative selection” by Gazal et al (2017)</p></li>
</ol>
<p>As stated in Sullivan et al 2015, confounding factors and polygenic effects can cause inflated test statistics and other methods cannot distinguish between inflation from confounding bias and a true signal. LD Score Regression (LDSC) is a technique that aims to identify the impact of confounding factors and polygenic effects using information from GWAS summary statistics.</p>
<p>This approach involves using regression to mesaure the relationship between Linkage Disequilibrium (LD) scores and test statistics of SNPs from the GWAS summary statistics. Variants in LD with a “causal” variant show an elevation in test statistics in association analysis proportional to their LD (measured by <span class="math notranslate nohighlight">\(r^2\)</span>) with the causal variant within a certain window size (could be 1 cM, 1kB, etc.). In contrast, inflation from confounders such as population stratification that occur purely from genetic drift will not correlate with LD. For a polygenic trait, SNPs with a high LD score will have more significant χ2 statistics on average than SNPs with a low LD score. Thus, if we regress the <span class="math notranslate nohighlight">\(\chi^2\)</span> statistics from GWAS against LD Score, the intercept minus one is an estimator of the mean contribution of confounding bias to the inflation in the test statistics. The regression model is known as LD Score regression.</p>
<section id="ldsc-model">
<h3>LDSC model<a class="headerlink" href="#ldsc-model" title="Permalink to this headline">¶</a></h3>
<p>Under a polygenic assumption, in which effect sizes for variants are drawn independently from distributions with variance proportional to  <span class="math notranslate nohighlight">\(1/(p(1-p))\)</span> where p is the minor allele frequency (MAF), the expected <span class="math notranslate nohighlight">\(\chi^2\)</span> statistic of variant j is:</p>
<div class="math notranslate nohighlight">
\[E[\chi^2|l_j] = Nh^2l_j/M + Na + 1 \quad (1)\]</div>
<p>where <span class="math notranslate nohighlight">\(N\)</span> is the sample size; <span class="math notranslate nohighlight">\(M\)</span> is the number of SNPs, such that <span class="math notranslate nohighlight">\(h^2/M\)</span> is the average heritability explained per SNP; <span class="math notranslate nohighlight">\(a\)</span> measures the contribution of confounding biases, such as cryptic relatedness and population stratification; and <span class="math notranslate nohighlight">\(l_j = \sum_k r^2_{jk}\)</span> is the LD Score of variant <span class="math notranslate nohighlight">\(j\)</span>, which measures the amount of genetic variation tagged by <span class="math notranslate nohighlight">\(j\)</span>. A full derivation of this equation is provided in the Supplementary Note of Sullivan et al (2015). An alternative derivation is provided in Supplementary Note of Zhu and Stephens (2017) AoAS.</p>
<p>From this we can see that LD Score regression can be used to compute SNP-based heritability for a phenotype or trait, from GWAS summary statistics and does not require genotype information like other methods such as REML do.</p>
</section>
<section id="stratified-ldsc">
<h3>Stratified LDSC<a class="headerlink" href="#stratified-ldsc" title="Permalink to this headline">¶</a></h3>
<p>Heritability is the proportion of phenotypic variation (VP) that is due to variation in genetic values (VG) and thus can tell us how much of the difference in observed phenotypes in a sample is due to difference in genetics in the sample. It can also be extended to analyze partitioned heritability for a phenotype/trait split over categories.</p>
<p>For Partitioned Heritability or Stratified LD Score Regression (S-LDSC) more power is added to our analysis by leveraging LD Score information as well as using SNPs that haven’t reached Genome Wide Significance to partition heritability for a trait over categories which many other methods do not do.</p>
<p>S-LDSC relies on the fact that the <span class="math notranslate nohighlight">\(\chi^2\)</span> association statistic for a given SNP includes the effects of all SNPs tagged by this SNP meaning that in a region of high LD in the genome the given SNP from the GWAS represents the effects of a group of SNPs in that region.</p>
<p>S-LDSC determines that a category of SNPs is enriched for heritability if SNPs with high LD to that category have more significant <span class="math notranslate nohighlight">\(\chi^2\)</span> statistics than SNPs with low LD to that category.</p>
<p>Here, enrichment of a category is defined as the proportion of SNP heritability in the category divided by the proportion of SNPs in that category.</p>
<p>More precisely, under a polygenic model, the expected <span class="math notranslate nohighlight">\(\chi^2\)</span> statistic of SNP <span class="math notranslate nohighlight">\(j\)</span> is</p>
<div class="math notranslate nohighlight">
\[E[\chi^2_j] = N\sum_CT_Cl(j,C) + Na + 1 \quad (2)\]</div>
<p>where <span class="math notranslate nohighlight">\(N\)</span> is sample size, C indexes categories, <span class="math notranslate nohighlight">\(ℓ(j, C)\)</span> is the LD score of SNP j with respect to category <span class="math notranslate nohighlight">\(l(j,C) = \sum_{k\epsilon C} r^2_{jk}\)</span>, <span class="math notranslate nohighlight">\(a\)</span> is a term that measures the contribution of confounding biases, and if the categories are disjoint, <span class="math notranslate nohighlight">\(\tau_C\)</span> is the per-SNP heritability in category <span class="math notranslate nohighlight">\(C\)</span>; if the categories overlap, then the per-SNP heritability of SNP j is <span class="math notranslate nohighlight">\(\sum_{C:j\epsilon C} \tau_C\)</span>.  Equation 2 allows us to estimate <span class="math notranslate nohighlight">\(\tau_C\)</span> via a (computationally simple) multiple regression of <span class="math notranslate nohighlight">\(\chi^2\)</span> against <span class="math notranslate nohighlight">\(ℓ(j, C)\)</span>, for either a quantitative or case-control study.</p>
<p>To see how these methods have been applied to real world data as well as a further discussion on methods and comparisons to other methods please read the three papers listed at the top of the document.</p>
</section>
</section>
<section id="command-interface">
<h2>Command Interface<a class="headerlink" href="#command-interface" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="err">!</span><span class="n">sos</span> <span class="n">run</span> <span class="n">LDSC_Code</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run LDSC.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  make_annot
  munge_sumstats_no_sign
  munge_sumstats_sign
  calc_ld_score
  calc_enrichment

Sections
  make_annot:
    Workflow Options:
      --bed VAL (as str, required)
                        path to bed file
      --bim VAL (as str, required)
                        path to bim file
      --annot VAL (as str, required)
                        name of output annotation file
  munge_sumstats_no_sign: This option is for when the summary statistic file
                        does not contain a signed summary statistic (Z or Beta).
                        In this case,the program will calculate Z for you based
                        on A1 being the risk allele
    Workflow Options:
      --sumst VAL (as str, required)
                        path to summary statistic file
      --alleles &#39;w_hm3.snplist&#39;
                        path to Hapmap3 SNPs file, keep all columns (SNP, A1,
                        and A2) for the munge_sumstats program
      --output VAL (as str, required)
                        path to output file
  munge_sumstats_sign:  This option is for when the summary statistic file does
                        contain a signed summary statistic (Z or Beta)
    Workflow Options:
      --sumst VAL (as str, required)
                        path to summary statistic file
      --alleles &#39;w_hm3.snplist&#39;
                        path to Hapmap3 SNPs file, keep all columns (SNP, A1,
                        and A2) for the munge_sumstats program
      --output VAL (as str, required)
                        path to output file
  calc_ld_score:        Calculate LD Scores **Make sure to delete SNP,CHR, and
                        BP columns from annotation files if they are present
                        otherwise this code will not work. Before deleting, if
                        these columns are present, make sure that the annotation
                        file is sorted.**
    Workflow Options:
      --bim VAL (as str, required)
                        Path to bim file
      --annot-file VAL (as str, required)
                        Path to annotation File. Make sure to remove the SNP,
                        CHR, and BP columns from the annotation file if present
                        before running.
      --output VAL (as str, required)
                        name of output file
      --snplist &#39;w_hm3.snplist&#39;
                        path to Hapmap3 SNPs file, remove the A1 and A2 columns
                        for the Calculate LD Scores program
  calc_enrichment:
    Workflow Options:
      --sumstats VAL (as str, required)
                        Path to Summary statistics File
      --ref-ld VAL (as str, required)
                        Path to Reference LD Scores Files (Base Annotation +
                        Annotation you want to analyze, format like minimal
                        working example)
      --w-ld VAL (as str, required)
                        Path to LD Weight Files (Format like minimal working
                        example)
      --frq-file VAL (as str, required)
                        path to frequency files (Format like minimal working
                        example)
      --output VAL (as str, required)
                        Output name
</pre></div>
</div>
</div>
</div>
</section>
<section id="make-annotation-file">
<h2>Make Annotation File<a class="headerlink" href="#make-annotation-file" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">make_annot</span><span class="p">]</span>

<span class="c1"># Make Annotated Bed File</span>

<span class="c1"># path to bed file</span>
<span class="kn">parameter:</span> <span class="n">bed</span> <span class="o">=</span> <span class="nb">str</span> 
<span class="c1">#path to bim file</span>
<span class="kn">parameter:</span> <span class="n">bim</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#name of output annotation file</span>
<span class="kn">parameter:</span> <span class="n">annot</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">make_annot</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">bed</span><span class="o">-</span><span class="n">file</span> <span class="p">{</span><span class="n">bed</span><span class="p">}</span> <span class="o">--</span><span class="n">bimfile</span> <span class="p">{</span><span class="n">bim</span><span class="p">}</span> <span class="o">--</span><span class="n">annot</span><span class="o">-</span><span class="n">file</span> <span class="p">{</span><span class="n">annot</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="munge-summary-statistics-option-1-no-signed-summary-statistic">
<h2>Munge Summary Statistics (Option 1: No Signed Summary Statistic)<a class="headerlink" href="#munge-summary-statistics-option-1-no-signed-summary-statistic" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#This option is for when the summary statistic file does not contain a signed summary statistic (Z or Beta). </span>
<span class="c1">#In this case,the program will calculate Z for you based on A1 being the risk allele</span>
<span class="p">[</span><span class="n">munge_sumstats_no_sign</span><span class="p">]</span>



<span class="c1">#path to summary statistic file</span>
<span class="kn">parameter:</span> <span class="n">sumst</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#path to Hapmap3 SNPs file, keep all columns (SNP, A1, and A2) for the munge_sumstats program</span>
<span class="kn">parameter:</span> <span class="n">alleles</span> <span class="o">=</span> <span class="s2">&quot;w_hm3.snplist&quot;</span>
<span class="c1">#path to output file</span>
<span class="kn">parameter:</span> <span class="n">output</span> <span class="o">=</span> <span class="nb">str</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">munge_sumstats</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">sumstats</span> <span class="p">{</span><span class="n">sumst</span><span class="p">}</span> <span class="o">--</span><span class="n">merge</span><span class="o">-</span><span class="n">alleles</span> <span class="p">{</span><span class="n">alleles</span><span class="p">}</span> <span class="o">--</span><span class="n">out</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span> <span class="o">--</span><span class="n">a1</span><span class="o">-</span><span class="n">inc</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="munge-summary-statistics-option-2-no-signed-summary-statistic">
<h2>Munge Summary Statistics (Option 2: No Signed Summary Statistic)<a class="headerlink" href="#munge-summary-statistics-option-2-no-signed-summary-statistic" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># This option is for when the summary statistic file does contain a signed summary statistic (Z or Beta)</span>
<span class="p">[</span><span class="n">munge_sumstats_sign</span><span class="p">]</span>



<span class="c1">#path to summary statistic file</span>
<span class="kn">parameter:</span> <span class="n">sumst</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#path to Hapmap3 SNPs file, keep all columns (SNP, A1, and A2) for the munge_sumstats program</span>
<span class="kn">parameter:</span> <span class="n">alleles</span> <span class="o">=</span> <span class="s2">&quot;w_hm3.snplist&quot;</span>
<span class="c1">#path to output file</span>
<span class="kn">parameter:</span> <span class="n">output</span> <span class="o">=</span> <span class="nb">str</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">munge_sumstats</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">sumstats</span> <span class="p">{</span><span class="n">sumst</span><span class="p">}</span> <span class="o">--</span><span class="n">merge</span><span class="o">-</span><span class="n">alleles</span> <span class="p">{</span><span class="n">alleles</span><span class="p">}</span> <span class="o">--</span><span class="n">out</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="calculate-ld-scores">
<h2>Calculate LD Scores<a class="headerlink" href="#calculate-ld-scores" title="Permalink to this headline">¶</a></h2>
<p><strong>Make sure to delete SNP,CHR, and BP columns from annotation files if they are present otherwise this code will not work. Before deleting, if these columns are present, make sure that the annotation file is sorted.</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#Calculate LD Scores</span>
<span class="c1">#**Make sure to delete SNP,CHR, and BP columns from annotation files if they are present otherwise this code will not work. Before deleting, if these columns are present, make sure that the annotation file is sorted.**</span>
<span class="p">[</span><span class="n">calc_ld_score</span><span class="p">]</span>

<span class="c1">#Path to bim file</span>
<span class="kn">parameter:</span> <span class="n">bim</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#Path to annotation File. Make sure to remove the SNP, CHR, and BP columns from the annotation file if present before running.</span>
<span class="kn">parameter:</span> <span class="n">annot_file</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#name of output file</span>
<span class="kn">parameter:</span> <span class="n">output</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#path to Hapmap3 SNPs file, remove the A1 and A2 columns for the Calculate LD Scores program </span>
<span class="kn">parameter:</span> <span class="n">snplist</span> <span class="o">=</span> <span class="s2">&quot;w_hm3.snplist&quot;</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">ldsc</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">bfile</span> <span class="p">{</span><span class="n">bim</span><span class="p">}</span> <span class="o">--</span><span class="n">l2</span> <span class="o">--</span><span class="n">ld</span><span class="o">-</span><span class="n">wind</span><span class="o">-</span><span class="n">cm</span> <span class="mi">1</span> <span class="o">--</span><span class="n">annot</span> <span class="p">{</span><span class="n">annot_file</span><span class="p">}</span> <span class="o">--</span><span class="n">thin</span><span class="o">-</span><span class="n">annot</span> <span class="o">--</span><span class="n">out</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span> <span class="o">--</span><span class="nb">print</span><span class="o">-</span><span class="n">snps</span> <span class="p">{</span><span class="n">snplist</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="calculate-functional-enrichment-using-annotations">
<h2>Calculate Functional Enrichment using Annotations<a class="headerlink" href="#calculate-functional-enrichment-using-annotations" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#Calculate Enrichment Scores for Functional Annotations</span>
<span class="p">[</span><span class="n">calc_enrichment</span><span class="p">]</span>

<span class="c1">#Path to Summary statistics File</span>
<span class="kn">parameter:</span> <span class="n">sumstats</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#Path to Reference LD Scores Files (Base Annotation + Annotation you want to analyze, format like minimal working example)</span>
<span class="kn">parameter:</span> <span class="n">ref_ld</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#Path to LD Weight Files (Format like minimal working example)</span>
<span class="kn">parameter:</span> <span class="n">w_ld</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#path to frequency files (Format like minimal working example)</span>
<span class="kn">parameter:</span> <span class="n">frq_file</span> <span class="o">=</span> <span class="nb">str</span>
<span class="c1">#Output name</span>
<span class="kn">parameter:</span> <span class="n">output</span> <span class="o">=</span> <span class="nb">str</span>

<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="n">ldsc</span><span class="o">.</span><span class="n">py</span> <span class="o">--</span><span class="n">h2</span> <span class="p">{</span><span class="n">sumstats</span><span class="p">}</span> <span class="o">--</span><span class="n">ref</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="p">{</span><span class="n">ref_ld</span><span class="p">}</span> <span class="o">--</span><span class="n">w</span><span class="o">-</span><span class="n">ld</span><span class="o">-</span><span class="nb">chr</span> <span class="p">{</span><span class="n">w_ld</span><span class="p">}</span> <span class="o">--</span><span class="n">overlap</span><span class="o">-</span><span class="n">annot</span> <span class="o">--</span><span class="n">frqfile</span><span class="o">-</span><span class="nb">chr</span> <span class="p">{</span><span class="n">frq_file</span><span class="p">}</span> <span class="o">--</span><span class="n">out</span> <span class="p">{</span><span class="n">output</span><span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./code/integrative_analysis/enrichment"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="gwas_enrichment.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Enrichment analysis: TORUS and GREGOR</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="../SuSiE_Ann/polyfun.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Fine-mapping with PolyFun</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>