{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9829cdaa-6ec2-450a-8b79-39e054e92d6a",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "# Fine-mapping for Summary Statistics"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c41fa20-1447-47f3-a518-f563fe0dfc1a",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Generation of LD (Skipable)\n",
    "If there is no LD matrixes provided, then in-Sample LD matrix shall be generated by our pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5475f16c-6f7f-46da-bbbc-ee9bf90d03c5",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "cd /mnt/vast/hpc/csg/molecular_phenotype_calling/LD\n",
    "\n",
    "sos run pipeline/genotype_formatting.ipynb ld_by_region_plink \\\n",
    "    --region_list ~/1300_hg38_EUR_LD_blocks.tsv --cwd output  \\\n",
    "    --genoFile ../genotype/ROSMAP_NIA_WGS.leftnorm.filtered.filtered.bed \\\n",
    "    --container ../containers/bioinfo.sif"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef5bc1b1-1937-47c8-af2a-e5fd55c05647",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Standardized of sumstat\n",
    "We should standardized our sumstat, and fix potential allele filp issue against the LD matrixs. After downloading the ADGWAS data, some processing are needed. Please refer to this [notebook](https://github.com/cumc/fungen-xqtl-analysis/blob/main/analysis/Wang_Columbia/GWAS/AD_GWAS_processing.ipynb) for processing that are not implemented in the pipeline to generate a yml list that guides the standardization of the sumstat.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5178c35b-de69-4b89-bc1f-6498d0b636a2",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "cd /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS\n",
    "sos run pipeline/yml_generator.ipynb yml_list \\\n",
    "    --sumstat-list ADGWAS2022.recipe  \\\n",
    "    --cwd  data_intergration_new/ADGWAS2022   &\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "79434f32-13e5-4154-a587-deaf8b225103",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "This is to create a reference data for the standardization based on our fasta files. IF an external LD panel are used, please manually extract the chr,pos,ref,alt of the external LD in the same format of our TARGET file. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39677156-fea8-4995-a7b4-131a4b987bd1",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run  pipeline/summary_stats_standardizer.ipynb   TARGET_generation  \\\n",
    "      --sumstat-list /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/qced_sumstat_list.txt    \\\n",
    "      --yml-list /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/yml_list.txt     \\\n",
    "      --fasta /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta \\\n",
    "      --cwd data_intergration/ADGWAS2022/  -J 22 -c /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/csg.yml -q csg2 --mem 50G --walltime 48h &\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50932a9f-404c-413f-8abc-b408bc26583a",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "Once the TARGET is availble, the standardizaion can begun. This is the same as the sumstat_standardization step that generates our QTL results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22c3593f-7c60-4a61-bc21-7dc1e23f5792",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run  pipeline/summary_stats_standardizer.ipynb   sumstat_standardization  \\\n",
    "      --sumstat-list /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/qced_sumstat_list.txt    \\\n",
    "      --yml-list /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/yml_list.txt     \\\n",
    "      --fasta /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/reference_data/GRCh38_full_analysis_set_plus_decoy_hla.noALT_noHLA_noDecoy_ERCC.fasta \\\n",
    "      --TARGET_list data_intergration/ADGWAS2022/TARGET.ref.list \\\n",
    "      --cwd data_intergration/ADGWAS2022/  -J 22 -c /mnt/vast/hpc/csg/xqtl_workflow_testing/finalizing/csg.yml -q csg2 --mem 50G --walltime 48h &"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16592af5-b776-4845-9703-b16296fe5e04",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## SuSiE RSS\n",
    "For both xQTL sumstat or the GWAS sumstat, as long as they are the output of sumstat_standardization, they can be fed into the susie_rss module directly. The fine mappling implembeted at the moment will be based on LD npz files that are from the first step of this notebook. If external LD are used, they would need to be splited as such."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2bf54611-f790-4049-88d2-c53db8af0816",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "cd /mnt/vast/hpc/csg/xqtl_workflow_testing/susie_rss\n",
    "sos run pipeline/SuSiE_RSS.ipynb SuSiE_RSS \\\n",
    "    --LD_list test.ld.list \\\n",
    "    --sumstat_list /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022/qced_sumstat_list.txt \\\n",
    "    --container containers/stephenslab.sif --impute --cwd output_impute_2 &"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53df1664-e9f5-434e-b75e-7ae23eb5a8c8",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## SuSiE results post processing\n",
    "The default output of susie or susie RSS are rds object, which are hard to intergrate. Therefore we have a utlity module to extract and save the info into three types of tables. One of the table, the lbf one, will have all the information needed for coloc analysis. For eQTL, the region list are the one used to do fine-mapping. For sumstat, the region list are the one used to call LD. When the input RDS have chr{1ï¼š22} in its name, their LBF table will be rbind into a lbf for the full chromosome."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ba889573-2e63-412a-a6d7-e24536920d58",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "cd /mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl\n",
    "\n",
    "sos run pipeline/SuSiE_post_processing.ipynb susie_to_tsv \\\n",
    "    --cwd output/test --rds_path `ls output/test/cache/*rds | head ` \\\n",
    "    --region-list <(head -50  ./dlpfc_region_list) --container containers/stephenslab.sif "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "71f1bacd-cb2b-4cd1-966e-7e4b28e0e8ea",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "cd /mnt/vast/hpc/csg/xqtl_workflow_testing/susie_rss\n",
    "\n",
    "sos run pipeline/SuSiE_post_processing.ipynb susie_to_tsv \\\n",
    "    --cwd output/ADGWAS_finemapping_extracted --rds_path `ls GWAS_Finemapping_Results/Bellenguez/ADG_*rds ` \\\n",
    "    --region-list ~/1300_hg38_EUR_LD_blocks_orig.tsv \\\n",
    "    --container containers/stephenslab.sif \n",
    "\n",
    "sos run pipeline/SuSiE_post_processing.ipynb susie_tsv_collapse \\\n",
    "    --cwd output/ADGWAS_finemapping_extracted --tsv_path `ls output/ADGWAS_finemapping_extracted/*lbf.tsv` \\\n",
    "    --container containers/stephenslab.sif "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a65421d3-f45b-4f6b-9fb5-b9a9e0fca932",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Coloc analysis\n",
    "The qtl_tsv are the lbf tsv who are not rbind into per chromosome, while the sumstat_tsv are per chromosome and have their chromosome it its name. The requirement is crucial for padding"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad6cd484-4618-4433-8ae2-a2b391463407",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run pipeline/coloc.ipynb coloc \\\n",
    "    --qtl_tsv `ls /mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl/output/susie_per_gene_tad/*f.tsv`   \\\n",
    "    --sumstat_tsv `ls output/ADGWAS_finemapping_extracted/*chr17.unisusie_rss.lbf.tsv` \\\n",
    "    --region_list test.region_list "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ]
   ],
   "version": "0.23.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
