

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>MASH analysis pipeline with posterior computation &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/multivariate_genome/MASH/mash_posterior';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="watershed analysis for rare xQTL" href="../../rare_xqtl/watershed.html" />
    <link rel="prev" title="MASH analysis pipeline with data-driven prior matrices" href="mash_fit.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../reference_data/reference_data.html">Reference Data</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/qtl_association_testing.html">QTL Association Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced cis-QTL Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/SuSiE.html">SuSiE fine-mapping workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/fSuSiE.html">fSuSiE workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mvSuSiE.html">Multivariate fine-mapping workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mrmash.html">Multivariate TWAS prediction model</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced Genome-wide Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../prototype_drafts/polyfun.html">Fine-mapping with PolyFun</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../prototype_drafts/MRAID_QTL.html">Mendelian Randomization using MRAID</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/ldsc.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL-GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../pecotmr_integration/pecotmr.html">The PECOTMR intergration framework</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../pecotmr_integration/SuSiE_enloc.html">GWAS integration: enrichment and colocalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../pecotmr_integration/twas_mr.html">GWAS integration: TWAS and MR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../pecotmr_integration/intact.html">TWAS and COLOC integration</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../MASH_pipeline.html">Multivariate association with MASH</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="mash_preprocessing.html">Extract genome-wide data for multivariate analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">MASH analysis pipeline with posterior computation</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Rare xQTL</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../rare_xqtl/watershed.html">watershed analysis for rare xQTL</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Post Processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../post_processing/sumstats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../post_processing/rds_to_vcf.html">Summary statistics in VCF format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../post_processing/association_scan_post_processing.html">Association result processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../post_processing/fine_mapping_post_processing.html">SuSIE results post process</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/edit/gh-pages/code/multivariate_genome/MASH/mash_posterior.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/multivariate_genome/MASH/mash_posterior.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/multivariate_genome/MASH/mash_posterior.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>MASH analysis pipeline with posterior computation</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example">example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior">posterior</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#contrast-analysis-works">contrast analysis: works</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-results">Posterior results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Posterior</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-contrast">Posterior contrast</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-score-from-contrast-results">Feature score from contrast results</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-score-meta">feature_score_meta</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-score-finemap">feature_score_finemap</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-score-nsig">feature_score_nsig</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-pval-pair">feature_pval_pair</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="mash-analysis-pipeline-with-posterior-computation">
<h1>MASH analysis pipeline with posterior computation<a class="headerlink" href="#mash-analysis-pipeline-with-posterior-computation" title="Permalink to this heading">#</a></h1>
<p>In the GTEx V6 paper we assumed one eQTL per gene and applied the model learned above to those SNPs. Under that assumption, the input data for posterior calculation will be the <code class="docutils literal notranslate"><span class="pre">dat$strong.*</span></code> matrices.
It is a fairly straightforward procedure as shown in <a class="reference external" href="https://stephenslab.github.io/mashr/articles/eQTL_outline.html">this vignette</a>.</p>
<p>But it is often more interesting to apply MASH to given list of eQTLs, eg, from those from fine-mapping results. In GTEx V8 analysis we obtain such gene-SNP pairs from DAP-G fine-mapping analysis. See <a class="reference external" href="https://stephenslab.github.io/gtex-eqtls/analysis/Independent_eQTL_Results.html">this notebook</a> for how the input data is prepared. The workflow below takes a number of input chunks (each chunk is a list of matrices <code class="docutils literal notranslate"><span class="pre">dat$Bhat</span></code> and <code class="docutils literal notranslate"><span class="pre">dat$Shat</span></code>)
and computes posterior for each chunk. It is therefore suited for running in parallel posterior computation for all gene-SNP pairs, if input data chunks are provided.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>JOB_OPT=&quot;-c midway2.yml -q midway2&quot;
DATA_DIR=/project/compbio/GTEx_eQTL/independent_eQTL
sos run workflows/mashr_flashr_workflow.ipynb posterior \
    $JOB_OPT \
    --posterior-input $DATA_DIR/DAPG_pip_gt_0.01-AllTissues/DAPG_pip_gt_0.01-AllTissues.*.rds \
                      $DATA_DIR/ConditionalAnalysis_AllTissues/ConditionalAnalysis_AllTissues.*.rds
</pre></div>
</div>
<section id="example">
<h2>example<a class="headerlink" href="#example" title="Permalink to this heading">#</a></h2>
<section id="posterior">
<h3>posterior<a class="headerlink" href="#posterior" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cd</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">Multivariate</span><span class="o">/</span><span class="n">MASH</span><span class="o">/</span><span class="n">MWE</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#slice</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_posterior</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">posterior</span>  \
    <span class="o">--</span><span class="n">analysis</span><span class="o">-</span><span class="n">units</span> <span class="o">&lt;</span><span class="p">(</span><span class="n">cat</span>  <span class="n">MWE</span><span class="o">.</span><span class="n">list</span><span class="o">|</span> <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">2</span><span class="p">)</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">MWE_udr</span>  \
    <span class="o">--</span><span class="n">output_prefix</span>  <span class="n">MWE_udr</span> \
    <span class="o">--</span><span class="n">vhat</span> <span class="n">mle</span> \
    <span class="o">--</span><span class="n">mash_model</span> <span class="n">MWE_udr</span><span class="o">/</span><span class="n">MWE_udr</span><span class="o">.</span><span class="n">EZ</span><span class="o">.</span><span class="n">V_mle</span><span class="o">.</span><span class="n">mash_model</span><span class="o">.</span><span class="n">rds</span> \
    <span class="o">--</span><span class="n">slice_method</span> <span class="kc">True</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">stephenslab</span><span class="o">.</span><span class="n">sif</span> 
</pre></div>
</div>
</div>
</div>
</section>
<section id="contrast-analysis-works">
<h3>contrast analysis: works<a class="headerlink" href="#contrast-analysis-works" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># no group</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_posterior</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">mash_posterior_contrast</span>   \
    <span class="o">--</span><span class="n">posterior_file</span> <span class="n">MWE_udr</span><span class="o">/</span><span class="n">mash_output_list_all</span> \
    <span class="o">--</span><span class="n">sum_file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">list</span>  \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">test</span> \
    <span class="o">--</span><span class="n">cells</span> <span class="s2">&quot;Ast&quot;</span><span class="p">,</span><span class="s2">&quot;Exc&quot;</span><span class="p">,</span><span class="s2">&quot;Inh&quot;</span><span class="p">,</span><span class="s2">&quot;Mic&quot;</span><span class="p">,</span><span class="s2">&quot;OPC&quot;</span><span class="p">,</span><span class="s2">&quot;Oli&quot;</span><span class="p">,</span><span class="s2">&quot;DLPFC_pQTL&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GFM&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GTS&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_SVZ&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_THA&quot;</span> 
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># 2 group</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_posterior</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">mash_posterior_contrast</span>   \
    <span class="o">--</span><span class="n">posterior_file</span> <span class="n">MWE_udr</span><span class="o">/</span><span class="n">mash_output_list_all</span> \
    <span class="o">--</span><span class="n">sum_file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">list</span>  \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">test_2group</span> \
    <span class="o">--</span><span class="n">cells</span> <span class="s2">&quot;Ast&quot;</span><span class="p">,</span><span class="s2">&quot;Exc&quot;</span><span class="p">,</span><span class="s2">&quot;Inh&quot;</span><span class="p">,</span><span class="s2">&quot;Mic&quot;</span><span class="p">,</span><span class="s2">&quot;OPC&quot;</span><span class="p">,</span><span class="s2">&quot;Oli&quot;</span><span class="p">,</span><span class="s2">&quot;DLPFC_pQTL&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GFM&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GTS&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_SVZ&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_THA&quot;</span> \
    <span class="o">--</span><span class="n">group1</span> <span class="s2">&quot;Mic&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GFM&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GTS&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_SVZ&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_THA&quot;</span> \
    <span class="o">--</span><span class="n">group2</span> <span class="s2">&quot;Ast&quot;</span><span class="p">,</span><span class="s2">&quot;Exc&quot;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># recipe with 2 group</span>
<span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mash_posterior</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">mash_posterior_contrast</span>   \
    <span class="o">--</span><span class="n">posterior_file</span> <span class="n">MWE_udr</span><span class="o">/</span><span class="n">mash_output_list_all</span> \
    <span class="o">--</span><span class="n">sum_file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">list</span>  \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">test_2group_recipe</span> \
    <span class="o">--</span><span class="n">cells</span> <span class="s2">&quot;Ast&quot;</span><span class="p">,</span><span class="s2">&quot;Exc&quot;</span><span class="p">,</span><span class="s2">&quot;Inh&quot;</span><span class="p">,</span><span class="s2">&quot;Mic&quot;</span><span class="p">,</span><span class="s2">&quot;OPC&quot;</span><span class="p">,</span><span class="s2">&quot;Oli&quot;</span><span class="p">,</span><span class="s2">&quot;DLPFC_pQTL&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GFM&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_GTS&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_SVZ&quot;</span><span class="p">,</span><span class="s2">&quot;MiGA_THA&quot;</span> \
    <span class="o">--</span><span class="n">grouping_recipe</span> <span class="n">recipe</span>
</pre></div>
</div>
</div>
</div>
<p>group recipe looks like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Ast</span><span class="p">,</span><span class="n">Exc</span>
<span class="n">Mic</span><span class="p">,</span><span class="n">MiGA_GFM</span><span class="p">,</span><span class="n">MiGA_GTS</span>
<span class="n">Oli</span><span class="p">,</span><span class="n">DLPFC_pQTL</span>
<span class="c1"># ... add more groups as needed, separated by commas</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="c1"># Work directory &amp; output directory</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;./&#39;</span><span class="p">)</span>
<span class="c1"># The filename prefix for output data</span>
<span class="kn">parameter:</span> <span class="kp">name</span><span class="o">=</span><span class="s2">&quot;test&quot;</span>
<span class="kn">parameter:</span> <span class="n">cells</span> <span class="o">=</span> <span class="p">[</span><span class="s2">&quot;Ast&quot;</span><span class="p">,</span><span class="s2">&quot;Exc&quot;</span><span class="p">,</span><span class="s2">&quot;Inh&quot;</span><span class="p">,</span><span class="s2">&quot;Mic&quot;</span><span class="p">,</span><span class="s2">&quot;OPC&quot;</span><span class="p">,</span><span class="s2">&quot;Oli&quot;</span><span class="p">,</span><span class="s2">&quot;DLPFC_pQTL&quot;</span><span class="p">]</span><span class="c1">#order is important</span>
<span class="kn">parameter:</span> <span class="n">group1</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">group2</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">group3</span> <span class="o">=</span> <span class="p">[]</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="c1"># handle N = per_chunk data-set in one job</span>
<span class="kn">parameter:</span> <span class="n">per_chunk</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1">###add for test</span>
<span class="kn">parameter:</span> <span class="n">output_prefix</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">parameter:</span> <span class="n">output_suffix</span> <span class="o">=</span> <span class="s1">&#39;all&#39;</span>
<span class="c1"># Exchangable effect (EE) or exchangable z-scores (EZ)</span>
<span class="kn">parameter:</span> <span class="n">effect_model</span> <span class="o">=</span> <span class="s1">&#39;EZ&#39;</span>
<span class="c1"># Identifier of $\hat{V}$ estimate file</span>
<span class="c1"># Options are &quot;identity&quot;, &quot;simple&quot;, &quot;mle&quot;, &quot;vhat_corshrink_xcondition&quot;, &quot;vhat_simple_specific&quot;</span>
<span class="kn">parameter:</span> <span class="n">vhat</span> <span class="o">=</span> <span class="s1">&#39;simple&#39;</span>
<span class="kn">parameter:</span> <span class="n">data</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;fastqtl_to_mash_output/FastQTLSumStats.mash.rds&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">p_cut</span><span class="o">=</span><span class="mf">0.00001</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>

<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">cwd</span><span class="o">.</span><span class="n">absolute</span><span class="p">()</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">output_prefix</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">output_prefix</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">data</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="n">vhat_data</span> <span class="o">=</span> <span class="n">file_target</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">effect_model</span><span class="si">}</span><span class="s2">.V_</span><span class="si">{</span><span class="n">vhat</span><span class="si">}</span><span class="s2">.rds&quot;</span><span class="p">)</span>
<span class="n">mash_model</span> <span class="o">=</span> <span class="n">file_target</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">effect_model</span><span class="si">}</span><span class="s2">.V_</span><span class="si">{</span><span class="n">vhat</span><span class="si">}</span><span class="s2">.mash_model.rds&quot;</span><span class="p">)</span>
<span class="n">posterior_list</span> <span class="o">=</span> <span class="n">file_target</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">output_prefix</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">effect_model</span><span class="si">}</span><span class="s2">.</span><span class="si">{</span><span class="n">output_suffix</span><span class="si">}</span><span class="s2">.posterior_list&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">sort_uniq</span><span class="p">(</span><span class="n">seq</span><span class="p">):</span>
    <span class="n">seen</span> <span class="o">=</span> <span class="nb">set</span><span class="p">()</span>
    <span class="k">return</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">seq</span> <span class="k">if</span> <span class="ow">not</span> <span class="p">(</span><span class="n">x</span> <span class="ow">in</span> <span class="n">seen</span> <span class="ow">or</span> <span class="n">seen</span><span class="o">.</span><span class="n">add</span><span class="p">(</span><span class="n">x</span><span class="p">))]</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="posterior-results">
<h2>Posterior results<a class="headerlink" href="#posterior-results" title="Permalink to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>The outcome of the <code class="docutils literal notranslate"><span class="pre">[posterior]</span></code> step should produce a number of serialized R objects <code class="docutils literal notranslate"><span class="pre">*.batch_*.posterior.rds</span></code> (can be loaded to R via <code class="docutils literal notranslate"><span class="pre">readRDS()</span></code>) â€“ I chopped data to batches to take advantage of computing in multiple cluster nodes. It should be self-explanary but please let me know otherwise.</p></li>
<li><p>Other posterior related files are:</p>
<ol class="arabic simple">
<li><p><code class="docutils literal notranslate"><span class="pre">*.batch_*.yaml</span></code>: gene-SNP pairs of interest, identified elsewhere (eg. fine-mapping analysis).</p></li>
<li><p>The corresponding univariate analysis summary statistics for gene-SNPs from <code class="docutils literal notranslate"><span class="pre">*.batch_*.yaml</span></code> are extracted and saved to <code class="docutils literal notranslate"><span class="pre">*.batch_*.rds</span></code>, creating input to the <code class="docutils literal notranslate"><span class="pre">[posterior]</span></code> step.</p></li>
<li><p>Note the <code class="docutils literal notranslate"><span class="pre">*.batch_*.stdout</span></code> file documents some SNPs found in fine-mapping results but not found in the original <code class="docutils literal notranslate"><span class="pre">fastqtl</span></code> output.</p></li>
</ol>
</li>
</ol>
</section>
<section id="id1">
<h2>Posterior<a class="headerlink" href="#id1" title="Permalink to this heading">#</a></h2>
<p>take all the 13K genes,
if <code class="docutils literal notranslate"><span class="pre">slice_method</span> <span class="pre">=</span> <span class="pre">True</span></code> for those with missing conditions we just drop those corresponding rows and cols in the prior model</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Apply posterior calculations with slice NA and set NaN/Inf 0/1E3, output_posterior_cov = T 
[posterior_1]
parameter: analysis_units = path
regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_units).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]
parameter: mash_model = path()
mash_model = mash_model.absolute()
parameter: posterior_input = [path(x[0]) for x in regions]
parameter: posterior_vhat_files = paths()
# eg, if data is saved in R list as data$strong, then
# when you specify `--data-table-name strong` it will read the data as
# readRDS(&#39;{_input:r}&#39;)$strong
parameter: data_table_name = &#39;&#39;
parameter: bhat_table_name = &#39;bhat&#39;
parameter: shat_table_name = &#39;sbhat&#39;
parameter: per_chunk = &#39;100&#39;
##  conditions can be excluded if needs arise. If nothing to exclude keep the default 0
parameter: exclude_condition = [&quot;1&quot;,&quot;3&quot;]

parameter: slice_method = False 
skip_if(len(posterior_input) == 0, msg = &quot;No posterior input data to compute on. Please specify it using --posterior-input.&quot;)
fail_if(len(posterior_vhat_files) &gt; 1 and len(posterior_vhat_files) != len(posterior_input), msg = &quot;length of --posterior-input and --posterior-vhat-files do not agree.&quot;)
for p in posterior_input:
    fail_if(not p.is_file(), msg = f&#39;Cannot find posterior input file ``{p}``&#39;)

input: posterior_input, group_by = per_chunk
output: f&quot;{cwd}/cache/mash_output_list_{_index+1}&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;  
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container, entrypoint=entrypoint
    library(mashr)
    library(dplyr)
    library(stringr)
    #library(ttt)
    handle_nan_etc = function(x) {
      x$bhat[which(is.nan(x$bhat))] = 0
      x$sbhat[which(is.nan(x$sbhat) | is.infinite(x$sbhat))] = 1E3
      return(x)
    }
    # Slice matrices
    slice_and_update_data &lt;- function(data, vhat, snps, samples) {
        data$bhat &lt;- data$bhat[snps, samples] %&gt;% as.matrix
        data$sbhat &lt;- data$sbhat[snps, samples] %&gt;% as.matrix
        data$Z &lt;- data$Z[snps, samples] %&gt;% as.matrix
        vhat &lt;- vhat[samples, samples] %&gt;% as.matrix

        # Filter SNPs and update column names
        data$snp &lt;- data$snp[data$snp %in% snps]
        colnames(data$bhat) &lt;- colnames(data$sbhat) &lt;- colnames(data$Z) &lt;- colnames(vhat) &lt;- samples

        return(list(data = data, vhat = vhat))
    }
  
    # Remove covariance matrices that are not needed
    remove_unnecessary_cov_matrices &lt;- function(cov, all_samples, samples) {
      unwanted_samples &lt;- setdiff(all.samples, samples)
      for (d in names(cov)) {
        if (d %in% unwanted_samples || d %in% paste0(&quot;ED_&quot;, unwanted_samples)) {
          cov[[d]] &lt;- NULL
        }
      }
      return(cov)
    }

    # Update or adjust the covariance matrices
    adjust_cov_matrices &lt;- function(cov, samples) {
      for (d in names(cov)) {
        if (d %in% samples) {
          cov[[d]] &lt;- matrix(0, length(samples), length(samples))
          cov[[d]][which(samples == d), which(samples == d)] &lt;- 1
        } else if (d == &quot;identity&quot;) {
          cov[[d]] &lt;- matrix(0, length(samples), length(samples))
          cov[[d]][1, 1] &lt;- 1  
        } else if (is.null(colnames(cov[[d]]))) {
          cov[[d]] &lt;- cov[[d]][1:length(samples), 1:length(samples)]
        } else {
          cov[[d]] &lt;- cov[[d]][samples, samples]
        }
        cov[[d]] &lt;- as.matrix(cov[[d]])
      }
      return(cov)
    }

    # Main function to update the covariance in the MASH model
    update_mash_model_cov &lt;- function(mash_model, all_samples, samples) {
      cov &lt;- mash_model$fitted_g$Ulist
      # Remove matrices that are not required
      cov &lt;- remove_unnecessary_cov_matrices(cov, all_samples, samples)
  
      # Update or reshape the covariance matrices
      cov &lt;- adjust_cov_matrices(cov, samples)
  
      # Update the covariance matrices in the model
      mash_model$fitted_g$Ulist &lt;- cov
  
      # Update the &#39;pi&#39; attribute of the model
      unwanted_samples &lt;- setdiff(all.samples, samples)
      for (s in unwanted_samples) {
        mash_model$fitted_g$pi &lt;- mash_model$fitted_g$pi[-grep(s, names(mash_model$fitted_g$pi))]
      }

      return(mash_model)
    }
  
    outlist = data.frame()
    for (f in c(${_input:r,})) try({

      data = readRDS(f)${(&#39;$&#39; + data_table_name) if data_table_name else &#39;&#39;}
      data &lt;- handle_nan_etc(data)

      if(c(${&quot;,&quot;.join(exclude_condition)})[1] &gt; 0 ){
        message(paste(&quot;Excluding condition ${exclude_condition} from the analysis&quot;))
        data$bhat = data$bhat[,-c(${&quot;,&quot;.join(exclude_condition)})]
        data$sbhat = data$sbhat[,-c(${&quot;,&quot;.join(exclude_condition)})]
        data$Z = data$Z[,-c(${&quot;,&quot;.join(exclude_condition)})]
      }

      vhat = readRDS(&quot;${vhat_data if len(posterior_vhat_files) == 0 else posterior_vhat_files[_index]}&quot;)
      mash_model &lt;- readRDS(&quot;${mash_model}&quot;)
  
      slice_method &lt;- ${&#39;TRUE&#39; if slice_method else &#39;FALSE&#39;}
      if(slice_method){
        # All additional operations from the second script go here

        all.samples &lt;- colnames(data$bhat)
        all.snps &lt;- rownames(data$bhat)    

        #remove the rows and cols containing NA
        na.test &lt;- data$bhat %&gt;% as.data.frame %&gt;% select_if(~any(!is.na(.))) %&gt;% na.omit %&gt;% as.matrix

        #recording meaningful rows and cols
        samples &lt;- colnames(na.test)
        snps &lt;- rownames(na.test)

        if(length(all.snps)!=length(snps) | length(all.samples)!=length(samples)){
            # slice data matrix
            data &lt;- slice_and_update_data(data, vhat, snps, samples)

            if(length(all.samples)!=length(samples)){
                ##slice the prior
                mash_model &lt;- update_mash_model_cov(mash_model, all_samples, samples)
            }
        }
      }

      mash_data = mash_set_data(data$${bhat_table_name}, Shat=data$${shat_table_name}, alpha=${1 if effect_model == &#39;EZ&#39; else 0}, V=vhat, zero_Bhat_Shat_reset = 1E3)
      mash_output = mash_compute_posterior_matrices(mash_model, mash_data, output_posterior_cov=TRUE)
      mash_output$snps = data$snps
      samplename &lt;- str_split(f, &quot;/&quot;, simplify = T) %&gt;% .[length(.)] %&gt;% gsub(&#39;.rds&#39;, &#39;&#39;, .)
      saveRDS(mash_output, paste0(&quot;${_output:d}&quot;, &quot;/&quot;, samplename, &quot;.posterior.rds&quot;))
      outlist &lt;- rbind(outlist, paste0(&quot;${_output:d}&quot;, &quot;/&quot;, samplename, &quot;.posterior.rds&quot;))

    })
    write.table(outlist, ${_output:r}, col.names=F, row.names=F, quote=F)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[posterior_2]
input: group_by = &quot;all&quot;
output:f&quot;{cwd}/mash_output_list_{output_suffix}&quot;
bash: expand =&#39;${ }&#39;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;
     cd ${_input[0]:d}
     cat mash_output_list_*[0-9] &gt;&gt; posterior_file_list
     awk -F &#39;cis_long_table.&#39; &#39;{print $2}&#39; posterior_file_list| awk -F &#39;.posterior.rds&#39; &#39;{print $1}&#39;|paste - posterior_file_list &gt; ${_output:r}
     rm posterior_file_list
</pre></div>
</div>
</div>
</div>
</section>
<section id="posterior-contrast">
<h2>Posterior contrast<a class="headerlink" href="#posterior-contrast" title="Permalink to this heading">#</a></h2>
<ul>
<li><p>Add group in this module</p>
<p>e.g. add MiGA-Microglia-scRNA data to this analysis to supplement the Mic cell count shortage, but the MiGA data comes from four sites with four datasets, and here I handled it in such a way that the different Micâ€™s were always in the same state during the deviation contrast analysis. If Mic is the one be compared, each Mic sample would be set as (n_populations - 1)/n_Mic instead of n_populations</p>
</li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># perform mash posterior contrast for sliced data
[mash_posterior_contrast_1]
parameter: grouping_recipe = &quot;&quot;
parameter: posterior_file = path
parameter: sum_file = path

# Extract data from posterior_file
paths_posterior = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split()[1] for x in open(posterior_file).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]
# Create a dictionary from sum_file for quick lookup
dict_sum = dict([(x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split()[0], x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split()[1]) for x in open(sum_file).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)])
# Use genes from posterior_file to fetch corresponding paths from sum_file
paths_sum = [dict_sum[x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split()[0]] for x in open(posterior_file).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]

input: paths_posterior, paired_with=&#39;paths_sum&#39;, group_by=1
output: f&quot;{cwd}/{_input:bnn}_posterior_contrast.rds&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;  
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container, entrypoint=entrypoint
    # Load necessary libraries
    library(mashr)
    library(RhpcBLASctl)
    library(magrittr)
    library(tidyverse)
    #library(ttt)

    # Set number of threads for BLAS operations
    blas_set_num_threads(1)

    # Create a function for pairwise contrast columns
    MakePairwiseContrastCols &lt;- function(contrast_left, orig_vector) {
        orig_vector[contrast_left[1]] &lt;- 1
        orig_vector[contrast_left[2]] &lt;- -1
        orig_vector
    }

    # Function to fit contrast data
    FitContrast &lt;- function(index, orig_mean, posterior_mean, posterior_vcov) {
        population_names &lt;- colnames(posterior_mean) %&gt;% str_remove_all(&quot;BETA_&quot;)

        orig_mean_vector &lt;- orig_mean[index,]
        names(orig_mean_vector) &lt;- population_names
        orig_mean_nonzero &lt;- as.vector(orig_mean_vector != 0)
        orig_mean_tested &lt;- names(orig_mean_vector[orig_mean_nonzero])
        
        if(length(orig_mean_tested)&gt;0){
            n_populations &lt;- length(orig_mean_tested)

            pairwise_vector &lt;- rep(0, n_populations)
            names(pairwise_vector) &lt;- orig_mean_tested

            grouping &lt;- grouping_all[orig_mean_tested]
            if (n_populations &gt; 1) {
                if (n_populations &gt; 2) {
                    #####1. deviation contrast
                    deviation_contrasts &lt;- rep(-1, n_populations^2) %&gt;% matrix(nrow = n_populations, ncol = n_populations)
                    diag(deviation_contrasts) &lt;- n_populations - 1
                    rownames(deviation_contrasts) &lt;- orig_mean_tested
                    colnames(deviation_contrasts) &lt;- orig_mean_tested
                    deviation_contrasts_tested &lt;- deviation_contrasts[, orig_mean_tested]

                    unique_groups &lt;- unique(grouping)
                    for (grp in unique_groups[unique_groups &gt; 0]) {
                       #same celltype (e.g. MIC) with different populations would get 1/n for their weight,
                        diag(deviation_contrasts_tested)[grouping == grp] &lt;- (n_populations - 1) / length(grouping[grouping == grp])
                        deviation_contrasts_tested[grouping == grp, grouping == grp] &lt;- (n_populations - 1) / length(grouping[grouping == grp])
                    }

                    colnames(deviation_contrasts_tested) %&lt;&gt;% str_c(&quot;_deviation&quot;)

                    ####2. pairwise contrast
                    two_combn &lt;- combn(orig_mean_tested, m = 2)
                    pairwise_names &lt;- apply(two_combn, 2, str_c, collapse = &quot;_vs_&quot;)
                    pairwise_contrast &lt;- apply(two_combn, 2, MakePairwiseContrastCols, pairwise_vector)

                    colnames(pairwise_contrast) &lt;- pairwise_names

                    # Create a new matrix to store the adjusted values
                    pairwise_contrast_new &lt;- pairwise_contrast

                    # Loop through each column to archieve such goal: e.g.
                    # microglia populations would get 1/n_Mic for their weight,
                    # and Mic vs Mic would still be 1 vs -1 to estimate the internal difference among microglia datasets
                    for (col in colnames(pairwise_contrast)) {
                      # Split column names to get group names
                      groups &lt;- strsplit(col, &quot;_vs_&quot;)[[1]]

                      # Get the grouping values for the two groups
                      group_values &lt;- grouping[names(grouping) %in% groups]

                      # Identify groups with non-zero grouping values
                      relevant_groups &lt;- names(group_values[group_values &gt; 0])

                      # Check if there are multiple distinct groups
                      if (length(unique(group_values)) &gt; 1 &amp;&amp; length(relevant_groups) &gt; 0) {
                        distinct_groups &lt;- unique(group_values[group_values &gt; 0])

                        for (distinct_grp in distinct_groups) {
                          # Identify rows belonging to the current group
                          rows_in_group &lt;- names(grouping[grouping == distinct_grp])

                          # Adjust the pairwise_contrast values for each row in the group
                          pairwise_contrast_new[rows_in_group, col] &lt;- pairwise_contrast[rows_in_group[rows_in_group %in% groups], col] / length(rows_in_group)
                        }
                      }
                    }

                    # Replace the original matrix with the new one
                    pairwise_contrast &lt;- pairwise_contrast_new

                    #### 3. combine them
                    contrast_design &lt;- cbind(deviation_contrasts_tested / (n_populations - 1), pairwise_contrast)

                } else {
                    pairwise_vector[orig_mean_tested[1]] &lt;- 1
                    pairwise_vector[orig_mean_tested[2]] &lt;- -1
                    contrast_design &lt;- as.matrix(pairwise_vector)
                    colnames(contrast_design) &lt;- str_c(orig_mean_tested[1], &quot;_vs_&quot;, orig_mean_tested[2])
                }

                posterior_mean_subset &lt;- posterior_mean[index,]
                posterior_mean_subset2 &lt;- posterior_mean_subset[orig_mean_tested]
                posterior_vcov_subset &lt;- posterior_vcov[,,index]
                posterior_vcov_subset2 &lt;- posterior_vcov_subset[orig_mean_tested,orig_mean_tested]

                contrast_diff &lt;- t(contrast_design) %*% posterior_mean_subset2
                contrast_vcov &lt;- t(contrast_design) %*% posterior_vcov_subset2 %*% contrast_design
                contrast_se &lt;- diag(contrast_vcov) %&gt;% sqrt

                contrast_p &lt;- 2 * (1 - pnorm(abs(contrast_diff) / contrast_se))

                contrast_diff_df &lt;- t(contrast_diff) %&gt;% as_tibble
                colnames(contrast_diff_df) %&lt;&gt;% str_c(&quot;mean_contrast_&quot;, .)
                contrast_se_df &lt;- t(contrast_se) %&gt;% as_tibble
                colnames(contrast_se_df) %&lt;&gt;% str_c(&quot;se_contrast_&quot;, .)
                contrast_p_df &lt;- t(contrast_p) %&gt;% as_tibble
                colnames(contrast_p_df) %&lt;&gt;% str_c(&quot;p_contrast_&quot;, .)

                contrast_df &lt;- bind_cols(contrast_diff_df, contrast_se_df, contrast_p_df)
            } else if(grouping[orig_mean_tested][1]!=grouping[orig_mean_tested][2]){
                contrast_vector &lt;- rep(NA, length(population_names))
                names(contrast_vector) &lt;- str_c(&quot;mean_contrast_&quot;, population_names, &quot;_deviation&quot;)
                contrast_df &lt;- t(contrast_vector) %&gt;% as_tibble
            }
         
        contrast_df &lt;- contrast_df %&gt;% as.data.frame
        rownames(contrast_df) &lt;- rownames(posterior_mean)[index]
        return(contrast_df)
        }
    
    }

    if(length(&quot;${cells}&quot;) &gt; 0){
        # All the cells
        cells &lt;- c(&quot;${&quot;, &quot;.join(cells)}&quot;) %&gt;% str_split(., &quot;,&quot;, simplify = TRUE) %&gt;% as.character 

        # Automatically set grouping categories based on the recipeï¼Œ set0 for the celltypes without multiple populations
        grouping_all &lt;- rep(0, length(cells))
        names(grouping_all) &lt;- cells
        
  
        # Read groupings from the recipe
        if(length(&quot;${group1}&quot;) &gt; 0){
            cell_groups &lt;- list(
              ${&quot;group1 = c(&quot; + &quot;, &quot;.join([&quot;&#39;&quot; + item + &quot;&#39;&quot; for item in group1]) + &quot;)&quot; if len(group1) &gt; 0 else &quot;&quot;} 
              ${&quot;, group2 = c(&quot; + &quot;, &quot;.join([&quot;&#39;&quot; + item + &quot;&#39;&quot; for item in group2]) + &quot;)&quot; if len(group2) &gt; 0 else &quot;&quot;} 
              ${&quot;, group3 = c(&quot; + &quot;, &quot;.join([&quot;&#39;&quot; + item + &quot;&#39;&quot; for item in group3]) + &quot;)&quot; if len(group3) &gt; 0 else &quot;&quot;}
            )
            if(!is.null(cell_groups)) {
              cell_groups &lt;- map(cell_groups, ~str_split(.x, &quot;,&quot;, simplify = TRUE) %&gt;% as.character())
            }
        }
  
        if(&quot;${grouping_recipe}&quot; != &quot;&quot;){
            cell_groups &lt;- readLines(&quot;${grouping_recipe}&quot;)
            cell_groups &lt;- lapply(cell_groups, function(g) strsplit(g, &quot;,&quot;)[[1]])
        }

        if(!is.null(cell_groups)){
            for(i in seq_along(cell_groups)) {
              grouping_all[cell_groups[[i]]] &lt;- i
            }
        }
    }

    
    # Read the data files
    orig_data &lt;- read_rds(&quot;${_paths_sum[0]}&quot;)$bhat
    posterior_data &lt;- read_rds(&quot;${_input}&quot;)
    posterior_mean &lt;- posterior_data$PosteriorMean
    posterior_cov &lt;- posterior_data$PosteriorCov

    # Align data and clean-up NaN values
    orig_data &lt;- orig_data[, colnames(posterior_mean), drop = FALSE]
    orig_data[which(is.nan(orig_data))] &lt;- 0 # Placeholder for NaNs

    # Apply the FitContrast function and consolidate results
    contrast_result &lt;- map(1:nrow(posterior_mean), FitContrast, orig_data, posterior_mean, posterior_cov) %&gt;% bind_rows %&gt;%
        select(matches(&quot;mean_contrast.*deviation&quot;), matches(&quot;mean_contrast.*_vs_&quot;), 
               matches(&quot;se_contrast.*deviation&quot;), matches(&quot;se_contrast.*_vs_&quot;), 
               matches(&quot;p_contrast.*deviation&quot;), matches(&quot;p_contrast.*_vs_&quot;))
    #rownames(contrast_result) &lt;- rownames(posterior_mean)

    write_rds(contrast_result,  ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># merge the contrast data with slice data
[mash_posterior_contrast_2]
input: group_by = &quot;all&quot;
output: f&quot;{cwd}/posterior_sum.csv&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = &#39;24h&#39;,  mem = &#39;10G&#39;, tags = f&#39;{_output:bn}&#39;  

R: expand = &quot;${ }&quot;,stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39; 
        library(dplyr)
        library(tidyverse)
        library(ggnewscale)
        
        all.list &lt;- stringr::str_split(&quot;${_input}&quot;, &quot; &quot;, simplify = T)
       
        p_cut &lt;- ${p_cut} %&gt;% as.numeric

        cells&lt;-c(&quot;${&quot;,&quot;.join(cells)}&quot;)%&gt;%str_split(.,&quot;,&quot;,simplify = T)%&gt;%as.character #ggnewscale cannot use a specified order, I can not find a good way to order them by category for now
        conditions &lt;-combn(cells, m = 2) %&gt;%apply(., 2, str_c, collapse = &quot;_vs_&quot;) 
        
        df &lt;- matrix(ncol = length(conditions), nrow = 4) %&gt;% as.data.frame()
        colnames(df) &lt;- conditions
        rownames(df) &lt;- c( &quot;n_sig_snp&quot;, &quot;n_snp&quot;,&quot;n_sig_feature&quot;,&quot;n_all_feature&quot;)
        for (con in conditions) {
            n.all.sig.snp &lt;- n.all.snp &lt;- n.all.sig.feature &lt;- n.all.feature &lt;- 0

            for (i in 1:length(all.list)) {
                tmp &lt;- readRDS(all.list[i])
                p.mtx &lt;- tmp %&gt;% select(matches(&quot;p_contrast.*_vs_&quot;))
                p.mtx.con &lt;- p.mtx %&gt;% select(matches(con))
                # print(n.sig.snp)
                if(ncol(p.mtx.con)&gt;0){
                    p.mtx.con&lt;-na.omit(p.mtx.con)
                    n.sig.snp &lt;- sum(p.mtx.con &lt; p_cut)
                    n.snp &lt;- nrow(p.mtx.con)
                } else {
                    n.sig.snp &lt;- n.snp &lt;-0
                }
                # print(n.snp)
                n.sig.feature &lt;- ifelse(n.sig.snp &gt; 0, 1, 0)
                n.feature&lt;- ifelse (n.snp &gt; 0 , 1, 0)

                n.all.sig.snp &lt;- n.sig.snp + n.all.sig.snp
                n.all.snp &lt;- n.all.snp + n.snp
                n.all.sig.feature &lt;- n.all.sig.feature + n.sig.feature
                n.all.feature &lt;- n.all.feature + n.feature
            }
            df[, con] &lt;- c( n.all.sig.snp,n.all.snp, n.all.sig.feature,n.all.feature)
        }
        write.csv(df, &quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># plot contrast result
[mash_posterior_contrast_3, posterior_cntrast_plot]
input: group_by = &quot;all&quot;
output: f&quot;{cwd}/posterior_sum.png&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = &#39;24h&#39;,  mem = &#39;10G&#39;, tags = f&#39;{_output:bn}&#39;  

R: expand = &quot;${ }&quot;,stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39; 
        library(dplyr)
        library(tidyverse)
        library(ggnewscale)
        df &lt;- read.csv(&quot;${_input:a}&quot;,row.names=1)
        colnames(df)&lt;-gsub(&quot;DLPFC_&quot;,&quot;&quot;,colnames(df))
        for (i in 1:ncol(df)) {
            con1 &lt;- stringr::str_split(colnames(df)[i], &quot;_vs_&quot;, simplify = T)[, 1]
            con2 &lt;- stringr::str_split(colnames(df)[i], &quot;_vs_&quot;, simplify = T)[, 2]

            if (con1 &gt; con2) {
                new.name &lt;- paste0(con2, &quot;_vs_&quot;, con1)
                colnames(df)[i] &lt;- new.name
            }
        }

        ## summarizxse with approach 1: snp-feature pair
        snp.ratio &lt;- df[&quot;n_sig_snp&quot;, ] / df[&quot;n_snp&quot;, ]
        snp.ratio &lt;- snp.ratio %&gt;%
            t() %&gt;%
            as.data.frame()
        colnames(snp.ratio) &lt;- &quot;ratio&quot;
        snp.ratio$group &lt;- &quot;snp&quot;

        ## summarize with approach 2: feature
        fet.ratio &lt;- df[&quot;n_sig_feature&quot;, ] / df[&quot;n_all_feature&quot;, ]
        fet.ratio &lt;- fet.ratio %&gt;%
            t() %&gt;%
            as.data.frame()
        rownames(fet.ratio) &lt;- paste0(stringr::str_split(rownames(fet.ratio), &quot;_vs_&quot;, simplify = T)[, 2], &quot;_vs_&quot;, stringr::str_split(rownames(fet.ratio), &quot;_vs_&quot;, simplify = T)[, 1])
        colnames(fet.ratio) &lt;- &quot;ratio&quot;
        fet.ratio$group &lt;- &quot;feature&quot;
        ratio &lt;- rbind(snp.ratio, fet.ratio)

        ## I need to add the below to make it Simmetrie

        cons &lt;- rownames(ratio) %&gt;%
            str_split(., &quot;_vs_&quot;, simplify = T) %&gt;%
            .[, 1] %&gt;%
            unique()
        for (i in 1:length(cons)) {
            new.name &lt;- paste0(cons[i], &quot;_vs_&quot;, cons[i])
            ratio[new.name, ] &lt;- 0
        }

        ratio$con1 &lt;- stringr::str_split(rownames(ratio), &quot;_vs_&quot;, simplify = T)[, 1]
        ratio$con2 &lt;- stringr::str_split(rownames(ratio), &quot;_vs_&quot;, simplify = T)[, 2]

        ## prepare for the plot, score1 is for snp-feature pair, score2 is for feature only
        ratio$score1 &lt;- ratio$score2 &lt;- 0
        ratio$score1[ratio$group == &quot;snp&quot;] &lt;- ratio$ratio[ratio$group == &quot;snp&quot;]
        ratio$score2[ratio$group == &quot;feature&quot;] &lt;- ratio$ratio[ratio$group == &quot;feature&quot;]
        ratio$label &lt;- paste0(round(ratio$ratio, 4) * 100, &quot;%&quot;)
        ratio$label[ratio$group == 0] &lt;- NA

        # plot
        num_cols &lt;- length(cons)
        height &lt;- width &lt;- 4 + num_cols * 0.5 
        ggplot(ratio[ratio$group == &quot;snp&quot;, ], aes(x = con1, y = con2)) +
            geom_tile(aes(fill = score1)) +
            scale_fill_gradient2(&quot;SNP_Feature pair&quot;,
                low = &quot;#762A83&quot;, mid = &quot;white&quot;, high = &quot;#1B7837&quot;
            ) +
            new_scale(&quot;fill&quot;) +
            geom_tile(aes(fill = score2), data = subset(ratio, group != &quot;snp&quot;)) +
            scale_fill_gradient2(&quot;Feature&quot;,
                low = &quot;#1B7837&quot;, mid = &quot;white&quot;, high = &quot;#762A83&quot;
            ) +
            geom_text(data = ratio, aes(label = label)) +
            theme_bw()
        #geom_text(data=ratio, aes(label = label, color = factor(group))) +theme_bw()
        #ggsave(gsub(&quot;.csv&quot;,&quot;.png&quot;,filename))

        ggsave(&quot;${_output}&quot;,width = width, height = height)
</pre></div>
</div>
</div>
</div>
</section>
<section id="feature-score-from-contrast-results">
<h2>Feature score from contrast results<a class="headerlink" href="#feature-score-from-contrast-results" title="Permalink to this heading">#</a></h2>
<p><strong>FIXME: Havenâ€™t cleaned up yet, need to talk to Gao to decide which one should be kept.</strong></p>
<section id="feature-score-meta">
<h3>feature_score_meta<a class="headerlink" href="#feature-score-meta" title="Permalink to this heading">#</a></h3>
<p>Meta approach: Meta analysis with each cell deviation contrast result of each feature (a loop for each column in deviation results). And then perform meta analysis with pairwise contrasts to get a pvalue to find to understand what the specific differences are. But there is a problem that meta analysis with so many snps would cost too much time to compute, I am trying to downsample for input, Dan suggest to LD prune with a more permissive threshold.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># compute feature score from contrast results
[feature_score_meta_1]
#regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]
parameter: plink_path = &quot;/mnt/mfs/cluster/bin/plink1.9.10/plink&quot;
parameter: bfile_path = &quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/WashU_genotype/genotype_qc/MAP_Brain-xQTL_Gwas_geno_0.1_maf_0.0005.filtered.&quot;
parameter: extract = &quot;/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/snp.list&quot;
parameter: window_size = &#39;100&#39;
parameter: step_size = &#39;10&#39;
parameter: r2_threshold = &#39;0.2&#39;
parameter: per_chunk = &#39;100&#39;
parameter: LD_prune = &quot;TRUE&quot;
parameter: downsample_ratio = &#39;1&#39;
parameter: meta_method = &#39;REML&#39;
parameter: LDcache = &quot;/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/&quot;
parameter: contrast_input = [path(x[0]) for x in regions]
input: contrast_input, group_by = per_chunk
output: f&quot;{cwd}/feature_score_metaLD/cache/mash_posterior_contrast_featurescore{_index+1}.rds&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;  
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container, entrypoint=entrypoint
    # Library Loading
    suppressMessages({
      library(data.table)
      library(tidyverse)
      library(metafor)
    })

    # Environment Configuration
    plink_path &lt;- &quot;${plink_path}&quot;
    bfile_path &lt;- &quot;${bfile_path}&quot;
    indep_pairwise &lt;- paste(&quot;${window_size}&quot;, &quot;${step_size}&quot;, &quot;${r2_threshold}&quot;)
    out &lt;- NULL
    set.seed(999)

    # Define helper functions

    plink_ld_pruning &lt;- function(res, chr, bfile_path, LDcache, gene, plink_path, indep_pairwise) {
      extract &lt;- str_c(&quot;${LDcache}&quot;, gene, &quot;_snp.list&quot;)
      output &lt;- str_c(&quot;${LDcache}&quot;, gene, &quot;_output&quot;)
      tmp_contrast_results &lt;- readRDS(res) %&gt;% as.matrix()
      write.table(rownames(tmp_contrast_results), extract, quote = F)
      # system(paste0(&quot;rm &quot;, output, &quot;*&quot;), intern = TRUE)
      bfile &lt;- str_c(bfile_path, chr)
      command &lt;- paste(plink_path, &quot;--bfile&quot;, bfile, &quot;--extract&quot;, extract, &quot;--indep-pairwise&quot;, indep_pairwise, &quot;--out&quot;, output)
      system(command, intern = TRUE)
      ld_output &lt;- read.table(str_c(&quot;${LDcache}&quot;, gene, &quot;_output.prune.in&quot;))
      return(tmp_contrast_results[ld_output$V1, ] %&gt;% as.data.table())
    }

    calculate_feature_scores &lt;- function(tmp_contrast_results, meta_method) {
      effect_sizes &lt;- tmp_contrast_results %&gt;% select(matches(&quot;mean_contrast.*deviation&quot;)) %&gt;% as.matrix()
      se_values &lt;- tmp_contrast_results %&gt;% select(matches(&quot;se_contrast.*deviation&quot;)) %&gt;% as.matrix()
      feature_scores &lt;- data.table()
      for (i in 1:ncol(effect_sizes)) {
        effect_sizes_condition &lt;- effect_sizes[, i]
        se_values_condition &lt;- se_values[, i]
        absolute_effect_sizes &lt;- abs(as.numeric(effect_sizes_condition))
        pairwise_standard_errors &lt;- as.numeric(se_values_condition)
        meta_result &lt;- rma(yi = absolute_effect_sizes, sei = pairwise_standard_errors, method = meta_method)
        z_scores &lt;- meta_result$b / meta_result$se
        feature_scores_condition &lt;- data.table(ZScore = z_scores)
        feature_scores &lt;- rbindlist(list(feature_scores, feature_scores_condition))
      }
      return(feature_scores)
    }

    # Main Loop
    for (res in c(${_input:r,})) try({
      chr &lt;- res %&gt;% basename() %&gt;% str_extract(., paste0(&quot;\\.(.*?)\\.&quot;)) %&gt;% str_replace_all(&quot;\\.&quot;, &quot;&quot;) %&gt;% str_replace_all(&quot;chr&quot;, &quot;&quot;)
      gene &lt;- basename(res) %&gt;% stringr::str_split(., &quot;norminal.cis_long_table.&quot;, simplify = TRUE) %&gt;% .[, 2] %&gt;% gsub(&quot;_posterior_contrast.rds&quot;, &quot;&quot;, .)

      tmp_contrast_results &lt;- readRDS(res)

      LD_prune &lt;- ${&#39;TRUE&#39; if LD_prune else &#39;FALSE&#39;}
      if (LD_prune) {
        tmp_contrast_results &lt;- plink_ld_pruning(res, chr, bfile_path, LDcache, gene, plink_path, indep_pairwise)
      }

      if (ncol(tmp_contrast_results %&gt;% select(matches(&quot;mean_contrast.*deviation&quot;))) &gt; 0) {
        feature_scores &lt;- calculate_feature_scores(tmp_contrast_results, &quot;${meta_method}&quot;)
        colnames(feature_scores) &lt;- gene
        condition_name &lt;- colnames(effect_sizes) %&gt;% gsub(&quot;mean_contrast_&quot;, &quot;&quot;, .) %&gt;% gsub(&quot;_deviation&quot;, &quot;&quot;, .)
        feature_scores &lt;- feature_scores[, condition := condition_name]
        saveRDS(feature_scores, str_c(&quot;${_output:d}&quot;, &quot;/&quot;, gsub(&quot;.rds&quot;, &quot;_featurescore.rds&quot;, basename(res))))
      }

      if (is.null(out)) {
        out &lt;- feature_scores
      } else {
        out &lt;- merge(out, feature_scores, by = &quot;condition&quot;, all = TRUE)
      }
    })
    saveRDS(out, &quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="feature-score-finemap">
<h3>feature_score_finemap<a class="headerlink" href="#feature-score-finemap" title="Permalink to this heading">#</a></h3>
<p>feature score with fine mapped QTL (the top 1 in each CS)</p>
<p>Fine-mapped approach(more recommend by Dan). pick the top SNP in each CS from each cell type fine mapped results. Then pick the most significant one from contrast result. Get the z-score from that snp, which should be the score of that cell type.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># compute feature score from contrast results with eQTL and pQTL finemapped signals
[feature_score_finemap_1]
#regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]
parameter: per_chunk = &#39;100&#39;
parameter: pfine_path = &quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/QTL_fine_mapping/pqtl.all_variants.tsv&quot;
parameter: efine_path = &quot;/mnt/vast/hpc/csg/hs3393/susie_eqtl/Result_new/tsv/&quot;
parameter: efine_suffix = &quot;.unisusie.fit.variant.tsv&quot;
parameter: gene_ref_path = &quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/reference_data/Homo_sapiens.GRCh38.103.chr.reformatted.collapse_only.gene.region_list&quot;
parameter: contrast_input = [path(x[0]) for x in regions]
input: contrast_input, group_by = per_chunk
output: f&quot;{cwd}/feature_score_finemap/cache/mash_posterior_contrast_featurescore{_index+1}.rds&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;  
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container, entrypoint=entrypoint
   # Load necessary libraries
    suppressMessages({
        library(data.table)
        library(tidyverse)
    })

    # Function to calculate scores from fine-mapping results
    score_from_cs &lt;- function(fine.file, contrast_results, con) {
        CSs &lt;- unique(fine.file$cs_order) %&gt;% .[. != 0]
        max.cs.df &lt;- NULL
        for (cs in CSs) {
            tmp &lt;- fine.file[fine.file$cs_order == cs, ]
            max.cs.df &lt;- rbind(max.cs.df, tmp[tmp$pip == max(tmp$pip), ])
        }
        snps &lt;- intersect(rownames(contrast_results), max.cs.df$variants)
        if (length(snps) == 0) return(NA)

        # Extract contrast p-values
        contrast_p &lt;- NULL
        if (ncol(select(contrast_results %&gt;% as.data.frame(), matches(str_c(&quot;p_contrast_&quot;, con, &quot;_deviation&quot;)))) &gt; 0) {
            contrast_p &lt;- contrast_results[snps, str_c(&quot;p_contrast_&quot;, con, &quot;_deviation&quot;), drop = F]
        } else if (ncol(select(contrast_results %&gt;% as.data.frame(), matches(&quot;p_contrast.*_vs_*&quot;))) == 1) {
            contrast_p &lt;- contrast_results %&gt;% as.data.frame() %&gt;% select(matches(&quot;p_contrast.*_vs_*&quot;)) %&gt;% .[snps, , drop = F]
        }

        if (is.null(contrast_p)) return(NA)

        max.snp &lt;- contrast_p[contrast_p[, 1] == max(contrast_p[, 1]), , drop = F] %&gt;% rownames()
        score &lt;- max(abs(contrast_results %&gt;% as.data.frame() %&gt;% select(matches(&quot;mean_contrast.*_vs_*&quot;)) %&gt;% .[max.snp, ]) /
                     (contrast_results %&gt;% as.data.frame() %&gt;% select(matches(&quot;se_contrast.*_vs_*&quot;)) %&gt;% .[max.snp, ]))
        return(score)
    }

    # Read pQTL file
    pfine.mapped.result &lt;- fread(&quot;${pfine_path}&quot;)
    pfine.mapped.result$Gene &lt;- str_split(pfine.mapped.result$molecular_trait_id, &quot;_&quot;, simplify = TRUE)[, 2]
    gene.ref &lt;- read.table(&quot;${gene_ref_path}&quot;)
    out &lt;- data.table()

    # Loop through each input and compute scores
    for (res in c(${_input:r,})) {
        tmp_contrast_results &lt;- readRDS(res) %&gt;% as.matrix()

        gene &lt;- basename(res) %&gt;%
            str_split(., &quot;norminal.cis_long_table.&quot;, simplify = TRUE) %&gt;%
            .[, 2] %&gt;%
            gsub(&quot;_posterior_contrast.rds&quot;, &quot;&quot;, .)
        ensemble &lt;- gene.ref[gene.ref$V5 == gene, ]$V4
        efine.mapped.result &lt;- list.files(path = &quot;${efine_path}&quot;, pattern = paste0(ensemble, &quot;${efine_suffix}&quot;), full.names = T)
        pfine.file &lt;- pfine.mapped.result[pfine.mapped.result$Gene == gene &amp; pfine.mapped.result$cs_order &gt; 0, ]

        df &lt;- data.frame()
        if (nrow(pfine.file) &gt; 0) {
            message(&quot;Extracting signal from pQTL in &quot;, gene)
            df[gene, &quot;DLPFC_pQTL&quot;] &lt;- score_from_cs(pfine.file, tmp_contrast_results, &quot;DLPFC_pQTL&quot;)
        }
        if (length(efine.mapped.result) &gt; 0) {
            fine.conditions &lt;- basename(efine.mapped.result) %&gt;%
                sub(&quot;demo.&quot;, &quot;&quot;, .) %&gt;%
                sub(paste0(&quot;.&quot;, ensemble, &quot;${efine_suffix}&quot;), &quot;&quot;, .) %&gt;%
                .[. != &quot;ALL&quot;] %&gt;%
                .[. != &quot;End&quot;]
            for (con in fine.conditions) {
                con.file &lt;- efine.mapped.result[grep(con, efine.mapped.result)] %&gt;% fread()
                con.fine.file &lt;- con.file[con.file$cs_order &gt; 0, ]
                if (nrow(con.fine.file) &gt; 0) {
                    df[gene, con] &lt;- score_from_cs(con.fine.file, tmp_contrast_results, con)
                }
            }
        }
        saveRDS(df, str_c(&quot;${cwd}&quot;,&quot;/feature_score_finemap/cache/&quot;,gsub(&quot;.rds&quot;,&quot;_featurescore.rds&quot;,basename(res))))
        out &lt;- rbindlist(list(out, as.data.table(df, keep.rownames = TRUE)), use.names = TRUE, fill = TRUE)
    }
    saveRDS(out, &quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># merge the feature score from contrast results with eQTL and pQTL finemapped signals
[feature_score_finemap_2]
input: group_by = &quot;all&quot;
output: f&quot;{cwd}/feature_score_finemap/posterior_feature_score_sum.csv&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = &#39;24h&#39;,  mem = &#39;10G&#39;, tags = f&#39;{_output:bn}&#39;  

R: expand = &quot;${ }&quot;,stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39; 
    suppressMessages(library(data.table))
    suppressMessages(library(tidyverse))
    
    out &lt;- data.table()

    all.list &lt;- stringr::str_split(&quot;${_input}&quot;, &quot; &quot;, simplify = T)
    for (i in all.list) {
        feature_scores &lt;- readRDS(i)
        # Print the feature scores for each condition
        if (!is.null(feature_scores)) {
           if (is.null(out)) {
                out &lt;- as.data.table(feature_scores, keep.rownames = TRUE)
              } else {
                out &lt;-  rbindlist(list(out, as.data.table(feature_scores, keep.rownames = TRUE)), use.names = TRUE, fill = TRUE)
              }
        }
    }
    write.csv(out, &quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="feature-score-nsig">
<h3>feature_score_nsig<a class="headerlink" href="#feature-score-nsig" title="Permalink to this heading">#</a></h3>
<p>Calculate the number of meaningful SNPs in each feature of each cell type (actually, ratio)</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># compute feature score from contrast results
[feature_score_nsig_1]
#regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]
parameter: per_chunk = &#39;100&#39;
parameter: contrast_input = [path(x[0]) for x in regions]
input: contrast_input, group_by = per_chunk
output: f&quot;{cwd}/feature_score_nsig/cache/mash_posterior_contrast_featurescore_nsig{_index+1}.rds&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;  
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container, entrypoint=entrypoint
   # Load required libraries
    suppressMessages({
        library(data.table)
        library(tidyverse)
        library(metafor)
    })

    # Convert p_cut to numeric
    p_cut &lt;- ${p_cut} %&gt;% as.numeric

    # Initialize the output data table
    out &lt;- NULL
    set.seed(999)

    # Function to extract the gene name from the file path
    get_gene_name &lt;- function(file_path) {
        basename(file_path) %&gt;%
            stringr::str_split(., &quot;norminal.cis_long_table.&quot;, simplify = TRUE) %&gt;%
            .[, 2] %&gt;%
            gsub(&quot;_posterior_contrast.rds&quot;, &quot;&quot;, .)
    }

    # Loop through each input RDS file
    for (res in c(${_input:r,})) {
        try({
            # Extract gene name
            gene &lt;- get_gene_name(res)
            print(gene)

            # Load the contrast results and filter columns of interest
            tmp_contrast_results &lt;- readRDS(res) %&gt;%
                select(matches(&quot;p_contrast_.*deviation&quot;)) %&gt;%
                as.matrix()

            # Initialize the feature output data table
            feature.out &lt;- data.table()

            # Calculate the ratio for each condition
            for (i in 1:ncol(tmp_contrast_results)) {
                condition_name &lt;- colnames(tmp_contrast_results)[i] %&gt;%
                    gsub(&quot;p_contrast_&quot;, &quot;&quot;, .) %&gt;%
                    gsub(&quot;_deviation&quot;, &quot;&quot;, .)

                n_sig_snp &lt;- sum(tmp_contrast_results[, i] &lt; p_cut, na.rm = TRUE)
                n_snp &lt;- sum(!is.na(tmp_contrast_results[, i]))

                ratio &lt;- n_sig_snp / n_snp
                feature.out &lt;- rbind(feature.out, data.table(gene = gene, condition = condition_name, ratio = ratio))
            }

            # Save individual gene results
            saveRDS(feature.out, paste0(&quot;${_output:d}&quot;, &quot;/&quot;, gsub(&quot;.rds&quot;, &quot;_n_sig_ratio.rds&quot;, basename(res))))

            # Merge with the overall output
            if (is.null(out)) {
                out &lt;- feature.out
            } else {
                out &lt;- merge(out, feature.out, by = &quot;condition&quot;, all = TRUE)
            }
        }, silent = TRUE) # Error handling to proceed even if one iteration fails
    }

    # Save the combined output
    saveRDS(out, &quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="feature-pval-pair">
<h3>feature_pval_pair<a class="headerlink" href="#feature-pval-pair" title="Permalink to this heading">#</a></h3>
<p>find the specific different cell type pair</p>
<p>perform meta analysis with pairwise contrasts to get a pvalue to find to understand what the specific differences are.</p>
<p>metaanalysis return a warning as â€œWarning message: â€œRatio of largest to smallest sampling variance extremely large. May not be able to obtain stable results.â€â€
Which is due to the big difference between max(pairwise_standard_errors^2) and min(pairwise_standard_errors^2) So I have delete the snps with small pairwise_standard_errors as Xuewei suggested</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># compute feature score from contrast results
[feature_pval_pair_1]
#regions = [x.replace(&quot;\&quot;&quot;,&quot;&quot;).strip().split() for x in open(analysis_unit).readlines() if x.strip() and not x.strip().startswith(&#39;#&#39;)]
parameter: plink_path = &quot;/mnt/mfs/cluster/bin/plink1.9.10/plink&quot;
parameter: bfile_path = &quot;/mnt/vast/hpc/csg/molecular_phenotype_calling/WashU_genotype/genotype_qc/MAP_Brain-xQTL_Gwas_geno_0.1_maf_0.0005.filtered.&quot;
parameter: extract = &quot;/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/snp.list&quot;
parameter: window_size = &#39;100&#39;
parameter: step_size = &#39;10&#39;
parameter: r2_threshold = &#39;0.2&#39;
parameter: per_chunk = &#39;100&#39;
parameter: LD_prune = True
parameter: downsample_ratio = &#39;1&#39;
parameter: meta_method = &#39;REML&#39;
parameter: se_cutoff = &#39;1E-03&#39;
parameter: LDcache = &quot;/home/rf2872/Work/Multivariate/MASH/From_SuSiE/2023.5_new/add_pQTL/LDcache/&quot;
parameter: contrast_input = [path(x[0]) for x in regions]
input: contrast_input, group_by = per_chunk
output: f&quot;{cwd}/feature_pval_pair/cache/mash_posterior_contrast_featurescore{_index+1}.rds&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;  
R: expand = &quot;${ }&quot;, workdir = cwd, stderr = f&quot;{_output:n}.stderr&quot;, stdout = f&quot;{_output:n}.stdout&quot;, container = container, entrypoint=entrypoint
    # Library Loading
    suppressMessages({
      library(data.table)
      library(tidyverse)
      library(metafor)
    })

    # Initializations
    out &lt;- NULL
    set.seed(999)

    # Helper function: extract gene name from result
    extract_gene &lt;- function(res) {
      basename(res) %&gt;%
        stringr::str_split(., &quot;norminal.cis_long_table.&quot;, simplify = TRUE) %&gt;%
        .[, 2] %&gt;%
        gsub(&quot;_posterior_contrast.rds&quot;, &quot;&quot;, .)
    }

    # Helper function: perform meta analysis per cell
    meta_analysis_per_cell &lt;- function(effect_sizes, se_values, gene, se_cutoff) {
      conditions &lt;- colnames(effect_sizes) %&gt;%
        sub(&quot;mean_contrast_&quot;, &quot;&quot;, .) %&gt;%
        unique()
      cells &lt;- c(sub(&quot;_vs_.*&quot;, &quot;&quot;, conditions), sub(&quot;.*_vs_&quot;, &quot;&quot;, conditions)) %&gt;% unique()

      df &lt;- data.table()
      for (cell in cells) {
        cell.b &lt;- effect_sizes[, grep(cell, colnames(effect_sizes)), drop = F]
        cell.se &lt;- se_values[, grep(cell, colnames(se_values)), drop = F]
        feature_pvals &lt;- data.table()
        for (i in 1:ncol(cell.b)) {
          effect_sizes_condition &lt;- cell.b[, i]
          se_values_condition &lt;- cell.b[, i]
          # Filter out based on se cutoff
          effect_sizes_condition &lt;- effect_sizes_condition[which(se_values_condition &gt; as.numeric(se_cutoff))]
          se_values_condition &lt;- se_values_condition[which(se_values_condition &gt; as.numeric(se_cutoff))]
          # Meta-analysis
          absolute_effect_sizes &lt;- abs(as.numeric(effect_sizes_condition))
          pairwise_standard_errors &lt;- as.numeric(se_values_condition)
          meta_result &lt;- rma(yi = absolute_effect_sizes, sei = pairwise_standard_errors, method = &quot;REML&quot;)
          feature_pval_condition &lt;- data.table(pavlue = meta_result$pval)
          feature_pvals &lt;- rbindlist(list(feature_pvals, feature_pval_condition))
        }
        colnames(feature_pvals) &lt;- gene
        condition_name &lt;- colnames(cell.b) %&gt;%
          gsub(&quot;mean_contrast_&quot;, &quot;&quot;, .)
        feature_pvals &lt;- feature_pvals[, condition := condition_name]
        df &lt;- rbindlist(list(df, feature_pvals))
      }
      return(df)
    }

    # Main Loop
    for (res in c(${_input:r,})) {
      gene &lt;- extract_gene(res)
      print(gene)
      tmp_contrast_results &lt;- readRDS(res)
      ld_output &lt;- read.table(str_c(&quot;${LDcache}&quot;, gene, &quot;_output.prune.in&quot;))
      tmp_contrast_results &lt;- tmp_contrast_results[ld_output$V1, ] %&gt;% as.data.table(keep.rownames = T)

      effect_sizes &lt;- tmp_contrast_results %&gt;% select(matches(&quot;mean_contrast.*_vs_*&quot;)) %&gt;% as.matrix()
      se_values &lt;- tmp_contrast_results %&gt;% select(matches(&quot;se_contrast.*_vs_*&quot;)) %&gt;% as.matrix()

      df &lt;- meta_analysis_per_cell(effect_sizes, se_values, gene, &quot;${se_cutoff}&quot;)
      saveRDS(df, str_c(&quot;${_output:d}&quot;, &quot;/&quot;, gsub(&quot;.rds&quot;, &quot;_featurescore_pw.rds&quot;, basename(res))))

      if (is.null(out)) {
        out &lt;- df
      } else {
        out &lt;- merge(out, df, by = &quot;condition&quot;, all = TRUE, allow.cartesian = TRUE)
      }
    }
    saveRDS(out, &quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># merge the feature score from contrast results
[feature_pval_pair_2, feature_score_meta_2, feature_score_nsig_2]
input: group_by = &quot;all&quot;
output: f&quot;{cwd}/feature_pval_pair/posterior_feature_score_sum.csv&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = &#39;24h&#39;,  mem = &#39;10G&#39;, tags = f&#39;{_output:bn}&#39;  

R: expand = &quot;${ }&quot;,stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39; 
    library(dplyr)
    library(tidyverse)
    out &lt;- NULL

    all.list &lt;- stringr::str_split(&quot;${_input}&quot;, &quot; &quot;, simplify = T)
    for (i in all.list) {
        feature_scores &lt;- readRDS(i)
        # Print the feature scores for each condition
        if (!is.null(feature_scores)) {
            if (is.null(out)) {
                out &lt;- feature_scores
            } else {
                out &lt;- merge(out, feature_scores, by = &quot;condition&quot;, all = TRUE)
            }
        }
    }
      out&lt;-t(out)
      #out&lt;-out[-1,]
    write.csv(out, &quot;${_output}&quot;,col.names=F)
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/multivariate_genome/MASH"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="mash_fit.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">MASH analysis pipeline with data-driven prior matrices</p>
      </div>
    </a>
    <a class="right-next"
       href="../../rare_xqtl/watershed.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">watershed analysis for rare xQTL</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#example">example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior">posterior</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#contrast-analysis-works">contrast analysis: works</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-results">Posterior results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">Posterior</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#posterior-contrast">Posterior contrast</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-score-from-contrast-results">Feature score from contrast results</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-score-meta">feature_score_meta</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-score-finemap">feature_score_finemap</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-score-nsig">feature_score_nsig</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#feature-pval-pair">feature_pval_pair</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      Â© Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>