
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>QTL Association Testing &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/association_scan/TensorQTL/TensorQTL';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="Quantile regression for QTL association testing" href="../quantile_models/qr_and_twas.html" />
    <link rel="prev" title="QTL Association Analysis" href="../qtl_association_testing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../qtl_association_testing.html">QTL Association Analysis</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../mnm_analysis/fine_mapping_miniprotocol.html">Fine-Mapping</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/association_scan/TensorQTL/TensorQTL.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/association_scan/TensorQTL/TensorQTL.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/association_scan/TensorQTL/TensorQTL.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>QTL Association Testing</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-cis-tensorqtl-command">i. Cis TensorQTL Command</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-trans-tensorqtl-command">ii. Trans TensorQTL Command</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#old-minimal-working-example">Old Minimal working example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-and-global-parameters">Setup and global parameters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cis-xqtl-association-testing">cis-xQTL association testing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#trans-xqtl-association-testing">Trans-xQTL association testing</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="qtl-association-testing">
<h1>QTL Association Testing<a class="headerlink" href="#qtl-association-testing" title="Link to this heading">#</a></h1>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Link to this heading">#</a></h2>
<p>We perform QTL association testing using TensorQTL [<a class="reference external" href="https://doi.org/10.1186/s13059-019-1836-7">cf. Taylor-Weiner et al (2019)</a>]. An additional protocol was added to test for quantile QTL associations.</p>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<ul class="simple">
<li><p>List of molecular phenotype files: a list of <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> files containing the table for the molecular phenotype. It should have a companion index file in <code class="docutils literal notranslate"><span class="pre">tbi</span></code> format. It is the output of gene_annotation or phenotype_by_chorm</p></li>
</ul>
<section id="example-phenotype-list">
<h3>Example phenotype list<a class="headerlink" href="#example-phenotype-list" title="Link to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start   end ID  path</span>
<span class="n">chr12</span>   <span class="mi">752578</span>  <span class="mi">752579</span>  <span class="n">ENSG00000060237</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">990508</span>  <span class="mi">990509</span>  <span class="n">ENSG00000082805</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">2794969</span> <span class="mi">2794970</span> <span class="n">ENSG00000004478</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">4649113</span> <span class="mi">4649114</span> <span class="n">ENSG00000139180</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6124769</span> <span class="mi">6124770</span> <span class="n">ENSG00000110799</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6534516</span> <span class="mi">6534517</span> <span class="n">ENSG00000111640</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
<div class="highlight-none notranslate"><div class="highlight"><pre><span></span>The header of the bed.gz is per the [TensorQTL](https://github.com/broadinstitute/tensorqtl) convention:

&gt;    Phenotypes must be provided in BED format, with a single header line starting with # and the first four columns corresponding to: chr, start, end, phenotype_id, with the remaining columns corresponding to samples (the identifiers must match those in the genotype input). The BED file should specify the center of the cis-window (usually the TSS), with start == end-1.
</pre></div>
</div>
<ul class="simple">
<li><p>List of genotypes in PLINK binary format (<code class="docutils literal notranslate"><span class="pre">bed</span></code>/<code class="docutils literal notranslate"><span class="pre">bim</span></code>/<code class="docutils literal notranslate"><span class="pre">fam</span></code>) for each chromosome, previously processed through our genotype QC pipelines.</p></li>
<li><p>Covariate file, a file with #id + samples name as colnames and each row a covariate: fixed and known covariates as well as hidden covariates recovered from factor analysis.</p></li>
<li><p>Optionally, a list of traits (genes, regions of molecular features etc) to analyze.</p></li>
</ul>
<p>For cis-analysis:</p>
<ul class="simple">
<li><p>Optionally, a list of genomic regions associate with each molecular features to analyze. The default cis-analysis will use a window around TSS. This can be customized to take given start and end genomic coordinates.</p></li>
</ul>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<p>For each chromosome, several of summary statistics files are generated, including both nominal test statistics for each test, as well as region (gene) level association evidence.</p>
<p>The columns of nominal association result are as follows:</p>
<ul class="simple">
<li><p>phenotype_id: Molecular trait identifier.(gene)</p></li>
<li><p>variant_id: ID of the variant (rsid or chr:position:ref:alt)</p></li>
<li><p>tss_distance: Distance of the SNP to the gene transcription start site (TSS)</p></li>
<li><p>af: The allele frequency of this SNPs</p></li>
<li><p>ma_samples: Number of samples carrying the minor allele</p></li>
<li><p>ma_count: Total number of minor alleles across individuals</p></li>
<li><p>pval: Nominal P-value from linear regression</p></li>
<li><p>beta: Slope of the linear regression</p></li>
<li><p>se: Standard error of beta</p></li>
<li><p>chr : Variant chromosome.</p></li>
<li><p>pos : Variant chromosomal position (basepairs).</p></li>
<li><p>ref : Variant reference allele (A, C, T, or G).</p></li>
<li><p>alt : Variant alternate allele.</p></li>
</ul>
<p>The columns of interaction association results are as follows:</p>
<p>Model: phenotype ~ <span class="math notranslate nohighlight">\(\beta\)</span>*snp + <span class="math notranslate nohighlight">\(\beta_{msex}\)</span>*msex + <span class="math notranslate nohighlight">\(\beta_{interaction}\)</span>*snp
*msex (taking msex as the interaction factor)</p>
<ul class="simple">
<li><p>chrom: chromosome number</p></li>
<li><p>pos : Variant chromosomal position (basepairs)</p></li>
<li><p>a1: Variant reference allele (A, C, T, or G)</p></li>
<li><p>a2: Variant alternate allele</p></li>
<li><p>molecular_trait_id: Molecular trait identifier, varies from phenotypes to phenotypes.</p></li>
<li><p>variant_id - ID of the top variant (rsid or chr:position:ref:alt)</p></li>
<li><p>af - Alternative allele frequency in MiGA cohort</p></li>
<li><p>ma_samples - Number of samples carrying the minor allele</p></li>
<li><p>ma_count: Total number of minor alleles across individuals</p></li>
<li><p>pvalue: P-value of the main effect from the nonlinear regression</p></li>
<li><p>beta: Slope of the main effect from the nonlinear regression</p></li>
<li><p>se: Standard error of beta</p></li>
<li><p>pvalue_msex: P-value of the msex term from the nonlinear regression</p></li>
<li><p>beta_msex: Slope of the msex term from the nonlinear regression</p></li>
<li><p>se_msex: Standard error of beta_msex</p></li>
<li><p>pvalue_msex_interaction: P-value of the interaction term from the nonlinear regression</p></li>
<li><p>beta_msex_interaction: Slope of the interaction term from the nonlinear regression</p></li>
<li><p>se_msex_interaction: Standard error of beta_msex_interaction</p></li>
<li><p>molecular_trait_object_id: an intermediate id, you can ignore</p></li>
<li><p>n: number of phenotype</p></li>
<li><p>qvalue_main: the false discovery rate of the main effect(pvalue)</p></li>
<li><p>qvalue_interaction: the false discovery rate of the interaction effect(pvalue_msex_interaction)</p></li>
</ul>
<p>The column specification of region (gene) level association evidence are as follows:</p>
<ul class="simple">
<li><p>phenotype_id - Molecular trait identifier. (gene)</p></li>
<li><p>num_var - Total number of variants tested in cis</p></li>
<li><p>beta_shape1 - First parameter value of the fitted beta distribution</p></li>
<li><p>beta_shape2 - Second parameter value of the fitted beta distribution</p></li>
<li><p>true_df - Effective degrees of freedom the beta distribution approximation</p></li>
<li><p>pval_true_df - Empirical P-value for the beta distribution approximation</p></li>
<li><p>variant_id - ID of the top variant (rsid or chr:position:ref:alt)</p></li>
<li><p>tss_distance - Distance of the SNP to the gene transcription start site (TSS)</p></li>
<li><p>ma_samples - Number of samples carrying the minor allele</p></li>
<li><p>ma_count - Total number of minor alleles across individuals</p></li>
<li><p>af - Alternative allele frequency in MiGA cohort</p></li>
<li><p>ref_factor - Flag indicating if the alternative allele is the minor allele in the cohort (1 if AF &lt;= 0.5, -1 if not)</p></li>
<li><p>pval_nominal - Nominal P-value from linear regression</p></li>
<li><p>slope - Slope of the linear regression</p></li>
<li><p>slope_se - Standard error of the slope</p></li>
<li><p>pval_perm - First permutation P-value directly obtained from the permutations with the direct method</p></li>
<li><p>pval_beta - Second permutation P-value obtained via beta approximation. This is the one to use for downstream analysis</p></li>
</ul>
<p>The columns of nominal association results for quantile regression are as follows:</p>
<ul class="simple">
<li><p>phenotype_id: Molecular trait identifier.(gene)</p></li>
<li><p>variant_id: ID of the variant (rsid or chr:position:ref:alt)</p></li>
<li><p>chr : Variant chromosome.</p></li>
<li><p>pos : Variant chromosomal position (basepairs).</p></li>
<li><p>ref : Variant reference allele (A, C, T, or G).</p></li>
<li><p>alt : Variant alternate allele.</p></li>
<li><p>af: Alternative allele frequency in MiGA cohort</p></li>
<li><p>combined_pval(composite-p value using cauchy combination method): the integrated QR p-value across multiple quantile levels.</p></li>
<li><p>qr_0.1_pval to qr_0.9_pval: quantile-specific QR p-values for the quantile levels 0.1, 0.2, …, 0.9.</p></li>
<li><p>qr_0.1_slope to qr_0.9_slope: quantile-specific QR coefficients for the quantile levels 0.1, 0.2, …, 0.9.</p></li>
<li><p>qr_0.1_tval to qr_0.9_tval: quantile-specific QR t-values for the quantile levels 0.1, 0.2, …, 0.9.</p></li>
</ul>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Link to this heading">#</a></h2>
<p>The data can be found on <a class="reference external" href="https://www.synapse.org/#!Synapse:syn36416559/files/">Synapse</a>.</p>
<section id="i-cis-tensorqtl-command">
<h3>i. Cis TensorQTL Command<a class="headerlink" href="#i-cis-tensorqtl-command" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/TensorQTL.ipynb cis \
    --genotype-file output/genotype_by_chrom/protocol_example.genotype.chr21_22.genotype_by_chrom_files.txt \
    --phenotype-file  output/phenotype_by_chrom/protocol_example.protein.bed.phenotype_by_chrom_files.region_list.txt \
    --covariate-file output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz \
    --customized_cis_windows prototype_example/protocol_example/protocol_example.protein.enhanced_cis_chr21_chr22.bed \
    --cwd output/cis_association/ \
    --MAC 5 \
    --container containers/TensorQTL.sif 
</pre></div>
</div>
</div>
</div>
</section>
<section id="ii-trans-tensorqtl-command">
<h3>ii. Trans TensorQTL Command<a class="headerlink" href="#ii-trans-tensorqtl-command" title="Link to this heading">#</a></h3>
<p>Note: Some protein is not in the customized cis windows list. There we will need to remove them from the analysis by create a region_list. Noted that the region list need to be a actual file. So <code class="docutils literal notranslate"><span class="pre">&lt;()</span></code> file is not acceptable.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>zcat output/protocol_example.protein.bed.gz | cut -f 1,2,3,4 | grep -v -e ENSG00000163554 \
    -e ENSG00000171564 -e ENSG00000171560 -e ENSG00000171557 &gt; output/protocol_example.protein.region_list
</pre></div>
</div>
</div>
</div>
<p>The following will take more than 180G of memory to run.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run xqtl-protocol/pipeline/TensorQTL.ipynb trans \
    --genotype-file output/genotype_by_chrom/protocol_example.genotype.chr21_22.genotype_by_chrom_files.txt \
    --phenotype-file  output/phenotype_by_chrom/protocol_example.protein.bed.phenotype_by_chrom_files.region_list.txt \
    --region-list output/protocol_example.protein.region_list \
    --covariate-file output/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.unrelated.plink_qc.prune.pca.Marchenko_PC.gz \
    --customized-cis-windows input/protocol_example.protein.enhanced_cis_chr21_chr22.bed \
    --cwd output/association/trans/ \
    --MAC 5 --numThreads 8 -J 1 -q csg --mem 240G -c /mnt/vast/hpc/csg/molecular_phenotype_calling/csg.yml \
    --container containers/TensorQTL.sif 
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Substep</p></th>
<th class="head"><p>Problem</p></th>
<th class="head"><p>Possible Reason</p></th>
<th class="head"><p>Solution</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="command-interface">
<h2>Command Interface<a class="headerlink" href="#command-interface" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run TensorQTL.ipynb -h
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run TensorQTL.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  cis
  trans

Global Workflow Options:
  --cwd output (as path)
                        Path to the work directory of the analysis.
  --phenotype-file VAL (as path, required)
                        Phenotype file, or a list of phenotype per region.
  --genotype-file VAL (as path, required)
                        A genotype file in PLINK binary format (bed/bam/fam)
                        format, or a list of genotype per chrom
  --covariate-file VAL (as path, required)
                        Covariate file
  --name  f&quot;{phenotype_file:bn}_{covariate_file:bn}&quot;

                        Prefix for the analysis output
  --region-list . (as path)
                        An optional subset of regions of molecular features to
                        analyze. The last column is the gene names
  --region-list-phenotype-column 4 (as int)
  --keep-sample . (as path)
                        Set list of sample to be keep
  --interaction &#39;&#39;
                        FIXME: please document
  --customized-cis-windows . (as path)
                        An optional list documenting the custom cis window for
                        each region to analyze, with four column, chr, start,
                        end, region ID (eg gene ID). If this list is not
                        provided, the default `window` parameter (see below)
                        will be used.
  --phenotype-group . (as path)
                        The phenotype group file to group molecule_trait into
                        molecule_trait_object This applies to multiple molecular
                        events in the same region, such as sQTL analysis.
  --chromosome  (as list)
                        The name of phenotype corresponding to gene_id or
                        gene_name in the region
  --MAC 0 (as int)
                        Minor allele count cutoff
  --window 1000000 (as int)
                        Specify the cis window for the up and downstream radius
                        to analyze around the region of interest in units of bp
                        This parameter will be set to zero if
                        `customized_cis_windows` is provided.
  --numThreads 8 (as int)
                        Number of threads
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 12h
  --mem 16G
  --container &#39;&#39;
                        Container option for software to run the analysis:
                        docker or singularity
  --entrypoint  (&#39;micromamba run -a &quot;&quot; -n&#39; + &#39; &#39; + re.sub(r&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;, &#39;&#39;, container.split(&#39;/&#39;)[-1])) if container else &quot;&quot;

  --maf-threshold  MAC/(2.0*N) 

                        Minor allele frequency cutoff. It will overwrite minor
                        allele cutoff. You may consider setting it to higher for
                        interaction analysis if you have statistical power
                        concerns

Sections
  cis_1:
    Workflow Options:
      --[no-]skip-nominal-if-exist (default to False)
                        parse input file lists skip nominal association results
                        if the files exists already This is false by default
                        which means to recompute everything This is only
                        relevant when the `parquet` files for nominal results
                        exist but not the other files and you want to avoid
                        computing the nominal results again
      --[no-]permutation (default to True)
  cis_2:
  trans_1:
    Workflow Options:
      --batch-size 50000 (as int)
      --pval-threshold 1.0 (as float)
      --[no-]permutation (default to False)
                        Permutation testing is incorrect when the analysis is
                        done by chrom
</pre></div>
</div>
</div>
</div>
</section>
<section id="old-minimal-working-example">
<h2>Old Minimal working example<a class="headerlink" href="#old-minimal-working-example" title="Link to this heading">#</a></h2>
<p>An MWE is uploaded to <a class="reference external" href="https://drive.google.com/drive/folders/1yjTwoO0DYGi-J9ouMsh9fHKfDmsXJ_4I?usp=sharing">google drive</a>.
The singularity image (sif) for running this MWE is uploaded to <a class="reference external" href="https://drive.google.com/drive/folders/1mLOS3AVQM8yTaWtCbO8Q3xla98Nr5bZQ">google drive</a></p>
<p>FIXME: need to update these links.</p>
<p>FIXME: Also need to update the example commands below using our new example dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/TensorQTL.ipynb cis \
    --genotype-file plink_files_list.txt \
    --phenotype-file MWE.bed.recipe \
    --covariate-file ALL.covariate.pca.BiCV.cov.gz \
    --cwd ./output/ \
    --container containers/TensorQTL.sif --MAC 5
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/TensorQTL.ipynb trans \
    --genotype-file plink_files_list.txt \
    --phenotype-file MWE.bed.recipe \
    --covariate-file ALL.covariate.pca.BiCV.cov.gz \
    --cwd ./output/ \
    --container containers/TensorQTL.sif --MAC 5 --region-name  gene_name
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/TensorQTL.ipynb trans \
    --genotype-file plink_files_list.txt \
    --phenotype-file MWE.bed.recipe \
    --covariate-file /mnt/vast/hpc/csg/snuc_pseudo_bulk/eight_tissue_analysis/output/data_preprocessing/ALL/covariates/ALL.log2cpm.ALL.covariate.pca.resid.PEER.cov.gz \
    --cwd ./output/trans_tensorQTL/ \
    --region-list /mnt/vast/hpc/csg/snuc_pseudo_bulk/eight_tissue_analysis/reference_data/AD_genes.region_list \
    --customized-cis-windows TADB_enhanced_cis.bed \
    --container containers/TensorQTL.sif --MAC 5 --region-name ID
</pre></div>
</div>
</div>
</div>
</section>
<section id="setup-and-global-parameters">
<h2>Setup and global parameters<a class="headerlink" href="#setup-and-global-parameters" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[global]
# Path to the work directory of the analysis.
parameter: cwd = path(&#39;output&#39;)
# Phenotype file, or a list of phenotype per region.
parameter: phenotype_file = path
# A genotype file in PLINK binary format (bed/bam/fam) format, or a list of genotype per chrom
parameter: genotype_file = path
# Covariate file
parameter: covariate_file = path
# Prefix for the analysis output
parameter: name = f&quot;{phenotype_file:bn}_{covariate_file:bn}&quot;
# An optional subset of regions of molecular features to analyze. The last column is the gene names
parameter: region_list = path()
parameter: region_list_phenotype_column = 4
# Set list of sample to be keep
parameter: keep_sample = path()
# FIXME: please document
parameter: interaction = &quot;&quot;

# An optional list documenting the custom cis window for each region to analyze, with four column, chr, start, end, region ID (eg gene ID).
# If this list is not provided, the default `window` parameter (see below) will be used.
parameter: customized_cis_windows = path()

# The phenotype group file to group molecule_trait into molecule_trait_object
# This applies to multiple molecular events in the same region, such as sQTL analysis.
parameter: phenotype_group = path() 

# The name of phenotype corresponding to gene_id or gene_name in the region
parameter: chromosome = []
# Minor allele count cutoff
parameter: MAC = 0

# Specify the cis window for the up and downstream radius to analyze around the region of interest in units of bp
# This parameter will be set to zero if `customized_cis_windows` is provided.
parameter: window = 1000000

# Number of threads
parameter: numThreads = 8
# For cluster jobs, number commands to run per job
parameter: job_size = 1
parameter: walltime = &#39;12h&#39;
parameter: mem = &#39;16G&#39;
# Container option for software to run the analysis: docker or singularity
parameter: container = &#39;&#39;
import re

# Use the header of the covariate file to decide the sample size
import pandas as pd
N = len(pd.read_csv(covariate_file, sep = &quot;\t&quot;,nrows = 1).columns) - 1

# Minor allele frequency cutoff. It will overwrite minor allele cutoff.
# You may consider setting it to higher for interaction analysis if you have statistical power concerns
parameter: maf_threshold = MAC/(2.0*N)

# Filtering significant trans associations (for trans_2 workflow)
parameter: pvalue_cutoff = &quot;&quot;
parameter: qvalue_cutoff = &quot;0.01&quot;  # default cutoff is 0.01


import os
import pandas as pd

def adapt_file_path(file_path, reference_file):
    &quot;&quot;&quot;
    Adapt a single file path based on its existence and a reference file&#39;s path.

    Args:
    - file_path (str): The file path to adapt.
    - reference_file (str): File path to use as a reference for adaptation.

    Returns:
    - str: Adapted file path.

    Raises:
    - FileNotFoundError: If no valid file path is found.
    &quot;&quot;&quot;
    reference_path = os.path.dirname(reference_file)

    # Check if the file exists
    if os.path.isfile(file_path):
        return file_path

    # Check file name without path
    file_name = os.path.basename(file_path)
    if os.path.isfile(file_name):
        return file_name

    # Check file name in reference file&#39;s directory
    file_in_ref_dir = os.path.join(reference_path, file_name)
    if os.path.isfile(file_in_ref_dir):
        return file_in_ref_dir

    # Check original file path prefixed with reference file&#39;s directory
    file_prefixed = os.path.join(reference_path, file_path)
    if os.path.isfile(file_prefixed):
        return file_prefixed

    # If all checks fail, raise an error
    raise FileNotFoundError(f&quot;No valid path found for file: {file_path}&quot;)

def adapt_file_path_all(df, column_name, reference_file):
    return df[column_name].apply(lambda x: adapt_file_path(x, reference_file))


if str(genotype_file).endswith(&quot;bed&quot;) and str(phenotype_file).endswith(&quot;bed.gz&quot;):
    input_files = [[phenotype_file, genotype_file]]
    if len(chromosome) &gt; 0:
        input_chroms = [int(x) for x in chromosome]
    else:
        input_chroms = [0]
else:
    import pandas as pd
    import os
    molecular_pheno_files = pd.read_csv(phenotype_file, sep = &quot;\t&quot;)
    if &quot;#chr&quot; in molecular_pheno_files.columns:
        molecular_pheno_files = molecular_pheno_files.groupby([&#39;#chr&#39;,&#39;path&#39;]).size().reset_index(name=&#39;count&#39;).drop(&quot;count&quot;,axis = 1).rename(columns = {&quot;#chr&quot;:&quot;#id&quot;})
    genotype_files = pd.read_csv(genotype_file,sep = &quot;\t&quot;)
    genotype_files[&quot;#id&quot;] = [x.replace(&quot;chr&quot;,&quot;&quot;) for x in genotype_files[&quot;#id&quot;].astype(str)] # e.g. remove chr1 to 1
    genotype_files[&quot;#path&quot;] = genotype_files[&quot;#path&quot;].apply(lambda x: adapt_file_path(x, genotype_file))
    molecular_pheno_files[&quot;#id&quot;] = [x.replace(&quot;chr&quot;,&quot;&quot;) for x in molecular_pheno_files[&quot;#id&quot;].astype(str)]
    input_files = molecular_pheno_files.merge(genotype_files, on = &quot;#id&quot;)
    # Only keep chromosome specified in --chromosome
    if len(chromosome) &gt; 0:
        input_files = input_files[input_files[&#39;#id&#39;].isin(chromosome)]
    input_files = input_files.values.tolist()
    input_chroms = [x[0] for x in input_files]
    input_files = [x[1:] for x in input_files]
</pre></div>
</div>
</div>
</div>
</section>
<section id="cis-xqtl-association-testing">
<h2>cis-xQTL association testing<a class="headerlink" href="#cis-xqtl-association-testing" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[cis_1]
# parse input file lists
# skip nominal association results if the files exists already
# This is false by default which means to recompute everything
# This is only relevant when the `parquet` files for nominal results exist but not the other files
# and you want to avoid computing the nominal results again
parameter: skip_nominal_if_exist = False
parameter: permutation = True
parameter: quantile_regression = False
parameter: qr_vcov= &#39;robust&#39;
parameter: qr_kernel= &#39;epa&#39;
parameter: qr_bandwidth= &#39;hsheather&#39;
parameter: qr_max_iter=200
parameter: qr_p_tol=0.000001
parameter: qr_regress_covariates = True
parameter: qr_add_intercept = False
parameter: qr_scale = False
# use -1 to set the maximum number of cores available
parameter: qr_threads = -1
if quantile_regression:
    permutation = False

# Extract interaction name
var_interaction = interaction
if os.path.isfile(interaction):
    interaction_s = pd.read_csv(interaction, sep=&#39;\t&#39;, index_col=0)
    var_interaction = interaction_s.columns[0] # interaction name
test_regional_association = permutation and len(var_interaction) == 0

input: input_files, group_by = len(input_files[0]), group_with = &quot;input_chroms&quot;
output_files = dict([(&quot;parquet&quot;, f&#39;{cwd:a}/{_input[0]:bnn}{&quot;_%s&quot; % var_interaction if interaction else &quot;&quot;}.cis_qtl_pairs.{&quot;&quot; if input_chroms[_index] == 0 else input_chroms[_index]}.parquet&#39;), # This convention is necessary to match the pattern of map_norminal output
                     (&quot;nominal&quot;, f&#39;{cwd:a}/{_input[0]:bnn}{&quot;&quot; if input_chroms[_index] == 0 else &quot;_chr%s&quot; % input_chroms[_index]}{&quot;_%s&quot; % var_interaction if interaction else &quot;&quot;}.cis_{&quot;quantile_&quot; if quantile_regression else &quot;&quot;}qtl.pairs.tsv.gz&#39;)])
if test_regional_association:
    output_files[&quot;regional&quot;] = f&#39;{cwd:a}/{_input[0]:bnn}{&quot;&quot; if input_chroms[_index] == 0 else &quot;_chr%s&quot; % input_chroms[_index]}{&quot;_%s&quot; % var_interaction if interaction else &quot;&quot;}.cis_qtl.regional.tsv.gz&#39;
output: output_files
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[&quot;nominal&quot;]:bnnn}&#39;
python: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[&quot;nominal&quot;]:nnn}.stderr&#39;, stdout = f&#39;{_output[&quot;nominal&quot;]:nnn}.stdout&#39; , container = container
    import pandas as pd
    import numpy as np
    import os
    import tensorqtl
    from tensorqtl import genotypeio, cis
    from scipy.stats import chi2
    import multiprocessing as mp
    from tqdm import tqdm

    ## Define paths
    plink_prefix_path = $[_input[1]:nar]
    expression_bed = $[_input[0]:ar]
    covariates_file = &quot;$[covariate_file:a]&quot;
    window = $[window]
    interaction = &quot;$[interaction]&quot;
    ## Load Data
    phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expression_bed)
    phenotype_id = phenotype_pos_df.index.name

    ## Analyze only the regions listed
    if $[region_list.is_file()]:
        region = pd.read_csv(&quot;$[region_list:a]&quot;, comment=&quot;#&quot;, header=None, sep=&quot;\t&quot; )
        phenotype_column = 1 if len(region.columns) == 1 else  $[region_list_phenotype_column]
        keep_region = region.iloc[:,phenotype_column-1].to_list()
        phenotype_df = phenotype_df[phenotype_df.index.isin(keep_region)]
        phenotype_pos_df = phenotype_pos_df[phenotype_pos_df.index.isin(keep_region)]
    covariates_df = pd.read_csv(covariates_file, sep=&#39;\t&#39;, index_col=0).T
    pr = genotypeio.PlinkReader(plink_prefix_path)
    genotype_df = pr.load_genotypes()
    variant_df = pr.bim.set_index(&#39;snp&#39;)[[&#39;chrom&#39;, &#39;pos&#39;,&#39;a0&#39;,&#39;a1&#39;]]
    
    ## use custom sample list to subset the covariates data
    if $[keep_sample.is_file()]:
        sample_list = pd.read_csv(&quot;$[keep_sample:a]&quot;, comment=&quot;#&quot;, header=None, names=[&quot;sample_id&quot;], sep=&quot;\t&quot;)
        covariates_df.loc[sample_list.sample_id]

    # Read interaction files or extract from covariates file.
    var_interaction = interaction
    interaction_s = []
    if os.path.isfile(interaction):
        # update var_interaction and interaction_s
        interaction_s = pd.read_csv(interaction, sep=&#39;\t&#39;, index_col=0)
        interaction_s = interaction_s[interaction_s.index.isin(covariates_df.index)] 
        var_interaction = interaction_s.columns[0] # interaction name
    # check if the interaction term in interaction table is in covariates file, if yes and interaction_s not yet loaded then, extract it out from covariates file
    if var_interaction in covariates_df.columns:
        # only load from covariate if it has not been loaded yet
        if len(interaction_s) == 0:
            interaction_s = covariates_df[var_interaction].to_frame()
        covariates_df = covariates_df.drop(columns=[var_interaction])
    if len(interaction) and len(interaction_s) == 0:
        raise ValueError(f&quot;Cannot find interaction variable or file {interaction}&quot;)

    # drop samples that with missing value in iteraction
    if len(interaction_s):
        interaction_s = interaction_s.dropna() 

    ## Retaining only common samples
    phenotype_df = phenotype_df[np.intersect1d(phenotype_df.columns, covariates_df.index)]
    phenotype_df = phenotype_df[np.intersect1d(phenotype_df.columns, genotype_df.columns)] 
    if len(interaction_s):
        phenotype_df = phenotype_df[np.intersect1d(phenotype_df.columns, interaction_s.index)]
        interaction_s = interaction_s[interaction_s.index.isin(phenotype_df.columns)]    
        interaction_s = interaction_s.loc[phenotype_df.columns]

    covariates_df = covariates_df.transpose()[np.intersect1d(phenotype_df.columns, covariates_df.index)].transpose()
    
    ## To simplify things, there should really not be &quot;chr&quot; prefix
    phenotype_pos_df.chr = phenotype_pos_df.chr.astype(str).str.replace(&quot;chr&quot;, &quot;&quot;)
    variant_df.chrom =  variant_df.chrom.astype(&quot;str&quot;).str.replace(&quot;chr&quot;, &quot;&quot;) 

    ## use custom cis windows list
    if $[customized_cis_windows.is_file()]:
        cis_list = pd.read_csv(&quot;$[customized_cis_windows:a]&quot;, comment=&quot;#&quot;, header=None, names=[&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;,phenotype_id], sep=&quot;\t&quot;)
        cis_list.chr = cis_list.chr.astype(str).str.replace(&quot;chr&quot;, &quot;&quot;)  ## Again to simplify things for chr format concordance.
        if cis_list[[&#39;chr&#39;, &#39;ID&#39;]].duplicated().sum() != 0: # if cis_list is not unique using identifier [&#39;#chr&#39;, &#39;ID&#39;]
                cis_list = cis_list.groupby([&#39;ID&#39;, &#39;chr&#39;]).agg({ # use union start-end position and make cis_list unique
                    &#39;start&#39;: &#39;min&#39;,
                    &#39;end&#39;: &#39;max&#39;
                }).reset_index()

                cis_list = cis_list[[&#39;chr&#39;, &#39;start&#39;, &#39;end&#39;, &#39;ID&#39;]]
                #cis_list = cis_list.set_index(&#39;chr&#39;)
        
        phenotype_pos_df = phenotype_pos_df.reset_index() #move the phenotype id index to a new column of the dataframe
        phenotype_df = phenotype_df.reset_index()
        # Ensure phenotype_pos_df is a subset of cis_list based on [&#39;#chr&#39;, &#39;ID&#39;]
        original_count = len(phenotype_pos_df)
        phenotype_pos_df = phenotype_pos_df[phenotype_pos_df.set_index([&#39;chr&#39;, &#39;ID&#39;]).index.isin(cis_list.set_index([cis_list.columns[0],cis_list.columns[3]]).index)] 
        phenotype_df = phenotype_df[phenotype_df.set_index(&#39;ID&#39;).index.isin(cis_list.set_index(cis_list.columns[3]).index)] 
        phenotype_df = phenotype_df.set_index(&#39;ID&#39;)
        removed_count = original_count - len(phenotype_pos_df)
        print(f&quot;{removed_count} rows were removed from phenotype_pos_df&quot;)

        # Merge the dataframes on &#39;chr&#39; and &#39;ID&#39;, including &#39;start&#39; and &#39;end&#39;
        phenotype_pos_df = phenotype_pos_df.merge(cis_list[[&#39;chr&#39;, &#39;ID&#39;, &#39;start&#39;, &#39;end&#39;]], 
                                        left_on = [&quot;chr&quot;,phenotype_id],
                                        right_on = [cis_list.columns[0],cis_list.columns[3]],
                                        suffixes=(&#39;_pheno&#39;, &#39;&#39;))

        # Function to decide whether to keep or rename columns: 
        # If &#39;start&#39; and &#39;end&#39; values are the same in both dataframes, we keep the original columns without suffixes; 
        # If they&#39;re different, we name columns from cis_list with &#39;_cis&#39; suffixes and keep the name still for columns from phentype_pos_df.
        # #in some cases (gene expression for eQTLs) the phenotype_id may be in the cis_list file
        def rename_if_different(row, col):
            if row[f&#39;{col}&#39;] == row[f&#39;{col}_pheno&#39;]:
                return row[col]
            else:
                return pd.Series({f&#39;{col}_pheno&#39;: row[f&#39;{col}_pheno&#39;], f&#39;{col}_cis&#39;: row[col]})
        
        # Apply the renaming logic
        for col in [&#39;start&#39;, &#39;end&#39;]:
            if f&#39;{col}_pheno&#39; in phenotype_pos_df.columns:# this condition is to not execute when the original phenotype_pos_df don&#39;t have start, end but pos column.
                temp = phenotype_pos_df.apply(lambda row: rename_if_different(row, col), axis=1)
                phenotype_pos_df = phenotype_pos_df.drop(columns=[f&#39;{col}_pheno&#39;]).join(temp)
                print(f&quot;Dropped columns due to value mismatch: {col}_pheno&quot;)
        
        phenotype_pos_df = phenotype_pos_df.set_index(phenotype_id)[[&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;]] # The final phenotype_pos_df will have three columns(chr, start, end) and index is the phenotype ID
        
        if len(phenotype_df.index) != len(phenotype_pos_df.index):
            raise ValueError(&quot;cannot uniquely match all the phentoype data in the input to the customized cis windows provided&quot;)
        window = 0 # In the updated tensorQTL, by default if there is a customized cis window, the actual cis window will be start - window &amp; end + window, so it is necessary to change the window parameter to 0

    ## Read phenotype group if availble
    if $[phenotype_group.is_file()]:
        group_s = pd.read_csv($[phenotype_group:r], sep=&#39;\t&#39;, header=None, index_col=0).squeeze()
    else:
        group_s = None

    if $[&quot;True&quot; if quantile_regression else &quot;False&quot;]:
        import statsmodels.api as sm
        import scipy.stats as ss
        from scipy.stats import cauchy
        import time
        import warnings
        import torch
        # Perform quantile regression
        start_time = time.time()
        def filter_maf_with_numpy(genotypes, variant_ids, maf_threshold, alleles=2):
            af = genotypes.sum(axis=1) / (alleles * genotypes.shape[1])
            maf = np.where(af &gt; 0.5, 1 - af, af)
            if maf_threshold &gt; 0:
                mask = maf &gt;= maf_threshold
                #mask = mask.astype(bool)
                genotypes = genotypes[mask]
                variant_ids = np.array(variant_ids)[mask]
                af = af[mask]
                maf = maf[mask]
            return genotypes, variant_ids, af, maf

        def remove_covariate_effects_np(X, Z, y, scale=False):
            # for X,Z,y, rows are samples
            if not np.all(Z[:, 0] == 1):
                Z = np.hstack([np.ones((Z.shape[0], 1)), Z])
            
            A = np.linalg.inv(Z.T @ Z)
            SZy = A @ Z.T @ y
            SZX = A @ Z.T @ X
            
            # Adjust y and X based on calculated effects
            y_adjusted = y - Z @ SZy
            X_adjusted = X - Z @ SZX
            
            if scale:
                # Standardize y and X
                y_scaled = (y_adjusted - y_adjusted.mean()) / y_adjusted.std()
                X_scaled = (X_adjusted - X_adjusted.mean(axis=0)) / X_adjusted.std(axis=0)
                return X_scaled, y_scaled
            else:
                return X_adjusted, y_adjusted

        def impute_mean(genotypes_t, missing=-1):#for torch tensor genotypes_t,cols are samples
            m = genotypes_t == missing
            ix = torch.nonzero(m, as_tuple=True)[0]
            if len(ix) &gt; 0:
                a = genotypes_t.sum(1)
                b = m.sum(1).float()
                mu = (a - missing*b) / (genotypes_t.shape[1] - b)
                genotypes_t[m] = mu[ix]

        def drop_zero_variance_rows(genotypes, variant_ids):
            # Calculate the variance of each row
            variances = np.var(genotypes, axis=1)
            # Find the indices of rows with non-zero variance
            non_zero_variance_indices = np.where(variances != 0)[0]
            # Select rows with non-zero variance in genotypes and corresponding variant_ids
            genotypes_updated = genotypes[non_zero_variance_indices, :]
            variant_ids_updated = [variant_ids[i] for i in non_zero_variance_indices.tolist()]
            return genotypes_updated, variant_ids_updated

        def cauchy_meta(pvals):
            # Check input
            pvals = np.array(pvals)
            pvals = pvals[~np.isnan(pvals)]
            if len(pvals) == 0:
                return np.nan
            # pvals[pvals == 0] = 2.2e-308
            # Convert to Cauchy
            cauchy_vals = np.zeros(pvals.shape)
            valid_indices = pvals &gt;= 1e-15
            cauchy_vals[valid_indices] = np.tan(np.pi * (0.5 - pvals[valid_indices]))
            stats = np.mean(cauchy_vals)
            p = cauchy.sf(stats)
            return p

        def fit_quantreg(phenotype_id, phenotype_array, genotype_matrix, variant_ids, af, covar_df, taus=[0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9], vcov=&#39;$[qr_vcov]&#39;, kernel=&#39;$[qr_kernel]&#39;, bandwidth=&#39;$[qr_bandwidth]&#39;, max_iter=$[qr_max_iter], p_tol=$[qr_p_tol]):
            results = []
            variant_pvals = {vid: [] for vid in variant_ids}
            
            if covar_df is not None:
                if genotype_matrix.shape[1] != covar_df.shape[0]:
                    raise ValueError(&quot;Mismatch in the number of samples between genotype matrix and covariates DataFrame&quot;)
            
            for tau in taus:
                for idx, genotype_row in enumerate(genotype_matrix):
                    current_af = af[idx]
                    if covar_df is not None:
                        X = np.c_[covar_df.values, genotype_row]

                    else:
                        # If covar_df is None, just use the genotype_row
                        X = genotype_row.reshape(-1, 1)
                        
                    if $[&quot;True&quot; if qr_add_intercept else &quot;False&quot;]:
                        X = sm.add_constant(X).astype(&#39;float&#39;)
                    
                    model = sm.QuantReg(phenotype_array, X)
                    try:
                        res = model.fit(q=tau, vcov=vcov, kernel=kernel, bandwidth=bandwidth, max_iter=max_iter, p_tol=p_tol)
                        pval = res.pvalues[-1]
                        variant_pvals[variant_ids[idx]].append(pval)
                        results.append({
                            &#39;phenotype_id&#39;: phenotype_id,
                            &#39;variant_id&#39;: variant_ids[idx],
                            &#39;af&#39;: current_af, 
                            &#39;tau&#39;: tau,
                            &#39;pval&#39;: pval,
                            &#39;slope&#39;: res.params[-1],
                            &#39;tval&#39;: res.tvalues[-1]
        
                        })
                    except Exception as e:
                        raise RuntimeError(f&quot;Error fitting model for variant {variant_ids[idx]} at tau {tau}: {e}&quot;)

            # Calculate combined p-values for each variant
            start_time_for_combined_p = time.time()
            for vid in variant_ids:
                combined_pval = cauchy_meta(variant_pvals[vid])
                for result in filter(lambda r: r[&#39;variant_id&#39;] == vid, results):
                    result[&#39;combined_pval&#39;] = combined_pval
            
            results_df = pd.DataFrame(results)
            end_time_for_combined_p = time.time()
            wide_df = pd.pivot_table(results_df, values=[&#39;pval&#39;, &#39;slope&#39;, &#39;tval&#39;], index=[&#39;phenotype_id&#39;, &#39;variant_id&#39;, &#39;af&#39;,&#39;combined_pval&#39;], columns=&#39;tau&#39;)
            wide_df.columns = [&#39;qr_{}_{}&#39;.format(col[1], col[0]) for col in wide_df.columns]
            wide_df.reset_index(inplace=True)

            variant_split = wide_df[&#39;variant_id&#39;].str.extract(r&#39;chr(\d+):(\d+)_([A-Z*]+)_([A-Z*]+)&#39;)
            variant_split.columns = [&#39;chr&#39;, &#39;pos&#39;, &#39;ref&#39;, &#39;alt&#39;]
            final_df = pd.concat([variant_split, wide_df], axis=1)
            return final_df

        # get same order and common samples of phenotype_df, genotype_df, and covariates_df
        covariates_df.index = covariates_df.index.astype(str)
        common_samples = np.intersect1d(phenotype_df.columns, genotype_df.columns)
        genotype_df = genotype_df[common_samples]          
        sample_order = covariates_df.index
        phenotype_df = phenotype_df[sample_order]
        genotype_df = genotype_df[sample_order]

        # quantile qtl mapping:
        device = torch.device(&quot;cuda&quot; if torch.cuda.is_available() else &quot;cpu&quot;)
        chrom = None
        igc = genotypeio.InputGeneratorCis(genotype_df, variant_df, phenotype_df, phenotype_pos_df, group_s=group_s, window=window)
        results_list = [] 

        def quantile_regression_wrapper(args):
            phenotype, genotypes, genotype_range, phenotype_id = args
            genotypes = torch.tensor(genotypes, dtype=torch.float).to(device)
            impute_mean(genotypes)
            if genotypes.is_cuda:
                genotypes = genotypes.cpu()
            genotypes = genotypes.numpy()
            # drop_zero_variance_rows
            variant_ids = variant_df.index[genotype_range[0]:genotype_range[-1]+1].tolist()
            genotypes, variant_ids = drop_zero_variance_rows(genotypes, variant_ids)   
            # filter_maf_with_numpy  
            genotypes, variant_ids, af, maf = filter_maf_with_numpy(genotypes, variant_ids, $[maf_threshold])
        
            # regress out covariates from y and X
            if $[&quot;True&quot; if qr_regress_covariates else &quot;False&quot;]:
                covar_np = covariates_df.to_numpy()
                X_adj, phenotype = remove_covariate_effects_np(genotypes.T, covar_np, phenotype, scale = &#39;$[qr_scale]&#39;)
                genotypes = X_adj.T    
                quantqtl_start_time = time.time()
                with warnings.catch_warnings():
                    warnings.simplefilter(&quot;ignore&quot;)
                    results_df = fit_quantreg(phenotype_id=phenotype_id,phenotype_array=phenotype,genotype_matrix=genotypes,covar_df=None,variant_ids=variant_ids,af=af
                    )                
            else:
                with warnings.catch_warnings():
                    warnings.simplefilter(&quot;ignore&quot;)
                    results_df = fit_quantreg(phenotype_id=phenotype_id,phenotype_array=phenotype,genotype_matrix=genotypes,covar_df=covariates_df,variant_ids=variant_ids,af=af
                    ) 
        
            quantqtl_end_time = time.time()
            runtime = quantqtl_end_time - quantqtl_start_time
            return results_df, runtime

        # quantile regression main function
        qr_threads = $[qr_threads]
        max_cores = mp.cpu_count()
        
        if qr_threads &lt;= 0:
            qr_threads = max_cores
        else:
            qr_threads = min(qr_threads, max_cores)

        results_list = []
        runtime_list = []
        if qr_threads == 1:
            for phenotype, genotypes, genotype_range, phenotype_id in igc.generate_data(chrom=chrom, verbose=True):
                results_df, runtime = quantile_regression_wrapper((phenotype, genotypes, genotype_range, phenotype_id))
                results_list.append(results_df)
                runtime_list.append(runtime)
        else:
            pool = mp.Pool(processes=qr_threads)       
            phenotype_args = [(phenotype, genotypes, genotype_range, phenotype_id)
                              for phenotype, genotypes, genotype_range, phenotype_id in igc.generate_data(chrom=chrom, verbose=True)]
        
            with tqdm(total=len(phenotype_args), desc=&quot;Quantile Regression&quot;) as progress_bar:
                for results_df, runtime in pool.imap_unordered(quantile_regression_wrapper, phenotype_args):
                    results_list.append(results_df)
                    runtime_list.append(runtime)
                    progress_bar.update(1)
        
            pool.close()
            pool.join()       

        quantile_reg_df = pd.concat(results_list, ignore_index=True)
        # sort the table if chrom and pos is not in ascending order
        quantile_reg_df[&#39;pos&#39;] = pd.to_numeric(quantile_reg_df[&#39;pos&#39;])
        if not all(quantile_reg_df[&#39;pos&#39;].iloc[i] &lt;= quantile_reg_df[&#39;pos&#39;].iloc[i+1] for i in range(len(quantile_reg_df)-1)):
            quantile_reg_df = quantile_reg_df.sort_values(by=[&#39;chr&#39;, &#39;pos&#39;])        
        end_time = time.time()  
        print(f&quot;Total execution time from the beginning to the quantile regression: {end_time - start_time} seconds&quot;)

        # save file
        quantile_reg_df.to_csv($[_output[&quot;nominal&quot;]:nr], sep=&#39;\t&#39;, index = None)
        # print general information of quantile_reg_df
        print(&#39;Output Information:&#39;)
        print(&quot;Output Rows:&quot;, len(quantile_reg_df))
        print(&quot;Output Columns:&quot;, quantile_reg_df.columns.tolist())
        print(&quot;Output Preview:&quot;, quantile_reg_df.iloc[1:5, 1:10])
        pass
    else:
        ## cis-QTL mapping: nominal associations for all variant-phenotype pairs
        if not ($[skip_nominal_if_exist] and $[_output[&quot;parquet&quot;].is_file()]):
            if len(interaction_s):
                cis.map_nominal(genotype_df, variant_df, 
                        phenotype_df, 
                        phenotype_pos_df, 
                        $[_output[&quot;parquet&quot;]:nnnr],
                        covariates_df=covariates_df,
                        interaction_df=interaction_s, 
                        maf_threshold_interaction=$[maf_threshold],
                        window=window,
                        group_s=group_s,
                        run_eigenmt=True, write_top=True, write_stats=True)
            else:
                cis.map_nominal(genotype_df, variant_df,
                    phenotype_df,
                    phenotype_pos_df,
                    $[_output[&quot;parquet&quot;]:nnnr],
                    covariates_df=covariates_df, 
                    window=window, 
                    maf_threshold = $[maf_threshold],
                    group_s=group_s)

        ## Load the parquet and save it as txt
        pairs_df = pd.read_parquet($[_output[&quot;parquet&quot;]:r])
        ## Remove rows whose &#39;pval_gi&#39; is null for following t_pval_conversion
        if len(interaction_s):
            pairs_df = pairs_df.dropna(subset=[&#39;pval_gi&#39;])
        # print general information of parquet
        print(&#39;Output Information:&#39;)
        print(&quot;This is the file containing the immediate output of TensorQTL&#39;s map_nominal function &quot;)
        print(os.path.getsize($[_output[&quot;parquet&quot;]:r]))

        ## Adds the group columns to pairs_df, if there is group_s use group_s, else use phenotype_id
        if group_s is not None:
            pairs_df = pairs_df.merge(pd.DataFrame( {&quot;molecular_trait_object_id&quot;: group_s}),left_on = &quot;phenotype_id&quot;, right_index = True)
        else:
            pairs_df[&quot;molecular_trait_object_id&quot;] = pairs_df.phenotype_id
        ## if pos in phenotype_pos_df(start distance and end distance are the same), 
        ## add the column &#39;end_distance&#39; with the same value as &#39;start_distance&#39; to avoid mismatch of column and names
        if &#39;end_distance&#39; not in pairs_df.columns:
            # Get the position of &#39;start_distance&#39; column
            start_pos = pairs_df.columns.get_loc(&#39;start_distance&#39;)
            # Create a new DataFrame with the same values as start_distance
            new_df = pairs_df.copy()
            # Insert end_distance column after start_distance
            new_df.insert(start_pos + 1, &#39;end_distance&#39;, pairs_df[&#39;start_distance&#39;])
            pairs_df = new_df

        # rename columns
        pairs_df.columns.values[0]  = &quot;molecular_trait_id&quot;
        pairs_df.columns.values[7]  = &quot;pvalue&quot;
        pairs_df.columns.values[8]  = &quot;beta&quot;
        pairs_df.columns.values[9]  = &quot;se&quot;
        pairs_df.rename(columns={&#39;TSS_D&#39;: &#39;tss_distance&#39;}, inplace=True)
        if len(interaction_s):
            # calculate genomic inflation factor lambda on interaction 
            lambda_col_interaction = pairs_df.groupby(&quot;molecular_trait_object_id&quot;).apply(lambda x: chi2.ppf(1. - np.median(x.pval_gi), 1)/chi2.ppf(0.5,1))
            pairs_df.columns.values[10]  = &quot;pvalue&quot; + &#39;_&#39; + var_interaction
            pairs_df.columns.values[11]  = &quot;beta&quot; + &#39;_&#39; + var_interaction
            pairs_df.columns.values[12]  = &quot;se&quot; + &#39;_&#39; + var_interaction
            pairs_df.columns.values[13]  = &quot;pvalue&quot; + &#39;_&#39; + var_interaction + &#39;_&#39; + &#39;interaction&#39;
            pairs_df.columns.values[14]  = &quot;beta&quot; + &#39;_&#39; + var_interaction + &#39;_&#39; + &#39;interaction&#39;
            pairs_df.columns.values[15]  = &quot;se&quot; + &#39;_&#39; + var_interaction + &#39;_&#39; + &#39;interaction&#39; 

        pairs_df[&quot;n&quot;] = len(phenotype_df.columns.values)
        pairs_df = variant_df.merge(pairs_df, right_on=&#39;variant_id&#39;, left_index=True)
        pairs_df.rename(columns={&#39;a1&#39;: &#39;a2&#39;, &#39;a0&#39;: &#39;a1&#39;}, inplace=True)
        # sort the table if chrom and pos is not in ascending order
        if not all(pairs_df[&#39;pos&#39;].iloc[i] &lt;= pairs_df[&#39;pos&#39;].iloc[i+1] for i in range(len(pairs_df)-1)):
            pairs_df = pairs_df.sort_values(by=[&#39;chrom&#39;, &#39;pos&#39;])
        # save file
        pairs_df.to_csv($[_output[&quot;nominal&quot;]:nr], sep=&#39;\t&#39;, index = None)
        # print general information of pairs_df
        print(&#39;Output Information:&#39;)
        print(&quot;Output Rows:&quot;, len(pairs_df))
        print(&quot;Output Columns:&quot;, pairs_df.columns.tolist())
        print(&quot;Output Preview:&quot;, pairs_df.iloc[1:5, 1:10])

        if $[test_regional_association]:
            # calculate genomic inflation factor lambda for main variant effect 
            lambda_col = pairs_df.groupby(&quot;molecular_trait_object_id&quot;).apply(lambda x: chi2.ppf(1. - np.median(x.pvalue), 1)/chi2.ppf(0.5,1))
            cis_df = cis.map_cis(genotype_df, 
                                variant_df, 
                                phenotype_df,
                                phenotype_pos_df,
                                covariates_df=covariates_df, 
                                seed=999, 
                                window=window, 
                                maf_threshold = $[maf_threshold],
                                group_s=group_s)
            cis_df.index.name = &quot;molecular_trait_id&quot;
            ## Add groups columns for eQTL analysis
            if &quot;group_id&quot; not in cis_df.columns:
                cis_df[&quot;group_id&quot;] = cis_df.index
                cis_df[&quot;group_size&quot;] = 1
            cis_df = cis_df.rename({&quot;group_id&quot;:&quot;molecular_trait_object_id&quot;,&quot;group_size&quot;:&quot;n_traits&quot;,&quot;num_var&quot; : &quot;n_variants&quot;,&quot;pval_perm&quot;:&quot;p_perm&quot;, &quot;pval_beta&quot;:&quot;p_beta&quot; },axis = 1)
            cis_df = cis_df.assign(genomic_inflation = lambda dataframe : dataframe[&quot;molecular_trait_object_id&quot;].map(lambda molecular_trait_object_id:lambda_col[molecular_trait_object_id]))
            # merge cis_df with variant_df
            cis_df = variant_df.merge(cis_df, right_on=&#39;variant_id&#39;, left_index=True)
            cis_df.rename(columns={&#39;a1&#39;: &#39;a2&#39;, &#39;a0&#39;: &#39;a1&#39;, &#39;variant_id&#39;: &#39;variant&#39;}, inplace=True)
            # sort the table if chrom and pos is not in ascending order
            if not all(cis_df[&#39;pos&#39;].iloc[i] &lt;= cis_df[&#39;pos&#39;].iloc[i+1] for i in range(len(cis_df)-1)):
                cis_df = cis_df.sort_values(by=[&#39;chrom&#39;, &#39;pos&#39;])
            # save file
            cis_df.to_csv(str($[_output[&quot;nominal&quot;]:nnnr])+str(&#39;.regional.tsv&#39;), sep=&#39;\t&#39;, index = None)
            # print general information of cis_df
            print(&#39;Output Information:&#39;)
            print(&quot;Output Rows:&quot;, len(cis_df))
            print(&quot;Output Columns:&quot;, cis_df.columns.tolist())
            print(&quot;Output Preview:&quot;, cis_df.iloc[0:5, 0:10])

R: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[&quot;nominal&quot;]:nnn}.stderr&#39;, stdout = f&#39;{_output[&quot;nominal&quot;]:nnn}.stdout&#39;, container = container
    library(purrr)
    library(tidyr)
    library(readr)
    library(dplyr)
    library(qvalue)
  
    pairs_df = read_delim($[_output[&quot;nominal&quot;]:nr,], delim = &#39;\t&#39;)
    compute_qvalues &lt;- function(pvalues) {
        tryCatch({
            if(length(pvalues) &lt; 2) {
                return(pvalues)
            } else {
                return(qvalue(pvalues)$qvalues)
            }
        }, error = function(e) {
            message(&quot;Too few p-values to calculate qvalue, fall back to BH&quot;)
            qvalue(pvalues, pi0 = 1)$qvalues
        })
    }
    
    quantile_regression &lt;- &quot;$[quantile_regression]&quot;
    if (quantile_regression) {
        pairs_df &lt;- pairs_df %&gt;%
            group_by(phenotype_id) %&gt;%
            mutate(
                across(
                    .cols = ends_with(&quot;_pval&quot;), 
                    .fns = ~ compute_qvalues(.x),
                    .names = &quot;{sub(&#39;_pval$&#39;, &#39;_qval&#39;, .col)}&quot;
                )
            )
    } else {
        var_interaction &lt;- &quot;$[interaction]&quot;
        # Check if &#39;interaction&#39; is a file
        if (file.exists(var_interaction)) {
            # Read the file into &#39;interaction_s&#39; dataframe
            interaction_s &lt;- read.delim(var_interaction, row.names = 1)
            # Update &#39;var_interaction&#39; to the first column name of &#39;interaction_s&#39;
            var_interaction &lt;- names(interaction_s)[1]
        }

        if (is.null(var_interaction) || var_interaction == &quot;&quot;) {
            pairs_df = pairs_df %&gt;% group_by(molecular_trait_id) %&gt;% mutate(qvalue = compute_qvalues(pvalue))
        } else {
            pairs_df = pairs_df %&gt;% group_by(molecular_trait_id) %&gt;% mutate(qvalue_main = compute_qvalues(pvalue), qvalue_interaction = compute_qvalues($[&quot;pvalue_%s_interaction&quot; % var_interaction]))         
        }
    }
    pairs_df %&gt;% write_delim($[_output[&quot;nominal&quot;]:nr],&quot;\t&quot;)
  
bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[&quot;nominal&quot;]:nnn}.stderr&#39;, stdout = f&#39;{_output[&quot;nominal&quot;]:nnn}.stdout&#39;, container = container
        bgzip --compress-level 9 $[_output[&quot;nominal&quot;]:n] 
        tabix -S 1 -s 1 -b 2 -e 2 $[_output[&quot;nominal&quot;]]

done_if(not test_regional_association)

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[&quot;nominal&quot;]:nnn}.stderr&#39;, stdout = f&#39;{_output[&quot;nominal&quot;]:nnn}.stdout&#39;, container = container
        bgzip --compress-level 9 $[_output[&quot;nominal&quot;]:nnn].regional.tsv
        tabix -S 1 -s 1 -b 2 -e 2 $[_output[&quot;nominal&quot;]:nnn].regional.tsv.gz
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[cis_2]
done_if(&quot;regional&quot; not in _input.labels)
input: group_by = &quot;all&quot;
output: f&#39;{cwd}/{_input[&quot;nominal&quot;][0]:bnnnn}.cis_qtl_regional.fdr.gz&#39;,
        f&#39;{cwd}/{_input[&quot;nominal&quot;][0]:bnnnn}.cis_qtl_regional.summary.txt&#39;
input_files = [str(x) for x in _input[&quot;regional&quot;]]
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;,container = container
    library(&quot;purrr&quot;)
    library(&quot;tidyr&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;)
    library(&quot;qvalue&quot;)
    emprical_pd = tibble(map(c($[_input[&quot;regional&quot;]:r,]), ~read_delim(.x,&quot;\t&quot;)))%&gt;%unnest()
    emprical_pd[&quot;q_beta&quot;] = tryCatch(qvalue(emprical_pd$p_beta)$qvalue, error = function(e){print(&quot;Too few pvalue to calculate qvalue, fall back to BH&quot;) 
                                                                                              qvalue(emprical_pd$p_beta,pi0 = 1 )$qvalue})  

    emprical_pd[&quot;q_perm&quot;] = tryCatch(qvalue(emprical_pd$p_perm)$qvalue, error = function(e){print(&quot;Too few pvalue to calculate qvalue, fall back to BH&quot;) 
                                                                                              qvalue(emprical_pd$p_perm,pi0 = 1 )$qvalue})
    emprical_pd[&quot;fdr_beta&quot;] = p.adjust(emprical_pd$p_beta,&quot;fdr&quot;)    
    emprical_pd[&quot;fdr_perm&quot;] = p.adjust(emprical_pd$p_perm,&quot;fdr&quot;)   


    # Calculate the global nominal p-value threshold based on q_beta at FDR 0.05
    if (!all(is.na(emprical_pd$p_beta))) {
      lb &lt;- emprical_pd %&gt;% 
        filter(q_beta &lt;= 0.05) %&gt;% 
        pull(p_beta) %&gt;% 
        sort()
      
      ub &lt;- emprical_pd %&gt;% 
        filter(q_beta &gt; 0.05) %&gt;% 
        pull(p_beta) %&gt;% 
        sort()
      
      if (length(lb) &gt; 0) {
        lb_val &lt;- tail(lb, 1)
        threshold &lt;- if (length(ub) &gt; 0) (lb_val + head(ub, 1)) / 2 else lb_val
        message(sprintf(&quot;min p-value threshold @ FDR 0.05: %g&quot;, threshold))
        
        emprical_pd &lt;- emprical_pd %&gt;% 
          mutate(pval_nominal_threshold = qbeta(threshold, beta_shape1, beta_shape2))
      }
    }

    summary = tibble(&quot;fdr_perm_0.05&quot; =  sum(emprical_pd[&quot;fdr_perm&quot;] &lt; 0.05) , 
                      &quot;fdr_beta_0.05&quot; = sum(emprical_pd[&quot;fdr_beta&quot;] &lt; 0.05),
                      &quot;q_perm_0.05&quot; = sum(emprical_pd[&quot;q_perm&quot;] &lt; 0.05) ,
                      &quot;q_beta_0.05&quot; = sum(emprical_pd[&quot;q_beta&quot;] &lt; 0.05) ,
                       &quot;q_perm_0.01&quot; = sum(emprical_pd[&quot;q_perm&quot;] &lt; 0.01) ,
                      &quot;q_beta_0.01&quot; = sum(emprical_pd[&quot;q_beta&quot;] &lt; 0.01)  )
    emprical_pd%&gt;%write_delim(&quot;$[_output[0]]&quot;,&quot;\t&quot;)
    summary%&gt;%write_delim(&quot;$[_output[1]]&quot;,&quot;\t&quot;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="trans-xqtl-association-testing">
<h2>Trans-xQTL association testing<a class="headerlink" href="#trans-xqtl-association-testing" title="Link to this heading">#</a></h2>
<p>For transQTL analysis, if you output all the p-values for many genes (default setting) it is suggested to provide the largest memory and CPU threads available on a compute node. eg 250G and &gt;32 threads.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[trans_1]
input: input_files, group_by = len(input_files[0]), group_with = &quot;input_chroms&quot;
output: nominal = f&#39;{cwd:a}/{_input[0]:bnn}{&quot;&quot; if input_chroms[_index] == 0 else &quot;_%s&quot; % input_chroms[_index]}.trans_qtl.pairs.tsv.gz&#39;

parameter: batch_size = 10000
parameter: pval_threshold = 1.0
# Permutation testing is incorrect when the analysis is done by chrom
parameter: permutation = False

task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
python: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;,container =container
    import pandas as pd
    import numpy as np
    import tensorqtl
    from tensorqtl import genotypeio, trans
    from scipy.stats import chi2

    ## Define paths
    plink_prefix_path = $[_input[1]:nar]
    expression_bed = $[_input[0]:ar]
    covariates_file = &quot;$[covariate_file:a]&quot;
    window = $[window]
    ## Loading Data
    phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expression_bed)
    phenotype_id = phenotype_pos_df.index.name

    ## Analyze only the regions listed
    if $[region_list.is_file()]:
        region = pd.read_csv(&quot;$[region_list:a]&quot;, comment=&quot;#&quot;, header=None,sep=&quot;\t&quot;)
        phenotype_column = 1 if len(region.columns) == 1 else  $[region_list_phenotype_column]
        keep_region = region.iloc[:,phenotype_column-1].to_list()
        phenotype_df = phenotype_df[phenotype_df.index.isin(keep_region)]
        phenotype_pos_df = phenotype_pos_df[phenotype_pos_df.index.isin(keep_region)]

    ## use custom cis windows
    if $[customized_cis_windows.is_file()]:
        cis_list = pd.read_csv(&quot;$[customized_cis_windows:a]&quot;, comment=&quot;#&quot;, header=None, names=[&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;,phenotype_id], sep=&quot;\t&quot;)
        phenotype_pos_df_reset = phenotype_pos_df.reset_index() #move the phenotype id index to a new column of the dataframe
        phenotype_pos_df = phenotype_pos_df_reset.merge(cis_list, left_on = [&quot;chr&quot;,phenotype_id],right_on = [cis_list.columns[0],cis_list.columns[3]])#in some cases (gene expression for eQTLs) the phenotype_id may be in the cis_list file
        if len(phenotype_df.index) - phenotype_pos_df_reset[~phenotype_pos_df_reset[phenotype_id].isin(cis_list[phenotype_id])].shape[0]!= len(phenotype_pos_df.index):
            raise ValueError(&quot;cannot uniquely match all the phentoype data in the input to the customized cis windows provided&quot;)
        phenotype_pos_df = phenotype_pos_df.set_index(phenotype_id)[[&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;]] # The final phenotype_pos_df will have three columns(chr, start, end) and index is the phenotype ID
        window = 0 # In the updated tensorQTL, by default if there is a customized cis window, the actual cis window will be start - window &amp; end + window, so it is necessary to change the window parameter to 0
        ## Retaining only traits in cis_list
        if phenotype_pos_df_reset[~phenotype_pos_df_reset[phenotype_id].isin(cis_list[phenotype_id])].shape[0] != 0:
            phenotype_df = phenotype_df.loc[phenotype_df.index.isin(cis_list[phenotype_id])]

    covariates_df = pd.read_csv(covariates_file, sep=&#39;\t&#39;, index_col=0).T
    ## use custom sample list to subset the covariates data
    if $[keep_sample.is_file()]:
        sample_list = pd.read_csv(&quot;$[keep_sample:a]&quot;, comment=&quot;#&quot;, header=None, names=[&quot;sample_id&quot;], sep=&quot;\t&quot;)
        covariates_df.loc[sample_list.sample_id]
    pr = genotypeio.PlinkReader(plink_prefix_path)
    genotype_df = pr.load_genotypes()
    variant_df = pr.bim.set_index(&#39;snp&#39;)[[&#39;chrom&#39;, &#39;pos&#39;,&#39;a0&#39;,&#39;a1&#39;]]

    ## Retaining only common samples
    phenotype_df = phenotype_df[np.intersect1d(phenotype_df.columns, covariates_df.index)]
    phenotype_df = phenotype_df[np.intersect1d(phenotype_df.columns, genotype_df.columns)]
    covariates_df = covariates_df.transpose()[np.intersect1d(phenotype_df.columns, covariates_df.index)].transpose()

    ## To simplify things, there should really not be &quot;chr&quot; prefix
    phenotype_pos_df.chr = [x.replace(&quot;chr&quot;,&quot;&quot;) for x in phenotype_pos_df.chr]
    variant_df.chrom = [x.replace(&quot;chr&quot;,&quot;&quot;) for x in variant_df.chrom]

    ## Trans analysis
    trans_df = trans.map_trans(genotype_df, 
                            phenotype_df,
                            covariates_df, 
                            batch_size=$[batch_size],
                            return_sparse=True, 
                            return_r2 = True,
                            pval_threshold=$[pval_threshold], 
                            maf_threshold=$[maf_threshold])

    ## Filter out cis signal, again if customized cis windows are used, the windows is [start-win,end + win] where win = 0, else it is [start - win, start + win]
    trans_df = trans.filter_cis(trans_df, phenotype_pos_df, variant_df, window=window)   

    ## Permutation
    if $[&#39;True&#39; if permutation else &#39;False&#39;]:
        perm_df = trans.map_permutations(genotype_df, covariates_df, batch_size=$[batch_size],
                             maf_threshold=$[maf_threshold])
        trans.apply_permutations(perm_df,trans_df)

    ## Output
    trans_df.columns.values[1]  = &quot;molecular_trait_id&quot;
    trans_df.columns.values[2]  = &quot;pvalue&quot;
    trans_df.columns.values[3]  = &quot;beta&quot;
    trans_df.columns.values[4]  = &quot;se&quot;
    trans_df[&quot;n&quot;] = len(phenotype_df.columns.values)
    trans_df = variant_df.merge(trans_df, right_on=&#39;variant_id&#39;, left_index=True)
    trans_df.rename(columns={&#39;a1&#39;: &#39;a2&#39;, &#39;a0&#39;: &#39;a1&#39;}, inplace=True)
    # genomic inflation factor
    lambda_col = trans_df.groupby(&quot;molecular_trait_id&quot;).apply(lambda x: chi2.ppf(1. - np.median(x.pvalue), 1)/chi2.ppf(0.5,1))
    # Convert the Series to a DataFrame
    lambda_col = lambda_col.reset_index()
    lambda_col.columns = [&#39;molecular_trait_id&#39;, &#39;genomic_inflation_lambda&#39;]
    lambda_col.to_csv(&quot;$[_output:nnn].genomic_inflation.tsv.gz&quot;, sep=&#39;\t&#39;, index = None, compression={&#39;method&#39;: &#39;gzip&#39;, &#39;compresslevel&#39;: 9})

    # sort the table if chrom and pos is not in ascending order
    if not all(trans_df[&#39;pos&#39;].iloc[i] &lt;= trans_df[&#39;pos&#39;].iloc[i+1] for i in range(len(trans_df)-1)):
        trans_df = trans_df.sort_values(by=[&#39;chrom&#39;, &#39;pos&#39;])
 
    # print general information of trans_df
    print(&#39;Output Information:&#39;)
    print(&quot;Output Rows:&quot;, len(trans_df))
    print(&quot;Output Columns:&quot;, trans_df.columns.tolist())
    print(&quot;Output Preview:&quot;, trans_df.iloc[0:5, 0:10])
    
    # save output
    trans_df.to_csv($[_output[&quot;nominal&quot;]:nr], sep=&#39;\t&#39;, index = None)


R: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;,container =container
    if ($[&#39;FALSE&#39; if permutation and pval_threshold &lt; 1.0 else &#39;TRUE&#39;]){ 
      library(purrr)
      library(tidyr)
      library(readr)
      library(dplyr)
      library(qvalue)
      # calculate qvalue
      pairs_df = read_delim($[_output[&quot;nominal&quot;]:nr,], delim = &#39;\t&#39;)  
      pairs_df = pairs_df %&gt;% group_by(molecular_trait_id) %&gt;% mutate(qvalue = qvalue(pvalue)$qvalues)
      pairs_df %&gt;% write_delim($[_output[&quot;nominal&quot;]:nr],&quot;\t&quot;)
    } 

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[&quot;nominal&quot;]:nnn}.stderr&#39;, stdout = f&#39;{_output[&quot;nominal&quot;]:nnn}.stdout&#39;, container = container
        bgzip --compress-level 9 $[_output[&quot;nominal&quot;]:n]
        tabix -S 1 -s 1 -b 2 -e 2 $[_output[&quot;nominal&quot;]]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[trans_2]
# This workflow filters significant trans-QTL associations based on p-value or q-value cutoffs
input: group_by = &quot;all&quot;
output: f&#39;{cwd}/qvalue_{qvalue_cutoff if qvalue_cutoff else &quot;pvalue_&quot; + pvalue_cutoff}_filtered/qvalue_{qvalue_cutoff if qvalue_cutoff else &quot;pvalue_&quot; + pvalue_cutoff}_filtered_pairs.txt&#39;

task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand=&quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;, container = container
    #!/bin/bash
    set -e
    home_dir=&quot;$[cwd]&quot;
    pvalue_cutoff=&quot;$[pvalue_cutoff]&quot;
    qvalue_cutoff=&quot;$[qvalue_cutoff]&quot;
    first_file=$(ls &quot;${home_dir}&quot;/*trans_qtl.pairs.tsv.gz | head -n 1)
    
    # Determine cutoff value 
    if [ -n &quot;$pvalue_cutoff&quot; ] &amp;&amp; [ -n &quot;$qvalue_cutoff&quot; ]; then
        echo &quot;Both p-value and q-value cutoffs provided, using q-value cutoff&quot;
        cutoff_value=&quot;$qvalue_cutoff&quot;
        cutoff_col=$(zcat &quot;$first_file&quot; | awk -F &#39;\t&#39; -v col=&#39;qvalue&#39; &#39;NR==1{for (i=1; i&lt;=NF; i++) if ($i==col) {print i;exit}}&#39;)
    elif [ -n &quot;$pvalue_cutoff&quot; ]; then
        cutoff_value=&quot;$pvalue_cutoff&quot;
        cutoff_col=$(zcat &quot;$first_file&quot; | awk -F &#39;\t&#39; -v col=&#39;pvalue&#39; &#39;NR==1{for (i=1; i&lt;=NF; i++) if ($i==col) {print i;exit}}&#39;)
    elif [ -n &quot;$qvalue_cutoff&quot; ]; then
        cutoff_value=&quot;$qvalue_cutoff&quot;
        cutoff_col=$(zcat &quot;$first_file&quot; | awk -F &#39;\t&#39; -v col=&#39;qvalue&#39; &#39;NR==1{for (i=1; i&lt;=NF; i++) if ($i==col) {print i;exit}}&#39;)
    else
        echo &quot;No cutoff provided, using default q-value cutoff of 0.01&quot;
        cutoff_value=&quot;0.01&quot;
        cutoff_col=$(zcat &quot;$first_file&quot; | awk -F &#39;\t&#39; -v col=&#39;qvalue&#39; &#39;NR==1{for (i=1; i&lt;=NF; i++) if ($i==col) {print i;exit}}&#39;)
    fi
    
    echo &quot;Using column $cutoff_col with cutoff value $cutoff_value&quot;

    output_file=&quot;$[_output[0]]&quot;
    mkdir -p $(dirname &quot;$output_file&quot;)
    zcat &quot;$first_file&quot; | head -n 1 &gt; &quot;$output_file&quot;
    
    for gz_file in &quot;${home_dir}&quot;/*trans_qtl.pairs.tsv.gz; do
        echo &quot;Processing $gz_file...&quot;
        # Append significant data from each .gz file to the collective file
        zcat &quot;$gz_file&quot; | awk -v cutoff=&quot;$cutoff_value&quot; -v col=&quot;$cutoff_col&quot; &#39;NR == 1 {next} NR &gt; 1 &amp;&amp; $col &lt; cutoff&#39; &gt;&gt; &quot;$output_file&quot;
    done
    
    echo &quot;Total significant associations found: $(wc -l &lt; &quot;$output_file&quot;) (including header)&quot;
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/association_scan/TensorQTL"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../qtl_association_testing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">QTL Association Analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="../quantile_models/qr_and_twas.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Quantile regression for QTL association testing</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-cis-tensorqtl-command">i. Cis TensorQTL Command</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-trans-tensorqtl-command">ii. Trans TensorQTL Command</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#old-minimal-working-example">Old Minimal working example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-and-global-parameters">Setup and global parameters</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cis-xqtl-association-testing">cis-xQTL association testing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#trans-xqtl-association-testing">Trans-xQTL association testing</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>