

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>TensorQTL QTL association testing &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=c5ced968eda925caa686" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=c5ced968eda925caa686" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=c5ced968eda925caa686" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=c5ced968eda925caa686" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=c5ced968eda925caa686" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=c5ced968eda925caa686" />
  <script src="../../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=c5ced968eda925caa686"></script>

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/association_scan/TensorQTL/TensorQTL';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="APEX QTL association testing" href="../APEX/APEX.html" />
    <link rel="prev" title="QTL association analysis" href="../qtl_association_testing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../data_preprocessing/reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden factor analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../qtl_association_testing.html">QTL association analysis</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">Fine-mapping with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/fastqtl_to_mash.html">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc_dap.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/association_scan/TensorQTL/TensorQTL.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/association_scan/TensorQTL/TensorQTL.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/association_scan/TensorQTL/TensorQTL.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>TensorQTL QTL association testing</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">TensorQTL QTL association testing</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#global-parameter-settings">Global parameter settings</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cis-xqtl-association-testing">cis-xQTL association testing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#transqtl-association-testing">TransQTL association testing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#association-results-processing">Association results processing</a></li>
</ul>
</li>
</ul>

            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="tensorqtl-qtl-association-testing">
<h1>TensorQTL QTL association testing<a class="headerlink" href="#tensorqtl-qtl-association-testing" title="Permalink to this heading">#</a></h1>
<p>This notebook implements a workflow using <a class="reference external" href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1836-7">tensorQTL</a> to perform QTL association testing.</p>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this heading">#</a></h2>
<ul>
<li><p>List of molecular phenotype files: a list of <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> files containing the table for the molecular phenotype. It should have a companion index file in <code class="docutils literal notranslate"><span class="pre">tbi</span></code> format.</p>
<p>The header of the bed.gz is per the <a class="reference external" href="https://github.com/broadinstitute/tensorqtl">TensorQTL</a> convention:</p>
<blockquote>
<div><p>Phenotypes must be provided in BED format, with a single header line starting with # and the first four columns corresponding to: chr, start, end, phenotype_id, with the remaining columns corresponding to samples (the identifiers must match those in the genotype input). The BED file should specify the center of the cis-window (usually the TSS), with start == end-1.</p>
</div></blockquote>
</li>
<li><p>List of genotypes in PLINK binary format (<code class="docutils literal notranslate"><span class="pre">bed</span></code>/<code class="docutils literal notranslate"><span class="pre">bim</span></code>/<code class="docutils literal notranslate"><span class="pre">fam</span></code>) for each chromosome, previously processed through our genotype QC pipelines.</p></li>
<li><p>Covariate file, a file with #id + samples name as colnames and each row a covariate: fixed and known covariates as well as hidden covariates recovered from factor analysis.</p></li>
<li><p>Optionally, a list of traits (genes, regions of molecular features etc) to analyze.</p></li>
</ul>
<p>For cis-analysis:</p>
<ul class="simple">
<li><p>Optionally, a list of genomic regions associate with each molecular features to analyze. The default cis-analysis will use a window around TSS. This can be customized to take given start and end genomic coordinates.</p></li>
</ul>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading">#</a></h2>
<p>For each chromosome, several of summary statistics files are generated, including both nominal test statistics for each test, as well as region (gene) level association evidence.</p>
<p>The columns of nominal association result are as follows:</p>
<ul class="simple">
<li><p>phenotype_id: Molecular trait identifier.(gene)</p></li>
<li><p>variant_id: ID of the variant (rsid or chr:position:ref:alt)</p></li>
<li><p>tss_distance: Distance of the SNP to the gene transcription start site (TSS)</p></li>
<li><p>af: The allele frequency of this SNPs</p></li>
<li><p>ma_samples: Number of samples carrying the minor allele</p></li>
<li><p>ma_count: Total number of minor alleles across individuals</p></li>
<li><p>pval: Nominal P-value from linear regression</p></li>
<li><p>beta: Slope of the linear regression</p></li>
<li><p>se: Standard error of beta</p></li>
<li><p>chr : Variant chromosome.</p></li>
<li><p>pos : Variant chromosomal position (basepairs).</p></li>
<li><p>ref : Variant reference allele (A, C, T, or G).</p></li>
<li><p>alt : Variant alternate allele.</p></li>
</ul>
<p>The column specification of region (gene) level association evidence are as follows:</p>
<ul class="simple">
<li><p>phenotype_id - Molecular trait identifier. (gene)</p></li>
<li><p>num_var - Total number of variants tested in cis</p></li>
<li><p>beta_shape1 - First parameter value of the fitted beta distribution</p></li>
<li><p>beta_shape2 - Second parameter value of the fitted beta distribution</p></li>
<li><p>true_df - Effective degrees of freedom the beta distribution approximation</p></li>
<li><p>pval_true_df - Empirical P-value for the beta distribution approximation</p></li>
<li><p>variant_id - ID of the top variant (rsid or chr:position:ref:alt)</p></li>
<li><p>tss_distance - Distance of the SNP to the gene transcription start site (TSS)</p></li>
<li><p>ma_samples - Number of samples carrying the minor allele</p></li>
<li><p>ma_count - Total number of minor alleles across individuals</p></li>
<li><p>af - Alternative allele frequency in MiGA cohort</p></li>
<li><p>ref_factor - Flag indicating if the alternative allele is the minor allele in the cohort (1 if AF &lt;= 0.5, -1 if not)</p></li>
<li><p>pval_nominal - Nominal P-value from linear regression</p></li>
<li><p>slope - Slope of the linear regression</p></li>
<li><p>slope_se - Standard error of the slope</p></li>
<li><p>pval_perm - First permutation P-value directly obtained from the permutations with the direct method</p></li>
<li><p>pval_beta - Second permutation P-value obtained via beta approximation. This is the one to use for downstream analysis</p></li>
</ul>
</section>
</section>
<section class="tex2jax_ignore mathjax_ignore" id="command-interface">
<h1>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this heading">#</a></h1>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run TensorQTL.ipynb -h
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run TensorQTL.ipynb [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  cis
  trans

Global Workflow Options:
  --covariate-file VAL (as path, required)
                        Covariate file
  --genotype-file VAL (as path, required)
                        A genotype file in PLINK binary format (bed/bam/fam)
                        format, per chrom
  --region-list . (as path)
                        An optional subset of regions of molecular features to
                        analyze. The last column is the  gene names
  --customized-cis-windows . (as path)
                        An optional list documenting the custom cis window for
                        each region to analyze, with four column, chr, start,
                        end, region ID (eg gene ID). If this list is not
                        provided, the default `window` parameter (see below)
                        will be used.
  --cwd output (as path)
                        Path to the work directory of the analysis.
  --phenotype-file VAL (as path, required)
                        Phenotype file, a list of phenotype per chrom.
  --name  f&quot;{phenotype_file:bn}_{covariate_file:bn}&quot;

                        Prefix for the analysis output
  --MAC 0 (as int)
                        Minor allele count cutoff
  --region-name &#39;gene_id&#39;
                        The name of phenotype corresponding to gene_id or
                        gene_name in the region
  --phenotype-group . (as path)
                        The phenotype group file to group molecule_trait into
                        molecule_trait_object This is applicable to when there
                        are multiple molecular events in the same region, such
                        as sQTL analysis.
  --window 1000000 (as int)
                        Specify the cis window for the up and downstream radius
                        to analyze around the region of interest, in units of bp
                        This parameter will be set to zero if
                        `customized_cis_windows` is provided.
  --numThreads 8 (as int)
                        Number of threads
  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 12h
  --mem 16G
  --container &#39;&#39;
                        Container option for software to run the analysis:
                        docker or singularity
  --entrypoint  (&#39;micromamba run -a &quot;&quot; -n&#39; + &#39; &#39; + re.sub(r&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;, &#39;&#39;, container.split(&#39;/&#39;)[-1])) if container else &quot;&quot;

  --maf-threshold  MAC/(2.0*N)

                        Minor allele frequency cutoff. It will overwrite minor
                        allele cutoff.

Sections
  cis_1:
    Workflow Options:
      --[no-]skip-nominal-if-exist (default to False)
                        skip nominal association results if the files exists
                        already This is false by default which means to
                        recompute everything This is only relevant when the
                        `parquet` files for nominal results exist but not the
                        other files and you want to avoid computing the nominal
                        results again
  trans_1:
    Workflow Options:
      --batch-size 50000 (as int)
      --pval-threshold 1.0 (as float)
  cis_2:
  trans_2:
</pre></div>
</div>
</div>
</div>
<section id="minimal-working-example">
<h2>Minimal working example<a class="headerlink" href="#minimal-working-example" title="Permalink to this heading">#</a></h2>
<p>An MWE is uploaded to <a class="reference external" href="https://drive.google.com/drive/folders/1yjTwoO0DYGi-J9ouMsh9fHKfDmsXJ_4I?usp=sharing">google drive</a>.
The singularity image (sif) for running this MWE is uploaded to <a class="reference external" href="https://drive.google.com/drive/folders/1mLOS3AVQM8yTaWtCbO8Q3xla98Nr5bZQ">google drive</a></p>
<p>FIXME: need to update these links.</p>
<p>FIXME: Also need to update the example commands below using our new example dataset.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">cis</span> \
    <span class="o">--</span><span class="n">genotype</span><span class="o">-</span><span class="n">file</span> <span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">recipe</span> \
    <span class="o">--</span><span class="n">covariate</span><span class="o">-</span><span class="n">file</span> <span class="n">ALL</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">BiCV</span><span class="o">.</span><span class="n">cov</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">./</span><span class="n">output</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">MAC</span> <span class="mi">5</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">trans</span> \
    <span class="o">--</span><span class="n">genotype</span><span class="o">-</span><span class="n">file</span> <span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">recipe</span> \
    <span class="o">--</span><span class="n">covariate</span><span class="o">-</span><span class="n">file</span> <span class="n">ALL</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">BiCV</span><span class="o">.</span><span class="n">cov</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">./</span><span class="n">output</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">MAC</span> <span class="mi">5</span> <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="kp">name</span>  <span class="n">gene_name</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">trans</span> \
    <span class="o">--</span><span class="n">genotype</span><span class="o">-</span><span class="n">file</span> <span class="n">plink_files_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">file</span> <span class="n">MWE</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">recipe</span> \
    <span class="o">--</span><span class="n">covariate</span><span class="o">-</span><span class="n">file</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">eight_tissue_analysis</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">ALL</span><span class="o">/</span><span class="n">covariates</span><span class="o">/</span><span class="n">ALL</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">ALL</span><span class="o">.</span><span class="n">covariate</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">resid</span><span class="o">.</span><span class="n">PEER</span><span class="o">.</span><span class="n">cov</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="o">./</span><span class="n">output</span><span class="o">/</span><span class="n">trans_tensorQTL</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">/</span><span class="n">eight_tissue_analysis</span><span class="o">/</span><span class="n">reference_data</span><span class="o">/</span><span class="n">AD_genes</span><span class="o">.</span><span class="n">region_list</span> \
    <span class="o">--</span><span class="n">customized</span><span class="o">-</span><span class="n">cis</span><span class="o">-</span><span class="n">windows</span> <span class="n">TADB_enhanced_cis</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">containers</span><span class="o">/</span><span class="n">TensorQTL</span><span class="o">.</span><span class="n">sif</span> <span class="o">--</span><span class="n">MAC</span> <span class="mi">5</span> <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="kp">name</span> <span class="n">ID</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="global-parameter-settings">
<h2>Global parameter settings<a class="headerlink" href="#global-parameter-settings" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Covariate file</span>
<span class="kn">parameter:</span> <span class="n">covariate_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># A genotype file in PLINK binary format (bed/bam/fam) format, per chrom</span>
<span class="kn">parameter:</span> <span class="n">genotype_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># An optional subset of regions of molecular features to analyze. The last column is the  gene names</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># An optional list documenting the custom cis window for each region to analyze, with four column, chr, start, end, region ID (eg gene ID).</span>
<span class="c1"># If this list is not provided, the default `window` parameter (see below) will be used.</span>
<span class="kn">parameter:</span> <span class="n">customized_cis_windows</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Path to the work directory of the analysis.</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;output&#39;</span><span class="p">)</span>
<span class="c1"># Phenotype file, a list of phenotype per chrom.</span>
<span class="kn">parameter:</span> <span class="n">phenotype_file</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># Prefix for the analysis output</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">phenotype_file</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">_</span><span class="si">{</span><span class="n">covariate_file</span><span class="si">:</span><span class="s2">bn</span><span class="si">}</span><span class="s2">&quot;</span>
<span class="c1"># Minor allele count cutoff</span>
<span class="kn">parameter:</span> <span class="n">MAC</span> <span class="o">=</span> <span class="mi">0</span>
<span class="c1"># The name of phenotype corresponding to gene_id or gene_name in the region</span>
<span class="kn">parameter:</span> <span class="n">region_name</span> <span class="o">=</span> <span class="s2">&quot;gene_id&quot;</span>
<span class="c1"># The phenotype group file to group molecule_trait into molecule_trait_object</span>
<span class="c1"># This is applicable to when there are multiple molecular events in the same region, such as sQTL analysis.</span>
<span class="kn">parameter:</span> <span class="n">phenotype_group</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span> 

<span class="c1"># Specify the cis window for the up and downstream radius to analyze around the region of interest, in units of bp</span>
<span class="c1"># This parameter will be set to zero if `customized_cis_windows` is provided.</span>
<span class="kn">parameter:</span> <span class="n">window</span> <span class="o">=</span> <span class="mi">1000000</span>

<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">8</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s1">&#39;12h&#39;</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s1">&#39;16G&#39;</span>
<span class="c1"># Container option for software to run the analysis: docker or singularity</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>

<span class="c1"># Use the header of covariate file to decide the sample size</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">N</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">covariate_file</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span><span class="n">nrows</span> <span class="o">=</span> <span class="mi">1</span><span class="p">)</span><span class="o">.</span><span class="n">columns</span><span class="p">)</span> <span class="o">-</span> <span class="mi">1</span>

<span class="c1"># Minor allele frequency cutoff. It will overwrite minor allele cutoff.</span>
<span class="kn">parameter:</span> <span class="n">maf_threshold</span> <span class="o">=</span> <span class="n">MAC</span><span class="o">/</span><span class="p">(</span><span class="mf">2.0</span><span class="o">*</span><span class="n">N</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="cis-xqtl-association-testing">
<h2>cis-xQTL association testing<a class="headerlink" href="#cis-xqtl-association-testing" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[cis_1]
# parse input file lists
import pandas as pd
molecular_pheno_files = pd.read_csv(phenotype_file, sep = &quot;\t&quot;)
genotype_files = pd.read_csv(genotype_file,sep = &quot;\t&quot;)
genotype_files[&quot;#id&quot;] = genotype_files[&quot;#id&quot;].astype(str)
molecular_pheno_files[&quot;#id&quot;] = molecular_pheno_files[&quot;#id&quot;].astype(str)

input_files = molecular_pheno_files.merge(genotype_files, on = &quot;#id&quot;)
input_files = input_files.values.tolist()
input_chroms = [x[0] for x in input_files]
input_files = [x[1:] for x in input_files]

# skip nominal association results if the files exists already
# This is false by default which means to recompute everything
# This is only relevant when the `parquet` files for nominal results exist but not the other files
# and you want to avoid computing the nominal results again
parameter: skip_nominal_if_exist = False
input: input_files, group_by = len(input_files[0]), group_with = &quot;input_chroms&quot; 
output: parquet = f&#39;{cwd:a}/{name}.cis_qtl_pairs.{_input_chroms}.parquet&#39;, # This convention is necessary to match the pattern of map_norminal output
        regional = f&#39;{cwd:a}/{name}.{_input_chroms}.cis_qtl_regional.gz&#39;,
        nominal = f&#39;{cwd:a}/{name}.{_input_chroms}.cis_qtl_nominal.gz&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
python: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39; , container = container, entrypoint = entrypoint
    import pandas as pd
    import numpy as np
    import tensorqtl
    from tensorqtl import genotypeio, cis, trans
    from scipy.stats import chi2
        
    ## Define paths
    plink_prefix_path = $[_input[1]:nar]
    expression_bed = $[_input[0]:ar]
    covariates_file = &quot;$[covariate_file:a]&quot;
    window = $[window]
    ## Load Data
    phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expression_bed)
    phenotype_id = phenotype_pos_df.index.name
    ## Analyze only the regions listed
    if $[region_list.is_file()]:
        region = pd.read_csv(&quot;$[region_list:a]&quot;, comment=&quot;#&quot;, header=None)
        keep_region = region[phenotype_id].to_list()
        phenotype_df = phenotype_df[phenotype_df.index.isin(keep_region)]
        phenotype_pos_df = phenotype_pos_df[phenotype_pos_df.index.isin(keep_region)]

    ## use custom cis windows
    if $[customized_cis_windows.is_file()]:
        cis_list = pd.read_csv(&quot;$[customized_cis_windows:a]&quot;, comment=&quot;#&quot;, header=None, names=[&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;,phenotype_id], sep=&quot;\t&quot;)
        phenotype_pos_df = phenotype_pos_df.reset_index() #move the phenotype id index to a new column of the dataframe
        phenotype_pos_df = phenotype_pos_df.merge(cis_list, left_on = [&quot;chr&quot;,phenotype_id],right_on = [cis_list.columns[0],cis_list.columns[3]])#in some cases (gene expression for eQTLs) the phenotype_id may be in the cis_list file
        phenotype_pos_df = phenotype_pos_df.set_index(phenotype_id)[[&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;]] # The final phenotype_pos_df will have three columns(chr, start, end) and index is the phenotype ID
        if len(phenotype_df.index) != len(phenotype_pos_df.index):
            raise ValueError(&quot;cannot uniquely match all the phentoype data in the input to the customized cis windows provided&quot;)
        window = 0 # In the updated tensorQTL, by default if there is a customized cis window, the actual cis window will be start - window &amp; end + window, so it is necessary to change the window parameter to 0
    
    covariates_df = pd.read_csv(covariates_file, sep=&#39;\t&#39;, index_col=0).T
    pr = genotypeio.PlinkReader(plink_prefix_path)
    genotype_df = pr.load_genotypes()
    variant_df = pr.bim.set_index(&#39;snp&#39;)[[&#39;chrom&#39;, &#39;pos&#39;]]
    ## Retaining only common samples
    phenotype_df = phenotype_df[np.intersect1d(phenotype_df.columns, covariates_df.index)]
    phenotype_df = phenotype_df[np.intersect1d(phenotype_df.columns, genotype_df.columns)]
    covariates_df = covariates_df.transpose()[np.intersect1d(phenotype_df.columns, covariates_df.index)].transpose()
    
    ## To simplify things, there should really not be &quot;chr&quot; prefix
    phenotype_pos_df.chr = [x.replace(&quot;chr&quot;,&quot;&quot;) for x in phenotype_pos_df.chr]
    variant_df.chrom = [x.replace(&quot;chr&quot;,&quot;&quot;) for x in variant_df.chrom]
    ## Read phenotype group if availble
    if $[phenotype_group.is_file()]:
        group_s = pd.read_csv($[phenotype_group:r], sep=&#39;\t&#39;, header=None, index_col=0, squeeze=True)
    else:
        group_s = None

    ## cis-QTL mapping: nominal associations for all variant-phenotype pairs
    if not ($[skip_nominal_if_exist] and $[_output[0].is_file()]):
        cis.map_nominal(genotype_df, variant_df,
                phenotype_df,
                phenotype_pos_df,
                &quot;$[_output[0]:nnn]&quot;, 
                covariates_df=covariates_df, 
                window=window, 
                maf_threshold = $[maf_threshold],
                group_s=group_s)

    ## Load the parquet and save it as txt
    pairs_df = pd.read_parquet(&quot;$[_output[0]]&quot;)
    ## Adds the group columns to pairs_df, if there is group_s use group_s, else use phenotype_id
    if group_s is not None:
        pairs_df = pairs_df.merge(pd.DataFrame( {&quot;molecular_trait_object_id&quot;: group_s}),left_on = &quot;phenotype_id&quot;, right_index = True)
    else:
        pairs_df[&quot;molecular_trait_object_id&quot;] = pairs_df.phenotype_id

    # genomic inflation factor lambda
    lambda_col = pairs_df.groupby(&quot;molecular_trait_object_id&quot;).apply( lambda x:  chi2.ppf(1. - np.median(x.pval_nominal), 1)/chi2.ppf(0.5,1))
    pairs_df.columns.values[0]  = &quot;molecular_trait_id&quot;
    pairs_df.columns.values[7]  = &quot;pvalue&quot;
    pairs_df.columns.values[8]  = &quot;beta&quot;
    pairs_df.columns.values[9]  = &quot;se&quot;
    # pairs_df = pairs_df.assign(maf = lambda dataframe: dataframe[&#39;af&#39;].map(lambda af:af if af &lt; 0.5 else 1-af) )#.drop(&quot;af&quot;,axis =  1)
    pairs_df[&quot;n&quot;] = len(phenotype_df.columns.values)
    pairs_df = pairs_df.assign(
    alt = lambda dataframe: dataframe[&#39;variant_id&#39;].map(lambda variant_id:variant_id.split(&quot;_&quot;)[-1])).assign(
    ref = lambda dataframe: dataframe[&#39;variant_id&#39;].map(lambda variant_id:variant_id.split(&quot;_&quot;)[-2])).assign(
    pos = lambda dataframe: dataframe[&#39;variant_id&#39;].map(lambda variant_id:variant_id.split(&quot;_&quot;)[0].split(&quot;:&quot;)[1])).assign(
    chrom = lambda dataframe: dataframe[&#39;variant_id&#39;].map(lambda variant_id:variant_id.split(&quot;:&quot;)[0]))
    pairs_df.to_csv(&quot;$[_output[2]]&quot;, sep=&#39;\t&#39;,index = None, compression=&#39;gzip&#39;)
    cis_df = cis.map_cis(genotype_df, 
                        variant_df, 
                        phenotype_df,
                        phenotype_pos_df,
                        covariates_df=covariates_df, 
                        seed=999, 
                        window=$[window], 
                        maf_threshold = $[maf_threshold],
                        group_s=group_s)
    
    cis_df.index.name = &quot;molecular_trait_id&quot;
    ## Add groups columns for eQTL analysis
    if &quot;group_id&quot; not in cis_df.columns:
        cis_df[&quot;group_id&quot;] = cis_df.index
        cis_df[&quot;group_size&quot;] = 1
    cis_df = cis_df.rename({&quot;group_id&quot;:&quot;molecular_trait_object_id&quot;,&quot;group_size&quot;:&quot;n_traits&quot;,&quot;num_var&quot; : &quot;n_variants&quot;,&quot;variant_id&quot;:&quot;variant&quot;,&quot;pval_perm&quot;:&quot;p_perm&quot;, &quot;pval_beta&quot;:&quot;p_beta&quot; },axis = 1)
    cis_df = cis_df.assign(genomic_inflation = lambda dataframe : dataframe[&quot;molecular_trait_object_id&quot;].map(lambda molecular_trait_object_id:lambda_col[molecular_trait_object_id]))
    cis_df.to_csv(&quot;$[_output[1]]&quot;, sep=&#39;\t&#39;, compression=&#39;gzip&#39;)

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;, container = container, entrypoint = entrypoint
        for i in $[_output[0]] ; do 
        echo &quot;output_info: $i &quot;
        echo &quot;This is the file containing the immediate output of TensorQTL&#39;s map_nominal function &quot;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;` 
        done
        for i in $[_output[1]] ; do 
        echo &quot;output_info: $i &quot; 
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;` 
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;` 
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l ` 
        echo &quot;output_column:&quot; `zcat $i | grep -V &quot;##&quot; | head -1 | wc -w `
        echo &quot;output_preview:&quot;
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6,7,8,9,10  ; done
        for i in $[_output[2]] ; do 
        echo &quot;output_info: $i &quot;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;` 
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;` 
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l `
        echo &quot;output_column:&quot; `zcat $i | grep -V &quot;##&quot; | head -1 | wc -w `
        echo &quot;output_preview:&quot; 
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6,7,8,9,10 ; done
</pre></div>
</div>
</div>
</div>
</section>
<section id="transqtl-association-testing">
<h2>TransQTL association testing<a class="headerlink" href="#transqtl-association-testing" title="Permalink to this heading">#</a></h2>
<p>For transQTL analysis, it is suggested to provide the largest memory and CPU threads available on a compute node. eg 250G and &gt;32 threads.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[trans_1]

fail_if(not region_list.is_file(), msg = &quot;For trans analysis, ``region_list`` must be provided.&quot;)


# parse input file lists
import pandas as pd
molecular_pheno_files = pd.read_csv(phenotype_file, sep = &quot;\t&quot;)
genotype_files = pd.read_csv(genotype_file,sep = &quot;\t&quot;)
genotype_files[&quot;#id&quot;] = genotype_files[&quot;#id&quot;].astype(str)
molecular_pheno_files[&quot;#id&quot;] = molecular_pheno_files[&quot;#id&quot;].astype(str)

input_files = molecular_pheno_files.merge(genotype_files, on = &quot;#id&quot;)
input_files = input_files.values.tolist()
input_chroms = [x[0] for x in input_files]
input_files = [x[1:] for x in input_files]

#input: phenotype_file, genotype_file
#output: nominal = f&#39;{cwd:a}/{_input[0]:bnn}.trans_qtl_nominal.gz&#39;
input: input_files, group_by = len(input_files[0]), group_with = &quot;input_chroms&quot; 
output: nominal = f&#39;{cwd:a}/{name}.{_input_chroms}.trans_qtl_nominal.gz&#39;
parameter: batch_size = 50000
parameter: pval_threshold = 1.0
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
python: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;,container =container, entrypoint = entrypoint
    import pandas as pd
    import numpy as np
    import tensorqtl
    from tensorqtl import genotypeio, cis, trans
    ## Define paths
    plink_prefix_path = $[_input[1]:nar]
    expression_bed = $[_input[0]:ar]
    covariates_file = &quot;$[covariate_file:a]&quot;
    window = $[window]
    ## Loading Data
    phenotype_df, phenotype_pos_df = tensorqtl.read_phenotype_bed(expression_bed)
    phenotype_id = phenotype_pos_df.index.name


    ## Analyze only the regions listed
    if $[region_list.is_file()]:
        region = pd.read_csv(&quot;$[region_list:a]&quot;, comment=&quot;#&quot;, header=None)
        keep_region = region[phenotype_id].to_list()
        phenotype_df = phenotype_df[phenotype_df.index.isin(keep_region)]
        phenotype_pos_df = phenotype_pos_df[phenotype_pos_df.index.isin(keep_region)]

    ## use custom cis windows
    if $[customized_cis_windows.is_file()]:
        cis_list = pd.read_csv(&quot;$[customized_cis_windows:a]&quot;, comment=&quot;#&quot;, header=None, names=[&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;,phenotype_id], sep=&quot;\t&quot;)
        phenotype_pos_df = phenotype_pos_df.reset_index() #move the phenotype id index to a new column of the dataframe
        phenotype_pos_df = phenotype_pos_df.merge(cis_list, left_on = [&quot;chr&quot;,phenotype_id],right_on = [cis_list.columns[0],cis_list.columns[3]])#in some cases (gene expression for eQTLs) the phenotype_id may be in the cis_list file
        phenotype_pos_df = phenotype_pos_df.set_index(phenotype_id)[[&quot;chr&quot;,&quot;start&quot;,&quot;end&quot;]] # The final phenotype_pos_df will have three columns(chr, start, end) and index is the phenotype ID
        if len(phenotype_df.index) != len(phenotype_pos_df.index):
            raise ValueError(&quot;cannot uniquely match all the phentoype data in the input to the customized cis windows provided&quot;)
        window = 0 # In the updated tensorQTL, by default if there is a customized cis window, the actual cis window will be start - window &amp; end + window, so it is necessary to change the window parameter to 0

    covariates_df = pd.read_csv(covariates_file, sep=&#39;\t&#39;, index_col=0).T
    pr = genotypeio.PlinkReader(plink_prefix_path)
    genotype_df = pr.load_genotypes()
    variant_df = pr.bim.set_index(&#39;snp&#39;)[[&#39;chrom&#39;, &#39;pos&#39;]]
    ## Retaining only common samples
    phenotype_df = phenotype_df[np.intersect1d(phenotype_df.columns, covariates_df.index)]

    phenotype_df = phenotype_df[np.intersect1d(phenotype_df.columns, genotype_df.columns)]
    covariates_df = covariates_df.transpose()[np.intersect1d(phenotype_df.columns, covariates_df.index)].transpose()

    ## To simplify things, there should really not be &quot;chr&quot; prefix
    phenotype_pos_df.chr = [x.replace(&quot;chr&quot;,&quot;&quot;) for x in phenotype_pos_df.chr]
    variant_df.chrom = [x.replace(&quot;chr&quot;,&quot;&quot;) for x in variant_df.chrom]

    ## Trans analysis
    trans_df = trans.map_trans(genotype_df, 
                            phenotype_df,
                            covariates_df, 
                            batch_size=$[batch_size],
                            return_sparse=True, 
                            return_r2 = True, 
                            pval_threshold=$[pval_threshold], 
                            maf_threshold=$[maf_threshold])

    ## Filter out cis signal, again if customized cis windows are used, the windows is [start-win,end + win] where win = 0, else it is [start - win, start + win]
    trans_df = trans.filter_cis(trans_df, phenotype_pos_df, variant_df, window=window)   

    ## Permutation
    perm_df = trans.map_permutations(genotype_df, covariates_df, batch_size=$[batch_size],
                             maf_threshold=$[maf_threshold])
    perm_output = trans.apply_permutations(perm_df,trans_df)

    ## Output
    trans_df.columns.values[1]  = &quot;molecular_trait_id&quot;
    trans_df.columns.values[2]  = &quot;pvalue&quot;
    trans_df.columns.values[3]  = &quot;beta&quot;
    trans_df.columns.values[4]  = &quot;se&quot;
    #trans_df = trans_df.assign(maf = lambda dataframe: dataframe[&#39;af&#39;].map(lambda af:af if af &lt; 0.5 else 1-af) )#.drop(&quot;af&quot;,axis =  1)
    trans_df = trans_df.assign(
    chrom = lambda dataframe: dataframe[&#39;variant_id&#39;].map(lambda variant_id:variant_id.split(&quot;:&quot;)[0])).assign(
    alt = lambda dataframe: dataframe[&#39;variant_id&#39;].map(lambda variant_id:variant_id.split(&quot;_&quot;)[2])).assign(
    ref = lambda dataframe: dataframe[&#39;variant_id&#39;].map(lambda variant_id:variant_id.split(&quot;_&quot;)[1])).assign(
    pos = lambda dataframe: dataframe[&#39;variant_id&#39;].map(lambda variant_id:variant_id.split(&quot;_&quot;)[0]))
    trans_df.to_csv(&quot;$[_output]&quot;, sep=&#39;\t&#39;,index = None)

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;, container = container, entrypoint = entrypoint
        for i in $[_output] ; do 
        echo &quot;output_info: $i &quot;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;` 
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l ` 
        echo &quot;output_column:&quot; `zcat $i | grep -V &quot;##&quot; | head -1 | wc -w ` 
        echo &quot;output_preview:&quot; 
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6,7,8,9,10; done
</pre></div>
</div>
</div>
</div>
</section>
<section id="association-results-processing">
<h2>Association results processing<a class="headerlink" href="#association-results-processing" title="Permalink to this heading">#</a></h2>
<p>For both cis and trans: Generate the recipe for yml processing
For cis: Also process the consolidates emprical data.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[cis_2]
input:  group_by = &quot;all&quot;
output: f&#39;{_input[&quot;nominal&quot;][0]:nnn}.cis_qtl.meta_info&#39;,
        f&#39;{_input[&quot;nominal&quot;][0]:nnn}.cis_qtl.column_info&#39;,
        f&#39;{_input[&quot;regional&quot;][0]:nnn}.fdr.gz&#39;,
        f&#39;{_input[&quot;regional&quot;][0]:nnn}.summary.txt&#39;   
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
python: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;,container = container, entrypoint = entrypoint
    import pandas as pd

    data_temp = pd.DataFrame({
    &quot;sumstat_dir&quot; : [$[_input[&quot;regional&quot;]:r,]],
    &quot;column_info&quot; : $[_output[1]:r]
    })
    column_info_df = pd.DataFrame( pd.Series( {&quot;ID&quot;: &quot;molecular_trait_id,chromosome,position,ref,alt&quot;,
      &quot;chromosome&quot;: &quot;chrom&quot;,
      &quot;position&quot;: &quot;pos&quot;,
      &quot;variant&quot;: &quot;variant_id&quot;,
      &quot;ref&quot;: &quot;ref&quot;,
      &quot;alt&quot;: &quot;alt&quot;,
      &quot;beta&quot;: &quot;beta&quot;,
      &quot;se&quot;: &quot;se&quot;,
      &quot;pvalue&quot;: &quot;pvalue&quot;,
      &quot;TSS_D&quot;: &quot;tss_distance&quot;,
      &quot;af&quot;: &quot;af&quot;,
      &quot;n&quot; : &quot;n&quot; ,
      &quot;ma_samples&quot;: &quot;ma_samples&quot;,
      &quot;ac&quot;: &quot;ma_count&quot;,
      &quot;molecular_trait_id&quot;: &quot;molecular_trait_id&quot;, 
      &quot;molecular_trait_object_id&quot;: &quot;molecular_trait_object_id&quot;}), columns = [&quot;TensorQTL&quot;] )

    data_temp[&quot;#chr&quot;] = [x.split(&quot;.&quot;)[-3].replace(&quot;chr&quot;,&quot;&quot;) for x in  [$[_input[&quot;nominal&quot;]:r,]]]
    data_temp = data_temp[[&#39;#chr&#39;, &#39;sumstat_dir&#39;, &#39;column_info&#39;]]
    data_temp.to_csv(&quot;$[_output[0]]&quot;,index = False,sep = &quot;\t&quot; )
    column_info_df.to_csv(&quot;$[_output[1]]&quot;,index = True,sep = &quot;\t&quot; )

R: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;,container =container, entrypoint = entrypoint 
    library(&quot;purrr&quot;)
    library(&quot;tidyr&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;)
    library(&quot;qvalue&quot;)
    emprical_pd = tibble(map(c($[_input[&quot;regional&quot;]:r,]), ~read_delim(.x,&quot;\t&quot;)))%&gt;%unnest()
    emprical_pd[&quot;q_beta&quot;] = tryCatch(qvalue(emprical_pd$p_beta)$qvalue, error = function(e){print(&quot;Too few pvalue to calculate qvalue, fall back to BH&quot;) 
                                                                                              qvalue(emprical_pd$p_beta,pi0 = 1 )$qvalue})  

    emprical_pd[&quot;q_perm&quot;] = tryCatch(qvalue(emprical_pd$p_perm)$qvalue, error = function(e){print(&quot;Too few pvalue to calculate qvalue, fall back to BH&quot;) 
                                                                                              qvalue(emprical_pd$p_perm,pi0 = 1 )$qvalue})
    emprical_pd[&quot;fdr_beta&quot;] = p.adjust(emprical_pd$p_beta,&quot;fdr&quot;)    
    emprical_pd[&quot;fdr_perm&quot;] = p.adjust(emprical_pd$p_perm,&quot;fdr&quot;)    
    summary = tibble(&quot;fdr_perm_0.05&quot; =  sum(emprical_pd[&quot;fdr_perm&quot;] &lt; 0.05) , 
                      &quot;fdr_beta_0.05&quot; = sum(emprical_pd[&quot;fdr_beta&quot;] &lt; 0.05),
                      &quot;q_perm_0.05&quot; = sum(emprical_pd[&quot;q_perm&quot;] &lt; 0.05) ,
                      &quot;q_beta_0.05&quot; = sum(emprical_pd[&quot;q_beta&quot;] &lt; 0.05) ,
                       &quot;q_perm_0.01&quot; = sum(emprical_pd[&quot;q_perm&quot;] &lt; 0.01) ,
                      &quot;q_beta_0.01&quot; = sum(emprical_pd[&quot;q_beta&quot;] &lt; 0.01)  )
    emprical_pd%&gt;%write_delim(&quot;$[_output[2]]&quot;,&quot;\t&quot;)
    summary%&gt;%write_delim(&quot;$[_output[3]]&quot;,&quot;\t&quot;)
bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;, container = container, entrypoint = entrypoint
        for i in $[_output] ; do 
        echo &quot;output_info: $i &quot;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;` 
        echo &quot;output_rows:&quot; `cat $i | wc -l  | cut -f 1 -d &quot; &quot;`
        echo &quot;output_headerow:&quot; `cat $i | grep &quot;##&quot; | wc -l `
        echo &quot;output_column:&quot; `cat $i | grep -V &quot;##&quot; | head -1 | wc -w `
        echo &quot;output_preview:&quot;
        cat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6,7,8,9,10 ; done
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[trans_2]
input:  group_by = &quot;all&quot;
output: f&#39;{_input[&quot;nominal&quot;][0]:nnn}.trans_qtl.meta_info&#39;,
        f&#39;{_input[&quot;nominal&quot;][0]:nnn}.trans_qtl.column_info&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
python: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;, container = container, entrypoint = entrypoint
    import pandas as pd 
    data_temp = pd.DataFrame({
    &quot;sumstat_dir&quot; : [$[_input[&quot;nominal&quot;]:r,]],
    &quot;column_info&quot; : $[_output[1]:r]
    })
    column_info_df = pd.DataFrame( pd.Series( {&quot;ID&quot;: &quot;molecular_trait_id,chromosome,position,ref,alt&quot;,
      &quot;chromosome&quot;: &quot;chrom&quot;,
      &quot;position&quot;: &quot;pos&quot;,
      &quot;ref&quot;: &quot;ref&quot;,
      &quot;alt&quot;: &quot;alt&quot;,
      &quot;variant&quot;: &quot;variant_id&quot;,
      &quot;beta&quot;: &quot;beta&quot;,
      &quot;se&quot;: &quot;se&quot;,
      &quot;pvalue&quot;: &quot;pval&quot;,
      &quot;af&quot;: &quot;af&quot;,
      &quot;molecular_trait_id&quot;: &quot;gene_ID&quot;}), columns = [&quot;TensorQTL&quot;] )
    data_temp.to_csv(&quot;$[_output[0]]&quot;,index = False,sep = &quot;\t&quot; )
    column_info_df.to_csv(&quot;$[_output[1]]&quot;,index = True,sep = &quot;\t&quot; )

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;, container = container, entrypoint = entrypoint
        for i in $[_output] ; do 
        echo &quot;output_info: $i &quot;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;` 
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;` 
        echo &quot;output_headerow:&quot; `zcat $i | grep &quot;##&quot; | wc -l ` 
        echo &quot;output_column:&quot; `zcat $i | grep -V &quot;##&quot; | head -1 | wc -w ` 
        echo &quot;output_preview:&quot; 
        zcat $i  | grep -v &quot;##&quot; | head  | cut -f 1,2,3,4,5,6,7,8,9,10 ; done
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/association_scan/TensorQTL"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../qtl_association_testing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">QTL association analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="../APEX/APEX.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">APEX QTL association testing</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#">TensorQTL QTL association testing</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
</ul>
</li>
<li class="toc-h1 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a><ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal working example</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#global-parameter-settings">Global parameter settings</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cis-xqtl-association-testing">cis-xQTL association testing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#transqtl-association-testing">TransQTL association testing</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#association-results-processing">Association results processing</a></li>
</ul>
</li>
</ul>

  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
       Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=c5ced968eda925caa686"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=c5ced968eda925caa686"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>