

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Fine-mapping with PolyFun &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/prototype_drafts/polyfun';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Mendelian Randomization using MRAID" href="MRAID_QTL.html" />
    <link rel="prev" title="High-dimensional regression with summary statistics" href="../mnm_analysis/rss_analysis.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced cis-QTL Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/SuSiE.html">SuSiE fine-mapping workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/fSuSiE.html">fSuSiE workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mvSuSiE.html">Multivariate fine-mapping workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mrmash.html">Multivariate TWAS prediction model</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced Genome-wide Analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">Fine-mapping with PolyFun</a></li>
<li class="toctree-l1"><a class="reference internal" href="MRAID_QTL.html">Mendelian Randomization using MRAID</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/ldsc.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL-GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../pecotmr_integration/pecotmr.html">The PECOTMR intergration framework</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">GWAS integration: enrichment and colocalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/twas_mr.html">GWAS integration: TWAS and MR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/intact.html">TWAS and COLOC integration</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">Extract genome-wide data for multivariate analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">MASH analysis pipeline with posterior computation</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Rare xQTL</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../rare_xqtl/watershed.html">watershed analysis for rare xQTL</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Post Processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../post_processing/sumstats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/rds_to_vcf.html">Summary statistics in VCF format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/association_scan_post_processing.html">Association result processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">SuSIE results post process</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/edit/gh-pages/code/prototype_drafts/polyfun.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/prototype_drafts/polyfun.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/prototype_drafts/polyfun.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Fine-mapping with PolyFun</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#aim">Aim</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods-overview">Methods Overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow">Workflow</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-obtain-functional-annotations">Step 1: Obtain functional annotations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-compute-ld-scores-for-annotations">Step 2: Compute LD-scores for annotations</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-1-compute-with-reference-panel-of-sequenced-individuals">Method 1: Compute with reference panel of sequenced individuals</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-2-compute-with-pre-computed-uk-biobank-ld-matrices">Method 2: Compute with pre-computed UK Biobank LD matrices</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-3-compute-with-own-pre-computed-ld-matrices">Method 3: Compute with own pre-computed LD matrices</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-compute-prior-causal-probabilities">Step 3: Compute Prior Causal Probabilities</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-1-use-precomputed-prior-causal-probabilities">Method 1: Use precomputed prior causal probabilities</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-2-compute-via-l2-regularized-extension-of-s-ldsc-preferred">Method 2: Compute via L2-regularized extension of S-LDSC (preferred)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-3-compute-non-parametrically">Method 3: Compute Non-parametrically</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-functionally-informed-fine-mapping-with-finemapper">Step 4: Functionally informed fine mapping with finemapper</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal Working Example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-1-functionally-informed-fine-mapping-using-summary-statistics-file-with-precomputed-prior-causal-probabilities">Example 1: Functionally-informed fine-mapping using summary statistics file with precomputed prior causal probabilities</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-2-functionally-informed-fine-mapping-using-summary-statistics-file-generated-from-pre-obtained-annotation-and-ld-score-files">Example 2: Functionally-informed fine-mapping using summary statistics file generated from pre-obtained annotation and LD-score files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary">Summary</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#example-1">Example 1:</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#example-2">Example 2:</a></li>
</ul>
</li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="fine-mapping-with-polyfun">
<h1>Fine-mapping with PolyFun<a class="headerlink" href="#fine-mapping-with-polyfun" title="Permalink to this heading">#</a></h1>
<section id="aim">
<h2>Aim<a class="headerlink" href="#aim" title="Permalink to this heading">#</a></h2>
<p>The purpose of this notebook ipmlements commands for <a class="reference external" href="https://github.com/omerwe/polyfun/wiki">a functionally-informed fine-mapping workflow using the PolyFun method</a>.</p>
</section>
<section id="methods-overview">
<h2>Methods Overview<a class="headerlink" href="#methods-overview" title="Permalink to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">PolyFun</span></code> offers the following features:</p>
<ol class="arabic simple">
<li><p>Using and/or creating Functional Annotations</p></li>
<li><p>Estimating Functional Enrichment Using <code class="docutils literal notranslate"><span class="pre">S-LDSC</span></code></p></li>
<li><p>Using and/or computating Prior Causal Probabilities from 1</p></li>
<li><p>Functionally Informed Fine Mapping with Finemapper</p></li>
<li><p>Polygenic Localization with <code class="docutils literal notranslate"><span class="pre">PolyLoc</span></code></p></li>
</ol>
<p><strong>Notice: this workflow does not implements 5 the <code class="docutils literal notranslate"><span class="pre">PolyLoc</span></code>.</strong></p>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this heading">#</a></h2>
<ol class="arabic">
<li><p>GWAS summary statistics including the following variables:</p>
<ul class="simple">
<li><p>variant_id - variant ID</p></li>
<li><p>P - p-value</p></li>
<li><p>CHR - chromosome number</p></li>
<li><p>BP - base pair position</p></li>
<li><p>A1 - The effect allele (i.e., the sign of the effect size is with respect to A1)</p></li>
<li><p>A2 - the second allele</p></li>
<li><p>MAF - minor allele frequency</p></li>
<li><p>BETA - effect size</p></li>
<li><p>SE - effect size standard error</p></li>
</ul>
</li>
<li><p>Functional annotation files including the following columns:</p>
<ul class="simple">
<li><p>CHR - chromosome number</p></li>
<li><p>BP base pair position (in hg19 coordinates)</p></li>
<li><p>SNP - dbSNP reference number</p></li>
<li><p>A1 - The effect allele</p></li>
<li><p>A2 - the second allele</p></li>
<li><p>Arbitrary additional columns representing annotations</p></li>
</ul>
</li>
<li><p>A <code class="docutils literal notranslate"><span class="pre">.l2.M</span></code> white-space delimited file containing a single line with the sums of the columns of each annotation</p></li>
<li><p>LD-score files</p>
<ul class="simple">
<li><p>Strongly recommended that LD-score files include A1,A2 columns</p></li>
</ul>
</li>
<li><p>LD information, taken from one of three possible data sources:</p>
<ul class="simple">
<li><p>plink files with genotypes from a reference panel</p></li>
<li><p>bgen file with genotypes from a reference panel</p></li>
<li><p>pre-computed LD matrix</p></li>
</ul>
<p>Optional if (4) is obtained and no plans to compute prior causal probabilities non-parametrically</p>
</li>
<li><p>Ld-score weights files.</p>
<ul class="simple">
<li><p>Strongly recommended that weight files include A1,A2 columns</p></li>
</ul>
</li>
</ol>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading">#</a></h2>
<p>A <code class="docutils literal notranslate"><span class="pre">.gz</span></code> file containing input summary statistics columns and additionally the following columns:</p>
<ul class="simple">
<li><p>PIP - posterior causal probability</p></li>
<li><p>BETA_MEAN - posterior mean of causal effect size (in standardized genotype scale)</p></li>
<li><p>BETA_SD - posterior standard deviation of causal effect size (in standardized genotype scale)</p></li>
<li><p>CREDIBLE_SET - the index of the first (typically smallest) credible set that the SNP belongs to (0 means none).</p></li>
</ul>
</section>
<section id="workflow">
<h2>Workflow<a class="headerlink" href="#workflow" title="Permalink to this heading">#</a></h2>
<p>Step 1 and 2 are optional if using pre-computed prior causal probabilities</p>
<section id="step-1-obtain-functional-annotations">
<h3>Step 1: Obtain functional annotations<a class="headerlink" href="#step-1-obtain-functional-annotations" title="Permalink to this heading">#</a></h3>
<p>For each chromosome, the following files need to be obtained:</p>
<ol class="arabic simple">
<li><p>A <code class="docutils literal notranslate"><span class="pre">.gz</span></code> or <code class="docutils literal notranslate"><span class="pre">.parquet</span></code> annotations file containing the following columns:</p></li>
</ol>
<ul class="simple">
<li><p>CHR - chromosome number</p></li>
<li><p>BP base pair position</p></li>
<li><p>SNP - dbSNP reference number</p></li>
<li><p>A1 - The effect allele</p></li>
<li><p>A2 - the second allele</p></li>
<li><p>Arbitrary additional columns representing annotations</p></li>
</ul>
<ol class="arabic simple" start="2">
<li><p>A <code class="docutils literal notranslate"><span class="pre">.l2.M</span></code> white-space delimited file containing a single line with the sums of the columns of each annotation</p></li>
<li><p>(Optional) A <code class="docutils literal notranslate"><span class="pre">l2.M_5_50</span></code> file that is the <code class="docutils literal notranslate"><span class="pre">.l2.M</span></code> file but only containing common SNPS (MAF between 5% and 50%)</p></li>
</ol>
<p>The above files can be obtained either by using existing function annotation files, or by creating your own through other software such as <code class="docutils literal notranslate"><span class="pre">TORUS</span></code>.</p>
<p>Existing function annotation files example: functional annotations for ~19 million UK Biobank imputed SNPs with MAF&gt;0.1%, based on the baseline-LF 2.2.UKB annotations.</p>
<p>Download (30G): <a class="reference external" href="https://data.broadinstitute.org/alkesgroup/LDSCORE/baselineLF_v2.2.UKB.polyfun.tar.gz">https://data.broadinstitute.org/alkesgroup/LDSCORE/baselineLF_v2.2.UKB.polyfun.tar.gz</a></p>
</section>
<section id="step-2-compute-ld-scores-for-annotations">
<h3>Step 2: Compute LD-scores for annotations<a class="headerlink" href="#step-2-compute-ld-scores-for-annotations" title="Permalink to this heading">#</a></h3>
<p>Precomputed LD-score files can be used. LD-score files can also be generated through the methods below:</p>
<section id="method-1-compute-with-reference-panel-of-sequenced-individuals">
<h4>Method 1: Compute with reference panel of sequenced individuals<a class="headerlink" href="#method-1-compute-with-reference-panel-of-sequenced-individuals" title="Permalink to this heading">#</a></h4>
<p>Reference panel should have at least 3000 sequenced individuals from target population.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;/mnt/mfs/statgen/containers/xqtl_pipeline_sif/polyfun.sif&quot;</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span><span class="p">{(</span><span class="s1">&#39;micromamba run -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">][:</span><span class="o">-</span><span class="mi">4</span><span class="p">])</span> <span class="k">if</span> <span class="n">container</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="s1">&#39;.sif&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="sa">f</span><span class="s1">&#39;&#39;</span><span class="p">}</span>
<span class="kn">parameter:</span> <span class="n">wd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">exe_dir</span> <span class="o">=</span> <span class="s2">&quot;/usr/local/bin/&quot;</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;demo&quot;</span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">annot_file</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">sumstats</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;./&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[ld_score]
input: annot_file, genoFile
output: f&#39;{wd:a}/{name}.ref.ldscore.parquet&#39;
task: trunk_workers = 1, trunk_size = 1, walltime = &#39;24h&#39;,  mem = &#39;30G&#39;, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39; , container = container, entrypoint = entrypoint
    python $[exe_dir]/compute_ldscores.py \
        --bfile $[_input[1]:n] \
        --annot $[_input[0]] \
        --out $[_output]
</pre></div>
</div>
</div>
</div>
</section>
<section id="method-2-compute-with-pre-computed-uk-biobank-ld-matrices">
<h4>Method 2: Compute with pre-computed UK Biobank LD matrices<a class="headerlink" href="#method-2-compute-with-pre-computed-uk-biobank-ld-matrices" title="Permalink to this heading">#</a></h4>
<p>Matrices download: <a class="reference external" href="https://data.broadinstitute.org/alkesgroup/UKBB_LD">https://data.broadinstitute.org/alkesgroup/UKBB_LD</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[ld_score_ukb]
input: annot_file
output: f&#39;{wd:a}/{name}.ukb.ldscore.parquet&#39;
task: trunk_workers = 1, trunk_size = 1, walltime = &#39;24h&#39;,  mem = &#39;30G&#39;, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39; , container = container, entrypoint = entrypoint
    python $[exe_dir]/compute_ldscores_from_ld.py \
        --annot $[_input[0]] \
        --ukb \
        --out $[_output]
</pre></div>
</div>
</div>
</div>
</section>
<section id="method-3-compute-with-own-pre-computed-ld-matrices">
<h4>Method 3: Compute with own pre-computed LD matrices<a class="headerlink" href="#method-3-compute-with-own-pre-computed-ld-matrices" title="Permalink to this heading">#</a></h4>
<p>Own pre-computed LD matrices should be in <code class="docutils literal notranslate"><span class="pre">.bcor</span></code> format.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[ld_score_own]
parameter: sample_size = int
parameter: bcor_files = paths
input: annot_file,bcor_files
output: f&#39;{wd:a}/{name}.original.ldscore.parquet&#39;
task: trunk_workers = 1, trunk_size = 1, walltime = &#39;24h&#39;,  mem = &#39;30G&#39;, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39; , container = container, entrypoint = entrypoint
    python $[exe_dir]/compute_ldscores_from_ld.py $[_input[1]] \
        --annot $[_input[0]] \
        --out $[_output] \
        --n $[sample_size] \
        
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-3-compute-prior-causal-probabilities">
<h3>Step 3: Compute Prior Causal Probabilities<a class="headerlink" href="#step-3-compute-prior-causal-probabilities" title="Permalink to this heading">#</a></h3>
<section id="method-1-use-precomputed-prior-causal-probabilities">
<h4>Method 1: Use precomputed prior causal probabilities<a class="headerlink" href="#method-1-use-precomputed-prior-causal-probabilities" title="Permalink to this heading">#</a></h4>
<p>Use precomputed prior causal probabilities of 19 million imputed UK Biobank SNPs with MAF&gt;0.1%, based on a meta-analysis of 15 UK Biobank traits.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[prior_causal_prob]
input: sumstats
output: f&#39;{wd:a}/{name}.pcp.gz&#39;
task: trunk_workers = 1, trunk_size = 1, walltime = &#39;24h&#39;,  mem = &#39;30G&#39;, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, container = container, entrypoint = entrypoint
    python $[exe_dir]/extract_snpvar.py \
        --sumstats $[_input] \
        --out $[_output] \
        --allow-missing
</pre></div>
</div>
</div>
</div>
</section>
<section id="method-2-compute-via-l2-regularized-extension-of-s-ldsc-preferred">
<h4>Method 2: Compute via L2-regularized extension of S-LDSC (preferred)<a class="headerlink" href="#method-2-compute-via-l2-regularized-extension-of-s-ldsc-preferred" title="Permalink to this heading">#</a></h4>
<p>Compute via an L2-regularized extension of stratified LD-score regression (S-LDSC). Use the annotation and LD-score files produced in Step1.</p>
<ol class="arabic simple">
<li><p>Create a munged summary statistics file in a PolyFun-friendly parquet format.</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[munged_sumstats]
parameter: sample_size = 472868 
parameter: min_info = 0.6
parameter: min_maf = 0.01
input: sumstats
output: f&#39;{wd:a}/{name}.sumstats_munged.parquet&#39;
task: trunk_workers = 1, trunk_size = 1, walltime = &#39;24h&#39;,  mem = &#39;127G&#39;, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39; , container = container, entrypoint = entrypoint
    python $[exe_dir]/munge_polyfun_sumstats.py \
      --sumstats $[_input] \
      --n $[sample_size] \
      --out $[_output] \
      --min-info $[min_info] \
      --min-maf $[min_maf]
</pre></div>
</div>
</div>
</div>
<ol class="arabic simple" start="2">
<li><p>Run PolyFun with L2-regularized S-LDSC</p></li>
</ol>
<ul class="simple">
<li><p>Require at least 45 GB of mem</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[L2_SLDSC]
# a ld score file with surfix l2.ldscore.parquet
parameter: ref_ld = path
# another ld score file with surfix l2.ldscore.parquet, different from ref_ld
parameter: ref_wgt = ref_ld
parameter: partitions = &quot;&quot;
input: ref_ld, ref_wgt,output_from(&quot;munged_sumstats&quot;)
# parameter: sumstat = _input[2]
output: f&#39;{wd:a}/{name}.ldsldsc.parquent&#39;
task: trunk_workers = 1, trunk_size = 1, walltime = &#39;24h&#39;,  mem = &#39;127G&#39;, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39; , container = container, entrypoint = entrypoint
    python $[exe_dir]/polyfun.py \
        --compute-h2-L2 \
        --output-prefix $[_output] \
        --sumstats $[_input[2]] \
        --ref-ld-chr $[_input[0]:nnnn].\
        --w-ld-chr $[_input[1]:nnnn]. \
        --allow-missing $[&quot;&quot; if partitions else &quot;--no-partitions&quot;]
</pre></div>
</div>
</div>
</div>
</section>
<section id="method-3-compute-non-parametrically">
<h4>Method 3: Compute Non-parametrically<a class="headerlink" href="#method-3-compute-non-parametrically" title="Permalink to this heading">#</a></h4>
<ol class="arabic simple">
<li><p>Create a munged summary statistics file in a PolyFun-friendly parquet format.
Duplicated cells are commented out, the input of [ld_snpbin] is the output from [L2_regu_SLDSC]</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1">#[munged_sumstats2]</span>
<span class="c1">#parameter: sumstats = AD_sumstats_Jansenetal_2019sept.txt.gz</span>
<span class="c1">#parameter: sample_size = int</span>
<span class="c1">#parameter: container = none</span>
<span class="c1">#bash: container = container </span>
<span class="c1">#    mkdir -p SLDSC_output</span>
<span class="c1">#    python munge_polyfun_sumstats.py \</span>
<span class="c1">#      --sumstats sumstats \</span>
<span class="c1">#      --n sample_size \</span>
<span class="c1">#      --out /SLDSC_output/sumstats_munged.parquet \</span>
<span class="c1">#      --min-info 0 \</span>
<span class="c1">#      --min-maf 0</span>
</pre></div>
</div>
</div>
</div>
<ol class="arabic simple" start="2">
<li><p>Run PolyFun with L2-regularized S-LDSC</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># [L2_regu_SLDSC2]</span>
<span class="c1"># </span>
<span class="c1"># parameter: container = none</span>
<span class="c1"># paramter: ref_ld = example_data/annotations.</span>
<span class="c1"># parameter: ref_wgt = example_data/weights.</span>
<span class="c1"># bash: container=container</span>
<span class="c1">#     python polyfun.py \</span>
<span class="c1">#     --compute-h2-L2 \</span>
<span class="c1">#     --output-prefix output/testrun \</span>
<span class="c1">#     --sumstats example_data/sumstats.parquet \</span>
<span class="c1">#     --ref-ld-chr ref_ld \</span>
<span class="c1">#     --w-ld-chr ref_wgt</span>
</pre></div>
</div>
</div>
</div>
<ol class="arabic simple" start="3">
<li><p>Compute LD-scores for each SNP bin</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[ld_snpbin]
depends: sos_step(&quot;L2_regu_SLDSC&quot;)
parameter: chrom = int
input: annot_file, genoFile
output: f&#39;{wd:a}/{name}.snpbin.ldscore.parquet&#39;
task: trunk_workers = 1, trunk_size = 1, walltime = &#39;24h&#39;,  mem = &#39;30G&#39;, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39; , container = container, entrypoint = entrypoint
     python $[exe_dir]/polyfun.py \
        --compute-ldscores \
        --bfile-chr $[_input[1]:n] \
        --output-prefix $[_output] \
        --chr $[chrom]
</pre></div>
</div>
</div>
</div>
<ol class="arabic simple" start="4">
<li><p>Re-estimate per-SNP heritabilities via S-LDSC</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>#[re_SLDSC]
#bash:
#    python polyfun.py \
#    --compute-h2-bins \
#    --output-prefix output/testrun \
#    --sumstats example_data/sumstats.parquet \
#    --w-ld-chr example_data/weights.

[L2_SLDSC_bins]
paramter: ref_ld = path
parameter: ref_wgt = ref_ld
parameter: partitions = &quot;&quot;
input: ref_ld, ref_wgt,output_from(&quot;munged_sumstats&quot;)
output: f&#39;{wd:a}/{name}.txt.gz&#39;
task: trunk_workers = 1, trunk_size = 1, walltime = &#39;24h&#39;,  mem = &#39;30G&#39;, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39; , container = container, entrypoint = entrypoint
    python $[exe_dir]/polyfun.py \
        --compute-h2-bins \
        --output-prefix $[_output] \
        --sumstats $[_input[2]] \
        --ref-ld-chr $[_input[0]]\
        --w-ld-chr $[_input[1]] 
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="step-4-functionally-informed-fine-mapping-with-finemapper">
<h3>Step 4: Functionally informed fine mapping with finemapper<a class="headerlink" href="#step-4-functionally-informed-fine-mapping-with-finemapper" title="Permalink to this heading">#</a></h3>
<p>Input summary statistics file must have <code class="docutils literal notranslate"><span class="pre">SNPVAR</span></code> column (per-SNP heritability) to perform functionally-informed fine-mapping. To fine-map without annotations, use additional parameter <code class="docutils literal notranslate"><span class="pre">--non-funct</span></code>. The summary statistical file then will not require the <code class="docutils literal notranslate"><span class="pre">SNPVAR</span></code> column.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[fine_mapping]
parameter: sample_size = 383290
parameter: chrom = 1
parameter: start = 46000001
parameter: end = 49000001
#parameter: output_path = &quot;output/finemap.1.46000001.49000001.gz&quot;
parameter: max_num_causal = 5
input: genoFile,sumstats
output: f&#39;{wd:a}/output/finemap.{chrom}.{start}.{end}.gz&#39;
task: trunk_workers = 1, trunk_size = 1, walltime = &#39;24h&#39;,  mem = &#39;30G&#39;, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: expand = &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39; , container = container, entrypoint = entrypoint
    python $[exe_dir]/finemapper.py \
        --geno $[_input[0]:n] \
        --sumstats $[_input[1]]  \
        --n $[sample_size] \
        --chr $[chrom] \
        --start $[start] \
        --end $[end] \
        --method susie \
        --max-num-causal $[max_num_causal] \
        --cache-dir $[_output:d]/cache \
        --out $[_output]
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="minimal-working-example">
<h2>Minimal Working Example<a class="headerlink" href="#minimal-working-example" title="Permalink to this heading">#</a></h2>
<section id="example-1-functionally-informed-fine-mapping-using-summary-statistics-file-with-precomputed-prior-causal-probabilities">
<h3>Example 1: Functionally-informed fine-mapping using summary statistics file with precomputed prior causal probabilities<a class="headerlink" href="#example-1-functionally-informed-fine-mapping-using-summary-statistics-file-with-precomputed-prior-causal-probabilities" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">protocol</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">integrative_analysis</span><span class="o">/</span><span class="n">SuSiE_Ann</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">prior_causal_prob</span> \
    <span class="o">--</span><span class="n">sumstats</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">at3535</span><span class="o">/</span><span class="n">polyfun</span><span class="o">/</span><span class="n">AD_sumstats_Jansenetal_2019sept</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span>  \
    <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> \
    <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">ADSPFG</span><span class="o">-</span><span class="n">xQTL</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">protocol</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">integrative_analysis</span><span class="o">/</span><span class="n">SuSiE_Ann</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fine_mapping</span> \
    <span class="o">--</span><span class="n">sumstats</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">at3535</span><span class="o">/</span><span class="n">polyfun</span><span class="o">/</span><span class="n">AD_sumstats_Jansenetal_2019sept</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span>  \
    <span class="o">--</span><span class="n">genoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">ROSMAP_xqtl</span><span class="o">/</span><span class="n">dataset</span><span class="o">/</span><span class="n">snvCombinedPlink</span><span class="o">/</span><span class="n">chr1</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> \
    <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">ADSPFG</span><span class="o">-</span><span class="n">xQTL</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="example-2-functionally-informed-fine-mapping-using-summary-statistics-file-generated-from-pre-obtained-annotation-and-ld-score-files">
<h3>Example 2: Functionally-informed fine-mapping using summary statistics file generated from pre-obtained annotation and LD-score files<a class="headerlink" href="#example-2-functionally-informed-fine-mapping-using-summary-statistics-file-generated-from-pre-obtained-annotation-and-ld-score-files" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">protocol</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">integrative_analysis</span><span class="o">/</span><span class="n">SuSiE_Ann</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">munged_sumstats</span>  \
<span class="o">--</span><span class="n">sumstats</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">at3535</span><span class="o">/</span><span class="n">polyfun</span><span class="o">/</span><span class="n">GCST90012877_buildGRCh37_colrenamed</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span> \
<span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">ADSPFG</span><span class="o">-</span><span class="n">xQTL</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">protocol</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">integrative_analysis</span><span class="o">/</span><span class="n">SuSiE_Ann</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">L2_SLDSC</span>  \
<span class="o">--</span><span class="n">sumstats</span> <span class="n">demo</span><span class="o">.</span><span class="n">sumstats_munged</span><span class="o">.</span><span class="n">parquet</span> \
<span class="o">--</span><span class="n">ref_ld</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">tl3030</span><span class="o">/</span><span class="n">baselineLF2</span><span class="mf">.2</span><span class="o">.</span><span class="n">UKB</span><span class="o">/</span><span class="n">baselineLF2</span><span class="mf">.2</span><span class="o">.</span><span class="n">UKB</span><span class="mf">.1</span><span class="o">.</span><span class="n">l2</span><span class="o">.</span><span class="n">ldscore</span><span class="o">.</span><span class="n">parquet</span> \
<span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">ADSPFG</span><span class="o">-</span><span class="n">xQTL</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">nohup</span> <span class="n">sos</span> <span class="n">run</span> <span class="o">~/</span><span class="n">GIT</span><span class="o">/</span><span class="n">xqtl</span><span class="o">-</span><span class="n">protocol</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">integrative_analysis</span><span class="o">/</span><span class="n">SuSiE_Ann</span><span class="o">/</span><span class="n">polyfun</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fine_mapping</span> \
    <span class="o">--</span><span class="n">sumstats</span> <span class="n">demo</span><span class="o">.</span><span class="n">ldsldsc</span><span class="o">.</span><span class="n">parquent</span>  \
    <span class="o">--</span><span class="n">genoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">ROSMAP_xqtl</span><span class="o">/</span><span class="n">dataset</span><span class="o">/</span><span class="n">snvCombinedPlink</span><span class="o">/</span><span class="n">chr1</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">-</span><span class="n">J</span> <span class="mi">200</span> <span class="o">-</span><span class="n">q</span> <span class="n">csg</span> <span class="o">-</span><span class="n">c</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">hs3163</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">ADSPFG</span><span class="o">-</span><span class="n">xQTL</span><span class="o">/</span><span class="n">code</span><span class="o">/</span><span class="n">csg</span><span class="o">.</span><span class="n">yml</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="summary">
<h3>Summary<a class="headerlink" href="#summary" title="Permalink to this heading">#</a></h3>
<section id="example-1">
<h4>Example 1:<a class="headerlink" href="#example-1" title="Permalink to this heading">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">bash:</span>
    <span class="n">gzcat</span> <span class="n">output</span><span class="o">/</span><span class="n">finemap</span><span class="mf">.1.46000001.49000001</span><span class="o">.</span><span class="n">gz</span> <span class="o">|</span> <span class="n">head</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>CHR	SNP	BP	A1	A2	SNPVAR	Z	N	P	PIP	BETA_MEAN	BETA_SD	DISTANCE_FROM_CENTER	CREDIBLE_SET
1	rs2088102	46032974	T	C	1.70060e-06	1.25500e+01	383290	3.97510e-36	1.00000e+00	-2.03917e-02	1.61901e-03	1456799	1
1	rs7528714	47966058	G	A	1.18040e-06	5.14320e+00	383290	2.70098e-07	9.97870e-01	7.42146e-03	1.62305e-03	476285	2
1	rs7528075	47870271	G	A	1.18040e-06	4.40160e+00	383290	1.07456e-05	9.76545e-01	-5.98945e-03	1.81667e-03	380498	3
1	rs212968	48734666	G	A	1.70060e-06	-3.01130e+00	383290	2.60132e-03	3.75823e-01	-1.56305e-03	2.23942e-03	1244893	0
1	rs2622911	47837404	C	A	1.70060e-06	3.12520e+00	383290	1.77684e-03	3.71804e-01	1.54312e-03	2.22812e-03	347631	0
1	rs4511165	48293181	G	A	1.70060e-06	-1.18940e+00	383290	2.34282e-01	5.75970e-02	1.60630e-04	7.52226e-04	803408	0
1	rs3766196	47284526	C	A	6.93040e-06	-5.92360e-02	383290	9.52764e-01	4.89776e-02	-5.06776e-07	3.48039e-04	205247	0
1	rs12567716	48197570	T	C	1.18040e-06	2.14810e+00	383290	3.17058e-02	4.45128e-02	-1.28281e-04	6.81457e-04	707797	0
1	rs4927234	48384796	G	A	1.70060e-06	-7.59600e-01	383290	4.47494e-01	3.41123e-02	7.74978e-05	5.04755e-04	895023	0
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;output/finemap.1.46000001.49000001.gz&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
    
<span class="n">num_var_cs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;CREDIBLE_SET&#39;</span><span class="p">])</span>
<span class="n">total_cs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">CREDIBLE_SET</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span><span class="o">-</span> <span class="mi">1</span>
<span class="n">avg_var_cs</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">num_var_cs</span><span class="p">)</span> <span class="o">/</span> <span class="n">total_cs</span>
<span class="n">pip50</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;PIP&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">pip95</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;PIP&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span><span class="mf">0.95</span><span class="p">)</span>

<span class="n">result</span> <span class="o">=</span> <span class="s2">&quot;Number of variants with PIP &gt; 0.5: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pip50</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;Number of variants with PIP &gt; 0.95: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pip95</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Number of variants that have credible sets: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num_var_cs</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Number of unique credible sets: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">total_cs</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Average number of variants per credible set: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">avg_var_cs</span><span class="p">)</span> 


<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;results.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">the_file</span><span class="p">:</span>
    <span class="n">the_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;results.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">contents</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">contents</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;Number of variants with PIP &gt; 0.5: 3\n&#39;, &#39;Number of variants with PIP &gt; 0.95: 3\n&#39;, &#39;Number of variants that have credible sets: 3\n&#39;, &#39;Number of unique credible sets: 3\n&#39;, &#39;Average number of variants per credible set: 1.0&#39;]
</pre></div>
</div>
</div>
</div>
</section>
<section id="example-2">
<h4>Example 2:<a class="headerlink" href="#example-2" title="Permalink to this heading">#</a></h4>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">bash:</span>
    <span class="n">gzcat</span> <span class="n">output</span><span class="o">/</span><span class="n">finemap</span><span class="mf">.1.460000010.49000000</span><span class="o">.</span><span class="n">gz</span> <span class="o">|</span> <span class="n">head</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="n">data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="s1">&#39;output/finemap.1.46000000.49000000.gz&#39;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="n">data</span><span class="o">.</span><span class="n">head</span><span class="p">(</span><span class="mi">5</span><span class="p">)</span>
    
<span class="n">num_var_cs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">count_nonzero</span><span class="p">(</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;CREDIBLE_SET&#39;</span><span class="p">])</span>
<span class="n">total_cs</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">data</span><span class="o">.</span><span class="n">CREDIBLE_SET</span><span class="o">.</span><span class="n">unique</span><span class="p">())</span><span class="o">-</span> <span class="mi">1</span>
<span class="n">avg_var_cs</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">num_var_cs</span><span class="p">)</span> <span class="o">/</span> <span class="n">total_cs</span>
<span class="n">pip50</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;PIP&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span><span class="mf">0.5</span><span class="p">)</span>
<span class="n">pip95</span> <span class="o">=</span> <span class="nb">sum</span><span class="p">(</span><span class="mi">1</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">data</span><span class="p">[</span><span class="s1">&#39;PIP&#39;</span><span class="p">]</span> <span class="k">if</span> <span class="n">i</span> <span class="o">&gt;</span><span class="mf">0.95</span><span class="p">)</span>

<span class="n">result</span> <span class="o">=</span> <span class="s2">&quot;Number of variants with PIP &gt; 0.5: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pip50</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;Number of variants with PIP &gt; 0.95: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pip95</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Number of variants that have credible sets: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">num_var_cs</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Number of unique credible sets: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">total_cs</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\n</span><span class="s2">&quot;</span> \
    <span class="o">+</span> <span class="s2">&quot;Average number of variants per credible set: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">avg_var_cs</span><span class="p">)</span> 


<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;results.txt&#39;</span><span class="p">,</span> <span class="s1">&#39;a&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">the_file</span><span class="p">:</span>
    <span class="n">the_file</span><span class="o">.</span><span class="n">write</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

<span class="k">with</span> <span class="nb">open</span><span class="p">(</span><span class="s1">&#39;results.txt&#39;</span><span class="p">)</span> <span class="k">as</span> <span class="n">f</span><span class="p">:</span>
    <span class="n">contents</span> <span class="o">=</span> <span class="n">f</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">contents</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>[&#39;Number of variants with PIP &gt; 0.5: 3\n&#39;, &#39;Number of variants with PIP &gt; 0.95: 3\n&#39;, &#39;Number of variants that have credible sets: 3\n&#39;, &#39;Number of unique credible sets: 3\n&#39;, &#39;Average number of variants per credible set: 1.0&#39;]
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/prototype_drafts"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="../mnm_analysis/rss_analysis.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">High-dimensional regression with summary statistics</p>
      </div>
    </a>
    <a class="right-next"
       href="MRAID_QTL.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Mendelian Randomization using MRAID</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#aim">Aim</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#methods-overview">Methods Overview</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow">Workflow</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-1-obtain-functional-annotations">Step 1: Obtain functional annotations</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-2-compute-ld-scores-for-annotations">Step 2: Compute LD-scores for annotations</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-1-compute-with-reference-panel-of-sequenced-individuals">Method 1: Compute with reference panel of sequenced individuals</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-2-compute-with-pre-computed-uk-biobank-ld-matrices">Method 2: Compute with pre-computed UK Biobank LD matrices</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-3-compute-with-own-pre-computed-ld-matrices">Method 3: Compute with own pre-computed LD matrices</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-3-compute-prior-causal-probabilities">Step 3: Compute Prior Causal Probabilities</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-1-use-precomputed-prior-causal-probabilities">Method 1: Use precomputed prior causal probabilities</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-2-compute-via-l2-regularized-extension-of-s-ldsc-preferred">Method 2: Compute via L2-regularized extension of S-LDSC (preferred)</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#method-3-compute-non-parametrically">Method 3: Compute Non-parametrically</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#step-4-functionally-informed-fine-mapping-with-finemapper">Step 4: Functionally informed fine mapping with finemapper</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example">Minimal Working Example</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-1-functionally-informed-fine-mapping-using-summary-statistics-file-with-precomputed-prior-causal-probabilities">Example 1: Functionally-informed fine-mapping using summary statistics file with precomputed prior causal probabilities</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-2-functionally-informed-fine-mapping-using-summary-statistics-file-generated-from-pre-obtained-annotation-and-ld-score-files">Example 2: Functionally-informed fine-mapping using summary statistics file generated from pre-obtained annotation and LD-score files</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#summary">Summary</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#example-1">Example 1:</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#example-2">Example 2:</a></li>
</ul>
</li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
       Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>