
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.17: http://docutils.sourceforge.net/" />

    <title>Summary statistics merger &#8212; ADSP-FG Project</title>
    
  <link href="../../_static/css/theme.css" rel="stylesheet">
  <link href="../../_static/css/index.ff1ffe594081f20da1ef19478df9384b.css" rel="stylesheet">

    
  <link rel="stylesheet"
    href="../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      

    
    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-book-theme.css?digest=c3fdc42140077d1ad13ad2f1588a4309" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../_static/js/index.be7d3bbb2ef33a8344ce.js">

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/togglebutton.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../_static/sphinx-book-theme.12a9622fbb08dcb3a2a40b2c02b83a57.js"></script>
    <script async="async" src="https://unpkg.com/thebe@0.5.1/lib/index.js"></script>
    <script>
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Summary statistics in VCF format" href="rds_to_vcf.html" />
    <link rel="prev" title="Fine-mapping with PolyFun" href="../integrative_analysis/SuSiE_Ann/polyfun.html" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="None">
    

    <!-- Google Analytics -->
    
  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    
    <div class="container-fluid" id="banner"></div>

    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
    <a class="navbar-brand text-wrap" href="../../index.html">
      
        <!-- `logo` is deprecated in Sphinx 4.0, so remove this when we stop supporting 3 -->
        
      
      
      <img src="../../_static/xqtl_wf.png" class="logo" alt="logo">
      
      
      <h1 class="site-logo" id="site-title">ADSP-FG Project</h1>
      
    </a>
</div><form class="bd-search d-flex align-items-center" action="../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item active">
        <ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../../README.html">
   xQTL analysis pipeline
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Complete Pipelines
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../complete_analysis/eQTL_analysis.html">
   Bulk RNA-seq eQTL analysis
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Molecular Phenotypes
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">
   RNA-seq Calling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">
   scRNA-seq expression calling
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../molecular_phenotypes/apa_calling.html">
   APA Calling
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Data Pre-processing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">
   Genotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/>
  <label for="toctree-checkbox-1">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">
     GWAS data QC workflow
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/PCA.html">
     Principal component analysis
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/GRM.html">
     Genomic relationship matrices
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">
     Genotype data formatting workflows
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">
   Phenotype data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/>
  <label for="toctree-checkbox-2">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/normalization.html">
     Molecular phenotype normalization
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">
     Phenotype data reformat
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">
   Covariate data preprocessing
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/>
  <label for="toctree-checkbox-3">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/PEER_factor.html">
     Factor analysis using PEER
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/BiCV_factor.html">
     Factor analysis using Bi-Cross validation
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../data_preprocessing/covariate/remove_covariates.html">
     Computing covariate residuals
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  QTL Association Testing
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../association_scan/cisQTL_scan.html">
   cisQTL analysis workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/>
  <label for="toctree-checkbox-4">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">
     TensorQTL QTL calling pipeline
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../association_scan/APEX/APEX.html">
     APEX QTL calling pipeline
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../association_scan/transQTL_scan.html">
   transQTL analysis workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multivariate Analysis
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/METAL_pipeline.html">
   Meta-analysis with METAL
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/>
  <label for="toctree-checkbox-5">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/METAL/METAL.html">
     Meta-analysis with METAL
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../multivariate/MASH_pipeline.html">
   Multivariate association with MASH
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/>
  <label for="toctree-checkbox-6">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/fastqtl_to_mash.html">
     eQTL summary statistics formatting
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mixture_prior.html">
     A multivariate EBNM approach for mixture multivariate distribution estimate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../multivariate/MASH/mashr.html">
     MASH analysis pipeline with data-driven prior matrices
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Statistical Fine-mapping
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../fine_mapping/univariate_fine_mapping.html">
   Univariate fine-mapping workflows
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/>
  <label for="toctree-checkbox-7">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../fine_mapping/SuSiE/SuSiE_RSS.html">
     Fine-mapping with SuSiE RSS model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../fine_mapping/SuSiE/SuSiE.html">
     Fine-mapping with SuSiE model
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../fine_mapping/multivariate_fine_mapping.html">
   Multivariate fine-mapping workflows
  </a>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Multi-omics integration
 </span>
</p>
<ul class="nav bd-sidenav">
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../integrative_analysis/colocalization.html">
   QTL and GWAS Colocalization
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/>
  <label for="toctree-checkbox-8">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/COLOC/fastenloc.html">
     Colocalization with FastENLOC
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/COLOC/mvenloc.html">
     Multivariate colocalization
    </a>
   </li>
  </ul>
 </li>
 <li class="toctree-l1 has-children">
  <a class="reference internal" href="../integrative_analysis/functional_annotation.html">
   Functional annotation integration
  </a>
  <input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/>
  <label for="toctree-checkbox-9">
   <i class="fas fa-chevron-down">
   </i>
  </label>
  <ul>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/enrichment/gwas_enrichment.html">
     Enrichment analysis: TORUS and GREGOR
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../integrative_analysis/SuSiE_Ann/polyfun.html">
     Fine-mapping with PolyFun
    </a>
   </li>
  </ul>
 </li>
</ul>
<p aria-level="2" class="caption" role="heading">
 <span class="caption-text">
  Utilities
 </span>
</p>
<ul class="current nav bd-sidenav">
 <li class="toctree-l1 current active">
  <a class="current reference internal" href="#">
   Summary statistics merger
  </a>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="rds_to_vcf.html">
   Summary statistics in VCF format
  </a>
 </li>
</ul>

    </div>
</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Supported by <a href="https://www.nia.nih.gov/research/ad-genetics">NIH/NIA</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../_sources/pipeline/misc/summary_stats_merger.ipynb"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.ipynb</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        
        <a class="issues-button"
            href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fpipeline/misc/summary_stats_merger.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/pipeline/misc/summary_stats_merger.ipynb"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>

            <!-- Full screen (wrap in <a> to have style consistency -->

<a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
        data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
        title="Fullscreen mode"><i
            class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
            <div class="tocsection onthispage pt-5 pb-3">
                <i class="fas fa-list"></i> Contents
            </div>
            <nav id="bd-toc-nav" aria-label="Page">
                <ul class="visible nav section-nav flex-column">
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#aim">
   Aim
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#notes">
   Notes
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#pre-requisites">
   Pre-requisites
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#input">
   Input
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h3 nav-item toc-entry">
    <a class="reference internal nav-link" href="#the-format-of-the-input-yaml-file">
     The format of the input yaml file
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#output">
   Output
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#example-command">
   Example command
  </a>
 </li>
 <li class="toc-h2 nav-item toc-entry">
  <a class="reference internal nav-link" href="#workflow-codes">
   Workflow codes
  </a>
 </li>
</ul>

            </nav>
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <section class="tex2jax_ignore mathjax_ignore" id="summary-statistics-merger">
<h1>Summary statistics merger<a class="headerlink" href="#summary-statistics-merger" title="Permalink to this headline">¶</a></h1>
<section id="aim">
<h2>Aim<a class="headerlink" href="#aim" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><a class="reference external" href="http://1.To">1.To</a> merge multiple summary statistic files to new summary statistic files with common SNPs</p></li>
<li><p><a class="reference external" href="http://2.To">2.To</a> deal with allele flip and reserve issues in the process of merging</p></li>
</ul>
</section>
<section id="notes">
<h2>Notes<a class="headerlink" href="#notes" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><ol class="simple">
<li><p>If there are duplicated <code class="docutils literal notranslate"><span class="pre">indels</span></code> in the summary statistics, they will be removed. For example, two SNPs at 10000 on chr1. one’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> is <code class="docutils literal notranslate"><span class="pre">T</span></code>, and <code class="docutils literal notranslate"><span class="pre">A1</span></code> is <code class="docutils literal notranslate"><span class="pre">TC</span></code>. Whereas the other one’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> is <code class="docutils literal notranslate"><span class="pre">TC</span></code>, and <code class="docutils literal notranslate"><span class="pre">A1</span></code> is <code class="docutils literal notranslate"><span class="pre">T</span></code>. Both of them will be removed. More about <code class="docutils literal notranslate"><span class="pre">indels</span></code> issues(<a class="reference external" href="https://github.com/statgenetics/UKBB_GWAS_dev/issues/81#issuecomment-1015556800">https://github.com/statgenetics/UKBB_GWAS_dev/issues/81#issuecomment-1015556800</a>).</p></li>
</ol>
</li>
<li><ol class="simple">
<li><p>If duplicated <code class="docutils literal notranslate"><span class="pre">chr:pos</span></code> (GWAS) or <code class="docutils literal notranslate"><span class="pre">gene:chr:pos</span></code> (TWAS) exist, run a recursive match for each pair of them between two summary statistic files (<code class="docutils literal notranslate"><span class="pre">query</span></code>(each of inputs) and <code class="docutils literal notranslate"><span class="pre">subject</span></code> (target file)).</p></li>
</ol>
</li>
<li><ol class="simple">
<li><p>under the same <code class="docutils literal notranslate"><span class="pre">chr:pos</span></code> or <code class="docutils literal notranslate"><span class="pre">gene:chr:pos</span></code>, The variants’ <code class="docutils literal notranslate"><span class="pre">A0</span></code> and <code class="docutils literal notranslate"><span class="pre">A1</span></code> are matched by exact, flip, reverse, or flip+reverse models. Only one of them is <code class="docutils literal notranslate"><span class="pre">True</span></code>, the variant in two files are matched. If they are matched by flip or flip+reverse, the sign of <code class="docutils literal notranslate"><span class="pre">query</span></code>’s <code class="docutils literal notranslate"><span class="pre">STAT</span></code> will be inversed. And the <code class="docutils literal notranslate"><span class="pre">query</span></code>’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> and <code class="docutils literal notranslate"><span class="pre">A1</span></code> will be the same as the <code class="docutils literal notranslate"><span class="pre">subject</span></code>’s <code class="docutils literal notranslate"><span class="pre">A0</span></code> and <code class="docutils literal notranslate"><span class="pre">A1</span></code>.</p></li>
</ol>
</li>
</ul>
</section>
<section id="pre-requisites">
<h2>Pre-requisites<a class="headerlink" href="#pre-requisites" title="Permalink to this headline">¶</a></h2>
<p>Make sure you install the pre-requisited before running this notebook:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">pip</span> <span class="n">install</span> <span class="n">LDtools</span>
</pre></div>
</div>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h2>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--cwd</span></code>, the path of working directory</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--yml_path</span></code>, the path of yaml file</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--keep-ambiguous</span></code>, boolean. default False. if add –keep-ambiguous parameter, keep ambiguous alleles which can not be decided from flip or reverse, such as A/T or C/G. Otherwise, remove them.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--intersect</span></code>, boolean. default False. if add –intersect parameter, output intersect SNPs in all input files.</p></li>
</ul>
<section id="the-format-of-the-input-yaml-file">
<h3>The format of the input yaml file<a class="headerlink" href="#the-format-of-the-input-yaml-file" title="Permalink to this headline">¶</a></h3>
<p>For GWAS summary statistics: <code class="docutils literal notranslate"><span class="pre">ID</span></code> is <code class="docutils literal notranslate"><span class="pre">CHR,POS,A0,A1</span></code>, which can be used as a unique label for each variant.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">INPUT:</span>
  <span class="o">-</span> <span class="o">./</span><span class="n">data</span><span class="o">/</span><span class="n">testflip</span><span class="o">/*.</span><span class="n">gz</span><span class="p">:</span>
        <span class="n">ID</span><span class="p">:</span> <span class="n">CHR</span><span class="p">,</span><span class="n">POS</span><span class="p">,</span><span class="n">A0</span><span class="p">,</span><span class="n">A1</span>
        <span class="n">CHR</span><span class="p">:</span> <span class="n">CHR</span>
        <span class="n">POS</span><span class="p">:</span> <span class="n">POS</span>
        <span class="n">A0</span><span class="p">:</span> <span class="n">REF</span>
        <span class="n">A1</span><span class="p">:</span> <span class="n">ALT</span>
        <span class="n">SNP</span><span class="p">:</span> <span class="n">SNP</span>
        <span class="n">STAT</span><span class="p">:</span> <span class="n">BETA</span>
        <span class="n">SE</span><span class="p">:</span> <span class="n">SE</span>
        <span class="n">P</span><span class="p">:</span> <span class="n">P</span>
  <span class="o">-</span> <span class="o">./</span><span class="n">data</span><span class="o">/</span><span class="n">testflip</span><span class="o">/</span><span class="n">flip</span><span class="o">/</span><span class="n">snps500_flip</span><span class="o">.</span><span class="n">regenie</span><span class="o">.</span><span class="n">snp_stats</span><span class="o">.</span><span class="n">gz</span><span class="p">:</span>
  
<span class="kn">TARGET:</span> 
  <span class="o">-</span> <span class="o">./</span><span class="n">data</span><span class="o">/</span><span class="n">testflip</span><span class="o">/</span><span class="n">snps500</span><span class="o">.</span><span class="n">regenie</span><span class="o">.</span><span class="n">snp_stats</span><span class="o">.</span><span class="n">gz</span><span class="p">:</span>
        <span class="n">ID</span><span class="p">:</span> <span class="n">CHR</span><span class="p">,</span><span class="n">POS</span><span class="p">,</span><span class="n">A0</span><span class="p">,</span><span class="n">A1</span>
        <span class="n">CHR</span><span class="p">:</span> <span class="n">CHR</span>
        <span class="n">POS</span><span class="p">:</span> <span class="n">POS</span>
        <span class="n">A0</span><span class="p">:</span> <span class="n">REF</span>
        <span class="n">A1</span><span class="p">:</span> <span class="n">ALT</span>
        <span class="n">SNP</span><span class="p">:</span> <span class="n">SNP</span>
        <span class="n">STAT</span><span class="p">:</span> <span class="n">BETA</span>
        <span class="n">SE</span><span class="p">:</span> <span class="n">SE</span>
        <span class="n">P</span><span class="p">:</span> <span class="n">P</span>
<span class="kn">OUTPUT:</span> <span class="n">data</span><span class="o">/</span><span class="n">testflip</span><span class="o">/</span><span class="n">output</span><span class="o">/</span>
</pre></div>
</div>
<p>For TWAS summary statistics: <code class="docutils literal notranslate"><span class="pre">ID</span></code> is <code class="docutils literal notranslate"><span class="pre">GENE,CHR,POS,A0,A1</span></code>, which add the <code class="docutils literal notranslate"><span class="pre">GENE</span></code> name because a variant can be made association with multiple genes.</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="kn">INPUT:</span>
  <span class="o">-</span> <span class="n">data</span><span class="o">/</span><span class="n">twas</span><span class="o">/*.</span><span class="n">txt</span><span class="p">:</span>
        <span class="n">ID</span><span class="p">:</span> <span class="n">GENE</span><span class="p">,</span><span class="n">CHR</span><span class="p">,</span><span class="n">POS</span><span class="p">,</span><span class="n">A0</span><span class="p">,</span><span class="n">A1</span>
        <span class="n">CHR</span><span class="p">:</span> <span class="n">chrom</span>
        <span class="n">POS</span><span class="p">:</span> <span class="n">pos</span>
        <span class="n">A0</span><span class="p">:</span> <span class="n">ref</span>
        <span class="n">A1</span><span class="p">:</span> <span class="n">alt</span>
        <span class="n">SNP</span><span class="p">:</span> <span class="n">variant_id</span>
        <span class="n">GENE</span><span class="p">:</span> <span class="n">gene</span>
        <span class="n">STAT</span><span class="p">:</span> <span class="n">beta</span>
        <span class="n">SE</span><span class="p">:</span> <span class="n">se</span>
        <span class="n">P</span><span class="p">:</span> <span class="n">pval</span>
 
  
<span class="kn">TARGET:</span> 
  <span class="o">-</span> <span class="n">data</span><span class="o">/</span><span class="n">twas</span><span class="o">/</span><span class="n">DLPFC</span><span class="o">.</span><span class="n">chr6</span><span class="o">.</span><span class="n">mol_phe</span><span class="o">.</span><span class="n">cis_long_table</span><span class="o">.</span><span class="n">reformated</span><span class="o">.</span><span class="n">txt</span><span class="p">:</span>
        <span class="n">ID</span><span class="p">:</span> <span class="n">GENE</span><span class="p">,</span><span class="n">CHR</span><span class="p">,</span><span class="n">POS</span><span class="p">,</span><span class="n">A0</span><span class="p">,</span><span class="n">A1</span>
        <span class="n">CHR</span><span class="p">:</span> <span class="n">chrom</span>
        <span class="n">POS</span><span class="p">:</span> <span class="n">pos</span>
        <span class="n">A0</span><span class="p">:</span> <span class="n">ref</span>
        <span class="n">A1</span><span class="p">:</span> <span class="n">alt</span>
        <span class="n">SNP</span><span class="p">:</span> <span class="n">variant_id</span>
        <span class="n">GENE</span><span class="p">:</span> <span class="n">gene</span>
        <span class="n">STAT</span><span class="p">:</span> <span class="n">beta</span>
        <span class="n">SE</span><span class="p">:</span> <span class="n">se</span>
        <span class="n">P</span><span class="p">:</span> <span class="n">pval</span>
<span class="kn">OUTPUT:</span> <span class="o">../</span><span class="n">data</span><span class="o">/</span><span class="n">twas</span><span class="o">/</span><span class="n">output</span><span class="o">/</span>
</pre></div>
</div>
<p>There are three parts in the input yaml file.</p>
<ul class="simple">
<li><p>INPUT</p>
<ul>
<li><p>A list of yml file, as the output from yml_generator, each yml file documents a set of input</p>
<ul>
<li><p>the input summary statistic files with the column names in below.</p></li>
<li><p>the input files can be from multiple directory and from different format. The input paths must follow the rules related to Unix shell. the format is to pair the column names with keys (CHR, POS, A0, A1, SNP, STAT, SE, P). if not provided, the column names of the input file will be considered as the default keys.</p></li>
<li><p>The input summary statistic file cannot have duplicated chr:pos</p></li>
<li><p>The input summary statstic file cannot have # in its header
-<code class="docutils literal notranslate"><span class="pre">ID</span></code> in yml is a unique identifier for each SNP. The default are CHR,POS. These duplicated identifier will only keep the first one in the file.</p></li>
</ul>
</li>
</ul>
</li>
<li><p>TARGET</p>
<ul>
<li><p>the target file is a reference summary statistic file or a file with chr, pos, a0, a1 columns at least, which the other files compare with.</p></li>
</ul>
</li>
<li><p>OUTPUT</p>
<ul>
<li><p>the path of an output directory for new summary statistic files</p></li>
</ul>
</li>
</ul>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h2>
<p>new summary statistic files with common SNPs in all input files. the sign of statistics has been corrected to make it consistent in different data.</p>
<ul class="simple">
<li><p>for each input sumstat file, a qced version will be generated.</p></li>
<li><p>The generated sumstat files will have header as “CHR  ,   POS  ,   A0   ,   A1    ,  SNP   ,  STAT ,   SE    ,  P” regardless of input header</p></li>
<li><p>The generated sumstat files will be in gz format.</p></li>
</ul>
</section>
<section id="example-command">
<h2>Example command<a class="headerlink" href="#example-command" title="Permalink to this headline">¶</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="o">./</span><span class="n">summary_stats_merger</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">--</span><span class="n">cwd</span> <span class="n">data</span> <span class="o">--</span><span class="n">yml_list</span> <span class="n">data</span><span class="o">/</span><span class="n">yml_list</span><span class="o">.</span><span class="n">txt</span> <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">ambiguous</span> <span class="o">--</span><span class="n">intersect</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Work directory where output will be saved to</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1">## path to a list of yml file , with columns #chr and dir</span>
<span class="kn">parameter:</span> <span class="n">yml_list</span> <span class="o">=</span> <span class="n">path</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="n">yml_path</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">yml_list</span><span class="p">,</span><span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span><span class="o">.</span><span class="n">values</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
<span class="c1">#if add --keep-ambiguous parameter, keep ambiguous alleles which can not be decided from flip or reverse, such as A/T or C/G. Otherwise, remove them.</span>
<span class="kn">parameter:</span> <span class="n">keep_ambiguous</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># if add --intersect parameter, output intersect SNPs in all input files.</span>
<span class="kn">parameter:</span> <span class="n">intersect</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># Containers that contains the necessary packages</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="nb">str</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="workflow-codes">
<h2>Workflow codes<a class="headerlink" href="#workflow-codes" title="Permalink to this headline">¶</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">default_1</span> <span class="p">(</span><span class="n">export</span> <span class="n">utils</span> <span class="n">script</span><span class="p">)]</span>
<span class="kn">depends:</span> <span class="n">Py_Module</span><span class="p">(</span><span class="s1">&#39;LDtools&#39;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/utils.py&#39;</span>
<span class="kn">report:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">output</span><span class="o">=</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/utils.py&#39;</span>

    <span class="kn">import</span> <span class="nn">os</span>
    <span class="kn">from</span> <span class="nn">LDtools.sumstat</span> <span class="kn">import</span> <span class="n">read_sumstat</span>
    <span class="kn">from</span> <span class="nn">LDtools.utils</span> <span class="kn">import</span> <span class="o">*</span>
    <span class="k">def</span> <span class="nf">merge_sumstats</span><span class="p">(</span><span class="n">yml</span><span class="p">,</span><span class="n">keep_ambiguous</span><span class="p">,</span><span class="n">intersect</span><span class="p">):</span>
        <span class="c1">#parse yaml</span>
        <span class="n">yml</span> <span class="o">=</span> <span class="n">load_yaml</span><span class="p">(</span><span class="n">yml</span><span class="p">)</span>
        <span class="n">input_dict</span> <span class="o">=</span> <span class="n">parse_input</span><span class="p">(</span><span class="n">yml</span><span class="p">[</span><span class="s1">&#39;INPUT&#39;</span><span class="p">])</span>
        <span class="n">target_dict</span> <span class="o">=</span> <span class="n">parse_input</span><span class="p">(</span><span class="n">yml</span><span class="p">[</span><span class="s1">&#39;TARGET&#39;</span><span class="p">])</span>
        <span class="n">output_path</span> <span class="o">=</span> <span class="n">yml</span><span class="p">[</span><span class="s1">&#39;OUTPUT&#39;</span><span class="p">]</span>

        <span class="n">input_dict</span><span class="p">[</span><span class="nb">list</span><span class="p">(</span><span class="n">target_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">target_dict</span><span class="o">.</span><span class="n">values</span><span class="p">())[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">lst_sumstats_file</span> <span class="o">=</span> <span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">i</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">input_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">()]</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Total number of sumstats: &#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">))</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">))</span><span class="o">&lt;</span><span class="nb">len</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">Exception</span><span class="p">(</span><span class="s2">&quot;There are duplicated names in </span><span class="si">{}</span><span class="s2">&quot;</span><span class="o">.</span><span class="n">format</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">))</span>
        <span class="c1">#read all sumstats</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">input_dict</span><span class="p">)</span>
        <span class="n">lst_sumstats</span> <span class="o">=</span> <span class="p">{</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">i</span><span class="p">):</span><span class="n">read_sumstat</span><span class="p">(</span><span class="n">i</span><span class="p">,</span><span class="n">j</span><span class="p">)</span> <span class="k">for</span> <span class="n">i</span><span class="p">,</span><span class="n">j</span> <span class="ow">in</span> <span class="n">input_dict</span><span class="o">.</span><span class="n">items</span><span class="p">()}</span>
        <span class="n">nqs</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="c1">#check duplicated indels and remove them.</span>
        <span class="n">subject</span> <span class="o">=</span> <span class="n">check_indels</span><span class="p">(</span><span class="n">lst_sumstats</span><span class="p">[</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">target_dict</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="mi">0</span><span class="p">])])</span>
        <span class="k">for</span> <span class="n">query</span> <span class="ow">in</span> <span class="n">lst_sumstats</span><span class="o">.</span><span class="n">values</span><span class="p">():</span>
            <span class="c1">#check duplicated indels and remove them.</span>
            <span class="n">query</span> <span class="o">=</span> <span class="n">check_indels</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
            <span class="c1">#under the same chr:pos or gene:chr:pos. match A0 and A1 by exact, flip, reverse, or flip+reverse.</span>
            <span class="c1">#if duplicated chr_pos or gene_chr_pos exist, run a recursive match for each pair of them between query and subject.</span>
            <span class="n">nq</span><span class="p">,</span><span class="n">_</span> <span class="o">=</span> <span class="n">snps_match</span><span class="p">(</span><span class="n">query</span><span class="p">,</span><span class="n">subject</span><span class="p">,</span><span class="n">keep_ambiguous</span><span class="p">)</span>
            <span class="n">nqs</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">nq</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">intersect</span><span class="p">:</span>
            <span class="c1">#get common snps</span>
            <span class="n">common_snps</span> <span class="o">=</span> <span class="nb">set</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="o">*</span><span class="p">[</span><span class="nb">set</span><span class="p">(</span><span class="n">nq</span><span class="o">.</span><span class="n">SNP</span><span class="p">)</span> <span class="k">for</span> <span class="n">nq</span> <span class="ow">in</span> <span class="n">nqs</span><span class="p">])</span>
            <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;Total number of common SNPs: &#39;</span><span class="p">,</span><span class="nb">len</span><span class="p">(</span><span class="n">common_snps</span><span class="p">))</span>
            <span class="c1">#write out new smustats</span>
            <span class="k">for</span> <span class="n">output_sumstats</span><span class="p">,</span><span class="n">nq</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">,</span><span class="n">nqs</span><span class="p">):</span>
                <span class="n">sumstats</span> <span class="o">=</span> <span class="n">nq</span><span class="p">[</span><span class="n">nq</span><span class="o">.</span><span class="n">SNP</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">common_snps</span><span class="p">)]</span>
                <span class="n">sumstats</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">output_sumstats</span><span class="p">),</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">output_sumstats</span><span class="p">,</span><span class="n">nq</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">lst_sumstats_file</span><span class="p">,</span><span class="n">nqs</span><span class="p">):</span>
                <span class="c1">#output match SNPs with target SNPs.</span>
                <span class="n">nq</span><span class="o">.</span><span class="n">to_csv</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">output_path</span><span class="p">,</span> <span class="n">output_sumstats</span><span class="p">),</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span><span class="n">compression</span><span class="o">=</span><span class="s1">&#39;gzip&#39;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="s1">&#39;All are done!!!&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">default_2</span> <span class="p">(</span><span class="n">merge</span> <span class="n">sumstats</span><span class="p">)]</span>
<span class="kn">depends:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/utils.py&#39;</span>
<span class="kn">input:</span> <span class="n">for_each</span> <span class="o">=</span> <span class="s2">&quot;yml_path&quot;</span>
<span class="kn">python:</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="nb">input</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/utils.py&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/output.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/output.stdout&#39;</span>

    <span class="n">yml</span> <span class="o">=</span> <span class="s2">&quot;$</span><span class="si">{_yml_path[1]}</span><span class="s2">&quot;</span>
    <span class="n">keep_ambiguous</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">keep_ambiguous</span><span class="p">}</span>
    <span class="n">intersect</span> <span class="o">=</span> <span class="err">$</span><span class="p">{</span><span class="n">intersect</span><span class="p">}</span>
    <span class="nb">print</span><span class="p">(</span><span class="n">yml</span><span class="p">,</span> <span class="n">keep_ambiguous</span><span class="p">,</span><span class="n">intersect</span><span class="p">)</span>
    <span class="n">merge_sumstats</span><span class="p">(</span><span class="n">yml</span><span class="p">,</span> <span class="n">keep_ambiguous</span><span class="p">,</span><span class="n">intersect</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            kernelName: "sos",
            path: "./pipeline/misc"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

              </div>
              
        
            <!-- Previous / next buttons -->
<div class='prev-next-area'> 
    <a class='left-prev' id="prev-link" href="../integrative_analysis/SuSiE_Ann/polyfun.html" title="previous page">
        <i class="fas fa-angle-left"></i>
        <div class="prev-next-info">
            <p class="prev-next-subtitle">previous</p>
            <p class="prev-next-title">Fine-mapping with PolyFun</p>
        </div>
    </a>
    <a class='right-next' id="next-link" href="rds_to_vcf.html" title="next page">
    <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Summary statistics in VCF format</p>
    </div>
    <i class="fas fa-angle-right"></i>
    </a>
</div>
        
        </div>
    </div>
    <footer class="footer">
    <div class="container">
      <p>
        
          By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics Consortium<br/>
        
            &copy; Copyright 2021, ADSP-FG xQTL methods and data integration subgroup.<br/>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>
  
  <script src="../../_static/js/index.be7d3bbb2ef33a8344ce.js"></script>

  </body>
</html>