{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mendelian Randomization using fine-mapping IV\n",
    "\n",
    "The procedures for PTWAS estimation and model validation that are implemented in the R script. The unit of analysis is a single gene-trait pair, whose information is summarized in a text file. First, we generate the units of analysis. Then, we conduct the causal effect estimation and validation procedure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overview\n",
    "\n",
    "* Organize the eQTL fine-mapping results, eQTL summary statistics, and GWAS summary statistics data into a format that can be processed by ptwas_est.  \n",
    "\n",
    "* run ptwas_est with R codes: causal effect estimation and validation."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Input\n",
    "\n",
    "* eQTL finemapping result:  \n",
    "\n",
    "    *   This result contain the information of credible sets and pip. <br/>\n",
    "<br/>\n",
    "* eqtl sumstats: se and beta\n",
    "\n",
    "* gwas sumstats: se and beta"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Output\n",
    "\n",
    "* data format for ptwas_est\n",
    "\n",
    "* ptwas_est resuls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Example\n",
    "\n",
    "### Generate the units of analysis\n",
    "\n",
    "####  Input\n",
    "* **eQTL finemapping**: /mnt/vast/hpc/csg/rf2872/Work/INTACT/fastenloc_test/eqtl_susie_table.txt\n",
    "\n",
    "* **eQTL sumstats**: /mnt/vast/hpc/csg/ftp_lisanwanglab_sync/ftp_fgc_xqtl/projects/rna-seq/BU/ROSMAP_DLPFC/eQTL/association_scan/norminal_qced_files\n",
    "\n",
    "* **GWAS sumstats**: /mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Output file format\n",
    "\n",
    " Each row of the file corresponds to a single SNP that is inferred as a memeber of a signal cluster, the columns are:\n",
    "\n",
    "* snp: SNP name\n",
    "\n",
    "* beta_eQTL: eQTL effect\n",
    "\n",
    "* se_eQTL: standard error of estimated eQTL effect\n",
    "\n",
    "* beta_GWAS: GWAS effect\n",
    "\n",
    "* se_GWAS: standard error of GWAS effect\n",
    "\n",
    "* cluster: Signal cluster ID (credible sets index)\n",
    "\n",
    "* pip: SNP posterior inclusiion probability (PIP)\n",
    "\n",
    "* gene_id: gene name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#######organize the data of eqtl finemapping, eqtl summary statistics and GWAS summary statistics\n",
    "library(data.table)\n",
    "library(tidyverse)\n",
    "library(stringr)\n",
    "\n",
    "#####load the the eqtl finemapping folders\n",
    "eqtl_table_list = fread(\"/mnt/vast/hpc/csg/rf2872/Work/INTACT/fastenloc_test/eqtl_susie_table.txt\")\n",
    "finemapping_folder = \"/mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl/output/susie_per_gene_tad/cache\"\n",
    "\n",
    "#####load the eqtl sumstats\n",
    "eqtl_folder = \"/mnt/vast/hpc/csg/ftp_lisanwanglab_sync/ftp_fgc_xqtl/projects/rna-seq/BU/ROSMAP_DLPFC/eQTL/association_scan/norminal_qced_files\"\n",
    "#####load the GWAS sumstats\n",
    "GWAS_folder = \"/mnt/vast/hpc/csg/xqtl_workflow_testing/ADGWAS/data_intergration/ADGWAS2022\"\n",
    "\n",
    "gene_name_list = eqtl_table_list$molecular_trait_i\n",
    "eqtl = data.frame()\n",
    "for (k in 1:100){\n",
    "gene_name = gene_name_list[k]\n",
    "eqtl_finemapp = readRDS(paste0(finemapping_folder,\"/demo.\",gene_name,\".unisusie.fit.rds\", sep=\"\"))\n",
    "cluster_num = eqtl_finemapp$dlpfc_eqtl$sets$cs_index\n",
    "if (!is.null(cluster_num)){\n",
    "    chr = str_sub(unique(str_split(eqtl_finemapp$dlpfc_eqtl$variable_name,\":\",simplify=T)[,1]),4)\n",
    "    eqtl_data = fread(file = paste0(eqtl_folder,\"/dlpfc_batch_all.rnaseqc.low_expression_filtered.outlier_removed.tmm.expression.bed.processed_phenotype.per_chrom_dlpfc_batch_all.rnaseqc.ROSMAP_covariates.ROSMAP_NIA_WGS.pca.PEER.txt.\",chr,\".norminal.cis_long_table.txt\",sep=\"\"))\n",
    "    GWAS_data = fread(paste0(GWAS_folder,\"/ADGWAS_Bellenguez_2022.\",chr,\"/ADGWAS2022.chr\",chr,\".sumstat.tsv\",sep=\"\"))\n",
    "    for (i in 1:length(cluster_num)){\n",
    "       cluster_id = cluster_num[i]\n",
    "       cluster_id_name  = paste0(\"L\",cluster_id,sep=\"\")\n",
    "       crediblesets = eqtl_finemapp$dlpfc_eqtl$sets$cs[[cluster_id_name]]\n",
    "       variants_name = eqtl_finemapp$dlpfc_eqtl$variable_name[crediblesets]\n",
    "       #####get the chromosome fo the variants_name\n",
    "       chr = str_sub(unique(str_split(variants_name,\":\",simplify=T)[,1]),4)\n",
    "       pip = gsub(\":\",\"\",eqtl_finemapp$dlpfc_eqtl$pip[crediblesets])\n",
    "       eqtl_cluster = tibble(snp = variants_name, cluster  = rep(cluster_id, length(variants_name)),pip = pip,gene_id = rep(gene_name,length(variants_name)))\n",
    "       eqtl_beta_se = eqtl_data%>%filter(molecular_trait_id%in%gene_name)%>%filter(variant%in%variants_name)%>%select(variant,beta,se)%>%rename(\"snp\" = \"variant\",\"beta_eQTL\"=\"beta\",\"se_eQTL\"=\"se\")\n",
    "       GWAS_beta_se = GWAS_data%>%mutate(snp = sub(\"_\",\":\",variant))%>%filter(snp%in%variants_name)%>%select(snp,beta,se)%>%rename(\"beta_GWAS\"=\"beta\",\"se_GWAS\"=\"se\")\n",
    "       df = list(eqtl_beta_se, GWAS_beta_se, eqtl_cluster)%>%reduce(full_join, by = \"snp\")%>%drop_na()\n",
    "       eqtl = rbind(eqtl, df)\n",
    "    }\n",
    "    rm(eqtl_data)\n",
    "    rm(GWAS_data)\n",
    "    fwrite(eqtl,file  = \"/home/aw3600/PTWAS/PTWAS_test_data.txt\")\n",
    "}\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Causal effect estimation and validation\n",
    "\n",
    "#### Estimation\n",
    "* First, we select strong eQTLs by thresholding on the corresponding signal-level PIPs (cpip = 0.5).\n",
    "\n",
    "* Then, we aggregate the estimates from the individual eQTLs using a fixed-effect meta-analysis procedure.\n",
    "\n",
    "#### Validation\n",
    "Identify severe violations of the exclusion restriction (ER) assumption and exclude them from the downstream estimation procedure.\n",
    "\n",
    "#### Inuput file format\n",
    "The output from the first step\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Output file format\n",
    "Each row of the file corresponds to a gene and the columns are:\n",
    "\n",
    "* gene_id: gene name\n",
    "\n",
    "* num_cluster: the number of credible sets of gene\n",
    "\n",
    "* num_instruments: the number of instruments included in the gene \n",
    "\n",
    "* spip: the sum of pip for credible sets of each gene\n",
    "\n",
    "* grp_beta: signal-level estimates (combine the SNP-level estimates from the member SNPs weighted by their corresponding PIPs)\n",
    "\n",
    "* grp_se: standard error of the signal-level estimates\n",
    "\n",
    "* meta: gene-level estimate of the causal effect (aggregate the signal-level estimates by applying a fixed-effect meta-analysis model)\n",
    "\n",
    "* se_meta: standard error of the gene-level estimate of the causal effect\n",
    "\n",
    "* Q: Cochran’s Q statistic\n",
    "\n",
    "* I2: $I^2$ statistics (quantify the heterogeneity, range from 0 to 1, $I^2 \\rightarrow 0$ indicates reasonable consistency among estimates by multiple eQTLs, whereas $I^2 \\rightarrow 1$ 1 implies severe departures from exclusion restriction (ER).)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Input"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 6 × 8</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>snp</th><th scope=col>beta_eQTL</th><th scope=col>se_eQTL</th><th scope=col>beta_GWAS</th><th scope=col>se_GWAS</th><th scope=col>cluster</th><th scope=col>pip</th><th scope=col>gene_id</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>chr1:169880000_A_G</td><td>-0.2883383</td><td>0.03356421</td><td>0.0031</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>\n",
       "\t<tr><td>chr1:169880823_G_A</td><td>-0.2883383</td><td>0.03356421</td><td>0.0034</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>\n",
       "\t<tr><td>chr1:169880877_T_C</td><td>-0.2883383</td><td>0.03356421</td><td>0.0033</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>\n",
       "\t<tr><td>chr1:169882224_A_G</td><td>-0.2859217</td><td>0.03388142</td><td>0.0024</td><td>0.0138</td><td>1</td><td>0.003048243</td><td>ENSG00000000457</td></tr>\n",
       "\t<tr><td>chr1:169882228_G_A</td><td>-0.2859217</td><td>0.03388142</td><td>0.0023</td><td>0.0138</td><td>1</td><td>0.003048243</td><td>ENSG00000000457</td></tr>\n",
       "\t<tr><td>chr1:169882581_G_T</td><td>-0.2883383</td><td>0.03356421</td><td>0.0025</td><td>0.0137</td><td>1</td><td>0.020186373</td><td>ENSG00000000457</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 6 × 8\n",
       "\\begin{tabular}{llllllll}\n",
       " snp & beta\\_eQTL & se\\_eQTL & beta\\_GWAS & se\\_GWAS & cluster & pip & gene\\_id\\\\\n",
       " <chr> & <dbl> & <dbl> & <dbl> & <dbl> & <int> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t chr1:169880000\\_A\\_G & -0.2883383 & 0.03356421 & 0.0031 & 0.0137 & 1 & 0.020186373 & ENSG00000000457\\\\\n",
       "\t chr1:169880823\\_G\\_A & -0.2883383 & 0.03356421 & 0.0034 & 0.0137 & 1 & 0.020186373 & ENSG00000000457\\\\\n",
       "\t chr1:169880877\\_T\\_C & -0.2883383 & 0.03356421 & 0.0033 & 0.0137 & 1 & 0.020186373 & ENSG00000000457\\\\\n",
       "\t chr1:169882224\\_A\\_G & -0.2859217 & 0.03388142 & 0.0024 & 0.0138 & 1 & 0.003048243 & ENSG00000000457\\\\\n",
       "\t chr1:169882228\\_G\\_A & -0.2859217 & 0.03388142 & 0.0023 & 0.0138 & 1 & 0.003048243 & ENSG00000000457\\\\\n",
       "\t chr1:169882581\\_G\\_T & -0.2883383 & 0.03356421 & 0.0025 & 0.0137 & 1 & 0.020186373 & ENSG00000000457\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 6 × 8\n",
       "\n",
       "| snp &lt;chr&gt; | beta_eQTL &lt;dbl&gt; | se_eQTL &lt;dbl&gt; | beta_GWAS &lt;dbl&gt; | se_GWAS &lt;dbl&gt; | cluster &lt;int&gt; | pip &lt;dbl&gt; | gene_id &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|\n",
       "| chr1:169880000_A_G | -0.2883383 | 0.03356421 | 0.0031 | 0.0137 | 1 | 0.020186373 | ENSG00000000457 |\n",
       "| chr1:169880823_G_A | -0.2883383 | 0.03356421 | 0.0034 | 0.0137 | 1 | 0.020186373 | ENSG00000000457 |\n",
       "| chr1:169880877_T_C | -0.2883383 | 0.03356421 | 0.0033 | 0.0137 | 1 | 0.020186373 | ENSG00000000457 |\n",
       "| chr1:169882224_A_G | -0.2859217 | 0.03388142 | 0.0024 | 0.0138 | 1 | 0.003048243 | ENSG00000000457 |\n",
       "| chr1:169882228_G_A | -0.2859217 | 0.03388142 | 0.0023 | 0.0138 | 1 | 0.003048243 | ENSG00000000457 |\n",
       "| chr1:169882581_G_T | -0.2883383 | 0.03356421 | 0.0025 | 0.0137 | 1 | 0.020186373 | ENSG00000000457 |\n",
       "\n"
      ],
      "text/plain": [
       "  snp                beta_eQTL  se_eQTL    beta_GWAS se_GWAS cluster\n",
       "1 chr1:169880000_A_G -0.2883383 0.03356421 0.0031    0.0137  1      \n",
       "2 chr1:169880823_G_A -0.2883383 0.03356421 0.0034    0.0137  1      \n",
       "3 chr1:169880877_T_C -0.2883383 0.03356421 0.0033    0.0137  1      \n",
       "4 chr1:169882224_A_G -0.2859217 0.03388142 0.0024    0.0138  1      \n",
       "5 chr1:169882228_G_A -0.2859217 0.03388142 0.0023    0.0138  1      \n",
       "6 chr1:169882581_G_T -0.2883383 0.03356421 0.0025    0.0137  1      \n",
       "  pip         gene_id        \n",
       "1 0.020186373 ENSG00000000457\n",
       "2 0.020186373 ENSG00000000457\n",
       "3 0.020186373 ENSG00000000457\n",
       "4 0.003048243 ENSG00000000457\n",
       "5 0.003048243 ENSG00000000457\n",
       "6 0.020186373 ENSG00000000457"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "formatted_input = fread(\"/home/aw3600/PTWAS/PTWAS_test_data.txt\")\n",
    "head(formatted_input)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### ptwas_est"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "#####define the threshold of the cumulative cluster pip\n",
    "cpip = 0.5\n",
    "\n",
    "######heterogeneity:  calculate I2 statistics based on the Cochran's Q statistic\n",
    "calc_I2 = function(Q, Est) {\n",
    "    Q = Q[[1]]\n",
    "    Est = length(unique(Est))\n",
    "    I2 = if (Q > 1e-3) (Q - Est + 1)/Q else 0\n",
    "    return(if (I2 < 0) 0 else I2)\n",
    "}\n",
    "\n",
    "#####ptwas_est\n",
    "output = formatted_input %>%\n",
    "    mutate(\n",
    "        beta_eQTL = beta_eQTL/se_eQTL,\n",
    "        se_eQTL = 1) %>%\n",
    "    group_by(gene_id, cluster) %>%\n",
    "    mutate(spip = sum(pip)) %>%\n",
    "    filter(spip >= cpip) %>% # Cumulative cluster pip greater than a user defined cumulative pip threshold\n",
    "    group_by(gene_id, cluster) %>%\n",
    "    mutate(\n",
    "        beta_yx = beta_GWAS/beta_eQTL,\n",
    "        se_yx = sqrt(\n",
    "            (se_GWAS^2/beta_eQTL^2) + ((beta_GWAS^2*se_eQTL^2)/beta_eQTL^4)),\n",
    "        grp_beta = sum((beta_yx*pip)/spip),\n",
    "        grp_se = sum((beta_yx^2 + se_yx^2)*pip/spip)) %>%\n",
    "    mutate(\n",
    "        grp_se = sqrt(grp_se - grp_beta^2),\n",
    "        wv = grp_se^-2) %>%\n",
    "    ungroup() %>%\n",
    "    group_by(gene_id) %>%\n",
    "    mutate(\n",
    "        meta = sum(unique(wv) * unique(grp_beta)),\n",
    "        sum_w = sum(unique(wv)),\n",
    "        se_meta = sqrt(sum_w^-1),\n",
    "        num_cluster = length(unique(cluster))) %>%\n",
    "    mutate(\n",
    "        num_instruments = length(snp),\n",
    "        meta = meta/sum_w,\n",
    "        ######sum(unique(wv)*(unique(grp_beta) - unique(meta))^2)\n",
    "        Q = sum(unique(wv)*(unique(grp_beta) - unique(meta))^2),\n",
    "        I2 = calc_I2(Q, grp_beta)) %>%\n",
    "        ungroup() %>%\n",
    "    distinct(gene_id, .keep_all = TRUE) %>%\n",
    "    mutate(\n",
    "        spip = round(spip, 3),\n",
    "        grp_beta = round(grp_beta, 3),\n",
    "        meta = round(meta, 3),\n",
    "        se_meta = round(se_meta, 3),\n",
    "        Q = round(Q, 3),\n",
    "        I2 = round(I2, 3)) %>%\n",
    "    select(gene_id, num_cluster, num_instruments, spip, grp_beta, grp_se, meta, se_meta, Q, I2)\n",
    "    fwrite(output,file  = \"/home/aw3600/PTWAS/PTWAS_est_data.txt\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A tibble: 51 × 10</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>gene_id</th><th scope=col>num_cluster</th><th scope=col>num_instruments</th><th scope=col>spip</th><th scope=col>grp_beta</th><th scope=col>grp_se</th><th scope=col>meta</th><th scope=col>se_meta</th><th scope=col>Q</th><th scope=col>I2</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>ENSG00000000457</td><td>2</td><td> 74</td><td>0.696</td><td> 0.000</td><td>0.0016118763</td><td>-0.001</td><td>0.001</td><td>0.157</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000000971</td><td>1</td><td> 22</td><td>0.739</td><td>-0.001</td><td>0.0010414938</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000001084</td><td>1</td><td> 10</td><td>0.881</td><td> 0.000</td><td>0.0010095378</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000001167</td><td>1</td><td> 16</td><td>0.737</td><td> 0.009</td><td>0.0029981468</td><td> 0.009</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000001460</td><td>1</td><td> 16</td><td>0.745</td><td>-0.001</td><td>0.0006707913</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000001461</td><td>1</td><td> 47</td><td>0.679</td><td>-0.005</td><td>0.0028313411</td><td>-0.005</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000001561</td><td>1</td><td>  9</td><td>0.586</td><td> 0.001</td><td>0.0006746664</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000001626</td><td>1</td><td> 28</td><td>0.801</td><td> 0.001</td><td>0.0011200443</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000001629</td><td>2</td><td> 48</td><td>0.678</td><td> 0.000</td><td>0.0019744363</td><td> 0.000</td><td>0.001</td><td>0.046</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000001630</td><td>1</td><td> 67</td><td>0.807</td><td> 0.001</td><td>0.0011870369</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000002822</td><td>2</td><td> 19</td><td>0.926</td><td>-0.001</td><td>0.0030673516</td><td> 0.001</td><td>0.002</td><td>0.897</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000002834</td><td>1</td><td>  1</td><td>0.999</td><td>-0.001</td><td>0.0012142297</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000002933</td><td>2</td><td> 40</td><td>0.985</td><td>-0.001</td><td>0.0013062166</td><td> 0.001</td><td>0.001</td><td>1.257</td><td>0.205</td></tr>\n",
       "\t<tr><td>ENSG00000003137</td><td>1</td><td> 21</td><td>0.595</td><td> 0.001</td><td>0.0014258335</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000003147</td><td>1</td><td>  8</td><td>0.901</td><td>-0.003</td><td>0.0015701290</td><td>-0.003</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000003249</td><td>2</td><td> 17</td><td>0.688</td><td> 0.002</td><td>0.0020332468</td><td> 0.002</td><td>0.001</td><td>0.096</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000003393</td><td>1</td><td>  5</td><td>0.901</td><td> 0.002</td><td>0.0011071698</td><td> 0.002</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000003402</td><td>1</td><td>  1</td><td>0.998</td><td> 0.002</td><td>0.0020686803</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000003509</td><td>1</td><td> 14</td><td>0.582</td><td>-0.001</td><td>0.0012611594</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000003987</td><td>1</td><td>  5</td><td>0.828</td><td> 0.001</td><td>0.0044738629</td><td> 0.001</td><td>0.004</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000003989</td><td>1</td><td>  3</td><td>0.985</td><td> 0.000</td><td>0.0007597256</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000004142</td><td>1</td><td>  3</td><td>0.985</td><td> 0.000</td><td>0.0005755454</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000004399</td><td>3</td><td>123</td><td>0.624</td><td> 0.003</td><td>0.0014414321</td><td> 0.003</td><td>0.001</td><td>0.489</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000004455</td><td>1</td><td> 72</td><td>0.755</td><td> 0.002</td><td>0.0015330947</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000004487</td><td>1</td><td> 12</td><td>0.550</td><td>-0.001</td><td>0.0013412799</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000004534</td><td>1</td><td> 28</td><td>0.871</td><td> 0.000</td><td>0.0003097528</td><td> 0.000</td><td>0.000</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000004700</td><td>1</td><td> 16</td><td>0.737</td><td> 0.000</td><td>0.0025675503</td><td> 0.000</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000004766</td><td>1</td><td>  1</td><td>0.957</td><td> 0.001</td><td>0.0035742160</td><td> 0.001</td><td>0.004</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005020</td><td>1</td><td>107</td><td>0.628</td><td> 0.000</td><td>0.0008228963</td><td> 0.000</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005059</td><td>2</td><td> 48</td><td>0.684</td><td>-0.001</td><td>0.0018060197</td><td> 0.000</td><td>0.002</td><td>2.463</td><td>0.594</td></tr>\n",
       "\t<tr><td>ENSG00000005075</td><td>1</td><td> 11</td><td>0.550</td><td> 0.002</td><td>0.0019552991</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005108</td><td>2</td><td> 36</td><td>0.604</td><td> 0.002</td><td>0.0013832122</td><td> 0.001</td><td>0.001</td><td>0.217</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005156</td><td>2</td><td> 30</td><td>0.890</td><td> 0.000</td><td>0.0016157681</td><td> 0.000</td><td>0.001</td><td>0.002</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005189</td><td>1</td><td> 88</td><td>0.783</td><td> 0.001</td><td>0.0026267194</td><td> 0.001</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005206</td><td>1</td><td>  7</td><td>0.949</td><td> 0.000</td><td>0.0015403997</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005249</td><td>1</td><td> 68</td><td>0.898</td><td> 0.000</td><td>0.0032628529</td><td> 0.000</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005436</td><td>1</td><td>  2</td><td>0.986</td><td> 0.001</td><td>0.0016112891</td><td> 0.001</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005448</td><td>1</td><td>  2</td><td>0.770</td><td> 0.007</td><td>0.0033168502</td><td> 0.007</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005471</td><td>1</td><td>  2</td><td>1.000</td><td> 0.000</td><td>0.0015087191</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005700</td><td>1</td><td> 19</td><td>0.537</td><td>-0.001</td><td>0.0029638984</td><td>-0.001</td><td>0.003</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000005981</td><td>2</td><td> 21</td><td>0.848</td><td> 0.001</td><td>0.0014226597</td><td> 0.001</td><td>0.001</td><td>0.598</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000006016</td><td>1</td><td>  2</td><td>0.832</td><td> 0.001</td><td>0.0008338203</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000006042</td><td>2</td><td> 15</td><td>0.730</td><td> 0.000</td><td>0.0009331976</td><td> 0.000</td><td>0.001</td><td>0.080</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000006116</td><td>1</td><td>  5</td><td>0.919</td><td>-0.004</td><td>0.0018333447</td><td>-0.004</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000006125</td><td>1</td><td> 79</td><td>0.788</td><td>-0.001</td><td>0.0014991831</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000006194</td><td>1</td><td> 23</td><td>0.605</td><td> 0.001</td><td>0.0013802571</td><td> 0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000006282</td><td>2</td><td>  6</td><td>1.000</td><td>-0.001</td><td>0.0005764921</td><td> 0.000</td><td>0.000</td><td>2.708</td><td>0.631</td></tr>\n",
       "\t<tr><td>ENSG00000006283</td><td>1</td><td>  7</td><td>0.853</td><td> 0.002</td><td>0.0016798225</td><td> 0.002</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000006453</td><td>1</td><td> 76</td><td>0.585</td><td>-0.001</td><td>0.0010935407</td><td>-0.001</td><td>0.001</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000006459</td><td>1</td><td> 50</td><td>0.704</td><td> 0.000</td><td>0.0018713348</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>\n",
       "\t<tr><td>ENSG00000006555</td><td>1</td><td> 26</td><td>0.776</td><td> 0.000</td><td>0.0024193024</td><td> 0.000</td><td>0.002</td><td>0.000</td><td>0.000</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A tibble: 51 × 10\n",
       "\\begin{tabular}{llllllllll}\n",
       " gene\\_id & num\\_cluster & num\\_instruments & spip & grp\\_beta & grp\\_se & meta & se\\_meta & Q & I2\\\\\n",
       " <chr> & <int> & <int> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\t ENSG00000000457 & 2 &  74 & 0.696 &  0.000 & 0.0016118763 & -0.001 & 0.001 & 0.157 & 0.000\\\\\n",
       "\t ENSG00000000971 & 1 &  22 & 0.739 & -0.001 & 0.0010414938 & -0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000001084 & 1 &  10 & 0.881 &  0.000 & 0.0010095378 &  0.000 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000001167 & 1 &  16 & 0.737 &  0.009 & 0.0029981468 &  0.009 & 0.003 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000001460 & 1 &  16 & 0.745 & -0.001 & 0.0006707913 & -0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000001461 & 1 &  47 & 0.679 & -0.005 & 0.0028313411 & -0.005 & 0.003 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000001561 & 1 &   9 & 0.586 &  0.001 & 0.0006746664 &  0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000001626 & 1 &  28 & 0.801 &  0.001 & 0.0011200443 &  0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000001629 & 2 &  48 & 0.678 &  0.000 & 0.0019744363 &  0.000 & 0.001 & 0.046 & 0.000\\\\\n",
       "\t ENSG00000001630 & 1 &  67 & 0.807 &  0.001 & 0.0011870369 &  0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000002822 & 2 &  19 & 0.926 & -0.001 & 0.0030673516 &  0.001 & 0.002 & 0.897 & 0.000\\\\\n",
       "\t ENSG00000002834 & 1 &   1 & 0.999 & -0.001 & 0.0012142297 & -0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000002933 & 2 &  40 & 0.985 & -0.001 & 0.0013062166 &  0.001 & 0.001 & 1.257 & 0.205\\\\\n",
       "\t ENSG00000003137 & 1 &  21 & 0.595 &  0.001 & 0.0014258335 &  0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000003147 & 1 &   8 & 0.901 & -0.003 & 0.0015701290 & -0.003 & 0.002 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000003249 & 2 &  17 & 0.688 &  0.002 & 0.0020332468 &  0.002 & 0.001 & 0.096 & 0.000\\\\\n",
       "\t ENSG00000003393 & 1 &   5 & 0.901 &  0.002 & 0.0011071698 &  0.002 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000003402 & 1 &   1 & 0.998 &  0.002 & 0.0020686803 &  0.002 & 0.002 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000003509 & 1 &  14 & 0.582 & -0.001 & 0.0012611594 & -0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000003987 & 1 &   5 & 0.828 &  0.001 & 0.0044738629 &  0.001 & 0.004 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000003989 & 1 &   3 & 0.985 &  0.000 & 0.0007597256 &  0.000 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000004142 & 1 &   3 & 0.985 &  0.000 & 0.0005755454 &  0.000 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000004399 & 3 & 123 & 0.624 &  0.003 & 0.0014414321 &  0.003 & 0.001 & 0.489 & 0.000\\\\\n",
       "\t ENSG00000004455 & 1 &  72 & 0.755 &  0.002 & 0.0015330947 &  0.002 & 0.002 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000004487 & 1 &  12 & 0.550 & -0.001 & 0.0013412799 & -0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000004534 & 1 &  28 & 0.871 &  0.000 & 0.0003097528 &  0.000 & 0.000 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000004700 & 1 &  16 & 0.737 &  0.000 & 0.0025675503 &  0.000 & 0.003 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000004766 & 1 &   1 & 0.957 &  0.001 & 0.0035742160 &  0.001 & 0.004 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000005020 & 1 & 107 & 0.628 &  0.000 & 0.0008228963 &  0.000 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000005059 & 2 &  48 & 0.684 & -0.001 & 0.0018060197 &  0.000 & 0.002 & 2.463 & 0.594\\\\\n",
       "\t ENSG00000005075 & 1 &  11 & 0.550 &  0.002 & 0.0019552991 &  0.002 & 0.002 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000005108 & 2 &  36 & 0.604 &  0.002 & 0.0013832122 &  0.001 & 0.001 & 0.217 & 0.000\\\\\n",
       "\t ENSG00000005156 & 2 &  30 & 0.890 &  0.000 & 0.0016157681 &  0.000 & 0.001 & 0.002 & 0.000\\\\\n",
       "\t ENSG00000005189 & 1 &  88 & 0.783 &  0.001 & 0.0026267194 &  0.001 & 0.003 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000005206 & 1 &   7 & 0.949 &  0.000 & 0.0015403997 &  0.000 & 0.002 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000005249 & 1 &  68 & 0.898 &  0.000 & 0.0032628529 &  0.000 & 0.003 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000005436 & 1 &   2 & 0.986 &  0.001 & 0.0016112891 &  0.001 & 0.002 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000005448 & 1 &   2 & 0.770 &  0.007 & 0.0033168502 &  0.007 & 0.003 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000005471 & 1 &   2 & 1.000 &  0.000 & 0.0015087191 &  0.000 & 0.002 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000005700 & 1 &  19 & 0.537 & -0.001 & 0.0029638984 & -0.001 & 0.003 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000005981 & 2 &  21 & 0.848 &  0.001 & 0.0014226597 &  0.001 & 0.001 & 0.598 & 0.000\\\\\n",
       "\t ENSG00000006016 & 1 &   2 & 0.832 &  0.001 & 0.0008338203 &  0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000006042 & 2 &  15 & 0.730 &  0.000 & 0.0009331976 &  0.000 & 0.001 & 0.080 & 0.000\\\\\n",
       "\t ENSG00000006116 & 1 &   5 & 0.919 & -0.004 & 0.0018333447 & -0.004 & 0.002 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000006125 & 1 &  79 & 0.788 & -0.001 & 0.0014991831 & -0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000006194 & 1 &  23 & 0.605 &  0.001 & 0.0013802571 &  0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000006282 & 2 &   6 & 1.000 & -0.001 & 0.0005764921 &  0.000 & 0.000 & 2.708 & 0.631\\\\\n",
       "\t ENSG00000006283 & 1 &   7 & 0.853 &  0.002 & 0.0016798225 &  0.002 & 0.002 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000006453 & 1 &  76 & 0.585 & -0.001 & 0.0010935407 & -0.001 & 0.001 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000006459 & 1 &  50 & 0.704 &  0.000 & 0.0018713348 &  0.000 & 0.002 & 0.000 & 0.000\\\\\n",
       "\t ENSG00000006555 & 1 &  26 & 0.776 &  0.000 & 0.0024193024 &  0.000 & 0.002 & 0.000 & 0.000\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A tibble: 51 × 10\n",
       "\n",
       "| gene_id &lt;chr&gt; | num_cluster &lt;int&gt; | num_instruments &lt;int&gt; | spip &lt;dbl&gt; | grp_beta &lt;dbl&gt; | grp_se &lt;dbl&gt; | meta &lt;dbl&gt; | se_meta &lt;dbl&gt; | Q &lt;dbl&gt; | I2 &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|---|\n",
       "| ENSG00000000457 | 2 |  74 | 0.696 |  0.000 | 0.0016118763 | -0.001 | 0.001 | 0.157 | 0.000 |\n",
       "| ENSG00000000971 | 1 |  22 | 0.739 | -0.001 | 0.0010414938 | -0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000001084 | 1 |  10 | 0.881 |  0.000 | 0.0010095378 |  0.000 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000001167 | 1 |  16 | 0.737 |  0.009 | 0.0029981468 |  0.009 | 0.003 | 0.000 | 0.000 |\n",
       "| ENSG00000001460 | 1 |  16 | 0.745 | -0.001 | 0.0006707913 | -0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000001461 | 1 |  47 | 0.679 | -0.005 | 0.0028313411 | -0.005 | 0.003 | 0.000 | 0.000 |\n",
       "| ENSG00000001561 | 1 |   9 | 0.586 |  0.001 | 0.0006746664 |  0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000001626 | 1 |  28 | 0.801 |  0.001 | 0.0011200443 |  0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000001629 | 2 |  48 | 0.678 |  0.000 | 0.0019744363 |  0.000 | 0.001 | 0.046 | 0.000 |\n",
       "| ENSG00000001630 | 1 |  67 | 0.807 |  0.001 | 0.0011870369 |  0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000002822 | 2 |  19 | 0.926 | -0.001 | 0.0030673516 |  0.001 | 0.002 | 0.897 | 0.000 |\n",
       "| ENSG00000002834 | 1 |   1 | 0.999 | -0.001 | 0.0012142297 | -0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000002933 | 2 |  40 | 0.985 | -0.001 | 0.0013062166 |  0.001 | 0.001 | 1.257 | 0.205 |\n",
       "| ENSG00000003137 | 1 |  21 | 0.595 |  0.001 | 0.0014258335 |  0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000003147 | 1 |   8 | 0.901 | -0.003 | 0.0015701290 | -0.003 | 0.002 | 0.000 | 0.000 |\n",
       "| ENSG00000003249 | 2 |  17 | 0.688 |  0.002 | 0.0020332468 |  0.002 | 0.001 | 0.096 | 0.000 |\n",
       "| ENSG00000003393 | 1 |   5 | 0.901 |  0.002 | 0.0011071698 |  0.002 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000003402 | 1 |   1 | 0.998 |  0.002 | 0.0020686803 |  0.002 | 0.002 | 0.000 | 0.000 |\n",
       "| ENSG00000003509 | 1 |  14 | 0.582 | -0.001 | 0.0012611594 | -0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000003987 | 1 |   5 | 0.828 |  0.001 | 0.0044738629 |  0.001 | 0.004 | 0.000 | 0.000 |\n",
       "| ENSG00000003989 | 1 |   3 | 0.985 |  0.000 | 0.0007597256 |  0.000 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000004142 | 1 |   3 | 0.985 |  0.000 | 0.0005755454 |  0.000 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000004399 | 3 | 123 | 0.624 |  0.003 | 0.0014414321 |  0.003 | 0.001 | 0.489 | 0.000 |\n",
       "| ENSG00000004455 | 1 |  72 | 0.755 |  0.002 | 0.0015330947 |  0.002 | 0.002 | 0.000 | 0.000 |\n",
       "| ENSG00000004487 | 1 |  12 | 0.550 | -0.001 | 0.0013412799 | -0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000004534 | 1 |  28 | 0.871 |  0.000 | 0.0003097528 |  0.000 | 0.000 | 0.000 | 0.000 |\n",
       "| ENSG00000004700 | 1 |  16 | 0.737 |  0.000 | 0.0025675503 |  0.000 | 0.003 | 0.000 | 0.000 |\n",
       "| ENSG00000004766 | 1 |   1 | 0.957 |  0.001 | 0.0035742160 |  0.001 | 0.004 | 0.000 | 0.000 |\n",
       "| ENSG00000005020 | 1 | 107 | 0.628 |  0.000 | 0.0008228963 |  0.000 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000005059 | 2 |  48 | 0.684 | -0.001 | 0.0018060197 |  0.000 | 0.002 | 2.463 | 0.594 |\n",
       "| ENSG00000005075 | 1 |  11 | 0.550 |  0.002 | 0.0019552991 |  0.002 | 0.002 | 0.000 | 0.000 |\n",
       "| ENSG00000005108 | 2 |  36 | 0.604 |  0.002 | 0.0013832122 |  0.001 | 0.001 | 0.217 | 0.000 |\n",
       "| ENSG00000005156 | 2 |  30 | 0.890 |  0.000 | 0.0016157681 |  0.000 | 0.001 | 0.002 | 0.000 |\n",
       "| ENSG00000005189 | 1 |  88 | 0.783 |  0.001 | 0.0026267194 |  0.001 | 0.003 | 0.000 | 0.000 |\n",
       "| ENSG00000005206 | 1 |   7 | 0.949 |  0.000 | 0.0015403997 |  0.000 | 0.002 | 0.000 | 0.000 |\n",
       "| ENSG00000005249 | 1 |  68 | 0.898 |  0.000 | 0.0032628529 |  0.000 | 0.003 | 0.000 | 0.000 |\n",
       "| ENSG00000005436 | 1 |   2 | 0.986 |  0.001 | 0.0016112891 |  0.001 | 0.002 | 0.000 | 0.000 |\n",
       "| ENSG00000005448 | 1 |   2 | 0.770 |  0.007 | 0.0033168502 |  0.007 | 0.003 | 0.000 | 0.000 |\n",
       "| ENSG00000005471 | 1 |   2 | 1.000 |  0.000 | 0.0015087191 |  0.000 | 0.002 | 0.000 | 0.000 |\n",
       "| ENSG00000005700 | 1 |  19 | 0.537 | -0.001 | 0.0029638984 | -0.001 | 0.003 | 0.000 | 0.000 |\n",
       "| ENSG00000005981 | 2 |  21 | 0.848 |  0.001 | 0.0014226597 |  0.001 | 0.001 | 0.598 | 0.000 |\n",
       "| ENSG00000006016 | 1 |   2 | 0.832 |  0.001 | 0.0008338203 |  0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000006042 | 2 |  15 | 0.730 |  0.000 | 0.0009331976 |  0.000 | 0.001 | 0.080 | 0.000 |\n",
       "| ENSG00000006116 | 1 |   5 | 0.919 | -0.004 | 0.0018333447 | -0.004 | 0.002 | 0.000 | 0.000 |\n",
       "| ENSG00000006125 | 1 |  79 | 0.788 | -0.001 | 0.0014991831 | -0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000006194 | 1 |  23 | 0.605 |  0.001 | 0.0013802571 |  0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000006282 | 2 |   6 | 1.000 | -0.001 | 0.0005764921 |  0.000 | 0.000 | 2.708 | 0.631 |\n",
       "| ENSG00000006283 | 1 |   7 | 0.853 |  0.002 | 0.0016798225 |  0.002 | 0.002 | 0.000 | 0.000 |\n",
       "| ENSG00000006453 | 1 |  76 | 0.585 | -0.001 | 0.0010935407 | -0.001 | 0.001 | 0.000 | 0.000 |\n",
       "| ENSG00000006459 | 1 |  50 | 0.704 |  0.000 | 0.0018713348 |  0.000 | 0.002 | 0.000 | 0.000 |\n",
       "| ENSG00000006555 | 1 |  26 | 0.776 |  0.000 | 0.0024193024 |  0.000 | 0.002 | 0.000 | 0.000 |\n",
       "\n"
      ],
      "text/plain": [
       "   gene_id         num_cluster num_instruments spip  grp_beta grp_se      \n",
       "1  ENSG00000000457 2            74             0.696  0.000   0.0016118763\n",
       "2  ENSG00000000971 1            22             0.739 -0.001   0.0010414938\n",
       "3  ENSG00000001084 1            10             0.881  0.000   0.0010095378\n",
       "4  ENSG00000001167 1            16             0.737  0.009   0.0029981468\n",
       "5  ENSG00000001460 1            16             0.745 -0.001   0.0006707913\n",
       "6  ENSG00000001461 1            47             0.679 -0.005   0.0028313411\n",
       "7  ENSG00000001561 1             9             0.586  0.001   0.0006746664\n",
       "8  ENSG00000001626 1            28             0.801  0.001   0.0011200443\n",
       "9  ENSG00000001629 2            48             0.678  0.000   0.0019744363\n",
       "10 ENSG00000001630 1            67             0.807  0.001   0.0011870369\n",
       "11 ENSG00000002822 2            19             0.926 -0.001   0.0030673516\n",
       "12 ENSG00000002834 1             1             0.999 -0.001   0.0012142297\n",
       "13 ENSG00000002933 2            40             0.985 -0.001   0.0013062166\n",
       "14 ENSG00000003137 1            21             0.595  0.001   0.0014258335\n",
       "15 ENSG00000003147 1             8             0.901 -0.003   0.0015701290\n",
       "16 ENSG00000003249 2            17             0.688  0.002   0.0020332468\n",
       "17 ENSG00000003393 1             5             0.901  0.002   0.0011071698\n",
       "18 ENSG00000003402 1             1             0.998  0.002   0.0020686803\n",
       "19 ENSG00000003509 1            14             0.582 -0.001   0.0012611594\n",
       "20 ENSG00000003987 1             5             0.828  0.001   0.0044738629\n",
       "21 ENSG00000003989 1             3             0.985  0.000   0.0007597256\n",
       "22 ENSG00000004142 1             3             0.985  0.000   0.0005755454\n",
       "23 ENSG00000004399 3           123             0.624  0.003   0.0014414321\n",
       "24 ENSG00000004455 1            72             0.755  0.002   0.0015330947\n",
       "25 ENSG00000004487 1            12             0.550 -0.001   0.0013412799\n",
       "26 ENSG00000004534 1            28             0.871  0.000   0.0003097528\n",
       "27 ENSG00000004700 1            16             0.737  0.000   0.0025675503\n",
       "28 ENSG00000004766 1             1             0.957  0.001   0.0035742160\n",
       "29 ENSG00000005020 1           107             0.628  0.000   0.0008228963\n",
       "30 ENSG00000005059 2            48             0.684 -0.001   0.0018060197\n",
       "31 ENSG00000005075 1            11             0.550  0.002   0.0019552991\n",
       "32 ENSG00000005108 2            36             0.604  0.002   0.0013832122\n",
       "33 ENSG00000005156 2            30             0.890  0.000   0.0016157681\n",
       "34 ENSG00000005189 1            88             0.783  0.001   0.0026267194\n",
       "35 ENSG00000005206 1             7             0.949  0.000   0.0015403997\n",
       "36 ENSG00000005249 1            68             0.898  0.000   0.0032628529\n",
       "37 ENSG00000005436 1             2             0.986  0.001   0.0016112891\n",
       "38 ENSG00000005448 1             2             0.770  0.007   0.0033168502\n",
       "39 ENSG00000005471 1             2             1.000  0.000   0.0015087191\n",
       "40 ENSG00000005700 1            19             0.537 -0.001   0.0029638984\n",
       "41 ENSG00000005981 2            21             0.848  0.001   0.0014226597\n",
       "42 ENSG00000006016 1             2             0.832  0.001   0.0008338203\n",
       "43 ENSG00000006042 2            15             0.730  0.000   0.0009331976\n",
       "44 ENSG00000006116 1             5             0.919 -0.004   0.0018333447\n",
       "45 ENSG00000006125 1            79             0.788 -0.001   0.0014991831\n",
       "46 ENSG00000006194 1            23             0.605  0.001   0.0013802571\n",
       "47 ENSG00000006282 2             6             1.000 -0.001   0.0005764921\n",
       "48 ENSG00000006283 1             7             0.853  0.002   0.0016798225\n",
       "49 ENSG00000006453 1            76             0.585 -0.001   0.0010935407\n",
       "50 ENSG00000006459 1            50             0.704  0.000   0.0018713348\n",
       "51 ENSG00000006555 1            26             0.776  0.000   0.0024193024\n",
       "   meta   se_meta Q     I2   \n",
       "1  -0.001 0.001   0.157 0.000\n",
       "2  -0.001 0.001   0.000 0.000\n",
       "3   0.000 0.001   0.000 0.000\n",
       "4   0.009 0.003   0.000 0.000\n",
       "5  -0.001 0.001   0.000 0.000\n",
       "6  -0.005 0.003   0.000 0.000\n",
       "7   0.001 0.001   0.000 0.000\n",
       "8   0.001 0.001   0.000 0.000\n",
       "9   0.000 0.001   0.046 0.000\n",
       "10  0.001 0.001   0.000 0.000\n",
       "11  0.001 0.002   0.897 0.000\n",
       "12 -0.001 0.001   0.000 0.000\n",
       "13  0.001 0.001   1.257 0.205\n",
       "14  0.001 0.001   0.000 0.000\n",
       "15 -0.003 0.002   0.000 0.000\n",
       "16  0.002 0.001   0.096 0.000\n",
       "17  0.002 0.001   0.000 0.000\n",
       "18  0.002 0.002   0.000 0.000\n",
       "19 -0.001 0.001   0.000 0.000\n",
       "20  0.001 0.004   0.000 0.000\n",
       "21  0.000 0.001   0.000 0.000\n",
       "22  0.000 0.001   0.000 0.000\n",
       "23  0.003 0.001   0.489 0.000\n",
       "24  0.002 0.002   0.000 0.000\n",
       "25 -0.001 0.001   0.000 0.000\n",
       "26  0.000 0.000   0.000 0.000\n",
       "27  0.000 0.003   0.000 0.000\n",
       "28  0.001 0.004   0.000 0.000\n",
       "29  0.000 0.001   0.000 0.000\n",
       "30  0.000 0.002   2.463 0.594\n",
       "31  0.002 0.002   0.000 0.000\n",
       "32  0.001 0.001   0.217 0.000\n",
       "33  0.000 0.001   0.002 0.000\n",
       "34  0.001 0.003   0.000 0.000\n",
       "35  0.000 0.002   0.000 0.000\n",
       "36  0.000 0.003   0.000 0.000\n",
       "37  0.001 0.002   0.000 0.000\n",
       "38  0.007 0.003   0.000 0.000\n",
       "39  0.000 0.002   0.000 0.000\n",
       "40 -0.001 0.003   0.000 0.000\n",
       "41  0.001 0.001   0.598 0.000\n",
       "42  0.001 0.001   0.000 0.000\n",
       "43  0.000 0.001   0.080 0.000\n",
       "44 -0.004 0.002   0.000 0.000\n",
       "45 -0.001 0.001   0.000 0.000\n",
       "46  0.001 0.001   0.000 0.000\n",
       "47  0.000 0.000   2.708 0.631\n",
       "48  0.002 0.002   0.000 0.000\n",
       "49 -0.001 0.001   0.000 0.000\n",
       "50  0.000 0.002   0.000 0.000\n",
       "51  0.000 0.002   0.000 0.000"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "output"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
