{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# APEX QTL association testing\n",
    "This notebook implements a workflow for using the APEX to conduct analysis. APEX implements a linear mixed model for association testing. This is potentially useful for analysis with related individuals.\n",
    "\n",
    "**FIXME: cite APEX paper**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Input\n",
    "- `--molecular-pheno`, The `bed.gz` file containing the table describing the molecular phenotype. It should have a companion index file in `tbi` format.\n",
    "\n",
    "- `genotype_list` a list of whole genome vcf file for each chromosome, those vcf are converted beforehand from plink trio in the data_processing sections\n",
    "\n",
    "- `grm_list` is a file containing list of grm matrixs that generated by the GRM module of this pipeline.\n",
    "\n",
    "- `covariate` is a file with #id + samples name as colnames and each row a covariate: fixed and known covariates as well as hidden covariates recovered from factor analysis.\n",
    "\n",
    "\n",
    "## Output\n",
    "\n",
    "A sets of summary statistics files for each chromosome, for both nomial significance for each test, as well as region (gene) level association evidence."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Command interface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sos run APEX.ipynb [workflow_name | -t targets] [options] [workflow_options]\n",
      "  workflow_name:        Single or combined workflows defined in this script\n",
      "  targets:              One or more targets to generate\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\n",
      "\n",
      "Workflows:\n",
      "  LMM_null\n",
      "  APEX_cis\n",
      "  cis\n",
      "  APEX_trans\n",
      "  trans\n",
      "\n",
      "Global Workflow Options:\n",
      "  --molecular-pheno-list VAL (as path, required)\n",
      "                        Path to the input molecular phenotype file, per chrm, in\n",
      "                        bed.gz format.\n",
      "  --covariate VAL (as path, required)\n",
      "                        Covariate file, in similar format as the molecular_pheno\n",
      "  --genotype-file-list VAL (as path, required)\n",
      "                        Genotype file in vcf format, per chrm\n",
      "  --grm-list VAL (as path, required)\n",
      "                        grm file in pairwise table format , per chrm\n",
      "  --cwd . (as path)\n",
      "                        Path to the work directory of the analysis.\n",
      "  --job-size 2 (as int)\n",
      "                        Specify the number of jobs per run.\n",
      "  --container ''\n",
      "                        Container option for software to run the analysis:\n",
      "                        docker or singularity\n",
      "  --name ROSMAP\n",
      "                        Prefix for the analysis output\n",
      "  --window 1000000 (as list)\n",
      "                        Specify the scanning window for the up and downstream\n",
      "                        radius to analyze around the region of interest, in\n",
      "                        units of bp\n",
      "\n",
      "Sections\n",
      "  APEX_*_1, LMM_null:\n",
      "  APEX_cis_2, cis:\n",
      "  APEX_trans_2, trans:\n",
      "    Workflow Options:\n",
      "      --keep-gene ''\n",
      "  APEX_*_3:\n",
      "  APEX_*_4:\n"
     ]
    }
   ],
   "source": [
    "sos run APEX.ipynb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Example\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run /home/hs3163/GIT/xqtl-pipeline/pipeline/APEX.ipynb APEX_cis   \\\n",
    "--genotype_file_list GRCh38_liftedover_sorted_all.add_chr.leftnorm.filtered.renamed.filtered.renamed.filtered.filtered.vcf_files_list.txt \\\n",
    "--molecular_pheno_list /mnt/mfs/statgen/snuc_pseudo_bulk/eight_tissue_analysis/data_preprocessing/ALL/phenotype_data/ALL.log2cpm.bed.processed_phenotype.per_chrom.recipe   \\\n",
    "--grm_list  data_preprocessing/genotype/grm/plink_files_list.loco_grm_list.txt  \\\n",
    "--covariate data_preprocessing/ALL/covariates/ALL.covariate.pca.BiCV.cov.gz     \\\n",
    "--cwd /mnt/mfs/statgen/snuc_pseudo_bulk/eight_tissue_analysis/Association_scan/ALL/APEX    \\\n",
    "--container /mnt/mfs/statgen/containers/apex.sif -J 50 -c csg.yml -q csg -n --no-LMM &"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Global parameter settings\n",
    "The section outlined the parameters that can be set in the command interface."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "**FIXME: please revisit this input list for various default parameter setting, in particular the list of genotypes to load. You can copy and paste the genotype load code from genotype_reformatting for now. I'll repurpose the LDtools package to include those codes in the future so we dont have to copy them over and over again.**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "**FIXME: can we combine APEX_3 and APEX_4**?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# Path to the input molecular phenotype file, per chrm, in bed.gz format.\n",
    "parameter: molecular_pheno_list = path\n",
    "# Covariate file, in similar format as the molecular_pheno\n",
    "parameter: covariate = path\n",
    "# Genotype file in vcf format, per chrm\n",
    "parameter: genotype_file_list = path\n",
    "# grm file in pairwise table format , per chrm\n",
    "parameter: grm_list = path(\"./\")\n",
    "# Path to the work directory of the analysis.\n",
    "parameter: cwd = path('./')\n",
    "# Specify the number of jobs per run.\n",
    "parameter: job_size = 2\n",
    "# Container option for software to run the analysis: docker or singularity\n",
    "parameter: container = ''\n",
    "# Prefix for the analysis output\n",
    "parameter: name = 'ROSMAP'\n",
    "# Specify the scanning window for the up and downstream radius to analyze around the region of interest, in units of bp\n",
    "parameter: window = 1000000\n",
    "# Number of threads\n",
    "parameter: numThreads = 1\n",
    "# For cluster jobs, number commands to run per job\n",
    "parameter: job_size = 1\n",
    "# Walltime \n",
    "parameter: walltime = '5h'\n",
    "parameter: mem = '80G'\n",
    "\n",
    "\n",
    "\n",
    "import pandas as pd\n",
    "molecular_pheno_chr_inv = pd.read_csv(molecular_pheno_list,sep = \"\\t\")\n",
    "geno_chr_inv = pd.read_csv(genotype_file_list,sep = \"\\t\")\n",
    "input_inv = molecular_pheno_chr_inv.merge(geno_chr_inv, on = \"#id\")\n",
    "parameter: LMM = False\n",
    "if LMM:\n",
    "    grm_chr_inv =  pd.read_csv(grm_list,sep = \"\\t\")\n",
    "    input_inv = input_inv.merge(grm_chr_inv,on = \"#id\")\n",
    "input_inv = input_inv.values.tolist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## LMM Regression  \n",
    "This step are done to precompute and store:\n",
    "1. LMM null models and trait residuals\n",
    "2. spline terms for LMM genotypic variances to speed up downstream analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[LMM_null]\n",
    "input: for_each = \"input_inv\"\n",
    "output: f'{cwd:a}/{path(_input_inv[1]):bnn}.theta.gz'\n",
    "task: trunk_workers = 1, trunk_size=job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "bash: expand= \"$[ ]\", stderr = f'{_output}.stderr', stdout = f'{_output}.stdout', container = container\n",
    "    apex lmm --vcf $[_input_inv[2]] \\\n",
    "    --bed $[_input_inv[1]] \\\n",
    "    --cov $[covariate] \\\n",
    "    --out $[_output[0]:nn] \\\n",
    "    --grm $[_input_inv[3]] \\\n",
    "    --get-theta "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## QTL asscoiation testing \n",
    "This step generate the cis-QTL and trans-QTL summary statistics and for downstream analysis from summary statistics. The analysis is done per chromosome to reduce running time."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[APEX_cis_1, cis]\n",
    "if LMM:\n",
    "    sos_run(\"LMM_null\")\n",
    "input: for_each = \"input_inv\"\n",
    "output:f'{cwd:a}/{path(_input_inv[1]):bnn}{\".LMM\" if LMM else \".OLS\"}.cis_long_table.txt.gz',\n",
    "       f'{cwd:a}/{path(_input_inv[1]):bnn}{\".LMM\" if LMM else \".OLS\"}.cis_gene_table.txt.gz',\n",
    "       f'{cwd:a}/{path(_input_inv[1]):bnn}{\".LMM\" if LMM else \".OLS\"}.cis_sumstats.txt.gz'\n",
    "parameter: theta = path(f'{cwd:a}/{path(_input_inv[1]):bnn}.theta.gz')\n",
    "task: trunk_workers = 1, trunk_size=job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "bash: expand= \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout', container = container\n",
    "    apex cis --vcf $[_input_inv[2]] \\\n",
    "    --bed $[_input_inv[1]] \\\n",
    "    --cov $[covariate] \\\n",
    "    --out $[_output[0]:nnn] \\\n",
    "    --long $[f'--theta {theta} --grm {_input_inv[3]}'  if LMM else \"\"] \\\n",
    "    --window $[window] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[APEX_trans_1, trans]\n",
    "if LMM:\n",
    "    sos_run(\"LMM_null\")\n",
    "parameter: keep_gene = \"\"\n",
    "input: for_each = \"input_inv\"\n",
    "output: f'{cwd:a}/{path(_input_inv[1]):bnn}{\".LMM\" if LMM else \".OLS\"}.trans_long_table.txt.gz',\n",
    "        f'{cwd:a}/{path(_input_inv[1]):bnn}{\".LMM\" if LMM else \".OLS\"}.trans_gene_table.txt.gz'\n",
    "parameter: theta = path(f'{cwd:a}/{path(_input_inv[1]):bnn}.theta.gz')\n",
    "task: trunk_workers = 1, trunk_size=job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "bash: expand= \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout', container = container\n",
    "    apex trans --vcf $[_input_inv[2]] \\\n",
    "    --bed $[_input_inv[1]] \\\n",
    "    --cov $[covariate] \\\n",
    "    --out $[_output[0]:nnn] \\\n",
    "    --gene \"$[keep_gene]\" \\\n",
    "    --long $[f'--theta {theta} --grm {_input_inv[3]}'  if LMM else \"\"] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[APEX_*_2]\n",
    "output: f'{_input[0]:nn}.reformated.txt'\n",
    "task: trunk_workers = 1, trunk_size=job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "R: expand = \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout',container = container\n",
    "    library(\"dplyr\")\n",
    "    library(\"tibble\")\n",
    "    library(\"readr\")\n",
    "    library(\"purrr\")\n",
    "    data <- read_delim(\"$[_input[0]]\",delim = \"\\t\")\n",
    "    data = data%>%mutate(variant_id = pmap_chr(list(a = `#chrom`,b = pos,c = ref, d = alt),function(a,b,c,d) paste(c(a,\":\",b,\"_\",c,\"_\",d),collapse = \"\")))%>%rename(chrom = `#chrom`)\n",
    "    data %>%write_delim(\"$[_output]\",delim = \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[APEX_*_3]\n",
    "input: group_by = \"all\"\n",
    "output: f'{cwd}/APEX_QTL_recipe.tsv',f'{cwd:a}/APEX_column_info.txt'\n",
    "python: expand = \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout'\n",
    "    import pandas as pd \n",
    "    data_tempt = pd.DataFrame({\n",
    "    \"#id\" : [int(x.split(\".\")[-6].replace(\"chr\",\"\")) for x in  [$[_input:r,]]],\n",
    "    \"sumstat_dir\" : [$[_input:r,]],\\\n",
    "    \"column_info\" : $[_output[1]:r]\n",
    "    })\n",
    "    column_info_df = pd.DataFrame( pd.Series( {\"ID\": \"GENE,CHR,POS,A0,A1\",\n",
    "          \"CHR\": \"chrom\",\n",
    "          \"POS\": \"pos\",\n",
    "          \"A0\": \"ref\",\n",
    "          \"A1\": \"alt\",\n",
    "          \"SNP\": \"variant_id\",\n",
    "          \"STAT\": \"beta\",\n",
    "          \"SE\": \"se\",\n",
    "          \"P\": \"pval\",\n",
    "          \"GENE\": \"gene\"}), columns = [\"APEX\"] )\n",
    "    data_tempt.to_csv(\"$[_output[0]]\",index = False,sep = \"\\t\" )\n",
    "    column_info_df.to_csv(\"$[_output[1]]\",index = True,sep = \"\\t\" )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "Bash",
     "#E6EEFF",
     ""
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.22.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
