{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Fine-mapping with SuSiE model\n",
    "This notebook conduct fine_mapping with complete data, unlike the susie_RSS module, this module perform analysis on 1 of the theme each time **FIXME: why saying this \\\"unlike the susie_RSS module, this module perform analysis on 1 of the theme each time\\\"? we should unify the interfaces if we can (let me try to see why we cannot do that)**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "sos run pipeline/SuSiE.ipynb uni_susie \\\n",
    "    --genoFile <(head GRCh38_plink_files_list.txt)  \\\n",
    "    --region-list ALL.log2cpm.region_list \\\n",
    "    --phenoFile ALL_pseudobulk_phenotype_list.txt \\\n",
    "    --covFile ALL_pseudobulk_phenotype_list.txt -J 800 -c csg.yml -q csg  &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run pipeline/SuSiE.ipynb uni_susie \\\n",
    "    --genoFile <(cat GRCh38_plink_files_list.txt | grep ENSG00000197943  )\\\n",
    "    --region-list ALL.log2cpm.region_list \\\n",
    "    --phenoFile ALL_pseudobulk_phenotype_list.txt \\\n",
    "    --covFile ALL_pseudobulk_covariate_list.txt --cwd ALL_PLCG2 --name ALL_pseudo_bulk  & "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run pipeline/per_gene_data_merger.ipynb data_merger \\\n",
    "    --genoFile <(cat GRCh38_plink_files_list.txt | grep ENSG00000197943  )\\\n",
    "    --region-list ALL.log2cpm.region_list \\\n",
    "    --phenoFile ALL_pseudobulk_phenotype_list.txt \\\n",
    "    --covFile ALL_pseudobulk_covariate_list.txt --cwd ALL_PLCG2 --name ALL_pseudo_bulk --cwd data_merger_test  & "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "## Input\n",
    "\n",
    "1. 1 region list documenting the regions to be analysised\n",
    "2. a list to the path where phenotype per gene data stored\n",
    "3. a list to the path where genotype per gene data stored\n",
    "\n",
    "By default the input 2 and 3 is the output from the data_preprocessing module.\n",
    "\n",
    "### Example genotype list\n",
    "\n",
    "region        dir\n",
    "ENSG00000000457 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000000457.bed\n",
    "ENSG00000000460 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000000460.bed\n",
    "ENSG00000000938 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000000938.bed\n",
    "ENSG00000000971 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000000971.bed\n",
    "ENSG00000001036 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000001036.bed\n",
    "ENSG00000001084 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000001084.bed\n",
    "ENSG00000001167 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000001167.bed\n",
    "ENSG00000001460 /mnt/mfs/statgen/xqtl_workflow_testing/demo/genotype_reformmating/demo_per_gene_plink/ENSG00000001460.bed\n",
    "\n",
    "\n",
    "### Example phenotype list\n",
    "\n",
    "\n",
    "region dir\n",
    "ENSG00000000457 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000000457.mol_phe.bed.gz\n",
    "ENSG00000000460 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000000460.mol_phe.bed.gz\n",
    "ENSG00000000938 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000000938.mol_phe.bed.gz\n",
    "ENSG00000000971 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000000971.mol_phe.bed.gz\n",
    "ENSG00000001036 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000001036.mol_phe.bed.gz\n",
    "ENSG00000001084 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000001084.mol_phe.bed.gz\n",
    "ENSG00000001167 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000001167.mol_phe.bed.gz\n",
    "ENSG00000001460 /mnt/mfs/statgen/xqtl_workflow_testing/demo/phenotype_reformat/demo.ENSG00000001460.mol_phe.bed.gz\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "## Output\n",
    "\n",
    "For each analysis unit we output:\n",
    "\n",
    "1. Analysis results in RDS format: A mvsusie Model\n",
    "2. A vcf file with selected snps\n",
    "        ES:PIP:CS"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "## MWE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run pipeline/SuSiE.ipynb uni_susie \\\n",
    "--genoFile /mnt/mfs/statgen/snuc_pseudo_bulk/Ast/genotype_per_gene/MWE.region_plink_files/plink_files_list.txt  \\\n",
    "--cwd MWE/susie_per_gene/ \\\n",
    "--region-list data/mwe/MWE.region.list \\\n",
    "--phenoFile data/mwe/MWE.phenotype.list \\\n",
    "--covFile data/mwe/MWE.covar.list -J 8 -c csg.yml -q csg  &\n",
    "\n",
    "sos run pipeline/SuSiE.ipynb mv_susie \\\n",
    "--genoFile ./mwe.region_plink_files/plink_files_list.txt \\\n",
    "--cwd MWE/rds_per_gene/ \\\n",
    "--region-list MWE.region.list \\\n",
    "--phenoFile MWE.phenotype.list \\\n",
    "--covFile MWE.covar.list &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "cd /mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl/\n",
    "sos run pipeline/SuSiE.ipynb uni_susie \\\n",
    "--genoFile /mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl/dlpfc_tad_gene_genotype_list  \\\n",
    "--cwd output/susie_per_gene_tad_2/ \\\n",
    "--region-list ./dlpfc_region_list \\\n",
    "--phenoFile phenotype_list \\\n",
    "--covFile covar_list -J 200 -c csg.yml -q csg2 -s build  --job_size 1 --mem 80G --numThreads 8 --container containers/stephenslab.sif --tad_list dlpfc_tad_region_list  &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## FSuSiE\n",
    "\n",
    "The interface of fsusie is identical as the uni_susie. However, instead of extracting 1 row out of each bed.gz file. It will extract all the rows which tabix believe they are contained by the region specified in the region list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "cd /mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl/\n",
    "sos run pipeline/SuSiE.ipynb uni_fsusie \\\n",
    "--genoFile /mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl/output/dlpfc_tad_region_list_plink_files/ROSMAP_NIA_WGS.leftnorm.filtered.filtered.plink_files_list.txt  \\\n",
    "--cwd output/f_susie_tad_eQTL/ --name eQTL \\\n",
    "--region-list ./dlpfc_tad_list \\\n",
    "--phenoFile phenotype_list \\\n",
    "--covFile covar_list  -s build"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run pipeline/SuSiE.ipynb uni_fsusie \\\n",
    "--genoFile /mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl/output/dlpfc_tad_region_list_plink_files/ROSMAP_NIA_WGS.leftnorm.filtered.filtered.plink_files_list.txt  \\\n",
    "--cwd output/f_susie_tad_haQTL_pos/ --name haQTL \\\n",
    "--region-list dlpfc_tad_list \\\n",
    "--phenoFile phenotype_list \\\n",
    "--covFile covar_list --prior mixture_normal_per_scale -J 200 --job_size 2 -c csg2.yml -q csg2 -s build  --job_size 1 --mem 35G  &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "sos run pipeline/SuSiE.ipynb uni_fsusie \\\n",
    "--genoFile /mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl/output/dlpfc_tad_region_list_plink_files/ROSMAP_NIA_WGS.leftnorm.filtered.filtered.plink_files_list.txt  \\\n",
    "--cwd output/f_susie_tad_haQTL_pos_check_pure_2/ --name haQTL \\\n",
    "--region-list dlpfc_tad_list \\\n",
    "--phenoFile phenotype_list \\\n",
    "--covFile covar_list --prior mixture_normal_per_scale -J 20 --job_size 2 -c csg2.yml -q csg2 -s build  --job_size 1 --mem 20G --container containers/stephenslab.sif &"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "import glob\n",
    "import pandas as pd\n",
    "# Input\n",
    "parameter: genoFile = path\n",
    "# A list of file path for phenotype and covariates, must be in the same order.\n",
    "parameter: phenoFile = path\n",
    "parameter: covFile = path\n",
    "parameter: region_list = path\n",
    "parameter: cwd = path(\"output\")\n",
    "parameter: name = \"demo\"\n",
    "region_tbl = pd.read_csv(region_list,sep = \"\\t\")\n",
    "id_name = region_tbl.columns.values[3]\n",
    "genoFile = pd.read_csv(genoFile,sep = \"\\t\",names = [id_name,\"path\"],header = 0).merge(region_tbl,on = id_name).to_dict(\"records\")\n",
    "\n",
    "## Path to work directory where output locates\n",
    "## Containers that contains the necessary packages\n",
    "parameter: container = \"\"\n",
    "# For cluster jobs, number commands to run per job\n",
    "parameter: job_size = 1\n",
    "# Wall clock time expected\n",
    "parameter: walltime = \"5h\"\n",
    "# Memory expected\n",
    "parameter: mem = \"16G\"\n",
    "# Number of threads\n",
    "parameter: numThreads = 2\n",
    "trait = pd.read_csv(phenoFile,sep = \"\\t\").tissue\n",
    "# use this function to edit memory string for PLINK input\n",
    "from sos.utils import expand_size"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Univariate fSuSiE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[uni_fsusie_1]\n",
    "parameter: max_L = 20\n",
    "# remove a variant if it has more than imiss missing individual data\n",
    "parameter: imiss = 0.1\n",
    "parameter: maf = 0.05\n",
    "parameter: prior  = [\"mixture_normal\", \"mixture_normal_per_scale\"]\n",
    "parameter: pheno_ID = \"ID\"\n",
    "parameter: max_SNP_EM = 1000\n",
    "input: phenoFile,covFile, for_each = \"genoFile\"\n",
    "output: [f'{cwd:a}/cache/{name}.{x}.{_genoFile[id_name]}.uni_Fsusie.{prior[0]}.rds' for x in trait ] \n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output[0]:nn}.stdout\", stderr = f\"{_output[0]:nn}.stderr\", container = container\n",
    "    ## Define function\n",
    "    compute_maf <- function(geno){\n",
    "      f <- mean(geno,na.rm = TRUE)/2\n",
    "      return(min(f, 1-f))\n",
    "    }\n",
    "\n",
    "    compute_missing <- function(geno){\n",
    "      miss <- sum(is.na(geno))/length(geno)\n",
    "      return(miss)\n",
    "    }\n",
    "    \n",
    "    mean_impute <- function(geno){\n",
    "      f <- apply(geno, 2, function(x) mean(x,na.rm = TRUE))\n",
    "      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] <- f[i]\n",
    "      return(geno)\n",
    "    }\n",
    "\n",
    "    is_zero_variance <- function(x) {\n",
    "      if (length(unique(x))==1) return(T)\n",
    "      else return(F)\n",
    "    }\n",
    "  \n",
    "    filter_X <- function(X, missing_rate_thresh, maf_thresh) {\n",
    "        rm_col <- which(apply(X, 2, compute_missing) > missing_rate_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, compute_maf) < maf_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, is_zero_variance))\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        return(mean_impute(X))\n",
    "    }\n",
    "  \n",
    "    compute_cov_flash <- function(Y, error_cache = NULL){\n",
    "        covar <- diag(ncol(Y))\n",
    "        tryCatch({\n",
    "        fl <- flashier::flash(Y, var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)\n",
    "        if(fl$n.factors==0){\n",
    "          covar <- diag(fl$residuals.sd^2)\n",
    "        } else {\n",
    "          fsd <- sapply(fl$fitted.g[[1]], '[[', \"sd\")\n",
    "          covar <- diag(fl$residuals.sd^2) + crossprod(t(fl$flash.fit$EF[[2]]) * fsd)\n",
    "        }\n",
    "        if (nrow(covar) == 0) {\n",
    "          covar <- diag(ncol(Y))\n",
    "          stop(\"Computed covariance matrix has zero rows\")\n",
    "        }\n",
    "        }, error = function(e) {\n",
    "          if (!is.null(error_cache)) {\n",
    "            saveRDS(list(data=Y, message=warning(e)), error_cache)\n",
    "            warning(\"FLASH failed. Using Identity matrix instead.\")\n",
    "            warning(e)\n",
    "          } else {\n",
    "            stop(e)\n",
    "          }\n",
    "        })\n",
    "        s <- apply(Y, 2, sd, na.rm=T)\n",
    "        if (length(s)>1) s = diag(s)\n",
    "        else s = matrix(s,1,1)\n",
    "        covar <- s%*%cov2cor(covar)%*%s\n",
    "        return(covar)\n",
    "    }\n",
    "\n",
    "    read_multi_pheno = function(path){\n",
    "      arg = paste0(c(\"tabix -h \",path,\" ${_genoFile[\"#chr\"]}:${_genoFile[\"start\"]}-${_genoFile[\"end\"]}\"),collapse = \"\")\n",
    "      result = system(arg,intern = T)\n",
    "        if (is.na(result[2]) == F){\n",
    "          output = read.table(text = result[2:length(result)], sep = \"\\t\")\n",
    "          colnames(output) = result[1]%>%stringr::str_split(\"\\t\")%>%unlist()\n",
    "        }\n",
    "        else {\n",
    "            output = NA\n",
    "            }\n",
    "        return(output) # add an if so that for each traits, those region in geno_file contains nothing in the pheno_file, can be corrctly left out\n",
    "    }       \n",
    "\n",
    "    \n",
    "    remove_covX = function(X,covar){\n",
    "    for ( i in 1:ncol(X) ) {\n",
    "        X[,i] = .lm.fit(x = cbind(1,covar), y = X[,i])$residuals\n",
    "        }\n",
    "     X = scale(X)\n",
    "    }\n",
    "    \n",
    "    ## Load Library\n",
    "    library(\"susiF.alpha\")\n",
    "    library(\"plink2R\")\n",
    "    library(\"dplyr\")\n",
    "    library(\"readr\")\n",
    "    library(\"stringr\")\n",
    "    library(\"purrr\")\n",
    "    ###\n",
    "    # Core code\n",
    "    ###\n",
    "    # Input\n",
    "    ### Genotype\n",
    "    geno = read_plink(\"${path(_genoFile[\"path\"]):n}\")\n",
    "    X = filter_X(geno$bed,${imiss}, ${maf} )\n",
    "\n",
    "    ### Phenotype\n",
    "    phenotype_list = read_delim(\"${_input[0]}\",\"\\t\")\n",
    "    traits = phenotype_list$tissue\n",
    "    covar_list = read_delim(\"${_input[1]}\",\"\\t\")\n",
    "    rownames(X) = read.table(text = rownames(X), sep= \":\" )$V2\n",
    "    covar_list = covar_list%>%mutate(covar = map(path, ~read_delim(.x,\"\\t\")%>%select(-`#id`)%>%na.omit%>%t()))\n",
    "    phenotype_list = inner_join(phenotype_list,covar_list, by = \"tissue\")\n",
    "    phenotype_list = phenotype_list %>% mutate(flag = map(path.x, read_multi_pheno)) %>% dplyr::filter(lengths(flag) >1 ) %>% mutate(Y = map2(path.x,covar, ~read_multi_pheno(.x)%>%select(start, rownames(.y))%>%t%>%as.matrix))%>%\n",
    "                                    mutate(covar = map2(covar, Y , ~.x[intersect(.x%>%rownames,rownames(.y)),] ), # add a flag so that those returning NA can be filtered out\n",
    "                                            pos = map(Y,~.x[\"start\",]),\n",
    "                                            X_data = map(covar,~X[intersect(rownames(.x),rownames(X)),]),\n",
    "                                            covar = map2(covar, X_data , ~.x[intersect(.x%>%rownames,rownames(.y)),] ),\n",
    "                                            Y = map2(covar, Y ,~ .y[intersect(.x%>%rownames,rownames(.y)),] ),\n",
    "                                            #### Get residue for each of tissue\n",
    "                                            Y_resid = map2(Y,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%>%scale%>%t%>%as_tibble))\n",
    "    \n",
    "    Y_resid = phenotype_list%>%pull(Y_resid)\n",
    "    names(Y_resid) = phenotype_list$tissue\n",
    "    X_list = phenotype_list%>%mutate( X_resid = map2(X_data,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%>%scale))%>%pull(X_resid)       # Remove covariate by tissues\n",
    "    non_missing = list()\n",
    "    fitted = list()\n",
    "    output_list = c(${_output:r})\n",
    "    rm_traits = traits[!traits %in% names(Y_resid)]\n",
    "    output_list_valid = output_list[traits %in% names(Y_resid)]\n",
    "    output_list_invalid = output_list[!traits %in% names(Y_resid)]\n",
    "    \n",
    "    # Fine-mapping with fSuSiE\n",
    "    for (r in 1:length(Y_resid)) {\n",
    "    Y_resid_mtx = (Y_resid[[r]])%>%t%>%as.matrix\n",
    "    non_missing[[r]] = which(rowSums(is.na(Y_resid_mtx)) == 0)\n",
    "    st = proc.time()\n",
    "    X = X_list[[r]]\n",
    "    print(paste(\"Dimension of X matrix:\", nrow(X), ncol(X)))\n",
    "    print(paste(\"Dimension of Y matrix:\", nrow(Y_resid_mtx), ncol(Y_resid_mtx)))\n",
    "    fitted[[r]] <- susiF(X = X[names(non_missing[[r]]),], Y_resid_mtx[non_missing[[r]],],pos = phenotype_list$pos[[r]],\n",
    "                       L=${max_L},prior = \"${prior[0]}\" ,  max_SNP_EM = ${max_SNP_EM} )\n",
    "    fitted[[r]]$time = proc.time() - st\n",
    "    fitted[[r]]$name = phenotype_list$tissue[r]\n",
    "    saveRDS(fitted[[r]], output_list_valid[r])\n",
    "    }\n",
    "    arg = paste0(c(\"touch \",output_list_invalid),collapse = \" \")\n",
    "    system(arg)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Multi F SuSiE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[multi_f_susie_1]\n",
    "parameter: max_L = 10\n",
    "# remove a variant if it has more than imiss missing individual data\n",
    "parameter: imiss = 0.1\n",
    "parameter: maf = 0.05\n",
    "parameter: prior  = [\"mixture_normal\", \"mixture_normal_per_scale\"]\n",
    "parameter: pheno_ID = \"ID\"\n",
    "input: phenoFile,covFile, for_each = \"genoFile\"\n",
    "output: f'{cwd:a}/cache/{name}.{x}.{_genoFile[id_name]}.multi_Fsusie.rds' for x in trait\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output[0]:nn}.stdout\", stderr = f\"{_output[0]:nn}.stderr\", container = container\n",
    "    ## Define function\n",
    "    compute_maf <- function(geno){\n",
    "      f <- mean(geno,na.rm = TRUE)/2\n",
    "      return(min(f, 1-f))\n",
    "    }\n",
    "\n",
    "    compute_missing <- function(geno){\n",
    "      miss <- sum(is.na(geno))/length(geno)\n",
    "      return(miss)\n",
    "    }\n",
    "    \n",
    "    mean_impute <- function(geno){\n",
    "      f <- apply(geno, 2, function(x) mean(x,na.rm = TRUE))\n",
    "      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] <- f[i]\n",
    "      return(geno)\n",
    "    }\n",
    "\n",
    "    is_zero_variance <- function(x) {\n",
    "      if (length(unique(x))==1) return(T)\n",
    "      else return(F)\n",
    "    }\n",
    "  \n",
    "    filter_X <- function(X, missing_rate_thresh, maf_thresh) {\n",
    "        rm_col <- which(apply(X, 2, compute_missing) > missing_rate_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, compute_maf) < maf_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, is_zero_variance))\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        return(mean_impute(X))\n",
    "    }\n",
    "  \n",
    "    compute_cov_flash <- function(Y, error_cache = NULL){\n",
    "        covar <- diag(ncol(Y))\n",
    "        tryCatch({\n",
    "        fl <- flashier::flash(Y, var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)\n",
    "        if(fl$n.factors==0){\n",
    "          covar <- diag(fl$residuals.sd^2)\n",
    "        } else {\n",
    "          fsd <- sapply(fl$fitted.g[[1]], '[[', \"sd\")\n",
    "          covar <- diag(fl$residuals.sd^2) + crossprod(t(fl$flash.fit$EF[[2]]) * fsd)\n",
    "        }\n",
    "        if (nrow(covar) == 0) {\n",
    "          covar <- diag(ncol(Y))\n",
    "          stop(\"Computed covariance matrix has zero rows\")\n",
    "        }\n",
    "        }, error = function(e) {\n",
    "          if (!is.null(error_cache)) {\n",
    "            saveRDS(list(data=Y, message=warning(e)), error_cache)\n",
    "            warning(\"FLASH failed. Using Identity matrix instead.\")\n",
    "            warning(e)\n",
    "          } else {\n",
    "            stop(e)\n",
    "          }\n",
    "        })\n",
    "        s <- apply(Y, 2, sd, na.rm=T)\n",
    "        if (length(s)>1) s = diag(s)\n",
    "        else s = matrix(s,1,1)\n",
    "        covar <- s%*%cov2cor(covar)%*%s\n",
    "        return(covar)\n",
    "    }\n",
    "\n",
    "    read_multi_pheno = function(path){\n",
    "      arg = paste0(c(\"tabix -h \",path,\" ${_genoFile[\"#chr\"]}:${_genoFile[\"start\"]}-${_genoFile[\"end\"]}\"),collapse = \"\")\n",
    "      result = system(arg,intern = T)\n",
    "        if (is.na(result[2]) == F){\n",
    "          output = read.table(text = result[2:length(result)], sep = \"\\t\")\n",
    "          colnames(output) = result[1]%>%stringr::str_split(\"\\t\")%>%unlist()\n",
    "        }\n",
    "        else {\n",
    "            output = NA\n",
    "            }\n",
    "        return(output) # add an if so that for each traits, those region in geno_file contains nothing in the pheno_file, can be corrctly left out\n",
    "    }       \n",
    "\n",
    "    \n",
    "    remove_covX = function(X,covar){\n",
    "    for ( i in 1:ncol(X) ) {\n",
    "        X[,i] = .lm.fit(x = cbind(1,covar), y = X[,i])$residuals\n",
    "        }\n",
    "     X = scale(X)\n",
    "    }\n",
    "    \n",
    "    ## Load Library\n",
    "    library(\"mvf.susie.alpha\")\n",
    "    library(\"susiF.alpha\")\n",
    "    library(\"plink2R\")\n",
    "    library(\"dplyr\")\n",
    "    library(\"readr\")\n",
    "    library(\"stringr\")\n",
    "    library(\"purrr\")\n",
    "    ###\n",
    "    # Core code\n",
    "    ###\n",
    "    # Input\n",
    "    ### Genotype\n",
    "    geno = read_plink(\"${path(_genoFile[\"path\"]):n}\")\n",
    "    X = filter_X(geno$bed,${imiss}, ${maf} )\n",
    "\n",
    "    ### Phenotype\n",
    "    phenotype_list = read_delim(\"${_input[0]}\",\"\\t\")\n",
    "    traits = phenotype_list$tissue\n",
    "    covar_list = read_delim(\"${_input[1]}\",\"\\t\")\n",
    "    rownames(X) = read.table(text = rownames(X), sep= \":\" )$V2\n",
    "    covar_list = covar_list%>%mutate(covar = map(path, ~read_delim(.x,\"\\t\")%>%select(-`#id`)%>%na.omit%>%t()))\n",
    "    phenotype_list = inner_join(phenotype_list,covar_list, by = \"tissue\")\n",
    "    phenotype_list = phenotype_list %>% mutate(flag = map(path.x, read_multi_pheno)) %>% dplyr::filter(lengths(flag) >1 ) %>% mutate(Y = map2(path.x,covar, ~read_multi_pheno(.x)%>%select(start, rownames(.y))%>%t%>%as.matrix))%>%\n",
    "                                    mutate(covar = map2(covar, Y , ~.x[intersect(.x%>%rownames,rownames(.y)),] ), # add a flag so that those returning NA can be filtered out\n",
    "                                            pos = map(Y,~.x[\"start\",]),\n",
    "                                            X_data = map(covar,~X[intersect(rownames(.x),rownames(X)),]),\n",
    "                                            covar = map2(covar, X_data , ~.x[intersect(.x%>%rownames,rownames(.y)),] ),\n",
    "                                            Y = map2(covar, Y ,~ .y[intersect(.x%>%rownames,rownames(.y)),] ),\n",
    "                                            #### Get residue for each of tissue\n",
    "                                            Y_resid = map2(Y,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%>%scale%>%t%>%as_tibble))\n",
    "    \n",
    "    Y_resid = phenotype_list%>%pull(Y_resid)\n",
    "    Y_resid_new = map(Y_resid, ~left_join(X[,1]%>%as.data.frame%>%rownames_to_column(\"rowname\"), .x%>%t%>%as.data.frame%>%rownames_to_column(\"rowname\") , by = \"rowname\")%>%select(-2)%>%column_to_rownames(\"rowname\")%>%as.matrix )\n",
    "\n",
    "  \n",
    "  \n",
    "  \n",
    "    m1 <- multfsusie(Y=list( Y_f = list(Y_resid_new[[1]],Y_resid_new[[3]]) , Y_u= Reduce(cbind, Y_resid_new[[2]]),pos = list(pos1 =  phenotype_list$pos[[1]],pos2 =  phenotype_list$pos[[3]]) ,\n",
    "                 X=X ,\n",
    "                 L=20 ,\n",
    "                 data.format=\"list_df\")\n",
    "    m1%>%saveRDS(\"${_output}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Univariate SuSiE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[uni_susie_1]\n",
    "parameter: max_L = 10\n",
    "# remove a variant if it has more than imiss missing individual data\n",
    "parameter: imiss = 0.1\n",
    "parameter: maf = 0.05\n",
    "import pandas as pd\n",
    "input: phenoFile,covFile, for_each = \"genoFile\"\n",
    "output: [f'{cwd:a}/cache/{name}.{x}.{_genoFile[id_name]}.unisusie.fit.rds' for x in trait  ] \n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output[0]:nn}.stdout\", stderr = f\"{_output[0]:nn}.stderr\", container = container\n",
    "    ## Define function\n",
    "    compute_maf <- function(geno){\n",
    "      f <- mean(geno,na.rm = TRUE)/2\n",
    "      return(min(f, 1-f))\n",
    "    }\n",
    "\n",
    "    compute_missing <- function(geno){\n",
    "      miss <- sum(is.na(geno))/length(geno)\n",
    "      return(miss)\n",
    "    }\n",
    "    \n",
    "    mean_impute <- function(geno){\n",
    "      f <- apply(geno, 2, function(x) mean(x,na.rm = TRUE))\n",
    "      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] <- f[i]\n",
    "      return(geno)\n",
    "    }\n",
    "\n",
    "    is_zero_variance <- function(x) {\n",
    "      if (length(unique(x))==1) return(T)\n",
    "      else return(F)\n",
    "    }\n",
    "  \n",
    "    filter_X <- function(X, missing_rate_thresh, maf_thresh) {\n",
    "        rm_col <- which(apply(X, 2, compute_missing) > missing_rate_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, compute_maf) < maf_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, is_zero_variance))\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        return(mean_impute(X))\n",
    "    }\n",
    "  \n",
    "    compute_cov_flash <- function(Y, error_cache = NULL){\n",
    "        covar <- diag(ncol(Y))\n",
    "        tryCatch({\n",
    "        fl <- flashier::flash(Y, var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)\n",
    "        if(fl$n.factors==0){\n",
    "          covar <- diag(fl$residuals.sd^2)\n",
    "        } else {\n",
    "          fsd <- sapply(fl$fitted.g[[1]], '[[', \"sd\")\n",
    "          covar <- diag(fl$residuals.sd^2) + crossprod(t(fl$flash.fit$EF[[2]]) * fsd)\n",
    "        }\n",
    "        if (nrow(covar) == 0) {\n",
    "          covar <- diag(ncol(Y))\n",
    "          stop(\"Computed covariance matrix has zero rows\")\n",
    "        }\n",
    "        }, error = function(e) {\n",
    "          if (!is.null(error_cache)) {\n",
    "            saveRDS(list(data=Y, message=warning(e)), error_cache)\n",
    "            warning(\"FLASH failed. Using Identity matrix instead.\")\n",
    "            warning(e)\n",
    "          } else {\n",
    "            stop(e)\n",
    "          }\n",
    "        })\n",
    "        s <- apply(Y, 2, sd, na.rm=T)\n",
    "        if (length(s)>1) s = diag(s)\n",
    "        else s = matrix(s,1,1)\n",
    "        covar <- s%*%cov2cor(covar)%*%s\n",
    "        return(covar)\n",
    "    }\n",
    "\n",
    "    read_gene_pheno = function(path){\n",
    "      arg = paste0(c(\"tabix -h \",path,\" ${_genoFile[\"#chr\"]}:${_genoFile[\"start\"]}-${_genoFile[\"start\"]+1}\"),collapse = \"\")\n",
    "      result = system(arg,intern = T)\n",
    "        if (is.na(result[2]) == F){\n",
    "          output = read.table(text = result[2], sep = \"\\t\")\n",
    "          colnames(output) = result[1]%>%stringr::str_split(\"\\t\")%>%unlist()\n",
    "        }\n",
    "        else {\n",
    "            output = NA\n",
    "            }\n",
    "        return(output) # add an if so that those genes in geno_file, but not presented in one specific tissue's pheno_file, can be corrctly left out\n",
    "    }       \n",
    "\n",
    "    \n",
    "    remove_covX = function(X,covar){\n",
    "    for ( i in 1:ncol(X) ) {\n",
    "        X[,i] = .lm.fit(x = cbind(1,covar), y = X[,i])$residuals\n",
    "        }\n",
    "     X = scale(X)\n",
    "    }\n",
    "    \n",
    "    ## Load Library\n",
    "    library(\"susieR\")\n",
    "    library(\"plink2R\")\n",
    "    library(\"dplyr\")\n",
    "    library(\"readr\")\n",
    "    library(\"stringr\")\n",
    "    library(\"purrr\")\n",
    "    ###\n",
    "    # Core code\n",
    "    ###\n",
    "    # Input\n",
    "    ### Genotype\n",
    "    geno = read_plink(\"${path(_genoFile[\"path\"]):n}\")\n",
    "    X = filter_X(geno$bed,${imiss}, ${maf} )\n",
    "\n",
    "    ### Phenotype\n",
    "    phenotype_list = read_delim(\"${_input[0]}\",\"\\t\")\n",
    "    traits = phenotype_list$tissue\n",
    "    covar_list = read_delim(\"${_input[1]}\",\"\\t\")\n",
    "    rownames(X) = read.table(text = rownames(X), sep= \":\" )$V2\n",
    "    covar_list = covar_list%>%mutate(covar = map(path, ~read_delim(.x,\"\\t\")%>%select(-`#id`)%>%na.omit%>%t()))\n",
    "    phenotype_list = inner_join(phenotype_list,covar_list, by = \"tissue\")\n",
    "    phenotype_list = phenotype_list %>% mutate(flag = map(path.x, read_gene_pheno)) %>% dplyr::filter(lengths(flag) >1 ) %>% mutate(Y = map2(path.x,covar, ~read_gene_pheno(.x)%>%select(-c(`#chr`,start,end,${id_name}), rownames(.y))%>%t%>%as.matrix))%>%\n",
    "                                    mutate(covar = map2(covar, Y , ~.x[intersect(.x%>%rownames,rownames(.y)),] ), # add a flag so that those returning NA can be filtered out\n",
    "                                            X_data = map(covar,~X[intersect(rownames(.x),rownames(X)),]),\n",
    "                                            covar = map2(covar, X_data , ~.x[intersect(.x%>%rownames,rownames(.y)),] ),\n",
    "                                            Y = map2(covar, Y ,~ .y[intersect(.x%>%rownames,rownames(.y)),] ),\n",
    "                                            #### Get residue for each of tissue\n",
    "                                            Y_resid = map2(Y,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%>%scale%>%t%>%as_tibble))\n",
    "    \n",
    "    Y_resid = phenotype_list%>%select(Y_resid)%>%tidyr::unnest(Y_resid)%>%t%>%as.matrix # in this part unnest will broad those matrix at smalled dimension (fewer rows/columns), so we should leave out those rows/ columns\n",
    "    colnames(Y_resid) = phenotype_list$tissue\n",
    "    \n",
    "    X_list = phenotype_list%>%mutate( X_resid = map2(X_data,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%>%scale))%>%pull(X_resid)       # Remove covariate by tissues\n",
    "    non_missing = list()\n",
    "    fitted = list()\n",
    "    output_list = c(${_output:r,})\n",
    "    rm_traits = traits[!traits %in% colnames(Y_resid)]\n",
    "    output_list_valid = output_list[traits %in% colnames(Y_resid)]\n",
    "    output_list_invalid = output_list[!traits %in% colnames(Y_resid)]\n",
    "    # Fine-mapping with SuSiE\n",
    "       for (r in 1:ncol(Y_resid)) {\n",
    "        non_missing[[r]] = which(!is.na(Y_resid[,r])) \n",
    "        st = proc.time()\n",
    "        X = X_list[[r]]\n",
    "    print(paste(\"Dimension of X matrix:\", nrow(X), ncol(X)))\n",
    "    print(paste(\"Dimension of Y matrix:\", nrow(Y_resid), ncol(Y_resid)))\n",
    "        fitted[[r]] <- susieR::susie(X[names(non_missing[[r]]),], Y_resid[non_missing[[r]],r],\n",
    "                           L=10,\n",
    "                           max_iter=1000,\n",
    "                           estimate_residual_variance=TRUE,\n",
    "                           estimate_prior_variance=TRUE,\n",
    "                           refine=TRUE,compute_univariate_zscore = TRUE, coverage = 0.95 )\n",
    "        fitted[[r]]$time = proc.time() - st\n",
    "        fitted[[r]]$cs_corr = susieR:::get_cs_correlation(fitted[[r]], X = X[names(non_missing[[r]]),])\n",
    "        fitted[[r]]$cs_snps = names(fitted[[r]]$X_column_scale_factors[unlist(fitted[[r]]$sets$cs)])\n",
    "        fitted[[r]]$variable_name =  names(fitted[[r]]$pip)\n",
    "        fitted[[r]]$coef =  coef.susie(fitted[[r]])\n",
    "        fitted[[r]]$input_data = list(\"X_resid\" = X[names(non_missing[[r]]),], \"Y_resid\" =Y_resid[non_missing[[r]],r] )\n",
    "        fitted[[r]]$name = phenotype_list$tissue[r]\n",
    "        saveRDS(fitted[[r]], output_list_valid[r])\n",
    "    }\n",
    "    arg = paste0(c(\"touch \",output_list_invalid),collapse = \" \")\n",
    "    system(arg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[uni_susie_sqtl_1]\n",
    "parameter: max_L = 10\n",
    "# remove a variant if it has more than imiss missing individual data\n",
    "parameter: imiss = 0.1\n",
    "parameter: maf = 0.05\n",
    "import pandas as pd\n",
    "input: phenoFile,covFile, for_each = \"genoFile\"\n",
    "output: f'{cwd:a}/cache/{name}.{_genoFile[id_name]}.unisusie.fit.rds'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output[0]:nn}.stdout\", stderr = f\"{_output[0]:nn}.stderr\", container = container\n",
    "    ## Define function\n",
    "    compute_maf <- function(geno){\n",
    "      f <- mean(geno,na.rm = TRUE)/2\n",
    "      return(min(f, 1-f))\n",
    "    }\n",
    "\n",
    "    compute_missing <- function(geno){\n",
    "      miss <- sum(is.na(geno))/length(geno)\n",
    "      return(miss)\n",
    "    }\n",
    "    \n",
    "    mean_impute <- function(geno){\n",
    "      f <- apply(geno, 2, function(x) mean(x,na.rm = TRUE))\n",
    "      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] <- f[i]\n",
    "      return(geno)\n",
    "    }\n",
    "\n",
    "    is_zero_variance <- function(x) {\n",
    "      if (length(unique(x))==1) return(T)\n",
    "      else return(F)\n",
    "    }\n",
    "  \n",
    "    filter_X <- function(X, missing_rate_thresh, maf_thresh) {\n",
    "        rm_col <- which(apply(X, 2, compute_missing) > missing_rate_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, compute_maf) < maf_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, is_zero_variance))\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        return(mean_impute(X))\n",
    "    }\n",
    "  \n",
    "    compute_cov_flash <- function(Y, error_cache = NULL){\n",
    "        covar <- diag(ncol(Y))\n",
    "        tryCatch({\n",
    "        fl <- flashier::flash(Y, var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)\n",
    "        if(fl$n.factors==0){\n",
    "          covar <- diag(fl$residuals.sd^2)\n",
    "        } else {\n",
    "          fsd <- sapply(fl$fitted.g[[1]], '[[', \"sd\")\n",
    "          covar <- diag(fl$residuals.sd^2) + crossprod(t(fl$flash.fit$EF[[2]]) * fsd)\n",
    "        }\n",
    "        if (nrow(covar) == 0) {\n",
    "          covar <- diag(ncol(Y))\n",
    "          stop(\"Computed covariance matrix has zero rows\")\n",
    "        }\n",
    "        }, error = function(e) {\n",
    "          if (!is.null(error_cache)) {\n",
    "            saveRDS(list(data=Y, message=warning(e)), error_cache)\n",
    "            warning(\"FLASH failed. Using Identity matrix instead.\")\n",
    "            warning(e)\n",
    "          } else {\n",
    "            stop(e)\n",
    "          }\n",
    "        })\n",
    "        s <- apply(Y, 2, sd, na.rm=T)\n",
    "        if (length(s)>1) s = diag(s)\n",
    "        else s = matrix(s,1,1)\n",
    "        covar <- s%*%cov2cor(covar)%*%s\n",
    "        return(covar)\n",
    "    }\n",
    "\n",
    "       read_multi_pheno = function(path){\n",
    "          arg = paste0(c(\"tabix -h \",path,\" ${_genoFile[\"#chr\"]}:${_genoFile[\"start\"]}-${_genoFile[\"end\"]}\"),collapse = \"\")\n",
    "          result = system(arg,intern = T)\n",
    "            if (is.na(result[2]) == F){\n",
    "              output = read.table(text = result[2:length(result)], sep = \"\\t\")\n",
    "              colnames(output) = result[1]%>%stringr::str_split(\"\\t\")%>%unlist()\n",
    "            }\n",
    "            else {\n",
    "                output = NA\n",
    "                }\n",
    "            return(output) # add an if so that for each traits, those region in geno_file contains nothing in the pheno_file, can be corrctly left out\n",
    "        }       \n",
    "        \n",
    "        remove_covX = function(X,covar){\n",
    "        for ( i in 1:ncol(X) ) {\n",
    "            X[,i] = .lm.fit(x = cbind(1,covar), y = X[,i])$residuals\n",
    "            }\n",
    "         X = scale(X)\n",
    "        }\n",
    "        \n",
    "        ## Load Library\n",
    "        library(\"susieR\")\n",
    "        library(\"plink2R\")\n",
    "        library(\"dplyr\")\n",
    "        library(\"readr\")\n",
    "        library(\"stringr\")\n",
    "        library(\"purrr\")\n",
    "        ###\n",
    "        # Core code\n",
    "        ###\n",
    "        # Input\n",
    "        ### Genotype\n",
    "        geno = read_plink(\"${path(_genoFile[\"path\"]):n}\")\n",
    "        X = filter_X(geno$bed,${imiss}, ${maf} )\n",
    "    \n",
    "        ### Phenotype\n",
    "        phenotype_list = read_delim(\"${_input[0]}\",\"\\t\")\n",
    "        traits = phenotype_list$tissue\n",
    "        covar_list = read_delim(\"${_input[1]}\",\"\\t\")\n",
    "        rownames(X) = read.table(text = rownames(X), sep= \":\" )$V2\n",
    "        covar_list = covar_list%>%mutate(covar = map(path, ~read_delim(.x,\"\\t\")%>%select(-`#id`)%>%na.omit%>%t()))\n",
    "        phenotype_list = inner_join(phenotype_list,covar_list, by = \"tissue\")\n",
    "        phenotype_list = phenotype_list %>% mutate(flag = map(path.x, read_multi_pheno)) %>% dplyr::filter(lengths(flag) >1 ) %>% mutate(Y = map2(path.x,covar, ~read_multi_pheno(.x)%>%select(start, rownames(.y))%>%t%>%as.matrix))%>%\n",
    "                                        mutate(covar = map2(covar, Y , ~.x[intersect(.x%>%rownames,rownames(.y)),] ), # add a flag so that those returning NA can be filtered out\n",
    "                                                pos = map(Y,~.x[\"start\",]),\n",
    "                                                X_data = map(covar,~X[intersect(rownames(.x),rownames(X)),]),\n",
    "                                                covar = map2(covar, X_data , ~.x[intersect(.x%>%rownames,rownames(.y)),] ),\n",
    "                                                Y = map2(covar, Y ,~ .y[intersect(.x%>%rownames,rownames(.y)),] ),\n",
    "                                                #### Get residue for each of tissue\n",
    "                                                Y_resid = map2(Y,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%>%scale%>%t%>%as_tibble))\n",
    "        \n",
    "        Y_resid = phenotype_list%>%select(Y_resid)%>%tidyr::unnest(Y_resid)%>%t%>%as.matrix # in this part unnest will broad those matrix at smalled dimension (fewer rows/columns), so we should leave out those rows/ columns\n",
    "        colnames(Y_resid) = phenotype_list$tissue\n",
    "        \n",
    "        X_list = phenotype_list%>%mutate( X_resid = map2(X_data,covar,~.lm.fit(x = cbind(1,.y), y = .x)$residuals%>%scale))%>%pull(X_resid)       # Remove covariate by tissues\n",
    "        non_missing = list()\n",
    "        fitted = list()\n",
    "        output_list = c(${_output:r,})\n",
    "        rm_traits = traits[!traits %in% colnames(Y_resid)]\n",
    "        # Fine-mapping with SuSiE\n",
    "           for (r in 1:ncol(Y_resid)) {\n",
    "            non_missing[[r]] = which(!is.na(Y_resid[,r])) \n",
    "            st = proc.time()\n",
    "            X = X_list[[1]]\n",
    "        print(paste(\"Dimension of X matrix:\", nrow(X), ncol(X)))\n",
    "        print(paste(\"Dimension of Y matrix:\", nrow(Y_resid), ncol(Y_resid)))\n",
    "            fitted[[r]] <- susieR::susie(X[names(non_missing[[r]]),], Y_resid[non_missing[[r]],r],\n",
    "                               L=10,\n",
    "                               max_iter=1000,\n",
    "                               estimate_residual_variance=TRUE,\n",
    "                               estimate_prior_variance=TRUE,\n",
    "                               refine=TRUE,compute_univariate_zscore = TRUE, coverage = 0.95 )\n",
    "            fitted[[r]]$time = proc.time() - st\n",
    "            fitted[[r]]$cs_corr = susieR:::get_cs_correlation(fitted[[r]], X = X[names(non_missing[[r]]),])\n",
    "            fitted[[r]]$cs_snps = names(fitted[[r]]$X_column_scale_factors[unlist(fitted[[r]]$sets$cs)])\n",
    "            fitted[[r]]$variable_name =  names(fitted[[r]]$pip)\n",
    "            fitted[[r]]$coef =  coef.susie(fitted[[r]])\n",
    "            fitted[[r]]$name = phenotype_list$tissue[r]\n",
    "        }  \n",
    "        saveRDS(fitted, ${_output})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "#[uni_susie_2]\n",
    "input: group_with = \"genoFile\"\n",
    "output: f\"{_input:n}.vcf.bgz\"\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output:nn}.stdout\", stderr = f\"{_output:nn}.stderr\", container = container\n",
    "   ## Define create_vcf function\n",
    "           create_vcf = function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL, \n",
    "        se = NULL, pval = NULL, name = NULL,cs = NULL, pip = NULL) \n",
    "    {\n",
    "        stopifnot(length(chrom) == length(pos))\n",
    "        if (is.null(snp)) {\n",
    "            snp <- paste0(chrom, \":\", pos)\n",
    "        }\n",
    "        snp <- paste0(chrom, \":\", pos)\n",
    "        nsnp <- length(chrom)\n",
    "        gen <- list()\n",
    "        ## Setupt data content for each sample column\n",
    "        if (!is.null(ea_af)) \n",
    "            gen[[\"AF\"]] <- matrix(ea_af, nsnp)\n",
    "        if (!is.null(effect)) \n",
    "            gen[[\"ES\"]] <- matrix(effect, nsnp)\n",
    "        if (!is.null(se)) \n",
    "            gen[[\"SE\"]] <- matrix(se, nsnp)\n",
    "        if (!is.null(pval)) \n",
    "            gen[[\"LP\"]] <- matrix(-log10(pval), nsnp)\n",
    "        if (!is.null(cs)) \n",
    "            gen[[\"CS\"]] <- matrix(cs, nsnp)\n",
    "        if (!is.null(pip)) \n",
    "            gen[[\"PIP\"]] <- matrix(pip, nsnp)\n",
    "        gen <- S4Vectors::SimpleList(gen)\n",
    "        \n",
    "      ## Setup snps info for the fix columns\n",
    "        gr <- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, \n",
    "            end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))\n",
    "         coldata <- S4Vectors::DataFrame(Studies = name, row.names = name)\n",
    "    ## Setup header informations\n",
    "        hdr <- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\", \n",
    "            row.names = \"fileformat\")), sample = name)\n",
    "        VariantAnnotation::geno(hdr) <- S4Vectors::DataFrame(Number = c(\"A\", \n",
    "            \"A\", \"A\", \"A\", \"A\", \"A\"), Type = c(\"Float\", \"Float\", \n",
    "            \"Float\", \"Float\", \"Float\", \"Float\"), Description = c(\"Effect size estimate relative to the alternative allele\", \n",
    "            \"Standard error of effect size estimate\", \"-log10 p-value for effect estimate\",  \n",
    "            \"Alternate allele frequency in the association study\",\n",
    "            \"The CS this variate are captured, 0 indicates not in any cs\", \"The posterior inclusion probability to a CS\"), \n",
    "            row.names = c(\"ES\", \"SE\", \"LP\", \"AF\", \"CS\", \"PIP\"))\n",
    "      ## Save only the meta information in the sample columns \n",
    "        VariantAnnotation::geno(hdr) <- subset(VariantAnnotation::geno(hdr), \n",
    "            rownames(VariantAnnotation::geno(hdr)) %in% names(gen))\n",
    "      ## Save VCF \n",
    "        vcf <- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, \n",
    "            exptData = list(header = hdr), geno = gen)\n",
    "        VariantAnnotation::alt(vcf) <- Biostrings::DNAStringSetList(as.list(ea))\n",
    "        VariantAnnotation::ref(vcf) <- Biostrings::DNAStringSet(nea)\n",
    "      ## Add fixed values\n",
    "        VariantAnnotation::fixed(vcf)$FILTER <- \"PASS\"\n",
    "          return(sort(vcf))\n",
    "        }\n",
    "    library(\"susieR\")\n",
    "    library(\"dplyr\")\n",
    "    library(\"tibble\")\n",
    "    library(\"purrr\")\n",
    "    library(\"readr\")\n",
    "    library(\"tidyr\")\n",
    "    library(\"stringr\")\n",
    "    \n",
    "    # Get list of cs snps\n",
    "    susie_list = readRDS(${_input:r})\n",
    "    susie_tb_ls = list()\n",
    "    for (i in 1:length(susie_list)){\n",
    "        susie_tb = tibble( snps =  names(susie_list[[1]]$pip)[which( susie_list[[i]]$pip >= 0)], snps_index = which(( susie_list[[i]]$pip >= 0))  )\n",
    "        susie_tb_ls[[i]]= susie_tb%>%mutate( cs = map(snps_index,~which( susie_list[[i]]$sets$cs %in% .x))%>%as.numeric%>%replace_na(0),\n",
    "                                 pip = map_dbl(snps_index,~( susie_list[[i]]$pip[.x])),\n",
    "                                 coef = map_dbl(snps_index,~(coef.susie( susie_list[[i]])[.x+1])))\n",
    "        }\n",
    "    if(length(susie_tb_ls) >= 2){ \n",
    "      for(i in 2:length(susie_tb_ls)){\n",
    "          susie_tb_ls[[i]] = full_join(susie_tb_ls[[i-1]],susie_tb_ls[[i]], by = \"snps\") \n",
    "        }\n",
    "    }\n",
    "    m = c(\"cs\",\"pip\",\"coef\")    \n",
    "    output = list()\n",
    "    for(i in m){\n",
    "    output[[i]] = susie_tb_ls[[length(susie_tb_ls)]]%>%select(contains(i))%>%as.matrix\n",
    "    }\n",
    "    snps_tb = susie_tb_ls[[length(susie_tb_ls)]]%>%mutate(\n",
    "                         chr = map_chr(snps,~read.table(text = .x,sep = \":\",as.is = T)$V1),\n",
    "                         pos_alt_ref = map_chr(snps,~read.table(text = .x,sep = \":\",as.is = TRUE)$V2),\n",
    "                         pos = map_dbl(pos_alt_ref,~read.table(text = .x,sep = \"_\",as.is = TRUE)$V1),\n",
    "                         alt = map_chr(pos_alt_ref,~read.table(text = .x,sep = \"_\",as.is = TRUE, colClass = \"character\")$V2),\n",
    "                         ref = map_chr(pos_alt_ref,~read.table(text = .x,sep = \"_\",as.is = TRUE, colClass = \"character\")$V3))\n",
    "    \n",
    "    snps_tb = snps_tb%>%filter(str_detect(ref, \"[ACTG]\") & str_detect(alt, \"[ACTG]\"))\n",
    "    output_vcf = create_vcf(\n",
    "            chrom = snps_tb$chr,\n",
    "             pos = snps_tb$pos,\n",
    "             ea = snps_tb$alt,\n",
    "             nea = snps_tb$ref,\n",
    "             effect = snps_tb%>%select(contains(\"coef\"))%>%as.matrix ,\n",
    "             pip = snps_tb%>%select(contains(\"pip\"))%>%as.matrix,\n",
    "             cs = snps_tb%>%select(contains(\"cs\"))%>%as.matrix,\n",
    "             name = names(susie_list))\n",
    "    VariantAnnotation::writeVcf(output_vcf,${_output:nr},index = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "#[*_susie_3]\n",
    "input: group_by = \"all\"\n",
    "output: f'{cwd}/{name}.susie.output_list.txt'\n",
    "python: expand= \"$[ ]\", stderr = f'{_output}.stderr', stdout = f'{_output}.stdout', container = container\n",
    "    import pandas as pd\n",
    "    pd.DataFrame({\"output\" : [$[_input:ar,]]}).to_csv(\"$[_output]\",index = False ,header = False, sep = \"t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[mv_susie_1]\n",
    "parameter: max_L = 10\n",
    "# remove a variant if it has more than imiss missing individual data\n",
    "parameter: imiss = 0.1\n",
    "parameter: maf = 0.05\n",
    "# Only analyze `cis` variants -- cis = N means using N variants around the center column of X matrix  \n",
    "parameter: cis = 'NULL'\n",
    "parameter: prior = path\n",
    "input: phenoFile,covFile, for_each = \"genoFile\"\n",
    "output: mvsusie = f'{wd:a}/{_input:bn}{(\"_cis_%s\" % cis) if cis != \"NULL\" else \"\"}_{name}.mvsusie.rds'\n",
    "task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output[0]:nn}.stdout\", stderr = f\"{_output[0]:nn}.stderr\", container = container\n",
    "    ###\n",
    "    # Utility functions\n",
    "    ###\n",
    "    compute_maf <- function(geno){\n",
    "      f <- mean(geno,na.rm = TRUE)/2\n",
    "      return(min(f, 1-f))\n",
    "    }\n",
    "\n",
    "    compute_missing <- function(geno){\n",
    "      miss <- sum(is.na(geno))/length(geno)\n",
    "      return(miss)\n",
    "    }\n",
    "    \n",
    "    mean_impute <- function(geno){\n",
    "      f <- apply(geno, 2, function(x) mean(x,na.rm = TRUE))\n",
    "      for (i in 1:length(f)) geno[,i][which(is.na(geno[,i]))] <- f[i]\n",
    "      return(geno)\n",
    "    }\n",
    "\n",
    "    is_zero_variance <- function(x) {\n",
    "      if (length(unique(x))==1) return(T)\n",
    "      else return(F)\n",
    "    }\n",
    "  \n",
    "    filter_X <- function(X, missing_rate_thresh, maf_thresh) {\n",
    "        rm_col <- which(apply(X, 2, compute_missing) > missing_rate_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, compute_maf) < maf_thresh)\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        rm_col <- which(apply(X, 2, is_zero_variance))\n",
    "        if (length(rm_col)) X <- X[, -rm_col]\n",
    "        return(mean_impute(X))\n",
    "    }\n",
    "\n",
    "    compute_cov_flash <- function(Y, error_cache = NULL){\n",
    "        covar <- diag(ncol(Y))\n",
    "        tryCatch({\n",
    "        fl <- flashier::flash(Y, var.type = 2, prior.family = c(flashier::prior.normal(), flashier::prior.normal.scale.mix()), backfit = TRUE, verbose.lvl=0)\n",
    "        if(fl$n.factors==0){\n",
    "          covar <- diag(fl$residuals.sd^2)\n",
    "        } else {\n",
    "          fsd <- sapply(fl$fitted.g[[1]], '[[', \"sd\")\n",
    "          covar <- diag(fl$residuals.sd^2) + crossprod(t(fl$flash.fit$EF[[2]]) * fsd)\n",
    "        }\n",
    "        if (nrow(covar) == 0) {\n",
    "          covar <- diag(ncol(Y))\n",
    "          stop(\"Computed covariance matrix has zero rows\")\n",
    "        }\n",
    "        }, error = function(e) {\n",
    "          if (!is.null(error_cache)) {\n",
    "            saveRDS(list(data=Y, message=warning(e)), error_cache)\n",
    "            warning(\"FLASH failed. Using Identity matrix instead.\")\n",
    "            warning(e)\n",
    "          } else {\n",
    "            stop(e)\n",
    "          }\n",
    "        })\n",
    "        s <- apply(Y, 2, sd, na.rm=T)\n",
    "        if (length(s)>1) s = diag(s)\n",
    "        else s = matrix(s,1,1)\n",
    "        covar <- s%*%cov2cor(covar)%*%s\n",
    "        return(covar)\n",
    "    }\n",
    "  \n",
    "    compute_cov_diag <- function(Y){\n",
    "        covar <- diag(apply(Y, 2, var, na.rm=T))\n",
    "        return(covar)\n",
    "    }\n",
    "\n",
    "    get_center <- function(k,n) {\n",
    "      ## For given number k, get the range k surrounding n/2\n",
    "      ## but have to make sure it does not go over the bounds\n",
    "      if (is.null(k)) {\n",
    "          return(1:n)\n",
    "      }\n",
    "      start = floor(n/2 - k/2)\n",
    "      end = floor(n/2 + k/2)\n",
    "      if (start<1) start = 1\n",
    "      if (end>n) end = n\n",
    "      return(start:end)\n",
    "    }\n",
    "    \n",
    "    get_prior_indices <- function(Y, U) {\n",
    "      # make sure the prior col/rows match the colnames of the Y matrix\n",
    "      y_names = colnames(Y)\n",
    "      u_names = colnames(U)\n",
    "      if (is.null(y_names) || is.null(u_names)) {\n",
    "          return(NULL)\n",
    "      } else if (identical(y_names, u_names)) {\n",
    "          return(NULL)\n",
    "      } else {\n",
    "          return(match(y_names, u_names))\n",
    "      }\n",
    "    }\n",
    "    \n",
    "    ###\n",
    "    # Core code\n",
    "    ###\n",
    "    # Input\n",
    "    ### Genotype\n",
    "    geno = read_plink(\"${path(_genoFile[\"path\"]):n}\")\n",
    "    X = filter_X(geno$bed,${imiss}, ${maf} )\n",
    "    X = X[,get_center(${cis}, ncol(X))]\n",
    "    ### Phenotype\n",
    "    phenotype_list = read_delim(\"${_input[0]}\",\"\\t\")\n",
    "    covar_list = read_delim(\"${_input[1]}\",\"\\t\")\n",
    "    covar_list = covar_list%>%mutate(covar = map(path, ~read_delim(.x,\"\\t\")%>%select(-`#id`)%>%na.omit%>%t()))\n",
    "    phenotype_list = inner_join(phenotype_list,covar_list, by = \"tissue\")\n",
    "    phenotype_list = phenotype_list%>%mutate(Y = map(path.x, ~read_gene_pheno(.x)%>%select(-c(`#chr`,start,end,${id_name}))%>%t%>%as.matrix))%>%mutate(\n",
    "                                            #### Get residue for each of tissue\n",
    "                                              Y_resid = map2(Y,covar,~.lm.fit(x = cbind(1,.y), y = .x[rownames(.y),])$residuals%>%t%>%as_tibble))\n",
    "    \n",
    "    Y_resid = phenotype_list%>%select(Y_resid)%>%tidyr::unnest(Y_resid)%>%t%>%as.matrix\n",
    "    colnames(Y_resid) = phenotype_list$tissue\n",
    "    # FIXME: handle it when prior does not exist\n",
    "    prior = readRDS(${prior:r})\n",
    "    print(paste(\"Number of components in the mixture prior:\", length(prior$U)))\n",
    "    prior = mvsusieR::create_mash_prior(mixture_prior=list(weights=prior$w, matrices=prior$U), include_indices = get_prior_indices(Y_resid, prior$U[[1]]), max_mixture_len=-1)\n",
    "    print(paste(\"Dimension of X matrix:\", nrow(X), ncol(X)))\n",
    "    print(paste(\"Dimension of Y matrix:\", nrow(Y_resid), ncol(Y_resid)))\n",
    "      \n",
    "\n",
    "    # GWAS Summary statistics\n",
    "    univariate_res = lapply(1:ncol(Y_resid), function(r) susieR:::univariate_regression(X[non_missing[[r]], ], Y_resid[non_missing[[r]], r]))\n",
    "    bhat = do.call(cbind, lapply(1:ncol(Y_resid), function(r) univariate_res[[r]]$betahat))\n",
    "    sbhat = do.call(cbind, lapply(1:ncol(Y_resid), function(r) univariate_res[[r]]$sebetahat))\n",
    "    saveRDS(list(bhat=bhat, sbhat=sbhat), \"${_output[0]:nn}.sumstat.rds\")\n",
    "    rm(bhat)\n",
    "    rm(sbhat)\n",
    "    # Multivariate fine-mapping\n",
    "    st = proc.time()\n",
    "    mv_res = mvsusieR::mvsusie(X, Y_resid, L=${max_L}, \n",
    "                              prior_variance=prior, residual_variance=resid_Y, \n",
    "                              precompute_covariances=F, compute_objective=T, \n",
    "                              estimate_residual_variance=F, estimate_prior_variance=T, estimate_prior_method='EM',\n",
    "                              max_iter = 100, n_thread=1, approximate=F)\n",
    "    mv_res$time = proc.time() - st\n",
    "    mv_res$cs_corr = susieR:::get_cs_correlation(mv_res, X=X)\n",
    "    saveRDS(mv_res, ${_output[0]:r})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[mv_susie_2]\n",
    "input: group_with = \"genoFile\"\n",
    "output: f\"{_input:n}.vcf.bgz\"\n",
    "task: trunk_workers = 1, trunk_size = 1, walltime = '2h', mem = '55G', cores = 1, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output:nn}.stdout\", stderr = f\"{_output:nn}.stderr\", container = container\n",
    "   ## Define create_vcf function\n",
    "           create_vcf = function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL, \n",
    "        se = NULL, pval = NULL, name = NULL,cs = NULL, pip = NULL) \n",
    "    {\n",
    "        stopifnot(length(chrom) == length(pos))\n",
    "        if (is.null(snp)) {\n",
    "            snp <- paste0(chrom, \":\", pos)\n",
    "        }\n",
    "        snp <- paste0(chrom, \":\", pos)\n",
    "        nsnp <- length(chrom)\n",
    "        gen <- list()\n",
    "        ## Setupt data content for each sample column\n",
    "        if (!is.null(ea_af)) \n",
    "            gen[[\"AF\"]] <- matrix(ea_af, nsnp)\n",
    "        if (!is.null(effect)) \n",
    "            gen[[\"ES\"]] <- matrix(effect, nsnp)\n",
    "        if (!is.null(se)) \n",
    "            gen[[\"SE\"]] <- matrix(se, nsnp)\n",
    "        if (!is.null(pval)) \n",
    "            gen[[\"LP\"]] <- matrix(-log10(pval), nsnp)\n",
    "        if (!is.null(cs)) \n",
    "            gen[[\"CS\"]] <- matrix(cs, nsnp)\n",
    "        if (!is.null(pip)) \n",
    "            gen[[\"PIP\"]] <- matrix(pip, nsnp)\n",
    "        gen <- S4Vectors::SimpleList(gen)\n",
    "        \n",
    "      ## Setup snps info for the fix columns\n",
    "        gr <- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, \n",
    "            end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))\n",
    "         coldata <- S4Vectors::DataFrame(Studies = name, row.names = name)\n",
    "    ## Setup header informations\n",
    "        hdr <- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = \"VCFv4.2\", \n",
    "            row.names = \"fileformat\")), sample = name)\n",
    "        VariantAnnotation::geno(hdr) <- S4Vectors::DataFrame(Number = c(\"A\", \n",
    "            \"A\", \"A\", \"A\", \"A\", \"A\"), Type = c(\"Float\", \"Float\", \n",
    "            \"Float\", \"Float\", \"Float\", \"Float\"), Description = c(\"Effect size estimate relative to the alternative allele\", \n",
    "            \"Standard error of effect size estimate\", \"-log10 p-value for effect estimate\",  \n",
    "            \"Alternate allele frequency in the association study\",\n",
    "            \"The CS this variate are captured, 0 indicates not in any cs\", \"The posterior inclusion probability to a CS\"), \n",
    "            row.names = c(\"ES\", \"SE\", \"LP\", \"AF\", \"CS\", \"PIP\"))\n",
    "      ## Save only the meta information in the sample columns \n",
    "        VariantAnnotation::geno(hdr) <- subset(VariantAnnotation::geno(hdr), \n",
    "            rownames(VariantAnnotation::geno(hdr)) %in% names(gen))\n",
    "      ## Save VCF \n",
    "        vcf <- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, \n",
    "            exptData = list(header = hdr), geno = gen)\n",
    "        VariantAnnotation::alt(vcf) <- Biostrings::DNAStringSetList(as.list(ea))\n",
    "        VariantAnnotation::ref(vcf) <- Biostrings::DNAStringSet(nea)\n",
    "      ## Add fixed values\n",
    "        VariantAnnotation::fixed(vcf)$FILTER <- \"PASS\"\n",
    "          return(sort(vcf))\n",
    "        }\n",
    "    library(\"susieR\")\n",
    "    library(\"dplyr\")\n",
    "    library(\"tibble\")\n",
    "    library(\"purrr\")\n",
    "    library(\"readr\")\n",
    "    library(\"tidyr\")\n",
    "    \n",
    "    # Get list of cs snps\n",
    "    res = readRDS(${_input:r})\n",
    "    output_snps = tibble( snps = res$variable_name[which(res$pip >= 0)], snps_index = which((res$pip >= 0))  )\n",
    "    output_snps = output_snps%>%mutate( cs = map(snps_index,~which(res$sets$cs %in% .x))%>%as.numeric%>%replace_na(0),\n",
    "                             pip = map_dbl(snps_index,~(res$pip[.x])),\n",
    "                     chr = map_chr(snps,~read.table(text = .x,sep = \":\",as.is = T)$V1),\n",
    "                     pos_alt_ref = map_chr(snps,~read.table(text = .x,sep = \":\",as.is = TRUE)$V2),\n",
    "                     pos = map_dbl(pos_alt_ref,~read.table(text = .x,sep = \"_\",as.is = TRUE)$V1),\n",
    "                     alt = map_chr(pos_alt_ref,~read.table(text = .x,sep = \"_\",as.is = TRUE, colClass = \"character\")$V2),\n",
    "                     ref = map_chr(pos_alt_ref,~read.table(text = .x,sep = \"_\",as.is = TRUE, colClass = \"character\")$V3))\n",
    "    \n",
    "    effect_mtr = res$coef[output_snps$snps_index+1]%>%as.matrix\n",
    "    colnames(effect_mtr) = \"${name}\"\n",
    "    rownames(effect_mtr) = output_snps$snps\n",
    "    cs_mtr = effect_mtr\n",
    "    for(i in 1:nrow(cs_mtr)) cs_mtr[i,] =  output_snps$cs[[i]]  \n",
    "    pip_mtr = effect_mtr\n",
    "    for(i in 1:nrow(pip_mtr)) pip_mtr[i,] =  output_snps$pip[[i]]  \n",
    "    \n",
    "    output_vcf = create_vcf(\n",
    "           chrom = output_snps$chr,\n",
    "            pos = output_snps$pos,\n",
    "            ea = output_snps$alt,\n",
    "            nea = output_snps$ref,\n",
    "            effect = effect_mtr ,\n",
    "            pip = pip_mtr,\n",
    "            cs = cs_mtr,\n",
    "            name = colnames(effect_mtr)\n",
    "              )\n",
    "    VariantAnnotation::writeVcf(output_vcf,${_output:nr},index = TRUE)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     ""
    ]
   ],
   "version": "0.23.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
