

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Advanced regression models for association analysis with individual level data &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/mnm_analysis/mnm_regression';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="SuSiE fine-mapping workflow" href="SuSiE.html" />
    <link rel="prev" title="Overview of advanced cis-QTL analysis" href="cis_advanced_overview.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced cis-QTL Analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="cis_advanced_overview.html">Overview of advanced cis-QTL analysis</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Advanced regression models for association analysis with individual level data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="SuSiE.html">SuSiE fine-mapping workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="fSuSiE.html">fSuSiE workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="mvSuSiE.html">Multivariate fine-mapping workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="mrmash.html">Multivariate TWAS prediction model</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced Genome-wide Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="rss_analysis.html">High-dimensional regression with summary statistics</a></li>
<li class="toctree-l1"><a class="reference internal" href="../prototype_drafts/polyfun.html">Fine-mapping with PolyFun</a></li>
<li class="toctree-l1"><a class="reference internal" href="../prototype_drafts/MRAID_QTL.html">Mendelian Randomization using MRAID</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/ldsc.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL-GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../pecotmr_integration/pecotmr.html">The PECOTMR intergration framework</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">GWAS integration: enrichment and colocalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/twas_mr.html">GWAS integration: TWAS and MR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/intact.html">TWAS and COLOC integration</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">Extract genome-wide data for multivariate analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">MASH analysis pipeline with posterior computation</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Rare xQTL</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../rare_xqtl/watershed.html">watershed analysis for rare xQTL</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Post Processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../post_processing/sumstats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/rds_to_vcf.html">Summary statistics in VCF format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/association_scan_post_processing.html">Association result processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">SuSIE results post process</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/mnm_analysis/mnm_regression.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/mnm_analysis/mnm_regression.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/mnm_analysis/mnm_regression.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Advanced regression models for association analysis with individual level data</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genotype-data">Example genotype data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-association-window-file">Example association-window file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#about-indels">About indels</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-susie-with-twas-weights-from-multiple-methods">i. SuSiE with TWAS weights from multiple methods</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-fsusie">ii. fSuSiE</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-mvsusie-with-twas-weights-from-mr-mash">iii. mvSuSiE with TWAS weights from mr.mash</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#univariate-analysis-susie-and-twas">Univariate analysis: SuSiE and TWAS</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multivariate-analysis-mvsusie-and-mr-mash">Multivariate analysis: mvSuSiE and mr.mash</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-for-epigenomic-qtl-fine-mapping">Functional regression fSuSiE for epigenomic QTL fine-mapping</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-with-other-modality">Functional regression fSuSiE with other modality</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="advanced-regression-models-for-association-analysis-with-individual-level-data">
<h1>Advanced regression models for association analysis with individual level data<a class="headerlink" href="#advanced-regression-models-for-association-analysis-with-individual-level-data" title="Permalink to this heading">#</a></h1>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Permalink to this heading">#</a></h2>
<p>This notebook performs various advanced statistical analysis on multiple xQTL in a given region. Current procedures implemented include:</p>
<ol class="arabic simple">
<li><p>Univariate analysis</p>
<ul class="simple">
<li><p>SuSiE</p></li>
<li><p>Univeriate TWAS weights: LASSO, Elastic net, mr.ash, SuSiE, Bayes alphabet soup</p></li>
<li><p>Cross validation of TWAS methods (optional but highly recommended if TWAS weights are computed)</p></li>
</ul>
</li>
<li><p>Functional data (epigenomic xQTL) analysis</p>
<ul class="simple">
<li><p>fSuSiE</p></li>
</ul>
</li>
<li><p>Multivariate analysis</p>
<ul class="simple">
<li><p>mvSuSiE</p></li>
<li><p>Multivariate TWAS weights: mvSuSiE and mr.mash</p></li>
</ul>
</li>
</ol>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>A list of regions to be analyzed (optional); the last column of this file should be region name.</p></li>
<li><p>Either a list of per chromosome genotype files, or one file for genotype data of the entire genome. Genotype data has to be in PLINK <code class="docutils literal notranslate"><span class="pre">bed</span></code> format.</p></li>
<li><p>Vector of lists of phenotype files per region to be analyzed, in UCSC <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> with index in <code class="docutils literal notranslate"><span class="pre">bed.gz.tbi</span></code> formats.</p></li>
<li><p>Vector of covariate files corresponding to the lists above.</p></li>
<li><p>Customized association windows file for variants (cis or trans). If it is not provided, a fixed sized window will be used around the region (a cis-window)</p></li>
<li><p>Optionally a vector of names of the phenotypic conditions in the form of <code class="docutils literal notranslate"><span class="pre">cond1</span> <span class="pre">cond2</span> <span class="pre">cond3</span></code> separated with whitespace.</p></li>
</ol>
<p>Input 2 and 3 should be outputs from <code class="docutils literal notranslate"><span class="pre">genotype_per_region</span></code> and <code class="docutils literal notranslate"><span class="pre">annotate_coord</span></code> modules in previous preprocessing steps. 4 should be output of <code class="docutils literal notranslate"><span class="pre">covariate_preprocessing</span></code> pipeline that contains genotype PC, phenotypic hidden confounders and fixed covariates.</p>
<section id="example-genotype-data">
<h3>Example genotype data<a class="headerlink" href="#example-genotype-data" title="Permalink to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr        path</span>
<span class="n">chr21</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21</span><span class="o">.</span><span class="n">bed</span>
<span class="n">chr22</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr22</span><span class="o">.</span><span class="n">bed</span>
</pre></div>
</div>
<p>Alternatively, simply use <code class="docutils literal notranslate"><span class="pre">protocol_example.genotype.chr21_22.bed</span></code> if all chromosomes are in the same file.</p>
</section>
<section id="example-phenotype-list">
<h3>Example phenotype list<a class="headerlink" href="#example-phenotype-list" title="Permalink to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start   end ID  path</span>
<span class="n">chr12</span>   <span class="mi">752578</span>  <span class="mi">752579</span>  <span class="n">ENSG00000060237</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">990508</span>  <span class="mi">990509</span>  <span class="n">ENSG00000082805</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">2794969</span> <span class="mi">2794970</span> <span class="n">ENSG00000004478</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">4649113</span> <span class="mi">4649114</span> <span class="n">ENSG00000139180</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6124769</span> <span class="mi">6124770</span> <span class="n">ENSG00000110799</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6534516</span> <span class="mi">6534517</span> <span class="n">ENSG00000111640</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
<section id="example-association-window-file">
<h3>Example association-window file<a class="headerlink" href="#example-association-window-file" title="Permalink to this heading">#</a></h3>
<p>It should have strictly 4 columns, with the header a commented out line:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start    end    gene_id</span>
<span class="n">chr10</span>   <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000008128</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000008130</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000067606</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">7101193</span>    <span class="n">ENSG00000069424</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">7960000</span>    <span class="n">ENSG00000069812</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000078369</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000078808</span>
</pre></div>
</div>
<p>The key is that the 4th column ID should match with the 4th column ID in the phenotype list. Otherwise the association-window to analyze will not be found.</p>
</section>
<section id="about-indels">
<h3>About indels<a class="headerlink" href="#about-indels" title="Permalink to this heading">#</a></h3>
<p>Option <code class="docutils literal notranslate"><span class="pre">--no-indel</span></code> will remove indel from analysis. FIXME: Gao need to provide more guidelines how to deal with indels in practice.</p>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading">#</a></h2>
<p>For each analysis region, the output is SuSiE model fitted and saved in RDS format.</p>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Permalink to this heading">#</a></h2>
<section id="i-susie-with-twas-weights-from-multiple-methods">
<h3>i. SuSiE with TWAS weights from multiple methods<a class="headerlink" href="#i-susie-with-twas-weights-from-multiple-methods" title="Permalink to this heading">#</a></h3>
<p>Timing [FIXME]</p>
<p>Below we duplicate the examples for phenotype and covariates to demonstrate that when there are multiple phenotypes for the same genotype it is possible to use this pipeline to analyze all of them (more than two is accepted as well).</p>
<p>Here using <code class="docutils literal notranslate"><span class="pre">--region-name</span></code> we focus the analysis on 3 genes. In practice if this parameter is dropped, the union of all regions in all phenotype region lists will be analyzed. It is possible for some of the regions there are no genotype data, in which case the pipeline will output RDS files with a warning message to indicate the lack of genotype data to analyze.</p>
<p><strong>Note:</strong> Suggested output naming convention is cohort_modality, eg ROSMAP_snRNA_pseudobulk.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span>  \
    <span class="o">--</span><span class="kp">name</span> <span class="n">protocol_example_protein</span>  \
    <span class="o">--</span><span class="n">genoFile</span> <span class="nb">input</span><span class="o">/</span><span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">bed</span>   \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
                <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
              <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span>  \
    <span class="o">--</span><span class="n">customized</span><span class="o">-</span><span class="n">association</span><span class="o">-</span><span class="n">windows</span> <span class="nb">input</span><span class="o">/</span><span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">enhanced_cis_chr21_chr22</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="kp">name</span> <span class="n">ENSG00000241973_P42356</span> <span class="n">ENSG00000160209_O00764</span> <span class="n">ENSG00000100412_Q99798</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">names</span> <span class="n">trait_A</span> <span class="n">trait_B</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">pecotmr_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
</div>
</div>
<p>It is also possible to analyze a selected list of regions using option <code class="docutils literal notranslate"><span class="pre">--region-list</span></code>. The last column of this file will be used for the list to analyze. Here for example use the same list of regions as we used for customized association-window:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span>  \
    <span class="o">--</span><span class="kp">name</span> <span class="n">protocol_example_protein</span>  \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">bed</span>   \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
                <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
              <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span>  \
    <span class="o">--</span><span class="n">customized</span><span class="o">-</span><span class="n">association</span><span class="o">-</span><span class="n">windows</span> <span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">enhanced_cis_chr21_chr22</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">enhanced_cis_chr21_chr22</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">names</span> <span class="n">trait_A</span> <span class="n">trait_B</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">pecotmr_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Note:</strong> When both <code class="docutils literal notranslate"><span class="pre">--region-name</span></code> and <code class="docutils literal notranslate"><span class="pre">--region-list</span></code> are used, the union of regions from these parameters will be analyzed.</p>
<p>FIXME: We should probably just explain these parameters, will work better for conversion script</p>
<p>To perform fine-mapping only without TWAS weights,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span> <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">twas</span><span class="o">-</span><span class="n">weights</span> <span class="o">...</span> <span class="c1"># rest of parameters the same. </span>
</pre></div>
</div>
<p>To perform fine-mapping and TWAS weights without cross validation,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span> <span class="o">--</span><span class="n">twas</span><span class="o">-</span><span class="n">cv</span><span class="o">-</span><span class="n">folds</span> <span class="mi">0</span> <span class="o">...</span> <span class="c1"># rest of parameters the same. </span>
</pre></div>
</div>
<p>It is also possible to specify a subset of samples to analyze, using <code class="docutils literal notranslate"><span class="pre">--keep-samples</span></code> parameter. For example we create a file to keep the ID of 50 samples,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">zcat</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">awk</span> <span class="s1">&#39;{for (i=2; i&lt;=51; i++) printf $i &quot; &quot;; print &quot;&quot;}&#39;</span><span class="o">&gt;</span> <span class="n">output</span><span class="o">/</span><span class="n">keep_samples</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
</div>
</div>
<p>then use them in our analysis,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span> <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">samples</span> <span class="n">output</span><span class="o">/</span><span class="n">keep_samples</span><span class="o">.</span><span class="n">txt</span> <span class="o">...</span> <span class="c1"># rest of parameters the same</span>
</pre></div>
</div>
</section>
<section id="ii-fsusie">
<h3>ii. fSuSiE<a class="headerlink" href="#ii-fsusie" title="Permalink to this heading">#</a></h3>
<p>Timing [FIXME]</p>
<p><strong>Note:</strong> Suggested output naming convention is cohort_modality, eg ROSMAP_snRNA_pseudobulk.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fsusie</span> \
    <span class="o">--</span><span class="kp">name</span> <span class="n">protocol_example_methylation</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">plink_per_chrom</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype_by_region</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">phenotype_by_region_files</span><span class="o">.</span><span class="n">txt</span> \
                <span class="n">output</span><span class="o">/</span><span class="n">phenotype_by_region</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">phenotype_by_region_files</span><span class="o">.</span><span class="n">txt</span>  \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
              <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">pecotmr_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="iii-mvsusie-with-twas-weights-from-mr-mash">
<h3>iii. mvSuSiE with TWAS weights from mr.mash<a class="headerlink" href="#iii-mvsusie-with-twas-weights-from-mr-mash" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">mnm</span> \
    <span class="o">--</span><span class="kp">name</span> <span class="n">multi_trait</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">FunGen_xQTL</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">Genotype</span><span class="o">/</span><span class="n">geno_by_chrom</span><span class="o">/</span><span class="n">ROSMAP_NIA_WGS</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">bcftools_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="mf">.11</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">wanggroup</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">pseudo_bulk_eqtl</span><span class="o">/</span><span class="n">Mic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">phenotype_data</span><span class="o">/</span><span class="n">Mic</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
      <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">wanggroup</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">pseudo_bulk_eqtl</span><span class="o">/</span><span class="n">Ast</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">phenotype_data</span><span class="o">/</span><span class="n">Ast</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
      <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">wanggroup</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">pseudo_bulk_eqtl</span><span class="o">/</span><span class="n">Oli</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">phenotype_data</span><span class="o">/</span><span class="n">Oli</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">covFile</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">wanggroup</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">pseudo_bulk_eqtl</span><span class="o">/</span><span class="n">Mic</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">covariate_data</span><span class="o">/</span><span class="n">Mic</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">Mic</span><span class="o">.</span><span class="n">rosmap_cov</span><span class="o">.</span><span class="n">ROSMAP_NIA_WGS</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">bcftools_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">snuc_pseudo_bulk</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">extracted</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
      <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">wanggroup</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">pseudo_bulk_eqtl</span><span class="o">/</span><span class="n">Ast</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">covariate_data</span><span class="o">/</span><span class="n">Ast</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">Ast</span><span class="o">.</span><span class="n">rosmap_cov</span><span class="o">.</span><span class="n">ROSMAP_NIA_WGS</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">bcftools_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">snuc_pseudo_bulk</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">extracted</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
      <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">wanggroup</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">pseudo_bulk_eqtl</span><span class="o">/</span><span class="n">Oli</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">data_preprocessing</span><span class="o">/</span><span class="n">covariate_data</span><span class="o">/</span><span class="n">Oli</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">Oli</span><span class="o">.</span><span class="n">rosmap_cov</span><span class="o">.</span><span class="n">ROSMAP_NIA_WGS</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">bcftools_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">snuc_pseudo_bulk</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">extracted</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">customized</span><span class="o">-</span><span class="n">association</span><span class="o">-</span><span class="n">windows</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">cl4215</span><span class="o">/</span><span class="n">mrmash</span><span class="o">/</span><span class="n">workflow</span><span class="o">/</span><span class="n">susie_twas</span><span class="o">/</span><span class="n">TADB_enhanced_cis</span><span class="o">.</span><span class="n">coding</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span> \
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="kp">name</span> <span class="n">ENSG00000073921</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">names</span> <span class="n">Mic_DeJager_eQTL</span> <span class="n">Ast_DeJager_eQTL</span> <span class="n">Oli_DeJager_eQTL</span> \
    <span class="o">--</span><span class="n">mixture_prior</span> <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">aw3600</span><span class="o">/</span><span class="n">MR_KMT_analysis</span><span class="o">/</span><span class="n">PTWAS</span><span class="o">/</span><span class="n">twas_mr_test</span><span class="o">/</span><span class="n">mash_test</span><span class="o">/</span><span class="n">mixture_prior_example</span><span class="o">.</span><span class="n">EZ</span><span class="o">.</span><span class="n">prior</span><span class="o">.</span><span class="n">rds</span> \
    <span class="o">--</span><span class="n">max_cv_variants</span> <span class="mi">1000</span> \
    <span class="o">--</span><span class="n">ld_reference_meta_file</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">data_public</span><span class="o">/</span><span class="mi">20240120</span><span class="n">_ADSP_LD_matrix</span><span class="o">/</span><span class="n">ld_meta_file</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Permalink to this heading">#</a></h2>
<table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Substep</p></th>
<th class="head"><p>Problem</p></th>
<th class="head"><p>Possible Reason</p></th>
<th class="head"><p>Solution</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="workflow-implementation">
<h2>Workflow implementation<a class="headerlink" href="#workflow-implementation" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># It is required to input the name of the analysis</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># A list of file paths for genotype data, or the genotype data itself. </span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># One or multiple lists of file paths for phenotype data.</span>
<span class="kn">parameter:</span> <span class="n">phenoFile</span> <span class="o">=</span> <span class="n">paths</span>
<span class="c1"># One or multiple lists of file paths for phenotype ID mapping file. The first column should be the original ID, the 2nd column should be the ID to be mapped to.</span>
<span class="kn">parameter:</span> <span class="n">phenoIDFile</span> <span class="o">=</span> <span class="n">paths</span><span class="p">()</span>
<span class="c1"># Covariate file path</span>
<span class="kn">parameter:</span> <span class="n">covFile</span> <span class="o">=</span> <span class="n">paths</span>
<span class="c1"># Optional: if a region list is provide the analysis will be focused on provided region. </span>
<span class="c1"># The LAST column of this list will contain the ID of regions to focus on</span>
<span class="c1"># Otherwise, all regions with both genotype and phenotype files will be analyzed</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Optional: if a region name is provided </span>
<span class="c1"># the analysis would be focused on the union of provides region list and region names</span>
<span class="kn">parameter:</span> <span class="n">region_name</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># Only focus on a subset of samples</span>
<span class="kn">parameter:</span> <span class="n">keep_samples</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># An optional list documenting the custom association window for each region to analyze, with four column, chr, start, end, region ID (eg gene ID).</span>
<span class="c1"># If this list is not provided, the default `window` parameter (see below) will be used.</span>
<span class="kn">parameter:</span> <span class="n">customized_association_windows</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Specify the cis window for the up and downstream radius to analyze around the region of interest in units of bp</span>
<span class="c1"># When this is set to negative, we will rely on using customized_association_windows</span>
<span class="kn">parameter:</span> <span class="n">cis_window</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="c1"># save data object or not</span>
<span class="kn">parameter:</span> <span class="n">save_data</span> <span class="o">=</span> <span class="kc">False</span>
<span class="c1"># Name of phenotypes</span>
<span class="kn">parameter:</span> <span class="n">phenotype_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">phenoFile</span><span class="p">]</span>
<span class="c1"># And indicator whether it is trans-analysis, ie, not using phenotypic coordinate information</span>
<span class="kn">parameter:</span> <span class="n">trans_analysis</span> <span class="o">=</span> <span class="kc">False</span>
<span class="kn">parameter:</span> <span class="n">seed</span> <span class="o">=</span> <span class="mi">999</span>

<span class="c1"># association analysis paramters</span>
<span class="c1"># initial number of single effects for SuSiE</span>
<span class="kn">parameter:</span> <span class="n">init_L</span> <span class="o">=</span> <span class="mi">8</span>
<span class="c1"># maximum number of single effects to use for SuSiE</span>
<span class="kn">parameter:</span> <span class="n">max_L</span> <span class="o">=</span> <span class="mi">30</span>
<span class="c1"># remove a variant if it has more than imiss missing individual level data</span>
<span class="kn">parameter:</span> <span class="n">imiss</span> <span class="o">=</span> <span class="mf">1.0</span>
<span class="c1"># MAF cutoff</span>
<span class="kn">parameter:</span> <span class="n">maf</span> <span class="o">=</span> <span class="mf">0.0025</span>
<span class="c1"># MAC cutoff, on top of MAF cutoff</span>
<span class="kn">parameter:</span> <span class="n">mac</span> <span class="o">=</span> <span class="mi">5</span>
<span class="c1"># Remove indels if indel = False</span>
<span class="kn">parameter:</span> <span class="n">indel</span> <span class="o">=</span> <span class="kc">True</span>
<span class="kn">parameter:</span> <span class="n">pip_cutoff</span> <span class="o">=</span> <span class="mf">0.025</span>
<span class="kn">parameter:</span> <span class="n">coverage</span> <span class="o">=</span> <span class="p">[</span><span class="mf">0.95</span><span class="p">,</span> <span class="mf">0.7</span><span class="p">,</span> <span class="mf">0.5</span><span class="p">]</span>
<span class="c1"># If this value is greater than 0, an initial single effect analysis will be performed </span>
<span class="c1"># to determine if follow up analysis will be continued or to simply return NULL</span>
<span class="kn">parameter:</span> <span class="n">skip_analysis_pip_cutoff</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># Perform Fine-mapping</span>
<span class="kn">parameter:</span> <span class="n">fine_mapping</span> <span class="o">=</span> <span class="kc">True</span>
<span class="c1"># Compute TWAS weights as well</span>
<span class="kn">parameter:</span> <span class="n">twas_weights</span> <span class="o">=</span> <span class="kc">True</span>
<span class="c1"># Perform K folds valiation CV for TWAS</span>
<span class="c1"># Set it to zero if this is to be skipped</span>
<span class="kn">parameter:</span> <span class="n">twas_cv_folds</span> <span class="o">=</span> <span class="mi">5</span>
<span class="kn">parameter:</span> <span class="n">twas_cv_threads</span> <span class="o">=</span> <span class="n">twas_cv_folds</span>
<span class="c1"># maximum number of variants to consider for CV</span>
<span class="c1"># We will randomly pick a subset of it for CV purpose</span>
<span class="c1"># We can set it to eg 5000 to save computational burden althought may risk overfitting for methods comparison purpose</span>
<span class="c1"># When set to -1 we don&#39;t use this feature</span>
<span class="kn">parameter:</span> <span class="n">max_cv_variants</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>
<span class="c1"># Further limit CV to only using common variants</span>
<span class="kn">parameter:</span> <span class="n">min_cv_maf</span> <span class="o">=</span> <span class="mf">0.05</span>
<span class="kn">parameter:</span> <span class="n">ld_reference_meta_file</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Analysis environment settings</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">200</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;1h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;20G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">1</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">skip_analysis_pip_cutoff</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">skip_analysis_pip_cutoff</span> <span class="o">=</span> <span class="p">[</span><span class="o">-</span><span class="mf">1.0</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">skip_analysis_pip_cutoff</span><span class="p">)</span> <span class="o">==</span> <span class="mi">1</span><span class="p">:</span>
    <span class="n">skip_analysis_pip_cutoff</span> <span class="o">=</span> <span class="n">skip_analysis_pip_cutoff</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">skip_analysis_pip_cutoff</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;``skip_analysis_pip_cutoff`` should have either length 1 or length the same as phenotype files (</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)</span><span class="si">}</span><span class="s2"> in this case)&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">group_by_region</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">partition</span><span class="p">):</span>
    <span class="c1"># from itertools import accumulate</span>
    <span class="c1"># partition = [len(x) for x in partition]</span>
    <span class="c1"># Compute the cumulative sums once</span>
    <span class="c1"># cumsum_vector = list(accumulate(partition))</span>
    <span class="c1"># Use slicing based on the cumulative sums</span>
    <span class="c1"># return [lst[(cumsum_vector[i-1] if i &gt; 0 else 0):cumsum_vector[i]] for i in range(len(partition))]</span>
    <span class="k">return</span> <span class="n">partition</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="k">def</span> <span class="nf">adapt_file_path</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">reference_file</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adapt a single file path based on its existence and a reference file&#39;s path.</span>

<span class="sd">    Args:</span>
<span class="sd">    - file_path (str): The file path to adapt.</span>
<span class="sd">    - reference_file (str): File path to use as a reference for adaptation.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - str: Adapted file path.</span>

<span class="sd">    Raises:</span>
<span class="sd">    - FileNotFoundError: If no valid file path is found.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">reference_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">reference_file</span><span class="p">)</span>

    <span class="c1"># Check if the file exists</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">file_path</span>

    <span class="c1"># Check file name without path</span>
    <span class="n">file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_name</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">file_name</span>

    <span class="c1"># Check file name in reference file&#39;s directory</span>
    <span class="n">file_in_ref_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reference_path</span><span class="p">,</span> <span class="n">file_name</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_in_ref_dir</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">file_in_ref_dir</span>

    <span class="c1"># Check original file path prefixed with reference file&#39;s directory</span>
    <span class="n">file_prefixed</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reference_path</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_prefixed</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">file_prefixed</span>

    <span class="c1"># If all checks fail, raise an error</span>
    <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No valid path found for file: </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">adapt_file_path_all</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">column_name</span><span class="p">,</span> <span class="n">reference_file</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">df</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">adapt_file_path</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">reference_file</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">get_analysis_regions</span><span class="p">:</span> <span class="kp">shared</span> <span class="o">=</span> <span class="s2">&quot;regional_data&quot;</span><span class="p">]</span>
<span class="c1"># input is genoFile, phenoFile, covFile and optionally region_list. If region_list presents then we only analyze what&#39;s contained in the list.</span>
<span class="c1"># regional_data should be a dictionary like:</span>
<span class="c1">#{&#39;data&#39;: [(&quot;genotype_1.bed&quot;, &quot;phenotype_1.bed.gz&quot;, &quot;covariate_1.gz&quot;), (&quot;genotype_2.bed&quot;, &quot;phenotype_1.bed.gz&quot;, &quot;phenotype_2.bed.gz&quot;, &quot;covariate_1.gz&quot;, &quot;covariate_2.gz&quot;) ... ],</span>
<span class="c1"># &#39;meta_info&#39;: [(&quot;chr12:752578-752579&quot;,&quot;chr12:752577-752580&quot;, &quot;gene_1&quot;, &quot;trait_1&quot;), (&quot;chr13:852580-852581&quot;,&quot;chr13:852579-852580&quot;, &quot;gene_2&quot;, &quot;trait_1&quot;, &quot;trait_2&quot;) ... ]}</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>

<span class="k">def</span> <span class="nf">preload_id_map</span><span class="p">(</span><span class="n">id_map_files</span><span class="p">):</span>
    <span class="n">id_maps</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">id_map_file</span> <span class="ow">in</span> <span class="n">id_map_files</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">id_map_file</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span> <span class="ow">and</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">id_map_file</span><span class="p">):</span>
            <span class="n">df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">id_map_file</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;old_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;new_ID&#39;</span><span class="p">])</span>
            <span class="n">id_maps</span><span class="p">[</span><span class="n">id_map_file</span><span class="p">]</span> <span class="o">=</span> <span class="n">df</span><span class="o">.</span><span class="n">set_index</span><span class="p">(</span><span class="s1">&#39;old_ID&#39;</span><span class="p">)[</span><span class="s1">&#39;new_ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">id_maps</span>

<span class="k">def</span> <span class="nf">load_and_apply_id_map</span><span class="p">(</span><span class="n">pheno_path</span><span class="p">,</span> <span class="n">id_map_path</span><span class="p">,</span> <span class="n">preloaded_id_maps</span><span class="p">):</span>
    <span class="n">pheno_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">pheno_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">pheno_df</span><span class="p">[</span><span class="s1">&#39;Original_ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pheno_df</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">id_map_path</span> <span class="ow">in</span> <span class="n">preloaded_id_maps</span><span class="p">:</span>
        <span class="n">id_map</span> <span class="o">=</span> <span class="n">preloaded_id_maps</span><span class="p">[</span><span class="n">id_map_path</span><span class="p">]</span>
        <span class="n">pheno_df</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">pheno_df</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">map</span><span class="p">(</span><span class="n">id_map</span><span class="p">)</span><span class="o">.</span><span class="n">fillna</span><span class="p">(</span><span class="n">pheno_df</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">])</span>
    <span class="k">return</span> <span class="n">pheno_df</span>

<span class="k">def</span> <span class="nf">filter_by_region_ids</span><span class="p">(</span><span class="n">data</span><span class="p">,</span> <span class="n">region_ids</span><span class="p">):</span>
    <span class="k">if</span> <span class="n">region_ids</span><span class="p">:</span>
        <span class="n">data</span> <span class="o">=</span> <span class="n">data</span><span class="p">[</span><span class="n">data</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)]</span>
    <span class="k">return</span> <span class="n">data</span>

<span class="k">def</span> <span class="nf">custom_join</span><span class="p">(</span><span class="n">series</span><span class="p">):</span>
    <span class="c1"># Initialize an empty list to hold the processed items</span>
    <span class="n">result</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">series</span><span class="p">:</span>
        <span class="k">if</span> <span class="s1">&#39;,&#39;</span> <span class="ow">in</span> <span class="n">item</span><span class="p">:</span>
            <span class="c1"># If the item contains commas, split by comma and convert to tuple</span>
            <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">tuple</span><span class="p">(</span><span class="n">item</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">)))</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># If the item does not contain commas, add it directly</span>
            <span class="n">result</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">item</span><span class="p">)</span>
    <span class="c1"># Convert the list of items to a tuple and return</span>
    <span class="k">return</span> <span class="nb">tuple</span><span class="p">(</span><span class="n">result</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">aggregate_phenotype_data</span><span class="p">(</span><span class="n">accumulated_pheno_df</span><span class="p">):</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">accumulated_pheno_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="n">accumulated_pheno_df</span> <span class="o">=</span> <span class="n">accumulated_pheno_df</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span><span class="s1">&#39;cond&#39;</span><span class="p">,</span><span class="s1">&#39;path&#39;</span><span class="p">,</span><span class="s1">&#39;cov_path&#39;</span><span class="p">],</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span>
            <span class="s1">&#39;#chr&#39;</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
            <span class="s1">&#39;ID&#39;</span><span class="p">:</span> <span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">str</span><span class="p">)[</span><span class="mi">0</span><span class="p">],</span>
            <span class="s1">&#39;Original_ID&#39;</span><span class="p">:</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">,</span>
            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="s1">&#39;min&#39;</span><span class="p">,</span>
            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="s1">&#39;max&#39;</span>
        <span class="p">})</span><span class="o">.</span><span class="n">groupby</span><span class="p">([</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span><span class="s1">&#39;ID&#39;</span><span class="p">],</span> <span class="n">as_index</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span><span class="o">.</span><span class="n">agg</span><span class="p">({</span>
            <span class="s1">&#39;cond&#39;</span><span class="p">:</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">,</span>
            <span class="s1">&#39;path&#39;</span><span class="p">:</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">,</span>
            <span class="s1">&#39;Original_ID&#39;</span><span class="p">:</span> <span class="n">custom_join</span><span class="p">,</span>
            <span class="s1">&#39;cov_path&#39;</span><span class="p">:</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">,</span>
            <span class="s1">&#39;start&#39;</span><span class="p">:</span> <span class="s1">&#39;min&#39;</span><span class="p">,</span>
            <span class="s1">&#39;end&#39;</span><span class="p">:</span> <span class="s1">&#39;max&#39;</span>
        <span class="p">})</span>
    <span class="k">return</span> <span class="n">accumulated_pheno_df</span>

<span class="k">def</span> <span class="nf">process_cis_files</span><span class="p">(</span><span class="n">pheno_files</span><span class="p">,</span> <span class="n">cov_files</span><span class="p">,</span> <span class="n">phenotype_names</span><span class="p">,</span> <span class="n">pheno_id_files</span><span class="p">,</span> <span class="n">region_ids</span><span class="p">,</span> <span class="n">preloaded_id_maps</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Example output:</span>
<span class="sd">    #chr    start      end    ID  Original_ID   path     cov_path             cond</span>
<span class="sd">    chr12   752578   752579  ENSG00000060237  Q9H4A3,P62873  protocol_example.protein_1.bed.gz,protocol_example.protein_2.bed.gz  covar_1.gz,covar_2.gz  trait_A,trait_B</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">accumulated_pheno_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="n">pheno_id_files</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">pheno_files</span><span class="p">)</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">pheno_id_files</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">pheno_id_files</span>

    <span class="k">for</span> <span class="n">pheno_path</span><span class="p">,</span> <span class="n">cov_path</span><span class="p">,</span> <span class="n">phenotype_name</span><span class="p">,</span> <span class="n">id_map_path</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">pheno_files</span><span class="p">,</span> <span class="n">cov_files</span><span class="p">,</span> <span class="n">phenotype_names</span><span class="p">,</span> <span class="n">pheno_id_files</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">cov_path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No valid path found for file: </span><span class="si">{</span><span class="n">cov_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">pheno_df</span> <span class="o">=</span> <span class="n">load_and_apply_id_map</span><span class="p">(</span><span class="n">pheno_path</span><span class="p">,</span> <span class="n">id_map_path</span><span class="p">,</span> <span class="n">preloaded_id_maps</span><span class="p">)</span>
        <span class="n">pheno_df</span> <span class="o">=</span> <span class="n">filter_by_region_ids</span><span class="p">(</span><span class="n">pheno_df</span><span class="p">,</span> <span class="n">region_ids</span><span class="p">)</span>
        
        <span class="k">if</span> <span class="ow">not</span> <span class="n">pheno_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">pheno_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">4</span><span class="p">]</span> <span class="o">=</span> <span class="n">adapt_file_path_all</span><span class="p">(</span><span class="n">pheno_df</span><span class="p">,</span> <span class="n">pheno_df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pheno_path</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">pheno_df</span> <span class="o">=</span> <span class="n">pheno_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">cov_path</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">cov_path</span><span class="p">),</span> <span class="n">cond</span><span class="o">=</span><span class="n">phenotype_name</span><span class="p">)</span>           
            <span class="n">accumulated_pheno_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">accumulated_pheno_df</span><span class="p">,</span> <span class="n">pheno_df</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">accumulated_pheno_df</span> <span class="o">=</span> <span class="n">aggregate_phenotype_data</span><span class="p">(</span><span class="n">accumulated_pheno_df</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">accumulated_pheno_df</span>

<span class="k">def</span> <span class="nf">process_trans_files</span><span class="p">(</span><span class="n">pheno_files</span><span class="p">,</span> <span class="n">cov_files</span><span class="p">,</span> <span class="n">phenotype_names</span><span class="p">,</span> <span class="n">pheno_id_files</span><span class="p">,</span> <span class="n">region_ids</span><span class="p">,</span> <span class="n">customized_association_windows</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Example output:</span>
<span class="sd">    #chr    start      end    ID  Original_ID   path     cov_path             cond</span>
<span class="sd">    chr21   0   0  chr21_18133254_19330300  carnitine,benzoate,hippurate  metabolon_1.bed.gz,metabolon_2.bed.gz  covar_1.gz,covar_2.gz  trait_A,trait_B</span>
<span class="sd">    &#39;&#39;&#39;</span>
    
    <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">customized_association_windows</span><span class="p">):</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Customized association analysis window must be specified for trans analysis.&quot;</span><span class="p">)</span>
    <span class="n">accumulated_pheno_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>
    <span class="n">pheno_id_files</span> <span class="o">=</span> <span class="p">[</span><span class="kc">None</span><span class="p">]</span> <span class="o">*</span> <span class="nb">len</span><span class="p">(</span><span class="n">pheno_files</span><span class="p">)</span> <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">pheno_id_files</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span> <span class="k">else</span> <span class="n">pheno_id_files</span>
    <span class="n">genotype_windows</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">customized_association_windows</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;#chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">,</span><span class="s2">&quot;ID&quot;</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">genotype_windows</span> <span class="o">=</span> <span class="n">filter_by_region_ids</span><span class="p">(</span><span class="n">genotype_windows</span><span class="p">,</span> <span class="n">region_ids</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">genotype_windows</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="k">return</span> <span class="n">accumulated_pheno_df</span>
    
    <span class="k">for</span> <span class="n">pheno_path</span><span class="p">,</span> <span class="n">cov_path</span><span class="p">,</span> <span class="n">phenotype_name</span><span class="p">,</span> <span class="n">id_map_path</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">pheno_files</span><span class="p">,</span> <span class="n">cov_files</span><span class="p">,</span> <span class="n">phenotype_names</span><span class="p">,</span> <span class="n">pheno_id_files</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">cov_path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No valid path found for file: </span><span class="si">{</span><span class="n">cov_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">pheno_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">pheno_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;\s+&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;Original_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;path&#39;</span><span class="p">])</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">pheno_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">pheno_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">adapt_file_path_all</span><span class="p">(</span><span class="n">pheno_df</span><span class="p">,</span> <span class="n">pheno_df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">],</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pheno_path</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
            <span class="n">pheno_df</span> <span class="o">=</span> <span class="n">pheno_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">cov_path</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">cov_path</span><span class="p">),</span> <span class="n">cond</span><span class="o">=</span><span class="n">phenotype_name</span><span class="p">)</span>
            <span class="c1"># Here we combine genotype_windows which contains &quot;#chr&quot; and &quot;ID&quot; to pheno_df by creating a cartesian product</span>
            <span class="n">pheno_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">genotype_windows</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">pheno_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span><span class="n">key</span><span class="o">=</span><span class="mi">1</span><span class="p">),</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;key&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">drop</span><span class="p">(</span><span class="s1">&#39;key&#39;</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>
            <span class="c1"># then set start and end columns to zero</span>
            <span class="n">pheno_df</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="n">pheno_df</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">if</span> <span class="n">id_map_path</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                <span class="c1"># Filter pheno_df by specific association-window and phenotype pairs</span>
                <span class="n">association_analysis_pair</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">id_map_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;\s+&#39;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s1">&#39;#&#39;</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Original_ID&#39;</span><span class="p">])</span>
                <span class="n">pheno_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">pheno_df</span><span class="p">,</span> <span class="n">association_analysis_pair</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">,</span> <span class="s1">&#39;Original_ID&#39;</span><span class="p">])</span>
            <span class="n">accumulated_pheno_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">accumulated_pheno_df</span><span class="p">,</span> <span class="n">pheno_df</span><span class="p">],</span> <span class="n">ignore_index</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

    <span class="n">accumulated_pheno_df</span> <span class="o">=</span> <span class="n">aggregate_phenotype_data</span><span class="p">(</span><span class="n">accumulated_pheno_df</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">accumulated_pheno_df</span>

<span class="c1"># Load phenotype meta data</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">covFile</span><span class="p">):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of input phenotypes files must match that of covariates files&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenotype_names</span><span class="p">):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of input phenotypes files must match the number of phenotype names&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoIDFile</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span> <span class="ow">and</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoIDFile</span><span class="p">):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of input phenotypes files must match the number of phenotype ID mapping files&quot;</span><span class="p">)</span>

<span class="c1"># Load genotype meta data</span>
<span class="k">if</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">:</span><span class="s2">x</span><span class="si">}</span><span class="s2">&quot;</span> <span class="o">==</span> <span class="s2">&quot;.bed&quot;</span><span class="p">:</span>
    <span class="n">geno_meta_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([(</span><span class="s2">&quot;chr&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[(</span><span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="s1">&#39;geno_path&#39;</span><span class="p">])</span>
<span class="kn">else:</span>
    <span class="n">geno_meta_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;\s+&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">geno_meta_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">adapt_file_path_all</span><span class="p">(</span><span class="n">geno_meta_data</span><span class="p">,</span> <span class="n">geno_meta_data</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">geno_meta_data</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="s1">&#39;geno_path&#39;</span><span class="p">]</span>
    <span class="n">geno_meta_data</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">geno_meta_data</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="sa">f</span><span class="s1">&#39;chr</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="c1"># Checking the DataFrame</span>
<span class="n">valid_chr_values</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;chr</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;chrX&#39;</span><span class="p">]</span>
<span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">value</span> <span class="ow">in</span> <span class="n">valid_chr_values</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">geno_meta_data</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid chromosome values found. Allowed values are chr1 to chr22 and chrX.&quot;</span><span class="p">)</span>

<span class="n">region_ids</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># If region_list is provided, read the file and extract IDs</span>
<span class="k">if</span> <span class="n">region_list</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
    <span class="n">region_list_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">region_list</span><span class="p">,</span> <span class="n">delim_whitespace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">comment</span> <span class="o">=</span> <span class="s2">&quot;#&quot;</span><span class="p">)</span>
    <span class="n">region_ids</span> <span class="o">=</span> <span class="n">region_list_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>  <span class="c1"># Extracting the last column for IDs</span>

<span class="c1"># If region_name is provided, include those IDs as well</span>
<span class="c1"># --region-name A B C will result in a list of [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;] here</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">region_name</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">region_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">region_name</span><span class="p">)))</span>

<span class="k">if</span> <span class="n">trans_analysis</span><span class="p">:</span>
    <span class="n">meta_data</span> <span class="o">=</span> <span class="n">process_trans_files</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">,</span> <span class="n">covFile</span><span class="p">,</span> <span class="n">phenotype_names</span><span class="p">,</span> <span class="n">phenoIDFile</span><span class="p">,</span> <span class="n">region_ids</span><span class="p">,</span> <span class="n">customized_association_windows</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="n">meta_data</span> <span class="o">=</span> <span class="n">process_cis_files</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">,</span> <span class="n">covFile</span><span class="p">,</span> <span class="n">phenotype_names</span><span class="p">,</span> <span class="n">phenoIDFile</span><span class="p">,</span> <span class="n">region_ids</span><span class="p">,</span> <span class="n">preload_id_map</span><span class="p">(</span><span class="n">phenoIDFile</span><span class="p">))</span>
    
<span class="k">if</span> <span class="ow">not</span> <span class="n">meta_data</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
    <span class="n">meta_data</span> <span class="o">=</span> <span class="n">meta_data</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">geno_meta_data</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">)</span>
    <span class="c1"># Adjust association-window</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">customized_association_windows</span><span class="p">):</span>
        <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loading customized association analysis window from </span><span class="si">{</span><span class="n">customized_association_windows</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">association_windows_list</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">customized_association_windows</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;#chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">,</span><span class="s2">&quot;ID&quot;</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">meta_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">meta_data</span><span class="p">,</span> <span class="n">association_windows_list</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="s1">&#39;ID&#39;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;_association&#39;</span><span class="p">))</span>
        <span class="n">mismatches</span> <span class="o">=</span> <span class="n">meta_data</span><span class="p">[</span><span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;start_association&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">mismatches</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;First 5 mismatches:&quot;</span><span class="p">)</span>
            <span class="nb">print</span><span class="p">(</span><span class="n">mismatches</span><span class="p">[[</span><span class="s1">&#39;ID&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">mismatches</span><span class="p">)</span><span class="si">}</span><span class="s2"> regions to analyze cannot be found in ``</span><span class="si">{</span><span class="n">customized_association_windows</span><span class="si">}</span><span class="s2">``. Please check your ``</span><span class="si">{</span><span class="n">customized_association_windows</span><span class="si">}</span><span class="s2">`` database to make sure it contains all association-window definitions. &quot;</span><span class="p">)</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">cis_window</span> <span class="o">&lt;</span> <span class="mi">0</span> <span class="p">:</span>
            <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Please either input valid path to association-window file via ``--customized-association-windows``, or set ``--cis-window`` to a non-negative integer.&quot;</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">cis_window</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;Warning: only variants within the range of start and end of molecular phenotype will be considered since cis_window is set to zero and no customized association window file was found. Please make sure this is by design.&quot;</span><span class="p">)</span>
        <span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;start_association&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">max</span><span class="p">(</span><span class="n">x</span> <span class="o">-</span> <span class="n">cis_window</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>
        <span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;end_association&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">cis_window</span>

    <span class="c1"># Example meta_data:</span>
    <span class="c1"># #chr    start      end    start_association       end_association           ID  Original_ID   path     cov_path             cond             coordinate     geno_path</span>
    <span class="c1"># 0  chr12   752578   752579  652578   852579  ENSG00000060237  Q9H4A3,P62873  protocol_example.protein_1.bed.gz,protocol_example.protein_2.bed.gz  covar_1.gz,covar_2.gz  trait_A,trait_B    chr12:752578-752579  protocol_example.genotype.chr21_22.bed       </span>
    <span class="c1"># Create the final dictionary</span>
    <span class="n">regional_data</span> <span class="o">=</span> <span class="p">{</span>
        <span class="s1">&#39;data&#39;</span><span class="p">:</span> <span class="p">[(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;geno_path&#39;</span><span class="p">],</span> <span class="o">*</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;path&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">),</span> <span class="o">*</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;cov_path&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">meta_data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()],</span>
        <span class="s1">&#39;meta_info&#39;</span><span class="p">:</span> <span class="p">[(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="c1"># this is the phenotypic region to extract data from</span>
                       <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start_association&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;end_association&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="c1"># this is the association window region</span>
                       <span class="n">row</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">],</span> <span class="n">row</span><span class="p">[</span><span class="s1">&#39;Original_ID&#39;</span><span class="p">],</span> <span class="o">*</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;cond&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">meta_data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()]</span>
    <span class="p">}</span>
<span class="kn">else:</span>
    <span class="n">regional_data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;data&#39;</span><span class="p">:[],</span> <span class="s1">&#39;meta_info&#39;</span><span class="p">:[]}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="univariate-analysis-susie-and-twas">
<h2>Univariate analysis: SuSiE and TWAS<a class="headerlink" href="#univariate-analysis-susie-and-twas" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[susie_twas_1]
depends: sos_variable(&quot;regional_data&quot;)
# Check if both &#39;data&#39; and &#39;meta_info&#39; are empty lists
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39;Either genotype or phenotype data are not available for region {&quot;, &quot;.join(region_name)}.&#39;)

meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name[:-2]}/{name}.{_meta_info[2]}.univariate{&quot;_susie&quot; if fine_mapping else &quot;&quot;}{&quot;_twas_weights&quot; if twas_weights else &quot;&quot;}.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    options(warn=1)
    library(pecotmr)
    # extract subset of samples
    keep_samples = NULL
    if (${&quot;TRUE&quot; if keep_samples.is_file() else &quot;FALSE&quot;}) {
      keep_samples = unlist(strsplit(readLines(${keep_samples:ar}), &quot;\\s+&quot;))
      message(paste(length(keep_samples), &quot;samples are selected to be loaded for analysis&quot;))
    }
    # Load regional association data
    tryCatch({
    fdat = load_regional_univariate_data(genotype = ${_input[0]:anr},
                                          phenotype = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1:len(_input)//2+1]])}),
                                          covariate = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[len(_input)//2+1:]])}),
                                          region = ${(&quot;&#39;%s&#39;&quot; % _meta_info[0]) if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &#39;NULL&#39;}, # if the end position is zero return NULL
                                          association_window = &quot;${_meta_info[1]}&quot;,
                                          conditions = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in _meta_info[4:]])}),
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss},
                                          keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;},
                                          keep_samples = keep_samples,
                                          extract_region_name = list(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])}),
                                          phenotype_header = ${&quot;4&quot; if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;},
                                          region_name_col = ${&quot;4&quot; if int(_meta_info[0].split(&#39;-&#39;)[-1])&gt;0 else &quot;1&quot;},
                                          scale_residuals = FALSE)
    }, NoSNPsError = function(e) {
        message(&quot;Error: &quot;, paste(e$message, &quot;${_meta_info[2] + &#39;@&#39; + _meta_info[1]}&quot;))
        #saveRDS(NULL, ${_output:ar})
        saveRDS(list(${_meta_info[2]} = e$message), ${_output:ar}, compress=&#39;xz&#39;)
        quit(save=&quot;no&quot;)
    })
  
    if (${&quot;TRUE&quot; if not (fine_mapping or twas_weights) else &quot;FALSE&quot;}) {
        # only export data
        saveRDS(list(${_meta_info[2]} = fdat), ${_output:ar}, compress=&#39;xz&#39;)
        quit(save=&quot;no&quot;)
    } else {
        if (${&quot;TRUE&quot; if save_data else &quot;FALSE&quot;}) {
            # save data object for debug purpose
            saveRDS(list(${_meta_info[2]} = fdat), &quot;${_output:ann}.univariate.rds&quot;, compress=&#39;xz&#39;)
        }
    }
    
    # setup univariate analysis pipeline options
    if (&quot;${_meta_info[2]}&quot; != &quot;${_meta_info[3]}&quot;) {
        region_name = c(&quot;${_meta_info[2]}&quot;, c(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])}))
    } else {
        region_name = &quot;${_meta_info[2]}&quot;
    }
  
    region_info = list(region_coord=parse_region(&quot;${_meta_info[0]}&quot;), grange=parse_region(&quot;${_meta_info[1]}&quot;), region_name=region_name)
    finemapping_opts = list(init_L=${init_L}, max_L=${max_L}, l_step = 5, coverage=${&quot;,&quot;.join([str(x) for x in coverage])}, signal_cutoff=${pip_cutoff})
    twas_weights_opts = list(cv_folds=${twas_cv_folds}, min_cv_maf=${min_cv_maf}, max_cv_variants=${max_cv_variants}, seed=${seed}, cv_threads=${twas_cv_threads}, 
                              ld_reference_meta_file=${(&#39;&quot;%s&quot;&#39; % ld_reference_meta_file) if not ld_reference_meta_file.is_dir() else &quot;NULL&quot;})
    
    fitted = list()
    condition_names = vector()
    empty_elements_cnt = 0
    r = 1
    while (r&lt;=length(fdat$residual_Y)) {
        dropped_samples = list(X=fdat$dropped_sample$dropped_samples_X[[r]], 
                             y=fdat$dropped_sample$dropped_samples_Y[[r]], 
                             covar=fdat$dropped_sample$dropped_samples_covar[[r]])
        results &lt;- lapply(1:ncol(fdat$residual_Y[[r]]), function(i) univariate_analysis_pipeline(fdat$residual_X[[r]], 
                                                                                    fdat$residual_Y[[r]][,i,drop=FALSE], 
                                                                                    fdat$residual_X_scalar[[r]], 
                                                                                    if (fdat$residual_Y_scalar[[r]] == 1) 1 else fdat$residual_Y_scalar[[r]][,i,drop=FALSE], 
                                                                                    fdat$maf[[r]], list(dropped_samples = dropped_samples),
                                                                                    pip_cutoff_to_skip = c(${&quot;,&quot;.join([&#39;%s&#39; % x for x in skip_analysis_pip_cutoff])})[r],
                                                                                    finemapping = ${&quot;TRUE&quot; if fine_mapping else &quot;FALSE&quot;},
                                                                                    twas_weights = ${&quot;TRUE&quot; if twas_weights else &quot;FALSE&quot;},
                                                                                    finemapping_opts = finemapping_opts,
                                                                                    twas_weights_opts = twas_weights_opts,
                                                                                    region_info = region_info))
        # Update condition names
        new_names = names(fdat$residual_Y)[r]
        ##FIXME:  residule Y lost their colname when there was only 1 column
        # new_col_names = colnames(fdat$residual_Y[[r]])
        new_col_names = list(${&quot;,&quot;.join([(&quot;c(&#39;&quot;+x+&quot;&#39;)&quot;) if isinstance(x, str) else (&quot;c&quot;+ str(x)) for x in _meta_info[3]])})[[r]]
        if (is.null(new_col_names)) {
          # column names does not exist, create generic names instead
          new_col_names = 1:ncol(fdat$residual_Y[[r]])
        }
        # if the name is different as region name ie, these original ID the same as gene ID, then give the new name
        if(!(identical(new_names, new_col_names)))
        new_names = paste(new_names, new_col_names, sep=&quot;_&quot;) # DLPFC_iso1 DLPFC_iso2

        empty_elements_idx &lt;- which(sapply(results, function(x) is.list(x) &amp;&amp; length(x) == 0))
        if (length(empty_elements_idx)&gt;0) {
          empty_elements_cnt &lt;- empty_elements_cnt + length(empty_elements_idx)
          results &lt;- results[-empty_elements_idx]
          new_names &lt;- new_names[-empty_elements_idx]
        }
        fitted = c(fitted, results)
        condition_names = c(condition_names, new_names)
        if (length(new_names)&gt;0) {
          message(&quot;Analysis completed for: &quot;, paste(new_names, collapse=&quot;,&quot;))
        }
        # original data no longer relevant, set to NA to release memory
        fdat$residual_X[[r]] &lt;- NA
        fdat$residual_Y[[r]] &lt;- NA
        r = r + 1
    }
    names(fitted) &lt;- condition_names
    saveRDS(list(&quot;${_meta_info[2]}&quot; = fitted), ${_output:ar}, compress=&#39;xz&#39;)
    if (empty_elements_cnt&gt;0) {
        message(empty_elements_cnt, &quot; out of &quot;, length(fitted), &quot; analysis are skipped for failing to pass initial screen for potential association signals&quot;)
    }
</pre></div>
</div>
</div>
</div>
</section>
<section id="multivariate-analysis-mvsusie-and-mr-mash">
<h2>Multivariate analysis: mvSuSiE and mr.mash<a class="headerlink" href="#multivariate-analysis-mvsusie-and-mr-mash" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[mnm_1]
# Prior model file generated from mashr. 
# Default will be used if it does not exist.
parameter: mixture_prior = path()
parameter: mixture_prior_cv = path()
parameter: prior_weights_min = 5e-4
parameter: prior_canonical_matrices = True
parameter: sample_partition = path() 
parameter: mvsusie_max_iter = 200
parameter: mrmash_max_iter = 5000
depends: sos_variable(&quot;regional_data&quot;)
# Check if both &#39;data&#39; and &#39;meta_info&#39; are empty lists
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39;Either genotype or phenotype data are not available for region {&quot;, &quot;.join(region_name)}.&#39;)

meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name[:-2]}/{name}.{_meta_info[0].split(&quot;:&quot;)[0]}_{_meta_info[2]}.mnm.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint

    library(pecotmr) 
    fdat = load_regional_multivariate_data(genotype = ${_input[0]:anr},
                                          phenotype = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1:len(_input)//2+1]])}),
                                          covariate = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[len(_input)//2+1:]])}),
                                          region = &quot;${_meta_info[0]}&quot;,
                                          conditions = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in _meta_info[4:]])}),
                                          association_window = &quot;${_meta_info[1]}&quot;,
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss},
                                          keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;})
    
    dd_prior &lt;- if (${mixture_prior:r} == &#39;.&#39; || ${mixture_prior:r} == &#39;&#39;) NULL else readRDS(${mixture_prior:r})
    dd_prior_cv &lt;- if (${mixture_prior_cv:r} == &#39;.&#39; || ${mixture_prior_cv:r} == &#39;&#39;) NULL else readRDS(${mixture_prior_cv:r})
    
    result &lt;- multivariate_analysis_pipeline(
                    X=fdat$X,
                    Y=fdat$residual_Y,
                    maf=fdat$maf,
                    max_L=${max_L},
                    ld_reference_meta_file = ${ld_reference_meta_file:r}, 
                    mvsusie_max_iter = ${mvsusie_max_iter}, 
                    mrmash_max_iter = ${mrmash_max_iter},
                    max_cv_variants = ${max_cv_variants}, 
                    pip_cutoff_to_skip = c(${&quot;,&quot;.join([&#39;%s&#39; % x for x in skip_analysis_pip_cutoff])}),
                    signal_cutoff = ${signal_cutoff},
                    data_driven_prior_matrices = dd_prior, 
                    data_driven_prior_matrices_cv = dd_prior_cv, 
                    canonical_prior_matrices = ${&quot;TRUE&quot; if prior_canonical_matrices else &quot;FALSE&quot;} || is.null(dd_prior),
                    sample_partition = ${&quot;NULL&quot; if sample_partition==&#39;.&#39; or sample_partition==&#39;&#39; else sample_partition},  
                    cv_folds = ${twas_cv_folds},
                    cv_seed = ${seed}, 
                    cv_threads = ${twas_cv_threads},
                    prior_weights_min = ${prior_weights_min}, 
                    twas_weights = ${&quot;TRUE&quot; if twas_weights else &quot;FALSE&quot;}
    )
  
    saveRDS(result, ${_output:ar})
</pre></div>
</div>
</div>
</div>
</section>
<section id="functional-regression-fsusie-for-epigenomic-qtl-fine-mapping">
<h2>Functional regression fSuSiE for epigenomic QTL fine-mapping<a class="headerlink" href="#functional-regression-fsusie-for-epigenomic-qtl-fine-mapping" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[fsusie_1]
# prior can be either of [&quot;mixture_normal&quot;, &quot;mixture_normal_per_scale&quot;]
parameter: prior = &quot;mixture_normal&quot;
parameter: max_SNP_EM = 100
# Max scale is such that 2^max_scale being the number of phenotypes in the transformed space. Default to 2^10  = 1024. Don&#39;t change it unless you know what you are doing. Max_scale should be at least larger than 5.
parameter:  max_scale = 10
# Purity and coverage used to call cs
parameter:  min_purity = 0.5
# Epigenetics mark filter
parameter: epigenetics_mark_treshold = 16
# Run susie for top pc of the fsusie input
parameter: susie_top_pc = 10
depends: sos_variable(&quot;regional_data&quot;)
# Check if both &#39;data&#39; and &#39;meta_info&#39; are empty lists
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39;Either genotype or phenotype data are not available for region {&quot;, &quot;.join(region_name)}.&#39;)
meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name[:-2]}/{name}.{_meta_info[0].replace(&quot;:&quot;, &quot;_&quot;).replace(&quot;-&quot;, &quot;_&quot;)}.fsusie_{prior}{&quot;_top_pc_weights&quot; if twas_weights else &quot;&quot;}.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    options(warn=1)
    # extract subset of samples
    keep_samples = NULL
    if (${&quot;TRUE&quot; if keep_samples.is_file() else &quot;FALSE&quot;}) {
      keep_samples = unlist(strsplit(readLines(${keep_samples:ar}), &quot;\\s+&quot;))
      message(paste(length(keep_samples), &quot;samples are selected to be loaded for analysis&quot;))
    }

    # Load regional functional data
    library(pecotmr)
    tryCatch({
    fdat = load_regional_functional_data(genotype = ${_input[0]:anr},
                                          phenotype = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1:len(_input)//2+1]])}),
                                          covariate = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[len(_input)//2+1:]])}),
                                          region = &quot;${_meta_info[0]}&quot;,
                                          association_window = &quot;${_meta_info[1]}&quot;,
                                          conditions = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in _meta_info[4:]])}),
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss},
                                          keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;},
                                          keep_samples = keep_samples,
                                          tabix_header = TRUE,
                                          phenotype_header = 4,
                                          region_name_col = 4,
                                          scale_residuals = FALSE)
    }, NoSNPsError = function(e) {
        message(&quot;Error: &quot;, paste(e$message, &quot;${_meta_info[2] + &#39;@&#39; + _meta_info[1]}&quot;))
        saveRDS(list(&quot;${_meta_info[0]}&quot; = e$message), ${_output:ar}, compress=&#39;xz&#39;)
        quit(save=&quot;no&quot;)
    })
    # Filter out list fdat that with less than a treshold of epigenomic marker.
    library(tidyverse)
    filter_fdat_except_specific_names &lt;- function(fdat, n) {
        # Identify which elements in list1 meet the row count criteria
        indices_to_keep &lt;- sapply(fdat$Y_coordinates, function(x) nrow(x) &gt;= n)
        fdat_filtered &lt;- map(fdat[!names(fdat) %in% c(&quot;dropped_sample&quot;, &quot;X&quot;, &quot;chrom&quot;)],~.x[indices_to_keep]) 
        return(c(fdat_filtered,fdat[names(fdat) %in% c(&quot;dropped_sample&quot;, &quot;X&quot;, &quot;chrom&quot;)]))
    }

    fdat = filter_fdat_except_specific_names(fdat, n = ${epigenetics_mark_treshold})
    # Check if Y_coordinates is empty after filtering
    if (length(fdat$Y_coordinates) == 0) {
        e_msg = paste0(&quot;None of the study have more than or equal to &quot;,${epigenetics_mark_treshold}, &quot; epigenetics marks, region skipped&quot;)
        message(e_msg)
        saveRDS(list(&quot;${_meta_info[0]}&quot; = e_msg ),  ${_output:ar}, compress=&#39;xz&#39;)
        quit(save=&quot;no&quot;)
    }

    if (${&quot;TRUE&quot; if save_data else &quot;FALSE&quot;}) {
      # save data object for debug purpose
      saveRDS(list(&quot;${_meta_info[0]}&quot; = fdat), &quot;${_output:ann}.${epigenetics_mark_treshold}_marks.dataset.rds&quot;, compress=&#39;xz&#39;)
    }
  
    fitted = setNames(replicate(length(fdat$residual_Y), list(), simplify = FALSE), names(fdat$residual_Y))
    for (r in 1:length(fitted)) {
        st = proc.time()
        fitted[[r]] = list()
        message(paste(&quot;Dimension of Y matrix is &quot;, nrow(fdat$residual_Y[[r]]), &quot;rows by&quot;, ncol(fdat$residual_Y[[r]]), &quot;columns.&quot;))
        
        # Run SuSiE on top K PC
        if(${&quot;TRUE&quot; if (susie_top_pc &gt;0 or twas_weights) else &quot;FALSE&quot;}) {
            top_pc_data &lt;- prcomp(fdat$residual_Y[[r]], center = TRUE, scale. = TRUE)$x
            if (${susie_top_pc} &lt; ncol(top_pc_data)) {
                top_pc_data &lt;- top_pc_data[,1:${susie_top_pc},drop=F]
            }
            fitted[[r]]$susie_on_top_pc &lt;- list()
            for (i in 1:ncol(top_pc_data)) {
                susie_on_top_pc &lt;- susie_wrapper(fdat$residual_X[[r]], top_pc_data[,i], init_L=${init_L}, max_L=${max_L}, refine=TRUE, coverage = ${coverage[0]})
                fitted[[r]]$susie_on_top_pc[[i]] &lt;- susie_post_processor(susie_on_top_pc, fdat$residual_X[[r]], top_pc_data[,i], fdat$residual_X_scalar[[r]], 1, fdat$maf[[r]],
                                           secondary_coverage = c(${&quot;,&quot;.join([str(x) for x in coverage[1:]])}), signal_cutoff = ${pip_cutoff},
                                           other_quantities = list(dropped_samples = list(X=fdat$dropped_sample$dropped_samples_X[[r]], 
                                                                   y=fdat$dropped_sample$dropped_samples_Y[[r]], 
                                                                   covar=fdat$dropped_sample$dropped_samples_covar[[r]])))
            }
            # Run TWAS weights on the first PC
            # Exactly the same codes copied from susie_twas
            if ( ${&quot;TRUE&quot; if twas_weights else &quot;FALSE&quot;} ) {
                twas_weights_output &lt;- twas_weights_pipeline(fdat$residual_X[[r]], top_pc_data[,1], fdat$maf[[r]], 
                                         susie_fit=fitted[[r]]$susie_on_top_pc[[1]]$susie_result_trimmed,
                                         ld_reference_meta_file = ${(&#39;&quot;%s&quot;&#39; % ld_reference_meta_file) if not ld_reference_meta_file.is_dir() else &quot;NULL&quot;},
                                         X_scalar = fdat$residual_X_scalar[[r]], y_scalar = fdat$residual_Y_scalar[[r]],
                                         cv_folds = ${twas_cv_folds}, coverage=${coverage[0]}, secondary_coverage=c(${&quot;,&quot;.join([str(x) for x in coverage[1:]])}), signal_cutoff = ${pip_cutoff},
                                         min_cv_maf=${min_cv_maf}, max_cv_variants=${max_cv_variants}, cv_seed=${seed}, cv_threads=${twas_cv_threads})
                # clean up the output database
                fitted[[r]] = c(fitted[[r]], twas_weights_output)
                fitted[[r]]$twas_weights = lapply(fitted[[r]]$twas_weights, function(x) { rownames(x) &lt;- NULL; return(x) })
            }
        }
        # Run fSuSiE -- this can take a while
        fitted[[r]]$fsusie_result &lt;- fsusie_wrapper(X = fdat$residual_X[[r]],
                                      Y = fdat$residual_Y[[r]],
                                      pos=fdat$Y_coordinates[[r]]$start,
                                      L=${max_L},
                                      prior=&quot;${prior}&quot;,
                                      max_SNP_EM=${max_SNP_EM}, 
                                      max_scale = ${max_scale},
                                      min.purity = ${min_purity},
                                      cov_lev = ${coverage[0]})
        fitted[[r]]$Y_coordinates = fdat$Y_coordinates[[r]] # Add Y coord 
        names(fitted[[r]]$fsusie_result$pip) = colnames(fdat$residual_X[[r]]) 
        fitted[[r]]$fsusie_summary &lt;- susie_post_processor(fitted[[r]]$fsusie_result, fdat$residual_X[[r]], NULL, fdat$residual_X_scalar[[r]], 1, fdat$maf[[r]], 
                                                          secondary_coverage = c(${&quot;,&quot;.join([str(x) for x in coverage[1:]])}), signal_cutoff = ${pip_cutoff},
                                                          other_quantities = list(dropped_samples = list(X=fdat$dropped_sample$dropped_samples_X[[r]], y=fdat$dropped_sample$dropped_samples_Y[[r]], 
                                                                                  covar=fdat$dropped_sample$dropped_samples_covar[[r]])))
        fitted[[r]]$total_time_elapsed = proc.time() - st
        fitted[[r]]$region_info = list(region_coord=parse_region(&quot;${_meta_info[0]}&quot;), grange=parse_region(&quot;${_meta_info[1]}&quot;), region_name=&quot;${_meta_info[2]}&quot;)
        # original data no longer relevant, set to NA to release memory
        fdat$residual_X[[r]] &lt;- NA
        fdat$residual_Y[[r]] &lt;- NA
    }
    saveRDS(list(&quot;${_meta_info[0]}&quot; = fitted), ${_output:ar}, compress=&#39;xz&#39;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="functional-regression-fsusie-with-other-modality">
<h2>Functional regression fSuSiE with other modality<a class="headerlink" href="#functional-regression-fsusie-with-other-modality" title="Permalink to this heading">#</a></h2>
<p><strong>This is still WIP  mvfSuSiE package is still being developed</strong></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[mvfsusie_1]
# prior can be either of [&quot;mixture_normal&quot;, &quot;mixture_normal_per_scale&quot;]
parameter: prior  = &quot;mixture_normal_per_scale&quot;
parameter: max_SNP_EM = 1000
depends: sos_variable(&quot;regional_data&quot;)
# Check if both &#39;data&#39; and &#39;meta_info&#39; are empty lists
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39;Either genotype or phenotype data are not available for region {&quot;, &quot;.join(region_name)}.&#39;)

meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name[:-2]}/{name}.{_meta_info[0]}.mvfsusie_{prior}.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    # Load regional association data
    fdat = load_regional_association_data(genotype = ${_input[0]:anr},
                                          phenotype = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1::2]])}),
                                          covariate = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[2::2]])}),
                                          region = ${&#39;&quot;%s:%s-%s&quot;&#39; % (_meta_info[1], _meta_info[2], _meta_info[3])},
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss})
    # Fine-mapping with mvfSuSiE
    library(&quot;mvf.susie.alpha&quot;)
    Y = map(fdat$residual_Y, ~left_join(fdat$X[,1]%&gt;%as.data.frame%&gt;%rownames_to_column(&quot;rowname&quot;), .x%&gt;%t%&gt;%as.data.frame%&gt;%rownames_to_column(&quot;rowname&quot;) , by = &quot;rowname&quot;)%&gt;%select(-2)%&gt;%column_to_rownames(&quot;rowname&quot;)%&gt;%as.matrix )
    fitted &lt;- multfsusie(Y_f = list(Y[[1]],Y[[3]]), 
                         Y_u = Reduce(cbind, Y[[2]]),
                         pos = list(pos1 =fdat$phenotype_coordiates[[1]], pos2 = fdat$phenotype_coordiates[[3]]),
                         X=X,
                         L=${max_L},
                         data.format=&quot;list_df&quot;)
    saveRDS(fitted, ${_output:ar})
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/mnm_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="cis_advanced_overview.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Overview of advanced cis-QTL analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="SuSiE.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">SuSiE fine-mapping workflow</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genotype-data">Example genotype data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-association-window-file">Example association-window file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#about-indels">About indels</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-susie-with-twas-weights-from-multiple-methods">i. SuSiE with TWAS weights from multiple methods</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-fsusie">ii. fSuSiE</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#iii-mvsusie-with-twas-weights-from-mr-mash">iii. mvSuSiE with TWAS weights from mr.mash</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#univariate-analysis-susie-and-twas">Univariate analysis: SuSiE and TWAS</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multivariate-analysis-mvsusie-and-mr-mash">Multivariate analysis: mvSuSiE and mr.mash</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-for-epigenomic-qtl-fine-mapping">Functional regression fSuSiE for epigenomic QTL fine-mapping</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-with-other-modality">Functional regression fSuSiE with other modality</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
       Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>