
<!DOCTYPE html>


<html lang="en" data-content_root="../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Multivariate Fine-Mapping for multiple genes &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css?v=be8a1c11" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/mnm_analysis/multivariate_multigene_fine_mapping_vignette';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE" href="univariate_fine_mapping_fsusie_vignette.html" />
    <link rel="prev" title="Univariate Fine-Mapping and TWAS with SuSiE" href="univariate_fine_mapping_twas_vignette.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="mnm_miniprotocol.html">Integrative Analysis with High-Dimensional Regression</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="mnm_postprocessing.html">Integrative Analysis Output Processing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/edit/gh-pages/code/mnm_analysis/multivariate_multigene_fine_mapping_vignette.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/mnm_analysis/multivariate_multigene_fine_mapping_vignette.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/mnm_analysis/multivariate_multigene_fine_mapping_vignette.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Multivariate Fine-Mapping for multiple genes</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-run-the-fine-mapping-with-mvsusie">ii. Run the Fine-Mapping with mvSuSiE</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#anticipated-results">Anticipated Results</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="multivariate-fine-mapping-for-multiple-genes">
<h1>Multivariate Fine-Mapping for multiple genes<a class="headerlink" href="#multivariate-fine-mapping-for-multiple-genes" title="Link to this heading">#</a></h1>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Link to this heading">#</a></h2>
<p>Multi gene fine-mapping and TWAS may also be conducted with our pipeline. This considers multiple genes jointly within specific TAD windows.</p>
<p>This step is similar to the multivariate fine-mapping with two main differences. 1) TAD windows with multiple genes need to be defined. The <code class="docutils literal notranslate"><span class="pre">--pheno_id_map_file</span></code> parameter is used for this. 2) To speed things up, the genes are filtered out if they don’t have a univariate fine mapped region. Genes may also be filtered out if they do have a univariate fine-mapped signal, but the signal is nowhere close to that of other genes.  The <code class="docutils literal notranslate"><span class="pre">--skip-analysis-pip-cutoff</span></code> parameter is used for this.</p>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<p><code class="docutils literal notranslate"><span class="pre">--genoFile</span></code>: path to a plink bed file containin genotypes. Include the <code class="docutils literal notranslate"><span class="pre">.bed</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">--phenoFile</span></code>: a tab delimited file containing chr, start, end, ID and path for the regions. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#chr    start   end     ID      path
chr12   389319  389320  ENSG00000073614 $PATH/snuc_pseudo_bulk.Mic.mega.normalized.log2cpm.bed.gz
chr12   752578  752579  ENSG00000060237 $PATH/snuc_pseudo_bulk.Mic.mega.normalized.log2cpm.bed.gz
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--covFile</span></code>: path to a gzipped file containing covariates in the rows, and sample ids in the columns.</p>
<p><code class="docutils literal notranslate"><span class="pre">--customized-association-windows</span></code>: a tab delimited file containing chr, start, end, and ID regions. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start   end     TAD_id</span>
<span class="n">chr1</span>    <span class="mi">0</span>       <span class="mi">10985501</span>        <span class="n">chr1_0_10985501</span>
<span class="n">chr1</span>    <span class="mi">5101787</span> <span class="mi">11630758</span>        <span class="n">chr1_5101787_11630758</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--phenotype-names</span></code>: the names of the phenotypes if multiple are included. There should be one for each phenotype file you include.</p>
<p><code class="docutils literal notranslate"><span class="pre">--max-cv-variants</span></code>: maximum number of variants for cross-validation.</p>
<p><code class="docutils literal notranslate"><span class="pre">--ld_reference_meta_file</span></code>: path to file containing chrom, start, end and path for linkage disequilibrium region information. For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chrom  start   end     path</span>
<span class="n">chr1</span>    <span class="mi">101384274</span>       <span class="mi">104443097</span>       <span class="n">chr1</span><span class="o">/</span><span class="n">chr1_101384274_104443097</span><span class="o">.</span><span class="n">cor</span><span class="o">.</span><span class="n">xz</span><span class="p">,</span><span class="n">chr1</span><span class="o">/</span><span class="n">chr1_101384274_104443097</span><span class="o">.</span><span class="n">cor</span><span class="o">.</span><span class="n">xz</span><span class="o">.</span><span class="n">bim</span>
<span class="n">chr1</span>    <span class="mi">104443097</span>       <span class="mi">106225286</span>       <span class="n">chr1</span><span class="o">/</span><span class="n">chr1_104443097_106225286</span><span class="o">.</span><span class="n">cor</span><span class="o">.</span><span class="n">xz</span><span class="p">,</span><span class="n">chr1</span><span class="o">/</span><span class="n">chr1_104443097_106225286</span><span class="o">.</span><span class="n">cor</span><span class="o">.</span><span class="n">xz</span><span class="o">.</span><span class="n">bim</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--independent_variant_list</span></code>: a gzipped file containing variant information. These should be independent from one another in terms of linkage disequilibrium.
For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">chrom</span>   <span class="n">pos</span>     <span class="n">alt</span>     <span class="n">ref</span>     <span class="n">variant_id</span>
<span class="n">chr1</span>    <span class="mi">16206</span>   <span class="n">T</span>       <span class="n">A</span>       <span class="n">chr1</span><span class="p">:</span><span class="mi">16206</span><span class="p">:</span><span class="n">T</span><span class="p">:</span><span class="n">A</span>
<span class="n">chr1</span>    <span class="mi">16433</span>   <span class="n">C</span>       <span class="n">G</span>       <span class="n">chr1</span><span class="p">:</span><span class="mi">16433</span><span class="p">:</span><span class="n">C</span><span class="p">:</span><span class="n">G</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--fine_mapping_meta</span></code>: A file containg a list of gene and region information and other conditions.
For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>#chr    start   end     region_id       TSS     original_data   combined_data   combined_data_sumstats  conditions      conditions_top_loci
chr1    0       6480000 ENSG00000008128 1724356 KNIGHT_pQTL.ENSG00000008128.univariate_susie_twas_weights.rds,MiGA_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds,MSBB_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds,ROSMAP_Bennett_Klein_pQTL.ENSG00000008128.univariate_susie_twas_weights.rds,ROSMAP_DeJager_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds,ROSMAP_Kellis_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds,ROSMAP_mega_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds,STARNET_eQTL.ENSG00000008128.univariate_susie_twas_weights.rds  $PATH/Fungen_xQTL.ENSG00000008128.cis_results_db.export.rds        $PATH/Fungen_xQTL.ENSG00000008128.cis_results_db.export_sumstats.rds       Knight_eQTL_brain,MiGA_GFM_eQTL,MiGA_GTS_eQTL,MiGA_SVZ_eQTL,MiGA_THA_eQTL,BM_10_MSBB_eQTL,BM_22_MSBB_eQTL,BM_36_MSBB_eQTL,BM_44_MSBB_eQTL,monocyte_ROSMAP_eQTL,Mic_DeJager_eQTL,Ast_DeJager_eQTL,Oli_DeJager_eQTL,Exc_DeJager_eQTL,Inh_DeJager_eQTL,DLPFC_DeJager_eQTL,PCC_DeJager_eQTL,AC_DeJager_eQTL,Mic_Kellis_eQTL,Ast_Kellis_eQTL,Oli_Kellis_eQTL,OPC_Kellis_eQTL,Exc_Kellis_eQTL,Inh_Kellis_eQTL,Ast_mega_eQTL,Exc_mega_eQTL,Inh_mega_eQTL,Oli_mega_eQTL,STARNET_eQTL_Mac       Knight_eQTL_brain,MiGA_GFM_eQTL,MiGA_GTS_eQTL,MiGA_SVZ_eQTL,MiGA_THA_eQTL,BM_10_MSBB_eQTL,BM_22_MSBB_eQTL,BM_36_MSBB_eQTL,BM_44_MSBB_eQTL,monocyte_ROSMAP_eQTL,Mic_DeJager_eQTL,Ast_DeJager_eQTL,Oli_DeJager_eQTL,Exc_DeJager_eQTL,Inh_DeJager_eQTL,DLPFC_DeJager_eQTL,PCC_DeJager_eQTL,AC_DeJager_eQTL,Mic_Kellis_eQTL,Ast_Kellis_eQTL,Oli_Kellis_eQTL,OPC_Kellis_eQTL,Exc_Kellis_eQTL,Inh_Kellis_eQTL,Ast_mega_eQTL,Exc_mega_eQTL,Inh_mega_eQTL,Oli_mega_eQTL,STARNET_eQTL_Mac
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--phenoIDFile</span></code>: A bed file containing a list of genes and their LD region.
For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">TAD_id</span>  <span class="n">ID</span>
<span class="n">chr19_0_13957223</span>        <span class="n">ENSG00000172270</span>
<span class="n">chr19_0_13957223</span>        <span class="n">ENSG00000099864</span>
<span class="n">chr19_0_13957223</span>        <span class="n">ENSG00000011304</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--skip-analysis-pip-cutoff</span></code>: A number of the pip cutoff.</p>
<p><code class="docutils literal notranslate"><span class="pre">--coverage</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">--maf</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">--pheno_id_map_file</span></code>: A file containing IDs and genes.
For example:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ID</span>      <span class="n">gene</span>
<span class="n">chr20</span><span class="p">:</span><span class="mi">50940933</span><span class="p">:</span><span class="mi">50941105</span><span class="p">:</span><span class="n">clu_44490_</span><span class="o">-</span><span class="p">:</span><span class="n">ENSG00000000419</span>     <span class="n">ENSG00000000419</span>
<span class="n">chr20</span><span class="p">:</span><span class="mi">50940933</span><span class="p">:</span><span class="mi">50941129</span><span class="p">:</span><span class="n">clu_44490_</span><span class="o">-</span><span class="p">:</span><span class="n">ENSG00000000419</span>     <span class="n">ENSG00000000419</span>
<span class="n">chr20</span><span class="p">:</span><span class="mi">50936262</span><span class="p">:</span><span class="mi">50942031</span><span class="p">:</span><span class="n">clu_44490_</span><span class="o">-</span><span class="p">:</span><span class="n">ENSG00000000419</span>     <span class="n">ENSG00000000419</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">--prior-canonical-matrices</span></code></p>
<p><code class="docutils literal notranslate"><span class="pre">--save-data</span></code>: whether to save intermediate data or not</p>
<p><code class="docutils literal notranslate"><span class="pre">--twas-cv-folds</span></code>: Perform K folds valiation CV for TWAS. Set this to zero to skip</p>
<p><code class="docutils literal notranslate"><span class="pre">--trans-analysis</span></code>: Include this if doing trans-analysis (not using phenotypic coordinate information)</p>
<p><code class="docutils literal notranslate"><span class="pre">--region-name</span></code>: if you only wish to analyze one region, then include the ID of a region found in the <code class="docutils literal notranslate"><span class="pre">customized-association-windows</span></code> file</p>
<p><code class="docutils literal notranslate"><span class="pre">--cwd</span></code>: output file path</p>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Link to this heading">#</a></h2>
<section id="ii-run-the-fine-mapping-with-mvsusie">
<h3>ii. Run the Fine-Mapping with mvSuSiE<a class="headerlink" href="#ii-run-the-fine-mapping-with-mvsusie" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="err">$</span><span class="n">PATH</span><span class="o">/</span><span class="n">protocol</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">mnm_regression</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">mnm_genes</span> \
    <span class="o">--</span><span class="n">name</span> <span class="n">ROSMAP_Ast_DeJager_eQTL</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="err">$</span><span class="n">PATH</span><span class="o">/</span><span class="n">ROSMAP_NIA_WGS</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">bcftools_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="mf">.11</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="err">$</span><span class="n">PATH</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">.</span><span class="n">Ast</span><span class="o">.</span><span class="n">mega</span><span class="o">.</span><span class="n">normalized</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">covFile</span> <span class="err">$</span><span class="n">PATH</span><span class="o">/</span><span class="n">snuc_pseudo_bulk</span><span class="o">.</span><span class="n">Ast</span><span class="o">.</span><span class="n">mega</span><span class="o">.</span><span class="n">normalized</span><span class="o">.</span><span class="n">log2cpm</span><span class="o">.</span><span class="n">rosmap_cov</span><span class="o">.</span><span class="n">ROSMAP_NIA_WGS</span><span class="o">.</span><span class="n">leftnorm</span><span class="o">.</span><span class="n">bcftools_qc</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">snuc_pseudo_bulk_mega</span><span class="o">.</span><span class="n">related</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">extracted</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">projected</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">customized</span><span class="o">-</span><span class="n">association</span><span class="o">-</span><span class="n">windows</span> <span class="err">$</span><span class="n">PATH</span><span class="o">/</span><span class="n">windows</span><span class="o">/</span><span class="n">TADB_sliding_window</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">names</span> <span class="n">Ast_DeJager_eQTL</span> \
    <span class="o">--</span><span class="nb">max</span><span class="o">-</span><span class="n">cv</span><span class="o">-</span><span class="n">variants</span> <span class="mi">5000</span> <span class="o">--</span><span class="n">ld_reference_meta_file</span> <span class="err">$</span><span class="n">PATH</span><span class="o">/</span><span class="n">ldref</span><span class="o">/</span><span class="n">ld_meta_file</span><span class="o">.</span><span class="n">tsv</span> \
    <span class="o">--</span><span class="n">independent_variant_list</span> <span class="err">$</span><span class="n">PATH</span><span class="o">/</span><span class="n">ld_pruned_variants</span><span class="o">.</span><span class="n">txt</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">fine_mapping_meta</span> <span class="err">$</span><span class="n">PATH</span><span class="o">/</span><span class="n">Fungen_xQTL</span><span class="o">.</span><span class="n">cis_results_db</span><span class="o">.</span><span class="n">new</span><span class="o">.</span><span class="n">tsv</span> \
    <span class="o">--</span><span class="n">phenoIDFile</span> <span class="err">$</span><span class="n">PATH</span><span class="o">/</span><span class="n">phenoIDFile_cis_extended_region</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">skip</span><span class="o">-</span><span class="n">analysis</span><span class="o">-</span><span class="n">pip</span><span class="o">-</span><span class="n">cutoff</span> <span class="mi">0</span> \
    <span class="o">--</span><span class="n">coverage</span> <span class="mf">0.95</span> \
    <span class="o">--</span><span class="n">maf</span> <span class="mf">0.01</span> \
    <span class="o">--</span><span class="n">pheno_id_map_file</span> <span class="err">$</span><span class="n">PATH</span><span class="o">/</span><span class="n">pheno_id_map_file</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">prior</span><span class="o">-</span><span class="n">canonical</span><span class="o">-</span><span class="n">matrices</span> \
    <span class="o">--</span><span class="n">save</span><span class="o">-</span><span class="n">data</span> \
    <span class="o">--</span><span class="n">twas</span><span class="o">-</span><span class="n">cv</span><span class="o">-</span><span class="n">folds</span> <span class="mi">0</span> \
    <span class="o">--</span><span class="n">trans</span><span class="o">-</span><span class="n">analysis</span> \
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="n">name</span> <span class="n">chr11_77324757_86627922</span> \ 
    <span class="o">--</span><span class="n">cwd</span> <span class="err">$</span><span class="n">PATH</span><span class="o">/</span><span class="n">output</span><span class="o">/</span> <span class="o">-</span><span class="n">s</span> <span class="n">force</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="anticipated-results">
<h2>Anticipated Results<a class="headerlink" href="#anticipated-results" title="Link to this heading">#</a></h2>
<p>For each gene and region, multivariate multigene finemapping will produce a file containing results for the top hits and a file containing twas weights produced by susie.</p>
<p><code class="docutils literal notranslate"><span class="pre">ROSMAP_Ast_DeJager_eQTL.chr11_77324757_86627922.multigene_bvrs.rds</span></code>:</p>
<ul class="simple">
<li><p>for each region name, includes:</p></li>
</ul>
<ol class="arabic simple">
<li><p>mrmash_fitted</p></li>
<li><p>reweighted_mixture_prior</p></li>
<li><p>reweighted_mixture_prior_cv</p></li>
<li><p>mvsusie_fitted</p></li>
<li><p>variant_names</p></li>
<li><p>analysis_script</p></li>
<li><p>other_quantitites</p></li>
<li><p>analysis_script</p></li>
<li><p>top_loci</p></li>
<li><p>susie_result_trimmed</p></li>
<li><p>total_time_elapsed</p></li>
<li><p>region_info</p></li>
</ol>
<p><code class="docutils literal notranslate"><span class="pre">ROSMAP_Ast_DeJager_eQTL.chr11_77324757_86627922.multigene_data.rds</span></code>:(from the –save-data argument)</p>
<ul class="simple">
<li><p>see <a class="reference external" href="https://github.com/cumc/pecotmr/blob/68d87ca1d0a059022bf4e55339621cbddc8993cc/R/file_utils.R#L461">pecotmr code</a> for description</p></li>
</ul>
<p><code class="docutils literal notranslate"><span class="pre">ROSMAP_Ast_DeJager_eQTL.chr11_77324757_86627922.multigene_twas_weights.rds</span></code>:</p>
<ul class="simple">
<li><p>for each region name and for each gene within that region, includes:</p></li>
</ul>
<ol class="arabic simple">
<li><p>twas_weights - from mrmash and mvsusie</p></li>
<li><p>twas_predictions - from mrmash and mvsusie</p></li>
<li><p>variant_names</p></li>
<li><p>total_time_elapsed</p></li>
<li><p>region_info</p></li>
</ol>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            name: "python3",
            path: "./code/mnm_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="univariate_fine_mapping_twas_vignette.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Univariate Fine-Mapping and TWAS with SuSiE</p>
      </div>
    </a>
    <a class="right-next"
       href="univariate_fine_mapping_fsusie_vignette.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-run-the-fine-mapping-with-mvsusie">ii. Run the Fine-Mapping with mvSuSiE</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#anticipated-results">Anticipated Results</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>