

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>SuSIE results post process &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/post_processing/fine_mapping_post_processing';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="prev" title="Association result processing" href="association_scan_post_processing.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced cis-QTL Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../mnm_analysis/cis_advanced_overview.html">Overview of advanced cis-QTL analysis</a><input class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul class="simple">
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/SuSiE.html">SuSiE fine-mapping workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/fSuSiE.html">fSuSiE workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mvSuSiE.html">Multivariate fine-mapping workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/mrmash.html">Multivariate TWAS prediction model</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced Genome-wide Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../mnm_analysis/rss_analysis.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l1"><a class="reference internal" href="../prototype_drafts/polyfun.html">Fine-mapping with PolyFun</a></li>
<li class="toctree-l1"><a class="reference internal" href="../prototype_drafts/MRAID_QTL.html">Mendelian Randomization using MRAID</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/ldsc.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL-GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../pecotmr_integration/pecotmr.html">The PECOTMR intergration framework</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">GWAS integration: enrichment and colocalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/twas_mr.html">GWAS integration: TWAS and MR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/intact.html">TWAS and COLOC integration</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">Extract genome-wide data for multivariate analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">MASH analysis pipeline with posterior computation</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Rare xQTL</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../rare_xqtl/watershed.html">watershed analysis for rare xQTL</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Post Processing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="sumstats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="rds_to_vcf.html">Summary statistics in VCF format</a></li>
<li class="toctree-l1"><a class="reference internal" href="association_scan_post_processing.html">Association result processing</a></li>
<li class="toctree-l1 current active"><a class="current reference internal" href="#">SuSIE results post process</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/post_processing/fine_mapping_post_processing.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/post_processing/fine_mapping_post_processing.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/post_processing/fine_mapping_post_processing.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>SuSIE results post process</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extracting-susie-results">Extracting susie results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extracting-susie-rss-results-for-adgwas">Extracting susie_rss results for ADGWAS</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extracting-fsusie-results">Extracting fsusie results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exporting-cis-analysis-susie-twas-results">Exporting cis_analysis susie_twas results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#export-gwas-data">Export gwas data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overlapped-gwas-data-and-eqtl-data">Overlapped gwas data and eQTL data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#plotting-the-pip-plot">Plotting the pip plot</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#into-vcf-format">Into VCF format</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cis-window-analysis-result-consolidation">Cis-window analysis Result consolidation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gwas-results-consolidation">GWAS results consolidation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overlap-qtl-and-gwas-results">Overlap QTL and GWAS results</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="susie-results-post-process">
<h1>SuSIE results post process<a class="headerlink" href="#susie-results-post-process" title="Permalink to this heading">#</a></h1>
<p>This notebook is to post-process the susie results into different text file</p>
<section id="extracting-susie-results">
<h2>Extracting susie results<a class="headerlink" href="#extracting-susie-results" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/SuSiE_post_processing.ipynb susie_to_tsv \
    --cwd output/test --rds_path `ls output/test/cache/*rds | head ` --region-list &lt;(head -50  ./dlpfc_region_list) --container containers/stephenslab.sif 
</pre></div>
</div>
</div>
</div>
</section>
<section id="extracting-susie-rss-results-for-adgwas">
<h2>Extracting susie_rss results for ADGWAS<a class="headerlink" href="#extracting-susie-rss-results-for-adgwas" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/SuSiE_post_processing.ipynb susie_to_tsv \
    --cwd output/ADGWAS_finemapping_extracted/Bellenguez/ --rds_path `ls GWAS_Finemapping_Results/Bellenguez/ADGWAS2022*rds ` \
    --region-list ~/1300_hg38_EUR_LD_blocks_orig.tsv \
    --container containers/stephenslab.sif 

sos run pipeline/SuSiE_post_processing.ipynb susie_tsv_collapse \
    --cwd output/ADGWAS_finemapping_extracted --tsv_path `ls output/ADGWAS_finemapping_extracted/*lbf.tsv` \
    --container containers/stephenslab.sif 
</pre></div>
</div>
</div>
</div>
</section>
<section id="extracting-fsusie-results">
<h2>Extracting fsusie results<a class="headerlink" href="#extracting-fsusie-results" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/SuSiE_post_processing.ipynb fsusie_to_tsv \
    --cwd output/f_susie_tad_haQTL_pos --rds_path `ls output/f_susie_tad_haQTL_pos/cache/*rds ` \
    --region-list ../eqtl/dlpfc_tad_list \
    --container containers/stephenslab.sif -s build
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/SuSiE_post_processing.ipynb fsusie_to_tsv \
    --cwd output/f_susie_tad_meQTL_pos_selected/ --rds_path `ls output/f_susie_tad_meQTL_pos_selected//cache/*1204*rds ` \
    --region-list ../eqtl/dlpfc_tad_list \
    --container containers/stephenslab.sif -s build
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/SuSiE_post_processing.ipynb fsusie_to_tsv \
    --cwd output/f_susie_tad_meQTL_pos_2/ --rds_path `ls output/f_susie_tad_meQTL_pos_2//cache/*rds ` \
    --region-list ../eqtl/dlpfc_tad_list \
    --container containers/stephenslab.sif -s build
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/SuSiE_post_processing.ipynb fsusie_to_tsv \
    --cwd output/f_susie_tad_meQTL_pos/ --rds_path `ls output/f_susie_tad_meQTL_pos//cache/*rds ` \
    --region-list ../eqtl/dlpfc_tad_list \
    --container containers/stephenslab.sif -s build
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/SuSiE_post_processing.ipynb fsusie_to_tsv \
    --cwd output/f_susie_tad_haQTL_pos_check_pure_2 --rds_path `ls output/f_susie_tad_haQTL_pos_check_pure_2/cache/*rds ` \
    --region-list ../eqtl/dlpfc_tad_list \
    --container containers/stephenslab.sif -s build
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run pipeline/SuSiE_post_processing.ipynb fsusie_to_tsv \
    --cwd output/f_susie_tad_meQTL_pos_2/ --rds_path `ls output/f_susie_tad_meQTL_pos_2//cache/*rds ` \
    --region-list ../eqtl/dlpfc_tad_list \
    --container containers/stephenslab.sif -s build
</pre></div>
</div>
</div>
</div>
</section>
<section id="exporting-cis-analysis-susie-twas-results">
<h2>Exporting cis_analysis susie_twas results<a class="headerlink" href="#exporting-cis-analysis-susie-twas-results" title="Permalink to this heading">#</a></h2>
<p>meta file is produced by:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">get_condition</span> <span class="o">&lt;-</span> <span class="n">function</span><span class="p">(</span><span class="n">conditions</span><span class="p">,</span> <span class="n">Author</span><span class="p">,</span> <span class="n">qtl_type</span><span class="p">){</span>
    <span class="n">strings</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">()</span>
    <span class="k">for</span><span class="p">(</span><span class="n">condition</span> <span class="ow">in</span> <span class="n">conditions</span><span class="p">){</span>
        <span class="n">string</span> <span class="o">=</span>  <span class="n">paste</span><span class="p">(</span><span class="n">condition</span><span class="p">,</span> <span class="n">Author</span><span class="p">,</span> <span class="n">qtl_type</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>  
        <span class="n">string</span> <span class="o">=</span> <span class="n">paste</span><span class="p">(</span><span class="n">unique</span><span class="p">(</span><span class="n">unlist</span><span class="p">(</span><span class="n">strsplit</span><span class="p">(</span><span class="n">string</span><span class="p">,</span> <span class="s2">&quot;_&quot;</span><span class="p">))),</span> <span class="n">collapse</span> <span class="o">=</span> <span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="n">strings</span> <span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="n">strings</span><span class="p">,</span> <span class="n">string</span><span class="p">)</span>
    <span class="p">}</span>
    <span class="k">return</span><span class="p">(</span><span class="n">strings</span><span class="p">)</span>
<span class="p">}</span>

<span class="n">raw_name</span><span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;Mic&quot;</span><span class="p">,</span><span class="s2">&quot;Ast&quot;</span><span class="p">,</span><span class="s2">&quot;Oli&quot;</span><span class="p">,</span><span class="s2">&quot;OPC&quot;</span><span class="p">,</span><span class="s2">&quot;Exc&quot;</span><span class="p">,</span><span class="s2">&quot;Inh&quot;</span><span class="p">,</span><span class="s2">&quot;DLPFC&quot;</span><span class="p">,</span><span class="s2">&quot;PCC&quot;</span><span class="p">,</span><span class="s2">&quot;AC&quot;</span><span class="p">)</span>
<span class="n">raw_name_kellis</span><span class="o">&lt;-</span> <span class="n">c</span><span class="p">(</span><span class="s2">&quot;Mic_Kellis&quot;</span><span class="p">,</span><span class="s2">&quot;Ast_Kellis&quot;</span><span class="p">,</span><span class="s2">&quot;Oli_Kellis&quot;</span><span class="p">,</span><span class="s2">&quot;OPC_Kellis&quot;</span><span class="p">,</span><span class="s2">&quot;Exc_Kellis&quot;</span><span class="p">,</span><span class="s2">&quot;Inh_Kellis&quot;</span><span class="p">,</span><span class="s2">&quot;Ast.10&quot;</span><span class="p">,</span><span class="s2">&quot;Mic.12&quot;</span><span class="p">,</span><span class="s2">&quot;Mic.13&quot;</span><span class="p">)</span>

<span class="n">dejager_name</span> <span class="o">&lt;-</span> <span class="n">get_condition</span><span class="p">(</span><span class="n">raw_name</span><span class="p">,</span> <span class="s2">&quot;De_Jager&quot;</span><span class="p">,</span><span class="s2">&quot;eQTL&quot;</span><span class="p">)</span>
<span class="n">kellis_name</span> <span class="o">&lt;-</span> <span class="n">get_condition</span><span class="p">(</span><span class="n">raw_name_kellis</span><span class="p">,</span> <span class="s2">&quot;Kellis&quot;</span><span class="p">,</span><span class="s2">&quot;eQTL&quot;</span><span class="p">)</span>
<span class="n">eQTL_meta</span> <span class="o">&lt;-</span> <span class="n">data</span><span class="o">.</span><span class="n">frame</span><span class="p">(</span><span class="n">raw_name</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">raw_name</span><span class="p">,</span> <span class="n">raw_name_kellis</span><span class="p">),</span> <span class="n">new_name</span> <span class="o">=</span> <span class="n">c</span><span class="p">(</span><span class="n">dejager_name</span><span class="p">,</span> <span class="n">kellis_name</span><span class="p">))</span>
<span class="n">write_delim</span><span class="p">(</span><span class="n">eQTL_meta</span><span class="p">,</span> <span class="s2">&quot;/mnt/vast/hpc/csg/rf2872/Work/Multivariate/susie_2024_new/eQTL_meta.tsv&quot;</span><span class="p">)</span>
</pre></div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">head</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">Multivariate</span><span class="o">/</span><span class="n">susie_2024_new</span><span class="o">/</span><span class="n">eQTL_meta</span><span class="o">.</span><span class="n">tsv</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>raw_name new_name
Mic Mic_De_Jager_eQTL
Ast Ast_De_Jager_eQTL
Oli Oli_De_Jager_eQTL
OPC OPC_De_Jager_eQTL
Exc Exc_De_Jager_eQTL
Inh Inh_De_Jager_eQTL
DLPFC DLPFC_De_Jager_eQTL
PCC PCC_De_Jager_eQTL
AC AC_De_Jager_eQTL
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">fine_mapping_post_processing</span><span class="o">.</span><span class="n">ipynb</span>  <span class="n">cis_results_export</span> \
    <span class="o">--</span><span class="n">region_file</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">resource</span><span class="o">/</span><span class="n">TADB_enhanced_cis</span><span class="o">.</span><span class="n">protein_coding</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">file_path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">rf2872</span><span class="o">/</span><span class="n">Work</span><span class="o">/</span><span class="n">Multivariate</span><span class="o">/</span><span class="n">susie_2024_new</span><span class="o">/</span><span class="n">rds_files</span><span class="o">/</span>     \
    <span class="o">--</span><span class="kp">name</span> <span class="n">Fungen_xQTL</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span> \
    <span class="o">--</span><span class="n">suffix</span> <span class="n">univariate_susie_twas_weights</span><span class="o">.</span><span class="n">rds</span>  \
    <span class="o">--</span><span class="n">prefix</span> <span class="n">MSBB_eQTL</span> <span class="n">MiGA_eQTL</span> <span class="n">ROSMAP_DeJager_eQTL</span> <span class="n">ROSMAP_Kellis_eQTL</span> <span class="n">ROSMAP_pQTL</span> <span class="n">ROSMAP_Bennett_Klein_pQTL</span> \
    <span class="o">--</span><span class="n">data_type</span> <span class="n">qtl</span> \
    <span class="o">--</span><span class="n">min_corr</span> <span class="mf">0.8</span> 
</pre></div>
</div>
</div>
</div>
</section>
<section id="export-gwas-data">
<h2>Export gwas data<a class="headerlink" href="#export-gwas-data" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">fine_mapping_post_processing</span><span class="o">.</span><span class="n">ipynb</span>  <span class="n">gwas_results_export</span> \
    <span class="o">--</span><span class="n">region_file</span> <span class="o">~/</span><span class="n">codes</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">resource</span><span class="o">/</span><span class="n">hg38_1362_blocks</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">file_path</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">hs3393</span><span class="o">/</span><span class="n">RSS_QC</span><span class="o">/</span><span class="n">GWAS_finemapping_Feb22_7dataset</span><span class="o">/</span><span class="n">SuSiE_RSS</span><span class="o">/</span> \
    <span class="o">--</span><span class="kp">name</span> <span class="n">ADGWAS_Feb22</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">ADGWAS_Mar7</span> \
    <span class="o">--</span><span class="n">data_type</span> <span class="n">gwas</span> \
    <span class="o">--</span><span class="n">min_corr</span> <span class="mf">0.8</span> \
    <span class="o">--</span><span class="n">suffix</span> <span class="n">susie_rss</span><span class="o">.</span><span class="n">rds</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="overlapped-gwas-data-and-eqtl-data">
<h2>Overlapped gwas data and eQTL data<a class="headerlink" href="#overlapped-gwas-data-and-eqtl-data" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>sos run ~/codes/xqtl-pipeline/pipeline/fine_mapping_post_processing.ipynb  overlap_qtl_gwas \
    --qtl-files `ls /mnt/vast/hpc/csg/rf2872/Work/Multivariate/susie_2024/data/*/*export.rds  ` \
    --gwas-union-file  /mnt/vast/hpc/csg/rf2872/Work/Multivariate/gwas/output/ADGWAS.union_export.tsv.gz \
    --name ROSMAP_eQTL \
    -c ~/env_files/csg.yml -J 50
</pre></div>
</div>
</div>
</div>
</section>
<section id="plotting-the-pip-plot">
<h2>Plotting the pip plot<a class="headerlink" href="#plotting-the-pip-plot" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE_post_processing</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_pip_landscape_plot</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">test</span><span class="o">/</span> <span class="o">--</span><span class="n">plot_list</span> <span class="n">plot_recipe</span> <span class="o">--</span><span class="n">annot_tibble</span> <span class="o">~/</span><span class="n">Annotatr_builtin_annotation_tibble</span><span class="o">.</span><span class="n">tsv</span> <span class="o">-</span><span class="n">s</span> <span class="n">force</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">SuSiE_post_processing</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_upsetR_plot</span> \
    <span class="o">--</span><span class="n">cwd</span> <span class="n">output</span><span class="o">/</span><span class="n">test</span><span class="o">/</span> <span class="o">--</span><span class="n">plot_list</span> <span class="n">plot_recipe_1</span>  <span class="o">-</span><span class="n">s</span> <span class="n">force</span> <span class="o">&amp;</span>
</pre></div>
</div>
</div>
</div>
<p>The required input for this step is a tab-delimited plot_recipe file that specifies the path to each of the variant.tsv files generated from this module. Each column represents a molecular phenotype, and each row indicates the files that share common variants. Since one TAD may correspond to multiple genes, additional eQTL are permitted. If there are additional molecular phenotypes or ADGWAS datasets, additional columns can be appended.</p>
<p>The built-in Annotatr_builtin_annotation_tibble.tsv can be downloaded from <a class="reference external" href="https://www.synapse.org/#!Synapse:syn51198526">synapse</a>, please download it and specify the path.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">cat</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">vast</span><span class="o">/</span><span class="n">hpc</span><span class="o">/</span><span class="n">csg</span><span class="o">/</span><span class="n">molecular_phenotype_calling</span><span class="o">/</span><span class="n">QTL_fine_mapping</span><span class="o">/</span><span class="n">plot_recipe</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>haQTL	mQTL	eQTL	eQTL	ADGWAS
/mnt/vast/hpc/csg/molecular_phenotype_calling/QTL_fine_mapping/output/f_susie_tad_meQTL_pos//meQTL.yuqi_mqtl.tad100.uni_Fsusie.mixture_normal_per_scale.variant.tsv	/mnt/vast/hpc/csg/molecular_phenotype_calling/QTL_fine_mapping/output/f_susie_tad_haQTL_pos//haQTL.rosmap_haqtl.tad100.uni_Fsusie.mixture_normal_per_scale.variant.tsv	/mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl//output/susie_per_gene_tad//demo.ENSG00000117322.unisusie.fit.variant.tsv	/mnt/vast/hpc/csg/molecular_phenotype_calling/eqtl//output/susie_per_gene_tad//demo.ENSG00000203710.unisusie.fit.variant.tsv	/mnt/vast/hpc/csg/xqtl_workflow_testing/susie_rss/output/ADGWAS_finemapping_extracted/Bellenguez/ADGWAS2022.chr1.sumstat.chr1_205972031_208461272.unisusie_rss.fit.variant.tsv
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">import</span> <span class="nn">glob</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="c1"># A region list file documenting the chr_pos_ref_alt of a susie_object</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="s2">&quot;demo&quot;</span>

<span class="c1">## Path to work directory where output locates</span>
<span class="c1">## Containers that contains the necessary packages</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">50</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;96h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;6G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">2</span>
<span class="kn">parameter:</span> <span class="n">windows</span> <span class="o">=</span> <span class="mi">1000000</span>
<span class="c1"># use this function to edit memory string for PLINK input</span>
<span class="kn">from</span> <span class="nn">sos.utils</span> <span class="kn">import</span> <span class="n">expand_size</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[susie_to_tsv_1]
# Input
# For complete susie, region_list or tad_list, for susie_rss , LD region list 
parameter: region_list = path
region_tbl = pd.read_csv(region_list,sep = &quot;\t&quot;)
parameter: rds_path = paths
input: rds_path, group_by = 1
output: f&quot;{cwd}/{_input:bn}.variant.tsv&quot;,f&quot;{cwd}/{_input:bn}.lbf.tsv&quot;,f&quot;{cwd}/{_input:bn}.effect.tsv&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.stdout&quot;, stderr = f&quot;{_output[0]:nn}.stderr&quot;, container = container, entrypoint = entrypoint
    library(&quot;dplyr&quot;)
    library(&quot;tibble&quot;)
    library(&quot;purrr&quot;)
    library(&quot;tidyr&quot;)
    library(&quot;readr&quot;)
    library(&quot;stringr&quot;)
    library(&quot;susieR&quot;)
    extract_lbf = function(susie_obj){
    
    if(&quot;variants&quot; %in% names(susie_obj) ){
    ss_bf = tibble(snps = susie_obj$variants, cs_index = ifelse(is.null(susie_obj$sets$cs_index), 0, paste0(susie_obj$sets$cs_index,collapse =&quot;,&quot;)),names = &quot;${f&#39;{_input:br}&#39;.split(&#39;.&#39;)[-4]}&quot;)
    }
      else 
      {
    ss_bf = tibble(snps = susie_obj$variable_name, cs_index = ifelse(is.null(susie_obj$sets$cs_index), 0, paste0(susie_obj$sets$cs_index,collapse =&quot;,&quot;)),names = &quot;${_input:bnnn}&quot;)
     }
    
    ss_bf = ss_bf%&gt;%cbind(susie_obj$lbf_variable%&gt;%t)%&gt;%as_tibble()
    
    return(ss_bf)
    }
    
    extract_variants_pip = function(susie_obj,region_list){
    susie_tb = tibble( variants =  names(susie_obj$pip)[which( susie_obj$pip &gt;= 0)],
                           snps_index = which(( susie_obj$pip &gt;= 0))) %&gt;%
        mutate(chromosome = map_chr(variants, ~read.table(text = .x, sep = &quot;:&quot;)$V1%&gt;%str_replace(&quot;chr&quot;,&quot;&quot;) ),
                position  = map_chr(variants, ~read.table(text = .x, sep = &quot;:&quot;)$V2  ),
                ref = map_chr(position , ~read.table(text = .x, sep = &quot;_&quot;,colClasses = &quot;character&quot;)$V2  ),
                alt = map_chr(position , ~read.table(text = .x, sep = &quot;_&quot;,colClasses = &quot;character&quot;)$V3  ),
                position  = map_dbl(position , ~read.table(text = .x, sep = &quot;_&quot;,as.is = T)$V1  )
                             )
      susie_tb = susie_tb%&gt;%mutate(cs_order =(map(susie_tb$snps_index , ~tryCatch(which(pmap(list( a= susie_obj$sets$cs) , function(a) .x %in% a )%&gt;%unlist()), error = function(e) return(0) )  ))%&gt;%as.character%&gt;%str_replace(&quot;integer\\(0\\)&quot;,&quot;0&quot;),
                         cs_id = map_chr(cs_order,~ifelse(.x ==&quot;0&quot;, &quot;None&quot; ,names(susie_obj$sets$cs)[.x%&gt;%str_split(&quot;:&quot;)%&gt;%unlist%&gt;%as.numeric] ) ),
                         log10_base_factor = map_chr(snps_index,~paste0( susie_obj$lbf_variable[,.x],  collapse = &quot;;&quot;)),
                         pip = susie_obj$pip,
                         posterior_mean = coef.susie(susie_obj)[-1],
                         posterior_sd = susie_get_posterior_sd(susie_obj),
                         z = posterior_mean/posterior_sd)
    
          susie_tb =  susie_tb%&gt;%mutate(  molecular_trait_id = region_list$molecular_trait_id,
                             finemapped_region_start = region_list$finemapped_region_start,
                             finemapped_region_end = region_list$finemapped_region_end)
          return(susie_tb)    }
          
        

     extract_effect_pip = function(susie_obj,region_list,susie_tb){
      result_tb =  tibble(phenotype = susie_obj$name,
        V = susie_obj$V,effect_id = paste0(&quot;L&quot;,1:length(V) ) ,
        cs_log10bf = susie_obj$lbf)
        if(is.null(susie_obj$sets$cs)){
            cs_min_r2 = cs_avg_r2 =  coverage =  0 
            cs = &quot;None&quot;} else {         cs = map_chr(susie_obj$sets$cs[result_tb$effect_id],~susie_tb$variants[.x]%&gt;%paste0(collapse = &quot;;&quot;))
        coverage = map(result_tb$effect_id, ~susie_obj$sets$coverage[which(names(susie_obj$sets$cs) == .x )])%&gt;%as.numeric%&gt;%replace_na(0)
        cs_min_r2  = (susie_obj$sets$purity[result_tb$effect_id,1])%&gt;%as.numeric%&gt;%replace_na(0)  
        cs_avg_r2  = (susie_obj$sets$purity[result_tb$effect_id,2])%&gt;%as.numeric%&gt;%replace_na(0) }
        result_tb = result_tb%&gt;%mutate(cs_min_r2 = cs_min_r2,cs_avg_r2 = cs_avg_r2 ,coverage = coverage%&gt;%unlist,cs = cs )            
      return(result_tb)
      }
       
  
    susie_obj = readRDS(&quot;${_input:a}&quot;)
    if(&quot;variants&quot; %in% names(susie_obj) ){susie_obj$variants = susie_obj$variants%&gt;%str_replace(&quot;_&quot;,&quot;:&quot;)}
    if(is.null(names(susie_obj$pip ))){names(susie_obj$pip) = susie_obj$variants}
    lbf = extract_lbf(susie_obj)
    region_list = read_delim(&quot;${region_list}&quot;,&quot;\t&quot;)
    if(ncol(region_list) == 3 ){   region_list =  region_list%&gt;%mutate(`#chr` = `#chr`%&gt;%str_remove_all(&quot; &quot;) , ID = paste0(`#chr`,&quot;_&quot;,start,&quot;_&quot;,end) ) } # LD_list 
    if(region_list$start[1] - region_list$end[1]  == -1 ){ 
        region_list = region_list%&gt;%mutate( start = start - ${windows} ,end = start +${windows}) # region_list for fix cis windows  
          } 
      if(&quot;gene_id&quot; %in% colnames(region_list)){region_list = region_list%&gt;%mutate(ID = gene_id)  } # region_list for gene
    region_list = region_list%&gt;%select(molecular_trait_id = ID, chromosome  = `#chr`,finemapped_region_start = start ,finemapped_region_end = end)  # Formatting
    region_list = region_list%&gt;%filter(molecular_trait_id == &quot;${f&#39;{_input:br}&#39;.split(&#39;.&#39;)[-4]}&quot;)
    variants_pip = extract_variants_pip( susie_obj , region_list)
    effect_pip = extract_effect_pip( susie_obj , region_list,variants_pip)
    lbf%&gt;%write_delim(&quot;${_output[1]}&quot;,&quot;\t&quot;)
    variants_pip%&gt;%write_delim(&quot;${_output[0]}&quot;,&quot;\t&quot;)
    effect_pip%&gt;%write_delim(&quot;${_output[2]}&quot;,&quot;\t&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[sQTL_susie_to_tsv_1]
# Input
# For complete susie, region_list or tad_list, for susie_rss , LD region list 
parameter: region_list = path
region_tbl = pd.read_csv(region_list,sep = &quot;\t&quot;)
parameter: rds_path = paths
input: rds_path, group_by = 1
input_name=f&quot;{_input:bn}&quot;
input_name=input_name.replace(&#39;*&#39;, &#39;N&#39;) # &quot;*&quot; in leafcutter2 would be ignored in shell and cause error 
output: f&quot;{cwd}/{input_name}.variant.tsv&quot;,f&quot;{cwd}/{input_name}.lbf.tsv&quot;,f&quot;{cwd}/{input_name}.effect.tsv&quot;
tags = f&#39;{step_name}_{_output[0]:bn}&#39;
tags = tags.replace(&#39;:&#39;, &#39;_&#39;).replace(&#39;+&#39;, &#39;ps&#39;) # also for other symbols in tag id 
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = tags
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.stdout&quot;, stderr = f&quot;{_output[0]:nn}.stderr&quot;, container = container
    library(&quot;dplyr&quot;)
    library(&quot;tibble&quot;)
    library(&quot;purrr&quot;)
    library(&quot;tidyr&quot;)
    library(&quot;readr&quot;)
    library(&quot;stringr&quot;)
    library(&quot;susieR&quot;)
    extract_lbf = function(susie_obj){
    
    if(&quot;variants&quot; %in% names(susie_obj) ){
    ss_bf = tibble(snps = susie_obj$variants, cs_index = ifelse(is.null(susie_obj$sets$cs_index), 0, paste0(susie_obj$sets$cs_index,collapse =&quot;,&quot;)),names = &quot;${f&#39;{_input:br}&#39;.split(&#39;.&#39;)[-4]}&quot;)
    }
      else 
      {
    ss_bf = tibble(snps = susie_obj$variable_name, cs_index = ifelse(is.null(susie_obj$sets$cs_index), 0, paste0(susie_obj$sets$cs_index,collapse =&quot;,&quot;)),names = &quot;${_input:bnnn}&quot;)
     }
    
    ss_bf = ss_bf%&gt;%cbind(susie_obj$lbf_variable%&gt;%t)%&gt;%as_tibble()
    
    return(ss_bf)
    }
    
    extract_variants_pip = function(susie_obj,region_list){
    susie_tb = tibble( variants =  names(susie_obj$pip)[which( susie_obj$pip &gt;= 0)],
                           snps_index = which(( susie_obj$pip &gt;= 0))) %&gt;%
        mutate(chromosome = map_chr(variants, ~read.table(text = .x, sep = &quot;:&quot;)$V1%&gt;%str_replace(&quot;chr&quot;,&quot;&quot;) ),
                position  = map_chr(variants, ~read.table(text = .x, sep = &quot;:&quot;)$V2  ),
                ref = map_chr(position , ~read.table(text = .x, sep = &quot;_&quot;,colClasses = &quot;character&quot;)$V2  ),
                alt = map_chr(position , ~read.table(text = .x, sep = &quot;_&quot;,colClasses = &quot;character&quot;)$V3  ),
                position  = map_dbl(position , ~read.table(text = .x, sep = &quot;_&quot;,as.is = T)$V1  )
                             )
      susie_tb = susie_tb%&gt;%mutate(cs_order =(map(susie_tb$snps_index , ~tryCatch(which(pmap(list( a= susie_obj$sets$cs) , function(a) .x %in% a )%&gt;%unlist()), error = function(e) return(0) )  ))%&gt;%as.character%&gt;%str_replace(&quot;integer\\(0\\)&quot;,&quot;0&quot;),
                         cs_id = map_chr(cs_order,~ifelse(.x ==&quot;0&quot;, &quot;None&quot; ,names(susie_obj$sets$cs)[.x%&gt;%str_split(&quot;:&quot;)%&gt;%unlist%&gt;%as.numeric] ) ),
                         log10_base_factor = map_chr(snps_index,~paste0( susie_obj$lbf_variable[,.x],  collapse = &quot;;&quot;)),
                         pip = susie_obj$pip,
                         posterior_mean = coef.susie(susie_obj)[-1],
                         posterior_sd = susie_get_posterior_sd(susie_obj),
                         z = posterior_mean/posterior_sd)
    
          susie_tb =  susie_tb%&gt;%mutate(  molecular_trait_id = region_list$molecular_trait_id,
                             finemapped_region_start = region_list$finemapped_region_start,
                             finemapped_region_end = region_list$finemapped_region_end)
          return(susie_tb)    }
          
        

     extract_effect_pip = function(susie_obj,region_list,susie_tb){
      result_tb =  tibble(phenotype = susie_obj$name,
        V = susie_obj$V,effect_id = paste0(&quot;L&quot;,1:length(V) ) ,
        cs_log10bf = susie_obj$lbf)
        if(is.null(susie_obj$sets$cs)){
            cs_min_r2 = cs_avg_r2 =  coverage =  0 
            cs = &quot;None&quot;} else {         cs = map_chr(susie_obj$sets$cs[result_tb$effect_id],~susie_tb$variants[.x]%&gt;%paste0(collapse = &quot;;&quot;))
        coverage = map(result_tb$effect_id, ~susie_obj$sets$coverage[which(names(susie_obj$sets$cs) == .x )])%&gt;%as.numeric%&gt;%replace_na(0)
        cs_min_r2  = (susie_obj$sets$purity[result_tb$effect_id,1])%&gt;%as.numeric%&gt;%replace_na(0)  
        cs_avg_r2  = (susie_obj$sets$purity[result_tb$effect_id,2])%&gt;%as.numeric%&gt;%replace_na(0) }
        result_tb = result_tb%&gt;%mutate(cs_min_r2 = cs_min_r2,cs_avg_r2 = cs_avg_r2 ,coverage = coverage%&gt;%unlist,cs = cs )            
      return(result_tb)
      }
       
  
    susie_obj = readRDS(&quot;${_input:a}&quot;)
    if(&quot;variants&quot; %in% names(susie_obj) ){susie_obj$variants = susie_obj$variants%&gt;%str_replace(&quot;_&quot;,&quot;:&quot;)}
    if(is.null(names(susie_obj$pip ))){names(susie_obj$pip) = susie_obj$variants}
    lbf = extract_lbf(susie_obj)
    region_list = read_delim(&quot;${region_list}&quot;,&quot;\t&quot;)
    if(ncol(region_list) == 3 ){   region_list =  region_list%&gt;%mutate(`#chr` = `#chr`%&gt;%str_remove_all(&quot; &quot;) , ID = paste0(`#chr`,&quot;_&quot;,start,&quot;_&quot;,end) ) } # LD_list 
    if(region_list$start[1] - region_list$end[1]  == -1 ){ 
        region_list = region_list%&gt;%mutate( start = start - ${windows} ,end = start +${windows}) # region_list for fix cis windows  
          } 
      if(&quot;gene_id&quot; %in% colnames(region_list)){region_list = region_list%&gt;%mutate(ID = gene_id)  } # region_list for gene
    region_list = region_list%&gt;%select(molecular_trait_id = ID, chromosome  = `#chr`,finemapped_region_start = start ,finemapped_region_end = end)  # Formatting
    mole_id = &quot;${f&#39;{_input:br}&#39;.split(&#39;.&#39;)[-4]}&quot;%&gt;%gsub(&quot;_N:&quot;,&quot;_*:&quot;,.)#for sQTL
    region_list = region_list%&gt;%filter(molecular_trait_id == mole_id)
    variants_pip = extract_variants_pip( susie_obj , region_list)
    effect_pip = extract_effect_pip( susie_obj , region_list,variants_pip)
    lbf%&gt;%write_delim(&quot;${_output[1]}&quot;,&quot;\t&quot;)
    variants_pip%&gt;%write_delim(&quot;${_output[0]}&quot;,&quot;\t&quot;)
    effect_pip%&gt;%write_delim(&quot;${_output[2]}&quot;,&quot;\t&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[fsusie_to_tsv_1]
# Input
# For complete susie, region_list or tad_list, for susie_rss , LD region list 
parameter: region_list = path
region_tbl = pd.read_csv(region_list,sep = &quot;\t&quot;)
parameter: rds_path = paths
input: rds_path, group_by = 1
output: f&quot;{cwd}/{_input:bn}.variant.tsv&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]:nn}.stdout&quot;, stderr = f&quot;{_output[0]:nn}.stderr&quot;, container = container, entrypoint = entrypoint
    library(&quot;dplyr&quot;)
    library(&quot;tibble&quot;)
    library(&quot;purrr&quot;)
    library(&quot;tidyr&quot;)
    library(&quot;readr&quot;)
    library(&quot;stringr&quot;)
    library(&quot;susieR&quot;)

    extract_variants_pip = function(susie_obj,region_list){
        susie_tb = tibble( variants =  names(susie_obj$csd_X),
                               snps_index = which(( susie_obj$pip &gt;= 0))) %&gt;%
            mutate(chromosome = map_chr(variants, ~read.table(text = .x, sep = &quot;:&quot;)$V1%&gt;%str_replace(&quot;chr&quot;,&quot;&quot;) ),
                    position  = map_chr(variants, ~read.table(text = .x, sep = &quot;:&quot;)$V2  ),
                    ref = map_chr(position , ~read.table(text = .x, sep = &quot;_&quot;,colClasses = &quot;character&quot;)$V2  ),
                    alt = map_chr(position , ~read.table(text = .x, sep = &quot;_&quot;,colClasses = &quot;character&quot;)$V3  ),
                    position  = map_dbl(position , ~read.table(text = .x, sep = &quot;_&quot;,as.is = T)$V1  )
                                 )
          susie_tb = susie_tb%&gt;%mutate(cs_order =(map(susie_tb$snps_index , ~tryCatch(which(pmap(list( a= susie_obj$cs) , function(a) .x %in% a )%&gt;%unlist()), error = function(e) return(0) )  ))%&gt;%as.character%&gt;%str_replace(&quot;integer\\(0\\)&quot;,&quot;0&quot;),
                             pip = susie_obj$pip)
          susie_tb =  susie_tb%&gt;%mutate(  molecular_trait_id = region_list$tad_index,
                                 finemapped_region_start = region_list$start,
                                 finemapped_region_end = region_list$end)
          if(&quot;purity&quot; %in% names(susie_obj)){
              susie_tb = susie_tb%&gt;%mutate(purity = map_dbl(susie_tb$cs_order, ~ifelse(.x%&gt;%as.numeric &gt; 0, susie_obj$purity[[as.numeric(.x)]], NA ) ), is_dummy = as.numeric(purity &lt; 0.5)  )
              }
    susie_tb = susie_tb%&gt;%mutate(effect_peak_pos = map_dbl(cs_order, ~ifelse(.x%&gt;%as.numeric &gt; 0, susie_obj$outing_grid[which(abs(susie_obj$fitted_func[[as.numeric(.x)]]) == max(abs(susie_obj$fitted_func[[as.numeric(.x)]])))] , NA ) )) 
    susie_tb_lbf = cbind(susie_tb%&gt;%select(molecular_trait_id,variants,cs_order),Reduce(cbind, susie_obj$lBF)%&gt;%as.tibble%&gt;%`colnames&lt;-`(1:length(susie_obj$lBF)))
          return(list(susie_tb, susie_tb_lbf))    }
    susie_obj = readRDS(&quot;${_input:a}&quot;)
    region_list = read_delim(&quot;${region_list}&quot;,&quot;\t&quot;)
    region_list = region_list%&gt;%filter(tad_index == &quot;${f&#39;{_input:br}&#39;.split(&#39;.&#39;)[-4]}&quot;)
    variants_pip = extract_variants_pip( susie_obj , region_list)[[1]]
    variants_lbf = extract_variants_pip( susie_obj , region_list)[[2]]
    print(paste0(&quot;fsusie run time is &quot;, round(susie_obj$runtime[[3]]/60),&quot;min&quot;))
    variants_pip%&gt;%write_delim(&quot;${_output}&quot;,&quot;\t&quot;)
    variants_pip%&gt;%write_delim(&quot;${_output:nn}.lbf.tsv&quot;,&quot;\t&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[*_to_tsv_2]
parameter: name = f&#39;{_input[0]:b}&#39;.split(&quot;.&quot;)[0]
input: group_by = &quot;all&quot;
output: f&quot;{cwd}/{name}.all_variants.tsv&quot;
bash: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    head -1 ${_input[0]} &gt; ${_output}
    cat ${_input[0]:d}/*variant.tsv | grep -v cs_order &gt;&gt; ${_output}
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[susie_tsv_collapse]
parameter: tsv_path = paths # TSV needs have the name ends with  *.chr1_2_3.unisusie(_rss).lbf.tsv
tsv_list  = pd.DataFrame({&quot;lbf_path&quot; : [str(x) for x in tsv_path]})
chromosome = list(set([f&#39;{x.split(&quot;.&quot;)[-5].split(&quot;_&quot;)[0].replace(&quot;chr&quot;,&quot;&quot;)}&#39;  for x in tsv_list.lbf_path ])) ## Add chr if there is no chr prefix. This is to accomodata chr XY and M
input: tsv_path, for_each = &quot;chromosome&quot;
output: f&#39;{cwd}/{_input[0]:bnnnnnnn}.chr{_chromosome}.unisusie_rss.lbf.tsv&#39;
bash: expand = &#39;${ }&#39;, stdout = f&quot;{_output}.stdout&quot;, stderr = f&quot;{_output}.stderr&quot;, container = container, entrypoint = entrypoint
        head -1 ${_input[0]} &gt; ${_output}
        cat ${_input[0]:d}/*.chr${_chromosome}_*lbf.tsv | grep -v cs_index &gt;&gt; ${_output}
        
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[susie_pip_landscape_plot]
parameter: plot_list = path
parameter: annot_tibble = path(&quot;~/Annotatr_builtin_annotation_tibble.tsv&quot;)
import pandas as pd
plot_list  = pd.read_csv(plot_list,sep = &quot;\t&quot;)
file_type = plot_list.columns.values.tolist()
file_type = [x.split(&quot;.&quot;)[0] for x in file_type ]
plot_list = plot_list.to_dict(&quot;records&quot;)
input: plot_list, group_by = len(file_type)
output: f&#39;{cwd}/{&quot;_&quot;.join(file_type)}.{str(_input[0]).split(&quot;.&quot;)[-5]}.pip_landscape_plot.rds&#39;,f&#39;{cwd}/{&quot;_&quot;.join(file_type)}.{str(_input[0]).split(&quot;.&quot;)[-5]}.pip_landscape_plot.pdf&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]}.stdout&quot;, stderr = f&quot;{_output[0]}.stderr&quot;, container = container, entrypoint = entrypoint
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;) 
    library(&quot;ggplot2&quot;)
    library(&quot;purrr&quot;)
    color = c(&quot;black&quot;, &quot;dodgerblue2&quot;, &quot;green4&quot;, &quot;#6A3D9A&quot;, 
          &quot;#FF7F00&quot;, &quot;gold1&quot;, &quot;skyblue2&quot;, &quot;#FB9A99&quot;, &quot;palegreen2&quot;,
          &quot;#CAB2D6&quot;, &quot;#FDBF6F&quot;, &quot;gray70&quot;, &quot;khaki2&quot;, &quot;maroon&quot;, &quot;orchid1&quot;,
          &quot;deeppink1&quot;, &quot;blue1&quot;, &quot;steelblue4&quot;, &quot;darkturquoise&quot;, &quot;green1&quot;, 
          &quot;yellow4&quot;, &quot;yellow3&quot;,&quot;darkorange4&quot;,&quot;brown&quot;,&quot;navyblue&quot;,&quot;#FF0000&quot;,
          &quot;darkgreen&quot;,&quot;#FFFF00&quot;,&quot;purple&quot;,&quot;#00FF00&quot;,&quot;pink&quot;,&quot;#0000FF&quot;,
          &quot;orange&quot;,&quot;#FF00FF&quot;,&quot;cyan&quot;,&quot;#00FFFF&quot;,&quot;#FFFFFF&quot;)
    extract_table = function(variant_df,type){ 
    if(&quot;purity&quot; %in% colnames(variant_df) ){
      variant_df$purity[is.na(variant_df$purity)] = 0
      variant_df[abs(variant_df$purity) &lt; 0.5,7] = 0
      }
     variant_df = variant_df%&gt;%mutate(CS = (cs_order%&gt;%as.factor%&gt;%as.numeric-1)%&gt;%as.factor)%&gt;%
          select( y = pip ,snp = variants,pos = position , CS, molecular_trait_id)%&gt;%mutate(molecular_trait_id = paste0(type,&quot;_&quot;,molecular_trait_id ) )
    return(variant_df)
    }
    plot_recipe = tibble( type =  c(&#39;${&quot;&#39;,&#39;&quot;.join(file_type) }&#39;), path = c(${_input:r,}))
    plot_list = map2(plot_recipe$type,plot_recipe$path, ~read_delim(.y, guess_max = 10000000)%&gt;%extract_table(.x) )
    plot_df = Reduce(rbind,plot_list)
    plot_range = (plot_df%&gt;%group_by(molecular_trait_id)%&gt;%summarize(start = (min(pos)), end = (max(pos)))%&gt;%mutate(start = median(start),end = median(end)))[1,c(2,3)]%&gt;%as.matrix
    plot_chr = (plot_df$snp[1]%&gt;%stringr::str_split(&quot;:&quot;))[[1]][1]
    plot_df = plot_df%&gt;%mutate(Shared = as.logical(map(snp, ~(plot_df%&gt;%filter( snp ==.x ,   CS%&gt;%as.numeric !=  1 )%&gt;%nrow()) &gt; 1  )))
    pip_plot &lt;- plot_df%&gt;%ggplot2::ggplot(aes(y = y, x = pos,
                                  col =  CS, shape = Shared )) + facet_grid(molecular_trait_id ~.)+
      geom_point(size = 7) +
      scale_color_manual(&quot;CS&quot;,values = color) +
      theme(axis.ticks.x = element_blank()) +
     ylab(&quot;Posterior Inclusion Probability (PIP)&quot;)+xlim(plot_range)+
      theme(axis.ticks.x = element_blank()) +
            theme(strip.text.y.right = element_text(angle = 0))+
            xlab(&quot;&quot;) + 
            theme(text = element_text(size = 30))+ggtitle(&quot;Overview of fine-mapping&quot;)
  
    annot = read_delim(&quot;${annot_tibble}&quot;)
    annot = annot%&gt;%filter(seqnames == plot_chr, start &gt; plot_range[1], end &lt; plot_range[2])
    annot_plot   = annot%&gt;%filter(!type%in%c(&quot;hg38_genes_introns&quot;,&quot;hg38_genes_1to5kb&quot;))%&gt;%
                        ggplot(aes())+
                        geom_segment( aes(x = start,xend = end, y = &quot;Regulartory Element&quot;, yend = &quot;Regulartory Element&quot;, color = type ), linewidth =10)+
                        ylab(&quot;&quot;)+xlab(&quot;&quot;)+xlim(plot_range)+theme(axis.text.x=element_blank(),text = element_text(size = 20))+scale_color_brewer(palette=&quot;Dark2&quot;)
    gene_plot = annot%&gt;%filter(type%in%c(&quot;hg38_genes_1to5kb&quot;))%&gt;%group_by(symbol)%&gt;%
                            summarise(start = min(start), end = max(end))%&gt;%na.omit%&gt;%
                            ggplot(aes())+geom_segment( aes(x = start,xend = end, y = &quot;Gene&quot;, yend = &quot;Gene&quot;, color = symbol ), linewidth =10)+
                            geom_label(aes(x = (start+end)/2,y = &quot;Gene&quot;, label = symbol ),size = 5)+ylab(&quot;&quot;)+xlab(&quot;POS&quot;)+
                            theme(legend.position=&quot;none&quot;)+theme(text = element_text(size = 20))+xlim(plot_range)
      
    list(pip_plot,plot_df,annot_plot,gene_plot)%&gt;%saveRDS(&quot;${_output[0]}&quot;)
    cowplot::plot_grid(plotlist = list(pip_plot,annot_plot,gene_plot),ncol = 1, align = &quot;v&quot;,axis = &quot;tlbr&quot;,rel_heights = c(8,1,1))%&gt;%ggsave(filename = &quot;${_output[1]}&quot;,device = &quot;pdf&quot;,dpi = &quot;retina&quot;,width = 30, height = 30)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[susie_upsetR_plot]
parameter: plot_list = path
import pandas as pd
plot_list  = pd.read_csv(plot_list, sep = &quot;\t&quot;)
file_type = plot_list.columns.values.tolist()
file_type = [x.split(&quot;.&quot;)[0] for x in file_type ]
plot_list = plot_list.to_dict(&quot;records&quot;)
input: plot_list
output: f&#39;{cwd}/{&quot;_&quot;.join(file_type)}.UpSetR.rds&#39;,f&#39;{cwd}/{&quot;_&quot;.join(file_type)}.UpSetR.pdf&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]}.stdout&quot;, stderr = f&quot;{_output[0]}.stderr&quot;, container = container, entrypoint = entrypoint
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;) 
    library(&quot;ggplot2&quot;)
    library(&quot;purrr&quot;)
    library(&quot;UpSetR&quot;)
    library(&quot;ComplexUpset&quot;)
    plot_recipe = tibble( type =  c(&#39;${&quot;&#39;,&#39;&quot;.join(file_type) }&#39;), path = c(${_input:r,}))
    plot_list = map2(plot_recipe$type,plot_recipe$path, ~read_delim(.y, guess_max = 10000000)%&gt;%mutate(cs = cs_order != 0 )%&gt;%filter(cs &gt; 0)%&gt;%select(variants,cs)%&gt;%`colnames&lt;-`(c(&quot;variants&quot;,.x))%&gt;%distinct() )
    cs_sharing = Reduce(full_join,plot_list)
    cs_upsetR_sharing = cs_sharing
    cs_upsetR_sharing[,2:ncol(cs_upsetR_sharing)]%&gt;%mutate_all(as.numeric)-&gt; cs_upsetR_sharing[,2:ncol(cs_upsetR_sharing)]
    a = upset(cs_upsetR_sharing%&gt;%as.data.frame,intersect = colnames(cs_upsetR_sharing[2:ncol(cs_upsetR_sharing)]),
      keep_empty_groups = F,
          base_annotations=list(`Intersection size` = intersection_size( bar_number_threshold = 1, position = position_dodge(0.5), width = 0.3 ,text = list(size = 5)   )  ) ,
              themes=upset_default_themes(axis.text=element_text(size=30))     ,
              min_degree = 1)
    
    a%&gt;%ggsave(filename = &quot;${_output[1]}&quot;,device = &quot;pdf&quot;,dpi = &quot;retina&quot;,width=18.5, height=10.5)
    list(cs_upsetR_sharing)%&gt;%saveRDS(&quot;${_output[0]}&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[susie_upsetR_cs_plot]
parameter: plot_list = path
import pandas as pd
plot_list  = pd.read_csv(plot_list, sep = &quot;\t&quot;)
file_type = plot_list.columns.values.tolist()
file_type = [x.split(&quot;.&quot;)[0] for x in file_type ]
plot_list = plot_list.to_dict(&quot;records&quot;)
parameter: trait_to_select =  1 
input: plot_list
output: f&#39;{cwd}/{&quot;_&quot;.join(file_type)}.UpSetR_{file_type[trait_to_select-1]}_cs.rds&#39;,f&#39;{cwd}/{&quot;_&quot;.join(file_type)}.UpSetR_{file_type[trait_to_select-1]}_cs.pdf&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]}.stdout&quot;, stderr = f&quot;{_output[0]}.stderr&quot;, container = container, entrypoint = entrypoint

    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;) 
    library(&quot;ggplot2&quot;)
    library(&quot;purrr&quot;)
    library(&quot;UpSetR&quot;)
    library(&quot;ComplexUpset&quot;)

    cs_sharing_identifer = function(upsetR_input,df){
    inner_join(upsetR_input, df%&gt;%select(variants,molecular_trait_id, cs_order)%&gt;%filter(cs_order != 0))%&gt;%select(-variants)-&gt; dfL_CS_sharing
    dfL_CS_sharing[is.na(dfL_CS_sharing)] = FALSE
    dfL_CS_sharing = dfL_CS_sharing%&gt;%group_by(molecular_trait_id,cs_order)%&gt;%summarize(across(everything(), list(mean))   )
    dfL_CS_sharing = dfL_CS_sharing%&gt;%mutate(across(colnames(dfL_CS_sharing)[3:ncol(dfL_CS_sharing)], ~.x != 0  ))%&gt;%`colnames&lt;-`(c(&quot;molecular_trait_id&quot;,&quot;cs_order&quot;,colnames(cs_sharing)[2:ncol(cs_sharing)]))
    }
  
  
    plot_recipe = tibble( type =  c(&#39;${&quot;&#39;,&#39;&quot;.join(file_type) }&#39;), path = c(${_input:r,}))
    plot_list = map2(plot_recipe$type,plot_recipe$path, ~read_delim(.y, guess_max = 10000000)%&gt;%mutate(cs = cs_order != 0 )%&gt;%filter(cs &gt; 0)%&gt;%select(variants,cs)%&gt;%`colnames&lt;-`(c(&quot;variants&quot;,.x))%&gt;%distinct() )
    cs_sharing = Reduce(full_join,plot_list)
    cs_upsetR_sharing = cs_sharing
    cs_upsetR_sharing[,2:ncol(cs_upsetR_sharing)]%&gt;%mutate_all(as.numeric)-&gt; cs_upsetR_sharing[,2:ncol(cs_upsetR_sharing)]

    df = read_delim(plot_recipe$path[[${trait_to_select}]]) 
    
    cs_sharing_df = cs_sharing_identifer(cs_upsetR_sharing,df)

    a = upset(cs_sharing_df%&gt;%as.data.frame,intersect = colnames(cs_sharing_df[3:ncol(cs_sharing_df)]),
          keep_empty_groups = F,
          base_annotations=list(`Intersection size` = intersection_size( bar_number_threshold = 1, position = position_dodge(0.5), width = 0.3 ,text = list(size = 8)   )  ),
          themes=upset_default_themes(axis.text=element_text(size=30)),
          set_size = F  ,  min_degree = 1,wrap = T) + ggtitle( paste0(plot_recipe$type[[${trait_to_select}]],&#39;CS shared with other phenotypes&#39;) )   + theme(plot.title = element_text(size = 40, face = &quot;bold&quot;))
  
    a%&gt;%ggsave(filename = &quot;${_output[1]}&quot;,device = &quot;pdf&quot;,dpi = &quot;retina&quot;,width=18.5, height=10.5)
    list(cs_sharing_df)%&gt;%saveRDS(&quot;${_output[0]}&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[tmp_annotatation_of_snps_1]
parameter: SNP_list = path
parameter: annotation_rds = path
input: SNP_list
output: f&#39;{cwd}/{_input:b}.annotated.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output}.stdout&quot;, stderr = f&quot;{_output}.stderr&quot;, container = container, entrypoint = entrypoint
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;) 
    library(&quot;purrr&quot;)
    library(&quot;stringr&quot;)
    sharing_snp = readRDS(&quot;${_input}&quot;)
    sharing_snp_fsusie = sharing_snp[[1]]%&gt;%filter(haQTL == 1 | mQTL == 1)
    sharing_snp_fsusie = sharing_snp_fsusie%&gt;%mutate(X1 =  read.table(text = sharing_snp_fsusie$variants, sep = &quot;:&quot;)$V1, X2 = read.table(text = read.table(text = sharing_snp_fsusie$variants, sep = &quot;:&quot;)$V2 , sep = &quot;_&quot;)$V1  )
    sharing_snp_fsusie = sharing_snp_fsusie%&gt;%select(variants,chr = X1, pos = X2)
    annotation = readRDS(&quot;${annotation_rds}&quot;)
    print(&quot;data loaded&quot;)
    result = sharing_snp_fsusie%&gt;%mutate(annot = map2( chr,pos , ~ annotation%&gt;%filter(X1 == .x, X2 &lt;= .y, X3 &gt;= .y)%&gt;%pull(X5)))%&gt;%mutate(annot = map_chr(annot, ~paste0(.x ,collapse = &quot;,&quot;)) )
    print(&quot;snp annotated&quot;)
    result%&gt;%saveRDS(&quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[tmp_annotatation_of_snps_2]
parameter: SNP_list = path
parameter: annotation_rds = path
output: f&#39;{cwd}/{_input:b}.annotated_rev.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output}.stdout&quot;, stderr = f&quot;{_output}.stderr&quot;, container = container, entrypoint = entrypoint
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;) 
    library(&quot;purrr&quot;)
    library(&quot;stringr&quot;)
    result = readRDS(${_input:r})
    result_rev = tibble(annot = unique(annotation$X5))%&gt;%mutate(variants = map(annot, ~  result%&gt;%filter( str_detect(annot,.x))%&gt;%pull(variants)) )%&gt;%mutate( variants = map_chr(variants,~paste0(.x ,collapse = &quot;,&quot;))  )
    result_rev%&gt;%saveRDS(&quot;${_output}&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[fsusie_extract_effect]
parameter: rds_path = paths
parameter: annot_tibble = path(&quot;~/Annotatr_builtin_annotation_tibble.tsv&quot;)
input: rds_path, group_by = 1
output: f&#39;{cwd}/{_input:bn}.estimated_effect.tsv&#39;,f&#39;{cwd}/{_input:bn}.estimated_effect.pdf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]}.stdout&quot;, stderr = f&quot;{_output[0]}.stderr&quot;, container = container, entrypoint = entrypoint
    library(&quot;stringr&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;) 
    library(&quot;ggplot2&quot;)
    library(&quot;purrr&quot;)
    library(&quot;tidyr&quot;)
    color = c(&quot;black&quot;, &quot;dodgerblue2&quot;, &quot;green4&quot;, &quot;#6A3D9A&quot;, 
          &quot;#FF7F00&quot;, &quot;gold1&quot;, &quot;skyblue2&quot;, &quot;#FB9A99&quot;, &quot;palegreen2&quot;,
          &quot;#CAB2D6&quot;, &quot;#FDBF6F&quot;, &quot;gray70&quot;, &quot;khaki2&quot;, &quot;maroon&quot;, &quot;orchid1&quot;,
          &quot;deeppink1&quot;, &quot;blue1&quot;, &quot;steelblue4&quot;, &quot;darkturquoise&quot;, &quot;green1&quot;, 
          &quot;yellow4&quot;, &quot;yellow3&quot;,&quot;darkorange4&quot;,&quot;brown&quot;,&quot;navyblue&quot;,&quot;#FF0000&quot;,
          &quot;darkgreen&quot;,&quot;#FFFF00&quot;,&quot;purple&quot;,&quot;#00FF00&quot;,&quot;pink&quot;,&quot;#0000FF&quot;,
          &quot;orange&quot;,&quot;#FF00FF&quot;,&quot;cyan&quot;,&quot;#00FFFF&quot;,&quot;#FFFFFF&quot;)
    
    effect_extract = function(fsusie){
        plot_df = fsusie$fitted_func%&gt;%as_tibble(.name_repair = &quot;universal&quot;)%&gt;%mutate(pos =  fsusie$outing_grid)%&gt;%`colnames&lt;-`(c(paste0(&quot;Effect_&quot;,1:length(fsusie$cs)),&quot;pos&quot;))%&gt;%mutate(`#chr` = str_split(names(fsusie$csd_X)[[1]],&quot;:&quot;)[[1]][[1]] )%&gt;%select(`#chr`, pos, everything())
        plot=  plot_df%&gt;%pivot_longer(cols = 3:ncol(plot_df),names_to = &quot;effect&quot;, values_to = &quot;values&quot;   ) %&gt;%ggplot(aes(x = pos, y = values,color = effect),linewidth = 7)+
        geom_line()+ylab(&quot;Estimated Effect&quot;) + xlab(&quot;POS&quot;)+facet_grid(effect~. )+
            scale_color_manual(&quot;Credible set&quot;,values = color[2:length(color)])+geom_line(aes(y = 0), color = &quot;black&quot;)
            theme(strip.text.y.right = element_text(angle = 0))+
            xlab(&quot;&quot;) + 
            ylab(&quot;Estimated Effect&quot;)+ 
            theme(text = element_text(size = 50))+
            ggtitle(paste0( &quot;Estimated effect for ${f&#39;{_input:br}&#39;.split(&#39;.&#39;)[-4]}&quot;))
        return(list(plot_df,plot))
        }
    susie_obj =  readRDS(&quot;${_input}&quot;)
    output = effect_extract(susie_obj)
    effect_tbl = output[[1]]
    annot = read_delim(&quot;${annot_tibble}&quot;)
    annot = annot%&gt;%filter(seqnames == (effect_tbl$`#chr`)[[1]], start &gt; min(effect_tbl$pos), end &lt; max(effect_tbl$pos))
    plot_range = c(min(effect_tbl$pos),  max(effect_tbl$pos))
    annot_plot   = annot%&gt;%filter(!type%in%c(&quot;hg38_genes_introns&quot;,&quot;hg38_genes_1to5kb&quot;))%&gt;%
                        ggplot(aes())+
                        geom_segment( aes(x = start,xend = end, y = &quot;Regulartory Element&quot;, yend = &quot;Regulartory Element&quot;, color = type ), linewidth =10)+
                        ylab(&quot;&quot;)+xlab(&quot;&quot;)+xlim(plot_range)+theme(axis.text.x=element_blank(),text = element_text(size = 20))+scale_color_brewer(palette=&quot;Dark2&quot;)
    gene_plot = annot%&gt;%filter(type%in%c(&quot;hg38_genes_1to5kb&quot;))%&gt;%group_by(symbol)%&gt;%
                            summarise(start = min(start), end = max(end))%&gt;%na.omit%&gt;%
                            ggplot(aes())+geom_segment( aes(x = start,xend = end, y = &quot;Gene&quot;, yend = &quot;Gene&quot;, color = symbol ), linewidth =10)+
                            geom_label(aes(x = (start+end)/2,y = &quot;Gene&quot;, label = symbol ),size = 5)+ylab(&quot;&quot;)+xlab(&quot;POS&quot;)+
                            theme(legend.position=&quot;none&quot;)+theme(text = element_text(size = 20))+xlim(plot_range)
    cowplot::plot_grid(plotlist = list(output[[2]]),ncol = 1, align = &quot;v&quot;,axis = &quot;tlbr&quot;,rel_heights = c(8))%&gt;%ggsave(filename = &quot;${_output[1]}&quot;,device = &quot;pdf&quot;,dpi = &quot;retina&quot;,width = 30, height = 30)
    effect_tbl%&gt;%write_delim(&quot;${_output[0]}&quot;,&quot;\t&quot;)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[fsusie_affected_region]
parameter: rds_path = paths
input: rds_path, group_by = 1
output: f&#39;{cwd}/{_input:bn}.affected_region.tsv&#39;,f&#39;{cwd}/{_input:bn}.affected_region.pdf&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output[0]}.stdout&quot;, stderr = f&quot;{_output[0]}.stderr&quot;, container = container, entrypoint = entrypoint
    library(&quot;stringr&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;) 
    library(&quot;ggplot2&quot;)
    library(&quot;purrr&quot;)
    library(&quot;tidyr&quot;)
    library(susiF.alpha)
    library(ashr)
    library(wavethresh)
    color = c(&quot;black&quot;, &quot;dodgerblue2&quot;, &quot;green4&quot;, &quot;#6A3D9A&quot;, 
          &quot;#FF7F00&quot;, &quot;gold1&quot;, &quot;skyblue2&quot;, &quot;#FB9A99&quot;, &quot;palegreen2&quot;,
          &quot;#CAB2D6&quot;, &quot;#FDBF6F&quot;, &quot;gray70&quot;, &quot;khaki2&quot;, &quot;maroon&quot;, &quot;orchid1&quot;,
          &quot;deeppink1&quot;, &quot;blue1&quot;, &quot;steelblue4&quot;, &quot;darkturquoise&quot;, &quot;green1&quot;, 
          &quot;yellow4&quot;, &quot;yellow3&quot;,&quot;darkorange4&quot;,&quot;brown&quot;,&quot;navyblue&quot;,&quot;#FF0000&quot;,
          &quot;darkgreen&quot;,&quot;#FFFF00&quot;,&quot;purple&quot;,&quot;#00FF00&quot;,&quot;pink&quot;,&quot;#0000FF&quot;,
          &quot;orange&quot;,&quot;#FF00FF&quot;,&quot;cyan&quot;,&quot;#00FFFF&quot;,&quot;#FFFFFF&quot;)
    ## Define Function
    update_cal_credible_band2 &lt;- function(susiF.obj )
    {
    
    
    
      if(sum( is.na(unlist(susiF.obj$alpha))))
      {
        stop(&quot;Error: some alpha value not updated, please update alpha value first&quot;)
      }
      temp &lt;- wavethresh::wd(rep(0, susiF.obj$n_wac))
    
    
      for ( l in 1:susiF.obj$L)
      {
        Smat &lt;-  susiF.obj$fitted_wc2[[l]]
        W1   &lt;- ((wavethresh::GenW(n=  ncol(Smat )  , filter.number = 10, family = &quot;DaubLeAsymm&quot;)))
        tt   &lt;- diag( W1%*%diag(c(susiF.obj$alpha[[l]]%*%Smat ))%*% t(W1 ))
    
        up                       &lt;-  susiF.obj$fitted_func[[l]]+ 3*sqrt(tt)
        low                      &lt;-  susiF.obj$fitted_func[[l]]- 3*sqrt(tt)
        susiF.obj$cred_band[[l]] &lt;- rbind(up, low)
      }
    
    
    
      return(susiF.obj)
    }
    
    affected_reg &lt;- function( susiF.obj){
      outing_grid &lt;- susiF.obj$outing_grid
    
      reg &lt;-  list()
      h &lt;- 1
      for (   l in 1:length(susiF.obj$cs)){
    
        pos_up &lt;-  which(susiF.obj$cred_band[[l]][1,]&lt;0)
        pos_low &lt;- which(susiF.obj$cred_band[[l]][2,]&gt;0)
    
    
        reg_up &lt;- split( pos_up,cumsum(c(1,diff( pos_up)!=1)))
    
        reg_low &lt;- split( pos_low,cumsum(c(1,diff( pos_low)!=1)))
        for( k in 1:length(reg_up)){
          reg[[h]] &lt;- c(l, outing_grid[reg_up[[k]][1]], outing_grid[reg_up[[k]][length(reg_up[[k]])]])
    
          h &lt;- h+1
        }
        for( k in 1:length(reg_low )){
          reg[[h]] &lt;- c(l, outing_grid[reg_low [[k]][1]], outing_grid[reg_low [[k]][length(reg_low [[k]])]])
    
          h &lt;- h+1
        }
    
    
      }
      reg &lt;-  do.call(rbind, reg)
      colnames(reg) &lt;- c(&quot;CS&quot;, &quot;Start&quot;,&quot;End&quot;)
      return(reg)
    }
    
    
    susiF_obj =  readRDS(&quot;${_input}&quot;)
    susiF_obj = update_cal_credible_band2(susiF_obj)
    affected_tbl = affected_reg(susiF_obj)
    affected_tbl = affected_tbl%&gt;%as_tibble%&gt;%mutate(analysis = susiF_obj$name,
                          chr = (names(susiF_obj$csd_X)[[1]]%&gt;%stringr::str_split(&quot;:&quot;))[[1]][[1]],
                          molecular_trait_id =  &quot;${f&#39;{_input:br}&#39;.split(&#39;.&#39;)[-4]}&quot;, 
                          purity  = purrr::map_dbl(CS,~susiF_obj$purity[[.x]] )  )
    
    affected_tbl%&gt;%as_tibble%&gt;%write_delim(&quot;${_output[0]}&quot;,&quot;\t&quot;)
    plt = plot_susiF(susiF_obj, cred.band = T)
    plt%&gt;%ggsave(filename = &quot;${_output[1]}&quot;,device = &quot;pdf&quot;,dpi = &quot;retina&quot;,width = 30, height = 30)
</pre></div>
</div>
</div>
</div>
</section>
<section id="into-vcf-format">
<h2>Into VCF format<a class="headerlink" href="#into-vcf-format" title="Permalink to this heading">#</a></h2>
<p>FIXME: These codes were moved from earlier workflows. To be cleaned up and tested.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>#[uni_susie_2]
input: group_with = &quot;genoFile&quot;
output: f&quot;{_input:n}.vcf.bgz&quot;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:nn}.stdout&quot;, stderr = f&quot;{_output:nn}.stderr&quot;, container = container, entrypoint = entrypoint
   ## Define create_vcf function
           create_vcf = function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL, 
        se = NULL, pval = NULL, name = NULL,cs = NULL, pip = NULL) 
    {
        stopifnot(length(chrom) == length(pos))
        if (is.null(snp)) {
            snp &lt;- paste0(chrom, &quot;:&quot;, pos)
        }
        snp &lt;- paste0(chrom, &quot;:&quot;, pos)
        nsnp &lt;- length(chrom)
        gen &lt;- list()
        ## Setupt data content for each sample column
        if (!is.null(ea_af)) 
            gen[[&quot;AF&quot;]] &lt;- matrix(ea_af, nsnp)
        if (!is.null(effect)) 
            gen[[&quot;ES&quot;]] &lt;- matrix(effect, nsnp)
        if (!is.null(se)) 
            gen[[&quot;SE&quot;]] &lt;- matrix(se, nsnp)
        if (!is.null(pval)) 
            gen[[&quot;LP&quot;]] &lt;- matrix(-log10(pval), nsnp)
        if (!is.null(cs)) 
            gen[[&quot;CS&quot;]] &lt;- matrix(cs, nsnp)
        if (!is.null(pip)) 
            gen[[&quot;PIP&quot;]] &lt;- matrix(pip, nsnp)
        gen &lt;- S4Vectors::SimpleList(gen)
        
      ## Setup snps info for the fix columns
        gr &lt;- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, 
            end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))
         coldata &lt;- S4Vectors::DataFrame(Studies = name, row.names = name)
    ## Setup header informations
        hdr &lt;- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = &quot;VCFv4.2&quot;, 
            row.names = &quot;fileformat&quot;)), sample = name)
        VariantAnnotation::geno(hdr) &lt;- S4Vectors::DataFrame(Number = c(&quot;A&quot;, 
            &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;), Type = c(&quot;Float&quot;, &quot;Float&quot;, 
            &quot;Float&quot;, &quot;Float&quot;, &quot;Float&quot;, &quot;Float&quot;), Description = c(&quot;Effect size estimate relative to the alternative allele&quot;, 
            &quot;Standard error of effect size estimate&quot;, &quot;-log10 p-value for effect estimate&quot;,  
            &quot;Alternate allele frequency in the association study&quot;,
            &quot;The CS this variate are captured, 0 indicates not in any cs&quot;, &quot;The posterior inclusion probability to a CS&quot;), 
            row.names = c(&quot;ES&quot;, &quot;SE&quot;, &quot;LP&quot;, &quot;AF&quot;, &quot;CS&quot;, &quot;PIP&quot;))
      ## Save only the meta information in the sample columns 
        VariantAnnotation::geno(hdr) &lt;- subset(VariantAnnotation::geno(hdr), 
            rownames(VariantAnnotation::geno(hdr)) %in% names(gen))
      ## Save VCF 
        vcf &lt;- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, 
            exptData = list(header = hdr), geno = gen)
        VariantAnnotation::alt(vcf) &lt;- Biostrings::DNAStringSetList(as.list(ea))
        VariantAnnotation::ref(vcf) &lt;- Biostrings::DNAStringSet(nea)
      ## Add fixed values
        VariantAnnotation::fixed(vcf)$FILTER &lt;- &quot;PASS&quot;
          return(sort(vcf))
        }
    library(&quot;susieR&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;tibble&quot;)
    library(&quot;purrr&quot;)
    library(&quot;readr&quot;)
    library(&quot;tidyr&quot;)
    library(&quot;stringr&quot;)
    
    # Get list of cs snps
    susie_list = readRDS(${_input:r})
    susie_tb_ls = list()
    for (i in 1:length(susie_list)){
        susie_tb = tibble( snps =  names(susie_list[[1]]$pip)[which( susie_list[[i]]$pip &gt;= 0)], snps_index = which(( susie_list[[i]]$pip &gt;= 0))  )
        susie_tb_ls[[i]]= susie_tb%&gt;%mutate( cs = map(snps_index,~which( susie_list[[i]]$sets$cs %in% .x))%&gt;%as.numeric%&gt;%replace_na(0),
                                 pip = map_dbl(snps_index,~( susie_list[[i]]$pip[.x])),
                                 coef = map_dbl(snps_index,~(coef.susie( susie_list[[i]])[.x+1])))
        }
    if(length(susie_tb_ls) &gt;= 2){ 
      for(i in 2:length(susie_tb_ls)){
          susie_tb_ls[[i]] = full_join(susie_tb_ls[[i-1]],susie_tb_ls[[i]], by = &quot;snps&quot;) 
        }
    }
    m = c(&quot;cs&quot;,&quot;pip&quot;,&quot;coef&quot;)    
    output = list()
    for(i in m){
    output[[i]] = susie_tb_ls[[length(susie_tb_ls)]]%&gt;%select(contains(i))%&gt;%as.matrix
    }
    snps_tb = susie_tb_ls[[length(susie_tb_ls)]]%&gt;%mutate(
                         chr = map_chr(snps,~read.table(text = .x,sep = &quot;:&quot;,as.is = T)$V1),
                         pos_alt_ref = map_chr(snps,~read.table(text = .x,sep = &quot;:&quot;,as.is = TRUE)$V2),
                         pos = map_dbl(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE)$V1),
                         alt = map_chr(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE, colClass = &quot;character&quot;)$V2),
                         ref = map_chr(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE, colClass = &quot;character&quot;)$V3))
    
    snps_tb = snps_tb%&gt;%filter(str_detect(ref, &quot;[ACTG]&quot;) &amp; str_detect(alt, &quot;[ACTG]&quot;))
    output_vcf = create_vcf(
            chrom = snps_tb$chr,
             pos = snps_tb$pos,
             ea = snps_tb$alt,
             nea = snps_tb$ref,
             effect = snps_tb%&gt;%select(contains(&quot;coef&quot;))%&gt;%as.matrix ,
             pip = snps_tb%&gt;%select(contains(&quot;pip&quot;))%&gt;%as.matrix,
             cs = snps_tb%&gt;%select(contains(&quot;cs&quot;))%&gt;%as.matrix,
             name = names(susie_list))
    VariantAnnotation::writeVcf(output_vcf,${_output:nr},index = TRUE)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[mv_susie_2]
input: group_with = &quot;genoFile&quot;
output: f&quot;{_input:n}.vcf.bgz&quot;
task: trunk_workers = 1, trunk_size = 1, walltime = &#39;2h&#39;, mem = &#39;55G&#39;, cores = 1, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:nn}.stdout&quot;, stderr = f&quot;{_output:nn}.stderr&quot;, container = container, entrypoint = entrypoint
   ## Define create_vcf function
           create_vcf = function (chrom, pos, nea, ea, snp = NULL, ea_af = NULL, effect = NULL, 
        se = NULL, pval = NULL, name = NULL,cs = NULL, pip = NULL) 
    {
        stopifnot(length(chrom) == length(pos))
        if (is.null(snp)) {
            snp &lt;- paste0(chrom, &quot;:&quot;, pos)
        }
        snp &lt;- paste0(chrom, &quot;:&quot;, pos)
        nsnp &lt;- length(chrom)
        gen &lt;- list()
        ## Setupt data content for each sample column
        if (!is.null(ea_af)) 
            gen[[&quot;AF&quot;]] &lt;- matrix(ea_af, nsnp)
        if (!is.null(effect)) 
            gen[[&quot;ES&quot;]] &lt;- matrix(effect, nsnp)
        if (!is.null(se)) 
            gen[[&quot;SE&quot;]] &lt;- matrix(se, nsnp)
        if (!is.null(pval)) 
            gen[[&quot;LP&quot;]] &lt;- matrix(-log10(pval), nsnp)
        if (!is.null(cs)) 
            gen[[&quot;CS&quot;]] &lt;- matrix(cs, nsnp)
        if (!is.null(pip)) 
            gen[[&quot;PIP&quot;]] &lt;- matrix(pip, nsnp)
        gen &lt;- S4Vectors::SimpleList(gen)
        
      ## Setup snps info for the fix columns
        gr &lt;- GenomicRanges::GRanges(chrom, IRanges::IRanges(start = pos, 
            end = pos + pmax(nchar(nea), nchar(ea)) - 1, names = snp))
         coldata &lt;- S4Vectors::DataFrame(Studies = name, row.names = name)
    ## Setup header informations
        hdr &lt;- VariantAnnotation::VCFHeader(header = IRanges::DataFrameList(fileformat = S4Vectors::DataFrame(Value = &quot;VCFv4.2&quot;, 
            row.names = &quot;fileformat&quot;)), sample = name)
        VariantAnnotation::geno(hdr) &lt;- S4Vectors::DataFrame(Number = c(&quot;A&quot;, 
            &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;, &quot;A&quot;), Type = c(&quot;Float&quot;, &quot;Float&quot;, 
            &quot;Float&quot;, &quot;Float&quot;, &quot;Float&quot;, &quot;Float&quot;), Description = c(&quot;Effect size estimate relative to the alternative allele&quot;, 
            &quot;Standard error of effect size estimate&quot;, &quot;-log10 p-value for effect estimate&quot;,  
            &quot;Alternate allele frequency in the association study&quot;,
            &quot;The CS this variate are captured, 0 indicates not in any cs&quot;, &quot;The posterior inclusion probability to a CS&quot;), 
            row.names = c(&quot;ES&quot;, &quot;SE&quot;, &quot;LP&quot;, &quot;AF&quot;, &quot;CS&quot;, &quot;PIP&quot;))
      ## Save only the meta information in the sample columns 
        VariantAnnotation::geno(hdr) &lt;- subset(VariantAnnotation::geno(hdr), 
            rownames(VariantAnnotation::geno(hdr)) %in% names(gen))
      ## Save VCF 
        vcf &lt;- VariantAnnotation::VCF(rowRanges = gr, colData = coldata, 
            exptData = list(header = hdr), geno = gen)
        VariantAnnotation::alt(vcf) &lt;- Biostrings::DNAStringSetList(as.list(ea))
        VariantAnnotation::ref(vcf) &lt;- Biostrings::DNAStringSet(nea)
      ## Add fixed values
        VariantAnnotation::fixed(vcf)$FILTER &lt;- &quot;PASS&quot;
          return(sort(vcf))
        }
    library(&quot;susieR&quot;)
    library(&quot;dplyr&quot;)
    library(&quot;tibble&quot;)
    library(&quot;purrr&quot;)
    library(&quot;readr&quot;)
    library(&quot;tidyr&quot;)
    
    # Get list of cs snps
    res = readRDS(${_input:r})
    output_snps = tibble( snps = res$variable_name[which(res$pip &gt;= 0)], snps_index = which((res$pip &gt;= 0))  )
    output_snps = output_snps%&gt;%mutate( cs = map(snps_index,~which(res$sets$cs %in% .x))%&gt;%as.numeric%&gt;%replace_na(0),
                             pip = map_dbl(snps_index,~(res$pip[.x])),
                     chr = map_chr(snps,~read.table(text = .x,sep = &quot;:&quot;,as.is = T)$V1),
                     pos_alt_ref = map_chr(snps,~read.table(text = .x,sep = &quot;:&quot;,as.is = TRUE)$V2),
                     pos = map_dbl(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE)$V1),
                     alt = map_chr(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE, colClass = &quot;character&quot;)$V2),
                     ref = map_chr(pos_alt_ref,~read.table(text = .x,sep = &quot;_&quot;,as.is = TRUE, colClass = &quot;character&quot;)$V3))
    
    effect_mtr = res$coef[output_snps$snps_index+1]%&gt;%as.matrix
    colnames(effect_mtr) = &quot;${name}&quot;
    rownames(effect_mtr) = output_snps$snps
    cs_mtr = effect_mtr
    for(i in 1:nrow(cs_mtr)) cs_mtr[i,] =  output_snps$cs[[i]]  
    pip_mtr = effect_mtr
    for(i in 1:nrow(pip_mtr)) pip_mtr[i,] =  output_snps$pip[[i]]  
    
    output_vcf = create_vcf(
           chrom = output_snps$chr,
            pos = output_snps$pos,
            ea = output_snps$alt,
            nea = output_snps$ref,
            effect = effect_mtr ,
            pip = pip_mtr,
            cs = cs_mtr,
            name = colnames(effect_mtr)
              )
    VariantAnnotation::writeVcf(output_vcf,${_output:nr},index = TRUE)
</pre></div>
</div>
</div>
</div>
</section>
<section id="cis-window-analysis-result-consolidation">
<h2>Cis-window analysis Result consolidation<a class="headerlink" href="#cis-window-analysis-result-consolidation" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Exporting cis susie_twas results
[cis_results_export_1, gwas_results_export_1]
# per chunk we process at most 200 datasets
parameter: per_chunk = 200
# Region list should have last column being region name. 
# Meta-data will be written to this column
parameter:region_file=path()
parameter:file_path=&#39;&#39;
parameter:prefix=[]
parameter:suffix=str
parameter: condition_meta = path()
# if the pip all variants in a cs &lt; this threshold, then remove this cs
parameter: pip_thres = 0.05
# only keep top cs_size variants in one cs 
parameter: cs_size = 3
# provide exported meta to filter the exported genes 
parameter: exported_file = path()
# optional: qtl or gwas, there is slightly different in qtl and gwas rds file
parameter: data_type = str
# Optional: if a region list is provide the analysis will be focused on provided region. 
# The LAST column of this list will contain the ID of regions to focus on
# Otherwise, all regions with both genotype and phenotype files will be analyzed
parameter: region_list = path()
# Optional: if a region name is provided 
# the analysis would be focused on the union of provides region list and region names
parameter: region_name = []
parameter: min_corr = 0.5
import pandas as pd
import os

region = pd.read_csv(region_file, sep=&#39;\t&#39;, names=[&#39;chr&#39;, &#39;start&#39;, &#39;end&#39;, &#39;id&#39;])
region_ids = []
# If region_list is provided, read the file and extract IDs
if not region_list.is_dir():
    if region_list.is_file():
        region_list_df = pd.read_csv(region_list, sep=&#39;\t&#39;, header=None, comment = &quot;#&quot;)
        region_ids = region_list_df.iloc[:, -1].unique()  # Extracting the last column for IDs
    else:
        raise ValueError(&quot;The region_list path provided is not a file.&quot;)
        
if len(region_name) &gt; 0:
    region_ids = list(set(region_ids).union(set(region_name)))
    
if len(region_ids) &gt; 0:
    region = region[region[&#39;id&#39;].isin(region_ids)]

# Function to create list of formatted strings for each row
def create_formatted_list(row):
    if len(prefix) &gt; 0:
        formatted_list = [f&quot;{file_path}/{p}.{row[&#39;id&#39;]}.{suffix}&quot; for p in prefix]
    else:
        formatted_list = [f&quot;{file_path}/{row[&#39;id&#39;]}.{suffix}&quot;]  # GWAS data do not have prefix         
    return formatted_list
# Apply the function to each row
region[&#39;original_data&#39;] = region.apply(create_formatted_list, axis=1)

def group_by_region(lst, partition):
    # from itertools import accumulate
    # partition = [len(x) for x in partition]
    # Compute the cumulative sums once
    # cumsum_vector = list(accumulate(partition))
    # Use slicing based on the cumulative sums
    # return [lst[(cumsum_vector[i-1] if i &gt; 0 else 0):cumsum_vector[i]] for i in range(len(partition))]
    return partition

def filter_existing_paths(row):
    existing_paths = [path for path in row if os.path.exists(path)]
    return existing_paths

def is_file_exported(paths, results_set):
    for path in paths:
        basename = os.path.basename(path)
        if basename not in results_set:
            return False
    return True

region[&#39;original_data&#39;] = region[&#39;original_data&#39;].apply(filter_existing_paths)
region = region[region[&#39;original_data&#39;].map(bool)]

# if provided exported meta, check if the original data are all exported already, isfo skip them 
if not exported_file.is_dir():
    if exported_file.is_file():
        results = pd.read_csv(exported_file, sep=&#39;\t&#39;)
        results_set = set(results[&#39;original_data&#39;])
        results_set = {item.strip() for sub in results_set for item in sub.split(&#39;,&#39;)}
        mask = region[&#39;original_data&#39;].apply(lambda paths: not is_file_exported(paths, results_set))
        region = region[mask]
    else:
        raise ValueError(&quot;The exported_file path provided is not a file.&quot;)

regional_data = {
    &#39;meta&#39;: [(row[&#39;chr&#39;], row[&#39;start&#39;], row[&#39;end&#39;], row[&#39;id&#39;]) for _, row in region.iterrows()],
    &#39;data&#39;: [(row[&#39;original_data&#39;]) for _, row in region.iterrows()]
}

meta_info = regional_data[&#39;meta&#39;]
stop_if(len(regional_data[&#39;data&#39;]) == 0, f&#39; All files have been exported already&#39;)

input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&quot;{cwd}/{name}_cache/{name}_batch{_index+1}.tsv&quot;
task: trunk_workers = job_size, walltime = walltime, trunk_size = job_size, mem = mem, cores = numThreads, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;,stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container, entrypoint=entrypoint
    library(tidyverse)
    # check top_loci existing or not
    has_rows &lt;- function(df) {
      !is.null(df) &amp;&amp; nrow(df) &gt; 0
    }
    # update top loci table
    update_top_loci_min_corr &lt;- function(dat_susie, top_loci_df, coverage_value, threshold = ${min_corr}) {
      if(coverage_value == &#39;cs_coverage_0.95&#39;) purity_res &lt;- dat_susie$sets$purity
       else purity_res &lt;- dat_susie$sets_secondary[[gsub(&#39;cs_&#39;,&#39;&#39;,coverage_value)]]$sets$purity
      not_pass_min_cs &lt;- rownames(purity_res)[purity_res$min.abs.corr &lt; threshold] %&gt;%
        gsub(&#39;L&#39;, &#39;&#39;, .)
      top_loci_df[[paste0( coverage_value, &quot;_min_corr&quot;)]] &lt;- top_loci_df[[coverage_value]]
      if (length(not_pass_min_cs) &gt; 0) {
        top_loci_df[[paste0( coverage_value, &quot;_min_corr&quot;)]][top_loci_df[[coverage_value]] %in% not_pass_min_cs] &lt;- 0
      }
        return(top_loci_df)
    }

    # function to filter the cs with all variants pip &lt; 0.05
    update_and_filter_cs_ids &lt;- function(dat_susie, df) {
      cs_columns &lt;- grep(&quot;^cs_coverage&quot;, names(df), value = TRUE)
      df$cs_all_non_zero_orig &lt;- ifelse(rowSums(df[cs_columns] == 0) == length(cs_columns), FALSE, TRUE)
      for (cs_column in cs_columns) {
        unique_cs_ids &lt;- unique(df[[cs_column]])
        for (cs_id in unique_cs_ids) {
          if (cs_id &gt; 0) {
            # pip check
            pip_check &lt;- df[df[[cs_column]] == cs_id, &quot;pip&quot;] &lt; 0.05
            if (all(pip_check, na.rm = TRUE)) {
              df[[cs_column]] &lt;- ifelse(df[[cs_column]] == cs_id, 0, df[[cs_column]])
            }

            # size check
            # new_col_name &lt;- paste0(cs_column, &quot;_refine&quot;)
            # df &lt;- df %&gt;%
            #   group_by(.data[[cs_column]]) %&gt;%
            #   mutate(rank = row_number(-pip)) %&gt;%
            #   mutate(!!new_col_name := if_else(rank &lt;= 3, .data[[cs_column]], as.integer(0))) %&gt;% # if we want to replace original cs_column, just replace new_col_name with cs_column here, would filter it
            #   ungroup() %&gt;%
            #   select(-rank)

          }
         df &lt;- update_top_loci_min_corr(dat_susie, top_loci_df = df,coverage_value = cs_column)   
        }
      }
      # Filter out rows where all cs_coverage columns are 0 and cs_all_non_zero_orig is TRUE
      df &lt;- df %&gt;% filter(!(cs_all_non_zero_orig &amp; rowSums(df[cs_columns] == 0) == length(cs_columns)))
      # Remove the helper column
      df &lt;- select(df, -cs_all_non_zero_orig)
      return(df)
    }

    refine_cs_size &lt;- function(df){
      return(x)
    }

    # function to decide run update_and_filter_cs_ids or not
    process_top_loci &lt;- function(dat_susie, data_frame) {
      if (has_rows(data_frame) &gt; 0) {
        return(update_and_filter_cs_ids(dat_susie, data_frame))
      } else {
        return(data_frame)
      }
    }

    get_pip_withname &lt;- function(susie_obj){
      pip = susie_obj$susie_result_trimmed$pip 
      names(pip) = susie_obj$variant_names
      return(pip)
    }

    # Process each path and collect results
    orig_files = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input])})
     # Extract info from each RDS file
    results &lt;- list()
    gene = &quot;${_meta_info[3]}&quot;

    # Post Processing: Extracting info
    # use for loop instead of apply to save memory
    res &lt;- res_sum &lt;- cons_top_loci &lt;- list()
    for(i in seq_along(orig_files)) {
      rds_path &lt;- orig_files[i]
       dat &lt;- tryCatch({
        readRDS(rds_path)
      }, error = function(e) {
        writeLines(rds_path %&gt;% basename, gsub(&quot;.tsv&quot;,&quot;_error&quot;,&quot;${_output}&quot;))
        return(NULL) # 
      })

      if(is.null(dat)) next
  
      #extract qtl type from susie rds file name, if we have set decent condtiton name, this could be removed 
      # qtl_type &lt;- sub(&quot;(\\.*?)\\..*&quot;, &quot;\\1&quot;, basename(rds_path)) %&gt;% sub(&quot;.*?_&quot;, &quot;\\1&quot;, .)

      temp_list &lt;- list() # Temporary list to store inner results
        for(s in names(dat)){
         dat_study &lt;- dat[[s]]
          temp_list &lt;- list()
          conditions &lt;- names(dat_study) %&gt;% .[!str_detect(pattern = &#39;sumstats&#39;,.)]
  
          if(length(conditions) &gt; 0) {
  
            for(condition in conditions) {
                dat_con &lt;- dat_study[[condition]]
                dat_susie &lt;- dat_con$susie_result_trimmed
  
                if(${&quot;TRUE&quot; if condition_meta.is_file() else &quot;FALSE&quot;}){
                      meta &lt;- suppressMessages(read_delim(&quot;${condition_meta}&quot;, col_names = F))
                      qtl_con &lt;- meta %&gt;% filter(X1 == condition) %&gt;% pull(X2)
                      if (length(qtl_con) == 0) {
                        qtl_con &lt;- condition
                        message(&quot;No matching entries found. qtl_con has been set to the condition value.&quot;)
                      }
                    } else {qtl_con &lt;- condition}
  
                res[[gene]][[s]][[qtl_con]] &lt;- list(
                  region_info = dat_con$region_info,
                  top_loci = process_top_loci(dat_susie, dat_con$top_loci),
                  preset_top_loci = process_top_loci(dat_susie, dat_con$preset_variants_result$top_loci),
                  pip = get_pip_withname(dat_con),
                  preset_pip = get_pip_withname(dat_con$preset_variants_result),
                  CV_table = dat_con$twas_cv_result$performance
                )
  
                res_sum[[gene]][[s]][[qtl_con]] &lt;- list(
                  variant_names = dat_con$variant_names,
                  sumstats = dat_con$sumstats
                )
                
                if(has_rows(dat_con$top_loci) || has_rows(dat_con$preset_top_loci)) cons_top_loci[[gene]][[s]][[qtl_con]] &lt;- qtl_con else  cons_top_loci[[gene]][[s]][[qtl_con]] &lt;- NULL
              }
            }
        }
    }
    cons_top_loci &lt;- cons_top_loci %&gt;% compact()  # Use &#39;compact&#39; to remove NULLs, assuming &#39;results&#39; has the correct structure  
    if (&#39;${data_type}&#39; == &#39;qtl&#39;){
      if(length(res) &gt; 0) res = res[[1]] else res = list() 
      if(length(res_sum) &gt; 0) res_sum = res_sum[[1]] else res_sum = list()
      if(length(cons_top_loci) &gt; 0) cons_top_loci = cons_top_loci[[1]] else cons_top_loci = NA
    }
  
    combine_data = paste0(&quot;${_output:add}&quot;,&quot;/&quot;,&quot;${name}&quot;, &quot;.&quot;, gene, &quot;.cis_results_db.export.rds&quot;)
    combine_data_sumstats = paste0(&quot;${_output:add}&quot;,&quot;/&quot;,&quot;${name}&quot;, &quot;.&quot;, gene, &quot;.cis_results_db.export_sumstats.rds&quot;)
  
    if (${&quot;TRUE&quot; if exported_file.is_file() else &quot;FALSE&quot;}){
        res_exp &lt;- readRDS(combine_data)
        res[[gene]] &lt;- c(res[[gene]], res_exp[[gene]])
        res_sum_exp &lt;- readRDS(combine_data_sumstats)
        res_sum[[gene]] &lt;- c(res_sum[[gene]], res_sum_exp[[gene]])
    }
    saveRDS(res, combine_data)
    saveRDS(res_sum, combine_data_sumstats)
    TSS &lt;- tryCatch({dat_con$region_info$region_coord$start},  error = function(e) {return(NA)})
  
    meta = data.frame(chr=&quot;${_meta_info[0]}&quot;, start=&quot;${_meta_info[1]}&quot;, end=&quot;${_meta_info[2]}&quot;, gene_id=&quot;${_meta_info[3]}&quot;, TSS =  if(is.null(TSS)) NA else TSS, 
                      original_data = paste(basename(orig_files), collapse = &quot;, &quot;), combined_data = basename(combine_data), combined_data_sumstats = basename(combine_data_sumstats), 
                      conditions = paste(names(res[[gene]]), collapse = &quot;,&quot;), 
                      conditions_top_loci = if(length(cons_top_loci) &gt; 0) cons_top_loci[[1]] %&gt;% unlist %&gt;% names %&gt;% as.character %&gt;% paste(., collapse = &#39;,&#39;) else &#39;&#39;)
    write_delim(meta, &quot;${_output}&quot;, delim = &#39;\t&#39;)
  
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[cis_results_export_2, gwas_results_export_2]
# provide exported meta to filter the exported genes 
parameter: exported_file = path()
# optional: qtl or gwas, there is slightly different in qtl and gwas rds file
parameter: data_type = str
input: group_by = &#39;all&#39;
output: f&quot;{cwd}/{name}.{&#39;cis_results_db&#39; if data_type == &#39;qtl&#39; else &#39;block_results_db&#39;}.tsv&quot; 
# stop_if(_input[0] not in locals().keys(), &#39;All files have been exported already&#39;) #FIXME should we remove to a separate file. sothat we can stop globally as above
task: trunk_workers = 1, walltime = &#39;1h&#39;, trunk_size = 1, mem = &#39;16G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
bash: expand = &quot;${ }&quot;, container = container, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    if [ -e &quot;${_output:ad}/${name}_cache/&quot; ]; then
        sed &#39;s/^chr/#chr/&#39; `ls ${_output:ad}/${name}_cache/*tsv |head -n1` | head -n 1 &gt; ${_output:an}.temp
        tail -n +2 -q ${_output:ad}/${name}_cache/*.tsv &gt;&gt; ${_output:an}.temp
        error_files=$(find &quot;${_output:ad}/${name}_cache/&quot; -type f -name &quot;*_error&quot;)

        if [[ -n $error_files ]]; then
            cat $error_files &gt;&gt; ${_output:an}.error_genes
        else
            echo &quot;No truncated files detected&quot;
        fi
    else
        echo &quot;All files have been exported already&quot;
    fi
    
R: expand = &quot;${ }&quot;, container = container, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    if (file.exists(paste0(${_output:anr},&quot;.temp&quot;))) {
        library(tidyverse)
        meta &lt;- read_delim(paste0(${_output:anr},&quot;.temp&quot;), delim = &#39;\t&#39;)

        if (${&quot;TRUE&quot; if exported_file.is_file() else &quot;FALSE&quot;}){
          exp_meta &lt;- read_delim(${_output:r}, delim = &#39;\t&#39;)
          meta &lt;- bind_rows(meta, exp_meta) %&gt;%
              group_by(gene_id) %&gt;%
              summarise(across(c(original_data, combined_data, combined_data_sumstats, conditions, conditions_top_loci), 
                               ~paste(unique(.), collapse = &quot;,&quot;)),
                        .groups = &#39;drop&#39;)
              }

        write_delim(meta, ${_output:r}, delim = &#39;\t&#39;)
    }
bash: expand = &quot;${ }&quot;, container = container, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint   
    rm -rf &quot;${_output:ad}/${name}_cache/&quot;
    rm -rf ${_output:an}.temp
</pre></div>
</div>
</div>
</div>
</section>
<section id="gwas-results-consolidation">
<h2>GWAS results consolidation<a class="headerlink" href="#gwas-results-consolidation" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>#get union of step1
#1200 blocks costed ~5mins with in one for loop
[gwas_results_export_3]
output: f&quot;{cwd}/{name}.union_export.tsv.gz&quot; 
task: trunk_workers = 1, walltime = &#39;1h&#39;, trunk_size = 1, mem = &#39;16G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, container = container, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    library(tidyverse)
    library(data.table)

    mtx &lt;- read_delim(${_input:r})
    files &lt;- mtx %&gt;% filter(!is.na(conditions_top_loci)) %&gt;% pull(combined_data) %&gt;% paste0(${_input[0]:dr},&#39;/&#39;,.)
    all_top_loci &lt;- data.frame()

    for (i in seq_along(files)) {
      file &lt;- files[i]
      res &lt;- readRDS(file)

      file_top_loci &lt;- lapply(names(res), function(block) {
        lapply(names(res[[block]]), function(study) {
          lapply(names(res[[block]][[study]]), function(method) {
            if (!is.null(res[[block]][[study]][[method]]$top_loci)) {
              temp_df &lt;- res[[block]][[study]][[method]]$top_loci
              mutate(temp_df, study = study, method = method, block = block)
            } else {
              NULL  
            }
          })
        }) %&gt;% bind_rows()  
      }) %&gt;% bind_rows() 

      all_top_loci &lt;- bind_rows(all_top_loci, file_top_loci)
      if (i %% 100 == 0) {
        message(sprintf(&quot;Have processed %d files.&quot;, i))
      }
    }
    fwrite(all_top_loci, ${_output:r})
</pre></div>
</div>
</div>
</div>
</section>
<section id="overlap-qtl-and-gwas-results">
<h2>Overlap QTL and GWAS results<a class="headerlink" href="#overlap-qtl-and-gwas-results" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[overlap_qtl_gwas_1]
parameter: per_chunk = 100
parameter: gwas_meta_path = path()
parameter: qtl_file_path = &#39;&#39;
parameter: gwas_file_path = &#39;&#39;
parameter: qtl_meta_path = path()
# Optional: if a region list is provide the analysis will be focused on provided region. 
# The LAST column of this list will contain the ID of regions to focus on
# Otherwise, all regions with both genotype and phenotype files will be analyzed
parameter: region_list = path()
# Optional: if a region name is provided 
# the analysis would be focused on the union of provides region list and region names
parameter: region_name = []
import pandas as pd
import numpy as np
from pathlib import Path

# Load the data, suppressing messages is not typically done in pandas as it does not inherently output messages when loading files
gwas_meta = pd.read_csv(gwas_meta_path, sep=&#39;\t&#39;, low_memory=False)
gwas_meta = gwas_meta[gwas_meta[&#39;conditions_top_loci&#39;].notna()]

qtl_meta = pd.read_csv(qtl_meta_path, sep=&#39;\t&#39;, low_memory=False)
qtl_meta = qtl_meta[qtl_meta[&#39;conditions_top_loci&#39;].notna()]

# Filter and mutate operations, translated to pandas
gwas_meta[&#39;combined_data_toploci&#39;] = gwas_meta.apply(lambda row: row[&#39;combined_data&#39;] if pd.notnull(row[&#39;conditions_top_loci&#39;]) else pd.NA, axis=1)

region_ids=[]
# If region_list is provided, read the file and extract IDs
if not region_list.is_dir():
    if region_list.is_file():
        region_list_df = pd.read_csv(region_list, sep=&#39;\t&#39;, header=None, comment = &quot;#&quot;)
        region_ids = region_list_df.iloc[:, -1].unique()  # Extracting the last column for IDs
    else:
        raise ValueError(&quot;The region_list path provided is not a file.&quot;)
        
if len(region_name) &gt; 0:
    region_ids = list(set(region_ids).union(set(region_name)))
    
if len(region_ids) &gt; 0:
    qtl_meta = qtl_meta[qtl_meta[&#39;gene_id&#39;].isin(region_ids)]
    
def group_by_region(lst, partition):
    # from itertools import accumulate
    # partition = [len(x) for x in partition]
    # Compute the cumulative sums once
    # cumsum_vector = list(accumulate(partition))
    # Use slicing based on the cumulative sums
    # return [lst[(cumsum_vector[i-1] if i &gt; 0 else 0):cumsum_vector[i]] for i in range(len(partition))]
    return partition

grouped_gwas_meta = {k: v for k, v in gwas_meta.groupby(&#39;#chr&#39;)}
def check_overlap(gene_row, grouped_gwas_meta):
    chr_group = gene_row[&#39;#chr&#39;]
    if chr_group in grouped_gwas_meta:
        block_region = grouped_gwas_meta[chr_group]
        overlaps = block_region[
            (block_region[&#39;start&#39;] &lt;= gene_row[&#39;end&#39;]) &amp;
            (block_region[&#39;end&#39;] &gt;= gene_row[&#39;start&#39;])
        ]
        if not overlaps.empty:
            return &#39;,&#39;.join(overlaps[&#39;combined_data_toploci&#39;].astype(str))
    return pd.NA

stop_if(len(qtl_meta) == 0, f&#39;No file left for analysis &#39;)

qtl_meta_cand = qtl_meta.apply(lambda row: pd.Series({
    &#39;gwas_file&#39;: check_overlap(row, grouped_gwas_meta)
}), axis=1)

# Concatenate the new columns to the original qtl_meta DataFrame
qtl_meta_cand = pd.concat([qtl_meta, qtl_meta_cand], axis=1)
qtl_meta_filtered = qtl_meta_cand[qtl_meta_cand[&#39;gwas_file&#39;].notna()]
qtl_meta_filtered = qtl_meta_filtered.dropna(subset=[&#39;gwas_file&#39;])


regional_data = {
    &#39;meta&#39;: [(row[&#39;#chr&#39;], row[&#39;start&#39;], row[&#39;end&#39;], row[&#39;gene_id&#39;], row[&#39;gwas_file&#39;].split(&#39;,&#39;)) for _, row in qtl_meta_filtered.iterrows()],
    &#39;qtl_data&#39;: [f&quot;{qtl_file_path}/{row[&#39;combined_data&#39;]}&quot; for _, row in qtl_meta_filtered.iterrows()]
}

meta_info = regional_data[&#39;meta&#39;]
stop_if(len(regional_data[&#39;qtl_data&#39;]) == 0, f&#39;No file left for analysis &#39;)

input: regional_data[&quot;qtl_data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;qtl_data&quot;]), group_with = &quot;meta_info&quot;
output: f&quot;{cwd}/gwas_qtl/cache/{name}_gwas_batch_meta{_index+1}.tsv&quot;
task: trunk_workers = 1, walltime = &#39;1h&#39;, trunk_size = job_size, mem = &#39;16G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, container = container, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    library(tidyverse)
    library(plyr)
    library(data.table)
    # Function to check if the variant or its flipped version is in qtl_var
    check_flip &lt;- function(variant, qtl_var) {
      # Strand flip function
      strand_flip &lt;- function(ref) {
        flip &lt;- ref
        flip[ref == &quot;A&quot;] &lt;- &quot;T&quot;
        flip[ref == &quot;T&quot;] &lt;- &quot;A&quot;
        flip[ref == &quot;G&quot;] &lt;- &quot;C&quot;
        flip[ref == &quot;C&quot;] &lt;- &quot;G&quot;
        return(flip)
      }
       # Split the variant string into parts
      variant_parts &lt;- str_split(variant, &#39;:&#39;, simplify = TRUE)[1, ]

      # Extract the reference and alternate alleles
      ref_allele &lt;- variant_parts[3]
      alt_allele &lt;- variant_parts[4]

      # Flip the alleles using the strand_flip function
      flipped_ref &lt;- strand_flip(ref_allele)
      flipped_alt &lt;- strand_flip(alt_allele)

      # Create the flipped variant string
      variant_flip &lt;- paste(variant_parts[1], variant_parts[2], alt_allele, ref_allele, sep = &quot;:&quot;) # chr:pos:A:C &amp; chr:pos:C:A
      variant_strand &lt;- paste(variant_parts[1], variant_parts[2], flipped_alt, flipped_ref, sep = &quot;:&quot;) #  chr:pos:A:C &amp; chr:pos:T:G
      variant_strand_flip &lt;- paste(variant_parts[1], variant_parts[2], flipped_ref, flipped_alt, sep = &quot;:&quot;) # chr:pos:A:C &amp; chr:pos:G:T

      # Check if either the original variant or the flipped variant is in qtl_var
      return(variant %in% qtl_var | variant_flip %in% qtl_var | variant_strand %in% qtl_var | variant_strand_flip %in% qtl_var)
    }


    # Function to add &#39;chr&#39; in variants
    add_chr_prefix &lt;- function(var) {
      if (any(grepl(&quot;chr&quot;, var))) {
        var &lt;- var
      } else {
        var &lt;- paste0(&quot;chr&quot;, var)
      }
      return(var)
    }

    # Function to check if a dataframe has rows
    has_rows &lt;- function(df) {
      !is.null(df) &amp;&amp; nrow(df) &gt; 0
    }


    extract_top_loci &lt;- function(res, include_method = FALSE) {
      all_top_loci &lt;- lapply(names(res), function(region) {
        lapply(names(res[[region]]), function(study) {
          if (include_method) {
            method_results &lt;- lapply(names(res[[region]][[study]]), function(method) {
              top_loci &lt;- NULL
              if (!is.null(res[[region]][[study]][[method]]$top_loci) &amp;&amp; nrow(res[[region]][[study]][[method]]$top_loci) &gt; 0) {
                top_loci &lt;- mutate(res[[region]][[study]][[method]]$top_loci, study = study, method = method, region = region)
              }
              return(top_loci)
            })
            return(bind_rows(method_results))
          } else {
            top_loci &lt;- list()
            if (!is.null(res[[region]][[study]]$top_loci) &amp;&amp; nrow(res[[region]][[study]]$top_loci) &gt; 0) {
              top_loci[[length(top_loci) + 1]] &lt;- mutate(res[[region]][[study]]$top_loci, study = study, region = region, method = &#39;top_loci&#39;)
            }
            if (!is.null(res[[region]][[study]]$preset_top_loci) &amp;&amp; nrow(res[[region]][[study]]$preset_top_loci) &gt; 0) {
              top_loci[[length(top_loci) + 1]] &lt;- mutate(res[[region]][[study]]$preset_top_loci, study = study, region = region, method = &#39;preset_top_loci&#39;)
            }
            return(bind_rows(top_loci))
          }
        })
      }) %&gt;% bind_rows() %&gt;% na.omit()

      return(all_top_loci)
    }

  
    # load data 
    qtl_file = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input])})
     # Extract info from each RDS file
    gwas_files = c(${&quot;,&quot;.join(&#39;&quot;%s&quot;&#39; % x for x in _meta_info[4])}) %&gt;% paste0(&#39;${gwas_file_path}&#39;,&#39;/&#39;,.)
    # Process GWAS files
    gwas_all_top_loci &lt;- do.call(rbind, lapply(gwas_files, function(file) {
      res &lt;- readRDS(file)
      gwas_all_top_loci &lt;- extract_top_loci(res, include_method = TRUE) 
    }))
    if(!is.null(gwas_all_top_loci) &amp;&amp; nrow(gwas_all_top_loci) &gt; 0){# fixme: could remove this judge if we get a solid enough meta
      gwas_all_top_loci &lt;- gwas_all_top_loci%&gt;% select(-c(&#39;z&#39;))

      # Process QTL file
      qtl &lt;- readRDS(qtl_file)
      qtl_all_top_loci &lt;- extract_top_loci(qtl) 
      if(!is.null(qtl_all_top_loci) &amp;&amp; nrow(qtl_all_top_loci) &gt; 0){# fixme: could remove this judge if we get a solid enough meta
        # qtl_all_top_loci &lt;- qtl_all_top_loci%&gt;% select(-c(&#39;betahat&#39;,&#39;sebetahat&#39;,&#39;maf&#39;))

        cs_cal &lt;- c(&#39;cs_coverage_0.95&#39;,&#39;cs_coverage_0.7&#39;,&#39;cs_coverage_0.5&#39;)

        qtl_all_var &lt;- qtl_all_top_loci %&gt;%
            #filter(rowSums(.[,cs_cal]) &gt; 0 ) %&gt;% #fixme
            pull(variant_id)

        gwas_all_var &lt;- gwas_all_top_loci %&gt;%
            #filter(rowSums(.[,cs_cal]) &gt; 0 ) %&gt;% #fixme
            pull(variant_id)

        gwas_all_var &lt;- if(any(grepl(&quot;chr&quot;, qtl_all_var))) add_chr_prefix(gwas_all_var) else gsub(&quot;chr&quot;, &quot;&quot;, gwas_all_var)
        gwas_all_top_loci$variant_id &lt;- gwas_all_var
        int_var &lt;- gwas_all_top_loci$variant_id[check_flip(gwas_all_var, qtl_all_var)]
        if(length(int_var) &gt; 0){
            gwas_all_top_loci &lt;- gwas_all_top_loci %&gt;% filter(check_flip(gwas_all_var, qtl_all_var))

            all_top_loci &lt;- rbind.fill(gwas_all_top_loci, qtl_all_top_loci)
            fwrite(all_top_loci,  gsub(&#39;_gwas_batch_meta&#39;,&#39;_gwas_batch_export&#39;,${_output:r}))

            new_gwas &lt;- split(gwas_all_top_loci, gwas_all_top_loci$study)
            new_gwas &lt;- lapply(new_gwas, function(df) {
              split(df, df$method)
            })

            qtl[[1]] &lt;- c(qtl[[1]], new_gwas)
            new_qtl_path &lt;-  paste0(${_output:ddr},&quot;/&quot;,gsub(&quot;.rds&quot;,&quot;.overlapped.gwas.rds&quot;,basename(qtl_file)))
            saveRDS(qtl, new_qtl_path)
  
            block_top_loci = gwas_all_top_loci$region %&gt;% unique %&gt;% paste(., collapse = &#39;,&#39;)
            final_combined_data = new_qtl_path %&gt;% basename
        } else {block_top_loci = final_combined_data = NA} # fixme: could remove this judge if we get a solid enough meta
      } else {block_top_loci = final_combined_data = NA} # fixme: could remove this judge if we get a solid enough meta
    } else {
        block_top_loci = final_combined_data = NA
    }
 
    qtl_meta &lt;- suppressMessages(read_delim(&#39;${qtl_meta_path}&#39;))
    qtl_meta &lt;- qtl_meta %&gt;% filter(gene_id == &#39;${_meta_info[3]}&#39;)  %&gt;% mutate(block_top_loci = block_top_loci,
                                                     final_combined_data = final_combined_data)
    fwrite(qtl_meta, ${_output:r})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[overlap_qtl_gwas_2]
input: group_by = &#39;all&#39;
output: f&quot;{cwd}/{name}.overlapped.gwas.tsv&quot;
task: trunk_workers = 1, walltime = &#39;1h&#39;, trunk_size = job_size, mem = &#39;16G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;
R: expand = &quot;${ }&quot;, container = container, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    library(data.table)
    exp_path &lt;- ${_input[0]:adr}
    meta_files &lt;- c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input])})
    exp_files &lt;- list.files(exp_path, &quot;_gwas_batch_export&quot;, full.names = T)
    meta_list &lt;- exp_list &lt;- list()
    meta_combined &lt;- rbindlist(lapply(meta_files, fread), fill = TRUE)
    exp_combined &lt;- rbindlist(lapply(exp_files, fread), fill = TRUE)
    fwrite(exp_combined, gsub(&quot;tsv&quot;,&quot;export.csv.gz&quot;,&quot;${_output}&quot;))
    fwrite(meta_combined, &quot;${_output}&quot;, sep = &#39;\t&#39;)
  
bash: expand = &quot;${ }&quot;, container = container, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint=entrypoint
    # rm -rf ${_input[0]:adr}
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/post_processing"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="association_scan_post_processing.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Association result processing</p>
      </div>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extracting-susie-results">Extracting susie results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extracting-susie-rss-results-for-adgwas">Extracting susie_rss results for ADGWAS</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#extracting-fsusie-results">Extracting fsusie results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#exporting-cis-analysis-susie-twas-results">Exporting cis_analysis susie_twas results</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#export-gwas-data">Export gwas data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overlapped-gwas-data-and-eqtl-data">Overlapped gwas data and eQTL data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#plotting-the-pip-plot">Plotting the pip plot</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#into-vcf-format">Into VCF format</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#cis-window-analysis-result-consolidation">Cis-window analysis Result consolidation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#gwas-results-consolidation">GWAS results consolidation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#overlap-qtl-and-gwas-results">Overlap QTL and GWAS results</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
       Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>