{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "# GWAS integration: TWAS and MR\n",
    "\n",
    "## Introduction\n",
    "\n",
    "This module provides software implementations for transcriptome-wide association analysis (TWAS), and Mendelian Randomization using fine-mapping instrumental variables (IV). The procedures implements the MR procedure described in Zhang et al 2020 for \"causal\" effects estimation and model validation, with the unit of analysis being a single gene-trait pair.\n",
    "\n",
    "This procedure is based on the SuSiE-TWAS workflow --- it assumes that xQTL fine-mapping has been performed (to be used for both TWAS and MR) and moleuclar traits prediction weights pre-computed (to be used for TWAS). Cross validation for TWAS weights is optional but highly recommended.\n",
    "\n",
    "GWAS data required are GWAS summary statistics and LD matrix for the region of interest.\n",
    "\n",
    "### Step 1: TWAS \n",
    "\n",
    "1. Extract GWAS z-score for region of interest and corresponding LD matrix.\n",
    "2. (Optional) perform allele matching QC for the LD matrix with summary stats.\n",
    "3. Process weights: for LASSO, Elastic Net and mr.ash we have to take the weights as is for QTL variants overlapping with GWAS variants. For SuSiE weights it can be adjusted to exactly match GWAS variants.\n",
    "4. Perofrm TWAS test for multiple sets of weights. \n",
    "5. For each gene, filter TWAS results by keeping the best model selected by CV. Drop the genes that don't show good evidence of TWAS prediction weights.\n",
    "\n",
    "### Step 2: MR for candidate genes\n",
    "\n",
    "1. Limit MR only to those showing some evidence of TWAS significance AND have strong instrumental variable (fine-mapping PIP or CS). \n",
    "2. Use fine-mapped xQTL with GWAS data to perform MR. \n",
    "3. For multiple IV, aggregate individual IV estimates using a fixed-effect meta-analysis procedure.\n",
    "4. Identify and exclude results with severe violations of the exclusion restriction (ER) assumption.\n",
    "\n",
    "### Input\n",
    "\n",
    "#### GWAS Data Input Interface (Similar to `susie_rss`)\n",
    "\n",
    "**GWAS Summary Statistics Files**\n",
    "- **Input**: Vector of files for one or more GWAS studies.\n",
    "- **Format**: \n",
    "  - Tab-delimited files.\n",
    "  - First 4 columns: `chr`, `pos`, `a0`, `a1`\n",
    "  - Additional columns can be loaded using column mapping file see below  \n",
    "- **Column Mapping files (optional)**:\n",
    "  - Optional YAML file for custom column mapping.\n",
    "  - Required columns: `chr`, `pos`, `a0`, `a1`, either `z` or (`betahat` and `sebetahat`).\n",
    "  - Optional columns: `n`, `var_y`.\n",
    "\n",
    "**GWAS Summary Statistics Meta-File**: this is optional and helpful when there are lots of GWAS data to process via the same command\n",
    "- **Columns**: `study_id`, chromosome number, path to summary statistics file, optional path to column mapping file.\n",
    "- **Note**: Chromosome number `0` indicates a genome-wide file.\n",
    "\n",
    "eg: `gwas_meta.tsv`\n",
    "\n",
    "```\n",
    "study_id    chrom    file_path                 column_mapping_file\n",
    "study1      1        gwas1.tsv.gz         column_mapping.yml\n",
    "study2      2        gwas2.tsv.gz         column_mapping.yml\n",
    "```\n",
    "\n",
    "**LD Reference Metadata File**\n",
    "- **Format**: Single TSV file.\n",
    "- **Contents**:\n",
    "  - Columns: `chr`, `start`, `end`, path to the LD matrix, genomic build.\n",
    "  - LD matrix path format: comma-separated, first entry is the LD matrix, second is the bim file.\n",
    "- **Documentation**: Refer to our LD reference preparation document for detailed information (Tosin pending update).\n",
    "\n",
    "#### Output of Fine-Mapping & TWAS Pipeline\n",
    "\n",
    "**xQTL Weight Database files**\n",
    "- path to various weight DB files, comma delimited.\n",
    "\n",
    "**xQTL Weight Database Metadata File**: this is optional and helpful when TWAS is done genome-wide for many regions via the same command\n",
    "- **Types**: Gene-based or TAD-based.\n",
    "- **Structure**: \n",
    "  - RDS format.\n",
    "  - Organized hierarchically: region → condition → weight matrix.\n",
    "  - Each column represents a different method.\n",
    "- **Format**: `chrom`, `start`, `end`, `region_id`, `condition` (e.g., tissue type, QTL), path to various weight DB files, comma delimited.\n",
    "\n",
    "eg: `xqtl_meta.tsv`\n",
    "\n",
    "```\n",
    "chrom    start    end    region_id    condition    file_path\n",
    "1        1000     5000   region1      cohor1:tissue1:eQTL      weight1.rds, weight2.rds\n",
    "2        2000     6000   region2      cohor1:tissue1:eQTL      weight3.rds\n",
    "3        3000     7000   region3      cohor1:tissue1:eQTL      weight4.rds, weight5.rds\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "## Output\n",
    "\n",
    "TWAS FIXME this is incorrect for now.\n",
    "\n",
    "- Each row corresponds to a single SNP inferred as a member of a signal cluster, with columns including:\n",
    "   - `snp`: SNP name.\n",
    "   - `beta_eQTL`: eQTL effect.\n",
    "   - `se_eQTL`: Standard error of estimated eQTL effect.\n",
    "   - `beta_GWAS`: GWAS effect.\n",
    "   - `se_GWAS`: Standard error of GWAS effect.\n",
    "   - `cluster`: Signal cluster ID (credible sets index).\n",
    "   - `pip`: SNP posterior inclusion probability (PIP).\n",
    "   - `gene_id`: Gene name.\n",
    "\n",
    "\n",
    "MR\n",
    "\n",
    "-  The output includes the following columns for each gene:\n",
    "    - `gene_id`: Gene name.\n",
    "    - `num_cluster`: Number of credible sets of the gene.\n",
    "    - `num_instruments`: Number of instruments included in the gene.\n",
    "    - `spip`: Sum of PIP for credible sets of each gene.\n",
    "    - `grp_beta`: Signal-level estimates, combining SNP-level estimates from member SNPs weighted by their PIPs.\n",
    "    - `grp_se`: Standard error of signal-level estimates.\n",
    "    - `meta`: Gene-level estimate of the causal effect, aggregating signal-level estimates using a fixed-effect meta-analysis model.\n",
    "    - `se_meta`: Standard error of the gene-level estimate of the causal effect.\n",
    "    - `Q`: Cochran’s Q statistic.\n",
    "    - `I2`: $I^2$ statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# Workdir\n",
    "parameter: gwas_meta_file = path()\n",
    "parameter: xqtl_meta_file = path()\n",
    "parameter: container = ''\n",
    "import re\n",
    "parameter: entrypoint= ('micromamba run -a \"\" -n' + ' ' + re.sub(r'(_apptainer:latest|_docker:latest|\\.sif)$', '', container.split('/')[-1])) if container else \"\"\n",
    "parameter: job_size = 100\n",
    "parameter: walltime = \"1h\"\n",
    "parameter: mem = \"16G\"\n",
    "parameter: numThreads = 2\n",
    "\n",
    "def group_by_region(lst, partition):\n",
    "    # from itertools import accumulate\n",
    "    # partition = [len(x) for x in partition]\n",
    "    # Compute the cumulative sums once\n",
    "    # cumsum_vector = list(accumulate(partition))\n",
    "    # Use slicing based on the cumulative sums\n",
    "    # return [lst[(cumsum_vector[i-1] if i > 0 else 0):cumsum_vector[i]] for i in range(len(partition))]\n",
    "    return partition"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS",
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "[get_analysis_regions: shared = \"regional_data\"]\n",
    "import pandas as pd\n",
    "import os\n",
    "\n",
    "def file_exists(file_path, relative_path=None):\n",
    "    \"\"\"Check if a file exists at the given path or relative to a specified path.\"\"\"\n",
    "    if os.path.exists(file_path) and os.path.isfile(file_path):\n",
    "        return True\n",
    "    elif relative_path:\n",
    "        # Constructing the relative path\n",
    "        relative_file_path = os.path.join(relative_path, file_path)\n",
    "        return os.path.exists(relative_file_path) and os.path.isfile(relative_file_path)\n",
    "    return False\n",
    "\n",
    "def check_required_columns(df, required_columns):\n",
    "    \"\"\"Check if the required columns are present in the dataframe.\"\"\"\n",
    "    missing_columns = [col for col in required_columns if col not in list(df.columns)]\n",
    "    if missing_columns:\n",
    "        raise ValueError(f\"Missing required columns: {', '.join(missing_columns)}\")\n",
    "\n",
    "def extract_regional_data(gwas_meta_file, xqtl_meta_file):\n",
    "    \"\"\"\n",
    "    Extracts data from GWAS and xQTL metadata files. Checks if files exist at specified paths,\n",
    "    or relative to the metadata file's directory if not found initially. Also checks for required columns in the input files.\n",
    "\n",
    "    Args:\n",
    "    - gwas_meta_file (str): File path to the GWAS metadata file.\n",
    "    - xqtl_meta_file (str): File path to the xQTL weight metadata file.\n",
    "\n",
    "    Returns:\n",
    "    - Tuple of two dictionaries:\n",
    "        - GWAS Dictionary: Maps study IDs to a list containing chromosome number, \n",
    "          GWAS file path, and optional column mapping file path.\n",
    "        - xQTL Dictionary: Nested dictionary with region IDs as keys. Each key maps to a \n",
    "          dictionary containing 'meta_info' (list of chromosome, start, end, condition)\n",
    "          and 'files' (list of file paths).\n",
    "\n",
    "    Raises:\n",
    "    - FileNotFoundError: If any specified file path does not exist, even when considered\n",
    "      relative to the metadata file's directory.\n",
    "    - ValueError: If required columns are missing in the input files.\n",
    "    \"\"\"\n",
    "    from collections import OrderedDict\n",
    "    # Required columns for each file type\n",
    "    required_gwas_columns = ['study_id', 'chrom', 'file_path']\n",
    "    required_xqtl_columns = ['region_id', 'chrom', 'start', 'end', 'condition', 'file_path']\n",
    "\n",
    "    # Base directory of the metadata files\n",
    "    gwas_base_dir = os.path.dirname(gwas_meta_file)\n",
    "    xqtl_base_dir = os.path.dirname(xqtl_meta_file)\n",
    "    \n",
    "    # Reading the GWAS metadata file\n",
    "    gwas_df = pd.read_csv(gwas_meta_file, sep=\"\\t\")\n",
    "    check_required_columns(gwas_df, required_gwas_columns)\n",
    "    gwas_dict = OrderedDict()\n",
    "    for _, row in gwas_df.iterrows():\n",
    "        file_path = row['file_path']\n",
    "        mapping_file = row.get('column_mapping_file')\n",
    "\n",
    "        # Check if the file and optional mapping file exist\n",
    "        if not file_exists(file_path, gwas_base_dir) or (mapping_file and not file_exists(mapping_file, gwas_base_dir)):\n",
    "            raise FileNotFoundError(f\"File {file_path} not found for {row['study_id']}\")\n",
    "        \n",
    "        # Adjust paths if necessary\n",
    "        file_path = file_path if file_exists(file_path) else os.path.join(gwas_base_dir, file_path)\n",
    "        if mapping_file:\n",
    "            mapping_file = mapping_file if file_exists(mapping_file) else os.path.join(gwas_base_dir, mapping_file)\n",
    "\n",
    "        gwas_dict[row['study_id']] = [row['chrom'], file_path, mapping_file]\n",
    "\n",
    "    # Reading the xQTL weight metadata file\n",
    "    xqtl_df = pd.read_csv(xqtl_meta_file, sep=\"\\t\")\n",
    "    check_required_columns(xqtl_df, required_xqtl_columns)\n",
    "    xqtl_dict = OrderedDict()\n",
    "    for _, row in xqtl_df.iterrows():\n",
    "        file_paths = [fp.strip() for fp in row['file_path'].split(',')]  # Splitting and stripping file paths\n",
    "\n",
    "        # Check if each file path exists\n",
    "        for i, fp in enumerate(file_paths):\n",
    "            if not file_exists(fp, xqtl_base_dir):\n",
    "                raise FileNotFoundError(f\"File {fp} not found in {row['region_id']}\")\n",
    "            file_paths[i] = fp if file_exists(fp) else os.path.join(xqtl_base_dir, fp)\n",
    "\n",
    "        xqtl_dict[row['region_id']] = {\"meta_info\": [row['chrom'], row['start'], row['end'], row['region_id'], row['condition']],\n",
    "                                       \"files\": file_paths}\n",
    "\n",
    "    return gwas_dict, xqtl_dict\n",
    "\n",
    "# Example usage\n",
    "gwas_dict, xqtl_dict = extract_regional_data(gwas_meta_file, xqtl_meta_file)\n",
    "regional_data = dict([(\"GWAS\", gwas_dict), (\"xQTL\", xqtl_dict)])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[twas_scan_1]\n",
    "depends: sos_variable(\"regional_data\")\n",
    "regions = regional_data['xQTL'].keys()\n",
    "meta_info = [x[\"meta_info\"] for x in regional_data['xQTL'].values()]\n",
    "xqtl_files = [x[\"files\"] for x in regional_data['xQTL'].values()]\n",
    "input: xqtl_files, group_by = lambda x: group_by_region(x, xqtl_files), group_with = \"meta_info\"\n",
    "#output: f'{cwd:a}/{step_name[:-2]}/{condition.replace(\":\", \"_\")}.{_meta_info[3]}.twas.rds'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "R: expand = '${ }', stdout = f\"{_output:n}.stdout\", stderr = f\"{_output:n}.stderr\", container = container, entrypoint = entrypoint\n",
    "    print(c(${_input:r,}))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "bash",
     "Bash",
     "#E6EEFF",
     "shell"
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.22.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
