{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "happy-vitamin",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "source": [
    "# Hidden Factor Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "looking-species",
   "metadata": {
    "kernel": "Bash",
    "tags": []
   },
   "source": [
    "## Description"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "60b51e84",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "We provide three different procedures for hidden factor analysis from omics data in our pipeline. The first is the [Probabilistic Estimation of Expression Residuals (PEER) method](https://github.com/PMBio/peer/wiki/Tutorial), a method also used for GTEx eQTL data analysis. The second is factor anallysis using Bi-Cross validation with the APEX software package [[cf. Owen et al., Statistical Science, 2016](https://doi.org/10.1214/15-STS539)] [[cf. Quick et al., bioRxiv, 2020](https://doi.org/10.1101/2020.12.18.423490)]. The third, and the one use for our main analyses, is a PCA based approach with automatic determination of the number of factors to use. This is mainly inspired by a recent benchmark from Jessica Li's group [[cf. Zhou et al., Genome Biology, 2022](https://doi.org/10.1186/s13059-022-02761-4)]. Please note that additional considerations should be taken for single-cell eQTL analysis as investigated by [[cf. Xue et al., Genome Biology, 2023](https://doi.org/10.1186/s13059-023-02873-5)]."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "instant-thickness",
   "metadata": {
    "kernel": "Bash",
    "tags": []
   },
   "source": [
    "## Input\n",
    "\n",
    "1. Molecular phenotype in `bed.gz` file format with indexing:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "built-array",
   "metadata": {
    "kernel": "R",
    "tags": []
   },
   "outputs": [],
   "source": [
    "readr::read_delim(\"data/example_data.bed.gz\",show_col_types= F)[1:3,1:8]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "mineral-links",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "2. A covariate file in text format with the first column `#id` that lists the name of all covariates, followed by other columns each being a sample with sample names matching that in the molecular phenotype file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "genuine-lemon",
   "metadata": {
    "kernel": "R"
   },
   "outputs": [],
   "source": [
    "readr::read_delim(\"data/example_cov.txt\",show_col_types= F)[,1:8]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "immediate-romance",
   "metadata": {
    "kernel": "R"
   },
   "source": [
    "Covariates file is optional to the PEER method but it should be supplied in order to construct a proper PEER model in the context of xQTL mapping."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aware-editor",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "3. For `APEX` we need an indexed VCF file to get the command to work, although this file is not used. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "05849177-3f3d-4650-b9aa-7380f6d46718",
   "metadata": {
    "kernel": "R"
   },
   "source": [
    "### Default Parameters: Number of Factors\n",
    "\n",
    "#### PEER\n",
    "\n",
    "By default the pipeline applies recommendations from [GTEx](https://gtexportal.org/home/documentationPage):\n",
    "\n",
    "> A set of covariates identified using the Probabilistic Estimation of Expression Residuals (PEER) method (Stegle et al., PLoS Comp. Biol., 2010 ), calculated for the normalized expression matrices (described below). For eQTL analyses, the number of PEER factors was determined as function of sample size (N): 15 factors for N<150, 30 factors for 150≤ N<250, 45 factors for 250≤ N<350, and 60 factors for N≥350, as a result of optimizing for the number of eGenes discovered. For sQTL analyses, 15 PEER factors were computed for each tissue.\n",
    "\n",
    "Although one can specify fixed number of factors using `--N` option in this pipeline.\n",
    "\n",
    "#### BiCV factor analysis with APEX\n",
    "\n",
    "We will do the same as PEER\n",
    "\n",
    "#### PCA with Marchenko's approach\n",
    "\n",
    "We will use Marchenko's approach to determine the number of PC to use. FIXME: citation."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "promotional-joyce",
   "metadata": {
    "kernel": "Markdown",
    "tags": []
   },
   "source": [
    "## Output\n",
    "\n",
    "### Primary Output\n",
    "\n",
    "* `*.cov.gz`: hidden factors inferred.\n",
    "\n",
    "\n",
    "### Secondary Outputs\n",
    "\n",
    "PEER:\n",
    "\n",
    "* Diagnosis plots for check of convergence\n",
    "* peer precision /residuals /weights"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "unable-psychiatry",
   "metadata": {
    "kernel": "Markdown",
    "tags": []
   },
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "accepting-springfield",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Minimal Working Example Steps\n",
    "An MWE is uploaded to [google drive](https://drive.google.com/drive/folders/1yjTwoO0DYGi-J9ouMsh9fHKfDmsXJ_4I?usp=sharing).\n",
    "The singularity image (sif) for running this MWE is uploaded to [google drive](https://drive.google.com/drive/folders/1mLOS3AVQM8yTaWtCbO8Q3xla98Nr5bZQ)\n",
    "\n",
    "FIXME: need updated links\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "829d44cc-0630-401d-9ead-adc1107ff283",
   "metadata": {
    "kernel": "R"
   },
   "source": [
    "### a. PEER Method"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "939b358e-5dc3-4ca0-9616-0d0567646b03",
   "metadata": {},
   "source": [
    "Because the MWE only contains 10 genes but 400+ samples. The computed N will be far greater than the number of genes. Therefore in the MWE the N is fixed to be 3."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ce63ce5-1ec6-4fec-8c81-e1ddbbf03dfe",
   "metadata": {},
   "source": [
    "Timing: X min"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "important-boating",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "FIXME: need updated examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "deadly-audit",
   "metadata": {
    "kernel": "Bash",
    "tags": []
   },
   "outputs": [],
   "source": [
    "sos run pipeline/PEER_factor.ipynb PEER \\\n",
    "   --cwd output \\\n",
    "   --phenoFile ALL.log2cpm.bed.chr12.mol_phe.bed.gz  \\\n",
    "   --container containers/PEER.sif  \\\n",
    "   --N 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "harmful-polish",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[01;34m./output\u001b[00m\n",
      "├── MWE.Cov_PEER.PEER.cov\n",
      "├── MWE.Cov_PEER.PEER.cov.stderr\n",
      "├── MWE.Cov_PEER.PEER.cov.stdout\n",
      "├── MWE.Cov_PEER.PEER_diagnosis.pdf\n",
      "├── MWE.Cov_PEER.PEER.peer_precisions.txt\n",
      "├── MWE.Cov_PEER.PEER.peer_residuals.txt\n",
      "└── MWE.Cov_PEER.PEER.peer_weights.txt\n",
      "\n",
      "0 directories, 7 files\n"
     ]
    }
   ],
   "source": [
    "tree ./output"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "supposed-hometown",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div class=\"sos_hint\">> ./peer_output/example_data.bed_diagnosis_plot.jpeg (22.2 KiB):</div>"
      ],
      "text/plain": [
       "\n",
       "> ./peer_output/example_data.bed_diagnosis_plot.jpeg (22.2 KiB):"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "image/jpeg": "/9j/4AAQSkZJRgABAQAAAQABAAD/2wBDAAgGBgcGBQgHBwcJCQgKDBQNDAsLDBkSEw8UHRofHh0aHBwgJC4nICIsIxwcKDcpLDAxNDQ0Hyc5PTgyPC4zNDL/2wBDAQkJCQwLDBgNDRgyIRwhMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjIyMjL/wAARCAHgAeADASIAAhEBAxEB/8QAHwAAAQUBAQEBAQEAAAAAAAAAAAECAwQFBgcICQoL/8QAtRAAAgEDAwIEAwUFBAQAAAF9AQIDAAQRBRIhMUEGE1FhByJxFDKBkaEII0KxwRVS0fAkM2JyggkKFhcYGRolJicoKSo0NTY3ODk6Q0RFRkdISUpTVFVWV1hZWmNkZWZnaGlqc3R1dnd4eXqDhIWGh4iJipKTlJWWl5iZmqKjpKWmp6ipqrKztLW2t7i5usLDxMXGx8jJytLT1NXW19jZ2uHi4+Tl5ufo6erx8vP09fb3+Pn6/8QAHwEAAwEBAQEBAQEBAQAAAAAAAAECAwQFBgcICQoL/8QAtREAAgECBAQDBAcFBAQAAQJ3AAECAxEEBSExBhJBUQdhcRMiMoEIFEKRobHBCSMzUvAVYnLRChYkNOEl8RcYGRomJygpKjU2Nzg5OkNERUZHSElKU1RVVldYWVpjZGVmZ2hpanN0dXZ3eHl6goOEhYaHiImKkpOUlZaXmJmaoqOkpaanqKmqsrO0tba3uLm6wsPExcbHyMnK0tPU1dbX2Nna4uPk5ebn6Onq8vP09fb3+Pn6/9oADAMBAAIRAxEAPwD3+iiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiqV9JqaOn2C1s5lx8xnuGjIPsAjZrkbW98U634r1vTV1O10pNNW3wkMAuQ5kVmJ3OFPYdqAO7ormf7F8U/9Dcv/AILI/wD4qo4tL8SzqzReMo3CsyErpsZwwOCPvdQRRbqB1VFcz/Yvin/obl/8Fkf/AMVR/Yvin/obl/8ABZH/APFUAdNRXM/2L4p/6G5f/BZH/wDFUf2L4p/6G5f/AAWR/wDxVAHTUVzP9i+Kf+huX/wWR/8AxVH9i+Kf+huX/wAFkf8A8VQB01Fcz/Yvin/obl/8Fkf/AMVR/Yvin/obl/8ABZH/APFUAdNRXM/2L4p/6G5f/BZH/wDFUf2L4p/6G5f/AAWR/wDxVAHTUVzP9i+Kf+huX/wWR/8AxVH9i+Kf+huX/wAFkf8A8VQB01Fcz/Yvin/obl/8Fkf/AMVR/Yvin/obl/8ABZH/APFUAdNRXM/2L4p/6G5f/BZH/wDFUf2L4p/6G5f/AAWR/wDxVAHTUVzDaP4oUEt4wQAdzpkf/wAVVGc6pbZ+0fEKwix/z0s4V/m9JtLcqMJT0irna0V51JrjxNhviVYE/wCzYRt/JqYPEJJwPiRaf+CxP8an2sP5kdCwOKeqpy+5/wCR6RRXnB8QOP8Amo9r/wCCtf8AGkHiBz0+Ituf+4Uv+NL2tP8AmX3j+oYr/n1L7n/kekUV5ufEEg/5qJB/4KRSjX5T0+IcJ/7hAo9rT/mX3h9QxX/PqX3P/I9Horzc+IJR/wA1Bj/8E1H/AAkMv/RQY/8AwTUe1p/zL7w+oYv/AJ9S/wDAX/kekUV5v/wkMv8A0UGP/wAE1H/CQy/9FBj/APBNR7Wn/MvvD6hi/wDn1L/wF/5HpFFeb/8ACQy/9FBj/wDBNR/wkMv/AEUGP/wTUe1p/wAy+8PqGL/59S/8Bf8AkekUV5v/AMJDL/0UGP8A8E1H/CQy/wDRQY//AATUe1p/zL7w+oYv/n1L/wABf+R6RRXm/wDwkMv/AEUGP/wTUf8ACQy/9FBj/wDBNR7Wn/MvvD6hi/8An1L/AMBf+R6RRXm//CQy/wDRQY//AATUf8JDL/0UGP8A8E1Htaf8y+8PqGL/AOfUv/AX/kekUV5v/wAJDL/0UGP/AME1H/CQy/8ARQY//BNR7Wn/ADL7w+oYv/n1L/wF/wCR6RRXm/8AwkMv/RQY/wDwTUf8JDL/ANFBj/8ABNR7Wn/MvvD6hi/+fUv/AAF/5HpFFeb/APCQy/8ARQY//BNR/wAJDL/0UGP/AME1Htaf8y+8PqGL/wCfUv8AwF/5HpFFeb/8JDL/ANFBj/8ABNR/wkMv/RQY/wDwTUe1p/zL7w+oYv8A59S/8Bf+R6RRXEaW2sa0JTp/jhJxFgP/AMSlVxnOOpHoa0f7F8U/9Dcv/gsj/wDiqtNNXRzzpzpy5Zqz7M6aiuP8N+J7uawuotQgvb65tL65tWuLe0wsgjlZQcA4BwBXWQS+fAkux03gHbIuGHsR2NMg4/xXLOmuxLqE+uW+ifZlKPpEUjEz7m3CQxKZANuzbjAOWz2q14Oku3m1ACTVZdIHl/Y5NVjZJi/zeYBvAcp9zBYZyW7YqPxBbeJL7X9mlX1xZwW8VtJGURDHMXmZZw5ZTkrEAQARyauaBb39hreq6fNd315YRxwS28958zB23+YgfA3AbUPtvIoA6KiuX8Zy6zp+nS6np2ri0hiEavG1osqqGkCtKx67VVtxA7Iaq+G9WurrxIbVfFFnr1n9jaR2tIUAgkDoF3MjEfMC2Bx909aAOyooooAKKKKACuP8Pf8AJSfGf+7Y/wDopq6O+i1OR0NheWcCgfMJ7VpST7ESLj9a5Xwml0nxB8ZC8mhmmxY5eGIxqf3bfwlmP60AdvWO3/Er18P0tdSIVvRLhV4P/AkGPqi92rYqrqNimo2Etq7FN4BWResbg5Vh7ggEe4rSlJJ2ls9H/XluJlqiqOk3z31iGmUJdRMYriMdFkXg49j1HqCD3q9Uyi4ycWMKKKKkAooooAKKKKACiiqmpapZaRam5v7lIIhwCx5Y+gHUn2FJtJXZUYSnJRirtlusbVPElrp9wLG3jkvtSYZWzt+WHux6Ivuf1qj5mueJOIhLouln/lowH2qYew6Rj3OW+lbOl6PYaNbmGxt1iDHLt1eQ+rMeWP1qOaUvh0Xc6vZUqGtV80v5V+r/AEWvmmZYbxlMAfL0O2B52s0spX642il/s7xRLzJ4gtIfaDT8/qzn+VdDRR7NdW/vF9bkvhjFf9up/nc57/hHdSm/4+vFOqN/1wSGL+SUn/CHWsh/0nVNZuR6SahIB+SkV0VFHso9Q+u118Lt6JL8kjnh4G8N5BfTFmPrPK8v/oTGrsHhvQrbHk6Np8ZHdbZM/nitSimqcFskRLGYiatKo382RxQRQLtiiSMeiKBUlFFWYNt6sKKKKBBRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAHLeBP+QfrH/Ycv/wD0e9dTXC+D9f0awttZt7zVrC2mGt35Mc1yiMAZ37E5rtoJ4rmBJ4JUlhkAZJI2DKw9QR1FAGVq3iSDS71LGOyvr+9aPzTBZQ72RCSAzEkKoJBAyecHHQ0zQtUt9Vvr+RDqMFygjE1jepsMAw21lHTDc8gkHb7Gl1Xw6b/UF1Gz1W+0u+8oQvNa+WwkQElVZZFZTgsxBxkbjzUujaEmky3NxJfXd/e3IQTXV0y7mVc7VARVVVG5jgAfeNAEfivVX0Tw3c36Tw2/ltErTzruSJWkVGcjI3YDEgZ5IA71l+FNSs7rUJIbXxfY6v8Aui5tbaKJCvK/OdnPfH41u67dJZaLcXEmqwaUqbc3k4UpHlgOdxA5zjn1rnPDWtz3/iRreHxPaa9Y/ZHkdrWBAIJA6BQzoSMsGbA4+6aAO0ooooA5g+JNXvJ7k6N4d+2WlvPJbmea9WHzHjYq+xcEkBgRk4zj05rY0bVY9Z0uO+jikh3M8bxS43RujlHU4JHDKwyDjiuHvLAavql/d6N4dv3t/tMkVxLDrstilzKjFJCI0bB+ZSMnaSR+Ndh4YmsZ/DtqdOs2sraPfD9mYDMTo7I6nBIJDq2Tk565OaANeuP8Pf8AJSfGf+7Y/wDopq6O+sbi7dGh1W8sgowVgWIhvc70b9MVyvhOCS2+IPjKOW7munAsf3swQMf3bf3FUfpQB29FFFAGPc/8SzXIrwcW18Vgn9Fl6Rv+P3D6ny/StioL2ziv7Ka0nB8uVCp2nBHuD2I6g9jVXRryW5s2iuiPttq5guMDALgD5gOwYFWHs1bS9+Cl1Wj9On+X3C6mjRRRWIwooooAKRmCqWYgADJJ7Vkar4jtNNnWzjWS81GQZjs7cbpD7nsq+5xVFdCv9cYS+I5lFvnK6ZbMREP+ujdZD7cL7Goc9bR1Z1Qw3uqdV8sfxfouvrovMWXxHcapM9p4at0u2U7ZL6XItoj7EcyH2X86tab4agtroahfzvqOp/8APzOBiP2jTog+nPvWxFFHBEkUMaRxoMKiDAUegAp9JQu7y1HLE8sXCiuVfi/V/orL1CiiitDkCiiigAooooAKKKKACiiigAooooAKKKKACiiigAoorm9ObX9TtTdrqtpCjSyqsf2EttCyMo58wZ4FAHSUVjfYfEP/AEG7T/wXn/45R9h8Q/8AQbtP/Bef/jlAGzRWN9h8Q/8AQbtP/Bef/jlH2HxD/wBBu0/8F5/+OUAbNFY32HxD/wBBu0/8F5/+OUfYfEP/AEG7T/wXn/45QBs0VjfYfEP/AEG7T/wXn/45R9h8Q/8AQbtP/Bef/jlAGzRWN9h8Q/8AQbtP/Bef/jlH2HxD/wBBu0/8F5/+OUAbNFY32HxD/wBBu0/8F5/+OUfYfEP/AEG7T/wXn/45QBs0VjfYfEP/AEG7T/wXn/45R9h8Q/8AQbtP/Bef/jlAGzRWN9h8Q/8AQbtP/Bef/jlH2HxD/wBBu0/8F5/+OUAbNFY32HxD/wBBu0/8F5/+OUfYfEP/AEG7T/wXn/45QBn+BP8AkH6x/wBhy/8A/R711Nct4Fhkt9P1WGWUSyLq92WcJtDFpCxOMnHJPeupoAKKKKAMjxPLfw+Hrp9M877SCnMCB5Fj3qJGRTkFwm4gYOSBwaxPC1zJLrskdhe6ze6T9mZppNUhdNk+5doQuiscrv3DkDC9MmrXjPWp7HTpLOyg1ZryQRsJLGwlmIjMgEgV1UqH2B8Z6EqTVDwvcP8A8JM1vZDxH/Zps2eX+2YZ8edvQL5byjIOC+Vzj7uBwaAO3ooooA84vLAavql/d6N4dv3t/tMkVxLDrstilzKjFJCI0bB+ZSMnaSR+Ndh4YmsZ/DtqdOs2sraPfD9mYDMTo7I6nBIJDq2Tk565Oa4+8sBq+qX93o3h2/e3+0yRXEsOuy2KXMqMUkIjRsH5lIydpJH412Hhiaxn8O2p06zayto98P2ZgMxOjsjqcEgkOrZOTnrk5oA164/w9/yUnxn/ALtj/wCimro77S7fUHRppLxCgwPIvJYR+IRhn8a5XwnaR2XxB8ZQxNMyAWPM0zyt/q2/ick/rQB29FFFABWPqH/Et1aDUxxBPttbv2yf3b/gxKn2fJ+7WxUVzbRXlrLbXCB4ZkKOp/iUjBFaU5qMtduv9f1qJktFZmi3MrW8llduXvLJvKkc9ZBjKSf8CXBPvuHatOpnBwk4sYVzV1e3uvapNpmk3TWtnbHbeX0agsX/AOeUZPAIHVuccd6k13Ubm5vF0DSJNt9Mu64uByLSL+9/vHoo/HtWvpunW2k6fDY2ceyGIYA6k+pJ7knkmsX77stv60O2CWHgqkleT2XZd3+i+fa8WlaLYaNC0dlAELndJKxLSSH1ZjyT9av0UVaSSsjknOVSTlN3bCiiimSFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAVj+GP+QEn/AF3n/wDRz1sVj+GP+QEn/Xef/wBHPQBsUVVu7CG8KmV7ldvTybmSL89jDP41X/sO0/57ah/4MZ//AIuk79DSKp295u/p/wAE0qKzf7DtP+e2of8Agxn/APi6vQQrbwrEhkKr0MkjO34sxJP4mhX6ikoJe62/lb9WVdR1WDTJbCOZJGN7dC1j2AHDFWbJyemEPr2p9/qmn6VAs2o31tZxM20SXEyxqT6ZYjmsfxdbX0w0W5sLCa+ay1JLiWGF41coI5FJHmMq9WHeqV4NUuda07Xv+Ecu5Fgt7i1awlmt/NQuYyJB+8KEEIVPzZweh5FMg6SXVtNgjlkm1C1jSII0jPMoCBzhCSTxuPT17Us+qafaxXMtxf2sMdqQs7yTKoiJAIDEn5SQwPPqPWvOW8AanNFo2mTxxnTrix8jVdkgxF5Rd4EUHBYBpSAQOiDpxU8XhnxHL4f02/vElTWE1R9Qvbe2khZ2zG0SbDIGjLKvln5uODyDg0AdzL4g0WC1huptXsI7edS8Ur3KBZFBAJUk4IBZRkeo9ai1vXYNCbTjc+WsF3dG3eaSUIsIEUkm8k8Y/d47dc54rnNE8NXFv4g0vUJLO8EcYvpZDfPAZI5ZTDjAh+Ubgsh+XPU5PNanjCwvL06FNZ6cL/7DqaXUsO9FOxYpRkbiAWDMuB646DJABsHWdLXSxqbalZjTyMi6M6+VjOM7846+9SQ6jZXPkeReW8v2iNpYfLlVvNRcAsuDyBuXJHqPWuJTQtWjMGrHSyxXW5NROlCWPeqNCYhg7vL8wN+8xuxljznml0zRdX0rU7TWP7KeQNJflrCCaPdbrcPE68swU/6olsE8vxnrQB10+vaPawRz3GrWEMMoZo5JLlFVwpAYgk4IBIB9CRT7rWNMsbWG6u9Rs7e3mIEU0s6ojkjI2knBz7Vynh7w9qVtP4Wmv7JEews71J/3it5UkjxlQDnnKh+R+OM1FpehappI0O5m0prwWlpd2j2qSR7ovMlVkYbmCkbU2nByARx1oA6zQdV/tvRbfURD5Pnbvk37sYYjrgelaNYHgrTbrR/B2m6feW629xAhV4kYMqfMSACOowRW/QAUUUUAcb4d0HR9QfWp73SbG5mOrXIMk1sjsQG45IzXXQwRW0CQQRJFFGAqRxqFVR6ADoKwfCP+q1n/ALC9z/6FXRUAZt/r+m6Z9sF1cbGs4I7iZQjMQjsypjA+YsyMABk5HTkU3RtftNa+0JDFdW9xblRNb3ULRSJuGVOD2ODgj0Poax/EOj6Fc62t5qOuPYztHBvgF1HGsqwytJGWVhk4dm6Vd0q60i98U6rNYXours2lqJ/LdXjRA02wAjvnzCc+1AF3X9Zh8P6PJqdzsFvDJEJWd9oRGkVWboc4DE46nGByaj0vxRoWt3DW+l6vZXkyp5hjgmVmC5AyQO2SPzo8TPfx+H7l9N877QCmTAoaUR718woDwXCbio55xXO+C9Qvr68jRX1OSxtkvEllv4ZELsbnNvgyAFiIg2fTKg88UAdxRRRQB5letpeo+IruCw0HWjcXDzlXtNVe0gu3hdY5iwSQAYYgZK5Y12vhaexuPDdo2n2TWVum+L7MwGYXRyjqcE5IdWBOTk8066tf7KtVk0mwjkne8Qsp5ws06mdwc8cFnx0yvSrthYW2mWgtbSMpEHd8FixLOxdiSeeWYn8aALNcf4e/5KT4z/3bH/0U1dHfaPpmpuj3+nWd2yDCmeBZCo9sjiuV8J2VrYfEHxlBZ20NtCBYkRwxhFH7tuw4oA7eiiigAooooAx9W/4l17BrC8RoBBef9cieHP8AuMc+ys9Lr2stpkMVvaRC41O7by7WDPU92b0VRyT/AI1Preo2emaTNPer5kTDyxCBuMzNwEA7k9MVgeDNPnt57qTV9zayiRx/vH3+XbkZjVT35DBj3ZWPpVzbqQtHdb+n/D6fNdjpoQjBe2qr3ei7vt6dX927NvQtGXR7NleU3F5O3m3Vyw+aWQ9T7AdAOwrUoorNJJWRjUqSqSc5vVhRRRTICiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACsfwx/yAk/67z/8Ao562Kx/DH/ICT/rvP/6OegDSnu7a2IE9xFEW6eY4XP51F/aunf8AP/a/9/l/xqWe0trkgz28UpXp5iBsfnUX9lad/wA+Fr/35X/CtI+zt717i1D+1dO/5/7X/v8AL/jViOWOaMSROroejKcg/jVf+ytO/wCfC1/78r/hViOKOGMRxIqIOiqMAfhSlyW924amR4i1m60ePT1srKG7ub68W1RJrgwopKO24sEc/wAHp3qjY+N7GWzdtQja2u47qW0a2tg90XePBYpsXcygMMnaMZwcVa8UeHk8RrpUE0NrPa21+tzcQ3K7lkQI642kEE5YHB44rmn8A31jqwn0vyDp9v5qWdmt/PZlI5Sjuu+EZAV0JAwwIfHG0GoGdV/wlejNNbxQ3Mtw1xGkqG2tpZlCOSFZ2RSEBweWI6H0qBfFWn20Ba9vFkke6uII0tbWV2Iicqw2AMx24AZsbc88AisSTwhqUJsxpdtZ6dLGqBryHUrgugEhdlKFcTj5mxvI+8eBVmz8M6tpWprqdobKefzr8NFLK6L5U84lUhghIYbQCMYOevAyAaXhrxNFrOmaZJPJELy/hmniSJW2tHG4UsDyP406nnPHtPL4p0mHT7e+824ktpwxjaCzmlOFOGJCKSoHqQBWHpfhjWdHtNAkhawuL3Tra5tpkaR443WV0fcpCscgxrwR3PPFRweFddt9G0exN3FItrDKtzFBfT2ivIzBlkDxjcQBuGw4HzdeKAN+x1z7f4jls7doZbH+zre9hmTJL+Y8oznOCuEUjjuasSa/pkNveTyXO1LO4W2n/dtlJW27VxjJJ3pjGc7hWJ4R8L32gPbG7mt5PK0i2sGMTMcvG8pJ5H3cOMd+tOv/AAtd3XjO31GOeFdKZori7gJO954Q4jI4xj5kJyf+WSe9AGtH4j0yTVhpYlmS6ZmRBJbSIkjKCWCOyhXIAJwpPQ+lU7bxbp2qjTpdMvA1vdXCxrJLaTKswMcjbY2IA3fIcnkDBBwSKyP+EV1yXxDYX91dpKtrqUl0ZWv5zviKyKiLBjy0Kh1GRknB55NW7PwvfW3h/wAJ2HnW4m0eRHnZSSrbYJIzs45+ZweccZoA0U8X6I63bNczRC1t3upPOtZYiYU+86BlG9R6rnqPUVNZ+JtIvluGhuyq28InkaeJ4R5RziQFwNycH5hkcda4yTwLr93FJ9qu4XuG0e8sJJZdQnm86aZVAk2su2Jcrkqg4yOuBW/rPhObWJbqNp44oLjRn08suSyuWyGxjBUfWgDSh8U6RNp9xf8AnzRWsAVnlntZYQQ33Su9RvyeBtzk49ataZrFlq8cr2cjkxPsljlheKSNsA4ZHAYcEHkcg1h6po+t+ItAlstSWxtLmOaCeB7S6lZXeNw/zHajICVA+UkjOc5FXPDWjzaY15Pc2scFxcFAxXUZ70sFBxl5QCOp4AoAy/Duki7fWpft19DnVrkbIZyq/e9K66CLyIEi3u+wAbpGyx9ye5rB8I/6rWf+wvc/+hV0VAGVc+HNKvdYbVLuziuLhrdbfEyK6hVZmGARwcs351HpmmabY67qMljpz2krwwRyskIjhkCmQrtxwWG9s/VaxPFcs6a7EuoT65b6J9mUo+kRSMTPubcJDEpkA27NuMA5bParXg6S7ebUAJNVl0geX9jk1WNkmL/N5gG8Byn3MFhnJbtigDU8TJcSeHbtbSwnvp8KUt7e8Nq7kMDxKCCuOvXnGO9cj4C1e5utfu7C78RyzzRW5c6PPCXe3+ZRv88qC45xjnrXSeNnRPCN6ZLeCeMmMOtxGZIkUyKDI6j7yoMuR3C1j+Dr2GDWG0jTtU0vVdPFqZ2l063jiW2cMoVG8s7TuDMQOCNh65FAHcUUUUAFFFFAFK+1jTNMdEv9Rs7RnGVE86xlh7ZPNcr4TvbW/wDiD4yns7mG5hIsQJIZA6n923ccV29cf4e/5KT4z/3bH/0U1AHYUUUUAFRXFxDaW0lxcSLHDEpd3Y4CgdTUtcof+Kx1Lb18P2cnPpezKf1jU/gT6gVEpW0W50UKKqNym7RW7/Reb6f5Jkmk282v6iniC/jZLaMH+zbVxyin/lqw/vMOnoPc1oa0DZPBrMYP+iZW4A/igbG//vnAf/gJA61r0hAIIIyD1BrSi/Zu+/fz7k16zqyvayWiXZf1v3eoAhgCCCDyCKWsjRSbKSfRpCf9Fw1sT/FA2do/4CQU+iqT1rXqqkOSVjFBRRRUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFcto2tW+m6ebS6tdUWaOebcE0u5ccysQQyxkEYI5BrqaKAMb/hKNP8A+ffV/wDwT3f/AMbo/wCEo0//AJ99X/8ABPd//G62aKAMb/hKNP8A+ffV/wDwT3f/AMbo/wCEo0//AJ99X/8ABPd//G62aKAMb/hKNP8A+ffV/wDwT3f/AMbo/wCEo0//AJ99X/8ABPd//G62aKAMb/hKNP8A+ffV/wDwT3f/AMbo/wCEo0//AJ99X/8ABPd//G62aKAMb/hKNP8A+ffV/wDwT3f/AMbo/wCEo0//AJ99X/8ABPd//G62aKAMb/hKNP8A+ffV/wDwT3f/AMbo/wCEo0//AJ99X/8ABPd//G62aKAMb/hKNP8A+ffV/wDwT3f/AMbo/wCEo0//AJ99X/8ABPd//G62aKAMb/hKNP8A+ffV/wDwT3f/AMbo/wCEo0//AJ99X/8ABPd//G62aKAMb/hKNP8A+ffV/wDwT3f/AMbo/wCEo0//AJ99X/8ABPd//G62aKAOI8JnUdQg1a607UIYLaTVbnak9g5cEPg5y6kcg8EDFdnAsqwIs8iySgAO6ptDH1AycfnXNeBP+QfrH/Ycv/8A0e9dTQAUUUUAZmr63Fo/k+ZZajc+bux9jtHn24x97aDjrxnrg+lY2l6tLqXi21Sy07U7LTo7G5acXNk9vG0pkg8v7wALYEv611lFABRRRQB5xeTn+1L8a/eeLLe6FzILWPTIJzAYdx8vYYUKsSu0nec7ie2K7Dww+ov4dtW1XzftZ38zqqyFN7eWXC8Bym0sB3zXH3k5/tS/Gv3niy3uhcyC1j0yCcwGHcfL2GFCrErtJ3nO4ntiuw8MPqL+HbVtV837Wd/M6qshTe3llwvAcptLAd80AWr7VbbT3RZ1uiXGR5NrLKPxKKcfjXK+E7uK9+IPjKaEShCtjjzYnjb/AFbfwsAf0rt64/w9/wAlJ8Z/7tj/AOimoA7CiisLX9WuIpYdI0ra2q3YJUkZW3j6GVvYdh3NTKSirs1o0pVZ8kf+G835Iravcza7qL+H9PlaOFADqV0h5jQ/8slP99h19BXQ21tDZ2sVtbRLFDEoREUYCgdqraRpNvo2nJaW+5sEtJI5y8rn7zse5Jq9SjF/E9zSvVi0qVP4V+L7v9Oy+bZRRRVnMZWtxSRJDqtuheexJZkUZMkJx5iD1OAGA7si1pRSxzxJLE6vG6hlZTkMDyCKfWPpX/EuvZ9HbiJQZ7P/AK5E/Mg/3GOPZWStl79O3WP5f8B/mxbM2KKKKxGFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAHC+D5tZS21lbOwsJof7bv8PNevGxPnv/AAiJh+tdtAZWgQzoiTEDesbl1B9ASBke+BXNeBP+QfrH/Ycv/wD0e9dTQAUVxfiuWdNdiXUJ9ct9E+zKUfSIpGJn3NuEhiUyAbdm3GActntVrwdJdvNqAEmqy6QPL+xyarGyTF/m8wDeA5T7mCwzkt2xQB1VFFFABRRRQB5xeTn+1L8a/eeLLe6FzILWPTIJzAYdx8vYYUKsSu0nec7ie2K7Dww+ov4dtW1XzftZ38zqqyFN7eWXC8Bym0sB3zXH3k5/tS/Gv3niy3uhcyC1j0yCcwGHcfL2GFCrErtJ3nO4ntiuw8MPqL+HbVtV837Wd/M6qshTe3llwvAcptLAd80AWr6/ls3RY9NvLsMMloPLwv13MK4zStUfTfHPia91DSdZt4b1bTyCunSzhtiMG5hV1GCR3716BRQBg/8ACXab/wA+2tf+CS8/+NVl6Vr2gWk97eQJrtzPdTsZZ20W8Y/KSvlgiHAC4K47EHvXZVm6HazWenyxTpsdry6lAyD8rzyOp49VYH8aTSbuXGcopxT0e5S/4THTP+fXXP8AwRXv/wAao/4THTP+fXXP/BFe/wDxqt+imQYH/CY6Z/z665/4Ir3/AONUf8Jjpn/Prrn/AIIr3/41W/RQBgf8Jjpn/Prrn/givf8A41VHUPEvhy9eCO9stYZ9x8kNol6rE4yQpEQPQcgeldbWbqFrNPqukzRpujgmkaQ5HygxOo/UinGTi7xdmBzv27w1/DZ+JV/3NO1Nf5JR/aHh7tH4uH0tdWH/ALLXaUVp7er/ADP72KyOL/tDw/2XxeP+3XVT/NaP7T0Yfdl8Xr/3D9RP84zXaUUe3q/zMLI4v+1tOH3b3xev/cIvD/OA0f2zbD7uq+L1/wC4DMf52prtKKPbz/pILHEQ+IbSeJZYdb8WSRsMq6aBIQR7EWlPOtwkf8hjxeP+5ek/+RK6Lw/azWWgWVtcJsmjjwy5BwfwrSo9tLsvuX+QWOKGtQjrrHi8/Xw9J/8AIlL/AG3B/wBBfxd/4T0n/wAiV2lFHtpdl9y/yCxxR1uEj/kMeLx/3L0n/wAiUDWoR11jxefr4ek/+RK7Wij20uy+5f5BY4g+IbRZlhOt+LBK6l1Q6BJuKggEgfZOg3Ln6j1p/wDbcH/QX8Xf+E9J/wDIldFPazP4msbtUzBFZ3ETvkcMzwlRjryEb8q0qPbS7L7l/kFji/7bg/6C/i7/AMJ6T/5Eo/tuD/oL+Lv/AAnpP/kSu0oo9tLsvuX+QWOL/tuD/oL+Lv8AwnpP/kSj+24P+gv4u/8ACek/+RK7Sij20uy+5f5BY4o65AASdY8XADqT4fk/+RKSLxBazRJLFrXiySN1DI66BIQwPQg/ZORXZzKXgkVRklSB+VU9Dt5bTQNNtp02TQ2sUbrkHDBQCOPej20uy+5f5BY5r+24P+gv4u/8J6T/AORKP7bg/wCgv4u/8J6T/wCRK7Sij20uy+5f5BY4v+24P+gv4u/8J6T/AORKP7bg/wCgv4u/8J6T/wCRK7Sij20uy+5f5BY4v+24P+gv4u/8J6T/AORKZ/wkNp5xh/tvxZ5oXfs/sCTdtzjOPsnSu3rNFrMPEz3ez9wbNYg+R97eTjHXoaPbS7L7l/kFjnf7bg/6C/i7/wAJ6T/5Eo/tuD/oL+Lv/Cek/wDkSu0oo9tLsvuX+QWOL/tuD/oL+Lv/AAnpP/kSj+24P+gv4u/8J6T/AORK7Sij20uy+5f5BY4v+24P+gv4u/8ACek/+RKZN4htLeGSabW/FkcUal3d9AkCqoGSSTacAV29ZviG1mvvDOq2lsm+eezmijTIG5mQgDJ46mj20uy+5f5BY53+24P+gv4u/wDCek/+RKP7bg/6C/i7/wAJ6T/5ErtKKPbS7L7l/kFji/7bg/6C/i7/AMJ6T/5Eo/tuD/oL+Lv/AAnpP/kSu0oo9tLsvuX+QWOL/tuD/oL+Lv8AwnpP/kSj+24P+gv4u/8ACek/+RK7Sij20uy+5f5BY5Sx8WabBJLBJc+IbyZQrlZtDuAyA5xwluvBwevoau/8Jdpv/PtrX/gkvP8A41Vy3tZk8R390yYgltrdEbI5ZWlLDHtuX860qzbu7sZw/hC/urOw1HfoWrFZ9VvLiJmtxGWR5mZTtkKsMgjqBXaQyGaBJDG8RYAlJMbl9jgkZqSikBxfiuWdNdiXUJ9ct9E+zKUfSIpGJn3NuEhiUyAbdm3GActntVrwdJdvNqAEmqy6QPL+xyarGyTF/m8wDeA5T7mCwzkt2xWjq3iSDS71LGOyvr+9aPzTBZQ72RCSAzEkKoJBAyecHHQ0zQtUt9Vvr+RDqMFygjE1jepsMAw21lHTDc8gkHb7GgDcooooAKKKKAODiudf1zVI9HuX1fT4Y7m+ee8htxFvjWYi2RZCpGDGQSRydo55Oek8Ly38ugQnUzK10kk0Rklj2PIiSsqOVwMFkVW6DrVvVNOXU7RLdpXiCXEFxuTqTFKkgH0JTB9iau0AFFFFADXdI42kkZURQSzMcAAdyar2GpWOqW5uNPvbe7gDFDJbyrIu4dRkEjNTzNEsEjTlBCFJcvjaFxznPbFcb8P9X0m+bXoNO1CyuCNUnlWO3mVsRnaAwAP3T2PSgDtaKKKACiiigAqi+taVHqa6Y+p2S37fdtWnUSnjPCZz+lXq8/8AEutaBqHiRPDT3+mWUsVzb3V9PPKkchdWV444wSCzttXLfwrx1IAAPQKKKKACiiigAqG6u7axtpLm7uIre3jGXlmcIqj3J4FTVS1a7s7HTJri/uktbZQA08mMRkkAHkEcEjrxQBJY6jY6pai50+8t7u3JIEtvKsiE/UEirNcX4CMD3fiCW2vpdThlu0cakwQJcHylBC7FVDtwASo5+ortKACiiigAooooAgub20s2gW6uoYGnkEUIlkC+Y56KuepODwOanrzDxlLqP/CS2V5daFfyxW2qWsVg8TwlCm4M5GZAQ7theQAAg5GTXp9ABRRRQAUUUUAVNR1TT9ItvtOp31rZW+4L5tzMsa5PQZYgZ4p9lfWepWcd3YXUF1bSZ2TQSB0bBwcMODyCPwrF8YJpxsLKXUdbGjLBdrLFdExgCQKwAzICvQnqO1P8H6ld6roRuLuQT7biWOG5EXli5iViEk29sj04PUcEUAb9FFFABRRRQAVSk1jTItTTTJNRs0v5BlLVp1ErD1CZyfyq7XlWkXd5pviKeE6ikmqXHiCVZtLaBCzWzO22bOPMwse0hs7cKFxmgD1WiiigAooooAKp6lq2m6NbrcapqFpYwM+xZLqZYlLYJwCxAzgE49jVys/XNUi0TQr3U5l3pbQtIEHVyBwo9ycAfWgCbT9TsNWtftWm31te25Yr5ttKsiZHUZUkZq1WP4W0qTRvDlnaXDBrsqZrpx/HO5LyH/vpm/CtigAooooAKKKQkhSQMkDp60AU7XWdLvrya0s9Ss7i5g/10MM6u8fOPmUHI/GrteSeBpg+s+GoPPtJvs9pcKLK2QrPpu7BK3LEkseNvITLc4bqPW6ACiiigDE1Xw6b/UF1Gz1W+0u+8oQvNa+WwkQElVZZFZTgsxBxkbjzUujaEmky3NxJfXd/e3IQTXV0y7mVc7VARVVVG5jgAfeNa1FABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAc3rnixdEn1aNrJphp+lf2jlWOZPmddmApx9z73OM8jirJ8V6XHYWt1MbuM3OfKgNlMZmwMsRHs3kDP3tuORVPWfDNxqupazKJoo4NQ0X+zlJyWVyZCWIxjGHHfsao6h4b1jUJdL1KSK2W+tLeS1ktodVuII3RihDCaNFYHKD5SpBz7A0AbieKtGmvbazgu2nnuYY7iJYIJJAYnYqrkqpAXKkEnAHfGRVK+8caVbaXqd1ALiaexs5LsW8ltNC0yKOqlk+Zc4BZQQM5NJoPhqbSNSkuVS3gibTorZY45nmKSLLNIx3OMsP3o5JySDwOKwB4F1u7P/EwvY2d9Ju7Cedr+e4MskyoPMCOAsY+UkqvtycUAdRH4ltZZbKVpxa201tPPIl3bSxPiPZlgXChVXdzkc5BHQ0g8aaAbK4vGvXiht/KMvnW8sbKJG2o21lDFWJwGAx154NZWq+FtU8RWtvFfmzs2XTbuxk8iZpgDL5WxhlFyP3ZyDjsOetMvvDGt63cSXuonToLkLaRRxW8ruhSO5SaRmYoDkhcBcHHqc5ABvjxRpBsbi8NxKkdvKsMqSW0qSq7Y2r5ZUOS25cADnPGav2N9DqNqtzbiYISRiaB4XBHqrgMPxFc/qPh26ubrW5hbWV5HfG2MUE87w4MYOSXVSUYHBBGTx2rR8NafqGm6UYNSufOlMrug855vKQn5U8xwGfH95hmgDYooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAorN17VRomktfsqFEmhR2dtqojyqjOT2Chix9hWRoPii51fWRC1vbjT7qK4msZY3Jdo4ZVj3N2w+8MuO1AHU0UUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAGD4qOvfYYRocVi2Jo3uGurgxYjWRCwGEYEFQ4bOMDpk8VT0K+vtS11Zbi80NY4LV4xa6beG4ZyWTDNlFKhdpAAz9/wClJ43ttWvtLktLeLR/7PPlSSS6hdvEN6yBgrKI2DIdqggkZDEcdaTw3dA6iUktvDFuWiIVtLvfMlY5HG3y1+Xqep6DigDrKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigDmfGsthc+H76wuL+wgdFhuJEvXxGUEy4WT0RypTOO5rjbLT9K8ZXl6tlF4ZsrtdOkigi0+dZ3WQshW4JCLt8sj5cc/OeRXoXiZ7+Pw/cvpvnfaAUyYFDSiPevmFAeC4TcVHPOK53wXqF9fXkaK+pyWNsl4kst/DIhdjc5t8GQAsREGz6ZUHnigDuKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigDJ8TJcSeHbtbSwnvp8KUt7e8Nq7kMDxKCCuOvXnGO9cj4C1e5utfu7C78RyzzRW5c6PPCXe3+ZRv88qC45xjnrXa6zHps2mNBq8UctlNJFE0ciF1dmkVUBHoWK+3rxU9nYWenQCCxtILWEdI4Iwij8AKALFFFFABRRRQAUUVhO+qXuuX1tbaglrDbLFtX7OHLFgSSST7UAbtFZH9n61/0HF/8A1/xo/s/Wv+g4v/AIBr/jQBr0Vkf2frX/QcX/wDX/Gj+z9a/wCg4v8A4Br/AI0Aa9FZH9n61/0HF/8AANf8aP7P1r/oOL/4Br/jQBr0Vkf2frX/AEHF/wDANf8AGj+z9a/6Di/+Aa/40Aa9FZH9n61/0HF/8A1/xo/s/Wv+g4v/AIBr/jQBr0Vkf2frX/QcX/wDX/Gj+z9a/wCg4v8A4Br/AI0Aa9FZH9n61/0HF/8AANf8aP7P1r/oOL/4Br/jQBr0Vkf2frX/AEHF/wDANf8AGj+z9a/6Di/+Aa/40Aa9FZH9n61/0HF/8A1/xo/s/Wv+g4v/AIBr/jQBr0Vkf2frX/QcX/wDX/Gj+z9a/wCg4v8A4Br/AI0Aa9FZH9n61/0HF/8AANf8a8z8T+OvGNh4n1PR9CNvejSrX7VeSzQomEChiVG4ZwGHqTzxXThcHVxU3CktlfVpLtu/UmUlHc9joryqP4lQweGdI1jU9cktm1JHKQpYCQqUba/IPTPSq/8Awt7Rf+hluv8AwVD/ABrdZTjne1KTs2tuzs/xF7SPc9doryL/AIW9ov8A0Mt1/wCCof40a/448TprmmaR4Wnh1O4vrMXqtNbrEPLILDAJHOFJ5NNZTjebllTcd3d6LTfV6B7SPc9doryTSPiY0/gyHX9X1hrMtdNaNFFYiTMgUNxz02kHn6c03/hb2i/9DLdf+Cof40f2RjuZxVJuztor7B7SPc9doryL/hb2i/8AQy3X/gqH+NLr3xA1mO30AeGb5NSuNaZhbia1WJRtbaQckc7sjrjihZRjeZRlTav30Wivu9Nkw9pHueuUV5HoPxF1CbQ9WvfEOojT59KuVtrhIbQSgsxIGMH1Vh6cdeaT/hb2i/8AQy3X/gqH+NDyjG8ziqbdu2q77ryYe0j3PXaK8i/4W9ov/Qy3X/gqH+NLrnxJu4PC1rq2g6p9vku7v7HFHLZiIBwMnOT7r+fXihZRjuZRdJq7tqrL72HtI9z1yivI9A8ea+9zrtn4ou49NutHiE03kWyyqVyBjgnn5l6E9e2KT/hb2i/9DLdf+Cof40PKcbzOMabla22q1V1qtNg9pHueu0V5F/wt7Rf+hluv/BUP8akv/iYB4Qv9c0nWnu2tZUh8mSwEfzv0yc9MAnj0o/snHXSdJq7S1VtXsHtI9z1mivH/AA7458Vz+IZNH8TSwadJ9iN7HJDAsg8sLu5AJ/hyfwxin/8AC3tF/wChluv/AAVD/GnLKMapcsabls9NVZ7aoPaR7nrtFeRf8Le0X/oZbr/wVD/GrUfxJtr3RNW1DTteluH06DznhfTwhbJ2gAk+pA/GpeU46PxUpL5d9A9pHuep0V4t4d+IHjK91/RrDWvs9pbazGZLOaGJXOOcEjceuMdjzV+b4taNDNJEfE1wxRipK6VwcfjVzyfGxnyKDk99NerXTzTD2ke561RXkX/C3tF/6GW6/wDBUP8AGr+l/Eax1l7qOy8RztJb20lyytpgXKIMtg564qJZVjoLmlSkl6MOePc9OorwrTfif4xmuNJu9Qjt4NG1S6NtBPHGrOCGCkkZ7E+gz2r17+z9a/6Di/8AgGv+NZ4vA18I0qy3vs09tHt1T3HGSlsa9FZeg3VzdafIbuVZZormeEuqbdwSRlBx9AK1K4ygooooAwPGjWw8J3i3dt9phlaKLyjcNArM8iqu6ReVUMQSR2B69Kx/DFjLoniY6dexqbmayeaKSHUrmdAiugZWjmYhTllww64YcYOdfxbd6jBp0UFhoUmrLcSxpOi+UVERkQOGV2GdyFwOoGMnAql4Y097HVG+x+FItAsGgbz8+SZJpdyeXgxs3yqvmdf7wxQB1tFFFABRRRQAVj6f/wAjLrP+7B/6Ca1nkRMb3Vc+pxWPprq/iTWSjBhiDkHP8JoA2qKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACvJNd8PeFPHvxPvtKki1GDULO2V7q6tZFEcuNo2sCDyAyjPsR2r1O/vYdN065vrhtsNtE0sh9FUEn+VebfBeymutP1jxTeL/pWsXjMD/sKTnHtuZh/wEV6+XOVCjVxcZNNJRVu7f42SbsZz1aid9ZeHtJsNJtdLisIGtLVdsUciB9vqee56k1J/Yek/wDQLsv/AAHT/Cr9FeY61RttyevmXZFD+w9J/wCgXZf+A6f4Vz/i/wCHGh+Mpbee9NxbXMCeWk1qwVin905BGOT+ddfRWlLFV6M1UpzaaBxTVmYuieE9H0HQotGtLRHtI2LkTgOXc9WbI5NW/wCw9J/6Bdl/4Dp/hV+iolXqyk5Sk7vV6hZFD+w9J/6Bdl/4Dp/hWP4r8B6L4vsLa2vUktzakm3ltSEaPOMgcEY4HGO1dPRVU8TWpTVSEmmtgcU9Gc74Z8E6L4V0eTTbKAzRTPvme5w7Sntu4xx2GK0/7D0n/oF2X/gOn+FX6KU8RWqTc5Sbb31BJLQof2HpP/QLsv8AwHT/AArP8ReDdG8TaGNJvLbyrdX8yI24CNE3PK8Y7ntW/RRDEVYSU4yaa21CyZyvhL4f6L4OgukslmuJboYnmumDs6/3eABjn0571t/2HpP/AEC7L/wHT/Cr9FOriq1WbqTk22CikrIof2HpP/QLsv8AwHT/AAqvqvhjSNY0O40e5sols5+WWFQhDDowx0IwOa16KmNerFqSk7rzCyON8IfDPQvBt5Le2bXVzdSJ5YlunDFE9AAAOwro/wCxNJ/6Bdl/4Dp/hV+irrYuvWm6lSbbYKKSsih/Yek/9Auy/wDAdP8ACmz6FpVxp11YPYW6210hjmSOMJvH4f5FaNFZ+1qb8z+8LI4Xwz8J/DvhfWl1W2e8ubiPIgFzIrLDnPQBRzyeua619G0uR2d9Ns2djks0Ckk/lV2ita+MxFefPVm29gUUtEUP7D0n/oF2X/gOn+FPj0rTolkWKwtYxIhRwkKruU9QcDkVcorF1Zvqwsjz7Svg54X0nX4tVh+2SeTJ5sNtLKGijbOQRxk44xknp3r0GiitcRi6+JalWk5NdwUVHYx/Dn/Hld/9hC6/9HPWxWP4c/48rv8A7CF1/wCjnrYrnGFFFFAGP4ptzceHbnF1b2phaO4Ety22IGKRZMOeynZgn0JrnvDGvx+J/FRuze6QJbWykiW0sL43LMGdC0jnauACgAGD95vpU/jfXdLtIv7M1V5LaNpbKaKcx7kkP2kZXsCF2KWGfuvxmovC0mnXHiW5ex121vIlS5kht4oCsiiaZZXLuThgHOFwBgMetAHb0UUUAFFFFAFO80nTdRZWvtPtbpkGFM8KuVHtkVzvhqztbHxr4phs7aG3iAtCEhjCL9xuwrrq5jRf+R78U/7tn/6LagDp6KKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAPOfjRq72Pgkabb5N1qs62yKvUrnLY/IL/wKuy8OaQmgeG9O0qPGLWBY2I/ibHzH8Tk/jXnWrf8AFW/HfTtOHz2egQ/aJR1HmcN/MxD/AICa9Zr1sb+4wlHD9X77+ei/BfiZx1k2FFFFeSaBRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAGP4c/48rv/sIXX/o562Kx/Dn/AB5Xf/YQuv8A0c9bFABRRRQBzfiHxbpWg3YtdXiZIj9neKR0yjs823jtmMhXPfBBHNXtM8TaNrNy1vp2oRXEyoXKJnIUEDPT1IqPxX9s/wCEbuTYpI86tE37qISSKokUuyKQQXVdzKMH5gKx/C1zJLrskdhe6ze6T9mZppNUhdNk+5doQuiscrv3DkDC9MmgDsqKKKAPOLyc/wBqX41+88WW90LmQWsemQTmAw7j5ewwoVYldpO853E9sV2Hhh9Rfw7atqvm/azv5nVVkKb28suF4DlNpYDvmuPvJz/al+NfvPFlvdC5kFrHpkE5gMO4+XsMKFWJXaTvOdxPbFdh4YfUX8O2rar5v2s7+Z1VZCm9vLLheA5TaWA75oA165jRf+R78U/7tn/6Latq+j1N3T7BdWcK4+YT27SEn2IdcVz3hlbtPGvikXs0Es2LTLQxGNcbG7Fm/nQB11FFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFQXt5Dp9hcXtw22G3iaWRvRVGT+gqevO/jPrD6f4HbT7fJutUmW2RV6lerY/IL/AMCrpwWHeJxEKK+0/wAOv4EyfKrlH4MWc17aa14rvF/0nV7xipPZFJJx7bmI/wCAivUqyvDOjp4f8M6dpKY/0WBUYju+MsfxYk/jWrWmY4hYjFTqR2vp6LRfgEFaKQUUUVxFBRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFAGP4c/wCPK7/7CF1/6Oetisfw5/x5Xf8A2ELr/wBHPWxQAUUUUAUdY1NNH0yS8eKSYh0jSKPG6R3cIijJAGWZRk8DNZ+j+JTqdxbW09k1tcTJdMVEgcKbecQuM4GclgR+NTeKZrOHw/N9vtluLWSSGGRGfYAHlRN5b+Hbu3ZHI254rN0fS9J8O+Jl02yinlubq0muXuLi7ed0USrlfnJKhmkZuOpVs5IoA6qiiigDjYIvFEkl/q0N7O88N7NGuk3EaJBLArkJsbbuDMmGD5IycEY6a3hG8vb/AMPi61CK5hnku7orHdR7JEj+0SeWCO2E2j6Y61uUUAFcxov/ACPfin/ds/8A0W1dPXMaL/yPfin/AHbP/wBFtQB09FFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFeTa1/xVvx10vSx89noMP2mYdQJOG/mYh+Br1O8uobGynvLhtsMEbSyN6KoyT+QrzL4M2s1/Drni27X/AEjVrxtmeyKSTj23MR/wCvWy79zRrYrsuVestPwVzOerUT1OiiivJNAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAx/Dn/Hld/wDYQuv/AEc9bFY/hz/jyu/+whdf+jnrYoAKKKKAMnxNeCx8P3MrW1vcqxSEx3LYixI6oTIcH5Buy3HQGsvwnoz6PcTrHZ+Hra3lTcw0yNldmBGNxPVQCfzq94vdU8NXCtZ2l4JJIYfKvIvMhBeVFDuvdVLbj/u9R1rkvAscdp4surUx6AlwkV1HLHp2mLbTQ+XOqrvIYna67XAPX3xmgD0miiigAooooAo3+s6dpbol7dxwM4yoc9RXPeGL611Dxp4pntJlmiItBuXp9xq6+uY0X/ke/FP+7Z/+i2oA6eiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAPPfjLrLab4EksoCTc6nKtqir1Knlv0G3/gVdX4W0ZfD3hfTdJUDNtAquR0L9WP4sSa891//AIqz45aRpA+ez0OL7VOOoD8N/Pyh+des162M/cYOjh+srzfz0j+C/EzjrJsKKKK8k0CiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigDH8Of8eV3/wBhC6/9HPWxWP4c/wCPK7/7CF1/6OetigAooooA57xbcDS9HutTl1S7tLdfIikaJ41EKmZQ0i7kbLYbpzkAAAE5qn4XvNIudTl+weMpdamMJLW7zQNtUEfPiONW4yBycfNWx4jk1CPRJTpZdbppIow8cYkaNGkVXcKcglULMAQelc54VOsReJ7q1v8AUdYvFiS5WRbyBViTEw8hkdY1DFojk4JwQenSgDuKKKKACiiigArmNF/5HvxT/u2f/otq6euY0X/ke/FP+7Z/+i2oA6eiorq6t7K1kubqaOCCJdzySMFVR6kms7RfE+ieIvN/sjU7e7MX31jb5l9yDzj3q1SnKLmouy3fQV1sa1FFFQMKKa7pFG0kjKiKCzMxwAB1JNY+k+LvD+u3klpper2t1cRjLRxvzj1HqPcVcaU5RcoxbS38vUV0bVFFFQMKKKxbTxd4dv8AVW0u01mzmvVJHkpKCSR1A7E+wq4UpzTcU3bfyFdI2qKKKgYUUViv4u8Ox6yNIfWbMahu2eQZRnd/d9M+3WrhSnUvyJu3YTaRtUUUVAwqG6uYrKznup22QwRtJI3oqjJP5Cpq4b4jaxp1xoF/4ai17TrHVryNY1S5m2gKWGQxGduVyOfWujC0HXrRprZvXyXV/IUnZXMT4N20upf294uu1xPqt2wjz2QEk49stj/gFeqVkeFtETw54X07SIyrfZoQrsvRnPLMPqxJ/GtetcxxCxGKnUj8N7L0Wi/AUFaNgoooriKCiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigDH8Of8eV3/2ELr/0c9bFY/hz/jyu/wDsIXX/AKOetigAooooAzdfiim0S4jnsbq9Q7f9HtH2SudwxtbcuCDg53DGOtY3haw8S2t/NJqNy6aU0WILO5uBczo+RyZQo4xkYLP1zu7VseINVbRtGlvY4kkkDxxosj7EDO6oCzYOFBYEnsAay9B8R3l/f29hex23ntHeNK1vkBTBciEcEk4YEkZ/umgDp6KKKAPOLzQj/al++s+D7/XriW5ke3vYbqIoIixMaAPKhi2rheBgkZyc12HhizvrDw7a22osxuU3na0plMaF2KIXPLFUKqW77c1r0UAUr6xuLt0aHVbyyCjBWBYiG9zvRv0xXO6eF8P+JdZlvJNWvGu1gKz/AGB5Q21WGMwxBRjP1rr6KAPOPibI3ibwRdadpcGqG6LpIsR0y5UShTkrkx49+e4FcD4Ehuk8WzatJoepaBaJYvZ/6HplzNvlHyFvlRuQwJOe6gc819C1m6HazWenyxTpsdry6lAyD8rzyOp49VYH8a9PD5rWoYaWGilZ3731tfrbp1TIcE5cx4+fDWsZ/wCSjeM//BFqP+NH/CM6x/0Ubxn/AOCLUf8AGvcqKr+2sb/Mv/AIf/Ih7OJ4v/ZN/b+F9esJfE/ijV57618uBLrRL4KjA56srfe+7261yngq0vovGGiXU/hrUNJi0hTHdzw6bcyPO+3PzBUJDEOv4HPoK+k6zdMtZrfUNZllTalxeLLEcg7lEESZ9vmRh+FXDO8TGnODs+ffS26tsrLbumL2cbpnkdz4e1aa6llT4heM0V3LBBoWoYUE9OMD8gPpUf8Awjuuf9FD8Yf+CLUa9yoqFnOMStdf+AQ/+RH7OJ4/o+n6jppvjd+L/FWorPZywJFNoV/tV2GA/Kt09hXnWhaRqaarolk/hjULGbTrwXF1qSWFy0jxl8rlAhOPlIHHOO3NfUtZtvazJ4jv7pkxBLbW6I2RyytKWGPbcv51pSz3FU+bRPm8rdGvs2XXqmJ0os8t1nStV1LWbu9tfHHi2zgnlLx2yaFf7YgeijAAwOnSqP8Awjuuf9FD8Yf+CLUa9yorGOcYuMVFNWX92P8AkP2cTyHw7Yalo+uW19e+M/Fmo28RJe1l0K/2yZBGDuDDvnp2rzCfQdX3T6N/wjWoS3ct8bhNXawuRJ5X3SdmzdtyQx4zk9OlfVtZr2sx8TQXYT9wtnJEXyOGLoQMdegNb0M+xVGbno27dLbXt8Nu/W4nSi9DzTxFY6jrGsPeWXjLxZp0DIii2i0K/wBqkKAT8qqOSM9O9ZX/AAjuuf8ARQ/GH/gi1GvcqKwp5viqcVCLVl/dj/kN04s8X0zR9XstUtbqfx34uuIYZVd4W0G/IkAOSpyCOenINcT4o03UX8Q+Ikj8MajqR1e582yvpNPuUeEbi5CqyAng7Tx/D6V9PVm31rNNrOlTxpmKBpTI2R8uUIH61tQz3FUqntNG7W2t1v8AZ5eonSi1YwvD/iCz0vw5pmn3MetSz21rHDI40S9O5lUAn/Ve1aP/AAl2m/8APtrX/gkvP/jVb1FeRObnJye7NEYP/CXab/z7a1/4JLz/AONUyTxppEOzzItYTewRd2i3g3MegH7rrXQ1m6vazXTaeYU3eVeJK/IGFAOT+tSBT/4S7Tf+fbWv/BJef/GqP+Eu03/n21r/AMEl5/8AGq3qKAMH/hLtN/59ta/8El5/8apkvjTSII/Mmi1iNAQNz6LeAZJwP+WXqQK6Gs3XrWa80l4LdN8hliYDIHCyKT19gaAKf/CXab/z7a1/4JLz/wCNUf8ACXab/wA+2tf+CS8/+NVvUUAYP/CXab/z7a1/4JLz/wCNU2XxnpMELzSw6xHGilndtFvAFA5JJ8rgV0FUNdt5bzw/qVtAm+aa1ljjXIGWKEAc+5oAoDxfphAIttaIP/UEvP8A41R/wl2m/wDPtrX/AIJLz/41W3EpWJFPUKAafQBg/wDCXab/AM+2tf8AgkvP/jVH/CXab/z7a1/4JLz/AONVvUUAc9D400i4gjmhi1iSKRQ6Omi3hVlIyCD5XIp//CXab/z7a1/4JLz/AONVc8PWs1j4Z0q0uU2TwWcMUiZB2sqAEZHHUVpUAYP/AAl2m/8APtrX/gkvP/jVH/CXab/z7a1/4JLz/wCNVvUUAc9F400idC8UWsSKGZCV0W8IDKSrD/VdQQQfcU//AIS7Tf8An21r/wAEl5/8aq5odrNZ6fLFOmx2vLqUDIPyvPI6nj1VgfxrSoAwf+Eu03/n21r/AMEl5/8AGqP+Eu03/n21r/wSXn/xqt6igDnk8aaRK8iRxawzRNskC6LeEo2A2D+64OCD9CKf/wAJdpv/AD7a1/4JLz/41VzTLWa31DWZZU2pcXiyxHIO5RBEmfb5kYfhWlQBg/8ACXab/wA+2tf+CS8/+NUf8Jdpv/PtrX/gkvP/AI1W9RQBhReL9KmmkhSPVfNjVWdDpF2GAOcHBj6HB/I1N/wktj/zw1X/AMFN1/8AG6lt7WZPEd/dMmIJba3RGyOWVpSwx7bl/OtKgDG8M7m0yeRopohLe3MirNE0bbWlYglWAIyOeRWzRRQAUUUUAZHiaaW30GaSO1FyvmRLPEYTLmEyKJTsHLYjLnGD06Gua8FWj6drDfZtKtrDT9SinuEhisfIaFY5VWHe3Ul0dn2tyuCBW/4wVW8MXJN59jdJIZIp/s5nKyLKjJiMEFmLBQBnkkcHoc/wvrGpX13CuoalFMJ4Z2S3/s028itDKInLHzXHDHGMfxA54xQB1tFFFABRRRQAUUUUANd0jjaSRlRFBLMxwAB3JqvYalY6pbm40+9t7uAMUMlvKsi7h1GQSM1PM0SwSNOUEIUly+NoXHOc9sVxvw/1fSb5teg07ULK4I1SeVY7eZWxGdoDAA/dPY9KAO1ooooAbJJHDE8srqkaAszscBQOpJ7Cqun6tpurxNLpuoWt7GjbWe2mWQKfQlSeadqVxZ2mmXVxqDRrZRRM85kGVCAZbI7jGa5LwVqWmeINZ1HXrO9sDLcwRQpZW06PJFAhcq8oUnDsXPH8IAHXNAHb0UUUAFULDXNJ1SaWHT9UsruWH/WJb3CSMn1AJxV8nAya8/0+80zUfiLY3WmawdWjFrcL5UIj8uwUlDyUUH5iAArknv60AegUUUUAFQSXtpDeQWkt1ClzOGMMLSAPIF5YqvU4zzjpU9eYPLqP/CydFvdR0K/S6nvLiGNy8JjW3EbKgXEmcAEyNkA5YgZwBQB6fRRRQAVmT+ItDtdTTTLjWdPh1B2VVtZLpFlYt90BCcknIxxzWnXn2u38OjeJbu40nX1k1e5urZZdEKI5lGEQ8Y8xfk+bcDtGOaAPQaKKKACqeoatpukRpJqWoWllG7bUa5mWMMfQFiMmrlea+MJZLHxtPfXOsLpkQ0cCwd4EkE0wdy8Y3g5JBjyq4ZhtweDQB6SrBlDKQQRkEd6WqmlzXFxpFlPdwfZ7mSBHlhAx5blQSv4HIq3QAVn6nr2j6J5X9ratY2HnZ8v7VcJFvxjONxGcZH5itCsLxfqE9h4cnFmQL+6ZbSz9ppSEU/8AAc7j7KaANe0vLa/tY7qzuIbm3lGY5YXDo49QRwamqppmnwaTpVpp1qu2C1hSGMf7KjA/lVugAqC8vrTTrV7q+uoLW3T70s8gRF+pPAqeuQ+IEnk2mizeZBbeVqccn226UtBakRyfPIuRlT93quCwORigDqbW7tr62jubS4iuLeQZSWFw6sPUEcGpq5P4euJNDvJAyS79RuHNzCNsNyS2fMiHZDnAGTyDyep6ygAqOaeK2geeeVIoo1LPJIwVVA6kk9BUlYHjS0ju/DEwlu7a1SGaG4Ml0cQny5VcLJ/sNtwfTNAGtY6hZanardafeW93bsSBLbyiRCR15BIqzXE+BJrrUNW8R6s32X7HeTw+SbNy8LssYV2Ryq7/AOEFtoBII5xmu2oAKqXGq6dZ3kFnc39rBdXBxDDLMqvKf9lScn8Kt1xPj7VNH0660A319Y21wNTgkzPKiP5QLZbk52gnk9KAOqvtW03THhS/1C0tGnbZEJ5ljMjei5PJ+lXK868S3ml22v61Pq7RPa6hoMcOnsw3Cc7pjJHH/eZt0RwOTx6V2ugxXUHh3TIr3P2tLSJZ9xyd4QBs/jmgDQqpcarp1neQWdzf2sF1cHEMMsyq8p/2VJyfwq3XE+PtU0fTrrQDfX1jbXA1OCTM8qI/lAtluTnaCeT0oA6u41XTrO8gs7m/tYLq4OIYZZlV5T/sqTk/hVuuH8daxolnL4elutR0+GRtRt5leWZFJhGcsCT9znr05rto5EmiSWN1eNwGVlOQwPQg0AOqpf6rp2lrG2oX9raLK2yM3Eyxh29BkjJq3XG/E2902z8HXJvrm1glcosHnuqsx3qSFz14GTigDpdR1jTNIWNtS1GzslkOENzOsYY+g3EZq4rK6B0YMrDIIOQRXH+KfGGhafYWMqXekzXOpK8djPczIINvG92kJxsHGQDljgD22/DFrZ2PhbS7TT7xL20gtkjiuUYMsoAxuBHGDjtQBrUUUUAFFFFAGV4ktGvdAuYEtZrmQlHSOCVYpAyurBkZvlDKQGGeMqK5nwWmnQazNbzX+qy60sUsv2bU4FheKOSXfIyBFCMGkIywLDgAYAxXSeJ5Y4tAn81bxg7xRKlnMYZXd5FVVDgjbliATkYBNZ/hfT2064lRfC1vpSOhLXK3KzSStkcO2Nzdzkk9KAOnooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKACiiigAooooAKKKKAMDxXHNd6b/AGeNBudWtbkETpBdrblNpUjLF1PJ9D2OetUPC1hJaalI7eH9V05fJKiW71c3SHlflCGV8HvnHYjPNddRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFc4mu6vd+JNQ02x0qye1sJYo5rie/aNzvRXJVBEwOA3dhnHaujrmIPB9o3inVtbvbOxnuJ5oZLOdog8kGyJV6kcHcpIwaAL9r4p0e7lniW6eJoYWnf7TBJADEv3nUyKAyjjLDIGR61Q1Dxxp1roj6nbQXtzGk8EW02c8e4SyBAy5jywwSQQCCQBnJFc83gDWdR2rqd8uW0y5sp7g309w0kkoTEgjcBIxlCSi4HIGTiuh1DT9f1rw/PZ3sOmQXSy28sBhuJHSRo5VkO7MYKAlAON2M96ALl54t0XTxEbu4mhMkXnlWtZd0Uecb5BtzEuQeX2jg+lLN4s0WC8vLWS7bzLJd1ywgkKQjy/MBZwu0AqcjJ56DJ4rJ1HQ9fuLm/urRNMSTVrBLS6Sad2Fuy+Zh0IQeYMSH5SE5HXk1PY+F5NPsddtfLtbyK98pYYrhjtdEto4cSEDjJRumeD+FAG5puq22qxPJbLcqqNtIuLWWBumfuyKpI98Yq7XP+F9K1LS47sX8oEUkim3tReSXQgULggSyAMcnnGMDtXQUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAFFFFABRRRQAUUUUAf/Z"
     },
     "metadata": {
      "image/png": {}
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div class=\"sos_hint\">> ./peer_output/example_data.bed_diagnosis_plot.jpeg:</div>"
      ],
      "text/plain": [
       ">>> ./peer_output/example_data.bed_diagnosis_plot.jpeg:\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash: ./peer_output/example_data.bed_diagnosis_plot.jpeg: Permission denied\n"
     ]
    }
   ],
   "source": [
    "%preview ./output/MWE.Cov_PEER.PEER_diagnosis.pdf -s png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "stuck-satisfaction",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PEER: loading input data... PEER: loading covariate file ...done!\n",
      "PEER: Input summary: 200 samples with 100 genes and 2 covariates \n",
      "Setting initialization parameters ...Done.\n",
      "PEER: estimating hidden confounders (30)\n",
      "\titeration 0/1000\n",
      "\titeration 1/1000\n",
      "\titeration 2/1000\n",
      "\titeration 3/1000\n",
      "\titeration 4/1000\n",
      "\titeration 5/1000\n",
      "Converged (var(residuals)) after 5 iterations\n",
      "\n",
      "Exporting results ... Done.\n"
     ]
    }
   ],
   "source": [
    "cat ./output/MWE.Cov_PEER.PEER.cov.stdout"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "higher-aquatic",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "### b. Marchenko Principal Components\n",
    "This step will compute residual on merged covariates (`Marchenko_PC_1`) and perform hidden factor analysis  (`Marchenko_PC_2`)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdf0ae4a",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "Timing: <1 min"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c2f2e436",
   "metadata": {
    "kernel": "SoS",
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "sos run pipeline/covariate_hidden_factor.ipynb Marchenko_PC \\\n",
    "   --cwd output/covariate \\\n",
    "   --phenoFile output/phenotype/protocol_example.protein.bed.gz  \\\n",
    "   --covFile output/covariate/protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.gz \\\n",
    "   --mean-impute-missing \\\n",
    "   --container containers/PCAtools.sif"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "479c6ba1",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "```\n",
    "INFO: Running computing residual on merged covariates:\n",
    "INFO: computing residual on merged covariates is completed.\n",
    "INFO: computing residual on merged covariates output:   output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.residual.bed.gz\n",
    "INFO: Running Marchenko_PC_2:\n",
    "INFO: Marchenko_PC_2 is completed.\n",
    "INFO: Marchenko_PC_2 output:   output/covariate/protocol_example.protein.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz\n",
    "INFO: Workflow Marchenko_PC (ID=we3d7583ffa60cb28) is executed successfully with 2 completed steps.\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "charming-console",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### c. BiCV Factor Analysis with APEX\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "791b9f2a-d471-4b58-8022-f3d2c04ee397",
   "metadata": {},
   "source": [
    "Timing: X min"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "24122e48-df7a-4c80-96aa-ca657e28d81d",
   "metadata": {},
   "source": [
    "FIXME"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sporting-creator",
   "metadata": {
    "kernel": "SoS",
    "vscode": {
     "languageId": "bat"
    }
   },
   "outputs": [],
   "source": [
    "sos run pipeline/BiCV_factor.ipynb BiCV \\\n",
    "   --cwd output \\\n",
    "   --phenoFile ALL.log2cpm.bed.chr12.mol_phe.bed.gz  \\\n",
    "   --container containers/apex.sif  \\\n",
    "   --N 3"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34432b3a-f4ca-43f5-ba49-6fe909da8b0a",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Troubleshooting"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a82d6d8-b932-46eb-9c41-a320b2168998",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "| Step | Substep | Problem | Possible Reason | Solution |\n",
    "|------|---------|---------|------------------|---------|\n",
    "|  |  |  |  |  |\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "extreme-horizontal",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "kernel": "Bash",
    "tags": []
   },
   "source": [
    "## Command Interface"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "exact-organic",
   "metadata": {
    "kernel": "Bash"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: sos run covariate_hidden_factor.ipynb\n",
      "               [workflow_name | -t targets] [options] [workflow_options]\n",
      "  workflow_name:        Single or combined workflows defined in this script\n",
      "  targets:              One or more targets to generate\n",
      "  options:              Single-hyphen sos parameters (see \"sos run -h\" for details)\n",
      "  workflow_options:     Double-hyphen workflow-specific parameters\n",
      "\n",
      "Workflows:\n",
      "  Marchenko_PC\n",
      "  PEER\n",
      "  BiCV\n",
      "\n",
      "Global Workflow Options:\n",
      "  --cwd output (as path)\n",
      "                        The output directory for generated files. MUST BE FULL\n",
      "                        PATH\n",
      "  --covFile VAL (as path, required)\n",
      "                        Merged Covariates File\n",
      "  --phenoFile VAL (as path, required)\n",
      "                        Path to the input molecular phenotype data.\n",
      "  --name  f'{phenoFile:bnn}.{covFile:bn}'\n",
      "\n",
      "  --job-size 1 (as int)\n",
      "                        For cluster jobs, number commands to run per job\n",
      "  --walltime 5h\n",
      "                        Wall clock time expected\n",
      "  --mem 16G\n",
      "                        Memory expected\n",
      "  --numThreads 8 (as int)\n",
      "                        Number of threads\n",
      "  --container ''\n",
      "                        Software container option\n",
      "  --entrypoint  ('micromamba run -a \"\" -n' + ' ' + re.sub(r'(_apptainer:latest|_docker:latest|\\.sif)$', '', container.split('/')[-1])) if container else \"\"\n",
      "\n",
      "\n",
      "Sections\n",
      "  *_1:\n",
      "    Workflow Options:\n",
      "      --[no-]mean-impute-missing (default to False)\n",
      "  Marchenko_PC_2:\n",
      "  PEER_2:\n",
      "    Workflow Options:\n",
      "      --N 0 (as int)\n",
      "                        N PEER factors, If do not specify or specified as 0,\n",
      "                        default values suggested by GTEx (based on different\n",
      "                        sample size) will be used\n",
      "      --iteration 1000 (as int)\n",
      "                        Default values from PEER software: The number of max\n",
      "                        iteration\n",
      "      --tol 0.001 (as float)\n",
      "                        Prior parameters parameter: Alpha_a = 0.001 parameter:\n",
      "                        Alpha_b = 0.1 parameter: Eps_a = 0.1 parameter: Eps_b =\n",
      "                        10.0 Tolarance parameters\n",
      "      --[no-]r2-tol (default to False)\n",
      "                        parameter: var_tol = 0.00001 minimum variance explained\n",
      "                        criteria to drop factors while training\n",
      "      --convergence-mode fast\n",
      "                        Convergence mode: Convergence mode for MOFAr \"slow\",\n",
      "                        \"medium\" or \"fast\", corresponding to 1e-5%, 1e-4% or\n",
      "                        1e-3% deltaELBO change.\n",
      "  PEER_3:\n",
      "  BiCV_2:\n",
      "  BiCV_3:\n",
      "    Workflow Options:\n",
      "      --N 0 (as int)\n",
      "                        N factors, if not specify, calculated based on sample\n",
      "                        size according to GTeX\n",
      "      --iteration 10 (as int)\n",
      "                        The number of iteration\n"
     ]
    }
   ],
   "source": [
    "!sos run covariate_hidden_factor.ipynb -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "liable-mainland",
   "metadata": {
    "kernel": "R"
   },
   "source": [
    "## Setup and global parameters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "systematic-douglas",
   "metadata": {
    "kernel": "SoS",
    "tags": []
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# The output directory for generated files. MUST BE FULL PATH\n",
    "parameter: cwd = path(\"output\")\n",
    "# Merged Covariates File\n",
    "parameter: covFile = path\n",
    "# Path to the input molecular phenotype data.\n",
    "parameter: phenoFile = path\n",
    "parameter: name = f'{phenoFile:bnn}.{covFile:bn}'\n",
    "# For cluster jobs, number commands to run per job\n",
    "parameter: job_size = 1\n",
    "# Wall clock time expected\n",
    "parameter: walltime = \"5h\"\n",
    "# Memory expected\n",
    "parameter: mem = \"16G\"\n",
    "# Number of threads\n",
    "parameter: numThreads = 8\n",
    "# Software container option\n",
    "parameter: container = \"\"\n",
    "import re\n",
    "parameter: entrypoint= ('micromamba run -a \"\" -n' + ' ' + re.sub(r'(_apptainer:latest|_docker:latest|\\.sif)$', '', container.split('/')[-1])) if container else \"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b9d4ef64-2a88-4225-8184-3d686d705d54",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "placed-mineral",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[*_1(computing residual on merged covariates)]\n",
    "parameter: mean_impute_missing = False\n",
    "input: phenoFile, covFile\n",
    "output: f'{cwd}/{name}.residual.bed.gz'\n",
    "task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f'{step_name}_{_output:bn}'\n",
    "R: expand = \"${ }\", stderr = f'{_output:nn}.stderr', stdout = f'{_output:nn}.stdout' , container = container, entrypoint = entrypoint\n",
    "\n",
    "    mean_impute <- function(d){\n",
    "      f <- apply(d, 2, function(x) mean(x, na.rm = TRUE))\n",
    "      for (i in 1:length(f)) d[,i][which(is.na(d[,i]))] <- f[i]\n",
    "      return(d)\n",
    "    }\n",
    "  \n",
    "    library(dplyr)\n",
    "    library(readr)\n",
    "    pheno = read_delim(${_input[0]:r},delim = \"\\t\")\n",
    "    covariate= read_delim(${_input[1]:r},delim = \"\\t\") \n",
    "\n",
    "    # Extract samples in both files (Outliers will be removed in here as they are no longer presented in the header of factor)\n",
    "    extraction_sample_list <- intersect(colnames(pheno), colnames(covariate)) \n",
    "    \n",
    "    \n",
    "    if(length(extraction_sample_list) == 0){\n",
    "      stop(\"No samples are overlapped in two files!\")\n",
    "    }\n",
    "    \n",
    "    ## Report sample counts:\n",
    "    print(paste((ncol(pheno) - 4), \"samples are in the phenotype file\", sep = \" \"))\n",
    "    print(paste((ncol(covariate) - 1), \"samples are in the covariate file\", sep = \" \"))\n",
    "\n",
    "    ## Report identical samples:\n",
    "    print(paste(length(extraction_sample_list), \"samples overlap between phenotype & covariate files and are included in the analysis:\", sep = \" \"))\n",
    "    print(extraction_sample_list)\n",
    "\n",
    "    ## Report non-overlapping samples :\n",
    "    covariate_missing = covariate %>% select(-all_of('#id')) %>% select(-all_of(extraction_sample_list))\n",
    "    print(paste(ncol(covariate_missing), \"samples in the covariate file are missing from the phenotype file:\", sep = \" \"))\n",
    "    print(colnames(covariate_missing))\n",
    "\n",
    "    pheno_missing = pheno %>% select(-c(1:4)) %>% select(-all_of(extraction_sample_list))\n",
    "    print(paste(ncol(pheno_missing), \"samples in the phenotype file are missing from the covariate file:\", sep = \" \"))\n",
    "    print(colnames(pheno_missing))\n",
    "\n",
    "    # Subset the data:\n",
    "    covariate = covariate[,extraction_sample_list]%>%as.matrix()%>%t()\n",
    "    pheno_id = pheno%>%select(1:4)\n",
    "    pheno = pheno%>%select(all_of(rownames(covariate)))%>%as.matrix()%>%t()\n",
    "    if (${\"T\" if mean_impute_missing else \"F\"}) {\n",
    "      pheno = mean_impute(pheno)\n",
    "    } else {\n",
    "      if(sum(is.na(pheno)) > 0){ stop(\"NA in phenotype input is not allowed!\") }\n",
    "    }\n",
    "    # Get residual \n",
    "    pheno_resid = .lm.fit(x = cbind(1,covariate), y = pheno)$residuals\n",
    "    pheno_output = cbind(pheno_id, pheno_resid%>%t())\n",
    "    pheno_output%>%write_delim(\"${_output:n}\",delim = \"\\t\")\n",
    "  \n",
    "bash: expand = \"${ }\", stderr = f'{_output:nn}.stderr', stdout = f'{_output:nn}.stdout', container = container, entrypoint = entrypoint\n",
    "    bgzip -f ${_output:n}\n",
    "    tabix -p bed ${_output}\n",
    "\n",
    "    stdout=${_output:n}.stdout \n",
    "    for i in ${_output} ; do \n",
    "    echo \"output_info: $i \" >> $stdout;\n",
    "    echo \"output_size:\" `ls -lh $i | cut -f 5  -d  \" \"`   >> $stdout;\n",
    "    echo \"output_rows:\" `zcat $i | wc -l  | cut -f 1 -d \" \"`   >> $stdout;\n",
    "    echo \"output_column:\" `zcat $i | head -1 | wc -w `   >> $stdout;\n",
    "    echo \"output_preview:\"   >> $stdout;\n",
    "    zcat $i | head  | cut -f 1,2,3,4,5,6   >> $stdout ; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "841fea8c-5777-48a2-b10c-98a110ad737a",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Marchenko Principal Components"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lesser-hygiene",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[Marchenko_PC_2]\n",
    "output: f'{cwd}/{_input:bnnn}.Marchenko_PC.gz'\n",
    "task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "R: container=container, expand= \"${ }\", stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout', entrypoint = entrypoint\n",
    "    library(\"dplyr\")\n",
    "    library(\"readr\")\n",
    "    library(\"PCAtools\")\n",
    "    library('BiocSingular')\n",
    "    residExpPath  = \"${_input}\"\n",
    "    covPath = \"${covFile}\"\n",
    "    residExpDF <- read_delim(residExpPath, show_col_types=FALSE)\n",
    "    covDF <- read_delim(covPath, show_col_types=FALSE)\n",
    "    commonMPSamples <- intersect(colnames(covDF), colnames(residExpDF))\n",
    "    covDFcommon <- cbind(covDF[, 1], covDF[, commonMPSamples])\n",
    "  \n",
    "    residExpPC <- pca(\n",
    "        residExpDF[,commonMPSamples], # The first four columns are: chr, start, end, and gene_id; so we skip those.\n",
    "        scale = TRUE,\n",
    "        center = TRUE,\n",
    "        BSPARAM = ExactParam())\n",
    "    M <- apply(residExpDF[, commonMPSamples], 1, function(X){ (X - mean(X))/sqrt(var(X))});\n",
    "    residSigma2 <- var(as.vector(M));\n",
    "    paste('sigma2:', residSigma2)\n",
    "    \n",
    "    MPPCNum <- chooseMarchenkoPastur(\n",
    "        .dim = dim(residExpDF[, commonMPSamples]), var.explained=residExpPC$sdev^2, noise=residSigma2)\n",
    "    \n",
    "    MPPCsDF <- as.data.frame(residExpPC$rotated[, 1:MPPCNum])\n",
    "\n",
    "    MPColMatrix <-  matrix(c(rep('Hidden_Factor_PC', times=MPPCNum), seq(1, MPPCNum)), ncol=2, nrow=MPPCNum)\n",
    "    colnames(MPPCsDF) <- apply(MPColMatrix, 1, function(X){return(paste0(X[1], X[2]))})\n",
    "    # Add #id Column\n",
    "    MPPCsDF <- as.data.frame(t(MPPCsDF))\n",
    "    MPPCsDF$id <- rownames(MPPCsDF)\n",
    "    MPPCsDF <- MPPCsDF %>% select(id, everything()) %>% rename(\"#id\" = \"id\")\n",
    "    write_delim((rbind(covDFcommon, MPPCsDF)), \"${_output}\", \"\\t\")\n",
    "bash: expand= \"$[ ]\", stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout', container = container, entrypoint = entrypoint\n",
    "        stdout=$[_output:n].stdout\n",
    "        for i in $[_output] ; do \n",
    "        echo \"output_info $i \" >> $stdout;\n",
    "        echo \"output_size:\" `ls -lh $i | cut -f 5  -d  \" \"`   >> $stdout;\n",
    "        echo \"output_rows:\" `zcat $i | wc -l  | cut -f 1 -d \" \"`   >> $stdout;\n",
    "        echo \"output_column:\" `zcat $i | head -1 | wc -w `   >> $stdout;\n",
    "        echo \"output_preview:\"   >> $stdout;\n",
    "        zcat $i | head  | cut -f 1,2,3,4,5,6   >> $stdout ; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e7d2ee6-54da-4d6a-85a5-c0e9d9664087",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### PEER Method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adaptive-saskatchewan",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[PEER_2]\n",
    "# N PEER factors, If do not specify or specified as 0, default values suggested by \n",
    "# GTEx (based on different sample size) will be used\n",
    "parameter: N = 0\n",
    "# Default values from PEER software:\n",
    "## The number of max iteration\n",
    "parameter: iteration = 1000\n",
    "### Prior parameters\n",
    "#parameter: Alpha_a = 0.001\n",
    "#parameter: Alpha_b = 0.1\n",
    "#parameter: Eps_a = 0.1\n",
    "#parameter: Eps_b = 10.0\n",
    "# Tolarance parameters\n",
    "parameter: tol = 0.001\n",
    "#parameter: var_tol = 0.00001\n",
    "# minimum variance explained criteria to drop factors while training\n",
    "parameter: r2_tol = False\n",
    "# Convergence mode: Convergence mode for MOFAr \"slow\", \"medium\" or \"fast\", corresponding to 1e-5%, 1e-4% or 1e-3% deltaELBO change.\n",
    "parameter: convergence_mode = \"fast\"\n",
    "# input is the residual file from the first step\n",
    "output: f'{cwd:a}/{_input:bnn}.PEER_MODEL.hd5'\n",
    "task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "python: container=container, expand= \"${ }\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout', entrypoint = entrypoint\n",
    "    from mofapy2.run.entry_point import entry_point\n",
    "    import pandas as pd\n",
    "    import os\n",
    "    ## Force the number of threads used in numpy.\n",
    "    os.environ[\"OMP_NUM_THREADS\"] = \"${numThreads}\" # export OMP_NUM_THREADS=4\n",
    "    os.environ[\"OPENBLAS_NUM_THREADS\"] = \"${numThreads}\" # export OPENBLAS_NUM_THREADS=4 \n",
    "    os.environ[\"MKL_NUM_THREADS\"] = \"${numThreads}\" # export MKL_NUM_THREADS=6\n",
    "    import numpy as np\n",
    "    ## Data\n",
    "    data = pd.read_csv(${_input:r},\"\\t\",index_col = 3).drop([\"#chr\",\"start\",\"end\"],axis = 1)\n",
    "    # initialise the entry point\n",
    "    ent = entry_point()\n",
    "    # Guess number of factor\n",
    "    # Suggest the number of factors to use if no input value\n",
    "    num_factor = ${N}\n",
    "    if num_factor == 0:\n",
    "        if len(data.columns) < 150:\n",
    "            num_factor = 15\n",
    "        elif len(data.columns) < 250:\n",
    "            num_factor = 30\n",
    "        elif len(data.columns) < 350:\n",
    "            num_factor = 45\n",
    "        else:\n",
    "            num_factor = 60\n",
    "    # Set data\n",
    "    # MOFA is a multi-view and multi-group inference framework. \n",
    "    # If usig only a single view and a single group (as in PEER), the data needs to be embedded into a nested list\n",
    "    ent.set_data_matrix([[data.transpose()]],samples_names=[data.columns.values.tolist()], features_names=[data.index.values.tolist()])\n",
    "    ent.set_model_options(factors= num_factor , spikeslab_weights=False, ard_weights=False)\n",
    "    ent.set_train_options(iter=${iteration}, convergence_mode=\"${convergence_mode}\" ${f',dropR2 = {r2_tol}' if r2_tol else \"\" }, startELBO=1, freqELBO=1,tolerance=${tol}, gpu_mode=False, verbose=True, seed=42)\n",
    "    ent.build()\n",
    "    ent.run()\n",
    "    ent.save(${_output:ar})\n",
    "    ## To fix issue https://github.com/cumc/xqtl-pipeline/issues/420\n",
    "    import h5py\n",
    "    right_name = [x.encode(\"UTF-8\") for x in  ent.data_opts[\"features_names\"][0]] \n",
    "    new_hd5 = h5py.File(${_output:ar}, \"r+\")\n",
    "    del new_hd5[\"features/view0\"]\n",
    "    new_hd5[\"features\"].create_dataset(\"view0\", data=np.array(right_name))\n",
    "    new_hd5.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "executive-temple",
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[PEER_3]\n",
    "output: f'{cwd:a}/{_input:bnn}.PEER.gz', f'{cwd:a}/{_input:bnn}.PEER.diag.pdf'\n",
    "task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "R:  container=container, expand= \"${ }\", stderr = f'{_output[0]:n}.stderr', stdout = f'{_output[0]:n}.stdout', entrypoint = entrypoint\n",
    "    ####################\n",
    "    ## Load libraries ##\n",
    "    ####################\n",
    "    \n",
    "    library(MOFA2)\n",
    "    library(dplyr)\n",
    "    library(readr)\n",
    "    ################\n",
    "    ## Load model ##\n",
    "    ################\n",
    "    \n",
    "    file <- \"${_input}\"\n",
    "    model <- load_model(file)\n",
    "    \n",
    "    ####################################\n",
    "    ## (Optional) add sample metadata ##\n",
    "    ####################################\n",
    "    \n",
    "    # Important: \n",
    "    # (1) the row names in the sample metadata data.frame have to match the sample names in the MOFA model\n",
    "    # (2) the sample name has to contain at least two columns: \n",
    "    # \t\t- sample: sample names\n",
    "    #\t\t- group: group name. Only required for multi-group inference. In your case just set the entire column to the same value\n",
    "    \n",
    "    # Load metadata\n",
    "    \n",
    "    # stopifnot(all(samples(model)[[1]]==metadata$sample))\n",
    "    # samples_metadata(model) <- metadata\n",
    "    \n",
    "    ###############################\n",
    "    ## (Optional) Subset factors ##\n",
    "    ###############################\n",
    "    \n",
    "    # We can remove factors that explain little variance (in this case, we require at least 0.01%)\n",
    "    # r2 <- model@cache$variance_explained$r2_per_factor\n",
    "    # factors <- sapply(r2, function(x) x[,1]>0.0001)\n",
    "    # model <- subset_factors(model, which(apply(factors,1,sum) >= 1))\n",
    "    \n",
    "    # fetch factors in matrix format, with dimensions (nfactors,nsamples)\n",
    "    factors <- get_factors(model)[[1]]\n",
    "    dim(factors)\n",
    "    factors = as_tibble(t(factors),rownames = \"#id\")\n",
    "    covPath = \"${covFile}\"\n",
    "    options(readr.show_col_types = FALSE)\n",
    "    covDF <- read_delim(covPath, delim = '\\t')\n",
    "    commonMPSamples <- intersect(colnames(covDF), colnames(factors))      \n",
    "    (rbind(covDF[, commonMPSamples], factors[, commonMPSamples])%>%write_delim(\"${_output[0]}\",\"\\t\"))\n",
    "    # Plot save to vcf\n",
    "    pdf(\"${_output[1]}\")\n",
    "    #############################\n",
    "    ## Plot variance explained ##\n",
    "    #############################\n",
    "    \n",
    "    # Plot variance explained using individual factors\n",
    "    plot_variance_explained(model, factors=\"all\")\n",
    "    plot_variance_explained(model, factors=c(1,2,3))\n",
    "    \n",
    "    # Plot total variance explained using all factors\n",
    "    plot_variance_explained(model, plot_total = TRUE)[[2]]\n",
    "    \n",
    "    ########################\n",
    "    ## Plot factor values ##\n",
    "    ########################\n",
    "    \n",
    "    plot_factor(model, factor = 1)\n",
    "    plot_factors(model, factor = c(1,2))\n",
    "    \n",
    "    ###########################\n",
    "    ## Plot feature loadings ##\n",
    "    ###########################\n",
    "    \n",
    "    # The weights or loadings provide a score for each gene on each factor. \n",
    "    # Genes with no association with the factor are expected to have values close to zero\n",
    "    # Genes with strong association with the factor are expected to have large absolute values. \n",
    "    # The sign of the loading indicates the direction of the effect: a positive loading indicates that the feature is more active in the cells with positive factor values, and viceversa.\n",
    "    \n",
    "    # Plot the distribution of loadings for Factor 1.\n",
    "    plot_weights(model,\n",
    "      view = 1,\n",
    "      factor = 1,\n",
    "      nfeatures = 10,     # Top number of features to highlight\n",
    "      scale = T           # Scale loadings from -1 to 1\n",
    "    )\n",
    "    \n",
    "    # If you are not interested in the full distribution, but just on the top loadings:\n",
    "    plot_top_weights(model, \n",
    "      view = 1, \n",
    "      factor = 1, \n",
    "      nfeatures = 10,\n",
    "      scale = T\n",
    "    )\n",
    "    \n",
    "    ######################################\n",
    "    ## Plot correlation between factors ##\n",
    "    ######################################\n",
    "    \n",
    "    plot_factor_cor(model)\n",
    "    \n",
    "    ###################\n",
    "    ## Fetch factors ##\n",
    "    ###################\n",
    "    invisible(dev.off())\n",
    "bash: expand= \"$[ ]\", stderr = f'{_output[0]:n}.stderr', stdout = f'{_output[0]:n}.stdout', container = container, entrypoint = entrypoint\n",
    "        for i in $[_output[0]] ; do \n",
    "        echo \"output_info $i \" >> $[_output[0]:n].stdout;\n",
    "        echo \"output_size:\" `ls -lh $i | cut -f 5  -d  \" \"`   >> $[_output[0]:n].stdout;\n",
    "        echo \"output_rows:\" `zcat $i | wc -l  | cut -f 1 -d \" \"`   >> $[_output[0]:n].stdout;\n",
    "        echo \"output_column:\" `zcat $i | head -1 | wc -w `   >> $[_output[0]:n].stdout;\n",
    "        echo \"output_preview:\"   >> $[_output[0]:n].stdout;\n",
    "        zcat $i | head  | cut -f 1,2,3,4,5,6   >> $[_output[0]:n].stdout ; done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "structured-paraguay",
   "metadata": {
    "kernel": "Bash"
   },
   "source": [
    "#### Reference\n",
    "* PEER code is adapted from [here](https://github.com/RTIInternational/biocloud_docker_tools/blob/master/peer/v1.3/run_peer.R)\n",
    "* GTEx recommandation of PEER factors is [here](https://gtexportal.org/home/documentationPage)\n",
    "* Examples by PEER is at [github](https://github.com/PMBio/peer/wiki/Tutorial)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "regulation-breeding",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### BiCV Factor Analysis with APEX\n",
    "\n",
    "Notice that the command options are different from those on the APEX website documentation. The commands on the documentation page does not work (last updated September 2021). The commands below were constructed and tested by our team based on our understanding of the program, without input from APEX authors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "according-copying",
   "metadata": {
    "kernel": "SoS",
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "[BiCV_2]\n",
    "# For cluster jobs, number commands to run per job\n",
    "import time\n",
    "output: f'{cwd:a}/{_input:bn}.fake.vcf.gz'\n",
    "#task: trunk_workers = 1,trunk_size = job_size , walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output:bn}'\n",
    "R: container=container, expand= \"$[ ]\", stderr = f'{_output}.stderr', stdout = f'{_output}.stdout', entrypoint = entrypoint\n",
    "    library(\"dplyr\")\n",
    "    library(\"readr\")\n",
    "    ## Add fake header\n",
    "    cat(paste(\"##fileformat=VCFv4.2\\n##fileDate=$[time.strftime(\"%Y%m%d\",time.localtime())]\\n##source=FAKE\\n\"), file=$[_output:nr], append=FALSE)\n",
    "    ## Add colnames based on bed\n",
    "    pheno = read_delim(\"$[_input]\", delim = \"\\t\",n_max = 1)\n",
    "    colnames(pheno)[1:3] = c(\"#CHROM\",\"POS\",\"ID\") \n",
    "    pheno = cbind(pheno[1:3]%>%mutate(REF = \"A\", ALT = \"C\", QUAL = \".\",FILTER = \".\", INFO = \"PR\", FORMAT = \"GT\"),pheno[,5:ncol(pheno)])\n",
    "    pheno%>%write_delim($[_output:nr],delim = \"\\t\",col_names = T, append = T)\n",
    "bash: container=container, expand= \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout', entrypoint = entrypoint\n",
    "    bgzip -f $[_output:n]\n",
    "    tabix -p vcf $[_output]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dynamic-columbus",
   "metadata": {
    "kernel": "SoS",
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [],
   "source": [
    "[BiCV_3]\n",
    "# N factors, if not specify, calculated based on sample size according to GTeX\n",
    "parameter: N = 0\n",
    "# The number of iteration\n",
    "parameter: iteration = 10\n",
    "\n",
    "import pandas as pd\n",
    "data = pd.read_csv(phenoFile,\"\\t\",index_col = 3).drop([\"#chr\",\"start\",\"end\"],axis = 1)\n",
    "if N == 0:\n",
    "    if len(data.columns) < 150:\n",
    "        N = 15\n",
    "    elif len(data.columns) < 250:\n",
    "        N = 30\n",
    "    elif len(data.columns) < 350:\n",
    "        N = 45\n",
    "    else:\n",
    "        N = 60\n",
    "input: output_from(1), output_from(2)\n",
    "output: f'{cwd:a}/{_input[0]:bnn}.BiCV.gz'\n",
    "task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f'{step_name}_{_output[0]:bn}'\n",
    "bash: container=container, expand= \"$[ ]\", stderr = f'{_output[0]}.stderr', stdout = f'{_output[0]}.stdout', entrypoint = entrypoint\n",
    "    apex factor \\\n",
    "        --out $[_output[0]:nn] \\\n",
    "        --iter $[iteration] \\\n",
    "        --factors $[N] \\\n",
    "        --bed $[_input[0]] \\\n",
    "        --vcf $[_input[1]] \\\n",
    "        --threads $[numThreads]  $[ f'--cov {covFile}' if covFile.is_file() else '']"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "Bash",
     "bash",
     "Bash",
     "#E6EEFF",
     "shell"
    ],
    [
     "R",
     "ir",
     "R",
     "#DCDCDA",
     "r"
    ],
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ]
   ],
   "version": "0.22.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
