
<!DOCTYPE html>


<html lang="en" data-content_root="../../../" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>Hidden Factor Analysis &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.5.2/css/all.min.css?digest=dfe6caa3a7d634c4db9b" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.5.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=03e43079" />
    <link rel="stylesheet" type="text/css" href="../../../_static/styles/sphinx-book-theme.css?v=eba8b062" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css?v=13237357" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css?v=76b2166b" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css?v=4fa983c6" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-design.min.css?v=95c83b7e" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b" />
  <script src="../../../_static/vendor/fontawesome/6.5.2/js/all.min.js?digest=dfe6caa3a7d634c4db9b"></script>

    <script src="../../../_static/documentation_options.js?v=9eb32ce0"></script>
    <script src="../../../_static/doctools.js?v=9a2dae69"></script>
    <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/clipboard.min.js?v=a7894cd8"></script>
    <script src="../../../_static/copybutton.js?v=f281be69"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?v=887ef09a"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js?v=4a39c7ea"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js?v=f930bc37"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script async="async" src="../../../_static/sphinx-thebe.js?v=c100c467"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"; const thebe_selector = ".thebe,.cell"; const thebe_selector_input = "pre"; const thebe_selector_output = ".output, .cell_output"</script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/data_preprocessing/covariate/covariate_hidden_factor';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="QTL Association Analysis" href="../../association_scan/qtl_association_testing.html" />
    <link rel="prev" title="Covariate Data Formatting" href="covariate_formatting.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <div id="pst-skip-link" class="skip-link d-print-none"><a href="#main-content">Skip to main content</a></div>
  
  <div id="pst-scroll-pixel-helper"></div>
  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>Back to top</button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-primary-sidebar-checkbox"/>
  <label class="overlay overlay-primary" for="pst-primary-sidebar-checkbox"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          id="pst-secondary-sidebar-checkbox"/>
  <label class="overlay overlay-secondary" for="pst-secondary-sidebar-checkbox"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>

  <div class="pst-async-banner-revealer d-none">
  <aside id="bd-header-version-warning" class="d-none d-print-none" aria-label="Version warning"></aside>
</div>

  
    <header class="bd-header navbar navbar-expand-lg bd-navbar d-print-none">
    </header>
  

  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  
    
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item">

 <script>
 document.write(`
   <button class="btn search-button-field search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
    <span class="search-button__default-text">Search</span>
    <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd class="kbd-shortcut__modifier">K</kbd></span>
   </button>
 `);
 </script></div>
        <div class="sidebar-primary-item"><nav class="bd-links bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../reference_data/reference_data.html">Reference Data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../genotype_preprocessing.html">Genotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../phenotype_preprocessing.html">Phenotype data preprocessing</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/phenotype_imputation.html">Phenotype data imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</details></li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../covariate_preprocessing.html">Covariate Data Preprocessing</a><details open="open"><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Hidden Factor Analysis</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/qtl_association_testing.html">QTL Association Analysis</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/quantile_models/qr_and_twas.html">Quantile regression for QTL association testing</a></li>
</ul>
</details></li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Mixture Model</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../multivariate_genome/multivariate_mixture_vignette.html">Mixture Multivariate Distribution Estimate</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multiomics Regression Models</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../mnm_analysis/fine_mapping_miniprotocol.html">Fine-Mapping</a><details><summary><span class="toctree-toggle" role="presentation"><i class="fa-solid fa-chevron-down"></i></span></summary><ul>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_twas_vignette.html">Univariate Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_multigene_fine_mapping_vignette.html">Multivariate Fine-Mapping for multiple genes</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/univariate_fine_mapping_fsusie_vignette.html">Univariate Fine-Mapping of Functional (Epigenomic) Data with fSuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/multivariate_fine_mapping_vignette.html">Multivariate Fine-Mapping with mvSuSiE and mr.mash</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/summary_stats_finemapping_vignette.html">Regression with Summary Statistics (RSS) Fine-Mapping and TWAS with SuSiE</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/mnm_regression.html">Advanced regression models for association analysis with individual level data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../mnm_analysis/mnm_methods/rss_analysis.html">High-dimensional regression with summary statistics</a></li>
</ul>
</details></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/SuSiE_enloc.html">xQTL-GWAS pairwise enrichment and colocalization</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../pecotmr_integration/twas_ctwas.html">TWAS, cTWAS and MR</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../mnm_analysis/mnm_methods/colocboost.html">Multi-trait colocalization using ColocBoost</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment and Validation</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/eoo_enrichment.html">Chromosome-Specific Enrichment Analysis of Annotations Using Block Jackknife</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gsea.html">Pathway Analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../enrichment/sldsc_enrichment.html">Stratified LD Score Regression</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main" role="main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article d-print-none">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><button class="sidebar-toggle primary-toggle btn btn-sm" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</button></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/edit/gh-pages/code/data_preprocessing/covariate/covariate_hidden_factor.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/statfungen/xqtl-protocol/issues/new?title=Issue%20on%20page%20%2Fcode/data_preprocessing/covariate/covariate_hidden_factor.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/data_preprocessing/covariate/covariate_hidden_factor.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm nav-link pst-navbar-icon theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="theme-switch fa-solid fa-sun fa-lg" data-mode="light"></i>
    <i class="theme-switch fa-solid fa-moon fa-lg" data-mode="dark"></i>
    <i class="theme-switch fa-solid fa-circle-half-stroke fa-lg" data-mode="auto"></i>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm pst-navbar-icon search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<button class="sidebar-toggle secondary-toggle btn btn-sm" title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</button>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Hidden Factor Analysis</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#default-parameters-number-of-factors">Default Parameters: Number of Factors</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#peer">PEER</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#bicv-factor-analysis-with-apex">BiCV factor analysis with APEX</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-with-marchenko-s-approach">PCA with Marchenko’s approach</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-with-buja-a-and-eyuboglu-n-permutation">PCA with Buja A and Eyuboglu N. permutation</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#primary-output">Primary Output</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#secondary-outputs">Secondary Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-generate-hidden-factors">ii. Generate Hidden Factors</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#a-peer-method">a. PEER Method</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#b-principal-components">b. Principal Components</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-and-global-parameters">Setup and global parameters</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#principal-components-analysis-on-molecular-phenotype-matrix">Principal Components Analysis on molecular phenotype matrix</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#peer-method">PEER Method</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#reference">Reference</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">BiCV Factor Analysis with APEX</a></li>
</ul>
</li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="hidden-factor-analysis">
<h1>Hidden Factor Analysis<a class="headerlink" href="#hidden-factor-analysis" title="Link to this heading">#</a></h1>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Link to this heading">#</a></h2>
<p>We provide two different procedures for hidden factor analysis from omics data in our pipeline. The first is the <a class="reference external" href="https://github.com/PMBio/peer/wiki/Tutorial">Probabilistic Estimation of Expression Residuals (PEER) method</a>, a method also used for GTEx eQTL data analysis. The second, and the one use for our main analyses, is a PCA based approach with automatic determination of the number of factors to use. This is mainly inspired by a recent benchmark from Jessica Li’s group [<a class="reference external" href="https://doi.org/10.1186/s13059-022-02761-4">cf. Zhou et al., Genome Biology, 2022</a>]. Please note that additional considerations should be taken for single-cell eQTL analysis as investigated by [<a class="reference external" href="https://doi.org/10.1186/s13059-023-02873-5">cf. Xue et al., Genome Biology, 2023</a>].</p>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Link to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>Molecular phenotype in <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> file format with indexing:</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>readr::read_delim(&quot;data/example_data.bed.gz&quot;,show_col_types= F)[1:3,1:8]
</pre></div>
</div>
</div>
</div>
<ol class="arabic simple" start="2">
<li><p>A covariate file in text format with the first column <code class="docutils literal notranslate"><span class="pre">#id</span></code> that lists the name of all covariates, followed by other columns each being a sample with sample names matching that in the molecular phenotype file:</p></li>
</ol>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>readr::read_delim(&quot;data/example_cov.txt&quot;,show_col_types= F)[,1:8]
</pre></div>
</div>
</div>
</div>
<p>Covariates file is optional to the PEER method but it should be supplied in order to construct a proper PEER model in the context of xQTL mapping.</p>
<ol class="arabic simple" start="3">
<li><p>For <code class="docutils literal notranslate"><span class="pre">APEX</span></code> we need an indexed VCF file to get the command to work, although this file is not used.</p></li>
</ol>
<section id="default-parameters-number-of-factors">
<h3>Default Parameters: Number of Factors<a class="headerlink" href="#default-parameters-number-of-factors" title="Link to this heading">#</a></h3>
<section id="peer">
<h4>PEER<a class="headerlink" href="#peer" title="Link to this heading">#</a></h4>
<p>By default the pipeline applies recommendations from <a class="reference external" href="https://gtexportal.org/home/documentationPage">GTEx</a>:</p>
<blockquote>
<div><p>A set of covariates identified using the Probabilistic Estimation of Expression Residuals (PEER) method (Stegle et al., PLoS Comp. Biol., 2010 ), calculated for the normalized expression matrices (described below). For eQTL analyses, the number of PEER factors was determined as function of sample size (N): 15 factors for N&lt;150, 30 factors for 150≤ N&lt;250, 45 factors for 250≤ N&lt;350, and 60 factors for N≥350, as a result of optimizing for the number of eGenes discovered. For sQTL analyses, 15 PEER factors were computed for each tissue.</p>
</div></blockquote>
<p>Although one can specify fixed number of factors using <code class="docutils literal notranslate"><span class="pre">--N</span></code> option in this pipeline.</p>
</section>
<section id="bicv-factor-analysis-with-apex">
<h4>BiCV factor analysis with APEX<a class="headerlink" href="#bicv-factor-analysis-with-apex" title="Link to this heading">#</a></h4>
<p>We will do the same as PEER</p>
</section>
<section id="pca-with-marchenko-s-approach">
<h4>PCA with Marchenko’s approach<a class="headerlink" href="#pca-with-marchenko-s-approach" title="Link to this heading">#</a></h4>
<p>We will use Marchenko’s approach to determine the number of PC to use.</p>
</section>
<section id="pca-with-buja-a-and-eyuboglu-n-permutation">
<h4>PCA with Buja A and Eyuboglu N. permutation<a class="headerlink" href="#pca-with-buja-a-and-eyuboglu-n-permutation" title="Link to this heading">#</a></h4>
<p>We also provide an option to use Buja A and Eyuboglu N. to determine the number of PC to use.</p>
</section>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Link to this heading">#</a></h2>
<section id="primary-output">
<h3>Primary Output<a class="headerlink" href="#primary-output" title="Link to this heading">#</a></h3>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">*.cov.gz</span></code>: hidden factors inferred.</p></li>
</ul>
</section>
<section id="secondary-outputs">
<h3>Secondary Outputs<a class="headerlink" href="#secondary-outputs" title="Link to this heading">#</a></h3>
<p>PEER:</p>
<ul class="simple">
<li><p>Diagnosis plots for check of convergence</p></li>
<li><p>peer precision /residuals /weights</p></li>
</ul>
</section>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Link to this heading">#</a></h2>
<p>An MWE is uploaded to <a class="reference external" href="https://drive.google.com/drive/folders/1yjTwoO0DYGi-J9ouMsh9fHKfDmsXJ_4I?usp=sharing">google drive</a>.</p>
<section id="ii-generate-hidden-factors">
<h3>ii. Generate Hidden Factors<a class="headerlink" href="#ii-generate-hidden-factors" title="Link to this heading">#</a></h3>
<section id="a-peer-method">
<h4>a. PEER Method<a class="headerlink" href="#a-peer-method" title="Link to this heading">#</a></h4>
<p>Because the MWE only contains 10 genes but 400+ samples. The computed N will be far greater than the number of genes. Therefore in the MWE the N is fixed to be 3.</p>
<p>Timing: X min</p>
<p>FIXME: error due to “import scipy as s” creating issue where a “s.shape” is called in <a class="reference external" href="https://github.com/bioFAM/mofapy2/blame/master/mofapy2/core/distributions/basic_distributions.py">mofapy2 code</a></p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run covariate_hidden_factor.ipynb PEER \
   --cwd ../../../output_test/peer_factors \
   --phenoFile ../../../mwe_data/xQTL_discovery/MWE.log2cpm.mol_phe.bed.gz \
   --covFile ../../../mwe_data/xQTL_discovery/MWE.covariate.cov.gz \
   --container oras://ghcr.io/statfungen/factor_analysis_apptainer:latest \
   --N 3 \
   -c ../../csg.yml -q neurology
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">computing residual on merged covariates</span>: 
INFO: t6daa86bd17ff21b8 <span class=" -Color -Color-Green">re-execute completed</span>
INFO: t6daa86bd17ff21b8 <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 6008724 (&quot;job_t6daa86bd17ff21b8&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">computing residual on merged covariates</span> output:   <span class=" -Color -Color-Green">../../../output_test/peer_factors/MWE.log2cpm.mol_phe.MWE.covariate.cov.residual.bed.gz</span>
INFO: Running <span class=" -Color -Color-Green">PEER_2</span>: 
INFO: t4b3df6fd53d8215c <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 6008740 (&quot;job_t4b3df6fd53d8215c&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
ERROR: [PEER_2]: [t4b3df6fd53d8215c]: Executing script in Singularity returns an error (exitcode=1, stderr=/restricted/projectnb/xqtl/xqtl_protocol/output_test/peer_factors/MWE.log2cpm.mol_phe.MWE.covariate.cov.residual.PEER_MODEL.hd5.stderr, stdout=/restricted/projectnb/xqtl/xqtl_protocol/output_test/peer_factors/MWE.log2cpm.mol_phe.MWE.covariate.cov.residual.PEER_MODEL.hd5.stdout).
The script has been saved to /usr3/graduate/fgrennjr/.sos/8a30b13094e465a1//usr3/graduate/fgrennjr/.sos/8a30b13094e465a1.To reproduce the error please run:
<span class=" -Color -Color-Green">singularity exec  /usr3/graduate/fgrennjr/.sos/singularity/library/ghcr.io-statfungen-factor_analysis_apptainer-latest.sif micromamba run -a &quot;&quot; -n factor_analysis python /usr3/graduate/fgrennjr/.sos/8a30b13094e465a1/singularity_run_854003.py</span>
[PEER]: Exits with 1 pending step (PEER_3)
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>tree ./output
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span><span class=" -Color -Color-Bold -Color-Bold-Blue">./output</span>
├── MWE.Cov_PEER.PEER.cov
├── MWE.Cov_PEER.PEER.cov.stderr
├── MWE.Cov_PEER.PEER.cov.stdout
├── MWE.Cov_PEER.PEER_diagnosis.pdf
├── MWE.Cov_PEER.PEER.peer_precisions.txt
├── MWE.Cov_PEER.PEER.peer_residuals.txt
└── MWE.Cov_PEER.PEER.peer_weights.txt

0 directories, 7 files
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>%preview ./output/MWE.Cov_PEER.PEER_diagnosis.pdf -s png
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output text_html"><div class="sos_hint">> ./peer_output/example_data.bed_diagnosis_plot.jpeg (22.2 KiB):</div></div><img alt="../../../_images/429dca9b037f6cebd6580ac9b87760d969f0c1f38fd5728bc7b6011c4f8f3929.jpg" src="../../../_images/429dca9b037f6cebd6580ac9b87760d969f0c1f38fd5728bc7b6011c4f8f3929.jpg" />
<div class="output text_html"><div class="sos_hint">> ./peer_output/example_data.bed_diagnosis_plot.jpeg:</div></div><div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>bash: ./peer_output/example_data.bed_diagnosis_plot.jpeg: Permission denied
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>cat ./output/MWE.Cov_PEER.PEER.cov.stdout
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>PEER: loading input data... PEER: loading covariate file ...done!
PEER: Input summary: 200 samples with 100 genes and 2 covariates 
Setting initialization parameters ...Done.
PEER: estimating hidden confounders (30)
	iteration 0/1000
	iteration 1/1000
	iteration 2/1000
	iteration 3/1000
	iteration 4/1000
	iteration 5/1000
Converged (var(residuals)) after 5 iterations

Exporting results ... Done.
</pre></div>
</div>
</div>
</div>
</section>
<section id="b-principal-components">
<h4>b. Principal Components<a class="headerlink" href="#b-principal-components" title="Link to this heading">#</a></h4>
<p>This step will compute residual on merged covariates (<code class="docutils literal notranslate"><span class="pre">PC_1</span></code>) and perform hidden factor analysis  (<code class="docutils literal notranslate"><span class="pre">PC_2</span></code>)</p>
<p>Timing: &lt;1 min</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run covariate_hidden_factor.ipynb PCA \
    --cwd ../../../output_test/machenko_pcs \
    --phenoFile ../../../output_test/impute/protocol_example.protein.bed.imputed.bed.gz  \
    --covFile ../../../mwe_data/protocol_data/output/covariate/protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.gz \
    --container oras://ghcr.io/statfungen/pcatools_apptainer:latest \
    -c ../../csg.yml -q neurology

</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>INFO: Running <span class=" -Color -Color-Green">computing residual on merged covariates</span>: 
INFO: t32689fab04863f6d <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 6050530 (&quot;job_t32689fab04863f6d&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">computing residual on merged covariates</span> output:   <span class=" -Color -Color-Green">../../../output_test/machenko_pcs/protocol_example.protein.bed.imputed.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.residual.bed.gz</span>
INFO: Running <span class=" -Color -Color-Green">Marchenko_PC_2</span>: 
INFO: t6e1a089c0a5a9fbb <span class=" -Color -Color-Green">submitted</span> to neurology with job id Your job 6050542 (&quot;job_t6e1a089c0a5a9fbb&quot;) has been submitted
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: Waiting for the completion of <span class=" -Color -Color-Green">1</span> task.
INFO: <span class=" -Color -Color-Green">Marchenko_PC_2</span> output:   <span class=" -Color -Color-Green">../../../output_test/machenko_pcs/protocol_example.protein.bed.imputed.protocol_example.samples.protocol_example.genotype.chr21_22.pQTL.plink_qc.prune.pca.Marchenko_PC.gz</span>
INFO: Workflow Marchenko_PC (ID=w5cd1816110d97780) is executed successfully with 2 completed steps and 2 completed tasks.
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>
<section id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Link to this heading">#</a></h2>
<div class="pst-scrollable-table-container"><table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Substep</p></th>
<th class="head"><p>Problem</p></th>
<th class="head"><p>Possible Reason</p></th>
<th class="head"><p>Solution</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</div>
</section>
<section id="command-interface">
<h2>Command Interface<a class="headerlink" href="#command-interface" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>!sos run covariate_hidden_factor.ipynb -h
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run covariate_hidden_factor.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  Marchenko_PC
  PEER
  BiCV

Global Workflow Options:
  --cwd output (as path)
                        The output directory for generated files. MUST BE FULL
                        PATH
  --covFile VAL (as path, required)
                        Merged Covariates File
  --phenoFile VAL (as path, required)
                        Path to the input molecular phenotype data.
  --name  f&#39;{phenoFile:bnn}.{covFile:bn}&#39;

  --job-size 1 (as int)
                        For cluster jobs, number commands to run per job
  --walltime 5h
                        Wall clock time expected
  --mem 16G
                        Memory expected
  --numThreads 8 (as int)
                        Number of threads
  --container &#39;&#39;
                        Software container option
  --entrypoint  (&#39;micromamba run -a &quot;&quot; -n&#39; + &#39; &#39; + re.sub(r&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;, &#39;&#39;, container.split(&#39;/&#39;)[-1])) if container else &quot;&quot;


Sections
  *_1:
    Workflow Options:
      --[no-]mean-impute-missing (default to False)
  Marchenko_PC_2:
  PEER_2:
    Workflow Options:
      --N 0 (as int)
                        N PEER factors, If do not specify or specified as 0,
                        default values suggested by GTEx (based on different
                        sample size) will be used
      --iteration 1000 (as int)
                        Default values from PEER software: The number of max
                        iteration
      --tol 0.001 (as float)
                        Prior parameters parameter: Alpha_a = 0.001 parameter:
                        Alpha_b = 0.1 parameter: Eps_a = 0.1 parameter: Eps_b =
                        10.0 Tolarance parameters
      --[no-]r2-tol (default to False)
                        parameter: var_tol = 0.00001 minimum variance explained
                        criteria to drop factors while training
      --convergence-mode fast
                        Convergence mode: Convergence mode for MOFAr &quot;slow&quot;,
                        &quot;medium&quot; or &quot;fast&quot;, corresponding to 1e-5%, 1e-4% or
                        1e-3% deltaELBO change.
  PEER_3:
  BiCV_2:
  BiCV_3:
    Workflow Options:
      --N 0 (as int)
                        N factors, if not specify, calculated based on sample
                        size according to GTeX
      --iteration 10 (as int)
                        The number of iteration
</pre></div>
</div>
</div>
</div>
</section>
<section id="setup-and-global-parameters">
<h2>Setup and global parameters<a class="headerlink" href="#setup-and-global-parameters" title="Link to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[global]
# The output directory for generated files. MUST BE FULL PATH
parameter: cwd = path(&quot;output&quot;)
# Merged Covariates File
parameter: covFile = path
# Path to the input molecular phenotype data.
parameter: phenoFile = path
parameter: name = f&#39;{phenoFile:bnn}.{covFile:bn}&#39;
# For cluster jobs, number commands to run per job
parameter: job_size = 1
# Wall clock time expected
parameter: walltime = &quot;5h&quot;
# Memory expected
parameter: mem = &quot;16G&quot;
# Number of threads
parameter: numThreads = 8
# Software container option
parameter: container = &quot;&quot;
parameter: entrypoint= &quot;&quot;
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[*_1(computing residual on merged covariates)]
parameter: mean_impute_missing = False
input: phenoFile, covFile
output: f&#39;{cwd}/{name}.residual.bed.gz&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime,  mem = mem, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &quot;${ }&quot;, stderr = f&#39;{_output:nn}.stderr&#39;, stdout = f&#39;{_output:nn}.stdout&#39; , container = container, entrypoint = entrypoint

    mean_impute &lt;- function(d){
      f &lt;- apply(d, 2, function(x) mean(x, na.rm = TRUE))
      for (i in 1:length(f)) d[,i][which(is.na(d[,i]))] &lt;- f[i]
      return(d)
    }
  
    library(dplyr)
    library(readr)
    pheno = read_delim(${_input[0]:r},delim = &quot;\t&quot;)
    covariate= read_delim(${_input[1]:r},delim = &quot;\t&quot;) 

    # Extract samples in both files (Outliers will be removed in here as they are no longer presented in the header of factor)
    extraction_sample_list &lt;- intersect(colnames(pheno), colnames(covariate)) 
    
    
    if(length(extraction_sample_list) == 0){
      stop(&quot;No samples are overlapped in two files!&quot;)
    }
    
    ## Report sample counts:
    print(paste((ncol(pheno) - 4), &quot;samples are in the phenotype file&quot;, sep = &quot; &quot;))
    print(paste((ncol(covariate) - 1), &quot;samples are in the covariate file&quot;, sep = &quot; &quot;))

    ## Report identical samples:
    print(paste(length(extraction_sample_list), &quot;samples overlap between phenotype &amp; covariate files and are included in the analysis:&quot;, sep = &quot; &quot;))
    print(extraction_sample_list)

    ## Report non-overlapping samples :
    covariate_missing = covariate %&gt;% select(-all_of(&#39;#id&#39;)) %&gt;% select(-all_of(extraction_sample_list))
    print(paste(ncol(covariate_missing), &quot;samples in the covariate file are missing from the phenotype file:&quot;, sep = &quot; &quot;))
    print(colnames(covariate_missing))

    pheno_missing = pheno %&gt;% select(-c(1:4)) %&gt;% select(-all_of(extraction_sample_list))
    print(paste(ncol(pheno_missing), &quot;samples in the phenotype file are missing from the covariate file:&quot;, sep = &quot; &quot;))
    print(colnames(pheno_missing))

    # Subset the data:
    covariate = covariate[,extraction_sample_list]%&gt;%as.matrix()%&gt;%t()
    pheno_id = pheno%&gt;%select(1:4)
    pheno = pheno%&gt;%select(all_of(rownames(covariate)))%&gt;%as.matrix()%&gt;%t()
    if (${&quot;T&quot; if mean_impute_missing else &quot;F&quot;}) {
      pheno = mean_impute(pheno)
    } else {
      if(sum(is.na(pheno)) &gt; 0){ stop(&quot;NA in phenotype input is not allowed!&quot;) }
    }
    # Get residual 
    pheno_resid = .lm.fit(x = cbind(1,covariate), y = pheno)$residuals
    pheno_output = cbind(pheno_id, pheno_resid%&gt;%t())
    pheno_output%&gt;%write_delim(&quot;${_output:n}&quot;,delim = &quot;\t&quot;)
  
bash: expand = &quot;${ }&quot;, stderr = f&#39;{_output:nn}.stderr&#39;, stdout = f&#39;{_output:nn}.stdout&#39;, container = container, entrypoint = entrypoint
    bgzip -f ${_output:n}
    tabix -p bed ${_output}

    stdout=${_output:n}.stdout 
    for i in ${_output} ; do 
    echo &quot;output_info: $i &quot; &gt;&gt; $stdout;
    echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
    echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
    echo &quot;output_column:&quot; `zcat $i | head -1 | wc -w `   &gt;&gt; $stdout;
    echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
    zcat $i | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
<section id="principal-components-analysis-on-molecular-phenotype-matrix">
<h3>Principal Components Analysis on molecular phenotype matrix<a class="headerlink" href="#principal-components-analysis-on-molecular-phenotype-matrix" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[Marchenko_PC_2, PCA_2]
# Marchenko or Buja_Eyuboglu
parameter: choose_k_method = &quot;Marchenko&quot;
output: f&#39;{cwd}/{_input:bnnn}.{choose_k_method}_PC.gz&#39;
task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, entrypoint = entrypoint
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;)
    library(&quot;PCAtools&quot;)
    library(&#39;BiocSingular&#39;)
    residExpPath  = &quot;${_input}&quot;
    covPath = &quot;${covFile}&quot;
    residExpDF &lt;- read_delim(residExpPath, show_col_types=FALSE)
    covDF &lt;- read_delim(covPath, show_col_types=FALSE)
    commonMPSamples &lt;- intersect(colnames(covDF), colnames(residExpDF))
    covDFcommon &lt;- cbind(covDF[, 1], covDF[, commonMPSamples])
  
    residExpPC &lt;- pca(
        residExpDF[,commonMPSamples], # The first four columns are: chr, start, end, and gene_id; so we skip those.
        scale = TRUE,
        center = TRUE,
        BSPARAM = ExactParam())

    # Choose the number of PC
    MPPCNum &lt;- 0
    if (${&quot;T&quot; if choose_k_method == &quot;Marchenko&quot; else &quot;F&quot;}) {
        M &lt;- apply(residExpDF[, commonMPSamples], 1, function(X){ (X - mean(X))/sqrt(var(X))});
        residSigma2 &lt;- var(as.vector(M));
        paste(&#39;sigma2:&#39;, residSigma2)
        MPPCNum &lt;- chooseMarchenkoPastur(
        .dim = dim(residExpDF[, commonMPSamples]), var.explained=residExpPC$sdev^2, noise=residSigma2)
    }
    if (${&quot;T&quot; if choose_k_method == &quot;Buja_Eyuboglu&quot; else &quot;F&quot;}) {
        MPPCNum &lt;- jackstraw::permutationPA(data.matrix(residExpDF[, commonMPSamples]), B = 100, threshold = 0.05, verbose = F)$r
    }
    if (MPPCNum == 0) stop(&quot;Invalid choice of methods to choose K for PCA: ${choose_k_method}&quot;)

    MPPCsDF &lt;- as.data.frame(residExpPC$rotated[, 1:MPPCNum])
    MPColMatrix &lt;-  matrix(c(rep(&#39;Hidden_Factor_PC&#39;, times=MPPCNum), seq(1, MPPCNum)), ncol=2, nrow=MPPCNum)
    colnames(MPPCsDF) &lt;- apply(MPColMatrix, 1, function(X){return(paste0(X[1], X[2]))})
    # Add #id Column
    MPPCsDF &lt;- as.data.frame(t(MPPCsDF))
    MPPCsDF$id &lt;- rownames(MPPCsDF)
    MPPCsDF &lt;- MPPCsDF %&gt;% select(id, everything()) %&gt;% rename(&quot;#id&quot; = &quot;id&quot;)
    write_delim((rbind(covDFcommon, MPPCsDF)), &quot;${_output}&quot;, &quot;\t&quot;)

bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container, entrypoint = entrypoint
        stdout=$[_output:n].stdout
        for i in $[_output] ; do 
        echo &quot;output_info $i &quot; &gt;&gt; $stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $stdout;
        echo &quot;output_column:&quot; `zcat $i | head -1 | wc -w `   &gt;&gt; $stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $stdout;
        zcat $i | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $stdout ; done
</pre></div>
</div>
</div>
</div>
</section>
<section id="peer-method">
<h3>PEER Method<a class="headerlink" href="#peer-method" title="Link to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[PEER_2]
# N PEER factors, If do not specify or specified as 0, default values suggested by 
# GTEx (based on different sample size) will be used
parameter: N = 0
# Default values from PEER software:
## The number of max iteration
parameter: iteration = 1000
### Prior parameters
#parameter: Alpha_a = 0.001
#parameter: Alpha_b = 0.1
#parameter: Eps_a = 0.1
#parameter: Eps_b = 10.0
# Tolarance parameters
parameter: tol = 0.001
#parameter: var_tol = 0.00001
# minimum variance explained criteria to drop factors while training
parameter: r2_tol = False
# Convergence mode: Convergence mode for MOFAr &quot;slow&quot;, &quot;medium&quot; or &quot;fast&quot;, corresponding to 1e-5%, 1e-4% or 1e-3% deltaELBO change.
parameter: convergence_mode = &quot;fast&quot;
# input is the residual file from the first step
output: f&#39;{cwd:a}/{_input:bnn}.PEER_MODEL.hd5&#39;
task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
python: container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;, entrypoint = entrypoint
    from mofapy2.run.entry_point import entry_point
    import pandas as pd
    import os
    ## Force the number of threads used in numpy.
    os.environ[&quot;OMP_NUM_THREADS&quot;] = &quot;${numThreads}&quot; # export OMP_NUM_THREADS=4
    os.environ[&quot;OPENBLAS_NUM_THREADS&quot;] = &quot;${numThreads}&quot; # export OPENBLAS_NUM_THREADS=4 
    os.environ[&quot;MKL_NUM_THREADS&quot;] = &quot;${numThreads}&quot; # export MKL_NUM_THREADS=6
    import numpy as np
    ## Data
    data = pd.read_csv(${_input:r},&quot;\t&quot;,index_col = 3).drop([&quot;#chr&quot;,&quot;start&quot;,&quot;end&quot;],axis = 1)
    # initialise the entry point
    ent = entry_point()
    # Guess number of factor
    # Suggest the number of factors to use if no input value
    num_factor = ${N}
    if num_factor == 0:
        if len(data.columns) &lt; 150:
            num_factor = 15
        elif len(data.columns) &lt; 250:
            num_factor = 30
        elif len(data.columns) &lt; 350:
            num_factor = 45
        else:
            num_factor = 60
    # Set data
    # MOFA is a multi-view and multi-group inference framework. 
    # If usig only a single view and a single group (as in PEER), the data needs to be embedded into a nested list
    ent.set_data_matrix([[data.transpose()]],samples_names=[data.columns.values.tolist()], features_names=[data.index.values.tolist()])
    ent.set_model_options(factors= num_factor , spikeslab_weights=False, ard_weights=False)
    ent.set_train_options(iter=${iteration}, convergence_mode=&quot;${convergence_mode}&quot; ${f&#39;,dropR2 = {r2_tol}&#39; if r2_tol else &quot;&quot; }, startELBO=1, freqELBO=1,tolerance=${tol}, gpu_mode=False, verbose=True, seed=42)
    ent.build()
    ent.run()
    ent.save(${_output:ar})
    ## To fix issue https://github.com/statfungen/xqtl-protocol/issues/420
    import h5py
    right_name = [x.encode(&quot;UTF-8&quot;) for x in  ent.data_opts[&quot;features_names&quot;][0]] 
    new_hd5 = h5py.File(${_output:ar}, &quot;r+&quot;)
    del new_hd5[&quot;features/view0&quot;]
    new_hd5[&quot;features&quot;].create_dataset(&quot;view0&quot;, data=np.array(right_name))
    new_hd5.close()
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[PEER_3]
output: f&#39;{cwd:a}/{_input:bnn}.PEER.gz&#39;, f&#39;{cwd:a}/{_input:bnn}.PEER.diag.pdf&#39;
task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
R:  container=container, expand= &quot;${ }&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, entrypoint = entrypoint
    ####################
    ## Load libraries ##
    ####################
    
    library(MOFA2)
    library(dplyr)
    library(readr)
    ################
    ## Load model ##
    ################
    
    file &lt;- &quot;${_input}&quot;
    model &lt;- load_model(file)
    
    ####################################
    ## (Optional) add sample metadata ##
    ####################################
    
    # Important: 
    # (1) the row names in the sample metadata data.frame have to match the sample names in the MOFA model
    # (2) the sample name has to contain at least two columns: 
    # 		- sample: sample names
    #		- group: group name. Only required for multi-group inference. In your case just set the entire column to the same value
    
    # Load metadata
    
    # stopifnot(all(samples(model)[[1]]==metadata$sample))
    # samples_metadata(model) &lt;- metadata
    
    ###############################
    ## (Optional) Subset factors ##
    ###############################
    
    # We can remove factors that explain little variance (in this case, we require at least 0.01%)
    # r2 &lt;- model@cache$variance_explained$r2_per_factor
    # factors &lt;- sapply(r2, function(x) x[,1]&gt;0.0001)
    # model &lt;- subset_factors(model, which(apply(factors,1,sum) &gt;= 1))
    
    # fetch factors in matrix format, with dimensions (nfactors,nsamples)
    factors &lt;- get_factors(model)[[1]]
    dim(factors)
    factors = as_tibble(t(factors),rownames = &quot;#id&quot;)
    covPath = &quot;${covFile}&quot;
    options(readr.show_col_types = FALSE)
    covDF &lt;- read_delim(covPath, delim = &#39;\t&#39;)
    commonMPSamples &lt;- intersect(colnames(covDF), colnames(factors))      
    (rbind(covDF[, commonMPSamples], factors[, commonMPSamples])%&gt;%write_delim(&quot;${_output[0]}&quot;,&quot;\t&quot;))
    # Plot save to vcf
    pdf(&quot;${_output[1]}&quot;)
    #############################
    ## Plot variance explained ##
    #############################
    
    # Plot variance explained using individual factors
    plot_variance_explained(model, factors=&quot;all&quot;)
    plot_variance_explained(model, factors=c(1,2,3))
    
    # Plot total variance explained using all factors
    plot_variance_explained(model, plot_total = TRUE)[[2]]
    
    ########################
    ## Plot factor values ##
    ########################
    
    plot_factor(model, factor = 1)
    plot_factors(model, factor = c(1,2))
    
    ###########################
    ## Plot feature loadings ##
    ###########################
    
    # The weights or loadings provide a score for each gene on each factor. 
    # Genes with no association with the factor are expected to have values close to zero
    # Genes with strong association with the factor are expected to have large absolute values. 
    # The sign of the loading indicates the direction of the effect: a positive loading indicates that the feature is more active in the cells with positive factor values, and viceversa.
    
    # Plot the distribution of loadings for Factor 1.
    plot_weights(model,
      view = 1,
      factor = 1,
      nfeatures = 10,     # Top number of features to highlight
      scale = T           # Scale loadings from -1 to 1
    )
    
    # If you are not interested in the full distribution, but just on the top loadings:
    plot_top_weights(model, 
      view = 1, 
      factor = 1, 
      nfeatures = 10,
      scale = T
    )
    
    ######################################
    ## Plot correlation between factors ##
    ######################################
    
    plot_factor_cor(model)
    
    ###################
    ## Fetch factors ##
    ###################
    invisible(dev.off())
bash: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]:n}.stderr&#39;, stdout = f&#39;{_output[0]:n}.stdout&#39;, container = container, entrypoint = entrypoint
        for i in $[_output[0]] ; do 
        echo &quot;output_info $i &quot; &gt;&gt; $[_output[0]:n].stdout;
        echo &quot;output_size:&quot; `ls -lh $i | cut -f 5  -d  &quot; &quot;`   &gt;&gt; $[_output[0]:n].stdout;
        echo &quot;output_rows:&quot; `zcat $i | wc -l  | cut -f 1 -d &quot; &quot;`   &gt;&gt; $[_output[0]:n].stdout;
        echo &quot;output_column:&quot; `zcat $i | head -1 | wc -w `   &gt;&gt; $[_output[0]:n].stdout;
        echo &quot;output_preview:&quot;   &gt;&gt; $[_output[0]:n].stdout;
        zcat $i | head  | cut -f 1,2,3,4,5,6   &gt;&gt; $[_output[0]:n].stdout ; done
</pre></div>
</div>
</div>
</div>
<section id="reference">
<h4>Reference<a class="headerlink" href="#reference" title="Link to this heading">#</a></h4>
<ul class="simple">
<li><p>PEER code is adapted from <a class="reference external" href="https://github.com/RTIInternational/biocloud_docker_tools/blob/master/peer/v1.3/run_peer.R">here</a></p></li>
<li><p>GTEx recommandation of PEER factors is <a class="reference external" href="https://gtexportal.org/home/documentationPage">here</a></p></li>
<li><p>Examples by PEER is at <a class="reference external" href="https://github.com/PMBio/peer/wiki/Tutorial">github</a></p></li>
</ul>
</section>
</section>
<section id="id1">
<h3>BiCV Factor Analysis with APEX<a class="headerlink" href="#id1" title="Link to this heading">#</a></h3>
<p>Factor anallysis using Bi-Cross validation with the APEX software package [<a class="reference external" href="https://doi.org/10.1214/15-STS539">cf. Owen et al., Statistical Science, 2016</a>] [<a class="reference external" href="https://doi.org/10.1101/2020.12.18.423490">cf. Quick et al., bioRxiv, 2020</a>]</p>
<p>Notice that the command options are different from those on the APEX website documentation. The commands on the documentation page does not work (last updated September 2021). The commands below were constructed and tested by our team based on our understanding of the program, without input from APEX authors.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[BiCV_2]
# For cluster jobs, number commands to run per job
import time
output: f&#39;{cwd:a}/{_input:bn}.fake.vcf.gz&#39;
#task: trunk_workers = 1,trunk_size = job_size , walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: container=container, expand= &quot;$[ ]&quot;, stderr = f&#39;{_output}.stderr&#39;, stdout = f&#39;{_output}.stdout&#39;, entrypoint = entrypoint
    library(&quot;dplyr&quot;)
    library(&quot;readr&quot;)
    ## Add fake header
    cat(paste(&quot;##fileformat=VCFv4.2\n##fileDate=$[time.strftime(&quot;%Y%m%d&quot;,time.localtime())]\n##source=FAKE\n&quot;), file=$[_output:nr], append=FALSE)
    ## Add colnames based on bed
    pheno = read_delim(&quot;$[_input]&quot;, delim = &quot;\t&quot;,n_max = 1)
    colnames(pheno)[1:3] = c(&quot;#CHROM&quot;,&quot;POS&quot;,&quot;ID&quot;) 
    pheno = cbind(pheno[1:3]%&gt;%mutate(REF = &quot;A&quot;, ALT = &quot;C&quot;, QUAL = &quot;.&quot;,FILTER = &quot;.&quot;, INFO = &quot;PR&quot;, FORMAT = &quot;GT&quot;),pheno[,5:ncol(pheno)])
    pheno%&gt;%write_delim($[_output:nr],delim = &quot;\t&quot;,col_names = T, append = T)
bash: container=container, expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;, entrypoint = entrypoint
    bgzip -f $[_output:n]
    tabix -p vcf $[_output]
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[BiCV_3]
# N factors, if not specify, calculated based on sample size according to GTeX
parameter: N = 0
# The number of iteration
parameter: iteration = 10

import pandas as pd
data = pd.read_csv(phenoFile,&quot;\t&quot;,index_col = 3).drop([&quot;#chr&quot;,&quot;start&quot;,&quot;end&quot;],axis = 1)
if N == 0:
    if len(data.columns) &lt; 150:
        N = 15
    elif len(data.columns) &lt; 250:
        N = 30
    elif len(data.columns) &lt; 350:
        N = 45
    else:
        N = 60
input: output_from(1), output_from(2)
output: f&#39;{cwd:a}/{_input[0]:bnn}.BiCV.gz&#39;
task: trunk_workers = 1, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output[0]:bn}&#39;
bash: container=container, expand= &quot;$[ ]&quot;, stderr = f&#39;{_output[0]}.stderr&#39;, stdout = f&#39;{_output[0]}.stdout&#39;, entrypoint = entrypoint
    apex factor \
        --out $[_output[0]:nn] \
        --iter $[iteration] \
        --factors $[N] \
        --bed $[_input[0]] \
        --vcf $[_input[1]] \
        --threads $[numThreads]  $[ f&#39;--cov {covFile}&#39; if covFile.is_file() else &#39;&#39;]
</pre></div>
</div>
</div>
</div>
</section>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/data_preprocessing/covariate"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer d-print-none">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="covariate_formatting.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Covariate Data Formatting</p>
      </div>
    </a>
    <a class="right-next"
       href="../../association_scan/qtl_association_testing.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">QTL Association Analysis</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">


  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#default-parameters-number-of-factors">Default Parameters: Number of Factors</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#peer">PEER</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#bicv-factor-analysis-with-apex">BiCV factor analysis with APEX</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-with-marchenko-s-approach">PCA with Marchenko’s approach</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#pca-with-buja-a-and-eyuboglu-n-permutation">PCA with Buja A and Eyuboglu N. permutation</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#primary-output">Primary Output</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#secondary-outputs">Secondary Outputs</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-generate-hidden-factors">ii. Generate Hidden Factors</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#a-peer-method">a. PEER Method</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#b-principal-components">b. Principal Components</a></li>
</ul>
</li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command Interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#setup-and-global-parameters">Setup and global parameters</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#principal-components-analysis-on-molecular-phenotype-matrix">Principal Components Analysis on molecular phenotype matrix</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#peer-method">PEER Method</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#reference">Reference</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#id1">BiCV Factor Analysis with APEX</a></li>
</ul>
</li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=dfe6caa3a7d634c4db9b"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=dfe6caa3a7d634c4db9b"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>