

<!DOCTYPE html>


<html lang="en" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>eQTL summary statistics formatting &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../../_static/styles/theme.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
<link href="../../../_static/styles/bootstrap.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
<link href="../../../_static/styles/pydata-sphinx-theme.css?digest=12da95d707ffb74b382d" rel="stylesheet" />

  
  <link href="../../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=12da95d707ffb74b382d" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css" />
    <link rel="stylesheet" href="../../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../../_static/scripts/bootstrap.js?digest=12da95d707ffb74b382d" />
<link rel="preload" as="script" href="../../../_static/scripts/pydata-sphinx-theme.js?digest=12da95d707ffb74b382d" />

    <script data-url_root="../../../" id="documentation_options" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <script>window.MathJax = {"options": {"processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
    <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/multivariate/MASH/fastqtl_to_mash';</script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />
    <link rel="next" title="A multivariate EBNM approach for mixture multivariate distribution estimate" href="mixture_prior.html" />
    <link rel="prev" title="Multivariate association with MASH" href="../MASH_pipeline.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">
  

<a class="navbar-brand logo" href="../../../README.html">
  
  
  
  
    
    
      
    
    
    <img src="../../../_static/xqtl_wf.png" class="logo__image only-light" alt="Logo image"/>
    <script>document.write(`<img src="../../../_static/xqtl_wf.png" class="logo__image only-dark" alt="Logo image"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../../README.html">
                    ADSP FunGen-xQTL computational protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../commands_generator/eQTL_analysis_commands.html">Bulk RNA-seq xQTL analysis</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../data_preprocessing/reference_data.html">Reference data standardization</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/VCF_QC.html">Genotype VCF file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK file quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/PCA.html">Principal component analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/GRM.html">Genomic relationship matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/genotype/genotype_formatting.html">Genotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/gene_annotation.html">Gene coordinate annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype data formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../data_preprocessing/covariate_preprocessing.html">Covariate data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/PEER_factor.html">Factor analysis using PEER</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/BiCV_factor.html">Factor analysis using Bi-Cross validation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../data_preprocessing/covariate/covariate_formatting.html">Covariate data formatting</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../association_scan/cisQTL_scan.html">cisQTL analysis workflows</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL association testing</a></li>

<li class="toctree-l2"><a class="reference internal" href="../../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../association_scan/transQTL_scan.html">transQTL analysis workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 current active has-children"><a class="reference internal" href="../MASH_pipeline.html">Multivariate association with MASH</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">eQTL summary statistics formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="mashr.html">MASH analysis pipeline with data-driven prior matrices</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Statistical Fine-mapping</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../fine_mapping/univariate_fine_mapping.html">Univariate fine-mapping workflows</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE_RSS.html">Fine-mapping with SuSiE RSS model</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../fine_mapping/SuSiE/SuSiE.html">SuSiE fine-mapping for individual level data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../fine_mapping/multivariate_fine_mapping.html">Multivariate fine-mapping workflows</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multi-omics integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/colocalization.html">QTL and GWAS Colocalization</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/fastenloc.html">Colocalization with FastENLOC</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/COLOC/mvenloc.html">Multivariate SuSiE and ENLOC model</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../../integrative_analysis/functional_annotation.html">Functional annotation integration</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/gwas_enrichment.html">Enrichment analysis: TORUS and GREGOR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/enrichment/ldsc.html">Stratified LD Score Regression</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../integrative_analysis/SuSiE_Ann/polyfun.html">Fine-mapping with PolyFun</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Utilities</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../../misc/summary_stats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../misc/rds_to_vcf.html">Summary statistics in VCF format</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/multivariate/MASH/fastqtl_to_mash.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/multivariate/MASH/fastqtl_to_mash.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../../_sources/code/multivariate/MASH/fastqtl_to_mash.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>


<script>
document.write(`
  <button class="theme-switch-button btn btn-sm btn-outline-primary navbar-btn rounded-circle" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch" data-mode="light"><i class="fa-solid fa-sun"></i></span>
    <span class="theme-switch" data-mode="dark"><i class="fa-solid fa-moon"></i></span>
    <span class="theme-switch" data-mode="auto"><i class="fa-solid fa-circle-half-stroke"></i></span>
  </button>
`);
</script>

<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>eQTL summary statistics formatting</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-data">Input data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-list-of-summary-statistics">A list of summary statistics</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-list-of-gene-names-optional">A list of gene names (optional)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#run-analysis">Run analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#note-on-job-option">Note on <code class="docutils literal notranslate"><span class="pre">JOB_OPTION</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#more-options">More options</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output-data">Output data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pipeline-details">Pipeline details</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hdf5-utilities">HDF5 utilities</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#convert-to-hdf5-format">Convert to HDF5 format</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#per-study-tissue-conversion">Per study (tissue) conversion</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#merge-per-study-tissue-hdf5-to-one-hdf5">Merge per study (tissue) HDF5 to one HDF5</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extract-data-for-mash-model">Extract data for MASH model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#compute-z-score-and-save-to-rds">Compute Z score and save to RDS</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#export-workflow-documentation-to-html-format">Export workflow documentation to HTML format</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="eqtl-summary-statistics-formatting">
<h1>eQTL summary statistics formatting<a class="headerlink" href="#eqtl-summary-statistics-formatting" title="Permalink to this heading">#</a></h1>
<p>This workflow converts <code class="docutils literal notranslate"><span class="pre">fastqtl</span></code> eQTL analysis summary statistics text output to formats more friendly to R analysis. In particular:</p>
<ol class="arabic simple">
<li><p>It converts single study results to HDF5 format grouped by genes.</p></li>
<li><p>It combines multiple studies into one single HDF5 file. In the context of GTEx each study is result from one tissue.</p></li>
<li><p>For MASH analysis in particular, it extracts from the complete data a subset of results to compute data driven MASH prior covariance, and to fit the MASH mixture model.</p></li>
</ol>
<section id="input-data">
<h2>Input data<a class="headerlink" href="#input-data" title="Permalink to this heading">#</a></h2>
<section id="a-list-of-summary-statistics">
<h3>A list of summary statistics<a class="headerlink" href="#a-list-of-summary-statistics" title="Permalink to this heading">#</a></h3>
<p>Summary statistics from <code class="docutils literal notranslate"><span class="pre">fasteqtl</span></code>-like output are in text format, one row per gene-snp pair. Columns in our example data-set are:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">gene_id</span> 
<span class="n">variant_id</span>      
<span class="n">tss_distance</span>    
<span class="n">ma_samples</span>      
<span class="n">ma_count</span>        
<span class="n">maf</span>     
<span class="n">pval_nominal</span>    
<span class="n">slope</span>   
<span class="n">slope_se</span>
</pre></div>
</div>
<p>This is format of GTEx version 8 summary statistics data. Each analysis (“tissue” for GTEx) has a separate text file. Additionally there are support files of gene transcription start site coordinates, and SNP coordinates.</p>
<p>The workflow takes a list of summary statistics file names, eg, <code class="docutils literal notranslate"><span class="pre">data/fastqtl/FastQTLSumStats.list</span></code> (can be configured) that has the contents:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">Tissue_1</span><span class="o">.</span><span class="n">fastqtl</span><span class="o">.</span><span class="n">gz</span>
<span class="n">Tissue_2</span><span class="o">.</span><span class="n">fastqtl</span><span class="o">.</span><span class="n">gz</span>
<span class="o">...</span>
</pre></div>
</div>
<p>The first two columns of these files have to be <code class="docutils literal notranslate"><span class="pre">gene_id</span></code> and <code class="docutils literal notranslate"><span class="pre">variant_id</span></code> (column name does not matter). For other contents we only need columns <span class="math notranslate nohighlight">\(\hat{\beta}\)</span>, <span class="math notranslate nohighlight">\(\text{SE}(\hat{\beta})\)</span> and p-value for the summary statistics. In the <code class="docutils literal notranslate"><span class="pre">fastqtl</span></code> output format above it is columns <code class="docutils literal notranslate"><span class="pre">8,</span> <span class="pre">9,</span> <span class="pre">7</span></code>. If your summary statistics file has a different format you can use <code class="docutils literal notranslate"><span class="pre">--cols</span></code> parameter to pass the proper column numbers. There are two ways to do it:</p>
<ol class="arabic simple">
<li><p>Specify 3 numbers for  <span class="math notranslate nohighlight">\(\hat{\beta}\)</span>, <span class="math notranslate nohighlight">\(\text{SE}(\hat{\beta})\)</span> and p-value</p></li>
<li><p>Specify 2 numbers for  <span class="math notranslate nohighlight">\(\hat{\beta}\)</span> and p-value. <span class="math notranslate nohighlight">\(\text{SE}(\hat{\beta})\)</span> will then be computed on the fly assuming that the test is two-sided test and that <span class="math notranslate nohighlight">\(\hat{\beta}/\text{SE}(\hat{\beta})\)</span> is standard normal.</p></li>
</ol>
<p>Current default is <code class="docutils literal notranslate"><span class="pre">--cols</span> <span class="pre">8</span> <span class="pre">9</span> <span class="pre">7</span></code>. <strong>Currently this pipeline only supports and requires summary statistics <span class="math notranslate nohighlight">\(\hat{\beta}\)</span>, <span class="math notranslate nohighlight">\(\text{SE}(\hat{\beta})\)</span> and p-value, not other quantities (eg t statistic)</strong>.</p>
</section>
<section id="a-list-of-gene-names-optional">
<h3>A list of gene names (optional)<a class="headerlink" href="#a-list-of-gene-names-optional" title="Permalink to this heading">#</a></h3>
<p>To speed up merging multiple HDF5 files it helps to provide a list of gene names. Otherwise it takes too much time to figure them out from individual HDF5 files before merger can happen. The gene list has the contents like:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">ENSG00000186092</span><span class="mf">.4</span>
<span class="n">ENSG00000227232</span><span class="mf">.5</span>
<span class="n">ENSG00000228463</span><span class="mf">.9</span>
<span class="n">ENSG00000241860</span><span class="mf">.6</span>
<span class="n">ENSG00000268903</span><span class="mf">.1</span>
<span class="n">ENSG00000269981</span><span class="mf">.1</span>
<span class="n">ENSG00000279457</span><span class="mf">.4</span>
<span class="n">ENSG00000279928</span><span class="mf">.2</span>
<span class="o">...</span>
</pre></div>
</div>
</section>
</section>
<section id="run-analysis">
<h2>Run analysis<a class="headerlink" href="#run-analysis" title="Permalink to this heading">#</a></h2>
<p>Under the same folder as this list file, you keep all these listed data files. Then you run:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>JOB_OPTION=&quot;-j 8&quot;
#JOB_OPTION=&quot;-q midway2 -c midway2.yml&quot;
sos run workflows/fastqtl_to_mash.ipynb convert \
    --data-list data/fastqtl/FastQTLSumStats.list \
    --gene-list data/fastqtl/GTEx_genes.txt \
    $JOB_OPTION
</pre></div>
</div>
<p>to convert to HDF5 only, and</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>sos run workflows/fastqtl_to_mash.ipynb \
    --data-list data/fastqtl/FastQTLSumStats.list \
    --gene-list data/fastqtl/GTEx_genes.txt \
    $JOB_OPTION
</pre></div>
</div>
<p>to convert to HDF5 AND extract MASH input.</p>
<section id="note-on-job-option">
<h3>Note on <code class="docutils literal notranslate"><span class="pre">JOB_OPTION</span></code><a class="headerlink" href="#note-on-job-option" title="Permalink to this heading">#</a></h3>
<p>With bash variable <code class="docutils literal notranslate"><span class="pre">JOB_OPTION</span></code> set to <code class="docutils literal notranslate"><span class="pre">-q</span> <span class="pre">midway2</span> <span class="pre">-c</span> <span class="pre">midway2.yml</span></code>, each workflow is configured to run on UChicago RCC’s midway2 cluster, via <code class="docutils literal notranslate"><span class="pre">task</span></code> option.
When executed as is, it takes 33hrs to convert the summary statistics for GTEx data (Release V8).
You might need to configure <code class="docutils literal notranslate"><span class="pre">task</span></code> differently for your computing environment if you are familiar with <code class="docutils literal notranslate"><span class="pre">SoS</span></code>.
Otherwise, you can use the <code class="docutils literal notranslate"><span class="pre">JOB_OPTION</span></code> <code class="docutils literal notranslate"><span class="pre">-j</span> <span class="pre">8</span></code> to use 8 CPU threads.</p>
</section>
</section>
<section id="more-options">
<h2>More options<a class="headerlink" href="#more-options" title="Permalink to this heading">#</a></h2>
<div class="cell tag_report_cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">fastqtl_to_mash</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
<div class="cell_output docutils container">
<div class="output stream highlight-myst-ansi notranslate"><div class="highlight"><pre><span></span>usage: sos run fastqtl_to_mash.ipynb
               [workflow_name | -t targets] [options] [workflow_options]
  workflow_name:        Single or combined workflows defined in this script
  targets:              One or more targets to generate
  options:              Single-hyphen sos parameters (see &quot;sos run -h&quot; for details)
  workflow_options:     Double-hyphen workflow-specific parameters

Workflows:
  convert
  default
  document_it

Global Workflow Options:
  --cwd fastqtl_to_mash_output (as path)
                        Work directory / output directory
  --data-list data/eQTL_summary_files.txt (as path)
                        A text file containing data-set names
  --gene-list . (as path)
                        Optionally, a list of gene names.
  --msg &#39;eQTL mapping summary statistics&#39;
                        Meta-info tag for HDF5 database
  --maxsize 1000 (as int)
                        maximum number of groups per HDF5 file
  --cols 8 9 7 (as list)
                        1-based indexing of 3 numbers indicating columns for
                        betahat, se(betahat) and p-value, or 2 numbesr for
                        betahat and p-value
  --keep-ensg-version 0 (as int)
  --common-suffix &#39;&#39;
                        for GTEx data, parameter: common_suffix =
                        &quot;.allpairs.txt&quot;

Sections
  convert_0, default_0: Generate utility functions
  convert_1, default_1: Convert summary stats gzip format to HDF5
  convert_2, default_2: Merge single study data to multivariate data
  default_3:            Extract data to fit MASH model
    Workflow Options:
      --random-per-gene 9 (as int)
                        Number of random SNPs to draw per gene for fitting MASH
                        mixture Use -1 to export all
      --best-per-gene 1 (as int)
                        Use 1 to extract strongest gene-snp pair Use 0 to avoid
                        extracting the info
  default_4:            Subset and split data, generate Z-score and save to RDS
    Workflow Options:
      --random-snp-size 0.5 (as float)
                        Size of mash random SNPs. This specifies the size of
                        `random` SNP set and the rest of random SNPs go into
                        `random_test` SNP set. if input is float it means half
                        random SNPs the input can also be a integer to use
                        specified number.
      --effects-list NULL (as path)
                        A list of effect names (SNP names) to include in mash
                        analysis.
      --conditions-list NULL (as path)
                        A list of condition names (tissue names) to include in
                        mash analysis
  document_it:          Export workflow to HTML document
</pre></div>
</div>
</div>
</div>
</section>
<section id="output-data">
<h2>Output data<a class="headerlink" href="#output-data" title="Permalink to this heading">#</a></h2>
<p>If you run the entire workflow, you should find under <code class="docutils literal notranslate"><span class="pre">./fastqtl_to_mash_output</span></code> (can be configured):</p>
<ul class="simple">
<li><p>Study (tissue) specific HDF5 files of summary statistics</p></li>
<li><p>Merged HDF5 from multiple studies</p></li>
<li><p>“*.portable.h5” data extracted for MASH computations</p></li>
<li><p>“*.mash.rds” data in RDS format splitted to <code class="docutils literal notranslate"><span class="pre">strong</span></code>/<code class="docutils literal notranslate"><span class="pre">random</span></code> and optionally <code class="docutils literal notranslate"><span class="pre">random_test</span></code>, with Z-scores computed from p-values.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Work directory / output directory</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;./fastqtl_to_mash_output&#39;</span><span class="p">)</span>
<span class="c1"># A text file containing data-set names</span>
<span class="kn">parameter:</span> <span class="n">data_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;data/eQTL_summary_files.txt&quot;</span><span class="p">)</span>
<span class="c1"># Optionally, a list of gene names.</span>
<span class="kn">parameter:</span> <span class="n">gene_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Meta-info tag for HDF5 database</span>
<span class="kn">parameter:</span> <span class="n">msg</span> <span class="o">=</span> <span class="s2">&quot;eQTL mapping summary statistics&quot;</span>
<span class="c1"># maximum number of groups per HDF5 file</span>
<span class="kn">parameter:</span> <span class="n">maxsize</span> <span class="o">=</span> <span class="mi">1000</span>
<span class="c1"># 1-based indexing of 3 numbers indicating columns for betahat, se(betahat) and p-value, </span>
<span class="c1"># or 2 numbesr for betahat and p-value</span>
<span class="kn">parameter:</span> <span class="n">cols</span> <span class="o">=</span> <span class="p">[</span><span class="mi">8</span><span class="p">,</span> <span class="mi">9</span><span class="p">,</span> <span class="mi">7</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">keep_ensg_version</span> <span class="o">=</span> <span class="mi">0</span>
<span class="c1"># for GTEx data, parameter: common_suffix = &quot;.allpairs.txt&quot;</span>
<span class="kn">parameter:</span> <span class="n">common_suffix</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">cols</span><span class="p">)</span> <span class="o">==</span> <span class="mi">2</span><span class="p">:</span>
    <span class="n">cols</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="mi">0</span><span class="p">)</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="pipeline-details">
<h2>Pipeline details<a class="headerlink" href="#pipeline-details" title="Permalink to this heading">#</a></h2>
<section id="hdf5-utilities">
<h3>HDF5 utilities<a class="headerlink" href="#hdf5-utilities" title="Permalink to this heading">#</a></h3>
<p>The HDF5 Python interface <code class="docutils literal notranslate"><span class="pre">pytables</span></code> is used to program the data format conversion workhorse. You can use command, eg, <code class="docutils literal notranslate"><span class="pre">pip</span> <span class="pre">install</span> <span class="pre">tables</span></code> to install <code class="docutils literal notranslate"><span class="pre">pytables</span></code> package.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Generate utility functions
[utils: provides = &#39;.sos/utils.py&#39;]
depends: Py_Module(&#39;tables&#39;), Py_Module(&#39;pandas&gt;0.23.0&#39;)
output: &#39;.sos/utils.py&#39;
report: expand = &quot;${ }&quot;, output = _output
    import sys, os, re, copy
    import numpy as np, pandas as pd, tables as tb
    tb.parameters.MAX_GROUP_WIDTH = 51200
    # tb.parameters.NODE_CACHE_SLOTS = -51200
    # tb.parameters.METADATA_CACHE_SIZE = 1048576 * 100000
    # tb.parameters.CHUNK_CACHE_SIZE = 2097152 * 100000
    # tb.parameters.CHUNK_CACHE_NELMTS = 521
    import scipy.stats as st

    class Environment:
        def __init__(self):
            self.float = np.float32
            self.duplicate_tag = &#39;_duplicated_&#39;
            self.common_suffix = &#39;${common_suffix}.h5&#39;

    env = Environment()

    class TBData(dict):
        def __init__(self, data, name, msg = None, root = &#39;/&#39;, complib = &#39;bzip2&#39;):
            &#39;&#39;&#39;bzip2 may not be compatible with other hdf5 applications; but zlib is fine&#39;&#39;&#39;
            self.__root = root.strip(&#39;/&#39;)
            self.__group = name
            self.__msg = msg
            try:
                if type(data) is dict:
                    self.update(data)
                elif type(data) is str:
                    # is file name
                    self.__load(tb.open_file(data))
                else:
                    # is file stream
                    self.__load(data)
            except tb.exceptions.NoSuchNodeError:
                raise ValueError(&#39;Cannot find dataset {}!&#39;.format(name))
            self.tb_filters = tb.Filters(complevel = 9, complib=complib)

        def sink(self, filename):
            with tb.open_file(filename, &#39;a&#39;) as f:
                if self.__root:
                    try:
                        f.create_group(&quot;/&quot;, self.__root)
                    except:
                        pass
                try:
                    # there is existing data -- have to merge with current data
                    # have to do this because the input file lines are not grouped by gene names!!
                    # use try ... except to hopefully faster than if ... else
                    # e.g., if not f.__contains__(&#39;/{}&#39;.format(self.__group)) ... else ...
                    for element in f.list_nodes(&#39;/{}/{}&#39;.format(self.__root, self.__group)):
                        if element.name != &#39;colnames&#39;:
                            self[element.name] = np.concatenate((element[:], self[element.name]))
                except tb.exceptions.NoSuchNodeError:
                    f.create_group(&quot;/&quot; + self.__root, self.__group,
                                   self.__msg if self.__msg else self.__group)
                for key in self:
                    self.__store_array(key, f)
                f.flush()

        def dump(self, table, output = False):
            if output:
                pd.DataFrame({self[&#39;colnames&#39;][i] : self[table][:,i] for i in range(len(self[&#39;colnames&#39;]))}, index = self[&#39;rownames&#39;]).to_csv(sys.stdout, na_rep = &#39;NA&#39;)
                return None
            else:
                return pd.DataFrame({self[&#39;colnames&#39;][i] : self[table][:,i] for i in range(len(self[&#39;colnames&#39;]))}, index = self[&#39;rownames&#39;])

        def __load(self, fstream):
            try:
                for element in fstream.list_nodes(&#39;/{}/{}&#39;.format(self.__root, self.__group)):
                    self[element.name] = element[:]
                fstream.close()
            except:
                fstream.close()
                raise

        def __roll_back(self, group, name):
            try:
                n = getattr(group, name)
                n._f_remove()
            except AttributeError:
                pass

        def __store_array(self, name, fstream):
            if self.__root:
                element = getattr(getattr(fstream.root, self.__root), self.__group)
            else:
                element = getattr(fstream.root, self.__group)
            arr = self[name]
            if type(arr) is list:
                arr = np.array(arr)
            self.__roll_back(element, name)
            #
            if arr.shape != (0,):
                ds = fstream.create_carray(element, name, tb.Atom.from_dtype(arr.dtype), arr.shape,
                                           filters = self.tb_filters)
                ds[:] = arr

    def get_tb_grps(filenames, group_name = None):
        if len(filenames) == 0:
            return []
        if isinstance(filenames, str):
            filenames = [filenames]
        names = set()
        for filename in filenames:
            with tb.open_file(filename) as f:
                names.update([node._v_name for node in (f.root if group_name is None else getattr(f.root, &#39;{}&#39;.format(group_name)))])
        return sorted(names)

    def get_se(bhat, pval):
        if bhat &lt; 0:
            z = st.norm.ppf(pval / 2)
        else:
            z = st.norm.ppf(1 - pval / 2)
        if z != 0:
            se = bhat/z
        else:
            se = np.nan
        return se

    class SSData:
        def __init__(self, header = False):
            self.data = {&#39;buffer&#39;:{&#39;data&#39;:[], &#39;rownames&#39;:[]}, &#39;output&#39;:{}}
            self.header = header
            self.previous_name = self.current_name = None
            self.min_line_len = max([${cols[0]}, ${cols[1]}, ${cols[2]}])

        def parse(self, line, ensg_version = 0, include = set()):
            # input line is snp, gene, beta, t, pval and optionally se(bhat)
            if not line:
                self.__reset()
                self.current_name = None
                return 1
            if isinstance(line, bytes):
                line = line.decode()
            line = line.strip().split()
            if self.header:
                # header line has to be accounted for
                self.header = False
                return 0
            #
            line[0] = line[0].strip(&#39;&quot;&#39;)
            if not line[0] in include and len(include)&gt;0:
                return -1
            # the line is not long enough ... there is format issues
            if len(line) &lt; self.min_line_len:
                return -1
            #
            if ensg_version == 0:
                line[0] = line[0].rsplit(&#39;.&#39;,1)[0]
            if self.previous_name is None:
                self.previous_name = line[0]
            self.current_name = line[0]
            if self.current_name != self.previous_name:
                self.__reset()
            if ${cols[2]} == 0:
                # the two numbers are bhat and p-value, we need to compute se(bhat)
                self.data[&#39;buffer&#39;][&#39;data&#39;].append([line[${cols[0]-1}], str(get_se(float(line[${cols[0]-1}]), float(line[${cols[1]-1}]))), line[${cols[1]-1}]])
            else:
                self.data[&#39;buffer&#39;][&#39;data&#39;].append([line[${cols[0]-1}], line[${cols[1]-1}], line[${cols[2]-1}]])
            self.data[&#39;buffer&#39;][&#39;rownames&#39;].append(self.__format_variant_id(line[1]))
            return 0
        
        @staticmethod
        def __format_variant_id(value):
            value = value.strip(&#39;&quot;&#39;).lstrip(&#39;chr&#39;).split(&#39;_&#39;)
            if len(value) &gt; 4:
                # keep it chr, pos, ref, alt
                value = value[:4]
            return &#39;_&#39;.join(value)

        def __reset(self):
            self.data[&#39;buffer&#39;][&#39;data&#39;] = np.array(self.data[&#39;buffer&#39;][&#39;data&#39;], dtype = env.float)
            self.data[&#39;buffer&#39;][&#39;rownames&#39;] = np.array(self.data[&#39;buffer&#39;][&#39;rownames&#39;])
            self.data[&#39;buffer&#39;][&#39;colnames&#39;] = np.array([&#39;beta&#39;,&#39;se&#39;,&#39;pval&#39;])
            self.data[&#39;output&#39;] = copy.deepcopy(self.data[&#39;buffer&#39;])
            self.data[&#39;buffer&#39;] = {&#39;data&#39;:[], &#39;rownames&#39;:[]}

        def dump(self):
            return self.data[&#39;output&#39;]

    class DataMerger(TBData):
        def __init__(self, files, name, msg = None):
            TBData.__init__(self, {}, name, msg, complib = &quot;zlib&quot;)
            self.files = sorted(files)
            self.__group = name

        def merge(self):
            data = {}
            one_snp = None
            failure_ct = 0
            # Collect data
            for item in self.files:
                tissue = re.sub(r&#39;{}$&#39;.format(env.common_suffix), &#39;&#39;, os.path.basename(item))
                try:
                    data[tissue] = TBData(item, self.__group)
                    if one_snp is None: one_snp = data[tissue][&#39;rownames&#39;][0]
                except ValueError:
                    data[tissue] = {&#39;data&#39; : np.array([[np.nan, np.nan, np.nan]]), &#39;rownames&#39;: None}
                    failure_ct += 1
                # Fix row name
                # Because in GTEx data file there are duplicated gene-snp pairs having different sumstats!!
                if data[tissue][&#39;rownames&#39;] is not None:
                    data[tissue][&#39;rownames&#39;] = self.__dedup(data[tissue][&#39;rownames&#39;], item)
            if failure_ct == len(self.files):
                return 1
            # Merge data
            for idx, item in enumerate([&#39;beta&#39;,&#39;se&#39;,&#39;pval&#39;]):
                self[item] = pd.concat([pd.DataFrame(
                    {tissue : data[tissue][&#39;data&#39;][:,idx]},
                    index = data[tissue][&#39;rownames&#39;] if data[tissue][&#39;rownames&#39;] is not None else [one_snp]
                    ) for tissue in sorted(data.keys())], sort=True, axis = 1)
                if &#39;rownames&#39; not in self:
                    self[&#39;rownames&#39;] = np.array(self[item].index, dtype = str)
                if &#39;colnames&#39; not in self:
                    self[&#39;colnames&#39;] = np.array(self[item].columns.values.tolist(), dtype = str)
                self[item] = np.array(self[item].values, dtype = env.float)
            # np.savetxt(sys.stdout, self[&#39;pval&#39;], fmt=&#39;%10.5f&#39;)
            # print(self[&#39;rownames&#39;])
            # print(self[&#39;colnames&#39;])
            return 0

        def __dedup(self, seq, filename):
            seen = {}
            dups = set()
            def __is_seen(x, seen):
                if x not in seen:
                    seen[x] = 0
                    return 0
                else:
                    seen[x] += 1
                    dups.add(x)
                    return 1
            # Tag them
            obs = [x if not __is_seen(x, seen) else &#39;%s%s%s&#39; % (x, env.duplicate_tag, seen[x]) for x in seq]
            # Log them
            if len(dups):
                filename = os.path.splitext(filename)[0]
                with open(filename + &#39;.error&#39;, &#39;a&#39;) as f:
                    for item in dups:
                        f.write(&#39;{}:{} appeared {} times in {}\n&#39;.\
                                format(self.__group, item, seen[item] + 1, filename))
            return obs

    def get_gs_pairs(data, name, num = (1, 9), method = &#39;equal_space&#39;):
        &#39;&#39;&#39;choose gene-snp pairs from data, controlled by num = (a, b)
        for the best gene-snp pair (a = 0 or 1), and b other random 
        gene-snp pairs&#39;&#39;&#39;

        def random_sample(x, k):
            if k &lt; 0:
                return sorted(x)
            return sorted(set(np.random.choice(x, min(len(x), k))))

        def equal_sample(x, k):
            if len(x) &lt; k or k &lt; 0:
                return sorted(x)
            f = lambda m, n: [i*n//m + n//(2*m) for i in range(m)]
            return sorted(set([x[i] for i in f(k, len(x))]))

        output = {&#39;colnames&#39; : data[&#39;colnames&#39;]}
        lp = data.dump(&#39;pval&#39;)
        shat = data.dump(&#39;se&#39;)
        # pval cannot be nan or zero (in fastqtl output, Shat nan has pval zero)
        lp = lp[np.all(np.isfinite(lp), axis=1) &amp; np.all(np.isfinite(shat), axis=1)]
        #
        if lp.empty and num[0] &gt; 0:
            return None
        # Find strong SNP-gene pair
        lp = -np.log10(lp)
        rowidx = np.where(data[&#39;rownames&#39;] == lp.max(axis=1).idxmax())[0][0]
        if num[0] != 0:
            output[&#39;strong_rownames&#39;] = [&#39;%s_%s&#39; % (name, data[&#39;rownames&#39;][rowidx].decode())]
            output[&#39;strong&#39;] = {}
            for k in [&#39;beta&#39;, &#39;pval&#39;, &#39;se&#39;]:
                output[&#39;strong&#39;][k] = data[k][rowidx, :]
        if num[1] != 0:
            all_randomidxes = [y for y, x in enumerate(data[&#39;rownames&#39;]) if x in lp.index and y != rowidx]
            sample_randomidxes = random_sample(all_randomidxes, num[1]) if method == &#39;random&#39; else equal_sample(all_randomidxes, num[1])
            output[&#39;random_rownames&#39;] = [&#39;%s_%s&#39; % (name, data[&#39;rownames&#39;][x].decode()) for x in sample_randomidxes]
            output[&#39;random&#39;] = {}
            for k in [&#39;beta&#39;, &#39;pval&#39;, &#39;se&#39;]:
                output[&#39;random&#39;][k] = data[k][sample_randomidxes, :]
        if not &#39;strong&#39; in output and not &#39;random&#39; in output:
            output = None
        return output

    def merge_tmp_h5(output, verbose = 0):
        from glob import glob
        tmpfiles = list(glob(output + &quot;_*.tmp&quot;))
        if os.path.isfile(output+&#39;.h5&#39;):
            os.remove(output+&#39;.h5&#39;)
        for item in sorted(tmpfiles):
            for name in get_tb_grps(item):
                cmd = &#39;h5copy -i {0} -o {2} -s &quot;/{1}&quot; -d &quot;/{1}&quot;&#39;.format(item, name, output+&#39;.h5&#39;)
                if verbose:
                    print(cmd)
                os.system(cmd)
</pre></div>
</div>
</div>
</div>
</section>
<section id="convert-to-hdf5-format">
<h3>Convert to HDF5 format<a class="headerlink" href="#convert-to-hdf5-format" title="Permalink to this heading">#</a></h3>
<section id="per-study-tissue-conversion">
<h4>Per study (tissue) conversion<a class="headerlink" href="#per-study-tissue-conversion" title="Permalink to this heading">#</a></h4>
<p>For per study conversion I use <code class="docutils literal notranslate"><span class="pre">bzip2</span></code> compression method and <code class="docutils literal notranslate"><span class="pre">float32</span></code> to achieve higher compression rate. This workflow step will generate one HDF5 file per summary statistics file. <code class="docutils literal notranslate"><span class="pre">h5copy</span></code> program can be installed via <code class="docutils literal notranslate"><span class="pre">conda</span> <span class="pre">install</span> <span class="pre">-c</span> <span class="pre">conda-forge</span> <span class="pre">hdf5</span></code>.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Convert summary stats gzip format to HDF5
[convert_1, default_1]
depends: &#39;.sos/utils.py&#39;, executable(&quot;h5copy&quot;)
fail_if(not data_list.is_file(), msg = &#39;Need data list file!&#39;)
data_files = set(get_output(f&quot;awk &#39;{{print $1}}&#39; {data_list:e}&quot;).strip().split(&#39;\n&#39;))
fail_if(len(data_files) == 0, msg = &#39;Need input data files!&#39;)
import os
input: [f&#39;{data_list:d}/{x}&#39; if os.path.isfile(f&#39;{data_list:d}/{x}&#39;) else x for x in data_files], group_by = 1
output: f&#39;{cwd:a}/{_input:bn}.h5&#39;
task: trunk_workers = 1, walltime = &#39;36h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;

bash: expand = True, workdir = cwd
    rm -f {_output:n}_*.tmp
    
python: expand = &quot;${ }&quot;, input = _depends[0]
    import warnings
    import gzip
    gene_names = set([x.strip() for x in open(${gene_list:r}).readlines() if x.strip()] if ${gene_list.is_file()} else [])
    ssp = SSData(header = True)
    group_counts = 0
    with gzip.open(${_input:r}) as f:
        while True:
            line = f.readline()
            quit = ssp.parse(line, ${keep_ensg_version}, include = gene_names)
            if quit == -1:
                continue
            if ssp.current_name != ssp.previous_name:
                group_counts += 1
                with warnings.catch_warnings():
                    warnings.filterwarnings(&quot;ignore&quot;, category = tb.FlavorWarning)
                    data = TBData(ssp.dump(), ssp.previous_name, &quot;${msg}&quot;)
                    data.sink(&quot;${_output:n}_%i.tmp&quot; % (np.ceil(group_counts / ${maxsize})) 
                                if ${maxsize} &gt; 0 else ${_output:r})
                ssp.previous_name = ssp.current_name
            if quit:
                break
    if ${maxsize} &gt; 0:
        merge_tmp_h5(${_output:nr})

bash: expand = True, workdir = cwd
    rm -f {_output:n}_*.tmp
</pre></div>
</div>
</div>
</div>
</section>
<section id="merge-per-study-tissue-hdf5-to-one-hdf5">
<h4>Merge per study (tissue) HDF5 to one HDF5<a class="headerlink" href="#merge-per-study-tissue-hdf5-to-one-hdf5" title="Permalink to this heading">#</a></h4>
<p>This step creates tables for each summary statistic per gene from multiple studies. Rows are effects (SNPs), columns are conditions (tissue names). This time I use <code class="docutils literal notranslate"><span class="pre">zlib</span></code> compression for better compatibility with other HDF5 routines (eg <code class="docutils literal notranslate"><span class="pre">rhdf5</span></code>).</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Merge single study data to multivariate data
[convert_2, default_2]
depends: &#39;.sos/utils.py&#39;, executable(&quot;h5copy&quot;)
input: group_by = &#39;all&#39;
output: f&#39;{cwd:a}/{data_list:bn}.h5&#39;
task: trunk_workers = 1, walltime = &#39;36h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;

bash: expand = True, workdir = cwd
    rm -f {_output:n}_*.tmp

python: expand = &#39;${ }&#39;, input = _depends[0]
    import warnings
    if ${gene_list.is_file()}:
        gene_names = [x.strip() for x in open(${gene_list:r}).readlines() if x.strip()]
    else:
        gene_names = get_tb_grps([${_input:r,}])
    if ${keep_ensg_version} == 0:
        gene_names = [os.path.splitext(x)[0] for x in gene_names]
    failure_ct = 0
    for idx, item in enumerate(gene_names):
        ssm = DataMerger([${_input:r,}], item, &quot;${msg}&quot;)
        with warnings.catch_warnings():
            warnings.filterwarnings(&quot;ignore&quot;, category = tb.FlavorWarning)
            if ssm.merge() == 0:
                ssm.sink(&quot;${_output:n}_%i.tmp&quot; % (np.ceil((idx + 1.0) / ${maxsize})) 
                        if ${maxsize} &gt; 0 else ${_output:r})
            else:
                failure_ct += 1
    with open(&quot;${_output:n}.log&quot;, &#39;w&#39;) as f:
        f.write(&quot;%s out of %s groups merged!\n&quot; % (len(gene_names) - failure_ct, len(gene_names)))
    if ${maxsize} &gt; 0:
        merge_tmp_h5(${_output:nr})
            
bash: expand = True, workdir = cwd
    rm -f {_output:n}_*.tmp
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="extract-data-for-mash-model">
<h3>Extract data for MASH model<a class="headerlink" href="#extract-data-for-mash-model" title="Permalink to this heading">#</a></h3>
<p>We need to extract the “top” signals based on single tissue analysis to compute MASH priors, as well as some random SNP sets to fit MASH mixture model.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Extract data to fit MASH model
[default_3]
depends: &#39;.sos/utils.py&#39;
# Number of random SNPs to draw per gene for fitting MASH mixture
# Use -1 to export all
parameter: random_per_gene = 9
# Use 1 to extract strongest gene-snp pair
# Use 0 to avoid extracting the info
parameter: best_per_gene = 1
output: f&quot;{_input:n}.portable.h5&quot;
task: trunk_workers = 1, walltime = &#39;36h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;

python: expand = &quot;${ }&quot;, input = _depends[0]
    import warnings
    if ${gene_list.is_file()}:
        gene_names = [x.strip() for x in open(${gene_list:r}).readlines() if x.strip()]
    else:
        gene_names = get_tb_grps([${_input:r,}])
    if ${keep_ensg_version} == 0:
        gene_names = [os.path.splitext(x)[0] for x in gene_names]
    output = dict()
    output[&#39;random&#39;] = {&#39;colnames&#39;: None, &#39;rownames&#39;: [], &#39;beta&#39;: None, &#39;se&#39;: None, &#39;pval&#39;: None}
    output[&#39;strong&#39;] = {&#39;colnames&#39;: None, &#39;rownames&#39;: [], &#39;beta&#39;: None, &#39;se&#39;: None, &#39;pval&#39;: None}
    failure_ct = 0
    for idx, name in enumerate(gene_names):
        # extract the best gene-snp pair or some random gene-snp pairs

        try:
            data = TBData(${_input:r}, name)
            res = get_gs_pairs(data, name, (${best_per_gene}, ${random_per_gene}))
        except ValueError:
            res = None
        #
        if res is None:
            failure_ct += 1
            continue
        for k in output:
            if not k in res:
                continue
            for kk in output[k]:
                if kk == &#39;rownames&#39;:
                    if output[k][&#39;rownames&#39;] is None:
                        output[k][&#39;rownames&#39;] = res[&quot;{}_rownames&quot;.format(k)]
                    else:
                        output[k][&#39;rownames&#39;].extend(res[&quot;{}_rownames&quot;.format(k)])
                elif kk == &#39;colnames&#39;:
                    if output[k][&#39;colnames&#39;] is None:
                        output[k][&#39;colnames&#39;] = res[&#39;colnames&#39;]
                else:
                    output[k][kk] = np.vstack((output[k][kk], res[k][kk])) if output[k][kk] is not None else res[k][kk]
    #
    if failure_ct &lt; len(gene_names):
        with warnings.catch_warnings():
            warnings.filterwarnings(&quot;ignore&quot;, category = tb.FlavorWarning)
            for k in output:
                TBData(dict(output[k]), k, msg = &quot;%s, %s gene-snp pair&quot; % (&quot;${msg}&quot;, k), complib = &#39;zlib&#39;).sink(${_output:r})
    with open(&quot;${_output:n}.log&quot;, &#39;w&#39;) as f:
        f.write(&quot;%s out of %s groups extracted!\n&quot; % (len(gene_names) - failure_ct, len(gene_names)))
</pre></div>
</div>
</div>
</div>
</section>
<section id="compute-z-score-and-save-to-rds">
<h3>Compute Z score and save to RDS<a class="headerlink" href="#compute-z-score-and-save-to-rds" title="Permalink to this heading">#</a></h3>
<p>Here I also implemented two command options to subset data:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">--effects-list</span></code>: a list of effect names (row names) to include from output MASH data-set. This option was used in MASH paper to remove SNPs in LD with each other.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">--conditions-list</span></code>: a list of conditions (column names) to include in output MASH data-set. Conditions not in this list are excluded from output. This option was used in MASH paper to focus analysis on brain / no-brain tissues.</p></li>
</ul>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span># Subset and split data, generate Z-score and save to RDS
[default_4]
# Size of mash random SNPs. 
# This specifies the size of `random` SNP set 
# and the rest of random SNPs go into `random_test` SNP set.
# if input is float it means half random SNPs
# the input can also be a integer to use specified number.
parameter: random_snp_size = 0.5
# A list of effect names (SNP names) to include in mash analysis.
parameter: effects_list = path(&#39;NULL&#39;)
# A list of condition names (tissue names) to include in mash analysis
parameter: conditions_list = path(&#39;NULL&#39;)

depends: R_library(&#39;rhdf5&#39;)
output: f&quot;{_input:n}.mash.rds&quot; if not str(_input).endswith(&#39;.portable.h5&#39;) else f&quot;{_input:nn}.mash.rds&quot;

task: trunk_workers = 1, walltime = &#39;36h&#39;, trunk_size = 1, mem = &#39;4G&#39;, cores = 1, tags = f&#39;{_output:bn}&#39;

R: expand = &quot;${ }&quot;
    ConvertP2Z &lt;- function(pval, beta) {
      z &lt;- abs(qnorm(pval / 2))
      z[which(beta &lt; 0)] &lt;- -1 * z[which(beta &lt; 0)]
      return(z)
    }

    handle_nan_b = function(x) {
          x[which(is.nan(x) | is.na(x))] = 0
          return(x)
    }

    handle_nan_s = function(x) {
          x[which(is.nan(x) | is.infinite(x) | is.na(x) | x == 0)] = 1E3
          return(x)
    }
  
    handle_nan_z = function(z, b, s) {
          z[which(is.nan(s) | is.infinite(s) | is.na(s) | s == 0)] = 0
          z[which(is.nan(b) | is.na(b) | b == 0)] = 0
          z[which(is.nan(z) | is.na(z))] = 0
          return(z)
    }
  
    GetSS &lt;- function(table, db) {
      dat &lt;- rhdf5::h5read(db, table)
      dat$&quot;z&quot; &lt;- ConvertP2Z(dat$&quot;pval&quot;, dat$&quot;beta&quot;)
      for (name in c(&quot;beta&quot;, &quot;se&quot;, &quot;pval&quot;, &quot;z&quot;)) {
        dat[[name]] &lt;- as.matrix(t(dat[[name]]))
        colnames(dat[[name]]) &lt;- dat$colnames
        rownames(dat[[name]]) &lt;- dat$rownames
      }
      dat$colnames &lt;- dat$rownames &lt;- NULL
      dat[[&#39;z&#39;]] = handle_nan_z(dat[[&#39;z&#39;]], dat[[&#39;beta&#39;]], dat[[&#39;se&#39;]])
      dat[[&#39;se&#39;]] = handle_nan_s(dat[[&#39;se&#39;]])
      dat[[&#39;beta&#39;]] = handle_nan_b(dat[[&#39;beta&#39;]])
      return(dat)
    }
  
    SplitTrainTest &lt;- function(dat, table) {
        # load data
        strong = dat$strong[[table]]
        random = dat$random[[table]]
        # select rows to keep
        num_train = ${random_snp_size}
        if (num_train &lt; 1)
          num_train = nrow(random) * num_train
        num_train = as.integer(num_train)
        if (num_train &gt; nrow(random)) {
            num_train = nrow(random)
        }
        train = random[1:num_train,]
        if (num_train == nrow(random)) {
            test = NULL
        } else {
            test = random[(num_train+1):nrow(random),]
        }
        if (${effects_list:r} != &#39;NULL&#39;) {
            pout = scan(${effects_list:ar}, what=&quot;character&quot;, sep=NULL)
            strong = strong[(rownames(strong) %in% pout),]
            train = train[(rownames(train) %in% pout),]
            if (!is.null(test))
                test = test[(rownames(test) %in% pout),]                                         
        }
        if (${conditions_list:r} != &#39;NULL&#39;) {
            rout = scan(${conditions_list:ar}, what=&quot;character&quot;, sep=NULL)
            strong = strong[,(colnames(strong) %in% rout),drop=F]
            train = train[,(colnames(train) %in% rout),drop=F]
            if (!is.null(test))
                test = test[,(colnames(test) %in% rout),drop=F]
        } 

        return(list(random = train,
               random.test = test,
               strong = strong))
    }
      
    SS_data = list(strong = GetSS(&#39;strong&#39;, ${_input:r}), random = GetSS(&#39;random&#39;, ${_input:r}))
    ztable = SplitTrainTest(SS_data, &quot;z&quot;)
    btable = SplitTrainTest(SS_data, &quot;beta&quot;)
    stable = SplitTrainTest(SS_data, &quot;se&quot;)
    # save output
    saveRDS(list(random.z = ztable$random,
                 random.test.z = ztable$random.test,
                 strong.z = ztable$strong,
                 random.b = btable$random,
                 random.test.b = btable$random.test,
                 strong.b = btable$strong,
                 random.s = stable$random,
                 random.test.s = stable$random.test,
                 strong.s = stable$strong), ${_output:r})
</pre></div>
</div>
</div>
</div>
<p>The output is an R list of data matrices:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="o">&gt;</span> <span class="n">dat</span> <span class="o">=</span> <span class="n">readRDS</span><span class="p">(</span><span class="s1">&#39;FastQTLSumStats.mash.rds&#39;</span><span class="p">)</span>
<span class="o">&gt;</span> <span class="n">names</span><span class="p">(</span><span class="n">dat</span><span class="p">)</span>
<span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="s2">&quot;random.z&quot;</span>      <span class="s2">&quot;random.test.z&quot;</span> <span class="s2">&quot;strong.z&quot;</span>      <span class="s2">&quot;random.b&quot;</span>     
<span class="p">[</span><span class="mi">5</span><span class="p">]</span> <span class="s2">&quot;random.test.b&quot;</span> <span class="s2">&quot;strong.b&quot;</span>      <span class="s2">&quot;random.s&quot;</span>      <span class="s2">&quot;random.test.s&quot;</span>
<span class="p">[</span><span class="mi">9</span><span class="p">]</span> <span class="s2">&quot;strong.s&quot;</span>
</pre></div>
</div>
<p><code class="docutils literal notranslate"><span class="pre">*.z</span></code>, <code class="docutils literal notranslate"><span class="pre">*.b</span></code> and <code class="docutils literal notranslate"><span class="pre">*.s</span></code> stands for z-scores, <span class="math notranslate nohighlight">\(\hat{\beta}\)</span> and <span class="math notranslate nohighlight">\(\text{SE}(\hat{\beta})\)</span> (standard error). <code class="docutils literal notranslate"><span class="pre">strong</span></code> is data of the “strongest” gene-SNP pair; <code class="docutils literal notranslate"><span class="pre">random</span></code> is a random subset of gene-SNP pairs. They are used to find pattern of sharing in strong signals and learning mixture weights. Additionally <code class="docutils literal notranslate"><span class="pre">random.test.*</span></code> is a seperate random data-set that can be used to assess fit / overfit of the MASH mixture model.</p>
</section>
</section>
<section id="export-workflow-documentation-to-html-format">
<h2>Export workflow documentation to HTML format<a class="headerlink" href="#export-workflow-documentation-to-html-format" title="Permalink to this heading">#</a></h2>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">workflow</span><span class="o">/</span><span class="n">fastqtl_to_mash</span> <span class="n">document_it</span>
</pre></div>
</div>
<p>will export the narrative in this workflow notebook to HTML format as a documentation file.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="c1"># Export workflow to HTML document</span>
<span class="p">[</span><span class="n">document_it</span><span class="p">]</span>
<span class="kn">input:</span> <span class="p">[</span><span class="n">item</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">)</span> <span class="k">if</span> <span class="n">item</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s1">&#39;.ipynb&#39;</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">output:</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">cwd</span><span class="si">:</span><span class="s1">a</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">item</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.html&#39;</span> <span class="k">for</span> <span class="n">item</span> <span class="ow">in</span> <span class="n">paths</span><span class="p">(</span><span class="n">sys</span><span class="o">.</span><span class="n">argv</span><span class="p">)</span> <span class="k">if</span> <span class="n">item</span><span class="o">.</span><span class="n">suffix</span> <span class="o">==</span> <span class="s1">&#39;.ipynb&#39;</span><span class="p">],</span> <span class="n">group_by</span> <span class="o">=</span> <span class="mi">2</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="kc">True</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="kc">False</span>
  <span class="n">sos</span> <span class="n">convert</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span> <span class="o">--</span><span class="n">template</span> <span class="n">sos</span><span class="o">-</span><span class="n">report</span><span class="o">-</span><span class="n">toc</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/multivariate/MASH"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
                <footer class="bd-footer-article">
                  <!-- Previous / next buttons -->
<div class="prev-next-area">
    <a class="left-prev"
       href="../MASH_pipeline.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Multivariate association with MASH</p>
      </div>
    </a>
    <a class="right-next"
       href="mixture_prior.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">A multivariate EBNM approach for mixture multivariate distribution estimate</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input-data">Input data</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-list-of-summary-statistics">A list of summary statistics</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#a-list-of-gene-names-optional">A list of gene names (optional)</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#run-analysis">Run analysis</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#note-on-job-option">Note on <code class="docutils literal notranslate"><span class="pre">JOB_OPTION</span></code></a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#more-options">More options</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output-data">Output data</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#pipeline-details">Pipeline details</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#hdf5-utilities">HDF5 utilities</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#convert-to-hdf5-format">Convert to HDF5 format</a><ul class="nav section-nav flex-column">
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#per-study-tissue-conversion">Per study (tissue) conversion</a></li>
<li class="toc-h4 nav-item toc-entry"><a class="reference internal nav-link" href="#merge-per-study-tissue-hdf5-to-one-hdf5">Merge per study (tissue) HDF5 to one HDF5</a></li>
</ul>
</li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#extract-data-for-mash-model">Extract data for MASH model</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#compute-z-score-and-save-to-rds">Compute Z score and save to RDS</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#export-workflow-documentation-to-html-format">Export workflow documentation to HTML format</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            <div class="bd-footer-content__inner">
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    
  <p class="copyright">
    
      © Copyright 2021+, FunGen xQTL Methods and Data Integration Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div></div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../../_static/scripts/bootstrap.js?digest=12da95d707ffb74b382d"></script>
<script src="../../../_static/scripts/pydata-sphinx-theme.js?digest=12da95d707ffb74b382d"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>