

<!DOCTYPE html>


<html lang="en" data-content_root="" >

  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

    <title>Advanced cis-QTL analysis with individual level data &#8212; FunGen-xQTL Consortium</title>
  
  
  
  <script data-cfasync="false">
    document.documentElement.dataset.mode = localStorage.getItem("mode") || "";
    document.documentElement.dataset.theme = localStorage.getItem("theme") || "light";
  </script>
  
  <!-- Loaded before other Sphinx assets -->
  <link href="../../_static/styles/theme.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/bootstrap.css?digest=5b4479735964841361fd" rel="stylesheet" />
<link href="../../_static/styles/pydata-sphinx-theme.css?digest=5b4479735964841361fd" rel="stylesheet" />

  
  <link href="../../_static/vendor/fontawesome/6.1.2/css/all.min.css?digest=5b4479735964841361fd" rel="stylesheet" />
  <link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-solid-900.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-brands-400.woff2" />
<link rel="preload" as="font" type="font/woff2" crossorigin href="../../_static/vendor/fontawesome/6.1.2/webfonts/fa-regular-400.woff2" />

    <link rel="stylesheet" type="text/css" href="../../_static/pygments.css" />
    <link rel="stylesheet" href="../../_static/styles/sphinx-book-theme.css?digest=14f4ca6b54d191a8c7657f6c759bf11a5fb86285" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/mystnb.4510f1fc1dee50b3e5859aac5469c37c29e427902b24a333a5f9fcb2f0b3ac41.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../_static/design-style.4045f2051d55cab465a707391d5b2007.min.css" />
  
  <!-- Pre-loaded scripts that we'll load fully later -->
  <link rel="preload" as="script" href="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd" />
<link rel="preload" as="script" href="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd" />
  <script src="../../_static/vendor/fontawesome/6.1.2/js/all.min.js?digest=5b4479735964841361fd"></script>

    <script data-url_root="../../" id="documentation_options" src="../../_static/documentation_options.js"></script>
    <script src="../../_static/jquery.js"></script>
    <script src="../../_static/underscore.js"></script>
    <script src="../../_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="../../_static/doctools.js"></script>
    <script src="../../_static/clipboard.min.js"></script>
    <script src="../../_static/copybutton.js"></script>
    <script src="../../_static/scripts/sphinx-book-theme.js?digest=5a5c038af52cf7bc1a1ec88eea08e6366ee68824"></script>
    <script>let toggleHintShow = 'Click to show';</script>
    <script>let toggleHintHide = 'Click to hide';</script>
    <script>let toggleOpenOnPrint = 'true';</script>
    <script src="../../_static/togglebutton.js"></script>
    <script>var togglebuttonSelector = '.toggle, .admonition.dropdown';</script>
    <script src="../../_static/design-tabs.js"></script>
    <script>const THEBE_JS_URL = "https://unpkg.com/thebe@0.8.2/lib/index.js"
const thebe_selector = ".thebe,.cell"
const thebe_selector_input = "pre"
const thebe_selector_output = ".output, .cell_output"
</script>
    <script async="async" src="../../_static/sphinx-thebe.js"></script>
    <script>DOCUMENTATION_OPTIONS.pagename = 'code/cis_analysis/cis_workhorse';</script>
    <link rel="index" title="Index" href="../../genindex.html" />
    <link rel="search" title="Search" href="../../search.html" />
    <link rel="next" title="SuSiE fine-mapping workflow" href="SuSiE.html" />
    <link rel="prev" title="Overview of advanced cis-QTL analysis" href="cis_advanced_overview.html" />
  <meta name="viewport" content="width=device-width, initial-scale=1"/>
  <meta name="docsearch:language" content="en"/>
  </head>
  
  
  <body data-bs-spy="scroll" data-bs-target=".bd-toc-nav" data-offset="180" data-bs-root-margin="0px 0px -60%" data-default-mode="">

  
  
  <a class="skip-link" href="#main-content">Skip to main content</a>
  
  <div id="pst-scroll-pixel-helper"></div>

  
  <button type="button" class="btn rounded-pill" id="pst-back-to-top">
    <i class="fa-solid fa-arrow-up"></i>
    Back to top
  </button>

  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__primary"
          id="__primary"/>
  <label class="overlay overlay-primary" for="__primary"></label>
  
  <input type="checkbox"
          class="sidebar-toggle"
          name="__secondary"
          id="__secondary"/>
  <label class="overlay overlay-secondary" for="__secondary"></label>
  
  <div class="search-button__wrapper">
    <div class="search-button__overlay"></div>
    <div class="search-button__search-container">
<form class="bd-search d-flex align-items-center"
      action="../../search.html"
      method="get">
  <i class="fa-solid fa-magnifying-glass"></i>
  <input type="search"
         class="form-control"
         name="q"
         id="search-input"
         placeholder="Search this book..."
         aria-label="Search this book..."
         autocomplete="off"
         autocorrect="off"
         autocapitalize="off"
         spellcheck="false"/>
  <span class="search-button__kbd-shortcut"><kbd class="kbd-shortcut__modifier">Ctrl</kbd>+<kbd>K</kbd></span>
</form></div>
  </div>
  
    <nav class="bd-header navbar navbar-expand-lg bd-navbar">
    </nav>
  
  <div class="bd-container">
    <div class="bd-container__inner bd-page-width">
      
      <div class="bd-sidebar-primary bd-sidebar">
        

  
  <div class="sidebar-header-items sidebar-primary__section">
    
    
    
    
  </div>
  
    <div class="sidebar-primary-items__start sidebar-primary__section">
        <div class="sidebar-primary-item">

  

<a class="navbar-brand logo" href="../../README.html">
  
  
  
  
  
    
    
      
    
    
    <img src="../../_static/xqtl_wf.png" class="logo__image only-light" alt="FunGen-xQTL Consortium - Home"/>
    <script>document.write(`<img src="../../_static/xqtl_wf.png" class="logo__image only-dark" alt="FunGen-xQTL Consortium - Home"/>`);</script>
  
  
</a></div>
        <div class="sidebar-primary-item"><nav class="bd-links" id="bd-docs-nav" aria-label="Main">
    <div class="bd-toc-item navbar-nav active">
        
        <ul class="nav bd-sidenav bd-sidenav__home-link">
            <li class="toctree-l1">
                <a class="reference internal" href="../../README.html">
                    FunGen-xQTL Computational Protocol
                </a>
            </li>
        </ul>
        <p aria-level="2" class="caption" role="heading"><span class="caption-text">Getting started</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../xqtl_protocol_demo.html">Illustration of xQTL protocol</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Command Generator</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/bulk_expression_commands.html">RNA-seq calling and QC</a></li>
<li class="toctree-l1"><a class="reference internal" href="../commands_generator/eQTL_analysis_commands.html">Univariate xQTL Discovery</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Reference data</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../reference_data/reference_data.html">Reference Data</a><input class="toctree-checkbox" id="toctree-checkbox-1" name="toctree-checkbox-1" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-1"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/reference_data_preparation.html">Reference Data Standardization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/generalized_TADB.html">Generation of Topologically Associated Domains and their Boundaries</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_prune_reference.html">Independent list of variants using LD clumping</a></li>
<li class="toctree-l2"><a class="reference internal" href="../reference_data/ld_reference_generation.html">Generating LD Reference Panel</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Molecular Phenotypes</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/bulk_expression.html">RNA-seq expression</a><input class="toctree-checkbox" id="toctree-checkbox-2" name="toctree-checkbox-2" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-2"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/RNA_calling.html">Quantifying expression from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_QC.html">Sample level RNA-seq quality control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/bulk_expression_normalization.html">Bulk RNA-seq counts normalization</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../molecular_phenotypes/scnuc_expression.html">scRNA-seq expression calling</a></li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/apa.html">Alternative polyadenylation</a><input class="toctree-checkbox" id="toctree-checkbox-3" name="toctree-checkbox-3" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-3"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/apa_calling.html">APA Calling</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/methylation.html">Methylation</a><input class="toctree-checkbox" id="toctree-checkbox-4" name="toctree-checkbox-4" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-4"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/methylation_calling.html">Quantification of methylation data</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../molecular_phenotypes/splicing.html">Alternative splicing from RNA-seq data</a><input class="toctree-checkbox" id="toctree-checkbox-5" name="toctree-checkbox-5" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-5"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/calling/splicing_calling.html">Quantifying alternative splicing from RNA-seq data</a></li>
<li class="toctree-l2"><a class="reference internal" href="../molecular_phenotypes/QC/splicing_normalization.html">Normalization and phenotype table generation for splicingQTL analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Data Pre-processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/genotype_preprocessing.html">Genotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-6" name="toctree-checkbox-6" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-6"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/VCF_QC.html">Genotype VCF File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GWAS_QC.html">Genotype PLINK File Quality Control</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/PCA.html">Principal Component Analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/GRM.html">Genomic Relationship Matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/genotype/genotype_formatting.html">Genotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/phenotype_preprocessing.html">Phenotype data preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-7" name="toctree-checkbox-7" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-7"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/gene_annotation.html">Gene Coordinate Annotation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_imputation.html">Phenotype Data Imputation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/phenotype/phenotype_formatting.html">Phenotype Data Formatting</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../data_preprocessing/covariate_preprocessing.html">Covariate Data Preprocessing</a><input class="toctree-checkbox" id="toctree-checkbox-8" name="toctree-checkbox-8" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-8"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_formatting.html">Covariate Data Formatting</a></li>
<li class="toctree-l2"><a class="reference internal" href="../data_preprocessing/covariate/covariate_hidden_factor.html">Hidden Factor Analysis</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">QTL Association Testing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../association_scan/qtl_association_testing.html">QTL Association Analysis</a><input class="toctree-checkbox" id="toctree-checkbox-9" name="toctree-checkbox-9" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-9"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/TensorQTL/TensorQTL.html">TensorQTL QTL Association Testing</a></li>
<li class="toctree-l2"><a class="reference internal" href="../association_scan/APEX/APEX.html">APEX QTL association testing</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../association_scan/qtl_association_postprocessing.html">QTL data postprocessing</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Advanced cis-QTL Analysis</span></p>
<ul class="current nav bd-sidenav">
<li class="toctree-l1 current active has-children"><a class="reference internal" href="cis_advanced_overview.html">Overview of advanced cis-QTL analysis</a><input checked="" class="toctree-checkbox" id="toctree-checkbox-10" name="toctree-checkbox-10" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-10"><i class="fa-solid fa-chevron-down"></i></label><ul class="current">
<li class="toctree-l2 current active"><a class="current reference internal" href="#">Advanced cis-QTL analysis with individual level data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="SuSiE.html">SuSiE fine-mapping workflow</a></li>
<li class="toctree-l1"><a class="reference internal" href="fSuSiE.html">fSuSiE workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="mvSuSiE.html">Multivariate fine-mapping workflows</a></li>
<li class="toctree-l1"><a class="reference internal" href="mrmash.html">Multivariate TWAS prediction model</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Enrichment Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../enrichment/gregor.html">GREGOR enrichment analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../enrichment/ldsc.html">Stratified LD Score Regression</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">xQTL-GWAS Integration</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../pecotmr_integration/pecotmr.html">The PECOTMR intergration framework</a><input class="toctree-checkbox" id="toctree-checkbox-11" name="toctree-checkbox-11" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-11"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/SuSiE_enloc.html">GWAS integration: enrichment and colocalization</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/twas_mr.html">GWAS integration: TWAS and MR</a></li>
<li class="toctree-l2"><a class="reference internal" href="../pecotmr_integration/intact.html">TWAS and COLOC integration</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Multivariate Analysis</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/METAL_pipeline.html">Meta-analysis with METAL</a><input class="toctree-checkbox" id="toctree-checkbox-12" name="toctree-checkbox-12" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-12"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/METAL/METAL.html">Meta-analysis with METAL</a></li>
</ul>
</li>
<li class="toctree-l1 has-children"><a class="reference internal" href="../multivariate_genome/MASH_pipeline.html">Multivariate association with MASH</a><input class="toctree-checkbox" id="toctree-checkbox-13" name="toctree-checkbox-13" type="checkbox"/><label class="toctree-toggle" for="toctree-checkbox-13"><i class="fa-solid fa-chevron-down"></i></label><ul>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_preprocessing.html">Extract genome-wide data for multivariate analysis</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mixture_prior.html">A multivariate EBNM approach for mixture multivariate distribution estimate</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_fit.html">MASH analysis pipeline with data-driven prior matrices</a></li>
<li class="toctree-l2"><a class="reference internal" href="../multivariate_genome/MASH/mash_posterior.html">MASH analysis pipeline with posterior computation</a></li>
</ul>
</li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Rare xQTL</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../rare_xqtl/watershed.html">watershed analysis for rare xQTL</a></li>
</ul>
<p aria-level="2" class="caption" role="heading"><span class="caption-text">Post Processing</span></p>
<ul class="nav bd-sidenav">
<li class="toctree-l1"><a class="reference internal" href="../post_processing/sumstats_standardizer.html">Summary statistics standardization and export</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/rds_to_vcf.html">Summary statistics in VCF format</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/association_scan_post_processing.html">Association result processing</a></li>
<li class="toctree-l1"><a class="reference internal" href="../post_processing/fine_mapping_post_processing.html">SuSIE results post process</a></li>
</ul>

    </div>
</nav></div>
    </div>
  
  
  <div class="sidebar-primary-items__end sidebar-primary__section">
  </div>
  
  <div id="rtd-footer-container"></div>


      </div>
      
      <main id="main-content" class="bd-main">
        
        

<div class="sbt-scroll-pixel-helper"></div>

          <div class="bd-content">
            <div class="bd-article-container">
              
              <div class="bd-header-article">
<div class="header-article-items header-article__inner">
  
    <div class="header-article-items__start">
      
        <div class="header-article-item"><label class="sidebar-toggle primary-toggle btn btn-sm" for="__primary" title="Toggle primary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
  <span class="fa-solid fa-bars"></span>
</label></div>
      
    </div>
  
  
    <div class="header-article-items__end">
      
        <div class="header-article-item">

<div class="article-header-buttons">





<div class="dropdown dropdown-source-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Source repositories">
    <i class="fab fa-github"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/edit/gh-pages/code/cis_analysis/cis_workhorse.ipynb" target="_blank"
   class="btn btn-sm btn-source-edit-button dropdown-item"
   title="Suggest edit"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-pencil-alt"></i>
  </span>
<span class="btn__text-container">Suggest edit</span>
</a>
</li>
      
      
      
      
      <li><a href="https://github.com/cumc/xqtl-pipeline/issues/new?title=Issue%20on%20page%20%2Fcode/cis_analysis/cis_workhorse.html&body=Your%20issue%20content%20here." target="_blank"
   class="btn btn-sm btn-source-issues-button dropdown-item"
   title="Open an issue"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-lightbulb"></i>
  </span>
<span class="btn__text-container">Open issue</span>
</a>
</li>
      
  </ul>
</div>






<div class="dropdown dropdown-download-buttons">
  <button class="btn dropdown-toggle" type="button" data-bs-toggle="dropdown" aria-expanded="false" aria-label="Download this page">
    <i class="fas fa-download"></i>
  </button>
  <ul class="dropdown-menu">
      
      
      
      <li><a href="../../_sources/code/cis_analysis/cis_workhorse.ipynb" target="_blank"
   class="btn btn-sm btn-download-source-button dropdown-item"
   title="Download source file"
   data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file"></i>
  </span>
<span class="btn__text-container">.ipynb</span>
</a>
</li>
      
      
      
      
      <li>
<button onclick="window.print()"
  class="btn btn-sm btn-download-pdf-button dropdown-item"
  title="Print to PDF"
  data-bs-placement="left" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-file-pdf"></i>
  </span>
<span class="btn__text-container">.pdf</span>
</button>
</li>
      
  </ul>
</div>




<button onclick="toggleFullScreen()"
  class="btn btn-sm btn-fullscreen-button"
  title="Fullscreen mode"
  data-bs-placement="bottom" data-bs-toggle="tooltip"
>
  

<span class="btn__icon-container">
  <i class="fas fa-expand"></i>
  </span>

</button>



<script>
document.write(`
  <button class="btn btn-sm navbar-btn theme-switch-button" title="light/dark" aria-label="light/dark" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="theme-switch nav-link" data-mode="light"><i class="fa-solid fa-sun fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="dark"><i class="fa-solid fa-moon fa-lg"></i></span>
    <span class="theme-switch nav-link" data-mode="auto"><i class="fa-solid fa-circle-half-stroke fa-lg"></i></span>
  </button>
`);
</script>


<script>
document.write(`
  <button class="btn btn-sm navbar-btn search-button search-button__button" title="Search" aria-label="Search" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <i class="fa-solid fa-magnifying-glass fa-lg"></i>
  </button>
`);
</script>
<label class="sidebar-toggle secondary-toggle btn btn-sm" for="__secondary"title="Toggle secondary sidebar" data-bs-placement="bottom" data-bs-toggle="tooltip">
    <span class="fa-solid fa-list"></span>
</label>
</div></div>
      
    </div>
  
</div>
</div>
              
              

<div id="jb-print-docs-body" class="onlyprint">
    <h1>Advanced cis-QTL analysis with individual level data</h1>
    <!-- Table of contents -->
    <div id="print-main-content">
        <div id="jb-print-toc">
            
            <div>
                <h2> Contents </h2>
            </div>
            <nav aria-label="Page">
                <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genotype-data">Example genotype data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-cis-window-file">Example cis-window file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#about-indels">About indels</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-susie-with-twas-weights">i. SuSiE with TWAS weights</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-fsusie">ii. fSuSiE</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#univariate-analysis-susie-and-twas">Univariate analysis: SuSiE and TWAS</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multivariate-analysis-mvsusie-and-mr-mash">Multivariate analysis: mvSuSiE and mr.mash</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mvsusie">mvSuSiE</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mr-mash">mr.mash</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-for-trait">Functional regression fSuSiE for trait</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-with-other-modality">Functional regression fSuSiE with other modality</a></li>
</ul>
            </nav>
        </div>
    </div>
</div>

              
                
<div id="searchbox"></div>
                <article class="bd-article" role="main">
                  
  <section class="tex2jax_ignore mathjax_ignore" id="advanced-cis-qtl-analysis-with-individual-level-data">
<h1>Advanced cis-QTL analysis with individual level data<a class="headerlink" href="#advanced-cis-qtl-analysis-with-individual-level-data" title="Permalink to this heading">#</a></h1>
<section id="description">
<h2>Description<a class="headerlink" href="#description" title="Permalink to this heading">#</a></h2>
<p>This notebook performs various advanced statistical analysis on multiple xQTL in a given region. Current procedures implemented include:</p>
<ol class="arabic simple">
<li><p>Univariate analysis</p>
<ul class="simple">
<li><p>SuSiE</p></li>
<li><p>Univeriate TWAS weights: LASSO, Elastic net, mr.mash and SuSiE (optional)</p></li>
<li><p>Cross validation of TWAS methods (optional but highly recommended if TWAS weights are computed)</p></li>
</ul>
</li>
<li><p>Functional data (epigenomic xQTL) analysis</p>
<ul class="simple">
<li><p>fSuSiE</p></li>
</ul>
</li>
<li><p>Multivariate analysis</p>
<ul class="simple">
<li><p>mvSuSiE</p></li>
<li><p>mr.mash</p></li>
</ul>
</li>
</ol>
</section>
<section id="input">
<h2>Input<a class="headerlink" href="#input" title="Permalink to this heading">#</a></h2>
<ol class="arabic simple">
<li><p>A list of regions to be analyzed (optional); the last column of this file should be region name.</p></li>
<li><p>Either a list of per chromosome genotype files, or one file for genotype data of the entire genome. Genotype data has to be in PLINK <code class="docutils literal notranslate"><span class="pre">bed</span></code> format.</p></li>
<li><p>Vector of lists of phenotype files per region to be analyzed, in UCSC <code class="docutils literal notranslate"><span class="pre">bed.gz</span></code> with index in <code class="docutils literal notranslate"><span class="pre">bed.gz.tbi</span></code> formats.</p></li>
<li><p>Vector of covariate files corresponding to the lists above.</p></li>
<li><p>Customized cis windows file. If it is not provided, a fixed sized cis-window will be used.</p></li>
<li><p>Optionally a vector of names of the phenotypic conditions in the form of <code class="docutils literal notranslate"><span class="pre">cond1</span> <span class="pre">cond2</span> <span class="pre">cond3</span></code> separated with whitespace.</p></li>
</ol>
<p>Input 2 and 3 should be outputs from <code class="docutils literal notranslate"><span class="pre">genotype_per_region</span></code> and <code class="docutils literal notranslate"><span class="pre">annotate_coord</span></code> modules in previous preprocessing steps. 4 should be output of <code class="docutils literal notranslate"><span class="pre">covariate_preprocessing</span></code> pipeline that contains genotype PC, phenotypic hidden confounders and fixed covariates.</p>
<section id="example-genotype-data">
<h3>Example genotype data<a class="headerlink" href="#example-genotype-data" title="Permalink to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr        path</span>
<span class="n">chr21</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21</span><span class="o">.</span><span class="n">bed</span>
<span class="n">chr22</span> <span class="o">/</span><span class="n">mnt</span><span class="o">/</span><span class="n">mfs</span><span class="o">/</span><span class="n">statgen</span><span class="o">/</span><span class="n">xqtl_workflow_testing</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr22</span><span class="o">.</span><span class="n">bed</span>
</pre></div>
</div>
<p>Alternatively, simply use <code class="docutils literal notranslate"><span class="pre">protocol_example.genotype.chr21_22.bed</span></code> if all chromosomes are in the same file.</p>
</section>
<section id="example-phenotype-list">
<h3>Example phenotype list<a class="headerlink" href="#example-phenotype-list" title="Permalink to this heading">#</a></h3>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start   end ID  path</span>
<span class="n">chr12</span>   <span class="mi">752578</span>  <span class="mi">752579</span>  <span class="n">ENSG00000060237</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">990508</span>  <span class="mi">990509</span>  <span class="n">ENSG00000082805</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">2794969</span> <span class="mi">2794970</span> <span class="n">ENSG00000004478</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">4649113</span> <span class="mi">4649114</span> <span class="n">ENSG00000139180</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6124769</span> <span class="mi">6124770</span> <span class="n">ENSG00000110799</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
<span class="n">chr12</span>   <span class="mi">6534516</span> <span class="mi">6534517</span> <span class="n">ENSG00000111640</span>  <span class="o">/</span><span class="n">home</span><span class="o">/</span><span class="n">gw</span><span class="o">/</span><span class="n">GIT</span><span class="o">/</span><span class="n">github</span><span class="o">/</span><span class="n">fungen</span><span class="o">-</span><span class="n">xqtl</span><span class="o">-</span><span class="n">analysis</span><span class="o">/</span><span class="n">analysis</span><span class="o">/</span><span class="n">Wang_Columbia</span><span class="o">/</span><span class="n">ROSMAP</span><span class="o">/</span><span class="n">MWE</span><span class="o">/</span><span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">gz</span>
</pre></div>
</div>
</section>
<section id="example-cis-window-file">
<h3>Example cis-window file<a class="headerlink" href="#example-cis-window-file" title="Permalink to this heading">#</a></h3>
<p>It should have strictly 4 columns, with the header a commented out line:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="c1">#chr    start    end    gene_id</span>
<span class="n">chr10</span>   <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000008128</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000008130</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000067606</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">7101193</span>    <span class="n">ENSG00000069424</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">7960000</span>    <span class="n">ENSG00000069812</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000078369</span>
<span class="n">chr1</span>    <span class="mi">0</span>    <span class="mi">6480000</span>    <span class="n">ENSG00000078808</span>
</pre></div>
</div>
<p>The key is that the 4th column ID should match with the 4th column ID in the phenotype list. Otherwise the cis-window to analyze will not be found.</p>
</section>
<section id="about-indels">
<h3>About indels<a class="headerlink" href="#about-indels" title="Permalink to this heading">#</a></h3>
<p>Option <code class="docutils literal notranslate"><span class="pre">--no-indel</span></code> will remove indel from analysis. FIXME: Gao need to provide more guidelines how to deal with indels in practice.</p>
</section>
</section>
<section id="output">
<h2>Output<a class="headerlink" href="#output" title="Permalink to this heading">#</a></h2>
<p>For each analysis region, the output is SuSiE model fitted and saved in RDS format.</p>
</section>
<section id="minimal-working-example-steps">
<h2>Minimal Working Example Steps<a class="headerlink" href="#minimal-working-example-steps" title="Permalink to this heading">#</a></h2>
<section id="i-susie-with-twas-weights">
<h3>i. SuSiE with TWAS weights<a class="headerlink" href="#i-susie-with-twas-weights" title="Permalink to this heading">#</a></h3>
<p>Timing [FIXME]</p>
<p>Below we duplicate the examples for phenotype and covariates to demonstrate that when there are multiple phenotypes for the same genotype it is possible to use this pipeline to analyze all of them (more than two is accepted as well).</p>
<p>Here using <code class="docutils literal notranslate"><span class="pre">--region-name</span></code> we focus the analysis on 3 genes. In practice if this parameter is dropped, the union of all regions in all phenotype region lists will be analyzed. It is possible for some of the regions there are no genotype data, in which case the pipeline will output RDS files with a warning message to indicate the lack of genotype data to analyze.</p>
<p><strong>Note:</strong> Suggested output naming convention is cohort_modality, eg ROSMAP_snRNA_pseudobulk.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">cis_workhorse</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span>  \
    <span class="o">--</span><span class="kp">name</span> <span class="n">protocol_example_protein</span>  \
    <span class="o">--</span><span class="n">genoFile</span> <span class="nb">input</span><span class="o">/</span><span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">bed</span>   \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
                <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
              <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span>  \
    <span class="o">--</span><span class="n">customized</span><span class="o">-</span><span class="n">cis</span><span class="o">-</span><span class="n">windows</span> <span class="nb">input</span><span class="o">/</span><span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">enhanced_cis_chr21_chr22</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="kp">name</span> <span class="n">ENSG00000241973_P42356</span> <span class="n">ENSG00000160209_O00764</span> <span class="n">ENSG00000100412_Q99798</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">names</span> <span class="n">trait_A</span> <span class="n">trait_B</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">pecotmr_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
</div>
</div>
<p>It is also possible to analyze a selected list of regions using option <code class="docutils literal notranslate"><span class="pre">--region-list</span></code>. The last column of this file will be used for the list to analyze. Here for example use the same list of regions as we used for customized cis-window:</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">cis_workhorse</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span>  \
    <span class="o">--</span><span class="kp">name</span> <span class="n">protocol_example_protein</span>  \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">bed</span>   \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
                <span class="n">output</span><span class="o">/</span><span class="n">phenotype</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">region_list</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
              <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span>  \
    <span class="o">--</span><span class="n">customized</span><span class="o">-</span><span class="n">cis</span><span class="o">-</span><span class="n">windows</span> <span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">enhanced_cis_chr21_chr22</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">region</span><span class="o">-</span><span class="nb">list</span> <span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">enhanced_cis_chr21_chr22</span><span class="o">.</span><span class="n">bed</span> \
    <span class="o">--</span><span class="n">phenotype</span><span class="o">-</span><span class="n">names</span> <span class="n">trait_A</span> <span class="n">trait_B</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">pecotmr_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
</div>
</div>
<p><strong>Note:</strong> When both <code class="docutils literal notranslate"><span class="pre">--region-name</span></code> and <code class="docutils literal notranslate"><span class="pre">--region-list</span></code> are used, the union of regions from these parameters will be analyzed.</p>
<p>FIXME: We should probably just explain these parameters, will work better for conversion script</p>
<p>To perform fine-mapping only without TWAS weights,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">cis_workhorse</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span> <span class="o">--</span><span class="n">no</span><span class="o">-</span><span class="n">twas</span><span class="o">-</span><span class="n">weights</span> <span class="o">...</span> <span class="c1"># rest of parameters the same. </span>
</pre></div>
</div>
<p>To perform fine-mapping and TWAS weights without cross validation,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">cis_workhorse</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span> <span class="o">--</span><span class="n">twas</span><span class="o">-</span><span class="n">cv</span><span class="o">-</span><span class="n">folds</span> <span class="mi">0</span> <span class="o">...</span> <span class="c1"># rest of parameters the same. </span>
</pre></div>
</div>
<p>It is also possible to specify a subset of samples to analyze, using <code class="docutils literal notranslate"><span class="pre">--keep-samples</span></code> parameter. For example we create a file to keep the ID of 50 samples,</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">zcat</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">protein</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> <span class="o">|</span> <span class="n">head</span> <span class="o">-</span><span class="mi">1</span> <span class="o">|</span> <span class="n">awk</span> <span class="s1">&#39;{for (i=2; i&lt;=51; i++) printf $i &quot; &quot;; print &quot;&quot;}&#39;</span><span class="o">&gt;</span> <span class="n">output</span><span class="o">/</span><span class="n">keep_samples</span><span class="o">.</span><span class="n">txt</span>
</pre></div>
</div>
</div>
</div>
<p>then use them in our analysis,</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">xqtl</span><span class="o">-</span><span class="n">pipeline</span><span class="o">/</span><span class="n">pipeline</span><span class="o">/</span><span class="n">cis_workhorse</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">susie_twas</span> <span class="o">--</span><span class="n">keep</span><span class="o">-</span><span class="n">samples</span> <span class="n">output</span><span class="o">/</span><span class="n">keep_samples</span><span class="o">.</span><span class="n">txt</span> <span class="o">...</span> <span class="c1"># rest of parameters the same</span>
</pre></div>
</div>
</section>
<section id="ii-fsusie">
<h3>ii. fSuSiE<a class="headerlink" href="#ii-fsusie" title="Permalink to this heading">#</a></h3>
<p>Timing [FIXME]</p>
<p><strong>Note:</strong> Suggested output naming convention is cohort_modality, eg ROSMAP_snRNA_pseudobulk.</p>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">pipeline</span><span class="o">/</span><span class="n">cis_workhorse</span><span class="o">.</span><span class="n">ipynb</span> <span class="n">fsusie</span> \
    <span class="o">--</span><span class="kp">name</span> <span class="n">protocol_example_methylation</span> \
    <span class="o">--</span><span class="n">genoFile</span> <span class="n">xqtl_association</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">plink_per_chrom</span><span class="o">.</span><span class="n">txt</span> \
    <span class="o">--</span><span class="n">phenoFile</span> <span class="n">output</span><span class="o">/</span><span class="n">phenotype_by_region</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">phenotype_by_region_files</span><span class="o">.</span><span class="n">txt</span> \
                <span class="n">output</span><span class="o">/</span><span class="n">phenotype_by_region</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">bed</span><span class="o">.</span><span class="n">phenotype_by_region_files</span><span class="o">.</span><span class="n">txt</span>  \
    <span class="o">--</span><span class="n">covFile</span> <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
              <span class="n">output</span><span class="o">/</span><span class="n">covariate</span><span class="o">/</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">methylation</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">samples</span><span class="o">.</span><span class="n">protocol_example</span><span class="o">.</span><span class="n">genotype</span><span class="o">.</span><span class="n">chr21_22</span><span class="o">.</span><span class="n">pQTL</span><span class="o">.</span><span class="n">plink_qc</span><span class="o">.</span><span class="n">prune</span><span class="o">.</span><span class="n">pca</span><span class="o">.</span><span class="n">Marchenko_PC</span><span class="o">.</span><span class="n">gz</span> \
    <span class="o">--</span><span class="n">container</span> <span class="n">oras</span><span class="p">:</span><span class="o">//</span><span class="n">ghcr</span><span class="o">.</span><span class="n">io</span><span class="o">/</span><span class="n">cumc</span><span class="o">/</span><span class="n">pecotmr_apptainer</span><span class="p">:</span><span class="n">latest</span>
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="troubleshooting">
<h2>Troubleshooting<a class="headerlink" href="#troubleshooting" title="Permalink to this heading">#</a></h2>
<table class="table">
<thead>
<tr class="row-odd"><th class="head"><p>Step</p></th>
<th class="head"><p>Substep</p></th>
<th class="head"><p>Problem</p></th>
<th class="head"><p>Possible Reason</p></th>
<th class="head"><p>Solution</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
</section>
<section id="command-interface">
<h2>Command interface<a class="headerlink" href="#command-interface" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="n">sos</span> <span class="n">run</span> <span class="n">cis_workhorse</span><span class="o">.</span><span class="n">ipynb</span> <span class="o">-</span><span class="n">h</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="workflow-implementation">
<h2>Workflow implementation<a class="headerlink" href="#workflow-implementation" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s2">&quot;output&quot;</span><span class="p">)</span>
<span class="c1"># A list of file paths for genotype data, or the genotype data itself. </span>
<span class="kn">parameter:</span> <span class="n">genoFile</span> <span class="o">=</span> <span class="n">path</span>
<span class="c1"># One or multiple lists of file paths for phenotype data.</span>
<span class="kn">parameter:</span> <span class="n">phenoFile</span> <span class="o">=</span> <span class="n">paths</span>
<span class="c1"># Covariate file path</span>
<span class="kn">parameter:</span> <span class="n">covFile</span> <span class="o">=</span> <span class="n">paths</span>
<span class="c1"># Optional: if a region list is provide the analysis will be focused on provided region. </span>
<span class="c1"># The LAST column of this list will contain the ID of regions to focus on</span>
<span class="c1"># Otherwise, all regions with both genotype and phenotype files will be analyzed</span>
<span class="kn">parameter:</span> <span class="n">region_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Optional: if a region name is provided </span>
<span class="c1"># the analysis would be focused on the union of provides region list and region names</span>
<span class="kn">parameter:</span> <span class="n">region_name</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># Only focus on a subset of samples</span>
<span class="kn">parameter:</span> <span class="n">keep_samples</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># An optional list documenting the custom cis window for each region to analyze, with four column, chr, start, end, region ID (eg gene ID).</span>
<span class="c1"># If this list is not provided, the default `window` parameter (see below) will be used.</span>
<span class="kn">parameter:</span> <span class="n">customized_cis_windows</span> <span class="o">=</span> <span class="n">path</span><span class="p">()</span>
<span class="c1"># Specify the cis window for the up and downstream radius to analyze around the region of interest in units of bp</span>
<span class="c1"># When this is zero, we will rely on customized_cis_windows</span>
<span class="kn">parameter:</span> <span class="n">window</span> <span class="o">=</span> <span class="mi">0</span>
<span class="c1"># It is required to input the name of the analysis</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="nb">str</span>
<span class="kn">parameter:</span> <span class="n">container</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
<span class="kn">import</span> <span class="nn">re</span>
<span class="kn">parameter:</span> <span class="n">entrypoint</span><span class="o">=</span> <span class="p">(</span><span class="s1">&#39;micromamba run -a &quot;&quot; -n&#39;</span> <span class="o">+</span> <span class="s1">&#39; &#39;</span> <span class="o">+</span> <span class="n">re</span><span class="o">.</span><span class="n">sub</span><span class="p">(</span><span class="sa">r</span><span class="s1">&#39;(_apptainer:latest|_docker:latest|\.sif)$&#39;</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="n">container</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]))</span> <span class="k">if</span> <span class="n">container</span> <span class="k">else</span> <span class="s2">&quot;&quot;</span>
<span class="c1"># For cluster jobs, number commands to run per job</span>
<span class="kn">parameter:</span> <span class="n">job_size</span> <span class="o">=</span> <span class="mi">200</span>
<span class="c1"># Wall clock time expected</span>
<span class="kn">parameter:</span> <span class="kp">walltime</span> <span class="o">=</span> <span class="s2">&quot;100h&quot;</span>
<span class="c1"># Memory expected</span>
<span class="kn">parameter:</span> <span class="kp">mem</span> <span class="o">=</span> <span class="s2">&quot;80G&quot;</span>
<span class="c1"># Number of threads</span>
<span class="kn">parameter:</span> <span class="n">numThreads</span> <span class="o">=</span> <span class="mi">1</span>
<span class="c1"># Name of phenotypes</span>
<span class="kn">parameter:</span> <span class="n">phenotype_names</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">x</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">phenoFile</span><span class="p">]</span>
<span class="kn">parameter:</span> <span class="n">seed</span> <span class="o">=</span> <span class="mi">999</span>

<span class="k">def</span> <span class="nf">group_by_region</span><span class="p">(</span><span class="n">lst</span><span class="p">,</span> <span class="n">partition</span><span class="p">):</span>
    <span class="c1"># from itertools import accumulate</span>
    <span class="c1"># partition = [len(x) for x in partition]</span>
    <span class="c1"># Compute the cumulative sums once</span>
    <span class="c1"># cumsum_vector = list(accumulate(partition))</span>
    <span class="c1"># Use slicing based on the cumulative sums</span>
    <span class="c1"># return [lst[(cumsum_vector[i-1] if i &gt; 0 else 0):cumsum_vector[i]] for i in range(len(partition))]</span>
    <span class="k">return</span> <span class="n">partition</span>

<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>

<span class="k">def</span> <span class="nf">adapt_file_path</span><span class="p">(</span><span class="n">file_path</span><span class="p">,</span> <span class="n">reference_file</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Adapt a single file path based on its existence and a reference file&#39;s path.</span>

<span class="sd">    Args:</span>
<span class="sd">    - file_path (str): The file path to adapt.</span>
<span class="sd">    - reference_file (str): File path to use as a reference for adaptation.</span>

<span class="sd">    Returns:</span>
<span class="sd">    - str: Adapted file path.</span>

<span class="sd">    Raises:</span>
<span class="sd">    - FileNotFoundError: If no valid file path is found.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">reference_path</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">reference_file</span><span class="p">)</span>

    <span class="c1"># Check if the file exists</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_path</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">file_path</span>

    <span class="c1"># Check file name without path</span>
    <span class="n">file_name</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">basename</span><span class="p">(</span><span class="n">file_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_name</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">file_name</span>

    <span class="c1"># Check file name in reference file&#39;s directory</span>
    <span class="n">file_in_ref_dir</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reference_path</span><span class="p">,</span> <span class="n">file_name</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_in_ref_dir</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">file_in_ref_dir</span>

    <span class="c1"># Check original file path prefixed with reference file&#39;s directory</span>
    <span class="n">file_prefixed</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">reference_path</span><span class="p">,</span> <span class="n">file_path</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">file_prefixed</span><span class="p">):</span>
        <span class="k">return</span> <span class="n">file_prefixed</span>

    <span class="c1"># If all checks fail, raise an error</span>
    <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No valid path found for file: </span><span class="si">{</span><span class="n">file_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">adapt_file_path_all</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">column_name</span><span class="p">,</span> <span class="n">reference_file</span><span class="p">):</span>
    <span class="k">return</span> <span class="n">df</span><span class="p">[</span><span class="n">column_name</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="n">adapt_file_path</span><span class="p">(</span><span class="n">x</span><span class="p">,</span> <span class="n">reference_file</span><span class="p">))</span>
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span><span class="p">[</span><span class="n">get_analysis_regions</span><span class="p">:</span> <span class="kp">shared</span> <span class="o">=</span> <span class="s2">&quot;regional_data&quot;</span><span class="p">]</span>
<span class="c1"># input is genoFile, phenoFile, covFile and optionally region_list. If region_list presents then we only analyze what&#39;s contained in the list.</span>
<span class="c1"># regional_data should be a dictionary like:</span>
<span class="c1">#{&#39;data&#39;: [(&quot;genotype_1.bed&quot;, &quot;phenotype_1.bed.gz&quot;, &quot;covariate_1.gz&quot;), (&quot;genotype_2.bed&quot;, &quot;phenotype_1.bed.gz&quot;, &quot;phenotype_2.bed.gz&quot;, &quot;covariate_1.gz&quot;, &quot;covariate_2.gz&quot;) ... ],</span>
<span class="c1"># &#39;meta_info&#39;: [(&quot;chr12:752578-752579&quot;, &quot;gene_1&quot;, &quot;trait_1&quot;), (&quot;chr13:852580-852581&quot;, &quot;gene_2&quot;, &quot;trait_1&quot;, &quot;trait_2&quot;) ... ]}</span>

<span class="k">def</span> <span class="nf">process_pheno_files</span><span class="p">(</span><span class="n">pheno_files</span><span class="p">,</span> <span class="n">cov_files</span><span class="p">,</span> <span class="n">phenotype_names</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Example output:</span>
<span class="sd">    #chr    start      end    start_cis       end_cis           ID  path     cov_path             cond             coordinate     geno_path</span>
<span class="sd">    0  chr12   752578   752579  652578   852579  ENSG00000060237_Q9H4A3  protocol_example.protein_1.bed.gz,protocol_example.protein_2.bed.gz  covar_1.gz,covar_2.gz  trait_A,trait_B    chr12:752578-752579  protocol_example.genotype.chr21_22.bed</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="c1"># Initialize an empty DataFrame for accumulation</span>
    <span class="n">accumulated_pheno_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">()</span>

    <span class="n">merge_keys</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="s1">&#39;start&#39;</span><span class="p">,</span> <span class="s1">&#39;end&#39;</span><span class="p">,</span> <span class="s1">&#39;ID&#39;</span><span class="p">]</span>

    <span class="k">for</span> <span class="n">pheno_path</span><span class="p">,</span> <span class="n">cov_path</span><span class="p">,</span> <span class="n">phenotype_name</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">pheno_files</span><span class="p">,</span> <span class="n">cov_files</span><span class="p">,</span> <span class="n">phenotype_names</span><span class="p">):</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">cov_path</span><span class="p">):</span>
            <span class="k">raise</span> <span class="ne">FileNotFoundError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;No valid path found for file: </span><span class="si">{</span><span class="n">cov_path</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

        <span class="c1"># Read and process each phenotype file</span>
        <span class="n">pheno_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">pheno_path</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
        
        <span class="c1"># Adapt pheno file paths and add additional information</span>
        <span class="n">pheno_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">4</span><span class="p">]</span> <span class="o">=</span> <span class="n">adapt_file_path_all</span><span class="p">(</span><span class="n">pheno_df</span><span class="p">,</span> <span class="n">pheno_df</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">pheno_path</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
        <span class="n">pheno_df</span> <span class="o">=</span> <span class="n">pheno_df</span><span class="o">.</span><span class="n">assign</span><span class="p">(</span>
            <span class="n">cov_path</span><span class="o">=</span><span class="nb">str</span><span class="p">(</span><span class="n">cov_path</span><span class="p">),</span> 
            <span class="n">cond</span><span class="o">=</span><span class="n">phenotype_name</span><span class="p">)</span>

        <span class="c1"># Merge with the accumulated DataFrame</span>
        <span class="k">if</span> <span class="n">accumulated_pheno_df</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
            <span class="n">accumulated_pheno_df</span> <span class="o">=</span> <span class="n">pheno_df</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="c1"># Merge on specified keys with default suffixes</span>
            <span class="n">merged_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">accumulated_pheno_df</span><span class="p">,</span> <span class="n">pheno_df</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="n">merge_keys</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;outer&#39;</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;_x&#39;</span><span class="p">,</span> <span class="s1">&#39;_y&#39;</span><span class="p">))</span>

            <span class="c1"># Determine non-key columns</span>
            <span class="n">non_key_columns</span> <span class="o">=</span> <span class="p">[</span><span class="n">col</span> <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">pheno_df</span><span class="o">.</span><span class="n">columns</span> <span class="k">if</span> <span class="n">col</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">merge_keys</span><span class="p">]</span>

            <span class="c1"># Concatenate non-key columns for matching keys</span>
            <span class="k">for</span> <span class="n">col</span> <span class="ow">in</span> <span class="n">non_key_columns</span><span class="p">:</span>
                <span class="n">col_x</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s1">_x&#39;</span>
                <span class="n">col_y</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">col</span><span class="si">}</span><span class="s1">_y&#39;</span>

                <span class="c1"># Handling concatenation for matching keys</span>
                <span class="n">merged_df</span><span class="p">[</span><span class="n">col</span><span class="p">]</span> <span class="o">=</span> <span class="n">merged_df</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span>
                    <span class="k">lambda</span> <span class="n">row</span><span class="p">:</span> <span class="n">row</span><span class="p">[</span><span class="n">col_x</span><span class="p">]</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">col_y</span><span class="p">])</span> <span class="k">else</span> 
                                <span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">col_y</span><span class="p">]</span> <span class="k">if</span> <span class="n">pd</span><span class="o">.</span><span class="n">isna</span><span class="p">(</span><span class="n">row</span><span class="p">[</span><span class="n">col_x</span><span class="p">])</span> <span class="k">else</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="n">col_x</span><span class="p">]</span><span class="si">}</span><span class="s1">,</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="n">col_y</span><span class="p">]</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

                <span class="c1"># Drop the temporary columns</span>
                <span class="n">merged_df</span><span class="o">.</span><span class="n">drop</span><span class="p">([</span><span class="n">col_x</span><span class="p">,</span> <span class="n">col_y</span><span class="p">],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>

            <span class="n">accumulated_pheno_df</span> <span class="o">=</span> <span class="n">merged_df</span>
    <span class="k">return</span> <span class="n">accumulated_pheno_df</span>

<span class="c1"># Load phenotype meta data</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">covFile</span><span class="p">):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of input phenotypes files must match that of covariates files&quot;</span><span class="p">)</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">)</span> <span class="o">!=</span> <span class="nb">len</span><span class="p">(</span><span class="n">phenotype_names</span><span class="p">):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Number of input phenotypes files must match the number of phenotype names&quot;</span><span class="p">)</span>
<span class="n">meta_data</span> <span class="o">=</span> <span class="n">process_pheno_files</span><span class="p">(</span><span class="n">phenoFile</span><span class="p">,</span> <span class="n">covFile</span><span class="p">,</span> <span class="n">phenotype_names</span><span class="p">)</span>

<span class="c1"># Load genotype meta data</span>
<span class="k">if</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">:</span><span class="s2">x</span><span class="si">}</span><span class="s2">&quot;</span> <span class="o">==</span> <span class="s2">&quot;.bed&quot;</span><span class="p">:</span>
    <span class="n">geno_meta_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">([(</span><span class="s2">&quot;chr&quot;</span><span class="o">+</span><span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">),</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span><span class="mi">23</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[(</span><span class="s2">&quot;chrX&quot;</span><span class="p">,</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)],</span> <span class="n">columns</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="s1">&#39;geno_path&#39;</span><span class="p">])</span>
<span class="kn">else:</span>
    <span class="n">geno_meta_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">geno_meta_data</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">adapt_file_path_all</span><span class="p">(</span><span class="n">geno_meta_data</span><span class="p">,</span> <span class="n">geno_meta_data</span><span class="o">.</span><span class="n">columns</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span> <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">genoFile</span><span class="si">:</span><span class="s2">a</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">geno_meta_data</span><span class="o">.</span><span class="n">columns</span> <span class="o">=</span> <span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="s1">&#39;geno_path&#39;</span><span class="p">]</span>
    <span class="n">geno_meta_data</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">geno_meta_data</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span> <span class="k">if</span> <span class="nb">str</span><span class="p">(</span><span class="n">x</span><span class="p">)</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;chr&#39;</span><span class="p">)</span> <span class="k">else</span> <span class="sa">f</span><span class="s1">&#39;chr</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

<span class="c1"># Checking the DataFrame</span>
<span class="n">valid_chr_values</span> <span class="o">=</span> <span class="p">[</span><span class="sa">f</span><span class="s1">&#39;chr</span><span class="si">{</span><span class="n">x</span><span class="si">}</span><span class="s1">&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">23</span><span class="p">)]</span> <span class="o">+</span> <span class="p">[</span><span class="s1">&#39;chrX&#39;</span><span class="p">]</span>
<span class="k">if</span> <span class="ow">not</span> <span class="nb">all</span><span class="p">(</span><span class="n">value</span> <span class="ow">in</span> <span class="n">valid_chr_values</span> <span class="k">for</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">geno_meta_data</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]):</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Invalid chromosome values found. Allowed values are chr1 to chr22 and chrX.&quot;</span><span class="p">)</span>

<span class="n">meta_data</span> <span class="o">=</span> <span class="n">meta_data</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">geno_meta_data</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;inner&#39;</span><span class="p">)</span>

<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">meta_data</span><span class="o">.</span><span class="n">index</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
    <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;No region overlap between genotype and any of the phenotypes&quot;</span><span class="p">)</span>

<span class="n">region_ids</span> <span class="o">=</span> <span class="p">[]</span>
<span class="c1"># If region_list is provided, read the file and extract IDs</span>
<span class="k">if</span> <span class="n">region_list</span><span class="o">.</span><span class="n">is_file</span><span class="p">():</span>
    <span class="n">region_list_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">region_list</span><span class="p">,</span> <span class="n">sep</span> <span class="o">=</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">comment</span> <span class="o">=</span> <span class="s2">&quot;#&quot;</span><span class="p">)</span>
    <span class="n">region_ids</span> <span class="o">=</span> <span class="n">region_list_df</span><span class="o">.</span><span class="n">iloc</span><span class="p">[:,</span> <span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span>  <span class="c1"># Extracting the last column for IDs</span>

<span class="c1"># If region_name is provided, include those IDs as well</span>
<span class="c1"># --region-name A B C will result in a list of [&quot;A&quot;, &quot;B&quot;, &quot;C&quot;] here</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">region_name</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">region_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)</span><span class="o">.</span><span class="n">union</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">region_name</span><span class="p">)))</span>

<span class="c1"># If either region_list or region_name is provided, filter the meta_data</span>
<span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
    <span class="n">meta_data</span> <span class="o">=</span> <span class="n">meta_data</span><span class="p">[</span><span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isin</span><span class="p">(</span><span class="n">region_ids</span><span class="p">)]</span>

<span class="c1"># Adjust cis-window</span>
<span class="k">if</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">isfile</span><span class="p">(</span><span class="n">customized_cis_windows</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Loading customized cis-window data from </span><span class="si">{</span><span class="n">customized_cis_windows</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">cis_list</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">customized_cis_windows</span><span class="p">,</span> <span class="n">comment</span><span class="o">=</span><span class="s2">&quot;#&quot;</span><span class="p">,</span> <span class="n">header</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">names</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;#chr&quot;</span><span class="p">,</span><span class="s2">&quot;start&quot;</span><span class="p">,</span><span class="s2">&quot;end&quot;</span><span class="p">,</span><span class="s2">&quot;ID&quot;</span><span class="p">],</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span><span class="p">)</span>
    <span class="n">meta_data</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">merge</span><span class="p">(</span><span class="n">meta_data</span><span class="p">,</span> <span class="n">cis_list</span><span class="p">,</span> <span class="n">on</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">,</span> <span class="s1">&#39;ID&#39;</span><span class="p">],</span> <span class="n">how</span><span class="o">=</span><span class="s1">&#39;left&#39;</span><span class="p">,</span> <span class="n">suffixes</span><span class="o">=</span><span class="p">(</span><span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="s1">&#39;_cis&#39;</span><span class="p">))</span> 
    <span class="n">mismatches</span> <span class="o">=</span> <span class="n">meta_data</span><span class="p">[</span><span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;start_cis&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">isna</span><span class="p">()]</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">mismatches</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
        <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;First 5 mismatches:&quot;</span><span class="p">)</span>
        <span class="nb">print</span><span class="p">(</span><span class="n">mismatches</span><span class="p">[[</span><span class="s1">&#39;ID&#39;</span><span class="p">]]</span><span class="o">.</span><span class="n">head</span><span class="p">())</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="nb">len</span><span class="p">(</span><span class="n">mismatches</span><span class="p">)</span><span class="si">}</span><span class="s2"> regions to analyze cannot be found in ``</span><span class="si">{</span><span class="n">customized_cis_windows</span><span class="si">}</span><span class="s2">``. Please check your ``</span><span class="si">{</span><span class="n">customized_cis_windows</span><span class="si">}</span><span class="s2">`` database to make sure it contains all cis-window definitions. &quot;</span><span class="p">)</span>
<span class="kn">else:</span>
    <span class="k">if</span> <span class="n">window</span> <span class="o">&lt;=</span><span class="mi">0</span> <span class="p">:</span>
        <span class="k">raise</span> <span class="ne">ValueError</span><span class="p">(</span><span class="s2">&quot;Please either input valid path to cis-window file via ``--customized-cis-windows``, or set ``--window`` to a positive integer&quot;</span><span class="p">)</span>
    <span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;start_cis&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">apply</span><span class="p">(</span><span class="k">lambda</span> <span class="n">x</span><span class="p">:</span> <span class="nb">max</span><span class="p">(</span><span class="n">x</span> <span class="o">-</span> <span class="n">window</span><span class="p">,</span> <span class="mi">0</span><span class="p">))</span>
    <span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;end_cis&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">meta_data</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span> <span class="o">+</span> <span class="n">window</span>
<span class="c1"># Create the final dictionary</span>
<span class="n">regional_data</span> <span class="o">=</span> <span class="p">{</span>
    <span class="s1">&#39;data&#39;</span><span class="p">:</span> <span class="p">[(</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;geno_path&#39;</span><span class="p">],</span> <span class="o">*</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;path&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">),</span> <span class="o">*</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;cov_path&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">meta_data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()],</span>
    <span class="s1">&#39;meta_info&#39;</span><span class="p">:</span> <span class="p">[(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;end&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="c1"># this is the phenotype region</span>
                   <span class="sa">f</span><span class="s2">&quot;</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;#chr&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">:</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;start_cis&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">-</span><span class="si">{</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;end_cis&#39;</span><span class="p">]</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">,</span> <span class="c1"># this is the cis-window region</span>
                   <span class="n">row</span><span class="p">[</span><span class="s1">&#39;ID&#39;</span><span class="p">],</span> <span class="o">*</span><span class="n">row</span><span class="p">[</span><span class="s1">&#39;cond&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;,&#39;</span><span class="p">))</span> <span class="k">for</span> <span class="n">_</span><span class="p">,</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">meta_data</span><span class="o">.</span><span class="n">iterrows</span><span class="p">()]</span>
<span class="p">}</span>
</pre></div>
</div>
</div>
</div>
</section>
<section id="univariate-analysis-susie-and-twas">
<h2>Univariate analysis: SuSiE and TWAS<a class="headerlink" href="#univariate-analysis-susie-and-twas" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[susie_twas_1]
# initial number of single effects for SuSiE
parameter: init_L = 8
# maximum number of single effects to use for SuSiE
parameter: max_L = 30
# remove a variant if it has more than imiss missing individual level data
parameter: imiss = 1.0
# MAF cutoff
parameter: maf = 0.005
# MAC cutoff, on top of MAF cutoff
parameter: mac = 5
# Remove indels if indel = False
parameter: indel = True
parameter: pip_cutoff = 0.1
parameter: coverage = [0.95, 0.7, 0.5]
# Compute TWAS weights as well
parameter: twas_weights = True
# Perform K folds valiation CV for TWAS
# Set it to zero if this is to be skipped
parameter: twas_cv_folds = 5
parameter: ld_reference_meta_file = path()
depends: sos_variable(&quot;regional_data&quot;)

meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name[:-2]}/{name}.{_meta_info[2]}.susie{&quot;_weights_db&quot; if twas_weights else &quot;&quot;}.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    library(pecotmr)
    # extract subset of samples
    keep_samples = NULL
    if (${&quot;TRUE&quot; if keep_samples.is_file() else &quot;FALSE&quot;}) {
      keep_samples = unlist(strsplit(readLines(${keep_samples:ar}), &quot;\\s+&quot;))
      message(paste(length(keep_samples), &quot;samples are selected to be loaded for analysis&quot;))
    }
    # Load regional association data
    tryCatch({
    fdat = load_regional_univariate_data(genotype = ${_input[0]:anr},
                                          phenotype = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1:len(_input)//2+1]])}),
                                          covariate = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[len(_input)//2+1:]])}),
                                          region = &quot;${_meta_info[0]}&quot;,
                                          cis_window = &quot;${_meta_info[1]}&quot;,
                                          conditions = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in _meta_info[3:]])}),
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss},
                                          keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;},
                                          keep_samples = keep_samples,
                                          scale_residuals = FALSE)
    }, NoSNPsError = function(e) {
        message(&quot;Error: &quot;, e$message)
        #saveRDS(NULL, ${_output:ar})
        saveRDS(list(&quot;${_meta_info[2] + &#39;@&#39; + _meta_info[1]}&quot; = e$message), ${_output:ar}, compress=&#39;xz&#39;)
        quit(save=&quot;no&quot;)
    })
    # Univeriate analysis suite
    fitted = list()
    for (r in 1:length(fdat$residual_Y)) {
      st = proc.time()
      fitted[[r]] = susie_wrapper(fdat$residual_X[[r]], fdat$residual_Y[[r]], ${init_L}, ${max_L}, ${coverage[0]})
      fitted[[r]] = susie_post_processor(fitted[[r]], fdat$residual_X[[r]], fdat$residual_Y[[r]], fdat$residual_X_scalar[[r]], fdat$residual_Y_scalar[[r]], 
                                       fdat$maf[[r]], secondary_coverage = c(${&quot;,&quot;.join([str(x) for x in coverage[1:]])}), signal_cutoff = ${pip_cutoff}, 
                                       other_quantities = list(dropped_samples = fdat$dropped_sample[[r]]))
      if ( ${&quot;TRUE&quot; if twas_weights else &quot;FALSE&quot;} ) {
          weight_methods = list(enet_weights = list(), lasso_weights = list(), mrash_weights = list(init_prior_sd = TRUE), susie_weights = list())
          if ( ${&quot;TRUE&quot; if ld_reference_meta_file.is_file() else &quot;FALSE&quot; } ) {
              X = filter_genotype_by_ld_reference(fdat$residual_X[[r]], ld_reference_meta_file)
          } else {
              X = fdat$residual_X[[r]]
          }
          fitted[[r]]$twas_weights = twas_weights(X, fdat$residual_Y[[r]], weight_methods = weight_methods)
          fitted[[r]]$twas_predictions = twas_predict(X, fitted[[r]]$twas_weights) # get TWAS predictions for possible next steps such as computing correlations between predicted expression values
          if ( ${&quot;TRUE&quot; if twas_cv_folds &gt; 0 else &quot;FALSE&quot;} ) {
              fitted[[r]]$twas_cv_result = twas_weights_cv(X, fdat$residual_Y[[r]], fold = ${twas_cv_folds}, weight_methods = weight_methods, seed = ${seed})
          }
      }
    }
    fitted[[r]]$total_time_elapsed = proc.time() - st
    names(fitted) &lt;- names(fdat$residual_Y)
    saveRDS(list(&quot;${_meta_info[2] + &#39;@&#39; + _meta_info[1]}&quot; = fitted), ${_output:ar}, compress=&#39;xz&#39;)
</pre></div>
</div>
</div>
</div>
</section>
<section id="multivariate-analysis-mvsusie-and-mr-mash">
<h2>Multivariate analysis: mvSuSiE and mr.mash<a class="headerlink" href="#multivariate-analysis-mvsusie-and-mr-mash" title="Permalink to this heading">#</a></h2>
<section id="mvsusie">
<h3>mvSuSiE<a class="headerlink" href="#mvsusie" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[mvsusie_1]
# Prior model file generated from mashr. 
# Default will be used if it does not exist.
parameter: mixture_prior = path()
parameter: max_L = 20
# remove a variant if it has more than imiss missing individual level data
parameter: imiss = 0.1
# MAF cutoff
parameter: maf = 0.0
# MAC cutoff, on top of MAF cutoff
# Here I set default to mac = 10 rather than using an MAF cutoff
# I don&#39;t set it to 5 because I&#39;m not so sure of performance of SuSiE on somewhat infrequent variants
# MAC = 10 would not be too infrequenty for xQTL data where sample size is about ~1,000 at most (as of 2022)
parameter: mac = 10
# Remove indels if indel = False
parameter: indel = True
depends: sos_variable(&quot;regional_data&quot;)

meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name[:-2]}/{name}.{_meta_info[0]}.susie_fitted.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
   
    get_prior_indices &lt;- function(Y, U) {
      # make sure the prior col/rows match the colnames of the Y matrix
      y_names = colnames(Y)
      u_names = colnames(U)
      if (is.null(y_names) || is.null(u_names)) {
          return(NULL)
      } else if (identical(y_names, u_names)) {
          return(NULL)
      } else {
          return(match(y_names, u_names))
      }
    }

    # Load regional association data
    fdat = load_regional_multivariate_data(genotype = ${_input[0]:anr},
                                          phenotype = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1::2]])}),
                                          covariate = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[2::2]])}),
                                          region = ${&#39;&quot;%s:%s-%s&quot;&#39; % (_meta_info[1], _meta_info[2], _meta_info[3])},
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss},
                                          keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;})

    # univariate summary statistics
    non_missing = lapply(1:ncol(fdat$residual_Y), function(r)) which(!is.na(fdat$residual_Y[,r]))
    univariate_res = lapply(1:ncol(fdat$residual_Y), function(r) susieR:::univariate_regression(X[non_missing[[r]], ], fdat$residual_Y[non_missing[[r]], r]))
    sumstat = list(bhat=do.call(cbind, lapply(1:ncol(fdat$residual_Y), function(r) univariate_res[[r]]$betahat)),
                   sbhat=do.call(cbind, lapply(1:ncol(fdat$residual_Y), function(r) univariate_res[[r]]$sebetahat)))
  
    # Multivariate fine-mapping
    # FIXME: handle it when prior does not exist
    prior = readRDS(${mixture_prior:r})
    print(paste(&quot;Number of components in the mixture prior:&quot;, length(prior$U)))
    prior = mvsusieR::create_mash_prior(mixture_prior=list(weights=prior$w, matrices=prior$U), include_indices = get_prior_indices(fdat$residual_Y, prior$U[[1]]), max_mixture_len=-1)   
    resid_Y = compute_cov_flash(fdat$residual_Y)
    st = proc.time()
    fitted = mvsusieR::mvsusie(fdat$X, 
                               fdat$residual_Y, 
                               L=${max_L}, 
                               prior_variance=prior, 
                               residual_variance=resid_Y, 
                               precompute_covariances=F, 
                               compute_objective=T, 
                               estimate_residual_variance=F, 
                               estimate_prior_variance=T, 
                               estimate_prior_method=&#39;EM&#39;,
                               max_iter = 200, 
                               n_thread=${numThreads}, 
                               approximate=F)
    fitted$analysis_time = proc.time() - st
    fitted$cs_corr = susieR::get_cs_correlation(fitted, X=fdat$X)
    fitted$cs_snps = names(fitted$X_column_scale_factors[unlist(fitted$sets$cs)])
    fitted$variable_name = names(fitted$pip)
    fitted$analysis_script = load_script()
    fitted$dropped_samples = fdat$dropped_sample
    fitted$sample_names = colnames(fdat$residual_Y)
    fitted$residual_y = resid_Y
    saveRDS(fitted, ${_output:ar})
</pre></div>
</div>
</div>
</div>
</section>
<section id="mr-mash">
<h3>mr.mash<a class="headerlink" href="#mr-mash" title="Permalink to this heading">#</a></h3>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[mrmash_1]
# Prior model file generated from mashr. 
# Default will be used if it does not exist.
parameter: mixture_prior = path()
# remove a variant if it has more than imiss missing individual level data
parameter: imiss = 0.1
# MAF cutoff
parameter: maf = 0.05
# MAC cutoff, on top of MAF cutoff
parameter: mac = 0
# Remove indels if indel = False
parameter: indel = True
# Path to prior grid data file: an RDS file with scaling factors
parameter: prior_grid = path(&#39;.&#39;)
# Path to prior weights data file: an RDS file with prior weights
parameter: prior_weights = path(&#39;.&#39;)
# Path to summary statistics directory
parameter: sumstats_file = path(&#39;.&#39;)
# Path to sample partition for cross validation
parameter: sample_partition = path(&#39;.&#39;)
parameter: fold = 1
parameter: var_cutoff = 0.05
parameter: n_nonmiss_Y = 100
parameter: canonical_mats = False
parameter: standardize = True
parameter: update_w0 = True
parameter: w0_threshold = 0.0
parameter: update_V = True
parameter: update_V_method = &quot;full&quot;
parameter: B_init_method = &quot;enet&quot;
parameter: max_iter = 5000
parameter: tol = 1e-2
parameter: verbose = False
parameter: save_model = False
parameter: glmnet_pred = False
depends: sos_variable(&quot;regional_data&quot;)

meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name[:-2]}_fold_{fold}/{name}.{_meta_info[0]}.mrmash.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    library(pecotmr)
    # Load regional association data
    fdat = load_regional_multivariate_data(genotype = ${_input[0]:anr},
                                          phenotype = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1::2]])}),
                                          covariate = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[2::2]])}),
                                          region = &quot;${_meta_info[0]}&quot;,
                                          conditions = c(&#39;${_meta_info[2]}&#39;),
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          xvar_cutoff = ${var_cutoff},
                                          imiss_cutoff = ${imiss},
                                          matrix_y_min_complete = ${n_nonmiss_Y},
                                          keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;})
    res = mrmash_wrapper(fdat$X, fdat$residual_Y, # here, X and Y are both matrices. X does not have missing data, Y may have it. Y is residualized (covariates removed) X is not.
                            ${sumstats_file:ar}, 
                            ${prior_matrices:ar}),
                            ${prior_weights:ar},
                            ${prior_grid:ar},
                            ${sample_partition:r}, # if available, will perform CV; otherwise will not perform CV
                            fold = ${fold},
                            nthreads = ${nthreads},
                            canonical_mats = ${&quot;T&quot; if canonical_mats else &quot;F&quot;},
                            standardize = ${&quot;T&quot; if standardize else &quot;F&quot;},
                            update_w0 = ${&quot;T&quot; if update_w0 else &quot;F&quot;},
                            w0_threshold = ${w0_threshold},
                            update_V = ${&quot;T&quot; if update_V else &quot;F&quot;},
                            update_V_method = &quot;${update_V_method}&quot;,
                            B_init_method = &quot;${B_init_method}&quot;,
                            max_iter = ${max_iter},
                            tol = ${tol},
                            verbose = ${&quot;T&quot; if verbose else &quot;F&quot;},
                            save_model = ${&quot;T&quot; if save_model else &quot;F&quot;},
                            glmnet_pred = ${&quot;T&quot; if glmnet_pred else &quot;F&quot;}
                          )
    st = proc.time()
    res$analysis_time = proc.time() - st
    res$analysis_script = load_script()
    res$dropped_samples = fdat$dropped_sample
    res$sample_names = colnames(fdat$residual_Y)
    saveRDS(res, ${_output:ar}, compress=&#39;xz&#39;)
</pre></div>
</div>
</div>
</div>
</section>
</section>
<section id="functional-regression-fsusie-for-trait">
<h2>Functional regression fSuSiE for trait<a class="headerlink" href="#functional-regression-fsusie-for-trait" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[fsusie_1]
parameter: max_L = 20
# remove a variant if it has more than imiss missing individual level data
parameter: imiss = 0.1
# MAF cutoff
parameter: maf = 0.0
# MAC cutoff, on top of MAF cutoff
# Here I set default to mac = 10 rather than using an MAF cutoff
# I don&#39;t set it to 5 because I&#39;m not so sure of performance of SuSiE on somewhat infrequent variants
# MAC = 10 would not be too infrequenty for xQTL data where sample size is about ~1,000 at most (as of 2022)
parameter: mac = 10
# prior can be either of [&quot;mixture_normal&quot;, &quot;mixture_normal_per_scale&quot;]
parameter: prior  = &quot;mixture_normal_per_scale&quot;
parameter: max_SNP_EM = 100
# Remove indels if indel = False
parameter: indel = True
depends: sos_variable(&quot;regional_data&quot;)

meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name[:-2]}/{name}.{_meta_info[0]}.fsusie_{prior}.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    library(pecotmr)
    library(dplyr)

    # Load regional functional data
    fdat = load_regional_functional_data(genotype = ${_input[0]:anr},
                                          phenotype = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1:len(_input)//2+1]])}),
                                          covariate = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[len(_input)//2+1:]])}),
                                          region = &quot;${_meta_info[0]}&quot;,
                                          conditions = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x for x in _meta_info[3:]])}),
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss},
                                          y_as_matrix = FALSE,
                                          keep_indel = ${&quot;TRUE&quot; if indel else &quot;FALSE&quot;})
    
    # Fine-mapping with fSuSiE
    library(&quot;susiF.alpha&quot;)
    fitted = list()
    for (r in 1:length(fdat$residual_Y)){
        st = proc.time()  
        fitted[[r]] &lt;- susiF_wrapper(X = fdat$residual_X[[r]],
                             Y = fdat$residual_Y[[r]],
                             pos=fdat$Y_coordinates[[r]]$start, #FIXME: this needs to be edited and added to load_regional_association_data
                             L=${max_L},
                             prior=&quot;${prior}&quot;,
                             max_SNP_EM=${max_SNP_EM}, secondary_coverage = c(${&quot;,&quot;.join([str(x) for x in coverage[1:]])}), signal_cutoff = ${pip_cutoff} )
    saveRDS(fitted, ${_output:ar}, compress=&#39;xz&#39;)
    
</pre></div>
</div>
</div>
</div>
</section>
<section id="functional-regression-fsusie-with-other-modality">
<h2>Functional regression fSuSiE with other modality<a class="headerlink" href="#functional-regression-fsusie-with-other-modality" title="Permalink to this heading">#</a></h2>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[mvfsusie_1]
parameter: max_L = 30
# remove a variant if it has more than imiss missing individual level data
parameter: imiss = 0.1
# MAF cutoff
parameter: maf = 0.0
# MAC cutoff, on top of MAF cutoff
# Here I set default to mac = 10 rather than using an MAF cutoff
# I don&#39;t set it to 5 because I&#39;m not so sure of performance of SuSiE on somewhat infrequent variants
# MAC = 10 would not be too infrequenty for xQTL data where sample size is about ~1,000 at most (as of 2022)
parameter: mac = 10
# prior can be either of [&quot;mixture_normal&quot;, &quot;mixture_normal_per_scale&quot;]
parameter: prior  = &quot;mixture_normal_per_scale&quot;
parameter: max_SNP_EM = 1000

depends: sos_variable(&quot;regional_data&quot;)

meta_info = regional_data[&#39;meta_info&#39;]
input: regional_data[&quot;data&quot;], group_by = lambda x: group_by_region(x, regional_data[&quot;data&quot;]), group_with = &quot;meta_info&quot;
output: f&#39;{cwd:a}/{step_name[:-2]}/{name}.{_meta_info[0]}.mvfsusie_{prior}.rds&#39;
task: trunk_workers = 1, trunk_size = job_size, walltime = walltime, mem = mem, cores = numThreads, tags = f&#39;{step_name}_{_output:bn}&#39;
R: expand = &#39;${ }&#39;, stdout = f&quot;{_output:n}.stdout&quot;, stderr = f&quot;{_output:n}.stderr&quot;, container = container, entrypoint = entrypoint
    # Load regional association data
    fdat = load_regional_association_data(genotype = ${_input[0]:anr},
                                          phenotype = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[1::2]])}),
                                          covariate = c(${&quot;,&quot;.join([&#39;&quot;%s&quot;&#39; % x.absolute() for x in _input[2::2]])}),
                                          region = ${&#39;&quot;%s:%s-%s&quot;&#39; % (_meta_info[1], _meta_info[2], _meta_info[3])},
                                          maf_cutoff = ${maf},
                                          mac_cutoff = ${mac},
                                          imiss_cutoff = ${imiss},
                                          y_as_matrix = FALSE)
    # Fine-mapping with mvfSuSiE
    library(&quot;mvf.susie.alpha&quot;)
    Y = map(fdat$residual_Y, ~left_join(fdat$X[,1]%&gt;%as.data.frame%&gt;%rownames_to_column(&quot;rowname&quot;), .x%&gt;%t%&gt;%as.data.frame%&gt;%rownames_to_column(&quot;rowname&quot;) , by = &quot;rowname&quot;)%&gt;%select(-2)%&gt;%column_to_rownames(&quot;rowname&quot;)%&gt;%as.matrix )
    fitted &lt;- multfsusie(Y_f = list(Y[[1]],Y[[3]]), 
                         Y_u = Reduce(cbind, Y[[2]]),
                         pos = list(pos1 =fdat$phenotype_coordiates[[1]], pos2 = fdat$phenotype_coordiates[[3]]),
                         X=X,
                         L=${max_L},
                         data.format=&quot;list_df&quot;)
    saveRDS(fitted, ${_output:ar})
</pre></div>
</div>
</div>
</div>
<div class="cell docutils container">
<div class="cell_input docutils container">
<div class="highlight-sos notranslate"><div class="highlight"><pre><span></span>[*_2]
input: group_by = &quot;all&quot;
output: f&#39;{cwd}/{name}.cis_workhorse_output.txt&#39;
python: expand= &quot;$[ ]&quot;, stderr = f&#39;{_output:n}.stderr&#39;, stdout = f&#39;{_output:n}.stdout&#39;, container = container, entrypoint = entrypoint
    import pandas as pd
    pd.DataFrame({&quot;output&quot; : [$[_input:ar,]]}).to_csv(&quot;$[_output]&quot;,index = False ,header = False, sep = &quot;\t&quot;)
</pre></div>
</div>
</div>
</div>
</section>
</section>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "sos"
        },
        kernelOptions: {
            name: "sos",
            path: "./code/cis_analysis"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'sos'</script>

                </article>
              

              
              
              
              
                <footer class="prev-next-footer">
                  
<div class="prev-next-area">
    <a class="left-prev"
       href="cis_advanced_overview.html"
       title="previous page">
      <i class="fa-solid fa-angle-left"></i>
      <div class="prev-next-info">
        <p class="prev-next-subtitle">previous</p>
        <p class="prev-next-title">Overview of advanced cis-QTL analysis</p>
      </div>
    </a>
    <a class="right-next"
       href="SuSiE.html"
       title="next page">
      <div class="prev-next-info">
        <p class="prev-next-subtitle">next</p>
        <p class="prev-next-title">SuSiE fine-mapping workflow</p>
      </div>
      <i class="fa-solid fa-angle-right"></i>
    </a>
</div>
                </footer>
              
            </div>
            
            
              
                <div class="bd-sidebar-secondary bd-toc"><div class="sidebar-secondary-items sidebar-secondary__inner">

  <div class="sidebar-secondary-item">
  <div class="page-toc tocsection onthispage">
    <i class="fa-solid fa-list"></i> Contents
  </div>
  <nav class="bd-toc-nav page-toc">
    <ul class="visible nav section-nav flex-column">
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#description">Description</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#input">Input</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-genotype-data">Example genotype data</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-phenotype-list">Example phenotype list</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#example-cis-window-file">Example cis-window file</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#about-indels">About indels</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#output">Output</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#minimal-working-example-steps">Minimal Working Example Steps</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#i-susie-with-twas-weights">i. SuSiE with TWAS weights</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#ii-fsusie">ii. fSuSiE</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#troubleshooting">Troubleshooting</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#command-interface">Command interface</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#workflow-implementation">Workflow implementation</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#univariate-analysis-susie-and-twas">Univariate analysis: SuSiE and TWAS</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#multivariate-analysis-mvsusie-and-mr-mash">Multivariate analysis: mvSuSiE and mr.mash</a><ul class="nav section-nav flex-column">
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mvsusie">mvSuSiE</a></li>
<li class="toc-h3 nav-item toc-entry"><a class="reference internal nav-link" href="#mr-mash">mr.mash</a></li>
</ul>
</li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-for-trait">Functional regression fSuSiE for trait</a></li>
<li class="toc-h2 nav-item toc-entry"><a class="reference internal nav-link" href="#functional-regression-fsusie-with-other-modality">Functional regression fSuSiE with other modality</a></li>
</ul>
  </nav></div>

</div></div>
              
            
          </div>
          <footer class="bd-footer-content">
            
<div class="bd-footer-content__inner container">
  
  <div class="footer-item">
    
<p class="component-author">
By The NIH/NIA Alzheimer's Disease Sequencing Project Functional Genomics xQTL Consortium
</p>

  </div>
  
  <div class="footer-item">
    

  <p class="copyright">
    
       Copyright 2021+, FunGen xQTL Analysis Working Group.
      <br/>
    
  </p>

  </div>
  
  <div class="footer-item">
    
  </div>
  
  <div class="footer-item">
    
  </div>
  
</div>
          </footer>
        

      </main>
    </div>
  </div>
  
  <!-- Scripts loaded after <body> so the DOM is not blocked -->
  <script src="../../_static/scripts/bootstrap.js?digest=5b4479735964841361fd"></script>
<script src="../../_static/scripts/pydata-sphinx-theme.js?digest=5b4479735964841361fd"></script>

  <footer class="bd-footer">
  </footer>
  </body>
</html>